
<!-- README.md is generated from README.Rmd. Please edit that file -->
<!-- This line is from RStudio -->

# Giotto

<!-- badges: start -->

[![License: GPL
v3](https://img.shields.io/badge/License-GPLv3-blue.svg)](https://www.gnu.org/licenses/gpl-3.0)
![Last Commit](https://badgen.net/github/last-commit/RubD/Giotto/suite)
<!-- badges: end -->

*Giotto Suite* is a comprehensive update to the *Giotto* package that
provides tools to process, analyze and visualize **single-cell and
subcellular spatial expression** data. The underlying framework is
generalizable to virtually all currently available spatial datasets. We
recently demonstrated the general applicability on 10 different datasets
created by 9 different state-of-the-art spatial technologies, including
*in situ* hybridization (seqFISH+, merFISH, osmFISH), sequencing
(Slide-seq, Visium, STARmap) and imaging-based multiplexing/proteomics
(CyCIF, MIBI, CODEX). These technologies differ in terms of resolution
(single cell vs multiple cells), spatial dimension (2D vs 3D), molecular
modality (protein vs RNA), and throughput (number of cells and genes).
More information and documentation about Giotto Suite can be found at
<http://giottosuite.com/index.html>.

The package is in heavy development. Please check back often!

<!-- <img src="inst/images/general_figs/overview_datasets.png" /> -->

## References

-   [Dries, R., Zhu, Q. et al. Giotto: a toolbox for integrative
    analysis and visualization of spatial expression data. Genome
    Biology
    (2021).](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-021-02286-2)
-   [Dries, R., Chen, J. et al. Advances in spatial transcriptomic data
    analysis. Genome Research
    (2021).](https://genome.cshlp.org/content/31/10/1706.long)
-   [Del Rossi, N., Chen, J. et al. Analyzing Spatial Transcriptomics
    Data Using Giotto. Current Protocols
    (2022).](https://currentprotocols.onlinelibrary.wiley.com/doi/abs/10.1002/cpz1.405)
