\documentclass[a4paper]{book}
\usepackage[times,inconsolata,hyper]{Rd}
\usepackage{makeidx}
\usepackage[utf8]{inputenc} % @SET ENCODING@
% \usepackage{graphicx} % @USE GRAPHICX@
\makeindex{}
\begin{document}
\chapter*{}
\begin{center}
{\textbf{\huge Package `Giotto'}}
\par\bigskip{\large \today}
\end{center}
\begin{description}
\raggedright{}
\inputencoding{utf8}
\item[Title]\AsIs{Spatial single-cell transcriptomics toolbox.}
\item[Version]\AsIs{0.3.2}
\item[Description]\AsIs{Toolbox to process, analyze and visualize spatial single-cell expression data.}
\item[License]\AsIs{MIT + file LICENSE}
\item[Encoding]\AsIs{UTF-8}
\item[LazyData]\AsIs{true}
\item[URL]\AsIs{}\url{https://rubd.github.io/Giotto/}\AsIs{, }\url{https://github.com/RubD/Giotto}\AsIs{}
\item[BugReports]\AsIs{}\url{https://github.com/RubD/Giotto/issues}\AsIs{}
\item[RoxygenNote]\AsIs{7.1.0}
\item[Depends]\AsIs{data.table (>= 1.12.2),
ggplot2 (>= 3.1.1),
base (>= 3.5.1),
utils (>= 3.5.1),
R (>= 3.5.1)}
\item[Imports]\AsIs{Matrix,
Rtsne (>= 0.15),
uwot (>= 0.0.0.9010),
FactoMineR (>= 1.34),
factoextra (>= 1.0.5),
cowplot (>= 0.9.4),
grDevices,
RColorBrewer (>= 1.1-2),
jackstraw (>= 1.3),
dbscan (>= 1.1-3),
ggalluvial (>= 0.9.1),
scales (>= 1.0.0),
ComplexHeatmap (>= 1.20.0),
qvalue (>= 2.14.1),
lfa (>= 1.12.0),
igraph (>= 1.2.4.1),
plotly,
reticulate (>= 1.14),
magrittr,
limma,
ggdendro,
smfishHmrf,
matrixStats (>= 0.55.0),
devtools,
reshape2,
ggraph,
Rcpp,
rlang (>= 0.4.3),
fitdistrplus}
\item[Suggests]\AsIs{knitr,
rmarkdown,
MAST,
scran (>= 1.10.1),
png,
tiff,
biomaRt,
trendsceek,
multinet (>= 3.0.2),
RTriangle (>= 1.6-0.10)}
\item[biocViews]\AsIs{}
\item[VignetteBuilder]\AsIs{knitr}
\item[LinkingTo]\AsIs{Rcpp,
RcppArmadillo}
\item[Remotes]\AsIs{lambdamoses/smfishhmrf-r}
\end{description}
\Rdcontents{\R{} topics documented:}
\inputencoding{utf8}
\HeaderA{adapt\_aspect\_ratio}{adapt\_aspect\_ratio}{adapt.Rul.aspect.Rul.ratio}
%
\begin{Description}\relax
adapt the aspact ratio after inserting cross section mesh grid lines
\end{Description}
%
\begin{Usage}
\begin{verbatim}
adapt_aspect_ratio(
  current_ratio,
  cell_locations,
  sdimx = NULL,
  sdimy = NULL,
  sdimz = NULL,
  mesh_obj = NULL
)
\end{verbatim}
\end{Usage}
\inputencoding{utf8}
\HeaderA{addCellIntMetadata}{addCellIntMetadata}{addCellIntMetadata}
%
\begin{Description}\relax
Creates an additional metadata column with information about interacting and non-interacting cell types of the
selected cell-cell interaction.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
addCellIntMetadata(
  gobject,
  spatial_network = "spatial_network",
  cluster_column,
  cell_interaction,
  name = "select_int",
  return_gobject = TRUE
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{spatial\_network}] name of spatial network to use

\item[\code{cluster\_column}] column of cell types

\item[\code{cell\_interaction}] cell-cell interaction to use

\item[\code{name}] name for the new metadata column

\item[\code{return\_gobject}] return an updated giotto object
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
This function will create an additional metadata column which selects interacting cell types for a specific cell-cell
interaction. For example, if you want to color interacting astrocytes and oligodendrocytes it will create a new metadata column with
the values "select\_astrocytes", "select\_oligodendrocytes", "other\_astrocytes", "other\_oligodendroyctes" and "other". Where "other" is all
other cell types found within the selected cell type column.
\end{Details}
%
\begin{Value}
Giotto object
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    addCellIntMetadata(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{addCellMetadata}{addCellMetadata}{addCellMetadata}
%
\begin{Description}\relax
adds cell metadata to the giotto object
\end{Description}
%
\begin{Usage}
\begin{verbatim}
addCellMetadata(
  gobject,
  new_metadata,
  by_column = FALSE,
  column_cell_ID = NULL
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{new\_metadata}] new cell metadata to use (data.table, data.frame, ...)

\item[\code{by\_column}] merge metadata based on cell\_ID column in pDataDT (default = FALSE)

\item[\code{column\_cell\_ID}] column name of new metadata to use if by\_column = TRUE
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
You can add additional cell metadata in two manners:
1. Provide a data.table or data.frame with cell annotations in the same order as the cell\_ID column in pDataDT(gobject)
2. Provide a data.table or data.frame with cell annotations and specificy which column contains the cell IDs,
these cell IDs need to match with the cell\_ID column in pDataDT(gobject)
\end{Details}
%
\begin{Value}
giotto object
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    addCellMetadata(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{addCellStatistics}{addCellStatistics}{addCellStatistics}
%
\begin{Description}\relax
adds cells statistics to the giotto object
\end{Description}
%
\begin{Usage}
\begin{verbatim}
addCellStatistics(
  gobject,
  expression_values = c("normalized", "scaled", "custom"),
  detection_threshold = 0,
  return_gobject = TRUE
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{expression\_values}] expression values to use

\item[\code{detection\_threshold}] detection threshold to consider a gene detected

\item[\code{return\_gobject}] boolean: return giotto object (default = TRUE)
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
This function will add the following statistics to cell metadata:
\begin{itemize}

\item{} nr\_genes: Denotes in how many genes are detected per cell
\item{} perc\_genes: Denotes what percentage of genes is detected per cell
\item{} total\_expr: Shows the total sum of gene expression per cell

\end{itemize}

\end{Details}
%
\begin{Value}
giotto object if return\_gobject = TRUE
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    addCellStatistics(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{addGeneMetadata}{addGeneMetadata}{addGeneMetadata}
%
\begin{Description}\relax
adds gene metadata to the giotto object
\end{Description}
%
\begin{Usage}
\begin{verbatim}
addGeneMetadata(gobject, new_metadata, by_column = F, column_gene_ID = NULL)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{new\_metadata}] new metadata to use

\item[\code{by\_column}] merge metadata based on gene\_ID column in fDataDT

\item[\code{column\_cell\_ID}] column name of new metadata to use if by\_column = TRUE
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
You can add additional gene metadata in two manners:
1. Provide a data.table or data.frame with gene annotations in the same order as the gene\_ID column in fDataDT(gobject)
2. Provide a data.table or data.frame with gene annotations and specificy which column contains the gene IDs,
these gene IDs need to match with the gene\_ID column in fDataDT(gobject)
\end{Details}
%
\begin{Value}
giotto object
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    addGeneMetadata(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{addGeneStatistics}{addGeneStatistics}{addGeneStatistics}
%
\begin{Description}\relax
adds gene statistics to the giotto object
\end{Description}
%
\begin{Usage}
\begin{verbatim}
addGeneStatistics(
  gobject,
  expression_values = c("normalized", "scaled", "custom"),
  detection_threshold = 0,
  return_gobject = TRUE
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{expression\_values}] expression values to use

\item[\code{detection\_threshold}] detection threshold to consider a gene detected

\item[\code{return\_gobject}] boolean: return giotto object (default = TRUE)
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
This function will add the following statistics to gene metadata:
\begin{itemize}

\item{} nr\_cells: Denotes in how many cells the gene is detected
\item{} per\_cells: Denotes in what percentage of cells the gene is detected
\item{} total\_expr: Shows the total sum of gene expression in all cells
\item{} mean\_expr: Average gene expression in all cells
\item{} mean\_expr\_det: Average gene expression in cells with detectable levels of the gene

\end{itemize}

\end{Details}
%
\begin{Value}
giotto object if return\_gobject = TRUE
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    addGeneStatistics(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{addHMRF}{addHMRF}{addHMRF}
%
\begin{Description}\relax
Add selected results from doHMRF to the giotto object
\end{Description}
%
\begin{Usage}
\begin{verbatim}
addHMRF(gobject, HMRFoutput, k = NULL, betas_to_add = NULL, hmrf_name = NULL)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{HMRFoutput}] HMRF output from doHMRF()

\item[\code{k}] number of domains

\item[\code{betas\_to\_add}] results from different betas that you want to add

\item[\code{name}] specify a custom name
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Description ...
\end{Details}
%
\begin{Value}
giotto object
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    addHMRF(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{addNetworkLayout}{addNetworkLayout}{addNetworkLayout}
%
\begin{Description}\relax
Add a network layout for a selected nearest neighbor network
\end{Description}
%
\begin{Usage}
\begin{verbatim}
addNetworkLayout(
  gobject,
  nn_network_to_use = "sNN",
  network_name = "sNN.pca",
  layout_type = c("drl"),
  options_list = NULL,
  layout_name = "layout",
  return_gobject = TRUE
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{nn\_network\_to\_use}] kNN or sNN

\item[\code{network\_name}] name of NN network to be used

\item[\code{layout\_type}] layout algorithm to use

\item[\code{options\_list}] list of options for selected layout

\item[\code{layout\_name}] name for layout

\item[\code{return\_gobject}] boolean: return giotto object (default = TRUE)
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
This function creates layout coordinates based on the provided kNN or sNN.
Currently only the force-directed graph layout "drl", see \code{\LinkA{layout\_with\_drl}{layout.Rul.with.Rul.drl}},
is implemented. This provides an alternative to tSNE or UMAP based visualizations.
\end{Details}
%
\begin{Value}
giotto object with updated layout for selected NN network
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    addNetworkLayout(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{addStatistics}{addStatistics}{addStatistics}
%
\begin{Description}\relax
adds genes and cells statistics to the giotto object
\end{Description}
%
\begin{Usage}
\begin{verbatim}
addStatistics(
  gobject,
  expression_values = c("normalized", "scaled", "custom"),
  detection_threshold = 0,
  return_gobject = TRUE
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{expression\_values}] expression values to use

\item[\code{detection\_threshold}] detection threshold to consider a gene detected

\item[\code{return\_gobject}] boolean: return giotto object (default = TRUE)
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
See \code{\LinkA{addGeneStatistics}{addGeneStatistics}} and \code{\LinkA{addCellStatistics}{addCellStatistics}}
\end{Details}
%
\begin{Value}
giotto object if return\_gobject = TRUE, else a list with results
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    addStatistics(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{adjustGiottoMatrix}{adjustGiottoMatrix}{adjustGiottoMatrix}
%
\begin{Description}\relax
normalize and/or scale expresion values of Giotto object
\end{Description}
%
\begin{Usage}
\begin{verbatim}
adjustGiottoMatrix(
  gobject,
  expression_values = c("normalized", "scaled", "custom"),
  batch_columns = NULL,
  covariate_columns = NULL,
  return_gobject = TRUE,
  update_slot = c("custom")
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{expression\_values}] expression values to use

\item[\code{batch\_columns}] metadata columns that represent different batch (max = 2)

\item[\code{covariate\_columns}] metadata columns that represent covariates to regress out

\item[\code{return\_gobject}] boolean: return giotto object (default = TRUE)

\item[\code{update\_slot}] expression slot that will be updated (default = custom)
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
This function implements the \code{\LinkA{ limma::removeBatchEffect}{ limma::removeBatchEffect}} function to
remove known batch effects and to adjust expression values according to provided covariates.
\end{Details}
%
\begin{Value}
giotto object
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    adjustGiottoMatrix(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{aes\_string2}{aes\_string2}{aes.Rul.string2}
%
\begin{Description}\relax
makes sure aes\_string can also be used with names that start with numeric values
\end{Description}
%
\begin{Usage}
\begin{verbatim}
aes_string2(...)
\end{verbatim}
\end{Usage}
\inputencoding{utf8}
\HeaderA{all\_plots\_save\_function}{all\_plots\_save\_function}{all.Rul.plots.Rul.save.Rul.function}
%
\begin{Description}\relax
Function to automatically save plots to directory of interest
\end{Description}
%
\begin{Usage}
\begin{verbatim}
all_plots_save_function(
  gobject,
  plot_object,
  save_dir = NULL,
  save_folder = NULL,
  save_name = NULL,
  default_save_name = "giotto_plot",
  save_format = NULL,
  show_saved_plot = F,
  ncol = 1,
  nrow = 1,
  scale = 1,
  base_width = NULL,
  base_height = NULL,
  base_aspect_ratio = NULL,
  units = NULL,
  dpi = NULL,
  limitsize = TRUE,
  ...
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{plot\_object}] object to plot

\item[\code{save\_dir}] directory to save to

\item[\code{save\_folder}] folder in save\_dir to save to

\item[\code{save\_name}] name of plot

\item[\code{save\_format}] format (e.g. png, tiff, pdf, ...)

\item[\code{show\_saved\_plot}] load \& display the saved plot

\item[\code{ncol}] number of columns

\item[\code{nrow}] number of rows

\item[\code{scale}] scale

\item[\code{base\_width}] width

\item[\code{base\_height}] height

\item[\code{base\_aspect\_ratio}] aspect ratio

\item[\code{units}] units

\item[\code{dpi}] Plot resolution

\item[\code{limitsize}] When TRUE (the default), ggsave will not save images larger than 50x50 inches, to prevent the common error of specifying dimensions in pixels.

\item[\code{...}] additional parameters to ggplot\_save\_function or general\_save\_function
\end{ldescription}
\end{Arguments}
%
\begin{SeeAlso}\relax
\code{\LinkA{general\_save\_function}{general.Rul.save.Rul.function}}
\end{SeeAlso}
%
\begin{Examples}
\begin{ExampleCode}
    all_plots_save_function(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{annotateGiotto}{annotateGiotto}{annotateGiotto}
%
\begin{Description}\relax
Converts cluster results into provided annotation.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
annotateGiotto(
  gobject,
  annotation_vector = NULL,
  cluster_column = NULL,
  name = "cell_types"
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{annotation\_vector}] named annotation vector (names = cluster ids)

\item[\code{cluster\_column}] cluster column to convert to annotation names

\item[\code{name}] new name for annotation column
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
You need to specifify which (cluster) column you want to annotate
and you need to provide an annotation vector like this:
\begin{itemize}

\item{} 1. identify the cell type of each cluster
\item{} 2. create a vector of these cell types, e.g. cell\_types =  c('T-cell', 'B-cell', 'Stromal')
\item{} 3. provide original cluster names to previous vector, e.g. names(cell\_types) = c(2, 1, 3)

\end{itemize}

\end{Details}
%
\begin{Value}
giotto object
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    annotateGiotto(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{annotateSpatialNetwork}{annotateSpatialNetwork}{annotateSpatialNetwork}
%
\begin{Description}\relax
Annotate spatial network with cell metadata information.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
annotateSpatialNetwork(
  gobject,
  spatial_network_name = "Delaunay_network",
  cluster_column,
  create_full_network = F
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{spatial\_network\_name}] name of spatial network to use

\item[\code{cluster\_column}] name of column to use for clusters

\item[\code{create\_full\_network}] convert from reduced to full network representation
\end{ldescription}
\end{Arguments}
%
\begin{Value}
annotated network in data.table format
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    annotateSpatialNetwork(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{annotate\_spatlocs\_with\_spatgrid\_2D}{annotate\_spatlocs\_with\_spatgrid\_2D}{annotate.Rul.spatlocs.Rul.with.Rul.spatgrid.Rul.2D}
%
\begin{Description}\relax
annotate spatial locations with 2D spatial grid information
\end{Description}
%
\begin{Usage}
\begin{verbatim}
annotate_spatlocs_with_spatgrid_2D(spatloc, spatgrid)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{spatloc}] spatial\_locs slot from giotto object

\item[\code{spatgrid}] selected spatial\_grid slot from giotto object
\end{ldescription}
\end{Arguments}
%
\begin{Value}
annotated spatial location data.table
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    annotate_spatlocs_with_spatgrid_2D()
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{annotate\_spatlocs\_with\_spatgrid\_3D}{annotate\_spatlocs\_with\_spatgrid\_3D}{annotate.Rul.spatlocs.Rul.with.Rul.spatgrid.Rul.3D}
%
\begin{Description}\relax
annotate spatial locations with 3D spatial grid information
\end{Description}
%
\begin{Usage}
\begin{verbatim}
annotate_spatlocs_with_spatgrid_3D(spatloc, spatgrid)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{spatloc}] spatial\_locs slot from giotto object

\item[\code{spatgrid}] selected spatial\_grid slot from giotto object
\end{ldescription}
\end{Arguments}
%
\begin{Value}
annotated spatial location data.table
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    annotate_spatlocs_with_spatgrid_3D()
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{average\_gene\_gene\_expression\_in\_groups}{average\_gene\_gene\_expression\_in\_groups}{average.Rul.gene.Rul.gene.Rul.expression.Rul.in.Rul.groups}
%
\begin{Description}\relax
calculate average expression per cluster
\end{Description}
%
\begin{Usage}
\begin{verbatim}
average_gene_gene_expression_in_groups(
  gobject,
  cluster_column = "cell_types",
  gene_set_1,
  gene_set_2
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object to use

\item[\code{cluster\_column}] cluster column with cell type information

\item[\code{gene\_set\_1}] first specific gene set from gene pairs

\item[\code{gene\_set\_2}] second specific gene set from gene pairs
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Details will follow soon.
\end{Details}
%
\begin{Value}
data.table with average expression scores for each cluster
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    average_gene_gene_expression_in_groups(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{binSpect}{binSpect}{binSpect}
%
\begin{Description}\relax
Previously: binGetSpatialGenes. BinSpect (Binary Spatial Extraction of genes) is a fast computational method
that identifies genes with a spatially coherent expression pattern.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
binSpect(
  gobject,
  bin_method = c("kmeans", "rank"),
  expression_values = c("normalized", "scaled", "custom"),
  subset_genes = NULL,
  spatial_network_name = "Delaunay_network",
  nstart = 3,
  iter_max = 10,
  percentage_rank = 30,
  do_fisher_test = TRUE,
  calc_hub = FALSE,
  hub_min_int = 3,
  get_av_expr = TRUE,
  get_high_expr = TRUE,
  do_parallel = TRUE,
  cores = NA,
  verbose = T
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{bin\_method}] method to binarize gene expression

\item[\code{expression\_values}] expression values to use

\item[\code{subset\_genes}] only select a subset of genes to test

\item[\code{spatial\_network\_name}] name of spatial network to use (default = 'spatial\_network')

\item[\code{nstart}] kmeans: nstart parameter

\item[\code{iter\_max}] kmeans: iter.max parameter

\item[\code{percentage\_rank}] percentage of top cells for binarization

\item[\code{do\_fisher\_test}] perform fisher test

\item[\code{calc\_hub}] calculate the number of hub cells

\item[\code{hub\_min\_int}] minimum number of cell-cell interactions for a hub cell

\item[\code{get\_av\_expr}] calculate the average expression per gene of the high expressing cells

\item[\code{get\_high\_expr}] calculate the number of high expressing cells  per gene

\item[\code{do\_parallel}] run calculations in parallel with mclapply

\item[\code{cores}] number of cores to use if do\_parallel = TRUE

\item[\code{verbose}] be verbose
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
We provide two ways to identify spatial genes based on gene expression binarization.
Both methods are identicial except for how binarization is performed.
\begin{itemize}

\item{} 1. binarize: Each gene is binarized (0 or 1) in each cell with \bold{kmeans} (k = 2) or based on \bold{rank} percentile
\item{} 2. network: Alll cells are connected through a spatial network based on the physical coordinates
\item{} 3. contingency table: A contingency table is calculated based on all edges of neighboring cells and the binarized expression (0-0, 0-1, 1-0 or 1-1)
\item{} 4. For each gene an odds-ratio (OR) and fisher.test (optional) is calculated

\end{itemize}

Other statistics are provided (optional):
\begin{itemize}

\item{} Number of cells with high expression (binary = 1)
\item{} Average expression of each gene within high expressing cells 
\item{} Number of hub cells, these are high expressing cells that have a user defined number of high expressing neighbors

\end{itemize}

By selecting a subset of likely spatial genes (e.g. soft thresholding highly variable genes) or using multiple cores can accelerate the speed.
\end{Details}
%
\begin{Value}
data.table with results (see details)
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    binSpect(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{calculateHVG}{calculateHVG}{calculateHVG}
%
\begin{Description}\relax
compute highly variable genes
\end{Description}
%
\begin{Usage}
\begin{verbatim}
calculateHVG(
  gobject,
  expression_values = c("normalized", "scaled", "custom"),
  method = c("cov_groups", "cov_loess"),
  reverse_log_scale = FALSE,
  logbase = 2,
  expression_threshold = 0,
  nr_expression_groups = 20,
  zscore_threshold = 1.5,
  HVGname = "hvg",
  difference_in_cov = 0.1,
  show_plot = NA,
  return_plot = NA,
  save_plot = NA,
  save_param = list(),
  default_save_name = "HVGplot",
  return_gobject = TRUE
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{expression\_values}] expression values to use

\item[\code{method}] method to calculate highly variable genes

\item[\code{reverse\_log\_scale}] reverse log-scale of expression values (default = FALSE)

\item[\code{logbase}] if reverse\_log\_scale is TRUE, which log base was used?

\item[\code{expression\_threshold}] expression threshold to consider a gene detected

\item[\code{nr\_expression\_groups}] number of expression groups for cov\_groups

\item[\code{zscore\_threshold}] zscore to select hvg for cov\_groups

\item[\code{HVGname}] name for highly variable genes in cell metadata

\item[\code{difference\_in\_cov}] minimum difference in coefficient of variance required

\item[\code{show\_plot}] show plot

\item[\code{return\_plot}] return ggplot object

\item[\code{save\_plot}] directly save the plot [boolean]

\item[\code{save\_param}] list of saving parameters from \code{\LinkA{all\_plots\_save\_function}{all.Rul.plots.Rul.save.Rul.function}}

\item[\code{default\_save\_name}] default save name for saving, don't change, change save\_name in save\_param

\item[\code{return\_gobject}] boolean: return giotto object (default = TRUE)
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Currently we provide 2 ways to calculate highly variable genes:
\bold{1. high coeff of variance (COV) within groups: } \\{}
First genes are binned (\emph{nr\_expression\_groups}) into average expression groups and
the COV for each gene is converted into a z-score within each bin. Genes with a z-score
higher than the threshold (\emph{zscore\_threshold}) are considered highly variable.  \\{}

\bold{2. high COV based on loess regression prediction: } \\{}
A predicted COV is calculated for each gene using loess regression (COV\textasciitilde{}log(mean expression))
Genes that show a higher than predicted COV (\emph{difference\_in\_cov}) are considered highly variable. \\{}
\end{Details}
%
\begin{Value}
giotto object highly variable genes appended to gene metadata (fDataDT)
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    calculateHVG(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{calculateMetaTable}{calculateMetaTable}{calculateMetaTable}
%
\begin{Description}\relax
calculates the average gene expression for one or more (combined) annotation columns.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
calculateMetaTable(
  gobject,
  expression_values = c("normalized", "scaled", "custom"),
  metadata_cols = NULL,
  selected_genes = NULL
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{expression\_values}] expression values to use

\item[\code{metadata\_cols}] annotation columns found in pDataDT(gobject)

\item[\code{selected\_genes}] subset of genes to use
\end{ldescription}
\end{Arguments}
%
\begin{Value}
data.table with average expression values for each gene per (combined) annotation
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    calculateMetaTable(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{calculateMetaTableCells}{calculateMetaTableCells}{calculateMetaTableCells}
%
\begin{Description}\relax
calculates the average metadata values for one or more (combined) annotation columns.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
calculateMetaTableCells(
  gobject,
  value_cols = NULL,
  metadata_cols = NULL,
  spat_enr_names = NULL
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{value\_cols}] metadata or enrichment value columns to use

\item[\code{metadata\_cols}] annotation columns found in pDataDT(gobject)

\item[\code{spat\_enr\_names}] which spatial enrichment results to include
\end{ldescription}
\end{Arguments}
%
\begin{Value}
data.table with average metadata values per (combined) annotation
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    calculateMetaTableCells(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{calculate\_distance\_and\_weight}{calculate\_distance\_and\_weight}{calculate.Rul.distance.Rul.and.Rul.weight}
%
\begin{Description}\relax
calculate\_distance\_and\_weight
\end{Description}
%
\begin{Usage}
\begin{verbatim}
calculate_distance_and_weight(
  networkDT,
  sdimx = "sdimx",
  sdimy = "sdimy",
  sdimz = "sdimz",
  d2_or_d3 = c(2, 3)
)
\end{verbatim}
\end{Usage}
\inputencoding{utf8}
\HeaderA{cellProximityBarplot}{cellProximityBarplot}{cellProximityBarplot}
%
\begin{Description}\relax
Create barplot from cell-cell proximity scores
\end{Description}
%
\begin{Usage}
\begin{verbatim}
cellProximityBarplot(
  gobject,
  CPscore,
  min_orig_ints = 5,
  min_sim_ints = 5,
  p_val = 0.05,
  show_plot = NA,
  return_plot = NA,
  save_plot = NA,
  save_param = list(),
  default_save_name = "cellProximityBarplot"
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{CPscore}] CPscore, output from cellProximityEnrichment()

\item[\code{min\_orig\_ints}] filter on minimum original cell-cell interactions

\item[\code{min\_sim\_ints}] filter on minimum simulated cell-cell interactions

\item[\code{p\_val}] p-value

\item[\code{show\_plot}] show plot

\item[\code{return\_plot}] return ggplot object

\item[\code{save\_plot}] directly save the plot [boolean]

\item[\code{save\_param}] list of saving parameters from \code{\LinkA{all\_plots\_save\_function}{all.Rul.plots.Rul.save.Rul.function}}

\item[\code{default\_save\_name}] default save name for saving, don't change, change save\_name in save\_param
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
This function creates a barplot that shows the  spatial proximity
enrichment or depletion of cell type pairs.
\end{Details}
%
\begin{Value}
ggplot barplot
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    cellProximityBarplot(CPscore)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{cellProximityEnrichment}{cellProximityEnrichment}{cellProximityEnrichment}
%
\begin{Description}\relax
Compute cell-cell interaction enrichment (observed vs expected)
\end{Description}
%
\begin{Usage}
\begin{verbatim}
cellProximityEnrichment(
  gobject,
  spatial_network_name = "Delaunay_network",
  cluster_column,
  number_of_simulations = 1000,
  adjust_method = c("none", "fdr", "bonferroni", "BH", "holm", "hochberg", "hommel",
    "BY")
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{spatial\_network\_name}] name of spatial network to use

\item[\code{cluster\_column}] name of column to use for clusters

\item[\code{number\_of\_simulations}] number of simulations to create expected observations
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Spatial proximity enrichment or depletion between pairs of cell types
is calculated by calculating the observed over the expected frequency
of cell-cell proximity interactions. The expected frequency is the average frequency
calculated from a number of spatial network simulations. Each individual simulation is
obtained by reshuffling the cell type labels of each node (cell)
in the spatial network.
\end{Details}
%
\begin{Value}
List of cell Proximity scores (CPscores) in data.table format. The first
data.table (raw\_sim\_table) shows the raw observations of both the original and
simulated networks. The second data.table (enrichm\_res) shows the enrichment results.
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    cellProximityEnrichment(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{cellProximityHeatmap}{cellProximityHeatmap}{cellProximityHeatmap}
%
\begin{Description}\relax
Create heatmap from cell-cell proximity scores
\end{Description}
%
\begin{Usage}
\begin{verbatim}
cellProximityHeatmap(
  gobject,
  CPscore,
  scale = T,
  order_cell_types = T,
  color_breaks = NULL,
  color_names = NULL,
  show_plot = NA,
  return_plot = NA,
  save_plot = NA,
  save_param = list(),
  default_save_name = "cellProximityHeatmap"
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{CPscore}] CPscore, output from cellProximityEnrichment()

\item[\code{scale}] scale cell-cell proximity interaction scores

\item[\code{order\_cell\_types}] order cell types based on enrichment correlation

\item[\code{color\_breaks}] numerical vector of length 3 to represent min, mean and maximum

\item[\code{color\_names}] character color vector of length 3

\item[\code{show\_plot}] show plot

\item[\code{return\_plot}] return ggplot object

\item[\code{save\_plot}] directly save the plot [boolean]

\item[\code{save\_param}] list of saving parameters from \code{\LinkA{all\_plots\_save\_function}{all.Rul.plots.Rul.save.Rul.function}}

\item[\code{default\_save\_name}] default save name for saving, don't change, change save\_name in save\_param
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
This function creates a heatmap that shows the  spatial proximity
enrichment or depletion of cell type pairs.
\end{Details}
%
\begin{Value}
ggplot heatmap
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    cellProximityHeatmap(CPscore)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{cellProximityNetwork}{cellProximityNetwork}{cellProximityNetwork}
%
\begin{Description}\relax
Create network from cell-cell proximity scores
\end{Description}
%
\begin{Usage}
\begin{verbatim}
cellProximityNetwork(
  gobject,
  CPscore,
  remove_self_edges = FALSE,
  self_loop_strength = 0.1,
  color_depletion = "lightgreen",
  color_enrichment = "red",
  rescale_edge_weights = TRUE,
  edge_weight_range_depletion = c(0.1, 1),
  edge_weight_range_enrichment = c(1, 5),
  layout = c("Fruchterman", "DrL", "Kamada-Kawai"),
  only_show_enrichment_edges = F,
  edge_width_range = c(0.1, 2),
  node_size = 4,
  node_text_size = 6,
  show_plot = NA,
  return_plot = NA,
  save_plot = NA,
  save_param = list(),
  default_save_name = "cellProximityNetwork"
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{CPscore}] CPscore, output from cellProximityEnrichment()

\item[\code{remove\_self\_edges}] remove enrichment/depletion edges with itself

\item[\code{self\_loop\_strength}] size of self-loops

\item[\code{color\_depletion}] color for depleted cell-cell interactions

\item[\code{color\_enrichment}] color for enriched cell-cell interactions

\item[\code{rescale\_edge\_weights}] rescale edge weights (boolean)

\item[\code{edge\_weight\_range\_depletion}] numerical vector of length 2 to rescale depleted edge weights

\item[\code{edge\_weight\_range\_enrichment}] numerical vector of length 2 to rescale enriched edge weights

\item[\code{layout}] layout algorithm to use to draw nodes and edges

\item[\code{only\_show\_enrichment\_edges}] show only the enriched pairwise scores

\item[\code{edge\_width\_range}] range of edge width

\item[\code{node\_size}] size of nodes

\item[\code{node\_text\_size}] size of node labels

\item[\code{show\_plot}] show plot

\item[\code{return\_plot}] return ggplot object

\item[\code{save\_plot}] directly save the plot [boolean]

\item[\code{save\_param}] list of saving parameters from \code{\LinkA{all\_plots\_save\_function}{all.Rul.plots.Rul.save.Rul.function}}

\item[\code{default\_save\_name}] default save name for saving, don't change, change save\_name in save\_param
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
This function creates a network that shows the  spatial proximity
enrichment or depletion of cell type pairs.
\end{Details}
%
\begin{Value}
igraph plot
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    cellProximityNetwork(CPscore)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{cellProximitySpatPlot}{cellProximitySpatPlot}{cellProximitySpatPlot}
%
\begin{Description}\relax
Visualize 2D cell-cell interactions according to spatial coordinates in ggplot mode
\end{Description}
%
\begin{Usage}
\begin{verbatim}
cellProximitySpatPlot(gobject, ...)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{interaction\_name}] cell-cell interaction name

\item[\code{cluster\_column}] cluster column with cell clusters

\item[\code{sdimx}] x-axis dimension name (default = 'sdimx')

\item[\code{sdimy}] y-axis dimension name (default = 'sdimy')

\item[\code{cell\_color}] color for cells (see details)

\item[\code{cell\_color\_code}] named vector with colors

\item[\code{color\_as\_factor}] convert color column to factor

\item[\code{show\_other\_cells}] decide if show cells not in network

\item[\code{show\_network}] show underlying spatial network

\item[\code{network\_color}] color of spatial network

\item[\code{spatial\_network\_name}] name of spatial network to use

\item[\code{show\_grid}] show spatial grid

\item[\code{grid\_color}] color of spatial grid

\item[\code{spatial\_grid\_name}] name of spatial grid to use

\item[\code{coord\_fix\_ratio}] fix ratio between x and y-axis

\item[\code{show\_legend}] show legend

\item[\code{point\_size\_select}] size of selected points

\item[\code{point\_select\_border\_col}] border color of selected points

\item[\code{point\_select\_border\_stroke}] stroke size of selected points

\item[\code{point\_size\_other}] size of other points

\item[\code{point\_other\_border\_col}] border color of other points

\item[\code{point\_other\_border\_stroke}] stroke size of other points

\item[\code{show\_plot}] show plots

\item[\code{return\_plot}] return ggplot object

\item[\code{save\_plot}] directly save the plot [boolean]

\item[\code{save\_param}] list of saving parameters from \code{\LinkA{all\_plots\_save\_function}{all.Rul.plots.Rul.save.Rul.function}}

\item[\code{default\_save\_name}] default save name for saving, don't change, change save\_name in save\_param
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Description of parameters.
\end{Details}
%
\begin{Value}
ggplot
\end{Value}
%
\begin{SeeAlso}\relax
\code{\LinkA{cellProximitySpatPlot2D}{cellProximitySpatPlot2D}} and \code{\LinkA{cellProximitySpatPlot3D}{cellProximitySpatPlot3D}} for 3D
\end{SeeAlso}
%
\begin{Examples}
\begin{ExampleCode}
    cellProximitySpatPlot(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{cellProximitySpatPlot2D}{cellProximitySpatPlot2D}{cellProximitySpatPlot2D}
%
\begin{Description}\relax
Visualize 2D cell-cell interactions according to spatial coordinates in ggplot mode
\end{Description}
%
\begin{Usage}
\begin{verbatim}
cellProximitySpatPlot2D(
  gobject,
  interaction_name = NULL,
  cluster_column = NULL,
  sdimx = "sdimx",
  sdimy = "sdimy",
  cell_color = NULL,
  cell_color_code = NULL,
  color_as_factor = T,
  show_other_cells = F,
  show_network = F,
  show_other_network = F,
  network_color = NULL,
  spatial_network_name = "Delaunay_network",
  show_grid = F,
  grid_color = NULL,
  spatial_grid_name = "spatial_grid",
  coord_fix_ratio = 1,
  show_legend = T,
  point_size_select = 2,
  point_select_border_col = "black",
  point_select_border_stroke = 0.05,
  point_size_other = 1,
  point_alpha_other = 0.3,
  point_other_border_col = "lightgrey",
  point_other_border_stroke = 0.01,
  show_plot = NA,
  return_plot = NA,
  save_plot = NA,
  save_param = list(),
  default_save_name = "cellProximitySpatPlot2D"
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{interaction\_name}] cell-cell interaction name

\item[\code{cluster\_column}] cluster column with cell clusters

\item[\code{sdimx}] x-axis dimension name (default = 'sdimx')

\item[\code{sdimy}] y-axis dimension name (default = 'sdimy')

\item[\code{cell\_color}] color for cells (see details)

\item[\code{cell\_color\_code}] named vector with colors

\item[\code{color\_as\_factor}] convert color column to factor

\item[\code{show\_other\_cells}] decide if show cells not in network

\item[\code{show\_network}] show underlying spatial network

\item[\code{network\_color}] color of spatial network

\item[\code{spatial\_network\_name}] name of spatial network to use

\item[\code{show\_grid}] show spatial grid

\item[\code{grid\_color}] color of spatial grid

\item[\code{spatial\_grid\_name}] name of spatial grid to use

\item[\code{coord\_fix\_ratio}] fix ratio between x and y-axis

\item[\code{show\_legend}] show legend

\item[\code{point\_size\_select}] size of selected points

\item[\code{point\_select\_border\_col}] border color of selected points

\item[\code{point\_select\_border\_stroke}] stroke size of selected points

\item[\code{point\_size\_other}] size of other points

\item[\code{point\_other\_border\_col}] border color of other points

\item[\code{point\_other\_border\_stroke}] stroke size of other points

\item[\code{show\_plot}] show plots

\item[\code{return\_plot}] return ggplot object

\item[\code{save\_plot}] directly save the plot [boolean]

\item[\code{save\_param}] list of saving parameters from \code{\LinkA{all\_plots\_save\_function}{all.Rul.plots.Rul.save.Rul.function}}

\item[\code{default\_save\_name}] default save name for saving, don't change, change save\_name in save\_param
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Description of parameters.
\end{Details}
%
\begin{Value}
ggplot
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    cellProximitySpatPlot2D(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{cellProximitySpatPlot3D}{cellProximitySpatPlot2D}{cellProximitySpatPlot3D}
%
\begin{Description}\relax
Visualize 3D cell-cell interactions according to spatial coordinates in plotly mode
\end{Description}
%
\begin{Usage}
\begin{verbatim}
cellProximitySpatPlot3D(
  gobject,
  interaction_name = NULL,
  cluster_column = NULL,
  sdimx = "sdimx",
  sdimy = "sdimy",
  sdimz = "sdimz",
  cell_color = NULL,
  cell_color_code = NULL,
  color_as_factor = T,
  show_other_cells = T,
  show_network = T,
  show_other_network = F,
  network_color = NULL,
  spatial_network_name = "Delaunay_network",
  show_grid = F,
  grid_color = NULL,
  spatial_grid_name = "spatial_grid",
  show_legend = T,
  point_size_select = 4,
  point_size_other = 2,
  point_alpha_other = 0.5,
  axis_scale = c("cube", "real", "custom"),
  custom_ratio = NULL,
  x_ticks = NULL,
  y_ticks = NULL,
  z_ticks = NULL,
  show_plot = NA,
  return_plot = NA,
  save_plot = NA,
  save_param = list(),
  default_save_name = "cellProximitySpatPlot3D",
  ...
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{interaction\_name}] cell-cell interaction name

\item[\code{cluster\_column}] cluster column with cell clusters

\item[\code{sdimx}] x-axis dimension name (default = 'sdimx')

\item[\code{sdimy}] y-axis dimension name (default = 'sdimy')

\item[\code{sdimz}] z-axis dimension name (default = 'sdimz')

\item[\code{cell\_color}] color for cells (see details)

\item[\code{cell\_color\_code}] named vector with colors

\item[\code{color\_as\_factor}] convert color column to factor

\item[\code{show\_other\_cells}] decide if show cells not in network

\item[\code{show\_network}] show underlying spatial network

\item[\code{network\_color}] color of spatial network

\item[\code{spatial\_network\_name}] name of spatial network to use

\item[\code{show\_grid}] show spatial grid

\item[\code{grid\_color}] color of spatial grid

\item[\code{spatial\_grid\_name}] name of spatial grid to use

\item[\code{show\_legend}] show legend

\item[\code{point\_size\_select}] size of selected points

\item[\code{point\_size\_other}] size of other points

\item[\code{show\_plot}] show plots

\item[\code{return\_plot}] return plotly object

\item[\code{save\_plot}] directly save the plot [boolean]

\item[\code{save\_param}] list of saving parameters from \code{\LinkA{all\_plots\_save\_function}{all.Rul.plots.Rul.save.Rul.function}}

\item[\code{default\_save\_name}] default save name for saving, don't change, change save\_name in save\_param
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Description of parameters.
\end{Details}
%
\begin{Value}
plotly
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    cellProximitySpatPlot3D(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{cellProximityVisPlot}{cellProximityVisPlot}{cellProximityVisPlot}
%
\begin{Description}\relax
Visualize cell-cell interactions according to spatial coordinates
\end{Description}
%
\begin{Usage}
\begin{verbatim}
cellProximityVisPlot(
  gobject,
  interaction_name = NULL,
  cluster_column = NULL,
  sdimx = NULL,
  sdimy = NULL,
  sdimz = NULL,
  cell_color = NULL,
  cell_color_code = NULL,
  color_as_factor = T,
  show_other_cells = F,
  show_network = F,
  show_other_network = F,
  network_color = NULL,
  spatial_network_name = "Delaunay_network",
  show_grid = F,
  grid_color = NULL,
  spatial_grid_name = "spatial_grid",
  coord_fix_ratio = 1,
  show_legend = T,
  point_size_select = 2,
  point_select_border_col = "black",
  point_select_border_stroke = 0.05,
  point_size_other = 1,
  point_alpha_other = 0.3,
  point_other_border_col = "lightgrey",
  point_other_border_stroke = 0.01,
  axis_scale = c("cube", "real", "custom"),
  custom_ratio = NULL,
  x_ticks = NULL,
  y_ticks = NULL,
  z_ticks = NULL,
  plot_method = c("ggplot", "plotly"),
  ...
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{interaction\_name}] cell-cell interaction name

\item[\code{cluster\_column}] cluster column with cell clusters

\item[\code{sdimx}] x-axis dimension name (default = 'sdimx')

\item[\code{sdimy}] y-axis dimension name (default = 'sdimy')

\item[\code{sdimz}] z-axis dimension name (default = 'sdimz')

\item[\code{cell\_color}] color for cells (see details)

\item[\code{cell\_color\_code}] named vector with colors

\item[\code{color\_as\_factor}] convert color column to factor

\item[\code{show\_network}] show underlying spatial network

\item[\code{network\_color}] color of spatial network

\item[\code{spatial\_network\_name}] name of spatial network to use

\item[\code{show\_grid}] show spatial grid

\item[\code{grid\_color}] color of spatial grid

\item[\code{spatial\_grid\_name}] name of spatial grid to use

\item[\code{coord\_fix\_ratio}] fix ratio between x and y-axis

\item[\code{show\_legend}] show legend

\item[\code{point\_size\_select}] size of selected points

\item[\code{point\_select\_border\_col}] border color of selected points

\item[\code{point\_select\_border\_stroke}] stroke size of selected points

\item[\code{point\_size\_other}] size of other points

\item[\code{point\_other\_border\_col}] border color of other points

\item[\code{point\_other\_border\_stroke}] stroke size of other points
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Description of parameters.
\end{Details}
%
\begin{Value}
ggplot or plotly
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    cellProximityVisPlot(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{cellProximityVisPlot\_2D\_ggplot}{cellProximityVisPlot\_2D\_ggplot}{cellProximityVisPlot.Rul.2D.Rul.ggplot}
%
\begin{Description}\relax
Visualize 2D cell-cell interactions according to spatial coordinates in ggplot mode
\end{Description}
%
\begin{Usage}
\begin{verbatim}
cellProximityVisPlot_2D_ggplot(
  gobject,
  interaction_name = NULL,
  cluster_column = NULL,
  sdimx = NULL,
  sdimy = NULL,
  cell_color = NULL,
  cell_color_code = NULL,
  color_as_factor = T,
  show_other_cells = F,
  show_network = F,
  show_other_network = F,
  network_color = NULL,
  spatial_network_name = "Delaunay_network",
  show_grid = F,
  grid_color = NULL,
  spatial_grid_name = "spatial_grid",
  coord_fix_ratio = 1,
  show_legend = T,
  point_size_select = 2,
  point_select_border_col = "black",
  point_select_border_stroke = 0.05,
  point_size_other = 1,
  point_alpha_other = 0.3,
  point_other_border_col = "lightgrey",
  point_other_border_stroke = 0.01,
  ...
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{interaction\_name}] cell-cell interaction name

\item[\code{cluster\_column}] cluster column with cell clusters

\item[\code{sdimx}] x-axis dimension name (default = 'sdimx')

\item[\code{sdimy}] y-axis dimension name (default = 'sdimy')

\item[\code{cell\_color}] color for cells (see details)

\item[\code{cell\_color\_code}] named vector with colors

\item[\code{color\_as\_factor}] convert color column to factor

\item[\code{show\_other\_cells}] decide if show cells not in network

\item[\code{show\_network}] show underlying spatial network

\item[\code{network\_color}] color of spatial network

\item[\code{spatial\_network\_name}] name of spatial network to use

\item[\code{show\_grid}] show spatial grid

\item[\code{grid\_color}] color of spatial grid

\item[\code{spatial\_grid\_name}] name of spatial grid to use

\item[\code{coord\_fix\_ratio}] fix ratio between x and y-axis

\item[\code{show\_legend}] show legend

\item[\code{point\_size\_select}] size of selected points

\item[\code{point\_select\_border\_col}] border color of selected points

\item[\code{point\_select\_border\_stroke}] stroke size of selected points

\item[\code{point\_size\_other}] size of other points

\item[\code{point\_other\_border\_col}] border color of other points

\item[\code{point\_other\_border\_stroke}] stroke size of other points
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Description of parameters.
\end{Details}
%
\begin{Value}
ggplot
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    cellProximityVisPlot_2D_ggplot(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{cellProximityVisPlot\_2D\_plotly}{cellProximityVisPlot\_2D\_plotly}{cellProximityVisPlot.Rul.2D.Rul.plotly}
%
\begin{Description}\relax
Visualize 2D cell-cell interactions according to spatial coordinates in plotly mode
\end{Description}
%
\begin{Usage}
\begin{verbatim}
cellProximityVisPlot_2D_plotly(
  gobject,
  interaction_name = NULL,
  cluster_column = NULL,
  sdimx = NULL,
  sdimy = NULL,
  cell_color = NULL,
  cell_color_code = NULL,
  color_as_factor = T,
  show_other_cells = F,
  show_network = F,
  show_other_network = F,
  network_color = NULL,
  spatial_network_name = "Delaunay_network",
  show_grid = F,
  grid_color = NULL,
  spatial_grid_name = "spatial_grid",
  show_legend = T,
  point_size_select = 2,
  point_size_other = 1,
  point_alpha_other = 0.3,
  axis_scale = c("cube", "real", "custom"),
  custom_ratio = NULL,
  x_ticks = NULL,
  y_ticks = NULL,
  ...
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{interaction\_name}] cell-cell interaction name

\item[\code{cluster\_column}] cluster column with cell clusters

\item[\code{sdimx}] x-axis dimension name (default = 'sdimx')

\item[\code{sdimy}] y-axis dimension name (default = 'sdimy')

\item[\code{cell\_color}] color for cells (see details)

\item[\code{cell\_color\_code}] named vector with colors

\item[\code{color\_as\_factor}] convert color column to factor

\item[\code{show\_other\_cells}] decide if show cells not in network

\item[\code{show\_network}] show underlying spatial network

\item[\code{network\_color}] color of spatial network

\item[\code{spatial\_network\_name}] name of spatial network to use

\item[\code{show\_grid}] show spatial grid

\item[\code{grid\_color}] color of spatial grid

\item[\code{spatial\_grid\_name}] name of spatial grid to use

\item[\code{show\_legend}] show legend

\item[\code{point\_size\_select}] size of selected points

\item[\code{coord\_fix\_ratio}] fix ratio between x and y-axis
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Description of parameters.
\end{Details}
%
\begin{Value}
plotly
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    cellProximityVisPlot_2D_plotly(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{cellProximityVisPlot\_3D\_plotly}{cellProximityVisPlot\_3D\_plotly}{cellProximityVisPlot.Rul.3D.Rul.plotly}
%
\begin{Description}\relax
Visualize 3D cell-cell interactions according to spatial coordinates in plotly mode
\end{Description}
%
\begin{Usage}
\begin{verbatim}
cellProximityVisPlot_3D_plotly(
  gobject,
  interaction_name = NULL,
  cluster_column = NULL,
  sdimx = NULL,
  sdimy = NULL,
  sdimz = NULL,
  cell_color = NULL,
  cell_color_code = NULL,
  color_as_factor = T,
  show_other_cells = F,
  show_network = F,
  show_other_network = F,
  network_color = NULL,
  spatial_network_name = "Delaunay_network",
  show_grid = F,
  grid_color = NULL,
  spatial_grid_name = "spatial_grid",
  show_legend = T,
  point_size_select = 2,
  point_size_other = 1,
  point_alpha_other = 0.5,
  axis_scale = c("cube", "real", "custom"),
  custom_ratio = NULL,
  x_ticks = NULL,
  y_ticks = NULL,
  z_ticks = NULL,
  ...
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{interaction\_name}] cell-cell interaction name

\item[\code{cluster\_column}] cluster column with cell clusters

\item[\code{sdimx}] x-axis dimension name (default = 'sdimx')

\item[\code{sdimy}] y-axis dimension name (default = 'sdimy')

\item[\code{sdimz}] z-axis dimension name (default = 'sdimz')

\item[\code{cell\_color}] color for cells (see details)

\item[\code{cell\_color\_code}] named vector with colors

\item[\code{color\_as\_factor}] convert color column to factor

\item[\code{show\_other\_cells}] decide if show cells not in network

\item[\code{show\_network}] show underlying spatial network

\item[\code{network\_color}] color of spatial network

\item[\code{spatial\_network\_name}] name of spatial network to use

\item[\code{show\_grid}] show spatial grid

\item[\code{grid\_color}] color of spatial grid

\item[\code{spatial\_grid\_name}] name of spatial grid to use

\item[\code{show\_legend}] show legend

\item[\code{point\_size\_select}] size of selected points

\item[\code{coord\_fix\_ratio}] fix ratio between x and y-axis
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Description of parameters.
\end{Details}
%
\begin{Value}
plotly
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    cellProximityVisPlot_3D_plotly(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{changeGiottoInstructions}{changeGiottoInstructions}{changeGiottoInstructions}
%
\begin{Description}\relax
Function to change one or more instructions from giotto object
\end{Description}
%
\begin{Usage}
\begin{verbatim}
changeGiottoInstructions(
  gobject,
  params = NULL,
  new_values = NULL,
  return_gobject = TRUE
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{params}] parameter(s) to change

\item[\code{new\_values}] new value(s) for parameter(s)

\item[\code{return\_gobject}] (boolean) return giotto object
\end{ldescription}
\end{Arguments}
%
\begin{Value}
named vector with giotto instructions
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    changeGiottoInstructions()
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{clusterCells}{clusterCells}{clusterCells}
%
\begin{Description}\relax
cluster cells using a variety of different methods
\end{Description}
%
\begin{Usage}
\begin{verbatim}
clusterCells(
  gobject,
  cluster_method = c("leiden", "louvain_community", "louvain_multinet", "randomwalk",
    "sNNclust", "kmeans", "hierarchical"),
  name = "cluster_name",
  nn_network_to_use = "sNN",
  network_name = "sNN.pca",
  pyth_leid_resolution = 1,
  pyth_leid_weight_col = "weight",
  pyth_leid_part_type = c("RBConfigurationVertexPartition",
    "ModularityVertexPartition"),
  pyth_leid_init_memb = NULL,
  pyth_leid_iterations = 1000,
  pyth_louv_resolution = 1,
  pyth_louv_weight_col = NULL,
  python_louv_random = F,
  python_path = NULL,
  louvain_gamma = 1,
  louvain_omega = 1,
  walk_steps = 4,
  walk_clusters = 10,
  walk_weights = NA,
  sNNclust_k = 20,
  sNNclust_eps = 4,
  sNNclust_minPts = 16,
  borderPoints = TRUE,
  expression_values = c("normalized", "scaled", "custom"),
  genes_to_use = NULL,
  dim_reduction_to_use = c("cells", "pca", "umap", "tsne"),
  dim_reduction_name = "pca",
  dimensions_to_use = 1:10,
  distance_method = c("original", "pearson", "spearman", "euclidean", "maximum",
    "manhattan", "canberra", "binary", "minkowski"),
  km_centers = 10,
  km_iter_max = 100,
  km_nstart = 1000,
  km_algorithm = "Hartigan-Wong",
  hc_agglomeration_method = c("ward.D2", "ward.D", "single", "complete", "average",
    "mcquitty", "median", "centroid"),
  hc_k = 10,
  hc_h = NULL,
  return_gobject = TRUE,
  set_seed = T,
  seed_number = 1234
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{cluster\_method}] community cluster method to use

\item[\code{name}] name for new clustering result

\item[\code{nn\_network\_to\_use}] type of NN network to use (kNN vs sNN)

\item[\code{network\_name}] name of NN network to use

\item[\code{pyth\_leid\_resolution}] resolution for leiden

\item[\code{pyth\_leid\_weight\_col}] column to use for weights

\item[\code{pyth\_leid\_part\_type}] partition type to use

\item[\code{pyth\_leid\_init\_memb}] initial membership

\item[\code{pyth\_leid\_iterations}] number of iterations

\item[\code{pyth\_louv\_resolution}] resolution for louvain

\item[\code{pyth\_louv\_weight\_col}] python louvain param: weight column

\item[\code{python\_louv\_random}] python louvain param: random

\item[\code{python\_path}] specify specific path to python if required

\item[\code{louvain\_gamma}] louvain param: gamma or resolution

\item[\code{louvain\_omega}] louvain param: omega

\item[\code{walk\_steps}] randomwalk: number of steps

\item[\code{walk\_clusters}] randomwalk: number of clusters

\item[\code{walk\_weights}] randomwalk: weight column

\item[\code{sNNclust\_k}] SNNclust: k neighbors to use

\item[\code{sNNclust\_eps}] SNNclust: epsilon

\item[\code{sNNclust\_minPts}] SNNclust: min points

\item[\code{borderPoints}] SNNclust: border points

\item[\code{expression\_values}] expression values to use

\item[\code{genes\_to\_use}] = NULL,

\item[\code{dim\_reduction\_to\_use}] dimension reduction to use

\item[\code{dim\_reduction\_name}] name of reduction 'pca',

\item[\code{dimensions\_to\_use}] dimensions to use

\item[\code{distance\_method}] distance method

\item[\code{km\_centers}] kmeans centers

\item[\code{km\_iter\_max}] kmeans iterations

\item[\code{km\_nstart}] kmeans random starting points

\item[\code{km\_algorithm}] kmeans algorithm

\item[\code{hc\_agglomeration\_method}] hierarchical clustering method

\item[\code{hc\_k}] hierachical number of clusters

\item[\code{hc\_h}] hierarchical tree cutoff

\item[\code{return\_gobject}] boolean: return giotto object (default = TRUE)

\item[\code{set\_seed}] set seed

\item[\code{seed\_number}] number for seed
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Wrapper for the different clustering methods.
\end{Details}
%
\begin{Value}
giotto object with new clusters appended to cell metadata
\end{Value}
%
\begin{SeeAlso}\relax
\code{\LinkA{doLeidenCluster}{doLeidenCluster}}, \code{\LinkA{doLouvainCluster\_community}{doLouvainCluster.Rul.community}}, \code{\LinkA{doLouvainCluster\_multinet}{doLouvainCluster.Rul.multinet}},
\code{\LinkA{doLouvainCluster}{doLouvainCluster}}, \code{\LinkA{doRandomWalkCluster}{doRandomWalkCluster}}, \code{\LinkA{doSNNCluster}{doSNNCluster}},
\code{\LinkA{doKmeans}{doKmeans}}, \code{\LinkA{doHclust}{doHclust}}
\end{SeeAlso}
%
\begin{Examples}
\begin{ExampleCode}
    clusterCells(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{clusterSpatialCorGenes}{clusterSpatialCorGenes}{clusterSpatialCorGenes}
%
\begin{Description}\relax
Cluster based on spatially correlated genes
\end{Description}
%
\begin{Usage}
\begin{verbatim}
clusterSpatialCorGenes(
  spatCorObject,
  name = "spat_clus",
  hclust_method = "ward.D",
  k = 10,
  return_obj = TRUE
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{spatCorObject}] spatial correlation object

\item[\code{name}] name for spatial clustering results

\item[\code{hclust\_method}] method for hierarchical clustering

\item[\code{k}] number of clusters to extract

\item[\code{return\_obj}] return spatial correlation object (spatCorObject)
\end{ldescription}
\end{Arguments}
%
\begin{Value}
spatCorObject or cluster results
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    clusterSpatialCorGenes(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{colSums\_giotto}{colSums\_giotto}{colSums.Rul.giotto}
%
\begin{Description}\relax
colSums\_giotto
\end{Description}
%
\begin{Usage}
\begin{verbatim}
colSums_giotto(mymatrix)
\end{verbatim}
\end{Usage}
\inputencoding{utf8}
\HeaderA{combCCcom}{combCCcom}{combCCcom}
%
\begin{Description}\relax
Combine spatial and expression based cell-cell communication data.tables
\end{Description}
%
\begin{Usage}
\begin{verbatim}
combCCcom(
  spatialCC,
  exprCC,
  min_lig_nr = 3,
  min_rec_nr = 3,
  min_padj_value = 1,
  min_log2fc = 0,
  min_av_diff = 0
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{spatialCC}] spatial cell-cell communication scores

\item[\code{exprCC}] expression cell-cell communication scores

\item[\code{min\_lig\_nr}] minimum number of ligand cells

\item[\code{min\_rec\_nr}] minimum number of receptor cells

\item[\code{min\_padj\_value}] minimum adjusted p-value

\item[\code{min\_log2fc}] minimum log2 fold-change

\item[\code{min\_av\_diff}] minimum average expression difference
\end{ldescription}
\end{Arguments}
%
\begin{Value}
combined data.table with spatial and expression communication data
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    combCCcom(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{combineCellProximityGenes}{combineCellProximityGenes}{combineCellProximityGenes}
%
\begin{Description}\relax
Combine CPG scores in a pairwise manner.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
combineCellProximityGenes(
  cpgObject,
  selected_ints = NULL,
  selected_genes = NULL,
  specific_genes_1 = NULL,
  specific_genes_2 = NULL,
  min_cells = 5,
  min_int_cells = 3,
  min_fdr = 0.05,
  min_spat_diff = 0,
  min_log2_fc = 0.5,
  do_parallel = TRUE,
  cores = NA,
  verbose = T
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{cpgObject}] cell proximity gene score object

\item[\code{selected\_ints}] subset of selected cell-cell interactions (optional)

\item[\code{selected\_genes}] subset of selected genes (optional)

\item[\code{specific\_genes\_1}] specific geneset combo (need to position match specific\_genes\_2)

\item[\code{specific\_genes\_2}] specific geneset combo (need to position match specific\_genes\_1)

\item[\code{min\_cells}] minimum number of target cell type

\item[\code{min\_int\_cells}] minimum number of interacting cell type

\item[\code{min\_fdr}] minimum adjusted p-value

\item[\code{min\_spat\_diff}] minimum absolute spatial expression difference

\item[\code{min\_log2\_fc}] minimum absolute log2 fold-change

\item[\code{do\_parallel}] run calculations in parallel with mclapply

\item[\code{cores}] number of cores to use if do\_parallel = TRUE

\item[\code{verbose}] verbose
\end{ldescription}
\end{Arguments}
%
\begin{Value}
cpgObject that contains the filtered differential gene scores
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    combineCellProximityGenes(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{combineCellProximityGenes\_per\_interaction}{combineCellProximityGenes\_per\_interaction}{combineCellProximityGenes.Rul.per.Rul.interaction}
%
\begin{Description}\relax
Combine CPG scores per interaction
\end{Description}
%
\begin{Usage}
\begin{verbatim}
combineCellProximityGenes_per_interaction(
  cpgObject,
  sel_int,
  selected_genes = NULL,
  specific_genes_1 = NULL,
  specific_genes_2 = NULL,
  min_cells = 5,
  min_int_cells = 3,
  min_fdr = 0.05,
  min_spat_diff = 0,
  min_log2_fc = 0.5
)
\end{verbatim}
\end{Usage}
%
\begin{Examples}
\begin{ExampleCode}
    combineCellProximityGenes_per_interaction()
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{combineCPG}{combineCPG}{combineCPG}
%
\begin{Description}\relax
Combine CPG scores in a pairwise manner.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
combineCPG(
  cpgObject,
  selected_ints = NULL,
  selected_genes = NULL,
  specific_genes_1 = NULL,
  specific_genes_2 = NULL,
  min_cells = 5,
  min_int_cells = 3,
  min_fdr = 0.05,
  min_spat_diff = 0,
  min_log2_fc = 0.5,
  do_parallel = TRUE,
  cores = NA,
  verbose = T
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{cpgObject}] cell proximity gene score object

\item[\code{selected\_ints}] subset of selected cell-cell interactions (optional)

\item[\code{selected\_genes}] subset of selected genes (optional)

\item[\code{specific\_genes\_1}] specific geneset combo (need to position match specific\_genes\_2)

\item[\code{specific\_genes\_2}] specific geneset combo (need to position match specific\_genes\_1)

\item[\code{min\_cells}] minimum number of target cell type

\item[\code{min\_int\_cells}] minimum number of interacting cell type

\item[\code{min\_fdr}] minimum adjusted p-value

\item[\code{min\_spat\_diff}] minimum absolute spatial expression difference

\item[\code{min\_log2\_fc}] minimum absolute log2 fold-change

\item[\code{do\_parallel}] run calculations in parallel with mclapply

\item[\code{cores}] number of cores to use if do\_parallel = TRUE

\item[\code{verbose}] verbose
\end{ldescription}
\end{Arguments}
%
\begin{Value}
cpgObject that contains the filtered differential gene scores
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    combineCPG(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{combineMetadata}{combineMetadata}{combineMetadata}
%
\begin{Description}\relax
This function combines the cell metadata with spatial locations and enrichment results from createSpatialEnrich
\end{Description}
%
\begin{Usage}
\begin{verbatim}
combineMetadata(gobject, spat_enr_names = NULL)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] Giotto object

\item[\code{spat\_enr\_names}] names of spatial enrichment results to include
\end{ldescription}
\end{Arguments}
%
\begin{Value}
Extended cell metadata in data.table format.
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    combineMetadata(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{convertEnsemblToGeneSymbol}{convertEnsemblToGeneSymbol}{convertEnsemblToGeneSymbol}
%
\begin{Description}\relax
This function convert ensembl gene IDs from a matrix to official gene symbols
\end{Description}
%
\begin{Usage}
\begin{verbatim}
convertEnsemblToGeneSymbol(matrix, species = c("mouse", "human"))
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{matrix}] an expression matrix with ensembl gene IDs as rownames

\item[\code{species}] species to use for gene symbol conversion
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
This function requires that the biomaRt library is installed
\end{Details}
%
\begin{Value}
expression matrix with gene symbols as rownames
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    convertEnsemblToGeneSymbol(matrix)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{convert\_to\_full\_spatial\_network}{convert\_to\_full\_spatial\_network}{convert.Rul.to.Rul.full.Rul.spatial.Rul.network}
%
\begin{Description}\relax
convert to a full spatial network
\end{Description}
%
\begin{Usage}
\begin{verbatim}
convert_to_full_spatial_network(reduced_spatial_network_DT)
\end{verbatim}
\end{Usage}
\inputencoding{utf8}
\HeaderA{convert\_to\_reduced\_spatial\_network}{convert\_to\_reduced\_spatial\_network}{convert.Rul.to.Rul.reduced.Rul.spatial.Rul.network}
%
\begin{Description}\relax
convert to a reduced spatial network
\end{Description}
%
\begin{Usage}
\begin{verbatim}
convert_to_reduced_spatial_network(full_spatial_network_DT)
\end{verbatim}
\end{Usage}
\inputencoding{utf8}
\HeaderA{createCrossSection}{createCrossSection}{createCrossSection}
%
\begin{Description}\relax
Create a virtual 2D cross section.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
createCrossSection(
  gobject,
  name = "cross_section",
  spatial_network_name = "Delaunay_network",
  thickness_unit = c("cell", "natural"),
  slice_thickness = 2,
  cell_distance_estimate_method = "mean",
  extend_ratio = 0.2,
  method = c("equation", "3 points", "point and norm vector",
    "point and two plane vectors"),
  equation = NULL,
  point1 = NULL,
  point2 = NULL,
  point3 = NULL,
  normVector = NULL,
  planeVector1 = NULL,
  planeVector2 = NULL,
  mesh_grid_n = 20,
  return_gobject = TRUE
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{name}] name of cress section object. (default = cross\_sectino)

\item[\code{spatial\_network\_name}] name of spatial network object. (default = Delaunay\_network)

\item[\code{thickness\_unit}] unit of the virtual section thickness. If "cell", average size of the observed cells is used as length unit. If "natural", the unit of cell location coordinates is used.(default = cell)

\item[\code{cell\_distance\_estimate\_method}] method to estimate average distance between neighobring cells. (default = mean)

\item[\code{extend\_ratio}] deciding the span of the cross section meshgrid, as a ratio of extension compared to the borders of the vitural tissue section. (default = 0.2)

\item[\code{method}] method to define the cross section plane.
If equation, the plane is defined by a four element numerical vector (equation) in the form of c(A,B,C,D), corresponding to a plane with equation Ax+By+Cz=D.
If 3 points, the plane is define by the coordinates of 3 points, as given by point1, point2, and point3.
If point and norm vector, the plane is defined by the coordinates of one point (point1) in the plane and the coordinates of one norm vector (normVector) to the plane.
If point and two plane vector, the plane is defined by the coordinates of one point (point1) in the plane and the coordinates of two vectors (planeVector1, planeVector2) in the plane.
(default = equation)

\item[\code{equation}] equation required by method "equation".equations needs to be a numerical vector of length 4, in the form of c(A,B,C,D), which defines plane Ax+By+Cz=D.

\item[\code{point1}] coordinates of the first point required by method "3 points","point and norm vector", and "point and two plane vectors".

\item[\code{point2}] coordinates of the second point required by method "3 points"

\item[\code{point3}] coordinates of the third point required by method "3 points"

\item[\code{normVector}] coordinates of the norm vector required by method "point and norm vector"

\item[\code{planeVector1}] coordinates of the first plane vector required by method "point and two plane vectors"

\item[\code{planeVector2}] coordinates of the second plane vector required by method "point and two plane vectors"

\item[\code{mesh\_grid\_n}] numer of meshgrid lines to generate along both directions for the cross section plane.

\item[\code{return\_gobject}] boolean: return giotto object (default = TRUE)
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Creates a virtual 2D cross section object for a given spatial network object. The users need to provide the definition of the cross section plane (see method).
\end{Details}
%
\begin{Value}
giotto object with updated spatial network slot
\end{Value}
\inputencoding{utf8}
\HeaderA{createGiottoInstructions}{createGiottoInstructions}{createGiottoInstructions}
%
\begin{Description}\relax
Function to set global instructions for giotto functions
\end{Description}
%
\begin{Usage}
\begin{verbatim}
createGiottoInstructions(
  python_path = NULL,
  show_plot = NULL,
  return_plot = NULL,
  save_plot = NULL,
  save_dir = NULL,
  plot_format = NULL,
  dpi = NULL,
  units = NULL,
  height = NULL,
  width = NULL
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{python\_path}] path to python binary to use

\item[\code{show\_plot}] print plot to console, default = TRUE

\item[\code{return\_plot}] return plot as object, default = TRUE

\item[\code{save\_plot}] automatically save plot, dafault = FALSE

\item[\code{save\_dir}] path to directory where to save plots

\item[\code{dpi}] resolution for raster images

\item[\code{height}] height of plots

\item[\code{width}] width of  plots
\end{ldescription}
\end{Arguments}
%
\begin{Value}
named vector with giotto instructions
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    createGiottoInstructions()
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{createGiottoObject}{create Giotto object}{createGiottoObject}
\keyword{giotto}{createGiottoObject}
%
\begin{Description}\relax
Function to create a giotto object
\end{Description}
%
\begin{Usage}
\begin{verbatim}
createGiottoObject(
  raw_exprs,
  spatial_locs = NULL,
  norm_expr = NULL,
  norm_scaled_expr = NULL,
  custom_expr = NULL,
  cell_metadata = NULL,
  gene_metadata = NULL,
  spatial_network = NULL,
  spatial_network_name = NULL,
  spatial_grid = NULL,
  spatial_grid_name = NULL,
  spatial_enrichment = NULL,
  spatial_enrichment_name = NULL,
  dimension_reduction = NULL,
  nn_network = NULL,
  offset_file = NULL,
  instructions = NULL,
  cores = NA
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{raw\_exprs}] matrix with raw expression counts [required]

\item[\code{spatial\_locs}] data.table or data.frame with coordinates for cell centroids

\item[\code{norm\_expr}] normalized expression values

\item[\code{norm\_scaled\_expr}] scaled expression values

\item[\code{custom\_expr}] custom expression values

\item[\code{cell\_metadata}] cell annotation metadata

\item[\code{gene\_metadata}] gene annotation metadata

\item[\code{spatial\_network}] list of spatial network(s)

\item[\code{spatial\_network\_name}] list of spatial network name(s)

\item[\code{spatial\_grid}] list of spatial grid(s)

\item[\code{spatial\_grid\_name}] list of spatial grid name(s)

\item[\code{spatial\_enrichment}] list of spatial enrichment score(s) for each spatial region

\item[\code{spatial\_enrichment\_name}] list of spatial enrichment name(s)

\item[\code{dimension\_reduction}] list of dimension reduction(s)

\item[\code{nn\_network}] list of nearest neighbor network(s)

\item[\code{offset\_file}] file used to stitch fields together (optional)

\item[\code{instructions}] list of instructions or output result from createGiottoInstructions

\item[\code{cores}] how many cores or threads to use to read data if paths are provided
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
[\strong{Requirements}] To create a giotto object you need to provide at least a matrix with genes as
row names and cells as column names. This matrix can be provided as a base matrix, sparse Matrix, data.frame,
data.table or as a path to any of those. 
To include spatial information about cells (or regions) you need to provide a matrix, data.table or data.frame (or path to them)
with coordinates for all spatial dimensions. This can be 2D (x and y) or 3D (x, y, x).
The row order for the cell coordinates should be the same as the column order for the provided expression data.

[\strong{Instructions}] Additionally an instruction file, generated manually or with \code{\LinkA{createGiottoInstructions}{createGiottoInstructions}}
can be provided to instructions, if not a default instruction file will be created
for the Giotto object.

[\strong{Multiple fields}] In case a dataset consists of multiple fields, like seqFISH+ for example,
an offset file can be provided to stitch the different fields together. \code{\LinkA{stitchFieldCoordinates}{stitchFieldCoordinates}}
can be used to generate such an offset file.

[\strong{Processed data}] Processed count data, such as normalized data, can be provided using
one of the different expression slots (norm\_expr, norm\_scaled\_expr, custom\_expr).

[\strong{Metadata}] Cell and gene metadata can be provided using the cell and gene metadata slots.
This data can also be added afterwards using the \code{\LinkA{addGeneMetadata}{addGeneMetadata}} or \code{\LinkA{addCellMetadata}{addCellMetadata}} functions.

[\strong{Other information}] Additional information can be provided through the appropriate slots:
\begin{itemize}

\item{} spatial networks
\item{} spatial girds
\item{} spatial enrichments
\item{} dimensions reductions
\item{} nearest neighbours networks

\end{itemize}

\end{Details}
%
\begin{Value}
giotto object
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    createGiottoObject(raw_exprs, spatial_locs)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{createHeatmap\_DT}{createHeatmap\_DT}{createHeatmap.Rul.DT}
%
\begin{Description}\relax
creates order for clusters
\end{Description}
%
\begin{Usage}
\begin{verbatim}
createHeatmap_DT(
  gobject,
  expression_values = c("normalized", "scaled", "custom"),
  genes,
  cluster_column = NULL,
  cluster_order = c("size", "correlation", "custom"),
  cluster_custom_order = NULL,
  cluster_cor_method = "pearson",
  cluster_hclust_method = "ward.D",
  gene_order = c("correlation", "custom"),
  gene_custom_order = NULL,
  gene_cor_method = "pearson",
  gene_hclust_method = "complete"
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{expression\_values}] expression values to use

\item[\code{genes}] genes to use

\item[\code{cluster\_column}] name of column to use for clusters

\item[\code{cluster\_order}] method to determine cluster order

\item[\code{cluster\_custom\_order}] custom order for clusters

\item[\code{cluster\_cor\_method}] method for cluster correlation

\item[\code{cluster\_hclust\_method}] method for hierarchical clustering of clusters

\item[\code{gene\_order}] method to determine gene order

\item[\code{gene\_custom\_order}] custom order for genes

\item[\code{gene\_cor\_method}] method for gene correlation

\item[\code{gene\_hclust\_method}] method for hierarchical clustering of genes
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Creates input data.tables for plotHeatmap function.
\end{Details}
%
\begin{Value}
list
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    createHeatmap_DT(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{createMetagenes}{createMetagenes}{createMetagenes}
%
\begin{Description}\relax
This function creates an average metagene for gene clusters.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
createMetagenes(
  gobject,
  expression_values = c("normalized", "scaled", "custom"),
  gene_clusters,
  name = "metagene",
  return_gobject = TRUE
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] Giotto object

\item[\code{expression\_values}] expression values to use

\item[\code{gene\_clusters}] numerical vector with genes as names

\item[\code{name}] name of the metagene results

\item[\code{return\_gobject}] return giotto object
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
An example for the 'gene\_clusters' could be like this:
cluster\_vector = c(1, 1, 2, 2); names(cluster\_vector) = c('geneA', 'geneB', 'geneC', 'geneD')
\end{Details}
%
\begin{Value}
giotto object
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    createMetagenes(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{createNearestNetwork}{createNearestNetwork}{createNearestNetwork}
%
\begin{Description}\relax
create a nearest neighbour (NN) network
\end{Description}
%
\begin{Usage}
\begin{verbatim}
createNearestNetwork(
  gobject,
  type = c("sNN", "kNN"),
  dim_reduction_to_use = "pca",
  dim_reduction_name = "pca",
  dimensions_to_use = 1:10,
  genes_to_use = NULL,
  expression_values = c("normalized", "scaled", "custom"),
  name = "sNN.pca",
  return_gobject = TRUE,
  k = 30,
  minimum_shared = 5,
  top_shared = 3,
  verbose = T,
  ...
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{type}] sNN or kNN

\item[\code{dim\_reduction\_to\_use}] dimension reduction method to use

\item[\code{dim\_reduction\_name}] name of dimension reduction set to use

\item[\code{dimensions\_to\_use}] number of dimensions to use as input

\item[\code{genes\_to\_use}] if dim\_reduction\_to\_use = NULL, which genes to use

\item[\code{expression\_values}] expression values to use

\item[\code{name}] arbitrary name for NN network

\item[\code{return\_gobject}] boolean: return giotto object (default = TRUE)

\item[\code{k}] number of k neighbors to use

\item[\code{minimum\_shared}] minimum shared neighbors

\item[\code{top\_shared}] keep at ...

\item[\code{verbose}] be verbose

\item[\code{...}] additional parameters for kNN and sNN functions from dbscan
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
This function creates a k-nearest neighbour (kNN) or shared nearest neighbour (sNN) network
based on the provided dimension reduction space. To run it directly on the gene expression matrix
set \emph{dim\_reduction\_to\_use = NULL}.

See also \code{\LinkA{kNN}{kNN}} and \code{\LinkA{sNN}{sNN}} for more information about
how the networks are created.

Output for kNN:
\begin{itemize}

\item{} from: cell\_ID for source cell
\item{} to: cell\_ID for target cell
\item{} distance: distance between cells
\item{} weight: weight = 1/(1 + distance)

\end{itemize}


Output for sNN:
\begin{itemize}

\item{} from: cell\_ID for source cell
\item{} to: cell\_ID for target cell
\item{} distance: distance between cells
\item{} weight: 1/(1 + distance)
\item{} shared: number of shared neighbours
\item{} rank: ranking of pairwise cell neighbours

\end{itemize}

For sNN networks two additional parameters can be set:
\begin{itemize}

\item{} minimum\_shared: minimum number of shared neighbours needed
\item{} top\_shared: keep this number of the top shared neighbours, irrespective of minimum\_shared setting

\end{itemize}

\end{Details}
%
\begin{Value}
giotto object with updated NN network
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    createNearestNetwork(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{createSpatialDelaunayNetwork}{createSpatialDelaunayNetwork}{createSpatialDelaunayNetwork}
%
\begin{Description}\relax
Create a spatial Delaunay network based on cell centroid physical distances.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
createSpatialDelaunayNetwork(
  gobject,
  method = c("delaunayn_geometry", "RTriangle", "deldir"),
  dimensions = "all",
  name = "delaunay_network",
  maximum_distance = "auto",
  minimum_k = 0,
  options = "Pp",
  Y = TRUE,
  j = TRUE,
  S = 0,
  verbose = T,
  return_gobject = TRUE,
  ...
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{dimensions}] which spatial dimensions to use (default = all)

\item[\code{name}] name for spatial network (default = 'delaunay\_network')

\item[\code{maximum\_distance}] distance cuttof for Delaunay neighbors to consider. If "auto", "upper wisker" value of the distance vector between neighbors is used; see the boxplotgraphics documentation for more details.(default = "auto")

\item[\code{minimum\_k}] minimum number of neigbhours if maximum\_distance != NULL

\item[\code{options}] (geometry) String containing extra control options for the underlying Qhull command; see the Qhull documentation (../doc/qhull/html/qdelaun.html) for the available options. (default = 'Pp', do not report precision problems)

\item[\code{Y}] (RTriangle) If TRUE prohibits the insertion of Steiner points on the mesh boundary.

\item[\code{j}] (RTriangle) If TRUE jettisons vertices that are not part of the final triangulation from the output.

\item[\code{S}] (RTriangle) Specifies the maximum number of added Steiner points.

\item[\code{verbose}] verbose

\item[\code{return\_gobject}] boolean: return giotto object (default = TRUE)

\item[\code{...}] Other parameters of the \code{\LinkA{triangulate}{triangulate}} function
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Creates a spatial Delaunay network as explained in \code{\LinkA{delaunayn}{delaunayn}} (default), \code{\LinkA{deldir}{deldir}}, or \code{\LinkA{triangulate}{triangulate}}.
\end{Details}
%
\begin{Value}
giotto object with updated spatial network slot
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    createSpatialDelaunayNetwork(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{createSpatialEnrich}{createSpatialEnrich}{createSpatialEnrich}
%
\begin{Description}\relax
Function to calculate gene signature enrichment scores per spatial position using a hypergeometric test.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
createSpatialEnrich(
  gobject,
  enrich_method = c("PAGE", "rank", "hypergeometric"),
  sign_matrix,
  expression_values = c("normalized", "scaled", "custom"),
  reverse_log_scale = TRUE,
  logbase = 2,
  p_value = TRUE,
  n_genes = 100,
  n_times = 1000,
  top_percentage = 5,
  output_enrichment = c("original", "zscore"),
  name = "PAGE",
  return_gobject = TRUE
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] Giotto object

\item[\code{enrich\_method}] method for gene signature enrichment calculation

\item[\code{sign\_matrix}] Matrix of signature genes for each cell type / process

\item[\code{expression\_values}] expression values to use

\item[\code{reverse\_log\_scale}] reverse expression values from log scale

\item[\code{logbase}] log base to use if reverse\_log\_scale = TRUE

\item[\code{p\_value}] calculate p-value (default = FALSE)

\item[\code{n\_times}] (page/rank) number of permutation iterations to calculate p-value

\item[\code{top\_percentage}] (hyper) percentage of cells that will be considered to have gene expression with matrix binarization

\item[\code{output\_enrichment}] how to return enrichment output

\item[\code{name}] to give to spatial enrichment results, default = PAGE

\item[\code{return\_gobject}] return giotto object
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
For details see the individual functions:
\begin{itemize}

\item{} PAGE: \code{\LinkA{PAGEEnrich}{PAGEEnrich}}
\item{} PAGE: \code{\LinkA{rankEnrich}{rankEnrich}}
\item{} PAGE: \code{\LinkA{hyperGeometricEnrich}{hyperGeometricEnrich}}

\end{itemize}

\end{Details}
%
\begin{Value}
Giotto object or enrichment results if return\_gobject = FALSE
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    createSpatialEnrich(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{createSpatialGrid}{createSpatialGrid}{createSpatialGrid}
%
\begin{Description}\relax
Create a spatial grid.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
createSpatialGrid(
  gobject,
  sdimx_stepsize = NULL,
  sdimy_stepsize = NULL,
  sdimz_stepsize = NULL,
  minimum_padding = 1,
  name = "spatial_grid",
  return_gobject = TRUE
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{sdimx\_stepsize}] stepsize along the x-axis

\item[\code{sdimy\_stepsize}] stepsize along the y-axis

\item[\code{sdimz\_stepsize}] stepsize along the z-axis

\item[\code{minimum\_padding}] minimum padding on the edges

\item[\code{name}] name for spatial grid (default = 'spatial\_grid')

\item[\code{return\_gobject}] boolean: return giotto object (default = TRUE)
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Creates a spatial grid with defined x, y (and z) dimensions.
The dimension units are based on the provided spatial location units.
\end{Details}
%
\begin{Value}
giotto object with updated spatial grid slot
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    createSpatialGrid(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{createSpatialGrid\_2D}{createSpatialGrid\_2D}{createSpatialGrid.Rul.2D}
%
\begin{Description}\relax
create a spatial grid for 2D spatial data.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
createSpatialGrid_2D(
  gobject,
  sdimx_stepsize = NULL,
  sdimy_stepsize = NULL,
  minimum_padding = 1,
  name = "spatial_grid",
  return_gobject = TRUE
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{sdimx\_stepsize}] stepsize along the x-axis

\item[\code{sdimy\_stepsize}] stepsize along the y-axis

\item[\code{minimum\_padding}] minimum padding on the edges

\item[\code{name}] name for spatial grid (default = 'spatial\_grid')

\item[\code{return\_gobject}] boolean: return giotto object (default = TRUE)
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Creates a spatial grid with defined x, y (and z) dimensions.
The dimension units are based on the provided spatial location units.
\end{Details}
%
\begin{Value}
giotto object with updated spatial grid slot
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    createSpatialGrid_2D(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{createSpatialGrid\_3D}{createSpatialGrid\_3D}{createSpatialGrid.Rul.3D}
%
\begin{Description}\relax
Create a spatial grid for 3D spatial data.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
createSpatialGrid_3D(
  gobject,
  sdimx_stepsize = NULL,
  sdimy_stepsize = NULL,
  sdimz_stepsize = NULL,
  minimum_padding = 1,
  name = "spatial_grid",
  return_gobject = TRUE
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{sdimx\_stepsize}] stepsize along the x-axis

\item[\code{sdimy\_stepsize}] stepsize along the y-axis

\item[\code{sdimz\_stepsize}] stepsize along the z-axis

\item[\code{minimum\_padding}] minimum padding on the edges

\item[\code{name}] name for spatial grid (default = 'spatial\_grid')

\item[\code{return\_gobject}] boolean: return giotto object (default = TRUE)
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Creates a spatial grid with defined x, y (and z) dimensions.
The dimension units are based on the provided spatial location units.
\end{Details}
%
\begin{Value}
giotto object with updated spatial grid slot
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    createSpatialGrid_3D(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{createSpatialKNNnetwork}{createSpatialKNNnetwork}{createSpatialKNNnetwork}
%
\begin{Description}\relax
Create a spatial knn network.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
createSpatialKNNnetwork(
  gobject,
  method = "dbscan",
  dimensions = "all",
  name = "knn_network",
  k = 4,
  maximum_distance = NULL,
  minimum_k = 0,
  verbose = F,
  return_gobject = TRUE,
  ...
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{method}] method to create kNN network

\item[\code{dimensions}] which spatial dimensions to use (default = all)

\item[\code{name}] name for spatial network (default = 'spatial\_network')

\item[\code{k}] number of nearest neighbors based on physical distance

\item[\code{maximum\_distance}] distance cuttof for nearest neighbors to consider for kNN network

\item[\code{minimum\_k}] minimum nearest neigbhours if maximum\_distance != NULL

\item[\code{verbose}] verbose

\item[\code{return\_gobject}] boolean: return giotto object (default = TRUE)
\end{ldescription}
\end{Arguments}
%
\begin{Value}
giotto object with updated spatial network slot

\strong{dimensions: } default = 'all' which takes all possible dimensions.
Alternatively you can provide a character vector that specififies the spatial dimensions to use, e.g. c("sdimx', "sdimy")
or a numerical vector, e.g. 2:3

\strong{maximum\_distance: } to create a network based on maximum distance only, you also need to set k to a very high value, e.g. k = 100
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    createSpatialKNNnetwork(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{createSpatialNetwork}{createSpatialNetwork}{createSpatialNetwork}
%
\begin{Description}\relax
Create a spatial network based on cell centroid physical distances.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
createSpatialNetwork(
  gobject,
  name = NULL,
  dimensions = "all",
  method = c("Delaunay", "kNN"),
  delaunay_method = c("delaunayn_geometry", "RTriangle", "deldir"),
  maximum_distance_delaunay = "auto",
  options = "Pp",
  Y = TRUE,
  j = TRUE,
  S = 0,
  minimum_k = 0,
  knn_method = "dbscan",
  k = 4,
  maximum_distance_knn = NULL,
  verbose = F,
  return_gobject = TRUE,
  ...
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{name}] name for spatial network (default = 'spatial\_network')

\item[\code{dimensions}] which spatial dimensions to use (default = all)

\item[\code{method}] which method to use to create a spatial network. (default = Delaunay)

\item[\code{delaunay\_method}] Delaunay method to use

\item[\code{maximum\_distance\_delaunay}] distance cuttof for nearest neighbors to consider for Delaunay network

\item[\code{options}] (geometry) String containing extra control options for the underlying Qhull command; see the Qhull documentation (../doc/qhull/html/qdelaun.html) for the available options. (default = 'Pp', do not report precision problems)

\item[\code{Y}] (RTriangle) If TRUE prohibits the insertion of Steiner points on the mesh boundary.

\item[\code{j}] (RTriangle) If TRUE jettisons vertices that are not part of the final triangulation from the output.

\item[\code{S}] (RTriangle) Specifies the maximum number of added Steiner points.

\item[\code{minimum\_k}] minimum nearest neigbhours if maximum\_distance != NULL

\item[\code{knn\_method}] method to create kNN network

\item[\code{k}] number of nearest neighbors based on physical distance

\item[\code{maximum\_distance\_knn}] distance cuttof for nearest neighbors to consider for kNN network

\item[\code{verbose}] verbose

\item[\code{return\_gobject}] boolean: return giotto object (default = TRUE)
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Creates a spatial network connecting single-cells based on their physical distance to each other.
For Delaunay method, neighbors will be decided by delaunay triangulation and a maximum distance criteria. For kNN method, number of neighbors can be determined by k, or maximum distance from each cell with or without
setting a minimum k for each cell.

\strong{dimensions: } default = 'all' which takes all possible dimensions.
Alternatively you can provide a character vector that specififies the spatial dimensions to use, e.g. c("sdimx', "sdimy")
or a numerical vector, e.g. 2:3
\end{Details}
%
\begin{Value}
giotto object with updated spatial network slot
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    createSpatialNetwork(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{create\_2d\_mesh\_grid\_line\_obj}{create\_2d\_mesh\_grid\_line\_obj}{create.Rul.2d.Rul.mesh.Rul.grid.Rul.line.Rul.obj}
%
\begin{Description}\relax
create 2d mesh grid line object
\end{Description}
%
\begin{Usage}
\begin{verbatim}
create_2d_mesh_grid_line_obj(x_min, x_max, y_min, y_max, mesh_grid_n)
\end{verbatim}
\end{Usage}
\inputencoding{utf8}
\HeaderA{create\_average\_detection\_DT}{create\_average\_detection\_DT}{create.Rul.average.Rul.detection.Rul.DT}
%
\begin{Description}\relax
calculates average gene detection for a cell metadata factor (e.g. cluster)
\end{Description}
%
\begin{Usage}
\begin{verbatim}
create_average_detection_DT(
  gobject,
  meta_data_name,
  expression_values = c("normalized", "scaled", "custom"),
  detection_threshold = 0
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{meta\_data\_name}] name of metadata column to use

\item[\code{expression\_values}] which expression values to use

\item[\code{detection\_threshold}] detection threshold to consider a gene detected
\end{ldescription}
\end{Arguments}
%
\begin{Value}
data.table with average gene epression values for each factor
\end{Value}
\inputencoding{utf8}
\HeaderA{create\_average\_DT}{create\_average\_DT}{create.Rul.average.Rul.DT}
%
\begin{Description}\relax
calculates average gene expression for a cell metadata factor (e.g. cluster)
\end{Description}
%
\begin{Usage}
\begin{verbatim}
create_average_DT(
  gobject,
  meta_data_name,
  expression_values = c("normalized", "scaled", "custom")
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{meta\_data\_name}] name of metadata column to use

\item[\code{expression\_values}] which expression values to use
\end{ldescription}
\end{Arguments}
%
\begin{Value}
data.table with average gene epression values for each factor
\end{Value}
\inputencoding{utf8}
\HeaderA{create\_cell\_type\_random\_cell\_IDs}{create\_cell\_type\_random\_cell\_IDs}{create.Rul.cell.Rul.type.Rul.random.Rul.cell.Rul.IDs}
%
\begin{Description}\relax
creates randomized cell ids within a selection of cell types
\end{Description}
%
\begin{Usage}
\begin{verbatim}
create_cell_type_random_cell_IDs(
  gobject,
  cluster_column = "cell_types",
  needed_cell_types
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object to use

\item[\code{cluster\_column}] cluster column with cell type information

\item[\code{needed\_cell\_types}] vector of cell type names for which a random id will be found
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Details will follow.
\end{Details}
%
\begin{Value}
list of randomly sampled cell ids with same cell type composition
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    create_cell_type_random_cell_IDs(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{create\_cluster\_matrix}{create\_cluster\_matrix}{create.Rul.cluster.Rul.matrix}
%
\begin{Description}\relax
creates aggregated matrix for a given clustering
\end{Description}
%
\begin{Usage}
\begin{verbatim}
create_cluster_matrix(
  gobject,
  expression_values = c("normalized", "scaled", "custom"),
  cluster_column,
  gene_subset = NULL
)
\end{verbatim}
\end{Usage}
%
\begin{Examples}
\begin{ExampleCode}
    create_cluster_matrix(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{create\_crossSection\_object}{create\_crossSection\_object}{create.Rul.crossSection.Rul.object}
%
\begin{Description}\relax
create a crossSection object
\end{Description}
%
\begin{Usage}
\begin{verbatim}
create_crossSection_object(
  name = NULL,
  method = NULL,
  thickness_unit = NULL,
  slice_thickness = NULL,
  plane_equation = NULL,
  mesh_grid_n = NULL,
  mesh_obj = NULL,
  cell_subset = NULL,
  cell_subset_spatial_locations = NULL,
  cell_subset_projection_locations = NULL,
  cell_subset_projection_PCA = NULL,
  cell_subset_projection_coords = NULL
)
\end{verbatim}
\end{Usage}
\inputencoding{utf8}
\HeaderA{create\_delaunayNetwork2D}{create\_delaunayNetwork2D}{create.Rul.delaunayNetwork2D}
%
\begin{Description}\relax
Create a spatial Delaunay network.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
create_delaunayNetwork2D(
  gobject,
  method = c("delaunayn_geometry", "RTriangle", "deldir"),
  sdimx = "sdimx",
  sdimy = "sdimy",
  name = "delaunay_network",
  maximum_distance = "auto",
  minimum_k = 0,
  options = "Pp",
  Y = TRUE,
  j = TRUE,
  S = 0,
  verbose = T,
  return_gobject = TRUE,
  ...
)
\end{verbatim}
\end{Usage}
%
\begin{Examples}
\begin{ExampleCode}
    create_delaunayNetwork2D(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{create\_delaunayNetwork3D}{create\_delaunayNetwork3D}{create.Rul.delaunayNetwork3D}
%
\begin{Description}\relax
Create a spatial Delaunay network.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
create_delaunayNetwork3D(
  gobject,
  method = "delaunayn_geometry",
  sdimx = "sdimx",
  sdimy = "sdimy",
  sdimz = "sdimz",
  name = "delaunay_network_3D",
  maximum_distance = "auto",
  minimum_k = 0,
  options = "Pp",
  return_gobject = TRUE,
  ...
)
\end{verbatim}
\end{Usage}
%
\begin{Examples}
\begin{ExampleCode}
    create_delaunayNetwork3D(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{create\_delaunayNetwork\_deldir}{create\_delaunayNetwork\_deldir}{create.Rul.delaunayNetwork.Rul.deldir}
%
\begin{Description}\relax
Create a spatial Delaunay network.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
create_delaunayNetwork_deldir(
  spatial_locations,
  sdimx = "sdimx",
  sdimy = "sdimy",
  ...
)
\end{verbatim}
\end{Usage}
%
\begin{Examples}
\begin{ExampleCode}
    create_delaunayNetwork_deldir(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{create\_delaunayNetwork\_geometry}{create\_delaunayNetwork\_geometry}{create.Rul.delaunayNetwork.Rul.geometry}
%
\begin{Description}\relax
Create a spatial Delaunay network.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
create_delaunayNetwork_geometry(
  spatial_locations,
  sdimx = "sdimx",
  sdimy = "sdimy",
  options = "Pp",
  ...
)
\end{verbatim}
\end{Usage}
%
\begin{Examples}
\begin{ExampleCode}
    create_delaunayNetwork_geometry(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{create\_delaunayNetwork\_geometry\_3D}{create\_delaunayNetwork\_geometry\_3D}{create.Rul.delaunayNetwork.Rul.geometry.Rul.3D}
%
\begin{Description}\relax
Create a spatial Delaunay network.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
create_delaunayNetwork_geometry_3D(
  spatial_locations,
  sdimx = "sdimx",
  sdimy = "sdimy",
  sdimz = "sdimz",
  options = options,
  ...
)
\end{verbatim}
\end{Usage}
%
\begin{Examples}
\begin{ExampleCode}
    create_delaunayNetwork_geometry_3D(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{create\_delaunayNetwork\_RTriangle}{create\_delaunayNetwork\_RTriangle}{create.Rul.delaunayNetwork.Rul.RTriangle}
%
\begin{Description}\relax
Create a spatial Delaunay network.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
create_delaunayNetwork_RTriangle(
  spatial_locations,
  sdimx = "sdimx",
  sdimy = "sdimy",
  Y = TRUE,
  j = TRUE,
  S = 0,
  ...
)
\end{verbatim}
\end{Usage}
%
\begin{Examples}
\begin{ExampleCode}
    create_delaunayNetwork_RTriangle(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{create\_dimObject}{create\_dimObject}{create.Rul.dimObject}
%
\begin{Description}\relax
Creates an object that stores a dimension reduction output
\end{Description}
%
\begin{Usage}
\begin{verbatim}
create_dimObject(
  name = "test",
  reduction_method = NULL,
  coordinates = NULL,
  misc = NULL,
  my_rownames = NULL
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{name}] arbitrary name for object

\item[\code{reduction\_method}] method used to reduce dimensions

\item[\code{coordinates}] accepts the coordinates after dimension reduction

\item[\code{misc}] any additional information will be added to this slot
\end{ldescription}
\end{Arguments}
%
\begin{Value}
number of distinct colors
\end{Value}
\inputencoding{utf8}
\HeaderA{create\_KNNnetwork\_dbscan}{create\_KNNnetwork\_dbscan}{create.Rul.KNNnetwork.Rul.dbscan}
%
\begin{Description}\relax
Create a spatial knn network.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
create_KNNnetwork_dbscan(
  spatial_locations,
  sdimx = "sdimx",
  sdimy = "sdimy",
  sdimz = "sdimz",
  k = 4,
  ...
)
\end{verbatim}
\end{Usage}
%
\begin{Examples}
\begin{ExampleCode}
    create_KNNnetwork_dbscan(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{create\_mesh\_grid\_lines}{create\_mesh\_grid\_lines}{create.Rul.mesh.Rul.grid.Rul.lines}
%
\begin{Description}\relax
create mesh grid lines for cross section
\end{Description}
%
\begin{Usage}
\begin{verbatim}
create_mesh_grid_lines(
  cell_subset_projection_locations,
  extend_ratio,
  mesh_grid_n
)
\end{verbatim}
\end{Usage}
\inputencoding{utf8}
\HeaderA{create\_spatialNetworkObject}{create\_spatialNetworkObject}{create.Rul.spatialNetworkObject}
%
\begin{Description}\relax
creates a spatial network object to store the created spatial network and additional information
\end{Description}
%
\begin{Usage}
\begin{verbatim}
create_spatialNetworkObject(
  name = NULL,
  method = NULL,
  parameters = NULL,
  outputObj = NULL,
  networkDT = NULL,
  cellShapeObj = NULL,
  networkDT_before_filter = NULL,
  crossSectionObjects = NULL,
  misc = NULL
)
\end{verbatim}
\end{Usage}
\inputencoding{utf8}
\HeaderA{crossSectionGenePlot}{crossSectionGenePlot}{crossSectionGenePlot}
%
\begin{Description}\relax
Visualize cells and gene expression in a virtual cross section according to spatial coordinates
\end{Description}
%
\begin{Usage}
\begin{verbatim}
crossSectionGenePlot(
  gobject = NULL,
  crossSection_obj = NULL,
  name = NULL,
  spatial_network_name = "Delaunay_network",
  expression_values = c("normalized", "scaled", "custom"),
  genes,
  genes_high_color = "red",
  genes_mid_color = "white",
  genes_low_color = "darkblue",
  show_network = F,
  network_color = NULL,
  edge_alpha = NULL,
  show_grid = F,
  grid_color = NULL,
  spatial_grid_name = "spatial_grid",
  midpoint = 0,
  scale_alpha_with_expression = FALSE,
  point_shape = c("border", "no_border"),
  point_size = 1,
  point_border_col = "black",
  point_border_stroke = 0.1,
  show_legend = T,
  legend_text = 8,
  background_color = "white",
  axis_text = 8,
  axis_title = 8,
  cow_n_col = 2,
  cow_rel_h = 1,
  cow_rel_w = 1,
  cow_align = "h",
  show_plot = NA,
  return_plot = NA,
  save_plot = NA,
  save_param = list(),
  default_save_name = "crossSectionGenePlot"
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{name}] name of virtual cross section to use

\item[\code{spatial\_network\_name}] name of spatial network to use

\item[\code{expression\_values}] gene expression values to use

\item[\code{genes}] genes to show

\item[\code{genes\_high\_color}] color represents high gene expression

\item[\code{genes\_mid\_color}] color represents middle gene expression

\item[\code{genes\_low\_color}] color represents low gene expression

\item[\code{show\_network}] show underlying spatial network

\item[\code{network\_color}] color of spatial network

\item[\code{show\_grid}] show spatial grid

\item[\code{grid\_color}] color of spatial grid

\item[\code{spatial\_grid\_name}] name of spatial grid to use

\item[\code{midpoint}] expression midpoint

\item[\code{scale\_alpha\_with\_expression}] scale expression with ggplot alpha parameter

\item[\code{point\_shape}] point with border or not (border or no\_border)

\item[\code{point\_size}] size of point (cell)

\item[\code{point\_border\_col}] color of border around points

\item[\code{point\_border\_stroke}] stroke size of border around points

\item[\code{show\_legend}] show legend

\item[\code{legend\_text}] size of legend text

\item[\code{background\_color}] color of plot background

\item[\code{axis\_text}] size of axis text

\item[\code{axis\_title}] size of axis title

\item[\code{cow\_n\_col}] cowplot param: how many columns

\item[\code{cow\_rel\_h}] cowplot param: relative height

\item[\code{cow\_rel\_w}] cowplot param: relative width

\item[\code{cow\_align}] cowplot param: how to align

\item[\code{show\_plot}] show plots

\item[\code{return\_plot}] return ggplot object

\item[\code{save\_plot}] directly save the plot [boolean]

\item[\code{save\_param}] list of saving parameters from \code{\LinkA{all\_plots\_save\_function}{all.Rul.plots.Rul.save.Rul.function}}

\item[\code{default\_save\_name}] default save name for saving, don't change, change save\_name in save\_param

\item[\code{...}] parameters for cowplot::save\_plot()
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Description of parameters.
\end{Details}
%
\begin{Value}
ggplot
\end{Value}
%
\begin{SeeAlso}\relax
\code{\LinkA{spatGenePlot3D}{spatGenePlot3D}} and \code{\LinkA{spatGenePlot2D}{spatGenePlot2D}}
\end{SeeAlso}
%
\begin{Examples}
\begin{ExampleCode}
    crossSectionGenePlot(gobject)

\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{crossSectionGenePlot3D}{crossSectionGenePlot3D}{crossSectionGenePlot3D}
%
\begin{Description}\relax
Visualize cells and gene expression in a virtual cross section according to spatial coordinates
\end{Description}
%
\begin{Usage}
\begin{verbatim}
crossSectionGenePlot3D(
  gobject,
  crossSection_obj = NULL,
  name = NULL,
  spatial_network_name = "Delaunay_network",
  expression_values = c("normalized", "scaled", "custom"),
  genes,
  show_network = F,
  network_color = NULL,
  edge_alpha = NULL,
  show_grid = F,
  cluster_column = NULL,
  select_cell_groups = NULL,
  select_cells = NULL,
  show_other_cells = T,
  other_cell_color = alpha("lightgrey", 0),
  other_point_size = 1,
  genes_high_color = "red",
  genes_mid_color = "white",
  genes_low_color = "darkblue",
  spatial_grid_name = "spatial_grid",
  point_size = 2,
  show_legend = T,
  axis_scale = c("cube", "real", "custom"),
  custom_ratio = NULL,
  x_ticks = NULL,
  y_ticks = NULL,
  z_ticks = NULL,
  show_plot = NA,
  return_plot = NA,
  save_plot = NA,
  save_param = list(),
  default_save_name = "crossSectionGenePlot3D"
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{name}] name of virtual cross section to use

\item[\code{spatial\_network\_name}] name of spatial network to use

\item[\code{expression\_values}] gene expression values to use

\item[\code{genes}] genes to show

\item[\code{show\_network}] show underlying spatial network

\item[\code{network\_color}] color of spatial network

\item[\code{show\_grid}] show spatial grid

\item[\code{genes\_high\_color}] color represents high gene expression

\item[\code{genes\_mid\_color}] color represents middle gene expression

\item[\code{genes\_low\_color}] color represents low gene expression

\item[\code{spatial\_grid\_name}] name of spatial grid to use

\item[\code{point\_size}] size of point (cell)

\item[\code{show\_legend}] show legend

\item[\code{show\_plot}] show plots

\item[\code{return\_plot}] return ggplot object

\item[\code{save\_plot}] directly save the plot [boolean]

\item[\code{save\_param}] list of saving parameters from \code{\LinkA{all\_plots\_save\_function}{all.Rul.plots.Rul.save.Rul.function}}

\item[\code{default\_save\_name}] default save name for saving, don't change, change save\_name in save\_param

\item[\code{grid\_color}] color of spatial grid

\item[\code{midpoint}] expression midpoint

\item[\code{scale\_alpha\_with\_expression}] scale expression with ggplot alpha parameter

\item[\code{...}] parameters for cowplot::save\_plot()
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Description of parameters.
\end{Details}
%
\begin{Value}
ggplot
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    crossSectionGenePlot3D(gobject)

\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{crossSectionPlot}{crossSectionPlot}{crossSectionPlot}
%
\begin{Description}\relax
Visualize cells in a virtual cross section according to spatial coordinates
\end{Description}
%
\begin{Usage}
\begin{verbatim}
crossSectionPlot(
  gobject,
  crossSection_obj = NULL,
  name = NULL,
  spatial_network_name = "Delaunay_network",
  group_by = NULL,
  group_by_subset = NULL,
  sdimx = "sdimx",
  sdimy = "sdimy",
  spat_enr_names = NULL,
  cell_color = NULL,
  color_as_factor = T,
  cell_color_code = NULL,
  cell_color_gradient = c("blue", "white", "red"),
  gradient_midpoint = NULL,
  gradient_limits = NULL,
  select_cell_groups = NULL,
  select_cells = NULL,
  point_shape = c("border", "no_border"),
  point_size = 3,
  point_border_col = "black",
  point_border_stroke = 0.1,
  show_cluster_center = F,
  show_center_label = F,
  center_point_size = 4,
  center_point_border_col = "black",
  center_point_border_stroke = 0.1,
  label_size = 4,
  label_fontface = "bold",
  show_network = F,
  network_color = NULL,
  network_alpha = 1,
  show_grid = F,
  spatial_grid_name = "spatial_grid",
  grid_color = NULL,
  show_other_cells = T,
  other_cell_color = "lightgrey",
  other_point_size = 1,
  other_cells_alpha = 0.1,
  coord_fix_ratio = NULL,
  title = NULL,
  show_legend = T,
  legend_text = 8,
  legend_symbol_size = 1,
  background_color = "white",
  axis_text = 8,
  axis_title = 8,
  cow_n_col = 2,
  cow_rel_h = 1,
  cow_rel_w = 1,
  cow_align = "h",
  show_plot = NA,
  return_plot = NA,
  save_plot = NA,
  save_param = list(),
  default_save_name = "crossSectionPlot"
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{name}] name of virtual cross section to use

\item[\code{spatial\_network\_name}] name of spatial network to use

\item[\code{group\_by\_subset}] subset the group\_by factor column

\item[\code{sdimx}] x-axis dimension name (default = 'sdimx')

\item[\code{sdimy}] y-axis dimension name (default = 'sdimy')

\item[\code{spat\_enr\_names}] names of spatial enrichment results to include

\item[\code{cell\_color}] color for cells (see details)

\item[\code{color\_as\_factor}] convert color column to factor

\item[\code{cell\_color\_code}] named vector with colors

\item[\code{cell\_color\_gradient}] vector with 3 colors for numeric data

\item[\code{gradient\_midpoint}] midpoint for color gradient

\item[\code{gradient\_limits}] vector with lower and upper limits

\item[\code{select\_cell\_groups}] select subset of cells/clusters based on cell\_color parameter

\item[\code{select\_cells}] select subset of cells based on cell IDs

\item[\code{point\_shape}] point with border or not (border or no\_border)

\item[\code{point\_size}] size of point (cell)

\item[\code{point\_border\_col}] color of border around points

\item[\code{point\_border\_stroke}] stroke size of border around points

\item[\code{show\_cluster\_center}] plot center of selected clusters

\item[\code{show\_center\_label}] plot label of selected clusters

\item[\code{center\_point\_size}] size of center points

\item[\code{label\_size}] size of labels

\item[\code{label\_fontface}] font of labels

\item[\code{show\_network}] show underlying spatial network

\item[\code{network\_color}] color of spatial network

\item[\code{network\_alpha}] alpha of spatial network

\item[\code{show\_grid}] show spatial grid

\item[\code{spatial\_grid\_name}] name of spatial grid to use

\item[\code{grid\_color}] color of spatial grid

\item[\code{show\_other\_cells}] display not selected cells

\item[\code{other\_cell\_color}] color of not selected cells

\item[\code{other\_point\_size}] point size of not selected cells

\item[\code{other\_cells\_alpha}] alpha of not selected cells

\item[\code{coord\_fix\_ratio}] fix ratio between x and y-axis

\item[\code{title}] title of plot

\item[\code{show\_legend}] show legend

\item[\code{legend\_text}] size of legend text

\item[\code{legend\_symbol\_size}] size of legend symbols

\item[\code{background\_color}] color of plot background

\item[\code{axis\_text}] size of axis text

\item[\code{axis\_title}] size of axis title

\item[\code{cow\_n\_col}] cowplot param: how many columns

\item[\code{cow\_rel\_h}] cowplot param: relative height

\item[\code{cow\_rel\_w}] cowplot param: relative width

\item[\code{cow\_align}] cowplot param: how to align

\item[\code{show\_plot}] show plot

\item[\code{return\_plot}] return ggplot object

\item[\code{save\_plot}] directly save the plot [boolean]

\item[\code{save\_param}] list of saving parameters from \code{\LinkA{all\_plots\_save\_function}{all.Rul.plots.Rul.save.Rul.function}}

\item[\code{default\_save\_name}] default save name for saving, don't change, change save\_name in save\_param

\item[\code{groub\_by}] create multiple plots based on cell annotation column
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Description of parameters.
\end{Details}
%
\begin{Value}
ggplot
\end{Value}
%
\begin{SeeAlso}\relax
\code{\LinkA{crossSectionPlot}{crossSectionPlot}}
\end{SeeAlso}
\inputencoding{utf8}
\HeaderA{crossSectionPlot3D}{crossSectionPlot3D}{crossSectionPlot3D}
%
\begin{Description}\relax
Visualize cells in a virtual cross section according to spatial coordinates
\end{Description}
%
\begin{Usage}
\begin{verbatim}
crossSectionPlot3D(
  gobject,
  crossSection_obj = NULL,
  name = NULL,
  spatial_network_name = "Delaunay_network",
  sdimx = "sdimx",
  sdimy = "sdimy",
  sdimz = "sdimz",
  point_size = 3,
  cell_color = NULL,
  cell_color_code = NULL,
  select_cell_groups = NULL,
  show_other_cells = T,
  other_cell_color = alpha("lightgrey", 0),
  other_point_size = 0.5,
  show_network = F,
  network_color = NULL,
  network_alpha = 1,
  other_cell_alpha = 0.5,
  show_grid = F,
  grid_color = NULL,
  spatial_grid_name = "spatial_grid",
  title = "",
  show_legend = T,
  axis_scale = c("cube", "real", "custom"),
  custom_ratio = NULL,
  x_ticks = NULL,
  y_ticks = NULL,
  z_ticks = NULL,
  show_plot = NA,
  return_plot = NA,
  save_plot = NA,
  save_param = list(),
  default_save_name = "crossSection3D"
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{name}] name of virtual cross section to use

\item[\code{spatial\_network\_name}] name of spatial network to use

\item[\code{sdimx}] x-axis dimension name (default = 'sdimx')

\item[\code{sdimy}] y-axis dimension name (default = 'sdimy')

\item[\code{sdimz}] z-axis dimension name (default = 'sdimy')

\item[\code{point\_size}] size of point (cell)

\item[\code{cell\_color}] color for cells (see details)

\item[\code{cell\_color\_code}] named vector with colors

\item[\code{select\_cell\_groups}] select subset of cells/clusters based on cell\_color parameter

\item[\code{show\_other\_cells}] display not selected cells

\item[\code{other\_cell\_color}] color of not selected cells

\item[\code{other\_point\_size}] point size of not selected cells

\item[\code{network\_color}] color of spatial network

\item[\code{show\_grid}] show spatial grid

\item[\code{grid\_color}] color of spatial grid

\item[\code{spatial\_grid\_name}] name of spatial grid to use

\item[\code{title}] title of plot

\item[\code{show\_legend}] show legend

\item[\code{axis\_scale}] the way to scale the axis

\item[\code{custom\_ratio}] customize the scale of the plot

\item[\code{x\_ticks}] set the number of ticks on the x-axis

\item[\code{y\_ticks}] set the number of ticks on the y-axis

\item[\code{z\_ticks}] set the number of ticks on the z-axis

\item[\code{show\_plot}] show plot

\item[\code{return\_plot}] return ggplot object

\item[\code{save\_plot}] directly save the plot [boolean]

\item[\code{save\_param}] list of saving parameters from \code{\LinkA{all\_plots\_save\_function}{all.Rul.plots.Rul.save.Rul.function}}

\item[\code{default\_save\_name}] default save name for saving, don't change, change save\_name in save\_param
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Description of parameters.
\end{Details}
%
\begin{Value}
ggplot
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    crossSectionPlot3D(gobject)

\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{decide\_cluster\_order}{decide\_cluster\_order}{decide.Rul.cluster.Rul.order}
%
\begin{Description}\relax
creates order for clusters
\end{Description}
%
\begin{Usage}
\begin{verbatim}
decide_cluster_order(
  gobject,
  expression_values = c("normalized", "scaled", "custom"),
  genes,
  cluster_column = NULL,
  cluster_order = c("size", "correlation", "custom"),
  cluster_custom_order = NULL,
  cor_method = "pearson",
  hclust_method = "ward.D"
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{expression\_values}] expression values to use

\item[\code{genes}] genes to use

\item[\code{cluster\_column}] name of column to use for clusters

\item[\code{cluster\_order}] method to determine cluster order

\item[\code{cluster\_custom\_order}] custom order for clusters

\item[\code{cor\_method}] method for correlation

\item[\code{hclust\_method}] method for hierarchical clustering
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Calculates order for clusters.
\end{Details}
%
\begin{Value}
custom
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    decide_cluster_order(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{detectSpatialCorGenes}{detectSpatialCorGenes}{detectSpatialCorGenes}
%
\begin{Description}\relax
Detect genes that are spatially correlated
\end{Description}
%
\begin{Usage}
\begin{verbatim}
detectSpatialCorGenes(
  gobject,
  method = c("grid", "network"),
  expression_values = c("normalized", "scaled", "custom"),
  subset_genes = NULL,
  spatial_network_name = "Delaunay_network",
  network_smoothing = NULL,
  spatial_grid_name = "spatial_grid",
  min_cells_per_grid = 4,
  cor_method = c("pearson", "kendall", "spearman")
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{method}] method to use for spatial averaging

\item[\code{expression\_values}] gene expression values to use

\item[\code{subset\_genes}] subset of genes to use

\item[\code{spatial\_network\_name}] name of spatial network to use

\item[\code{network\_smoothing}] smoothing factor beteen 0 and 1 (default: automatic)

\item[\code{spatial\_grid\_name}] name of spatial grid to use

\item[\code{min\_cells\_per\_grid}] minimum number of cells to consider a grid

\item[\code{b}] smoothing factor beteen 0 and 1 (default: automatic)
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
For method = network, it expects a fully connected spatial network. You can make sure to create a
fully connected network by setting minimal\_k > 0 in the \code{\LinkA{createSpatialNetwork}{createSpatialNetwork}} function.
\begin{itemize}

\item{} 1. grid-averaging: average gene expression values within a predefined spatial grid
\item{} 2. network-averaging: smoothens the gene expression matrix by averaging the expression within one cell
by using the neighbours within the predefined spatial network. b is a smoothening factor
that defaults to 1 - 1/k, where k is the median number of  k-neighbors in the
selected spatial network. Setting b = 0 means no smoothing and b = 1 means no contribution
from its own expression.

\end{itemize}

The spatCorObject can be further explored with showSpatialCorGenes()
\end{Details}
%
\begin{Value}
returns a spatial correlation object: "spatCorObject"
\end{Value}
%
\begin{SeeAlso}\relax
\code{\LinkA{showSpatialCorGenes}{showSpatialCorGenes}}
\end{SeeAlso}
%
\begin{Examples}
\begin{ExampleCode}
    detectSpatialCorGenes(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{detectSpatialPatterns}{detectSpatialPatterns}{detectSpatialPatterns}
%
\begin{Description}\relax
Identify spatial patterns through PCA on average expression in a spatial grid.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
detectSpatialPatterns(
  gobject,
  expression_values = c("normalized", "scaled", "custom"),
  spatial_grid_name = "spatial_grid",
  min_cells_per_grid = 4,
  scale_unit = F,
  ncp = 100,
  show_plot = T,
  PC_zscore = 1.5
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{expression\_values}] expression values to use

\item[\code{spatial\_grid\_name}] name of spatial grid to use (default = 'spatial\_grid')

\item[\code{min\_cells\_per\_grid}] minimum number of cells in a grid to be considered

\item[\code{scale\_unit}] scale features

\item[\code{ncp}] number of principal components to calculate

\item[\code{show\_plot}] show plots

\item[\code{PC\_zscore}] minimum z-score of variance explained by a PC
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Steps to identify spatial patterns:
\begin{itemize}

\item{} 1. average gene expression for cells within a grid, see createSpatialGrid
\item{} 2. perform PCA on the average grid expression profiles
\item{} 3. convert variance of principlal components (PCs) to z-scores and select PCs based on a z-score threshold

\end{itemize}

\end{Details}
%
\begin{Value}
spatial pattern object 'spatPatObj'
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    detectSpatialPatterns(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{dimCellPlot}{dimCellPlot}{dimCellPlot}
%
\begin{Description}\relax
Visualize cells according to dimension reduction coordinates
\end{Description}
%
\begin{Usage}
\begin{verbatim}
dimCellPlot(
  gobject,
  dim_reduction_to_use = "umap",
  dim_reduction_name = "umap",
  dim1_to_use = 1,
  dim2_to_use = 2,
  spat_enr_names = NULL,
  cell_annotation_values = NULL,
  show_NN_network = F,
  nn_network_to_use = "sNN",
  network_name = "sNN.pca",
  cell_color_gradient = c("blue", "white", "red"),
  gradient_midpoint = NULL,
  gradient_limits = NULL,
  select_cell_groups = NULL,
  select_cells = NULL,
  show_other_cells = T,
  other_cell_color = "lightgrey",
  other_point_size = 0.5,
  show_cluster_center = F,
  show_center_label = T,
  center_point_size = 4,
  center_point_border_col = "black",
  center_point_border_stroke = 0.1,
  label_size = 4,
  label_fontface = "bold",
  edge_alpha = NULL,
  point_shape = c("border", "no_border"),
  point_size = 1,
  point_border_col = "black",
  point_border_stroke = 0.1,
  show_legend = T,
  legend_text = 8,
  legend_symbol_size = 1,
  background_color = "white",
  axis_text = 8,
  axis_title = 8,
  cow_n_col = 2,
  cow_rel_h = 1,
  cow_rel_w = 1,
  cow_align = "h",
  show_plot = NA,
  return_plot = NA,
  save_plot = NA,
  save_param = list(),
  default_save_name = "dimCellPlot"
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{dim\_reduction\_to\_use}] dimension reduction to use

\item[\code{dim\_reduction\_name}] dimension reduction name

\item[\code{dim1\_to\_use}] dimension to use on x-axis

\item[\code{dim2\_to\_use}] dimension to use on y-axis

\item[\code{spat\_enr\_names}] names of spatial enrichment results to include

\item[\code{cell\_annotation\_values}] numeric cell annotation columns

\item[\code{show\_NN\_network}] show underlying NN network

\item[\code{nn\_network\_to\_use}] type of NN network to use (kNN vs sNN)

\item[\code{network\_name}] name of NN network to use, if show\_NN\_network = TRUE

\item[\code{cell\_color\_gradient}] vector with 3 colors for numeric data

\item[\code{gradient\_midpoint}] midpoint for color gradient

\item[\code{gradient\_limits}] vector with lower and upper limits

\item[\code{select\_cell\_groups}] select subset of cells/clusters based on cell\_color parameter

\item[\code{select\_cells}] select subset of cells based on cell IDs

\item[\code{show\_other\_cells}] display not selected cells

\item[\code{other\_cell\_color}] color of not selected cells

\item[\code{other\_point\_size}] size of not selected cells

\item[\code{show\_cluster\_center}] plot center of selected clusters

\item[\code{show\_center\_label}] plot label of selected clusters

\item[\code{center\_point\_size}] size of center points

\item[\code{label\_size}] size of labels

\item[\code{label\_fontface}] font of labels

\item[\code{edge\_alpha}] column to use for alpha of the edges

\item[\code{point\_shape}] point with border or not (border or no\_border)

\item[\code{point\_size}] size of point (cell)

\item[\code{point\_border\_col}] color of border around points

\item[\code{point\_border\_stroke}] stroke size of border around points

\item[\code{show\_legend}] show legend

\item[\code{legend\_text}] size of legend text

\item[\code{legend\_symbol\_size}] size of legend symbols

\item[\code{background\_color}] color of plot background

\item[\code{axis\_text}] size of axis text

\item[\code{axis\_title}] size of axis title

\item[\code{show\_plot}] show plot

\item[\code{return\_plot}] return ggplot object

\item[\code{save\_plot}] directly save the plot [boolean]

\item[\code{save\_param}] list of saving parameters from \code{\LinkA{all\_plots\_save\_function}{all.Rul.plots.Rul.save.Rul.function}}

\item[\code{default\_save\_name}] default save name for saving, don't change, change save\_name in save\_param

\item[\code{cell\_color}] color for cells (see details)

\item[\code{color\_as\_factor}] convert color column to factor

\item[\code{cell\_color\_code}] named vector with colors

\item[\code{title}] title for plot, defaults to cell\_color parameter
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Description of parameters. For 3D plots see \code{\LinkA{dimCellPlot2D}{dimCellPlot2D}}
\end{Details}
%
\begin{Value}
ggplot
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    dimCellPlot(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{dimCellPlot2D}{dimCellPlot2D}{dimCellPlot2D}
%
\begin{Description}\relax
Visualize cells according to dimension reduction coordinates
\end{Description}
%
\begin{Usage}
\begin{verbatim}
dimCellPlot2D(
  gobject,
  dim_reduction_to_use = "umap",
  dim_reduction_name = "umap",
  dim1_to_use = 1,
  dim2_to_use = 2,
  spat_enr_names = NULL,
  cell_annotation_values = NULL,
  show_NN_network = F,
  nn_network_to_use = "sNN",
  network_name = "sNN.pca",
  cell_color_gradient = c("blue", "white", "red"),
  gradient_midpoint = NULL,
  gradient_limits = NULL,
  select_cell_groups = NULL,
  select_cells = NULL,
  show_other_cells = T,
  other_cell_color = "lightgrey",
  other_point_size = 0.5,
  show_cluster_center = F,
  show_center_label = T,
  center_point_size = 4,
  center_point_border_col = "black",
  center_point_border_stroke = 0.1,
  label_size = 4,
  label_fontface = "bold",
  edge_alpha = NULL,
  point_shape = c("border", "no_border"),
  point_size = 1,
  point_border_col = "black",
  point_border_stroke = 0.1,
  show_legend = T,
  legend_text = 8,
  legend_symbol_size = 1,
  background_color = "white",
  axis_text = 8,
  axis_title = 8,
  cow_n_col = 2,
  cow_rel_h = 1,
  cow_rel_w = 1,
  cow_align = "h",
  show_plot = NA,
  return_plot = NA,
  save_plot = NA,
  save_param = list(),
  default_save_name = "dimCellPlot2D"
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{dim\_reduction\_to\_use}] dimension reduction to use

\item[\code{dim\_reduction\_name}] dimension reduction name

\item[\code{dim1\_to\_use}] dimension to use on x-axis

\item[\code{dim2\_to\_use}] dimension to use on y-axis

\item[\code{spat\_enr\_names}] names of spatial enrichment results to include

\item[\code{cell\_annotation\_values}] numeric cell annotation columns

\item[\code{show\_NN\_network}] show underlying NN network

\item[\code{nn\_network\_to\_use}] type of NN network to use (kNN vs sNN)

\item[\code{network\_name}] name of NN network to use, if show\_NN\_network = TRUE

\item[\code{cell\_color\_gradient}] vector with 3 colors for numeric data

\item[\code{gradient\_midpoint}] midpoint for color gradient

\item[\code{gradient\_limits}] vector with lower and upper limits

\item[\code{select\_cell\_groups}] select subset of cells/clusters based on cell\_color parameter

\item[\code{select\_cells}] select subset of cells based on cell IDs

\item[\code{show\_other\_cells}] display not selected cells

\item[\code{other\_cell\_color}] color of not selected cells

\item[\code{other\_point\_size}] size of not selected cells

\item[\code{show\_cluster\_center}] plot center of selected clusters

\item[\code{show\_center\_label}] plot label of selected clusters

\item[\code{center\_point\_size}] size of center points

\item[\code{label\_size}] size of labels

\item[\code{label\_fontface}] font of labels

\item[\code{edge\_alpha}] column to use for alpha of the edges

\item[\code{point\_shape}] point with border or not (border or no\_border)

\item[\code{point\_size}] size of point (cell)

\item[\code{point\_border\_col}] color of border around points

\item[\code{point\_border\_stroke}] stroke size of border around points

\item[\code{show\_legend}] show legend

\item[\code{legend\_text}] size of legend text

\item[\code{legend\_symbol\_size}] size of legend symbols

\item[\code{background\_color}] color of plot background

\item[\code{axis\_text}] size of axis text

\item[\code{axis\_title}] size of axis title

\item[\code{show\_plot}] show plot

\item[\code{return\_plot}] return ggplot object

\item[\code{save\_plot}] directly save the plot [boolean]

\item[\code{save\_param}] list of saving parameters from \code{\LinkA{all\_plots\_save\_function}{all.Rul.plots.Rul.save.Rul.function}}

\item[\code{default\_save\_name}] default save name for saving, don't change, change save\_name in save\_param

\item[\code{cell\_color}] color for cells (see details)

\item[\code{color\_as\_factor}] convert color column to factor

\item[\code{cell\_color\_code}] named vector with colors

\item[\code{title}] title for plot, defaults to cell\_color parameter
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Description of parameters. For 3D plots see \code{\LinkA{dimPlot3D}{dimPlot3D}}
\end{Details}
%
\begin{Value}
ggplot
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    dimCellPlot2D(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{dimGenePlot}{dimGenePlot}{dimGenePlot}
%
\begin{Description}\relax
Visualize cells and gene expression according to dimension reduction coordinates
\end{Description}
%
\begin{Usage}
\begin{verbatim}
dimGenePlot(
  gobject,
  expression_values = c("normalized", "scaled", "custom"),
  genes = NULL,
  dim_reduction_to_use = "umap",
  dim_reduction_name = "umap",
  dim1_to_use = 1,
  dim2_to_use = 2,
  show_NN_network = F,
  nn_network_to_use = "sNN",
  network_name = "sNN.pca",
  network_color = "lightgray",
  edge_alpha = NULL,
  scale_alpha_with_expression = FALSE,
  point_shape = c("border", "no_border"),
  point_size = 1,
  cell_color_gradient = c("blue", "white", "red"),
  gradient_midpoint = NULL,
  gradient_limits = NULL,
  point_border_col = "black",
  point_border_stroke = 0.1,
  show_legend = T,
  legend_text = 8,
  background_color = "white",
  axis_text = 8,
  axis_title = 8,
  cow_n_col = 2,
  cow_rel_h = 1,
  cow_rel_w = 1,
  cow_align = "h",
  show_plot = NA,
  return_plot = NA,
  save_plot = NA,
  save_param = list(),
  default_save_name = "dimGenePlot"
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{expression\_values}] gene expression values to use

\item[\code{genes}] genes to show

\item[\code{dim\_reduction\_to\_use}] dimension reduction to use

\item[\code{dim\_reduction\_name}] dimension reduction name

\item[\code{dim1\_to\_use}] dimension to use on x-axis

\item[\code{dim2\_to\_use}] dimension to use on y-axis

\item[\code{show\_NN\_network}] show underlying NN network

\item[\code{nn\_network\_to\_use}] type of NN network to use (kNN vs sNN)

\item[\code{network\_name}] name of NN network to use, if show\_NN\_network = TRUE

\item[\code{edge\_alpha}] column to use for alpha of the edges

\item[\code{scale\_alpha\_with\_expression}] scale expression with ggplot alpha parameter

\item[\code{point\_size}] size of point (cell)

\item[\code{cell\_color\_gradient}] vector with 3 colors for numeric data

\item[\code{gradient\_midpoint}] midpoint for color gradient

\item[\code{gradient\_limits}] vector with lower and upper limits

\item[\code{point\_border\_col}] color of border around points

\item[\code{point\_border\_stroke}] stroke size of border around points

\item[\code{show\_legend}] show legend

\item[\code{cow\_n\_col}] cowplot param: how many columns

\item[\code{cow\_rel\_h}] cowplot param: relative height

\item[\code{cow\_rel\_w}] cowplot param: relative width

\item[\code{cow\_align}] cowplot param: how to align

\item[\code{show\_plot}] show plots

\item[\code{return\_plot}] return ggplot object

\item[\code{save\_plot}] directly save the plot [boolean]

\item[\code{save\_param}] list of saving parameters from \code{\LinkA{all\_plots\_save\_function}{all.Rul.plots.Rul.save.Rul.function}}

\item[\code{default\_save\_name}] default save name for saving, don't change, change save\_name in save\_param

\item[\code{...}] parameters for cowplot::save\_plot()
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Description of parameters.
\end{Details}
%
\begin{Value}
ggplot
\end{Value}
%
\begin{SeeAlso}\relax
\code{\LinkA{dimGenePlot3D}{dimGenePlot3D}}
\end{SeeAlso}
%
\begin{Examples}
\begin{ExampleCode}
    dimGenePlot(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{dimGenePlot2D}{dimGenePlot2D}{dimGenePlot2D}
%
\begin{Description}\relax
Visualize cells and gene expression according to dimension reduction coordinates
\end{Description}
%
\begin{Usage}
\begin{verbatim}
dimGenePlot2D(
  gobject,
  expression_values = c("normalized", "scaled", "custom"),
  genes = NULL,
  dim_reduction_to_use = "umap",
  dim_reduction_name = "umap",
  dim1_to_use = 1,
  dim2_to_use = 2,
  show_NN_network = F,
  nn_network_to_use = "sNN",
  network_name = "sNN.pca",
  network_color = "lightgray",
  edge_alpha = NULL,
  scale_alpha_with_expression = FALSE,
  point_shape = c("border", "no_border"),
  point_size = 1,
  cell_color_gradient = c("blue", "white", "red"),
  gradient_midpoint = NULL,
  gradient_limits = NULL,
  point_border_col = "black",
  point_border_stroke = 0.1,
  show_legend = T,
  legend_text = 8,
  background_color = "white",
  axis_text = 8,
  axis_title = 8,
  cow_n_col = 2,
  cow_rel_h = 1,
  cow_rel_w = 1,
  cow_align = "h",
  show_plot = NA,
  return_plot = NA,
  save_plot = NA,
  save_param = list(),
  default_save_name = "dimGenePlot2D"
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{expression\_values}] gene expression values to use

\item[\code{genes}] genes to show

\item[\code{dim\_reduction\_to\_use}] dimension reduction to use

\item[\code{dim\_reduction\_name}] dimension reduction name

\item[\code{dim1\_to\_use}] dimension to use on x-axis

\item[\code{dim2\_to\_use}] dimension to use on y-axis

\item[\code{show\_NN\_network}] show underlying NN network

\item[\code{nn\_network\_to\_use}] type of NN network to use (kNN vs sNN)

\item[\code{network\_name}] name of NN network to use, if show\_NN\_network = TRUE

\item[\code{edge\_alpha}] column to use for alpha of the edges

\item[\code{scale\_alpha\_with\_expression}] scale expression with ggplot alpha parameter

\item[\code{point\_shape}] point with border or not (border or no\_border)

\item[\code{point\_size}] size of point (cell)

\item[\code{cell\_color\_gradient}] vector with 3 colors for numeric data

\item[\code{gradient\_midpoint}] midpoint for color gradient

\item[\code{gradient\_limits}] vector with lower and upper limits

\item[\code{point\_border\_col}] color of border around points

\item[\code{point\_border\_stroke}] stroke size of border around points

\item[\code{show\_legend}] show legend

\item[\code{legend\_text}] size of legend text

\item[\code{background\_color}] color of plot background

\item[\code{axis\_text}] size of axis text

\item[\code{axis\_title}] size of axis title

\item[\code{cow\_n\_col}] cowplot param: how many columns

\item[\code{cow\_rel\_h}] cowplot param: relative height

\item[\code{cow\_rel\_w}] cowplot param: relative width

\item[\code{cow\_align}] cowplot param: how to align

\item[\code{show\_plot}] show plots

\item[\code{return\_plot}] return ggplot object

\item[\code{save\_plot}] directly save the plot [boolean]

\item[\code{save\_param}] list of saving parameters from \code{\LinkA{all\_plots\_save\_function}{all.Rul.plots.Rul.save.Rul.function}}

\item[\code{default\_save\_name}] default save name for saving, don't change, change save\_name in save\_param

\item[\code{...}] parameters for cowplot::save\_plot()
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Description of parameters.
\end{Details}
%
\begin{Value}
ggplot
\end{Value}
%
\begin{SeeAlso}\relax
\code{\LinkA{dimGenePlot3D}{dimGenePlot3D}}
\end{SeeAlso}
%
\begin{Examples}
\begin{ExampleCode}
    dimGenePlot2D(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{dimGenePlot3D}{dimGenePlot3D}{dimGenePlot3D}
%
\begin{Description}\relax
Visualize cells and gene expression according to dimension reduction coordinates
\end{Description}
%
\begin{Usage}
\begin{verbatim}
dimGenePlot3D(
  gobject,
  expression_values = c("normalized", "scaled", "custom"),
  genes = NULL,
  dim_reduction_to_use = "umap",
  dim_reduction_name = "umap",
  dim1_to_use = 1,
  dim2_to_use = 2,
  dim3_to_use = 3,
  show_NN_network = F,
  nn_network_to_use = "sNN",
  network_name = "sNN.pca",
  network_color = "lightgray",
  cluster_column = NULL,
  select_cell_groups = NULL,
  select_cells = NULL,
  show_other_cells = T,
  other_cell_color = "lightgrey",
  other_point_size = 1,
  edge_alpha = NULL,
  point_size = 2,
  genes_high_color = NULL,
  genes_mid_color = "white",
  genes_low_color = "blue",
  show_legend = T,
  show_plot = NA,
  return_plot = NA,
  save_plot = NA,
  save_param = list(),
  default_save_name = "dimGenePlot3D"
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{expression\_values}] gene expression values to use

\item[\code{genes}] genes to show

\item[\code{dim\_reduction\_to\_use}] dimension reduction to use

\item[\code{dim\_reduction\_name}] dimension reduction name

\item[\code{dim1\_to\_use}] dimension to use on x-axis

\item[\code{dim2\_to\_use}] dimension to use on y-axis

\item[\code{dim3\_to\_use}] dimension to use on z-axis

\item[\code{show\_NN\_network}] show underlying NN network

\item[\code{nn\_network\_to\_use}] type of NN network to use (kNN vs sNN)

\item[\code{network\_name}] name of NN network to use, if show\_NN\_network = TRUE

\item[\code{edge\_alpha}] column to use for alpha of the edges

\item[\code{point\_size}] size of point (cell)

\item[\code{show\_legend}] show legend

\item[\code{show\_plot}] show plots

\item[\code{return\_plot}] return ggplot object

\item[\code{save\_plot}] directly save the plot [boolean]

\item[\code{save\_param}] list of saving parameters from \code{\LinkA{all\_plots\_save\_function}{all.Rul.plots.Rul.save.Rul.function}}

\item[\code{default\_save\_name}] default save name for saving, don't change, change save\_name in save\_param

\item[\code{...}] parameters for cowplot::save\_plot()
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Description of parameters.
\end{Details}
%
\begin{Value}
ggplot
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    dimGenePlot3D(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{dimPlot}{dimPlot}{dimPlot}
%
\begin{Description}\relax
Visualize cells according to dimension reduction coordinates
\end{Description}
%
\begin{Usage}
\begin{verbatim}
dimPlot(
  gobject,
  group_by = NULL,
  group_by_subset = NULL,
  dim_reduction_to_use = "umap",
  dim_reduction_name = "umap",
  dim1_to_use = 1,
  dim2_to_use = 2,
  spat_enr_names = NULL,
  show_NN_network = F,
  nn_network_to_use = "sNN",
  network_name = "sNN.pca",
  cell_color = NULL,
  color_as_factor = T,
  cell_color_code = NULL,
  cell_color_gradient = c("blue", "white", "red"),
  gradient_midpoint = NULL,
  gradient_limits = NULL,
  select_cell_groups = NULL,
  select_cells = NULL,
  show_other_cells = T,
  other_cell_color = "lightgrey",
  other_point_size = 0.5,
  show_cluster_center = F,
  show_center_label = T,
  center_point_size = 4,
  center_point_border_col = "black",
  center_point_border_stroke = 0.1,
  label_size = 4,
  label_fontface = "bold",
  edge_alpha = NULL,
  point_shape = c("border", "no_border"),
  point_size = 1,
  point_border_col = "black",
  point_border_stroke = 0.1,
  show_legend = T,
  legend_text = 8,
  legend_symbol_size = 1,
  background_color = "white",
  axis_text = 8,
  axis_title = 8,
  title = NULL,
  cow_n_col = 2,
  cow_rel_h = 1,
  cow_rel_w = 1,
  cow_align = "h",
  show_plot = NA,
  return_plot = NA,
  save_plot = NA,
  save_param = list(),
  default_save_name = "dimPlot"
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{group\_by\_subset}] subset the group\_by factor column

\item[\code{dim\_reduction\_to\_use}] dimension reduction to use

\item[\code{dim\_reduction\_name}] dimension reduction name

\item[\code{dim1\_to\_use}] dimension to use on x-axis

\item[\code{dim2\_to\_use}] dimension to use on y-axis

\item[\code{spat\_enr\_names}] names of spatial enrichment results to include

\item[\code{show\_NN\_network}] show underlying NN network

\item[\code{nn\_network\_to\_use}] type of NN network to use (kNN vs sNN)

\item[\code{network\_name}] name of NN network to use, if show\_NN\_network = TRUE

\item[\code{cell\_color}] color for cells (see details)

\item[\code{color\_as\_factor}] convert color column to factor

\item[\code{cell\_color\_code}] named vector with colors

\item[\code{cell\_color\_gradient}] vector with 3 colors for numeric data

\item[\code{gradient\_midpoint}] midpoint for color gradient

\item[\code{gradient\_limits}] vector with lower and upper limits

\item[\code{select\_cell\_groups}] select subset of cells/clusters based on cell\_color parameter

\item[\code{select\_cells}] select subset of cells based on cell IDs

\item[\code{show\_other\_cells}] display not selected cells

\item[\code{other\_cell\_color}] color of not selected cells

\item[\code{other\_point\_size}] size of not selected cells

\item[\code{show\_cluster\_center}] plot center of selected clusters

\item[\code{show\_center\_label}] plot label of selected clusters

\item[\code{center\_point\_size}] size of center points

\item[\code{label\_size}] size of labels

\item[\code{label\_fontface}] font of labels

\item[\code{edge\_alpha}] column to use for alpha of the edges

\item[\code{point\_shape}] point with border or not (border or no\_border)

\item[\code{point\_size}] size of point (cell)

\item[\code{point\_border\_col}] color of border around points

\item[\code{point\_border\_stroke}] stroke size of border around points

\item[\code{show\_legend}] show legend

\item[\code{legend\_text}] size of legend text

\item[\code{legend\_symbol\_size}] size of legend symbols

\item[\code{background\_color}] color of plot background

\item[\code{axis\_text}] size of axis text

\item[\code{axis\_title}] size of axis title

\item[\code{title}] title for plot, defaults to cell\_color parameter

\item[\code{cow\_n\_col}] cowplot param: how many columns

\item[\code{cow\_rel\_h}] cowplot param: relative height

\item[\code{cow\_rel\_w}] cowplot param: relative width

\item[\code{cow\_align}] cowplot param: how to align

\item[\code{show\_plot}] show plot

\item[\code{return\_plot}] return ggplot object

\item[\code{save\_plot}] directly save the plot [boolean]

\item[\code{save\_param}] list of saving parameters from \code{\LinkA{all\_plots\_save\_function}{all.Rul.plots.Rul.save.Rul.function}}

\item[\code{default\_save\_name}] default save name for saving, don't change, change save\_name in save\_param

\item[\code{groub\_by}] create multiple plots based on cell annotation column
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Description of parameters, see \code{\LinkA{dimPlot2D}{dimPlot2D}}. For 3D plots see \code{\LinkA{dimPlot3D}{dimPlot3D}}
\end{Details}
%
\begin{Value}
ggplot
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    dimPlot(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{dimPlot2D}{dimPlot2D}{dimPlot2D}
%
\begin{Description}\relax
Visualize cells according to dimension reduction coordinates
\end{Description}
%
\begin{Usage}
\begin{verbatim}
dimPlot2D(
  gobject,
  group_by = NULL,
  group_by_subset = NULL,
  dim_reduction_to_use = "umap",
  dim_reduction_name = "umap",
  dim1_to_use = 1,
  dim2_to_use = 2,
  spat_enr_names = NULL,
  show_NN_network = F,
  nn_network_to_use = "sNN",
  network_name = "sNN.pca",
  cell_color = NULL,
  color_as_factor = T,
  cell_color_code = NULL,
  cell_color_gradient = c("blue", "white", "red"),
  gradient_midpoint = NULL,
  gradient_limits = NULL,
  select_cell_groups = NULL,
  select_cells = NULL,
  show_other_cells = T,
  other_cell_color = "lightgrey",
  other_point_size = 0.5,
  show_cluster_center = F,
  show_center_label = T,
  center_point_size = 4,
  center_point_border_col = "black",
  center_point_border_stroke = 0.1,
  label_size = 4,
  label_fontface = "bold",
  edge_alpha = NULL,
  point_shape = c("border", "no_border"),
  point_size = 1,
  point_border_col = "black",
  point_border_stroke = 0.1,
  title = NULL,
  show_legend = T,
  legend_text = 8,
  legend_symbol_size = 1,
  background_color = "white",
  axis_text = 8,
  axis_title = 8,
  cow_n_col = 2,
  cow_rel_h = 1,
  cow_rel_w = 1,
  cow_align = "h",
  show_plot = NA,
  return_plot = NA,
  save_plot = NA,
  save_param = list(),
  default_save_name = "dimPlot2D"
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{group\_by\_subset}] subset the group\_by factor column

\item[\code{dim\_reduction\_to\_use}] dimension reduction to use

\item[\code{dim\_reduction\_name}] dimension reduction name

\item[\code{dim1\_to\_use}] dimension to use on x-axis

\item[\code{dim2\_to\_use}] dimension to use on y-axis

\item[\code{spat\_enr\_names}] names of spatial enrichment results to include

\item[\code{show\_NN\_network}] show underlying NN network

\item[\code{nn\_network\_to\_use}] type of NN network to use (kNN vs sNN)

\item[\code{network\_name}] name of NN network to use, if show\_NN\_network = TRUE

\item[\code{cell\_color}] color for cells (see details)

\item[\code{color\_as\_factor}] convert color column to factor

\item[\code{cell\_color\_code}] named vector with colors

\item[\code{cell\_color\_gradient}] vector with 3 colors for numeric data

\item[\code{gradient\_midpoint}] midpoint for color gradient

\item[\code{gradient\_limits}] vector with lower and upper limits

\item[\code{select\_cell\_groups}] select subset of cells/clusters based on cell\_color parameter

\item[\code{select\_cells}] select subset of cells based on cell IDs

\item[\code{show\_other\_cells}] display not selected cells

\item[\code{other\_cell\_color}] color of not selected cells

\item[\code{other\_point\_size}] size of not selected cells

\item[\code{show\_cluster\_center}] plot center of selected clusters

\item[\code{show\_center\_label}] plot label of selected clusters

\item[\code{center\_point\_size}] size of center points

\item[\code{label\_size}] size of labels

\item[\code{label\_fontface}] font of labels

\item[\code{edge\_alpha}] column to use for alpha of the edges

\item[\code{point\_shape}] point with border or not (border or no\_border)

\item[\code{point\_size}] size of point (cell)

\item[\code{point\_border\_col}] color of border around points

\item[\code{point\_border\_stroke}] stroke size of border around points

\item[\code{title}] title for plot, defaults to cell\_color parameter

\item[\code{show\_legend}] show legend

\item[\code{legend\_text}] size of legend text

\item[\code{legend\_symbol\_size}] size of legend symbols

\item[\code{background\_color}] color of plot background

\item[\code{axis\_text}] size of axis text

\item[\code{axis\_title}] size of axis title

\item[\code{cow\_n\_col}] cowplot param: how many columns

\item[\code{cow\_rel\_h}] cowplot param: relative height

\item[\code{cow\_rel\_w}] cowplot param: relative width

\item[\code{cow\_align}] cowplot param: how to align

\item[\code{show\_plot}] show plot

\item[\code{return\_plot}] return ggplot object

\item[\code{save\_plot}] directly save the plot [boolean]

\item[\code{save\_param}] list of saving parameters from \code{\LinkA{all\_plots\_save\_function}{all.Rul.plots.Rul.save.Rul.function}}

\item[\code{default\_save\_name}] default save name for saving, don't change, change save\_name in save\_param

\item[\code{groub\_by}] create multiple plots based on cell annotation column
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Description of parameters. For 3D plots see \code{\LinkA{dimPlot3D}{dimPlot3D}}
\end{Details}
%
\begin{Value}
ggplot
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    dimPlot2D(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{dimPlot2D\_single}{dimPlot2D\_single}{dimPlot2D.Rul.single}
%
\begin{Description}\relax
Visualize cells according to dimension reduction coordinates
\end{Description}
%
\begin{Usage}
\begin{verbatim}
dimPlot2D_single(
  gobject,
  dim_reduction_to_use = "umap",
  dim_reduction_name = "umap",
  dim1_to_use = 1,
  dim2_to_use = 2,
  spat_enr_names = NULL,
  show_NN_network = F,
  nn_network_to_use = "sNN",
  network_name = "sNN.pca",
  cell_color = NULL,
  color_as_factor = T,
  cell_color_code = NULL,
  cell_color_gradient = c("blue", "white", "red"),
  gradient_midpoint = NULL,
  gradient_limits = NULL,
  select_cell_groups = NULL,
  select_cells = NULL,
  show_other_cells = T,
  other_cell_color = "lightgrey",
  other_point_size = 0.5,
  show_cluster_center = F,
  show_center_label = T,
  center_point_size = 4,
  center_point_border_col = "black",
  center_point_border_stroke = 0.1,
  label_size = 4,
  label_fontface = "bold",
  edge_alpha = NULL,
  point_shape = c("border", "no_border"),
  point_size = 1,
  point_border_col = "black",
  point_border_stroke = 0.1,
  title = NULL,
  show_legend = T,
  legend_text = 8,
  legend_symbol_size = 1,
  background_color = "white",
  axis_text = 8,
  axis_title = 8,
  show_plot = NA,
  return_plot = NA,
  save_plot = NA,
  save_param = list(),
  default_save_name = "dimPlot2D_single"
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{dim\_reduction\_to\_use}] dimension reduction to use

\item[\code{dim\_reduction\_name}] dimension reduction name

\item[\code{dim1\_to\_use}] dimension to use on x-axis

\item[\code{dim2\_to\_use}] dimension to use on y-axis

\item[\code{spat\_enr\_names}] names of spatial enrichment results to include

\item[\code{show\_NN\_network}] show underlying NN network

\item[\code{nn\_network\_to\_use}] type of NN network to use (kNN vs sNN)

\item[\code{network\_name}] name of NN network to use, if show\_NN\_network = TRUE

\item[\code{cell\_color}] color for cells (see details)

\item[\code{color\_as\_factor}] convert color column to factor

\item[\code{cell\_color\_code}] named vector with colors

\item[\code{cell\_color\_gradient}] vector with 3 colors for numeric data

\item[\code{gradient\_midpoint}] midpoint for color gradient

\item[\code{gradient\_limits}] vector with lower and upper limits

\item[\code{select\_cell\_groups}] select subset of cells/clusters based on cell\_color parameter

\item[\code{select\_cells}] select subset of cells based on cell IDs

\item[\code{show\_other\_cells}] display not selected cells

\item[\code{other\_cell\_color}] color of not selected cells

\item[\code{other\_point\_size}] size of not selected cells

\item[\code{show\_cluster\_center}] plot center of selected clusters

\item[\code{show\_center\_label}] plot label of selected clusters

\item[\code{center\_point\_size}] size of center points

\item[\code{label\_size}] size of labels

\item[\code{label\_fontface}] font of labels

\item[\code{edge\_alpha}] column to use for alpha of the edges

\item[\code{point\_shape}] point with border or not (border or no\_border)

\item[\code{point\_size}] size of point (cell)

\item[\code{point\_border\_col}] color of border around points

\item[\code{point\_border\_stroke}] stroke size of border around points

\item[\code{title}] title for plot, defaults to cell\_color parameter

\item[\code{show\_legend}] show legend

\item[\code{legend\_text}] size of legend text

\item[\code{legend\_symbol\_size}] size of legend symbols

\item[\code{background\_color}] color of plot background

\item[\code{axis\_text}] size of axis text

\item[\code{axis\_title}] size of axis title

\item[\code{show\_plot}] show plot

\item[\code{return\_plot}] return ggplot object

\item[\code{save\_plot}] directly save the plot [boolean]

\item[\code{save\_param}] list of saving parameters from \code{\LinkA{all\_plots\_save\_function}{all.Rul.plots.Rul.save.Rul.function}}

\item[\code{default\_save\_name}] default save name for saving, don't change, change save\_name in save\_param
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Description of parameters. For 3D plots see \code{\LinkA{dimPlot3D}{dimPlot3D}}
\end{Details}
%
\begin{Value}
ggplot
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    dimPlot2D_single(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{dimPlot3D}{dimPlot3D}{dimPlot3D}
%
\begin{Description}\relax
Visualize cells according to dimension reduction coordinates
\end{Description}
%
\begin{Usage}
\begin{verbatim}
dimPlot3D(
  gobject,
  dim_reduction_to_use = "umap",
  dim_reduction_name = "umap",
  dim1_to_use = 1,
  dim2_to_use = 2,
  dim3_to_use = 3,
  select_cell_groups = NULL,
  select_cells = NULL,
  show_other_cells = T,
  other_cell_color = "lightgrey",
  other_point_size = 2,
  show_NN_network = F,
  nn_network_to_use = "sNN",
  network_name = "sNN.pca",
  color_as_factor = T,
  cell_color = NULL,
  cell_color_code = NULL,
  show_cluster_center = F,
  show_center_label = T,
  center_point_size = 4,
  label_size = 4,
  edge_alpha = NULL,
  point_size = 3,
  show_plot = NA,
  return_plot = NA,
  save_plot = NA,
  save_param = list(),
  default_save_name = "dim3D"
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{dim\_reduction\_to\_use}] dimension reduction to use

\item[\code{dim\_reduction\_name}] dimension reduction name

\item[\code{dim1\_to\_use}] dimension to use on x-axis

\item[\code{dim2\_to\_use}] dimension to use on y-axis

\item[\code{dim3\_to\_use}] dimension to use on z-axis

\item[\code{select\_cell\_groups}] select subset of cells/clusters based on cell\_color parameter

\item[\code{select\_cells}] select subset of cells based on cell IDs

\item[\code{show\_other\_cells}] display not selected cells

\item[\code{other\_cell\_color}] color of not selected cells

\item[\code{other\_point\_size}] size of not selected cells

\item[\code{show\_NN\_network}] show underlying NN network

\item[\code{nn\_network\_to\_use}] type of NN network to use (kNN vs sNN)

\item[\code{network\_name}] name of NN network to use, if show\_NN\_network = TRUE

\item[\code{color\_as\_factor}] convert color column to factor

\item[\code{cell\_color}] color for cells (see details)

\item[\code{cell\_color\_code}] named vector with colors

\item[\code{show\_cluster\_center}] plot center of selected clusters

\item[\code{show\_center\_label}] plot label of selected clusters

\item[\code{center\_point\_size}] size of center points

\item[\code{label\_size}] size of labels

\item[\code{edge\_alpha}] column to use for alpha of the edges

\item[\code{point\_size}] size of point (cell)

\item[\code{show\_plot}] show plot

\item[\code{return\_plot}] return ggplot object

\item[\code{save\_plot}] directly save the plot [boolean]

\item[\code{save\_param}] list of saving parameters from \code{\LinkA{all\_plots\_save\_function}{all.Rul.plots.Rul.save.Rul.function}}

\item[\code{default\_save\_name}] default save name for saving, don't change, change save\_name in save\_param

\item[\code{show\_legend}] show legend
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Description of parameters.
\end{Details}
%
\begin{Value}
plotly
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    dimPlot3D(gobject)

\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{doHclust}{doHclust}{doHclust}
%
\begin{Description}\relax
cluster cells using hierarchical clustering algorithm
\end{Description}
%
\begin{Usage}
\begin{verbatim}
doHclust(
  gobject,
  expression_values = c("normalized", "scaled", "custom"),
  genes_to_use = NULL,
  dim_reduction_to_use = c("cells", "pca", "umap", "tsne"),
  dim_reduction_name = "pca",
  dimensions_to_use = 1:10,
  distance_method = c("pearson", "spearman", "original", "euclidean", "maximum",
    "manhattan", "canberra", "binary", "minkowski"),
  agglomeration_method = c("ward.D2", "ward.D", "single", "complete", "average",
    "mcquitty", "median", "centroid"),
  k = 10,
  h = NULL,
  name = "hclust",
  return_gobject = TRUE,
  set_seed = T,
  seed_number = 1234
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{expression\_values}] expression values to use

\item[\code{genes\_to\_use}] subset of genes to use

\item[\code{dim\_reduction\_to\_use}] dimension reduction to use

\item[\code{dim\_reduction\_name}] dimensions reduction name

\item[\code{dimensions\_to\_use}] dimensions to use

\item[\code{distance\_method}] distance method

\item[\code{agglomeration\_method}] agglomeration method for hclust

\item[\code{k}] number of final clusters

\item[\code{h}] cut hierarchical tree at height = h

\item[\code{name}] name for hierarchical clustering

\item[\code{return\_gobject}] boolean: return giotto object (default = TRUE)

\item[\code{set\_seed}] set seed

\item[\code{seed\_number}] number for seed
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Description on how to use Kmeans clustering method.
\end{Details}
%
\begin{Value}
giotto object with new clusters appended to cell metadata
\end{Value}
%
\begin{SeeAlso}\relax
\code{\LinkA{hclust}{hclust}}
\end{SeeAlso}
%
\begin{Examples}
\begin{ExampleCode}
    doHclust(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{doHMRF}{doHMRF}{doHMRF}
%
\begin{Description}\relax
Run HMRF
\end{Description}
%
\begin{Usage}
\begin{verbatim}
doHMRF(
  gobject,
  expression_values = c("normalized", "scaled", "custom"),
  spatial_network_name = "Delaunay_network",
  spatial_genes = NULL,
  spatial_dimensions = c("sdimx", "sdimy", "sdimz"),
  dim_reduction_to_use = NULL,
  dim_reduction_name = "pca",
  dimensions_to_use = 1:10,
  name = "test",
  k = 10,
  betas = c(0, 2, 50),
  tolerance = 1e-10,
  zscore = c("none", "rowcol", "colrow"),
  numinit = 100,
  python_path = NULL,
  output_folder = NULL,
  overwrite_output = TRUE
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{expression\_values}] expression values to use

\item[\code{spatial\_network\_name}] name of spatial network to use for HMRF

\item[\code{spatial\_genes}] spatial genes to use for HMRF

\item[\code{spatial\_dimensions}] select spatial dimensions to use, default is all possible dimensions

\item[\code{dim\_reduction\_to\_use}] use another dimension reduction set as input

\item[\code{dim\_reduction\_name}] name of dimension reduction set to use

\item[\code{dimensions\_to\_use}] number of dimensions to use as input

\item[\code{name}] name of HMRF run

\item[\code{k}] number of HMRF domains

\item[\code{betas}] betas to test for

\item[\code{tolerance}] tolerance

\item[\code{zscore}] zscore

\item[\code{numinit}] number of initializations

\item[\code{python\_path}] python path to use

\item[\code{output\_folder}] output folder to save results

\item[\code{overwrite\_output}] overwrite output folder
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Description of HMRF parameters ...
\end{Details}
%
\begin{Value}
Creates a directory with results that can be viewed with viewHMRFresults
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    doHMRF(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{doKmeans}{doKmeans}{doKmeans}
%
\begin{Description}\relax
cluster cells using kmeans algorithm
\end{Description}
%
\begin{Usage}
\begin{verbatim}
doKmeans(
  gobject,
  expression_values = c("normalized", "scaled", "custom"),
  genes_to_use = NULL,
  dim_reduction_to_use = c("cells", "pca", "umap", "tsne"),
  dim_reduction_name = "pca",
  dimensions_to_use = 1:10,
  distance_method = c("original", "pearson", "spearman", "euclidean", "maximum",
    "manhattan", "canberra", "binary", "minkowski"),
  centers = 10,
  iter_max = 100,
  nstart = 1000,
  algorithm = "Hartigan-Wong",
  name = "kmeans",
  return_gobject = TRUE,
  set_seed = T,
  seed_number = 1234
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{expression\_values}] expression values to use

\item[\code{genes\_to\_use}] subset of genes to use

\item[\code{dim\_reduction\_to\_use}] dimension reduction to use

\item[\code{dim\_reduction\_name}] dimensions reduction name

\item[\code{dimensions\_to\_use}] dimensions to use

\item[\code{distance\_method}] distance method

\item[\code{centers}] number of final clusters

\item[\code{iter\_max}] kmeans maximum iterations

\item[\code{nstart}] kmeans nstart

\item[\code{algorithm}] kmeans algorithm

\item[\code{name}] name for kmeans clustering

\item[\code{return\_gobject}] boolean: return giotto object (default = TRUE)

\item[\code{set\_seed}] set seed

\item[\code{seed\_number}] number for seed
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Description on how to use Kmeans clustering method.
\end{Details}
%
\begin{Value}
giotto object with new clusters appended to cell metadata
\end{Value}
%
\begin{SeeAlso}\relax
\code{\LinkA{kmeans}{kmeans}}
\end{SeeAlso}
%
\begin{Examples}
\begin{ExampleCode}
    doKmeans(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{doLeidenCluster}{doLeidenCluster}{doLeidenCluster}
%
\begin{Description}\relax
cluster cells using a NN-network and the Leiden community detection algorithm
\end{Description}
%
\begin{Usage}
\begin{verbatim}
doLeidenCluster(
  gobject,
  name = "leiden_clus",
  nn_network_to_use = "sNN",
  network_name = "sNN.pca",
  python_path = NULL,
  resolution = 1,
  weight_col = "weight",
  partition_type = c("RBConfigurationVertexPartition", "ModularityVertexPartition"),
  init_membership = NULL,
  n_iterations = 1000,
  return_gobject = TRUE,
  set_seed = T,
  seed_number = 1234,
  ...
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{name}] name for cluster

\item[\code{nn\_network\_to\_use}] type of NN network to use (kNN vs sNN)

\item[\code{network\_name}] name of NN network to use

\item[\code{python\_path}] specify specific path to python if required

\item[\code{resolution}] resolution

\item[\code{weight\_col}] weight column to use for edges

\item[\code{partition\_type}] The type of partition to use for optimisation.

\item[\code{init\_membership}] initial membership of cells for the partition

\item[\code{n\_iterations}] number of interations to run the Leiden algorithm.
If the number of iterations is negative, the Leiden algorithm is run until
an iteration in which there was no improvement.

\item[\code{return\_gobject}] boolean: return giotto object (default = TRUE)

\item[\code{set\_seed}] set seed

\item[\code{seed\_number}] number for seed
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
This function is a wrapper for the Leiden algorithm implemented in python,
which can detect communities in graphs of millions of nodes (cells),
as long as they can fit in memory. See the \url{https://github.com/vtraag/leidenalg}leidenalg
github page or the \url{https://leidenalg.readthedocs.io/en/stable/index.html}readthedocs
page for more information.

Partition types available and information:
\begin{itemize}

\item{} RBConfigurationVertexPartition: Implements Reichardt and Bornholdts Potts model
with a configuration null model. This quality function is well-defined only for positive edge weights.
This quality function uses a linear resolution parameter.
\item{} ModularityVertexPartition: Implements modularity.
This quality function is well-defined only for positive edge weights. It does \emph{not} use the resolution parameter

\end{itemize}


Set \emph{weight\_col = NULL} to give equal weight (=1) to each edge.
\end{Details}
%
\begin{Value}
giotto object with new clusters appended to cell metadata
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    doLeidenCluster(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{doLeidenSubCluster}{doLeidenSubCluster}{doLeidenSubCluster}
%
\begin{Description}\relax
Further subcluster cells using a NN-network and the Leiden algorithm
\end{Description}
%
\begin{Usage}
\begin{verbatim}
doLeidenSubCluster(
  gobject,
  name = "sub_pleiden_clus",
  cluster_column = NULL,
  selected_clusters = NULL,
  hvg_param = list(reverse_log_scale = T, difference_in_variance = 1, expression_values
    = "normalized"),
  hvg_min_perc_cells = 5,
  hvg_mean_expr_det = 1,
  use_all_genes_as_hvg = FALSE,
  min_nr_of_hvg = 5,
  pca_param = list(expression_values = "normalized", scale_unit = T),
  nn_param = list(dimensions_to_use = 1:20),
  k_neighbors = 10,
  resolution = 0.5,
  n_iterations = 500,
  python_path = NULL,
  nn_network_to_use = "sNN",
  network_name = "sNN.pca",
  return_gobject = TRUE,
  verbose = T
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{name}] name for new clustering result

\item[\code{cluster\_column}] cluster column to subcluster

\item[\code{selected\_clusters}] only do subclustering on these clusters

\item[\code{hvg\_param}] parameters for calculateHVG

\item[\code{hvg\_min\_perc\_cells}] threshold for detection in min percentage of cells

\item[\code{hvg\_mean\_expr\_det}] threshold for mean expression level in cells with detection

\item[\code{use\_all\_genes\_as\_hvg}] forces all genes to be HVG and to be used as input for PCA

\item[\code{min\_nr\_of\_hvg}] minimum number of HVG, or all genes will be used as input for PCA

\item[\code{pca\_param}] parameters for runPCA

\item[\code{nn\_param}] parameters for parameters for createNearestNetwork

\item[\code{k\_neighbors}] number of k for createNearestNetwork

\item[\code{resolution}] resolution of Leiden clustering

\item[\code{n\_iterations}] number of interations to run the Leiden algorithm.

\item[\code{python\_path}] specify specific path to python if required

\item[\code{nn\_network\_to\_use}] type of NN network to use (kNN vs sNN)

\item[\code{network\_name}] name of NN network to use

\item[\code{return\_gobject}] boolean: return giotto object (default = TRUE)

\item[\code{verbose}] verbose
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
This function performs subclustering using the Leiden algorithm on selected clusters.
The systematic steps are:
\begin{itemize}

\item{} 1. subset Giotto object
\item{} 2. identify highly variable genes
\item{} 3. run PCA
\item{} 4. create nearest neighbouring network
\item{} 5. do Leiden clustering

\end{itemize}

\end{Details}
%
\begin{Value}
giotto object with new subclusters appended to cell metadata
\end{Value}
%
\begin{SeeAlso}\relax
\code{\LinkA{doLeidenCluster}{doLeidenCluster}}
\end{SeeAlso}
%
\begin{Examples}
\begin{ExampleCode}
    doLeidenSubCluster(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{doLouvainCluster}{doLouvainCluster}{doLouvainCluster}
%
\begin{Description}\relax
cluster cells using a NN-network and the Louvain algorithm.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
doLouvainCluster(
  gobject,
  version = c("community", "multinet"),
  name = "louvain_clus",
  nn_network_to_use = "sNN",
  network_name = "sNN.pca",
  python_path = NULL,
  resolution = 1,
  weight_col = NULL,
  gamma = 1,
  omega = 1,
  louv_random = F,
  return_gobject = TRUE,
  set_seed = F,
  seed_number = 1234,
  ...
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{version}] implemented version of Louvain clustering to use

\item[\code{name}] name for cluster

\item[\code{nn\_network\_to\_use}] type of NN network to use (kNN vs sNN)

\item[\code{network\_name}] name of NN network to use

\item[\code{python\_path}] [community] specify specific path to python if required

\item[\code{resolution}] [community] resolution

\item[\code{gamma}] [multinet] Resolution parameter for modularity in the generalized louvain method.

\item[\code{omega}] [multinet] Inter-layer weight parameter in the generalized louvain method.

\item[\code{return\_gobject}] boolean: return giotto object (default = TRUE)

\item[\code{set\_seed}] set seed

\item[\code{seed\_number}] number for seed
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Louvain clustering using the community or multinet implementation of the louvain clustering algorithm.
\end{Details}
%
\begin{Value}
giotto object with new clusters appended to cell metadata
\end{Value}
%
\begin{SeeAlso}\relax
\code{\LinkA{doLouvainCluster\_community}{doLouvainCluster.Rul.community}} and \code{\LinkA{doLouvainCluster\_multinet}{doLouvainCluster.Rul.multinet}}
\end{SeeAlso}
%
\begin{Examples}
\begin{ExampleCode}
    doLouvainCluster(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{doLouvainCluster\_community}{doLouvainCluster\_community}{doLouvainCluster.Rul.community}
%
\begin{Description}\relax
cluster cells using a NN-network and the Louvain algorithm from the community module in Python
\end{Description}
%
\begin{Usage}
\begin{verbatim}
doLouvainCluster_community(
  gobject,
  name = "louvain_clus",
  nn_network_to_use = "sNN",
  network_name = "sNN.pca",
  python_path = NULL,
  resolution = 1,
  weight_col = NULL,
  louv_random = F,
  return_gobject = TRUE,
  set_seed = F,
  seed_number = 1234,
  ...
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{name}] name for cluster

\item[\code{nn\_network\_to\_use}] type of NN network to use (kNN vs sNN)

\item[\code{network\_name}] name of NN network to use

\item[\code{python\_path}] specify specific path to python if required

\item[\code{resolution}] resolution

\item[\code{weight\_col}] weight column to use for edges

\item[\code{louv\_random}] Will randomize the node evaluation order and the community evaluation
order to get different partitions at each call

\item[\code{return\_gobject}] boolean: return giotto object (default = TRUE)

\item[\code{set\_seed}] set seed

\item[\code{seed\_number}] number for seed
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
This function is a wrapper for the Louvain algorithm implemented in Python,
which can detect communities in graphs of nodes (cells).
See the \url{https://python-louvain.readthedocs.io/en/latest/index.html}readthedocs
page for more information.

Set \emph{weight\_col = NULL} to give equal weight (=1) to each edge.
\end{Details}
%
\begin{Value}
giotto object with new clusters appended to cell metadata
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    doLouvainCluster_community(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{doLouvainCluster\_multinet}{doLouvainCluster\_multinet}{doLouvainCluster.Rul.multinet}
%
\begin{Description}\relax
cluster cells using a NN-network and the Louvain algorithm from the multinet package in R.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
doLouvainCluster_multinet(
  gobject,
  name = "louvain_clus",
  nn_network_to_use = "sNN",
  network_name = "sNN.pca",
  gamma = 1,
  omega = 1,
  return_gobject = TRUE,
  set_seed = F,
  seed_number = 1234,
  ...
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{name}] name for cluster

\item[\code{nn\_network\_to\_use}] type of NN network to use (kNN vs sNN)

\item[\code{network\_name}] name of NN network to use

\item[\code{gamma}] Resolution parameter for modularity in the generalized louvain method.

\item[\code{omega}] Inter-layer weight parameter in the generalized louvain method.

\item[\code{return\_gobject}] boolean: return giotto object (default = TRUE)

\item[\code{set\_seed}] set seed

\item[\code{seed\_number}] number for seed
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
See \code{\LinkA{glouvain\_ml}{glouvain.Rul.ml}} from the multinet package in R for
more information.
\end{Details}
%
\begin{Value}
giotto object with new clusters appended to cell metadata
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    doLouvainCluster_multinet(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{doLouvainSubCluster}{doLouvainSubCluster}{doLouvainSubCluster}
%
\begin{Description}\relax
subcluster cells using a NN-network and the Louvain algorithm
\end{Description}
%
\begin{Usage}
\begin{verbatim}
doLouvainSubCluster(
  gobject,
  name = "sub_louvain_clus",
  version = c("community", "multinet"),
  cluster_column = NULL,
  selected_clusters = NULL,
  hvg_param = list(reverse_log_scale = T, difference_in_variance = 1, expression_values
    = "normalized"),
  hvg_min_perc_cells = 5,
  hvg_mean_expr_det = 1,
  use_all_genes_as_hvg = FALSE,
  min_nr_of_hvg = 5,
  pca_param = list(expression_values = "normalized", scale_unit = T),
  nn_param = list(dimensions_to_use = 1:20),
  k_neighbors = 10,
  resolution = 0.5,
  gamma = 1,
  omega = 1,
  python_path = NULL,
  nn_network_to_use = "sNN",
  network_name = "sNN.pca",
  return_gobject = TRUE,
  verbose = T
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{name}] name for new clustering result

\item[\code{version}] version of Louvain algorithm to use

\item[\code{cluster\_column}] cluster column to subcluster

\item[\code{selected\_clusters}] only do subclustering on these clusters

\item[\code{hvg\_param}] parameters for calculateHVG

\item[\code{hvg\_min\_perc\_cells}] threshold for detection in min percentage of cells

\item[\code{hvg\_mean\_expr\_det}] threshold for mean expression level in cells with detection

\item[\code{use\_all\_genes\_as\_hvg}] forces all genes to be HVG and to be used as input for PCA

\item[\code{min\_nr\_of\_hvg}] minimum number of HVG, or all genes will be used as input for PCA

\item[\code{pca\_param}] parameters for runPCA

\item[\code{nn\_param}] parameters for parameters for createNearestNetwork

\item[\code{k\_neighbors}] number of k for createNearestNetwork

\item[\code{resolution}] resolution for community algorithm

\item[\code{gamma}] gamma

\item[\code{omega}] omega

\item[\code{python\_path}] specify specific path to python if required

\item[\code{nn\_network\_to\_use}] type of NN network to use (kNN vs sNN)

\item[\code{network\_name}] name of NN network to use

\item[\code{return\_gobject}] boolean: return giotto object (default = TRUE)

\item[\code{verbose}] verbose
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
This function performs subclustering using the Louvain algorithm on selected clusters.
The systematic steps are:
\begin{itemize}

\item{} 1. subset Giotto object
\item{} 2. identify highly variable genes
\item{} 3. run PCA
\item{} 4. create nearest neighbouring network
\item{} 5. do Louvain clustering

\end{itemize}

\end{Details}
%
\begin{Value}
giotto object with new subclusters appended to cell metadata
\end{Value}
%
\begin{SeeAlso}\relax
\code{\LinkA{doLouvainCluster\_multinet}{doLouvainCluster.Rul.multinet}} and \code{\LinkA{doLouvainCluster\_community}{doLouvainCluster.Rul.community}}
\end{SeeAlso}
%
\begin{Examples}
\begin{ExampleCode}
    doLouvainSubCluster(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{doLouvainSubCluster\_community}{doLouvainSubCluster\_community}{doLouvainSubCluster.Rul.community}
%
\begin{Description}\relax
subcluster cells using a NN-network and the Louvain community detection algorithm
\end{Description}
%
\begin{Usage}
\begin{verbatim}
doLouvainSubCluster_community(
  gobject,
  name = "sub_louvain_comm_clus",
  cluster_column = NULL,
  selected_clusters = NULL,
  hvg_param = list(reverse_log_scale = T, difference_in_variance = 1, expression_values
    = "normalized"),
  hvg_min_perc_cells = 5,
  hvg_mean_expr_det = 1,
  use_all_genes_as_hvg = FALSE,
  min_nr_of_hvg = 5,
  pca_param = list(expression_values = "normalized", scale_unit = T),
  nn_param = list(dimensions_to_use = 1:20),
  k_neighbors = 10,
  resolution = 0.5,
  python_path = NULL,
  nn_network_to_use = "sNN",
  network_name = "sNN.pca",
  return_gobject = TRUE,
  verbose = T
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{name}] name for new clustering result

\item[\code{cluster\_column}] cluster column to subcluster

\item[\code{selected\_clusters}] only do subclustering on these clusters

\item[\code{hvg\_param}] parameters for calculateHVG

\item[\code{hvg\_min\_perc\_cells}] threshold for detection in min percentage of cells

\item[\code{hvg\_mean\_expr\_det}] threshold for mean expression level in cells with detection

\item[\code{use\_all\_genes\_as\_hvg}] forces all genes to be HVG and to be used as input for PCA

\item[\code{min\_nr\_of\_hvg}] minimum number of HVG, or all genes will be used as input for PCA

\item[\code{pca\_param}] parameters for runPCA

\item[\code{nn\_param}] parameters for parameters for createNearestNetwork

\item[\code{k\_neighbors}] number of k for createNearestNetwork

\item[\code{resolution}] resolution

\item[\code{python\_path}] specify specific path to python if required

\item[\code{nn\_network\_to\_use}] type of NN network to use (kNN vs sNN)

\item[\code{network\_name}] name of NN network to use

\item[\code{return\_gobject}] boolean: return giotto object (default = TRUE)

\item[\code{verbose}] verbose
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
This function performs subclustering using the Louvain community algorithm on selected clusters.
The systematic steps are:
\begin{itemize}

\item{} 1. subset Giotto object
\item{} 2. identify highly variable genes
\item{} 3. run PCA
\item{} 4. create nearest neighbouring network
\item{} 5. do Louvain community clustering

\end{itemize}

\end{Details}
%
\begin{Value}
giotto object with new subclusters appended to cell metadata
\end{Value}
%
\begin{SeeAlso}\relax
\code{\LinkA{doLouvainCluster\_community}{doLouvainCluster.Rul.community}}
\end{SeeAlso}
%
\begin{Examples}
\begin{ExampleCode}
    doLouvainSubCluster_community(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{doLouvainSubCluster\_multinet}{doLouvainSubCluster\_multinet}{doLouvainSubCluster.Rul.multinet}
%
\begin{Description}\relax
subcluster cells using a NN-network and the Louvain multinet detection algorithm
\end{Description}
%
\begin{Usage}
\begin{verbatim}
doLouvainSubCluster_multinet(
  gobject,
  name = "sub_louvain_mult_clus",
  cluster_column = NULL,
  selected_clusters = NULL,
  hvg_param = list(reverse_log_scale = T, difference_in_variance = 1, expression_values
    = "normalized"),
  hvg_min_perc_cells = 5,
  hvg_mean_expr_det = 1,
  use_all_genes_as_hvg = FALSE,
  min_nr_of_hvg = 5,
  pca_param = list(expression_values = "normalized", scale_unit = T),
  nn_param = list(dimensions_to_use = 1:20),
  k_neighbors = 10,
  gamma = 1,
  omega = 1,
  nn_network_to_use = "sNN",
  network_name = "sNN.pca",
  return_gobject = TRUE,
  verbose = T
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{name}] name for new clustering result

\item[\code{cluster\_column}] cluster column to subcluster

\item[\code{selected\_clusters}] only do subclustering on these clusters

\item[\code{hvg\_param}] parameters for calculateHVG

\item[\code{hvg\_min\_perc\_cells}] threshold for detection in min percentage of cells

\item[\code{hvg\_mean\_expr\_det}] threshold for mean expression level in cells with detection

\item[\code{use\_all\_genes\_as\_hvg}] forces all genes to be HVG and to be used as input for PCA

\item[\code{min\_nr\_of\_hvg}] minimum number of HVG, or all genes will be used as input for PCA

\item[\code{pca\_param}] parameters for runPCA

\item[\code{nn\_param}] parameters for parameters for createNearestNetwork

\item[\code{k\_neighbors}] number of k for createNearestNetwork

\item[\code{gamma}] gamma

\item[\code{omega}] omega

\item[\code{nn\_network\_to\_use}] type of NN network to use (kNN vs sNN)

\item[\code{network\_name}] name of NN network to use

\item[\code{return\_gobject}] boolean: return giotto object (default = TRUE)

\item[\code{verbose}] verbose

\item[\code{python\_path}] specify specific path to python if required
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
This function performs subclustering using the Louvain multinet algorithm on selected clusters.
The systematic steps are:
\begin{itemize}

\item{} 1. subset Giotto object
\item{} 2. identify highly variable genes
\item{} 3. run PCA
\item{} 4. create nearest neighbouring network
\item{} 5. do Louvain multinet clustering

\end{itemize}

\end{Details}
%
\begin{Value}
giotto object with new subclusters appended to cell metadata
\end{Value}
%
\begin{SeeAlso}\relax
\code{\LinkA{doLouvainCluster\_multinet}{doLouvainCluster.Rul.multinet}}
\end{SeeAlso}
%
\begin{Examples}
\begin{ExampleCode}
    doLouvainSubCluster_multinet(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{doRandomWalkCluster}{doRandomWalkCluster}{doRandomWalkCluster}
%
\begin{Description}\relax
Cluster cells using a random walk approach.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
doRandomWalkCluster(
  gobject,
  name = "random_walk_clus",
  nn_network_to_use = "sNN",
  network_name = "sNN.pca",
  walk_steps = 4,
  walk_clusters = 10,
  walk_weights = NA,
  return_gobject = TRUE,
  set_seed = F,
  seed_number = 1234,
  ...
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{name}] name for cluster

\item[\code{nn\_network\_to\_use}] type of NN network to use (kNN vs sNN)

\item[\code{network\_name}] name of NN network to use

\item[\code{walk\_steps}] number of walking steps

\item[\code{walk\_clusters}] number of final clusters

\item[\code{walk\_weights}] cluster column defining the walk weights

\item[\code{return\_gobject}] boolean: return giotto object (default = TRUE)

\item[\code{set\_seed}] set seed

\item[\code{seed\_number}] number for seed
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
See \code{\LinkA{cluster\_walktrap}{cluster.Rul.walktrap}} function from the igraph
package in R for more information.
\end{Details}
%
\begin{Value}
giotto object with new clusters appended to cell metadata
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    doRandomWalkCluster(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{doSNNCluster}{doSNNCluster}{doSNNCluster}
%
\begin{Description}\relax
Cluster cells using a SNN cluster approach.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
doSNNCluster(
  gobject,
  name = "sNN_clus",
  nn_network_to_use = "kNN",
  network_name = "kNN.pca",
  k = 20,
  eps = 4,
  minPts = 16,
  borderPoints = TRUE,
  return_gobject = TRUE,
  set_seed = F,
  seed_number = 1234,
  ...
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{name}] name for cluster

\item[\code{nn\_network\_to\_use}] type of NN network to use (only works on kNN)

\item[\code{network\_name}] name of kNN network to use

\item[\code{k}] Neighborhood size for nearest neighbor sparsification to create the shared NN graph.

\item[\code{eps}] Two objects are only reachable from each other if they share at least eps nearest neighbors.

\item[\code{minPts}] minimum number of points that share at least eps nearest neighbors for a point to be considered a core points.

\item[\code{borderPoints}] should borderPoints be assigned to clusters like in DBSCAN?

\item[\code{return\_gobject}] boolean: return giotto object (default = TRUE)

\item[\code{set\_seed}] set seed

\item[\code{seed\_number}] number for seed
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
See \code{\LinkA{sNNclust}{sNNclust}} from dbscan package
\end{Details}
%
\begin{Value}
giotto object with new clusters appended to cell metadata
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    doSNNCluster(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{do\_cell\_proximity\_test}{do\_cell\_proximity\_test}{do.Rul.cell.Rul.proximity.Rul.test}
%
\begin{Description}\relax
Performs a selected differential test on subsets of a matrix
\end{Description}
%
\begin{Usage}
\begin{verbatim}
do_cell_proximity_test(
  expr_values,
  select_ind,
  other_ind,
  diff_test = c("permutation", "limma", "t.test", "wilcox"),
  mean_method = c("arithmic", "geometric"),
  offset = 0.1,
  n_perm = 100,
  adjust_method = c("bonferroni", "BH", "holm", "hochberg", "hommel", "BY", "fdr",
    "none"),
  cores = 2
)
\end{verbatim}
\end{Usage}
%
\begin{Examples}
\begin{ExampleCode}
    do_cell_proximity_test()
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{do\_limmatest}{do\_limmatest}{do.Rul.limmatest}
%
\begin{Description}\relax
Performs limma t.test on subsets of a matrix
\end{Description}
%
\begin{Usage}
\begin{verbatim}
do_limmatest(expr_values, select_ind, other_ind, mean_method, offset = 0.1)
\end{verbatim}
\end{Usage}
%
\begin{Examples}
\begin{ExampleCode}
    do_limmatest()
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{do\_multi\_permuttest\_random}{do\_multi\_permuttest\_random}{do.Rul.multi.Rul.permuttest.Rul.random}
%
\begin{Description}\relax
calculate multiple random values
\end{Description}
%
\begin{Usage}
\begin{verbatim}
do_multi_permuttest_random(
  expr_values,
  select_ind,
  other_ind,
  mean_method,
  offset = 0.1,
  n = 100,
  cores = 2
)
\end{verbatim}
\end{Usage}
%
\begin{Examples}
\begin{ExampleCode}
    do_multi_permuttest_random()
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{do\_page\_permutation}{do\_page\_permutation}{do.Rul.page.Rul.permutation}
%
\begin{Description}\relax
creates permutation for the PAGEEnrich test
\end{Description}
%
\begin{Usage}
\begin{verbatim}
do_page_permutation(gobject, sig_gene, ntimes)
\end{verbatim}
\end{Usage}
%
\begin{Examples}
\begin{ExampleCode}
    do_page_permutation()
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{do\_permuttest\_original}{do\_permuttest\_original}{do.Rul.permuttest.Rul.original}
%
\begin{Description}\relax
calculate original values
\end{Description}
%
\begin{Usage}
\begin{verbatim}
do_permuttest_original(
  expr_values,
  select_ind,
  other_ind,
  name = "orig",
  mean_method,
  offset = 0.1
)
\end{verbatim}
\end{Usage}
%
\begin{Examples}
\begin{ExampleCode}
    do_permuttest_original()
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{do\_permuttest\_random}{do\_permuttest\_random}{do.Rul.permuttest.Rul.random}
\aliasA{do\_permuttest}{do\_permuttest\_random}{do.Rul.permuttest}
%
\begin{Description}\relax
calculate random values

Performs permutation test on subsets of a matrix
\end{Description}
%
\begin{Usage}
\begin{verbatim}
do_permuttest_random(
  expr_values,
  select_ind,
  other_ind,
  name = "perm_1",
  mean_method,
  offset = 0.1
)

do_permuttest(
  expr_values,
  select_ind,
  other_ind,
  n_perm = 1000,
  adjust_method = "fdr",
  mean_method,
  offset = 0.1,
  cores = 2
)
\end{verbatim}
\end{Usage}
%
\begin{Examples}
\begin{ExampleCode}
    do_permuttest_random()
    do_permuttest_random()
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{do\_rank\_permutation}{do\_rank\_permutation}{do.Rul.rank.Rul.permutation}
%
\begin{Description}\relax
creates permutation for the rankEnrich test
\end{Description}
%
\begin{Usage}
\begin{verbatim}
do_rank_permutation(sc_gene, n)
\end{verbatim}
\end{Usage}
%
\begin{Examples}
\begin{ExampleCode}
    do_rank_permutation()
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{do\_spatial\_grid\_averaging}{do\_spatial\_grid\_averaging}{do.Rul.spatial.Rul.grid.Rul.averaging}
%
\begin{Description}\relax
smooth gene expression over a defined spatial grid
\end{Description}
%
\begin{Usage}
\begin{verbatim}
do_spatial_grid_averaging(
  gobject,
  expression_values = c("normalized", "scaled", "custom"),
  subset_genes = NULL,
  spatial_grid_name = "spatial_grid",
  min_cells_per_grid = 4
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{expression\_values}] gene expression values to use

\item[\code{subset\_genes}] subset of genes to use

\item[\code{spatial\_grid\_name}] name of spatial grid to use

\item[\code{min\_cells\_per\_grid}] minimum number of cells to consider a grid
\end{ldescription}
\end{Arguments}
%
\begin{Value}
matrix with smoothened gene expression values based on spatial grid
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    do_spatial_grid_averaging(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{do\_spatial\_knn\_smoothing}{do\_spatial\_knn\_smoothing}{do.Rul.spatial.Rul.knn.Rul.smoothing}
%
\begin{Description}\relax
smooth gene expression over a kNN spatial network
\end{Description}
%
\begin{Usage}
\begin{verbatim}
do_spatial_knn_smoothing(
  gobject,
  expression_values = c("normalized", "scaled", "custom"),
  subset_genes = NULL,
  spatial_network_name = "Delaunay_network",
  b = NULL
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{expression\_values}] gene expression values to use

\item[\code{subset\_genes}] subset of genes to use

\item[\code{spatial\_network\_name}] name of spatial network to use

\item[\code{b}] smoothing factor beteen 0 and 1 (default: automatic)
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
This function will smoothen the gene expression values per cell according to
its neighbors in the selected spatial network. \\{}
b is a smoothening factor that defaults to 1 - 1/k, where k is the median number of
k-neighbors in the selected spatial network. Setting b = 0 means no smoothing and b = 1
means no contribution from its own expression.
\end{Details}
%
\begin{Value}
matrix with smoothened gene expression values based on kNN spatial network
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    do_spatial_knn_smoothing(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{do\_ttest}{do\_ttest}{do.Rul.ttest}
\aliasA{do\_wilctest}{do\_ttest}{do.Rul.wilctest}
%
\begin{Description}\relax
Performs t.test on subsets of a matrix

Performs wilcoxon on subsets of a matrix
\end{Description}
%
\begin{Usage}
\begin{verbatim}
do_ttest(
  expr_values,
  select_ind,
  other_ind,
  adjust_method,
  mean_method,
  offset = 0.1
)

do_wilctest(
  expr_values,
  select_ind,
  other_ind,
  adjust_method,
  mean_method,
  offset = 0.1
)
\end{verbatim}
\end{Usage}
%
\begin{Examples}
\begin{ExampleCode}
    do_ttest()
    do_ttest()
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{DT\_removeNA}{DT\_removeNA}{DT.Rul.removeNA}
%
\begin{Description}\relax
set NA values to 0 in a data.table object
\end{Description}
%
\begin{Usage}
\begin{verbatim}
DT_removeNA(DT)
\end{verbatim}
\end{Usage}
\inputencoding{utf8}
\HeaderA{dt\_to\_matrix}{dt\_to\_matrix}{dt.Rul.to.Rul.matrix}
%
\begin{Description}\relax
converts data.table to matrix
\end{Description}
%
\begin{Usage}
\begin{verbatim}
dt_to_matrix(x)
\end{verbatim}
\end{Usage}
%
\begin{Examples}
\begin{ExampleCode}
    dt_to_matrix(x)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{estimateCellCellDistance}{estimateCellCellDistance}{estimateCellCellDistance}
%
\begin{Description}\relax
estimate average distance between neighboring cells
\end{Description}
%
\begin{Usage}
\begin{verbatim}
estimateCellCellDistance(
  gobject,
  spatial_network_name = "Delaunay_network",
  method = c("mean", "median")
)
\end{verbatim}
\end{Usage}
\inputencoding{utf8}
\HeaderA{evaluate\_expr\_matrix}{evaluate\_expr\_matrix}{evaluate.Rul.expr.Rul.matrix}
%
\begin{Description}\relax
Evaluate expression matrices.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
evaluate_expr_matrix(inputmatrix, sparse = TRUE, cores = NA)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{inputmatrix}] inputmatrix to evaluate
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
The inputmatrix can be a matrix, sparse matrix, data.frame, data.table or path to any of these.
\end{Details}
%
\begin{Value}
sparse matrix
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    evaluate_expr_matrix()
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{exportGiottoViewer}{exportGiottoViewer}{exportGiottoViewer}
%
\begin{Description}\relax
compute highly variable genes
\end{Description}
%
\begin{Usage}
\begin{verbatim}
exportGiottoViewer(
  gobject,
  output_directory = NULL,
  spat_enr_names = NULL,
  factor_annotations = NULL,
  numeric_annotations = NULL,
  dim_reductions,
  dim_reduction_names,
  expression_values = c("scaled", "normalized", "custom"),
  dim_red_rounding = NULL,
  dim_red_rescale = c(-20, 20),
  expression_rounding = 2,
  overwrite_dir = T,
  verbose = T
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{output\_directory}] directory where to save the files

\item[\code{spat\_enr\_names}] spatial enrichment results to include for annotations

\item[\code{factor\_annotations}] giotto cell annotations to view as factor

\item[\code{numeric\_annotations}] giotto cell annotations to view as numeric

\item[\code{dim\_reductions}] high level dimension reductions to view

\item[\code{dim\_reduction\_names}] specific dimension reduction names

\item[\code{expression\_values}] expression values to use in Viewer

\item[\code{dim\_red\_rounding}] numerical indicating how to round the coordinates

\item[\code{dim\_red\_rescale}] numericals to rescale the coordinates

\item[\code{expression\_rounding}] numerical indicating how to round the expression data

\item[\code{overwrite\_dir}] overwrite files in the directory if it already existed

\item[\code{verbose}] be verbose
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Giotto Viewer expects the results from Giotto Analyzer in a specific format,
which is provided by this function. To include enrichment results from \code{\LinkA{createSpatialEnrich}{createSpatialEnrich}}
include the provided spatial enrichment name (default PAGE or rank)
and add the gene signature names (.e.g cell types) to the numeric annotations parameter.
\end{Details}
%
\begin{Value}
writes the necessary output to use in Giotto Viewer
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    exportGiottoViewer(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{exprCellCellcom}{exprCellCellcom}{exprCellCellcom}
%
\begin{Description}\relax
Cell-Cell communication scores based on expression only
\end{Description}
%
\begin{Usage}
\begin{verbatim}
exprCellCellcom(
  gobject,
  cluster_column = "cell_types",
  random_iter = 1000,
  gene_set_1,
  gene_set_2,
  log2FC_addendum = 0.1,
  adjust_method = c("fdr", "bonferroni", "BH", "holm", "hochberg", "hommel", "BY",
    "none"),
  adjust_target = c("genes", "cells"),
  verbose = T
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object to use

\item[\code{cluster\_column}] cluster column with cell type information

\item[\code{random\_iter}] number of iterations

\item[\code{gene\_set\_1}] first specific gene set from gene pairs

\item[\code{gene\_set\_2}] second specific gene set from gene pairs

\item[\code{log2FC\_addendum}] addendum to add when calculating log2FC

\item[\code{adjust\_method}] which method to adjust p-values

\item[\code{adjust\_target}] adjust multiple hypotheses at the cell or gene level

\item[\code{verbose}] verbose
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Statistical framework to identify if pairs of genes (such as ligand-receptor combinations)
are expressed at higher levels than expected based on a reshuffled null distribution of gene expression values,
without considering the spatial position of cells.
More details will follow soon.
\end{Details}
%
\begin{Value}
Cell-Cell communication scores for gene pairs based on expression only
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    exprCellCellcom(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{extended\_gini\_fun}{extended\_gini\_fun}{extended.Rul.gini.Rul.fun}
%
\begin{Description}\relax
calculate gini coefficient on a minimum length vector
\end{Description}
%
\begin{Usage}
\begin{verbatim}
extended_gini_fun(x, weights = rep(1, length = length(x)), minimum_length = 16)
\end{verbatim}
\end{Usage}
%
\begin{Value}
gini coefficient
\end{Value}
\inputencoding{utf8}
\HeaderA{extend\_vector}{extend\_vector}{extend.Rul.vector}
%
\begin{Description}\relax
extend the range of a vector by a given ratio
\end{Description}
%
\begin{Usage}
\begin{verbatim}
extend_vector(x, extend_ratio)
\end{verbatim}
\end{Usage}
\inputencoding{utf8}
\HeaderA{extractNearestNetwork}{extractNearestNetwork}{extractNearestNetwork}
%
\begin{Description}\relax
Extracts a NN-network from a Giotto object
\end{Description}
%
\begin{Usage}
\begin{verbatim}
extractNearestNetwork(
  gobject,
  nn_network_to_use = "sNN",
  network_name = "sNN.pca",
  output = c("igraph", "data.table")
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{nn\_network\_to\_use}] kNN or sNN

\item[\code{network\_name}] name of NN network to be used

\item[\code{output}] return a igraph or data.table object
\end{ldescription}
\end{Arguments}
%
\begin{Value}
igraph or data.table object
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    extractNearestNetwork(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{fDataDT}{fDataDT}{fDataDT}
%
\begin{Description}\relax
show gene metadata
\end{Description}
%
\begin{Usage}
\begin{verbatim}
fDataDT(gobject)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object
\end{ldescription}
\end{Arguments}
%
\begin{Value}
data.table with gene metadata
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    pDataDT(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{filterCellProximityGenes}{filterCellProximityGenes}{filterCellProximityGenes}
%
\begin{Description}\relax
Filter cell proximity gene scores.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
filterCellProximityGenes(
  cpgObject,
  min_cells = 4,
  min_cells_expr = 1,
  min_int_cells = 4,
  min_int_cells_expr = 1,
  min_fdr = 0.1,
  min_spat_diff = 0.2,
  min_log2_fc = 0.2,
  min_zscore = 2,
  zscores_column = c("cell_type", "genes"),
  direction = c("both", "up", "down")
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{cpgObject}] cell proximity gene score object

\item[\code{min\_cells}] minimum number of source cell type

\item[\code{min\_cells\_expr}] minimum expression level for source cell type

\item[\code{min\_int\_cells}] minimum number of interacting neighbor cell type

\item[\code{min\_int\_cells\_expr}] minimum expression level for interacting neighbor cell type

\item[\code{min\_fdr}] minimum adjusted p-value

\item[\code{min\_spat\_diff}] minimum absolute spatial expression difference

\item[\code{min\_log2\_fc}] minimum log2 fold-change

\item[\code{min\_zscore}] minimum z-score change

\item[\code{zscores\_column}] calculate z-scores over cell types or genes

\item[\code{direction}] differential expression directions to keep
\end{ldescription}
\end{Arguments}
%
\begin{Value}
cpgObject that contains the filtered differential gene scores
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    filterCellProximityGenes(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{filterCombinations}{filterCombinations}{filterCombinations}
%
\begin{Description}\relax
Shows how many genes and cells are lost with combinations of thresholds.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
filterCombinations(
  gobject,
  expression_values = c("raw", "normalized", "scaled", "custom"),
  expression_thresholds = c(1, 2),
  gene_det_in_min_cells = c(5, 50),
  min_det_genes_per_cell = c(200, 400),
  scale_x_axis = "identity",
  x_axis_offset = 0,
  scale_y_axis = "identity",
  y_axis_offset = 0,
  show_plot = TRUE,
  return_plot = FALSE,
  save_plot = NA,
  save_param = list(),
  default_save_name = "filterCombinations"
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{expression\_values}] expression values to use

\item[\code{expression\_thresholds}] all thresholds to consider a gene expressed

\item[\code{gene\_det\_in\_min\_cells}] minimum number of cells that should express a gene to consider that gene further

\item[\code{min\_det\_genes\_per\_cell}] minimum number of expressed genes per cell to consider that cell further

\item[\code{scale\_x\_axis}] ggplot transformation for x-axis (e.g. log2)

\item[\code{x\_axis\_offset}] x-axis offset to be used together with the scaling transformation

\item[\code{scale\_y\_axis}] ggplot transformation for y-axis (e.g. log2)

\item[\code{y\_axis\_offset}] y-axis offset to be used together with the scaling transformation

\item[\code{show\_plot}] show plot

\item[\code{return\_plot}] return only ggplot object

\item[\code{save\_plot}] directly save the plot [boolean]

\item[\code{save\_param}] list of saving parameters from \code{\LinkA{all\_plots\_save\_function}{all.Rul.plots.Rul.save.Rul.function}}

\item[\code{default\_save\_name}] default save name for saving, don't change, change save\_name in save\_param
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Creates a scatterplot that visualizes the number of genes and cells that are
lost with a specific combination of a gene and cell threshold given an arbitrary cutoff
to call a gene expressed. This function can be used to make an informed decision at the
filtering step with filterGiotto.
\end{Details}
%
\begin{Value}
list of data.table and ggplot object
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    filterCombinations(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{filterCPG}{filterCPG}{filterCPG}
%
\begin{Description}\relax
Filter cell proximity gene scores.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
filterCPG(
  cpgObject,
  min_cells = 4,
  min_cells_expr = 1,
  min_int_cells = 4,
  min_int_cells_expr = 1,
  min_fdr = 0.1,
  min_spat_diff = 0.2,
  min_log2_fc = 0.2,
  min_zscore = 2,
  zscores_column = c("cell_type", "genes"),
  direction = c("both", "up", "down")
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{cpgObject}] cell proximity gene score object

\item[\code{min\_cells}] minimum number of source cell type

\item[\code{min\_cells\_expr}] minimum expression level for source cell type

\item[\code{min\_int\_cells}] minimum number of interacting neighbor cell type

\item[\code{min\_int\_cells\_expr}] minimum expression level for interacting neighbor cell type

\item[\code{min\_fdr}] minimum adjusted p-value

\item[\code{min\_spat\_diff}] minimum absolute spatial expression difference

\item[\code{min\_log2\_fc}] minimum log2 fold-change

\item[\code{min\_zscore}] minimum z-score change

\item[\code{zscores\_column}] calculate z-scores over cell types or genes

\item[\code{direction}] differential expression directions to keep
\end{ldescription}
\end{Arguments}
%
\begin{Value}
cpgObject that contains the filtered differential gene scores
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    filterCPG(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{filterDistributions}{filterDistributions}{filterDistributions}
%
\begin{Description}\relax
show gene or cell distribution after filtering on expression threshold
\end{Description}
%
\begin{Usage}
\begin{verbatim}
filterDistributions(
  gobject,
  expression_values = c("raw", "normalized", "scaled", "custom"),
  expression_threshold = 1,
  detection = c("genes", "cells"),
  plot_type = c("histogram", "violin"),
  nr_bins = 30,
  fill_color = "lightblue",
  scale_axis = "identity",
  axis_offset = 0,
  show_plot = NA,
  return_plot = NA,
  save_plot = NA,
  save_param = list(),
  default_save_name = "filterDistributions"
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{expression\_values}] expression values to use

\item[\code{expression\_threshold}] threshold to consider a gene expressed

\item[\code{detection}] consider genes or cells

\item[\code{plot\_type}] type of plot

\item[\code{nr\_bins}] number of bins for histogram plot

\item[\code{fill\_color}] fill color for plots

\item[\code{scale\_axis}] ggplot transformation for axis (e.g. log2)

\item[\code{axis\_offset}] offset to be used together with the scaling transformation

\item[\code{show\_plot}] show plot

\item[\code{return\_plot}] return ggplot object

\item[\code{save\_plot}] directly save the plot [boolean]

\item[\code{save\_param}] list of saving parameters from \code{\LinkA{all\_plots\_save\_function}{all.Rul.plots.Rul.save.Rul.function}}

\item[\code{default\_save\_name}] default save name for saving, don't change, change save\_name in save\_param
\end{ldescription}
\end{Arguments}
%
\begin{Value}
ggplot object
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    filterDistributions(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{filterGiotto}{filterGiotto}{filterGiotto}
%
\begin{Description}\relax
filter Giotto object based on expression threshold
\end{Description}
%
\begin{Usage}
\begin{verbatim}
filterGiotto(
  gobject,
  expression_values = c("raw", "normalized", "scaled", "custom"),
  expression_threshold = 1,
  gene_det_in_min_cells = 100,
  min_det_genes_per_cell = 100,
  verbose = F
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{expression\_values}] expression values to use

\item[\code{expression\_threshold}] threshold to consider a gene expressed

\item[\code{gene\_det\_in\_min\_cells}] minimum \# of cells that need to express a gene

\item[\code{min\_det\_genes\_per\_cell}] minimum \# of genes that need to be detected in a cell

\item[\code{verbose}] verbose
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
The function \code{\LinkA{filterCombinations}{filterCombinations}} can be used to explore the effect of different parameter values.
\end{Details}
%
\begin{Value}
giotto object
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    filterGiotto(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{filter\_network}{filter\_network}{filter.Rul.network}
%
\begin{Description}\relax
function to filter a spatial network
\end{Description}
%
\begin{Usage}
\begin{verbatim}
filter_network(networkDT, maximum_distance = NULL, minimum_k = NULL)
\end{verbatim}
\end{Usage}
\inputencoding{utf8}
\HeaderA{findCellProximityGenes}{findCellProximityGenes}{findCellProximityGenes}
%
\begin{Description}\relax
Identifies genes that are differentially expressed due to proximity to other cell types.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
findCellProximityGenes(
  gobject,
  expression_values = "normalized",
  selected_genes = NULL,
  cluster_column,
  spatial_network_name = "Delaunay_network",
  minimum_unique_cells = 1,
  minimum_unique_int_cells = 1,
  diff_test = c("permutation", "limma", "t.test", "wilcox"),
  mean_method = c("arithmic", "geometric"),
  offset = 0.1,
  adjust_method = c("bonferroni", "BH", "holm", "hochberg", "hommel", "BY", "fdr",
    "none"),
  nr_permutations = 1000,
  exclude_selected_cells_from_test = T,
  do_parallel = TRUE,
  cores = NA
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{expression\_values}] expression values to use

\item[\code{selected\_genes}] subset of selected genes (optional)

\item[\code{cluster\_column}] name of column to use for cell types

\item[\code{spatial\_network\_name}] name of spatial network to use

\item[\code{minimum\_unique\_cells}] minimum number of target cells required

\item[\code{minimum\_unique\_int\_cells}] minimum number of interacting cells required

\item[\code{diff\_test}] which differential expression test

\item[\code{mean\_method}] method to use to calculate the mean

\item[\code{offset}] offset value to use when calculating log2 ratio

\item[\code{adjust\_method}] which method to adjust p-values

\item[\code{nr\_permutations}] number of permutations if diff\_test = permutation

\item[\code{exclude\_selected\_cells\_from\_test}] exclude interacting cells other cells

\item[\code{do\_parallel}] run calculations in parallel with mclapply

\item[\code{cores}] number of cores to use if do\_parallel = TRUE
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Function to calculate if genes are differentially expressed in cell types
when they interact (approximated by physical proximity) with other cell types.
The results data.table in the cpgObject contains - at least - the following columns:
\begin{itemize}

\item{} genes: All or selected list of tested genes
\item{} sel: average gene expression in the interacting cells from the target cell type 
\item{} other: average gene expression in the NOT-interacting cells from the target cell type 
\item{} log2fc: log2 fold-change between sel and other
\item{} diff: spatial expression difference between sel and other
\item{} p.value: associated p-value
\item{} p.adj: adjusted p-value
\item{} cell\_type: target cell type
\item{} int\_cell\_type: interacting cell type
\item{} nr\_select: number of cells for selected target cell type
\item{} int\_nr\_select: number of cells for interacting cell type
\item{} nr\_other: number of other cells of selected target cell type
\item{} int\_nr\_other: number of other cells for interacting cell type
\item{} unif\_int: cell-cell interaction

\end{itemize}

\end{Details}
%
\begin{Value}
cpgObject that contains the differential gene scores
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    findCellProximityGenes(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{findCellProximityGenes\_per\_interaction}{findCellProximityGenes\_per\_interaction}{findCellProximityGenes.Rul.per.Rul.interaction}
%
\begin{Description}\relax
Identifies genes that are differentially expressed due to proximity to other cell types.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
findCellProximityGenes_per_interaction(
  expr_values,
  cell_metadata,
  annot_spatnetwork,
  sel_int,
  minimum_unique_cells = 1,
  minimum_unique_int_cells = 1,
  exclude_selected_cells_from_test = T,
  diff_test = c("permutation", "limma", "t.test", "wilcox"),
  mean_method = c("arithmic", "geometric"),
  offset = 0.1,
  adjust_method = "bonferroni",
  nr_permutations = 100,
  cores = 1
)
\end{verbatim}
\end{Usage}
%
\begin{Examples}
\begin{ExampleCode}
    findCellProximityGenes_per_interaction()
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{findCPG}{findCPG}{findCPG}
%
\begin{Description}\relax
Identifies genes that are differentially expressed due to proximity to other cell types.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
findCPG(
  gobject,
  expression_values = "normalized",
  selected_genes = NULL,
  cluster_column,
  spatial_network_name = "Delaunay_network",
  minimum_unique_cells = 1,
  minimum_unique_int_cells = 1,
  diff_test = c("permutation", "limma", "t.test", "wilcox"),
  mean_method = c("arithmic", "geometric"),
  offset = 0.1,
  adjust_method = c("bonferroni", "BH", "holm", "hochberg", "hommel", "BY", "fdr",
    "none"),
  nr_permutations = 100,
  exclude_selected_cells_from_test = T,
  do_parallel = TRUE,
  cores = NA
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{expression\_values}] expression values to use

\item[\code{selected\_genes}] subset of selected genes (optional)

\item[\code{cluster\_column}] name of column to use for cell types

\item[\code{spatial\_network\_name}] name of spatial network to use

\item[\code{minimum\_unique\_cells}] minimum number of target cells required

\item[\code{minimum\_unique\_int\_cells}] minimum number of interacting cells required

\item[\code{diff\_test}] which differential expression test

\item[\code{mean\_method}] method to use to calculate the mean

\item[\code{offset}] offset value to use when calculating log2 ratio

\item[\code{adjust\_method}] which method to adjust p-values

\item[\code{nr\_permutations}] number of permutations if diff\_test = permutation

\item[\code{exclude\_selected\_cells\_from\_test}] exclude interacting cells other cells

\item[\code{do\_parallel}] run calculations in parallel with mclapply

\item[\code{cores}] number of cores to use if do\_parallel = TRUE
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Function to calculate if genes are differentially expressed in cell types
when they interact (approximated by physical proximity) with other cell types.
The results data.table in the cpgObject contains - at least - the following columns:
\begin{itemize}

\item{} genes: All or selected list of tested genes
\item{} sel: average gene expression in the interacting cells from the target cell type 
\item{} other: average gene expression in the NOT-interacting cells from the target cell type 
\item{} log2fc: log2 fold-change between sel and other
\item{} diff: spatial expression difference between sel and other
\item{} p.value: associated p-value
\item{} p.adj: adjusted p-value
\item{} cell\_type: target cell type
\item{} int\_cell\_type: interacting cell type
\item{} nr\_select: number of cells for selected target cell type
\item{} int\_nr\_select: number of cells for interacting cell type
\item{} nr\_other: number of other cells of selected target cell type
\item{} int\_nr\_other: number of other cells for interacting cell type
\item{} unif\_int: cell-cell interaction

\end{itemize}

\end{Details}
%
\begin{Value}
cpgObject that contains the differential gene scores
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    findCPG(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{findGiniMarkers}{findGiniMarkers}{findGiniMarkers}
%
\begin{Description}\relax
Identify marker genes for selected clusters based on gini detection and expression scores.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
findGiniMarkers(
  gobject,
  expression_values = c("normalized", "scaled", "custom"),
  cluster_column,
  subset_clusters = NULL,
  group_1 = NULL,
  group_2 = NULL,
  min_expr_gini_score = 0.2,
  min_det_gini_score = 0.2,
  detection_threshold = 0,
  rank_score = 1,
  min_genes = 5
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{expression\_values}] gene expression values to use

\item[\code{cluster\_column}] clusters to use

\item[\code{subset\_clusters}] selection of clusters to compare

\item[\code{group\_1}] group 1 cluster IDs from cluster\_column for pairwise comparison

\item[\code{group\_2}] group 2 cluster IDs from cluster\_column for pairwise comparison

\item[\code{min\_expr\_gini\_score}] filter on minimum gini coefficient for expression

\item[\code{min\_det\_gini\_score}] filter on minimum gini coefficient for detection

\item[\code{detection\_threshold}] detection threshold for gene expression

\item[\code{rank\_score}] rank scores for both detection and expression to include

\item[\code{min\_genes}] minimum number of top genes to return
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Detection of marker genes using the \url{https://en.wikipedia.org/wiki/Gini_coefficient}gini
coefficient is based on the following steps/principles per gene:
\begin{itemize}

\item{} 1. calculate average expression per cluster
\item{} 2. calculate detection fraction per cluster
\item{} 3. calculate gini-coefficient for av. expression values over all clusters
\item{} 4. calculate gini-coefficient for detection fractions over all clusters
\item{} 5. convert gini-scores to rank scores
\item{} 6. for each gene create combined score = detection rank x expression rank x expr gini-coefficient x detection gini-coefficient
\item{} 7. for each gene sort on expression and detection rank and combined score

\end{itemize}


As a results "top gini" genes are genes that are very selectivily expressed in a specific cluster,
however not always expressed in all cells of that cluster. In other words highly specific, but
not necessarily sensitive at the single-cell level.

To perform differential expression between cluster groups you need to specificy cluster IDs
to the parameters \emph{group\_1} and \emph{group\_2}.
\end{Details}
%
\begin{Value}
data.table with marker genes
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    findGiniMarkers(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{findGiniMarkers\_one\_vs\_all}{findGiniMarkers\_one\_vs\_all}{findGiniMarkers.Rul.one.Rul.vs.Rul.all}
%
\begin{Description}\relax
Identify marker genes for all clusters in a one vs all manner based on gini detection and expression scores.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
findGiniMarkers_one_vs_all(
  gobject,
  expression_values = c("normalized", "scaled", "custom"),
  cluster_column,
  subset_clusters = NULL,
  min_expr_gini_score = 0.5,
  min_det_gini_score = 0.5,
  detection_threshold = 0,
  rank_score = 1,
  min_genes = 4,
  verbose = TRUE
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{expression\_values}] gene expression values to use

\item[\code{cluster\_column}] clusters to use

\item[\code{subset\_clusters}] selection of clusters to compare

\item[\code{min\_expr\_gini\_score}] filter on minimum gini coefficient on expression

\item[\code{min\_det\_gini\_score}] filter on minimum gini coefficient on detection

\item[\code{detection\_threshold}] detection threshold for gene expression

\item[\code{rank\_score}] rank scores for both detection and expression to include

\item[\code{min\_genes}] minimum number of top genes to return

\item[\code{verbose}] be verbose
\end{ldescription}
\end{Arguments}
%
\begin{Value}
data.table with marker genes
\end{Value}
%
\begin{SeeAlso}\relax
\code{\LinkA{findGiniMarkers}{findGiniMarkers}}
\end{SeeAlso}
%
\begin{Examples}
\begin{ExampleCode}
    findGiniMarkers_one_vs_all(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{findMarkers}{findMarkers}{findMarkers}
%
\begin{Description}\relax
Identify marker genes for selected clusters.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
findMarkers(
  gobject,
  expression_values = c("normalized", "scaled", "custom"),
  cluster_column,
  method = c("scran", "gini", "mast"),
  subset_clusters = NULL,
  group_1 = NULL,
  group_2 = NULL,
  min_expr_gini_score = 0.5,
  min_det_gini_score = 0.5,
  detection_threshold = 0,
  rank_score = 1,
  min_genes = 4,
  group_1_name = NULL,
  group_2_name = NULL,
  adjust_columns = NULL,
  ...
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{expression\_values}] gene expression values to use

\item[\code{cluster\_column}] clusters to use

\item[\code{method}] method to use to detect differentially expressed genes

\item[\code{subset\_clusters}] selection of clusters to compare

\item[\code{group\_1}] group 1 cluster IDs from cluster\_column for pairwise comparison

\item[\code{group\_2}] group 2 cluster IDs from cluster\_column for pairwise comparison

\item[\code{min\_expr\_gini\_score}] gini: filter on minimum gini coefficient for expression

\item[\code{min\_det\_gini\_score}] gini: filter minimum gini coefficient for detection

\item[\code{detection\_threshold}] gini: detection threshold for gene expression

\item[\code{rank\_score}] gini: rank scores to include

\item[\code{min\_genes}] minimum number of top genes to return (for gini)

\item[\code{group\_1\_name}] mast: custom name for group\_1 clusters

\item[\code{group\_2\_name}] mast: custom name for group\_2 clusters

\item[\code{adjust\_columns}] mast: column in pDataDT to adjust for (e.g. detection rate)

\item[\code{...}] additional parameters for the findMarkers function in scran or zlm function in MAST
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Wrapper for all individual functions to detect marker genes for clusters.
\end{Details}
%
\begin{Value}
data.table with marker genes
\end{Value}
%
\begin{SeeAlso}\relax
\code{\LinkA{findScranMarkers}{findScranMarkers}}, \code{\LinkA{findGiniMarkers}{findGiniMarkers}} and \code{\LinkA{findMastMarkers}{findMastMarkers}}
\end{SeeAlso}
%
\begin{Examples}
\begin{ExampleCode}
    findMarkers(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{findMarkers\_one\_vs\_all}{findMarkers\_one\_vs\_all}{findMarkers.Rul.one.Rul.vs.Rul.all}
%
\begin{Description}\relax
Identify marker genes for all clusters in a one vs all manner.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
findMarkers_one_vs_all(
  gobject,
  expression_values = c("normalized", "scaled", "custom"),
  cluster_column,
  subset_clusters = NULL,
  method = c("scran", "gini", "mast"),
  pval = 0.01,
  logFC = 0.5,
  min_genes = 10,
  min_expr_gini_score = 0.5,
  min_det_gini_score = 0.5,
  detection_threshold = 0,
  rank_score = 1,
  adjust_columns = NULL,
  verbose = TRUE,
  ...
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{expression\_values}] gene expression values to use

\item[\code{cluster\_column}] clusters to use

\item[\code{subset\_clusters}] selection of clusters to compare

\item[\code{method}] method to use to detect differentially expressed genes

\item[\code{pval}] scran \& mast: filter on minimal p-value

\item[\code{logFC}] scan \& mast: filter on logFC

\item[\code{min\_genes}] minimum genes to keep per cluster, overrides pval and logFC

\item[\code{min\_expr\_gini\_score}] gini: filter on minimum gini coefficient for expression

\item[\code{min\_det\_gini\_score}] gini: filter minimum gini coefficient for detection

\item[\code{detection\_threshold}] gini: detection threshold for gene expression

\item[\code{rank\_score}] gini: rank scores to include

\item[\code{adjust\_columns}] mast: column in pDataDT to adjust for (e.g. detection rate)

\item[\code{verbose}] be verbose

\item[\code{...}] additional parameters for the findMarkers function in scran or zlm function in MAST
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Wrapper for all one vs all functions to detect marker genes for clusters.
\end{Details}
%
\begin{Value}
data.table with marker genes
\end{Value}
%
\begin{SeeAlso}\relax
\code{\LinkA{findScranMarkers\_one\_vs\_all}{findScranMarkers.Rul.one.Rul.vs.Rul.all}}, \code{\LinkA{findGiniMarkers\_one\_vs\_all}{findGiniMarkers.Rul.one.Rul.vs.Rul.all}} and \code{\LinkA{findMastMarkers\_one\_vs\_all}{findMastMarkers.Rul.one.Rul.vs.Rul.all}}
\end{SeeAlso}
%
\begin{Examples}
\begin{ExampleCode}
    findMarkers_one_vs_all(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{findMastMarkers}{findMastMarkers}{findMastMarkers}
%
\begin{Description}\relax
Identify marker genes for selected clusters based on the MAST package.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
findMastMarkers(
  gobject,
  expression_values = c("normalized", "scaled", "custom"),
  cluster_column,
  group_1 = NULL,
  group_1_name = NULL,
  group_2 = NULL,
  group_2_name = NULL,
  adjust_columns = NULL,
  ...
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{expression\_values}] gene expression values to use

\item[\code{cluster\_column}] clusters to use

\item[\code{group\_1}] group 1 cluster IDs from cluster\_column for pairwise comparison

\item[\code{group\_1\_name}] custom name for group\_1 clusters

\item[\code{group\_2}] group 2 cluster IDs from cluster\_column for pairwise comparison

\item[\code{group\_2\_name}] custom name for group\_2 clusters

\item[\code{adjust\_columns}] column in pDataDT to adjust for (e.g. detection rate)

\item[\code{...}] additional parameters for the zlm function in MAST
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
This is a minimal convenience wrapper around the \code{\LinkA{zlm}{zlm}}
from the MAST package to detect differentially expressed genes.
\end{Details}
%
\begin{Value}
data.table with marker genes
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    findMastMarkers(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{findMastMarkers\_one\_vs\_all}{findMastMarkers\_one\_vs\_all}{findMastMarkers.Rul.one.Rul.vs.Rul.all}
%
\begin{Description}\relax
Identify marker genes for all clusters in a one vs all manner based on the MAST package.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
findMastMarkers_one_vs_all(
  gobject,
  expression_values = c("normalized", "scaled", "custom"),
  cluster_column,
  subset_clusters = NULL,
  adjust_columns = NULL,
  pval = 0.001,
  logFC = 1,
  min_genes = 10,
  verbose = TRUE,
  ...
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{expression\_values}] gene expression values to use

\item[\code{cluster\_column}] clusters to use

\item[\code{subset\_clusters}] selection of clusters to compare

\item[\code{adjust\_columns}] column in pDataDT to adjust for (e.g. detection rate)

\item[\code{pval}] filter on minimal p-value

\item[\code{logFC}] filter on logFC

\item[\code{min\_genes}] minimum genes to keep per cluster, overrides pval and logFC

\item[\code{verbose}] be verbose

\item[\code{...}] additional parameters for the zlm function in MAST
\end{ldescription}
\end{Arguments}
%
\begin{Value}
data.table with marker genes
\end{Value}
%
\begin{SeeAlso}\relax
\code{\LinkA{findMastMarkers}{findMastMarkers}}
\end{SeeAlso}
%
\begin{Examples}
\begin{ExampleCode}
    findMastMarkers_one_vs_all(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{findNetworkNeighbors}{findNetworkNeighbors}{findNetworkNeighbors}
%
\begin{Description}\relax
Find the spatial neighbors for a selected group of cells within the selected spatial network.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
findNetworkNeighbors(
  gobject,
  spatial_network_name,
  source_cell_ids = NULL,
  name = "nb_cells"
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] Giotto object

\item[\code{spatial\_network\_name}] name of spatial network

\item[\code{source\_cell\_ids}] cell ids for which you want to know the spatial neighbors

\item[\code{name}] name of the results
\end{ldescription}
\end{Arguments}
%
\begin{Value}
data.table
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    findNetworkNeighbors(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{findScranMarkers}{findScranMarkers}{findScranMarkers}
%
\begin{Description}\relax
Identify marker genes for all or selected clusters based on scran's implementation of findMarkers.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
findScranMarkers(
  gobject,
  expression_values = c("normalized", "scaled", "custom"),
  cluster_column,
  subset_clusters = NULL,
  group_1 = NULL,
  group_2 = NULL,
  ...
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{expression\_values}] gene expression values to use

\item[\code{cluster\_column}] clusters to use

\item[\code{subset\_clusters}] selection of clusters to compare

\item[\code{group\_1}] group 1 cluster IDs from cluster\_column for pairwise comparison

\item[\code{group\_2}] group 2 cluster IDs from cluster\_column for pairwise comparison

\item[\code{...}] additional parameters for the findMarkers function in scran
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
This is a minimal convenience wrapper around
the \code{\LinkA{findMarkers}{findMarkers}} function from the scran package.

To perform differential expression between cluster groups you need to specificy cluster IDs
to the parameters \emph{group\_1} and \emph{group\_2}.
\end{Details}
%
\begin{Value}
data.table with marker genes
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    findScranMarkers(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{findScranMarkers\_one\_vs\_all}{findScranMarkers\_one\_vs\_all}{findScranMarkers.Rul.one.Rul.vs.Rul.all}
%
\begin{Description}\relax
Identify marker genes for all clusters in a one vs all manner based on scran's implementation of findMarkers.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
findScranMarkers_one_vs_all(
  gobject,
  expression_values = c("normalized", "scaled", "custom"),
  cluster_column,
  subset_clusters = NULL,
  pval = 0.01,
  logFC = 0.5,
  min_genes = 10,
  verbose = TRUE,
  ...
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{expression\_values}] gene expression values to use

\item[\code{cluster\_column}] clusters to use

\item[\code{subset\_clusters}] subset of clusters to use

\item[\code{pval}] filter on minimal p-value

\item[\code{logFC}] filter on logFC

\item[\code{min\_genes}] minimum genes to keep per cluster, overrides pval and logFC

\item[\code{verbose}] be verbose

\item[\code{...}] additional parameters for the findMarkers function in scran
\end{ldescription}
\end{Arguments}
%
\begin{Value}
data.table with marker genes
\end{Value}
%
\begin{SeeAlso}\relax
\code{\LinkA{findScranMarkers}{findScranMarkers}}
\end{SeeAlso}
%
\begin{Examples}
\begin{ExampleCode}
    findScranMarkers_one_vs_all(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{find\_grid\_2D}{find\_grid\_2D}{find.Rul.grid.Rul.2D}
%
\begin{Description}\relax
find grid location in 2D
\end{Description}
%
\begin{Usage}
\begin{verbatim}
find_grid_2D(grid_DT, x_loc, y_loc)
\end{verbatim}
\end{Usage}
\inputencoding{utf8}
\HeaderA{find\_grid\_3D}{find\_grid\_3D}{find.Rul.grid.Rul.3D}
%
\begin{Description}\relax
find grid location in 3D
\end{Description}
%
\begin{Usage}
\begin{verbatim}
find_grid_3D(grid_DT, x_loc, y_loc, z_loc)
\end{verbatim}
\end{Usage}
\inputencoding{utf8}
\HeaderA{find\_grid\_x}{find\_grid\_x}{find.Rul.grid.Rul.x}
%
\begin{Description}\relax
find grid location on x-axis
\end{Description}
%
\begin{Usage}
\begin{verbatim}
find_grid_x(grid_DT, x_loc)
\end{verbatim}
\end{Usage}
\inputencoding{utf8}
\HeaderA{find\_grid\_y}{find\_grid\_y}{find.Rul.grid.Rul.y}
%
\begin{Description}\relax
find grid location on y-axis
\end{Description}
%
\begin{Usage}
\begin{verbatim}
find_grid_y(grid_DT, y_loc)
\end{verbatim}
\end{Usage}
\inputencoding{utf8}
\HeaderA{find\_grid\_z}{find\_grid\_z}{find.Rul.grid.Rul.z}
%
\begin{Description}\relax
find grid location on z-axis
\end{Description}
%
\begin{Usage}
\begin{verbatim}
find_grid_z(grid_DT, z_loc)
\end{verbatim}
\end{Usage}
\inputencoding{utf8}
\HeaderA{find\_x\_y\_ranges}{find\_x\_y\_ranges}{find.Rul.x.Rul.y.Rul.ranges}
%
\begin{Description}\relax
get the extended ranges of x and y
\end{Description}
%
\begin{Usage}
\begin{verbatim}
find_x_y_ranges(data, extend_ratio)
\end{verbatim}
\end{Usage}
\inputencoding{utf8}
\HeaderA{general\_save\_function}{general\_save\_function}{general.Rul.save.Rul.function}
%
\begin{Description}\relax
Function to automatically save plots to directory of interest
\end{Description}
%
\begin{Usage}
\begin{verbatim}
general_save_function(
  gobject,
  plot_object,
  save_dir = NULL,
  save_folder = NULL,
  save_name = NULL,
  default_save_name = "giotto_plot",
  save_format = c("png", "tiff", "pdf", "svg"),
  show_saved_plot = F,
  base_width = NULL,
  base_height = NULL,
  base_aspect_ratio = NULL,
  units = NULL,
  dpi = NULL,
  ...
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{plot\_object}] non-ggplot object to plot

\item[\code{save\_dir}] directory to save to

\item[\code{save\_folder}] folder in save\_dir to save to

\item[\code{save\_name}] name of plot

\item[\code{save\_format}] format (e.g. png, tiff, pdf, ...)

\item[\code{show\_saved\_plot}] load \& display the saved plot

\item[\code{base\_width}] width

\item[\code{base\_height}] height

\item[\code{base\_aspect\_ratio}] aspect ratio

\item[\code{units}] units

\item[\code{dpi}] Plot resolution
\end{ldescription}
\end{Arguments}
%
\begin{Examples}
\begin{ExampleCode}
    general_save_function(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{get10Xmatrix}{get10Xmatrix}{get10Xmatrix}
%
\begin{Description}\relax
This function creates an expression matrix from a 10X structured folder
\end{Description}
%
\begin{Usage}
\begin{verbatim}
get10Xmatrix(path_to_data, gene_column_index = 1)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{path\_to\_data}] path to the 10X folder

\item[\code{gene\_column\_index}] which column from the features or genes .tsv file to use for row ids
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
A typical 10X folder is named raw\_feature\_bc\_matrix or raw\_feature\_bc\_matrix and tt has 3 files:
\begin{itemize}

\item{} barcodes.tsv(.gz)
\item{} features.tsv(.gz) or genes.tsv(.gz)
\item{} matrix.mtx(.gz)

\end{itemize}

By default the first column of the features or genes .tsv file will be used, however if multiple
annotations are provided (e.g. ensembl gene ids and gene symbols) the user can select another column.
\end{Details}
%
\begin{Value}
expression matrix from 10X
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    get10Xmatrix(10Xmatrix)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{getClusterSimilarity}{getClusterSimilarity}{getClusterSimilarity}
%
\begin{Description}\relax
Creates data.table with pairwise correlation scores between each cluster.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
getClusterSimilarity(
  gobject,
  expression_values = c("normalized", "scaled", "custom"),
  cluster_column,
  cor = c("pearson", "spearman")
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{expression\_values}] expression values to use

\item[\code{cluster\_column}] name of column to use for clusters

\item[\code{cor}] correlation score to calculate distance
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Creates data.table with pairwise correlation scores between each cluster and
the group size (\# of cells) for each cluster. This information can be used together
with mergeClusters to combine very similar or small clusters into bigger clusters.
\end{Details}
%
\begin{Value}
data.table
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    getClusterSimilarity(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{getDendrogramSplits}{getDendrogramSplits}{getDendrogramSplits}
%
\begin{Description}\relax
Split dendrogram at each node and keep the leave (label) information..
\end{Description}
%
\begin{Usage}
\begin{verbatim}
getDendrogramSplits(
  gobject,
  expression_values = c("normalized", "scaled", "custom"),
  cluster_column,
  cor = c("pearson", "spearman"),
  distance = "ward.D",
  h = NULL,
  h_color = "red",
  show_dend = TRUE,
  verbose = TRUE
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{expression\_values}] expression values to use

\item[\code{cluster\_column}] name of column to use for clusters

\item[\code{cor}] correlation score to calculate distance

\item[\code{distance}] distance method to use for hierarchical clustering

\item[\code{h}] height of horizontal lines to plot

\item[\code{h\_color}] color of horizontal lines

\item[\code{show\_dend}] show dendrogram

\item[\code{verbose}] be verbose
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Creates a data.table with three columns and each row represents a node in the
dendrogram. For each node the height of the node is given together with the two
subdendrograms. This information can be used to determine in a hierarchical manner
differentially expressed marker genes at each node.
\end{Details}
%
\begin{Value}
data.table object
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    getDendrogramSplits(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{getDistinctColors}{getDistinctColors}{getDistinctColors}
%
\begin{Description}\relax
Returns a number of distint colors based on the RGB scale
\end{Description}
%
\begin{Usage}
\begin{verbatim}
getDistinctColors(n)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{n}] number of colors wanted
\end{ldescription}
\end{Arguments}
%
\begin{Value}
number of distinct colors
\end{Value}
\inputencoding{utf8}
\HeaderA{get\_cross\_section\_coordinates}{get\_cross\_section\_coordinates}{get.Rul.cross.Rul.section.Rul.coordinates}
%
\begin{Description}\relax
get local coordinates within cross section plane
\end{Description}
%
\begin{Usage}
\begin{verbatim}
get_cross_section_coordinates(cell_subset_projection_locations)
\end{verbatim}
\end{Usage}
\inputencoding{utf8}
\HeaderA{get\_distance}{get\_distance}{get.Rul.distance}
%
\begin{Description}\relax
estimate average distance between neighboring cells with network table as input
\end{Description}
%
\begin{Usage}
\begin{verbatim}
get_distance(networkDT, method = c("mean", "median"))
\end{verbatim}
\end{Usage}
\inputencoding{utf8}
\HeaderA{get\_sectionThickness}{get\_sectionThickness}{get.Rul.sectionThickness}
%
\begin{Description}\relax
get section thickness
\end{Description}
%
\begin{Usage}
\begin{verbatim}
get_sectionThickness(
  gobject,
  thickness_unit = c("cell", "natural"),
  slice_thickness = 2,
  spatial_network_name = "Delaunay_network",
  cell_distance_estimate_method = c("mean", "median"),
  plane_equation = NULL
)
\end{verbatim}
\end{Usage}
\inputencoding{utf8}
\HeaderA{ggplot\_save\_function}{ggplot\_save\_function}{ggplot.Rul.save.Rul.function}
%
\begin{Description}\relax
Function to automatically save plots to directory of interest
\end{Description}
%
\begin{Usage}
\begin{verbatim}
ggplot_save_function(
  gobject,
  plot_object,
  save_dir = NULL,
  save_folder = NULL,
  save_name = NULL,
  default_save_name = "giotto_plot",
  save_format = NULL,
  show_saved_plot = F,
  ncol = 1,
  nrow = 1,
  scale = 1,
  base_width = NULL,
  base_height = NULL,
  base_aspect_ratio = NULL,
  units = NULL,
  dpi = NULL,
  limitsize = TRUE,
  ...
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{plot\_object}] ggplot object to plot

\item[\code{save\_dir}] directory to save to

\item[\code{save\_folder}] folder in save\_dir to save to

\item[\code{save\_name}] name of plot

\item[\code{save\_format}] format (e.g. png, tiff, pdf, ...)

\item[\code{show\_saved\_plot}] load \& display the saved plot

\item[\code{ncol}] number of columns

\item[\code{nrow}] number of rows

\item[\code{scale}] scale

\item[\code{base\_width}] width

\item[\code{base\_height}] height

\item[\code{base\_aspect\_ratio}] aspect ratio

\item[\code{units}] units

\item[\code{dpi}] Plot resolution

\item[\code{limitsize}] When TRUE (the default), ggsave will not save images larger than 50x50 inches, to prevent the common error of specifying dimensions in pixels.
\end{ldescription}
\end{Arguments}
%
\begin{SeeAlso}\relax
\code{\LinkA{cowplot::save\_plot}{cowplot::save.Rul.plot}}
\end{SeeAlso}
%
\begin{Examples}
\begin{ExampleCode}
    ggplot_save_function(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{giotto-class}{S4 giotto Class}{giotto.Rdash.class}
\aliasA{giotto}{giotto-class}{giotto}
\keyword{giotto,}{giotto-class}
\keyword{object}{giotto-class}
%
\begin{Description}\relax
Framework of giotto object to store and work with spatial expression data
\end{Description}
%
\begin{Section}{Slots}

\begin{description}

\item[\code{raw\_exprs}] raw expression counts

\item[\code{norm\_expr}] normalized expression counts

\item[\code{norm\_scaled\_expr}] normalized and scaled expression counts

\item[\code{custom\_expr}] custom normalized counts

\item[\code{spatial\_locs}] spatial location coordinates for cells

\item[\code{cell\_metadata}] metadata for cells

\item[\code{gene\_metadata}] metadata for genes

\item[\code{cell\_ID}] unique cell IDs

\item[\code{gene\_ID}] unique gene IDs

\item[\code{spatial\_network}] spatial network in data.table/data.frame format

\item[\code{spatial\_grid}] spatial grid in data.table/data.frame format

\item[\code{dimension\_reduction}] slot to save dimension reduction coordinates

\item[\code{nn\_network}] nearest neighbor network in igraph format

\item[\code{parameters}] slot to save parameters that have been used

\item[\code{instructions}] slot for global function instructions

\item[\code{offset\_file}] offset file used to stitch together image fields

\item[\code{OS\_platform}] Operating System to run Giotto analysis on

\end{description}
\end{Section}
\inputencoding{utf8}
\HeaderA{heatmSpatialCorGenes}{heatmSpatialCorGenes}{heatmSpatialCorGenes}
%
\begin{Description}\relax
Create heatmap of spatially correlated genes
\end{Description}
%
\begin{Usage}
\begin{verbatim}
heatmSpatialCorGenes(
  gobject,
  spatCorObject,
  use_clus_name = NULL,
  show_cluster_annot = TRUE,
  show_row_dend = T,
  show_column_dend = F,
  show_row_names = F,
  show_column_names = F,
  show_plot = NA,
  return_plot = NA,
  save_plot = NA,
  save_param = list(),
  default_save_name = "heatmSpatialCorGenes",
  ...
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{spatCorObject}] spatial correlation object

\item[\code{use\_clus\_name}] name of clusters to visualize (from clusterSpatialCorGenes())

\item[\code{show\_cluster\_annot}] show cluster annotation on top of heatmap

\item[\code{show\_row\_dend}] show row dendrogram

\item[\code{show\_column\_dend}] show column dendrogram

\item[\code{show\_row\_names}] show row names

\item[\code{show\_column\_names}] show column names

\item[\code{show\_plot}] show plot

\item[\code{return\_plot}] return ggplot object

\item[\code{save\_plot}] directly save the plot [boolean]

\item[\code{save\_param}] list of saving parameters from \code{\LinkA{all\_plots\_save\_function}{all.Rul.plots.Rul.save.Rul.function}}

\item[\code{default\_save\_name}] default save name for saving, don't change, change save\_name in save\_param

\item[\code{...}] additional parameters to the \code{\LinkA{Heatmap}{Heatmap}} function from ComplexHeatmap
\end{ldescription}
\end{Arguments}
%
\begin{Value}
Heatmap generated by ComplexHeatmap
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    heatmSpatialCorGenes(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{hyperGeometricEnrich}{hyperGeometricEnrich}{hyperGeometricEnrich}
%
\begin{Description}\relax
Function to calculate gene signature enrichment scores per spatial position using a hypergeometric test.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
hyperGeometricEnrich(
  gobject,
  sign_matrix,
  expression_values = c("normalized", "scaled", "custom"),
  reverse_log_scale = TRUE,
  logbase = 2,
  top_percentage = 5,
  output_enrichment = c("original", "zscore")
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] Giotto object

\item[\code{sign\_matrix}] Matrix of signature genes for each cell type / process

\item[\code{expression\_values}] expression values to use

\item[\code{reverse\_log\_scale}] reverse expression values from log scale

\item[\code{logbase}] log base to use if reverse\_log\_scale = TRUE

\item[\code{top\_percentage}] percentage of cells that will be considered to have gene expression with matrix binarization

\item[\code{output\_enrichment}] how to return enrichment output
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
The enrichment score is calculated based on the p-value from the
hypergeometric test, -log10(p-value).
\end{Details}
%
\begin{Value}
data.table with enrichment results
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    hyperGeometricEnrich(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{insertCrossSectionGenePlot3D}{insertCrossSectionGenePlot3D}{insertCrossSectionGenePlot3D}
%
\begin{Description}\relax
Visualize cells and gene expression in a virtual cross section according to spatial coordinates
\end{Description}
%
\begin{Usage}
\begin{verbatim}
insertCrossSectionGenePlot3D(
  gobject,
  crossSection_obj = NULL,
  name = NULL,
  spatial_network_name = "Delaunay_network",
  mesh_grid_color = "#1f77b4",
  mesh_grid_width = 3,
  mesh_grid_style = "dot",
  sdimx = "sdimx",
  sdimy = "sdimy",
  sdimz = "sdimz",
  expression_values = c("normalized", "scaled", "custom"),
  genes,
  show_network = F,
  network_color = NULL,
  edge_alpha = NULL,
  show_grid = F,
  cluster_column = NULL,
  select_cell_groups = NULL,
  select_cells = NULL,
  show_other_cells = F,
  other_cell_color = "lightgrey",
  other_point_size = 1,
  genes_high_color = NULL,
  genes_mid_color = "white",
  genes_low_color = "darkblue",
  spatial_grid_name = "spatial_grid",
  point_size = 2,
  show_legend = T,
  axis_scale = c("cube", "real", "custom"),
  custom_ratio = NULL,
  x_ticks = NULL,
  y_ticks = NULL,
  z_ticks = NULL,
  show_plot = NA,
  return_plot = NA,
  save_plot = NA,
  save_param = list(),
  default_save_name = "spatGenePlot3D_with_cross_section"
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{name}] name of virtual cross section to use

\item[\code{spatial\_network\_name}] name of spatial network to use

\item[\code{mesh\_grid\_color}] color for the meshgrid lines

\item[\code{mesh\_grid\_width}] width for the meshgrid lines

\item[\code{mesh\_grid\_style}] style for the meshgrid lines

\item[\code{sdimx}] x-axis dimension name (default = 'sdimx')

\item[\code{sdimy}] y-axis dimension name (default = 'sdimy')

\item[\code{sdimz}] z-axis dimension name (default = 'sdimy')

\item[\code{expression\_values}] gene expression values to use

\item[\code{genes}] genes to show

\item[\code{show\_network}] show underlying spatial network

\item[\code{network\_color}] color of spatial network

\item[\code{show\_grid}] show spatial grid

\item[\code{genes\_high\_color}] color represents high gene expression

\item[\code{genes\_mid\_color}] color represents middle gene expression

\item[\code{genes\_low\_color}] color represents low gene expression

\item[\code{spatial\_grid\_name}] name of spatial grid to use

\item[\code{point\_size}] size of point (cell)

\item[\code{show\_legend}] show legend

\item[\code{show\_plot}] show plots

\item[\code{return\_plot}] return ggplot object

\item[\code{save\_plot}] directly save the plot [boolean]

\item[\code{save\_param}] list of saving parameters from \code{\LinkA{all\_plots\_save\_function}{all.Rul.plots.Rul.save.Rul.function}}

\item[\code{default\_save\_name}] default save name for saving, don't change, change save\_name in save\_param

\item[\code{grid\_color}] color of spatial grid

\item[\code{midpoint}] expression midpoint

\item[\code{scale\_alpha\_with\_expression}] scale expression with ggplot alpha parameter

\item[\code{...}] parameters for cowplot::save\_plot()
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Description of parameters.
\end{Details}
%
\begin{Value}
ggplot
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    insertCrossSectionGenePlot3D(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{insertCrossSectionSpatPlot3D}{insertCrossSectionSpatPlot3D}{insertCrossSectionSpatPlot3D}
%
\begin{Description}\relax
Visualize the meshgrid lines of cross section together with cells
\end{Description}
%
\begin{Usage}
\begin{verbatim}
insertCrossSectionSpatPlot3D(
  gobject,
  crossSection_obj = NULL,
  name = NULL,
  spatial_network_name = "Delaunay_network",
  mesh_grid_color = "#1f77b4",
  mesh_grid_width = 3,
  mesh_grid_style = "dot",
  sdimx = "sdimx",
  sdimy = "sdimy",
  sdimz = "sdimz",
  point_size = 2,
  cell_color = NULL,
  cell_color_code = NULL,
  select_cell_groups = NULL,
  select_cells = NULL,
  show_other_cells = T,
  other_cell_color = "lightgrey",
  other_point_size = 0.5,
  show_network = F,
  network_color = NULL,
  network_alpha = 1,
  other_cell_alpha = 0.5,
  show_grid = F,
  grid_color = NULL,
  spatial_grid_name = "spatial_grid",
  title = "",
  show_legend = T,
  axis_scale = c("cube", "real", "custom"),
  custom_ratio = NULL,
  x_ticks = NULL,
  y_ticks = NULL,
  z_ticks = NULL,
  show_plot = NA,
  return_plot = NA,
  save_plot = NA,
  save_param = list(),
  default_save_name = "spat3D_with_cross_section"
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{name}] name of virtual cross section to use

\item[\code{spatial\_network\_name}] name of spatial network to use

\item[\code{mesh\_grid\_color}] color for the meshgrid lines

\item[\code{mesh\_grid\_width}] width for the meshgrid lines

\item[\code{mesh\_grid\_style}] style for the meshgrid lines

\item[\code{sdimx}] x-axis dimension name (default = 'sdimx')

\item[\code{sdimy}] y-axis dimension name (default = 'sdimy')

\item[\code{sdimz}] z-axis dimension name (default = 'sdimy')

\item[\code{point\_size}] size of point (cell)

\item[\code{cell\_color}] color for cells (see details)

\item[\code{cell\_color\_code}] named vector with colors

\item[\code{select\_cell\_groups}] select subset of cells/clusters based on cell\_color parameter

\item[\code{show\_other\_cells}] display not selected cells

\item[\code{other\_cell\_color}] color of not selected cells

\item[\code{other\_point\_size}] point size of not selected cells

\item[\code{network\_color}] color of spatial network

\item[\code{show\_grid}] show spatial grid

\item[\code{grid\_color}] color of spatial grid

\item[\code{spatial\_grid\_name}] name of spatial grid to use

\item[\code{title}] title of plot

\item[\code{show\_legend}] show legend

\item[\code{axis\_scale}] the way to scale the axis

\item[\code{custom\_ratio}] customize the scale of the plot

\item[\code{x\_ticks}] set the number of ticks on the x-axis

\item[\code{y\_ticks}] set the number of ticks on the y-axis

\item[\code{z\_ticks}] set the number of ticks on the z-axis

\item[\code{show\_plot}] show plot

\item[\code{return\_plot}] return ggplot object

\item[\code{save\_plot}] directly save the plot [boolean]

\item[\code{save\_param}] list of saving parameters from \code{\LinkA{all\_plots\_save\_function}{all.Rul.plots.Rul.save.Rul.function}}

\item[\code{default\_save\_name}] default save name for saving, don't change, change save\_name in save\_param
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Description of parameters.
\end{Details}
%
\begin{Value}
ggplot
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    insertCrossSectionSpatPlot3D(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{kmeans\_binarize}{kmeans\_binarize}{kmeans.Rul.binarize}
%
\begin{Description}\relax
create binarized scores from a vector using kmeans
\end{Description}
%
\begin{Usage}
\begin{verbatim}
kmeans_binarize(x, nstart = 3, iter.max = 10)
\end{verbatim}
\end{Usage}
\inputencoding{utf8}
\HeaderA{loadHMRF}{loadHMRF}{loadHMRF}
%
\begin{Description}\relax
load previous HMRF
\end{Description}
%
\begin{Usage}
\begin{verbatim}
loadHMRF(
  name_used = "test",
  output_folder_used,
  k_used = 10,
  betas_used,
  python_path_used
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{name\_used}] name of HMRF that was run

\item[\code{output\_folder\_used}] output folder that was used

\item[\code{k\_used}] number of HMRF domains that was tested

\item[\code{betas\_used}] betas that were tested

\item[\code{python\_path\_used}] python path that was used
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Description of HMRF parameters ...
\end{Details}
%
\begin{Value}
reloads a previous ran HMRF from doHRMF
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    loadHMRF(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{makeSignMatrixPAGE}{makeSignMatrixPAGE}{makeSignMatrixPAGE}
%
\begin{Description}\relax
Function to convert a list of signature genes (e.g. for cell types or processes) into
a binary matrix format that can be used with the PAGE enrichment option. Each cell type or process should
have a vector of cell-type or process specific genes. These vectors need to be combined into a list (sign\_list).
The names of the cell types or processes that are provided in the list need to be given (sign\_names).
\end{Description}
%
\begin{Usage}
\begin{verbatim}
makeSignMatrixPAGE(sign_names, sign_list)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{sign\_names}] vector with names for each provided gene signature

\item[\code{sign\_list}] list of genes (signature)
\end{ldescription}
\end{Arguments}
%
\begin{Value}
matrix
\end{Value}
%
\begin{SeeAlso}\relax
\code{\LinkA{PAGEEnrich}{PAGEEnrich}}
\end{SeeAlso}
%
\begin{Examples}
\begin{ExampleCode}
    makeSignMatrixPAGE()
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{makeSignMatrixRank}{makeSignMatrixRank}{makeSignMatrixRank}
%
\begin{Description}\relax
Function to convert a single-cell count matrix
and a corresponding single-cell cluster vector into
a rank matrix that can be used with the Rank enrichment option.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
makeSignMatrixRank(sc_matrix, sc_cluster_ids, gobject = NULL)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{sc\_matrix}] matrix of single-cell RNAseq expression data

\item[\code{sc\_cluster\_ids}] vector of cluster ids

\item[\code{gobject}] if giotto object is given then only genes present in both datasets will be considered
\end{ldescription}
\end{Arguments}
%
\begin{Value}
matrix
\end{Value}
%
\begin{SeeAlso}\relax
\code{\LinkA{rankEnrich}{rankEnrich}}
\end{SeeAlso}
%
\begin{Examples}
\begin{ExampleCode}
    makeSignMatrixRank()
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{make\_simulated\_network}{make\_simulated\_network}{make.Rul.simulated.Rul.network}
%
\begin{Description}\relax
Simulate random network.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
make_simulated_network(
  gobject,
  spatial_network_name = "Delaunay_network",
  cluster_column,
  number_of_simulations = 100
)
\end{verbatim}
\end{Usage}
%
\begin{Examples}
\begin{ExampleCode}
    make_simulated_network(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{mergeClusters}{mergeClusters}{mergeClusters}
%
\begin{Description}\relax
Merge selected clusters based on pairwise correlation scores and size of cluster.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
mergeClusters(
  gobject,
  expression_values = c("normalized", "scaled", "custom"),
  cluster_column,
  cor = c("pearson", "spearman"),
  new_cluster_name = "merged_cluster",
  min_cor_score = 0.8,
  max_group_size = 20,
  force_min_group_size = 10,
  return_gobject = TRUE,
  verbose = TRUE
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{expression\_values}] expression values to use

\item[\code{cluster\_column}] name of column to use for clusters

\item[\code{cor}] correlation score to calculate distance

\item[\code{new\_cluster\_name}] new name for merged clusters

\item[\code{min\_cor\_score}] min correlation score to merge pairwise clusters

\item[\code{max\_group\_size}] max cluster size that can be merged

\item[\code{force\_min\_group\_size}] size of clusters that will be merged with their most similar neighbor(s)

\item[\code{return\_gobject}] return giotto object

\item[\code{verbose}] be verbose
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Merge selected clusters based on pairwise correlation scores and size of cluster.
To avoid large clusters to merge the max\_group\_size can be lowered. Small clusters can
be forcibly merged with their most similar pairwise cluster by adjusting the
force\_min\_group\_size parameter. Clusters smaller than this value will be merged
independent on the provided min\_cor\_score value. \\{}
A giotto object is returned by default, if FALSE then the merging vector will be returned.
\end{Details}
%
\begin{Value}
Giotto object
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    mergeClusters(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{mygini\_fun}{mygini\_fun}{mygini.Rul.fun}
%
\begin{Description}\relax
calculate gini coefficient
\end{Description}
%
\begin{Usage}
\begin{verbatim}
mygini_fun(x, weights = rep(1, length(x)))
\end{verbatim}
\end{Usage}
%
\begin{Value}
gini coefficient
\end{Value}
\inputencoding{utf8}
\HeaderA{my\_arowMeans}{my\_arowMeans}{my.Rul.arowMeans}
%
\begin{Description}\relax
arithmic rowMeans that works for a single column
\end{Description}
%
\begin{Usage}
\begin{verbatim}
my_arowMeans(x)
\end{verbatim}
\end{Usage}
%
\begin{Examples}
\begin{ExampleCode}
    my_arowMeans(x)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{my\_growMeans}{my\_growMeans}{my.Rul.growMeans}
%
\begin{Description}\relax
geometric rowMeans that works for a single column
\end{Description}
%
\begin{Usage}
\begin{verbatim}
my_growMeans(x, offset = 0.1)
\end{verbatim}
\end{Usage}
%
\begin{Examples}
\begin{ExampleCode}
    my_growMeans(x)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{my\_rowMeans}{my\_rowMeans}{my.Rul.rowMeans}
%
\begin{Description}\relax
arithmic or geometric rowMeans that works for a single column
\end{Description}
%
\begin{Usage}
\begin{verbatim}
my_rowMeans(x, method = c("arithmic", "geometric"), offset = 0.1)
\end{verbatim}
\end{Usage}
%
\begin{Examples}
\begin{ExampleCode}
    my_rowMeans(x)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{nnDT\_to\_kNN}{nnDT\_to\_kNN}{nnDT.Rul.to.Rul.kNN}
%
\begin{Description}\relax
Convert a nearest network data.table to a kNN object
\end{Description}
%
\begin{Usage}
\begin{verbatim}
nnDT_to_kNN(nnDT)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{nnDT}] nearest neighbor network in data.table format
\end{ldescription}
\end{Arguments}
%
\begin{Value}
kNN object
\end{Value}
\inputencoding{utf8}
\HeaderA{node\_clusters}{node\_clusters}{node.Rul.clusters}
%
\begin{Description}\relax
Merge selected clusters based on pairwise correlation scores and size of cluster.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
node_clusters(hclus_obj, verbose = TRUE)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{hclus\_obj}] hclus object

\item[\code{verbose}] be verbose
\end{ldescription}
\end{Arguments}
%
\begin{Value}
list of splitted dendrogram nodes from high to low node height
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    node_clusters(hclus_obj)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{normalizeGiotto}{normalizeGiotto}{normalizeGiotto}
%
\begin{Description}\relax
fast normalize and/or scale expresion values of Giotto object
\end{Description}
%
\begin{Usage}
\begin{verbatim}
normalizeGiotto(
  gobject,
  norm_methods = c("standard", "osmFISH"),
  library_size_norm = TRUE,
  scalefactor = 6000,
  log_norm = TRUE,
  log_offset = 1,
  logbase = 2,
  scale_genes = T,
  scale_cells = T,
  scale_order = c("first_genes", "first_cells"),
  verbose = F
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{norm\_methods}] normalization method to use

\item[\code{library\_size\_norm}] normalize cells by library size

\item[\code{scalefactor}] scale factor to use after library size normalization

\item[\code{log\_norm}] transform values to log-scale

\item[\code{log\_offset}] offset value to add to expression matrix, default = 1

\item[\code{logbase}] log base to use to log normalize expression values

\item[\code{scale\_genes}] z-score genes over all cells

\item[\code{scale\_cells}] z-score cells over all genes

\item[\code{scale\_order}] order to scale genes and cells

\item[\code{verbose}] be verbose
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Currently there are two 'methods' to normalize your raw counts data.

A. The standard method follows the standard protocol which can be adjusted using
the provided parameters and follows the following order: \\{}
\begin{itemize}

\item{} 1. Data normalization for total library size and scaling by a custom scale-factor.
\item{} 2. Log transformation of data.
\item{} 3. Z-scoring of data by genes and/or cells.

\end{itemize}

B. The normalization method as provided by the osmFISH paper is also implemented: \\{}
\begin{itemize}

\item{} 1. First normalize genes, for each gene divide the counts by the total gene count and
multiply by the total number of genes.
\item{} 2. Next normalize cells, for each cell divide the normalized gene counts by the total
counts per cell and multiply by the total number of cells.

\end{itemize}

This data will be saved in the Giotto slot for custom expression.
\end{Details}
%
\begin{Value}
giotto object
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    normalizeGiotto(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{normalizeGiottoOld}{normalizeGiotto}{normalizeGiottoOld}
%
\begin{Description}\relax
normalize and/or scale expresion values of Giotto object
\end{Description}
%
\begin{Usage}
\begin{verbatim}
normalizeGiottoOld(
  gobject,
  norm_methods = c("standard", "osmFISH"),
  library_size_norm = TRUE,
  scalefactor = 6000,
  log_norm = TRUE,
  logbase = 2,
  scale_genes = T,
  scale_cells = T,
  scale_order = c("first_genes", "first_cells"),
  verbose = F
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{norm\_methods}] normalization method to use

\item[\code{library\_size\_norm}] normalize cells by library size

\item[\code{scalefactor}] scale factor to use after library size normalization

\item[\code{log\_norm}] transform values to log-scale

\item[\code{logbase}] log base to use to log normalize expression values

\item[\code{scale\_genes}] z-score genes over all cells

\item[\code{scale\_cells}] z-score cells over all genes

\item[\code{scale\_order}] order to scale genes and cells

\item[\code{verbose}] be verbose
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Currently there are two 'methods' to normalize your raw counts data.

A. The standard method follows the standard protocol which can be adjusted using
the provided parameters and follows the following order: \\{}
\begin{itemize}

\item{} 1. Data normalization for total library size and scaling by a custom scale-factor.
\item{} 2. Log transformation of data.
\item{} 3. Z-scoring of data by genes and/or cells.

\end{itemize}

B. The normalization method as provided by the osmFISH paper is also implemented: \\{}
\begin{itemize}

\item{} 1. First normalize genes, for each gene divide the counts by the total gene count and
multiply by the total number of genes.
\item{} 2. Next normalize cells, for each cell divide the normalized gene counts by the total
counts per cell and multiply by the total number of cells.

\end{itemize}

This data will be saved in the Giotto slot for custom expression.
\end{Details}
%
\begin{Value}
giotto object
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    normalizeGiotto(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{PAGEEnrich}{PAGEEnrich}{PAGEEnrich}
%
\begin{Description}\relax
Function to calculate gene signature enrichment scores per spatial position using PAGE.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
PAGEEnrich(
  gobject,
  sign_matrix,
  expression_values = c("normalized", "scaled", "custom"),
  reverse_log_scale = TRUE,
  logbase = 2,
  output_enrichment = c("original", "zscore")
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] Giotto object

\item[\code{sign\_matrix}] Matrix of signature genes for each cell type / process

\item[\code{expression\_values}] expression values to use

\item[\code{reverse\_log\_scale}] reverse expression values from log scale

\item[\code{logbase}] log base to use if reverse\_log\_scale = TRUE

\item[\code{output\_enrichment}] how to return enrichment output
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
sign\_matrix: a binary matrix with genes as row names and cell-types as column names.
Alternatively a list of signature genes can be provided to makeSignMatrixPAGE, which will create
the matrix for you. \\{}

The enrichment Z score is calculated by using method (PAGE) from
Kim SY et al., BMC bioinformatics, 2005 as \eqn{Z = ((Sm  mu)*m^(1/2)) / delta}{}.
For each gene in each spot, mu is the fold change values versus the mean expression
and delta is the standard deviation. Sm is the mean fold change value of a specific marker gene set
and  m is the size of a given marker gene set.
\end{Details}
%
\begin{Value}
data.table with enrichment results
\end{Value}
%
\begin{SeeAlso}\relax
\code{\LinkA{makeSignMatrixPAGE}{makeSignMatrixPAGE}}
\end{SeeAlso}
%
\begin{Examples}
\begin{ExampleCode}
    PAGEEnrich(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{pDataDT}{pDataDT}{pDataDT}
%
\begin{Description}\relax
show cell metadata
\end{Description}
%
\begin{Usage}
\begin{verbatim}
pDataDT(gobject)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object
\end{ldescription}
\end{Arguments}
%
\begin{Value}
data.table with cell metadata
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    pDataDT(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{plotCCcomDotplot}{plotCCcomDotplot}{plotCCcomDotplot}
%
\begin{Description}\relax
Plots dotplot for ligand-receptor communication scores in cell-cell interactions
\end{Description}
%
\begin{Usage}
\begin{verbatim}
plotCCcomDotplot(
  gobject,
  comScores,
  selected_LR = NULL,
  selected_cell_LR = NULL,
  show_LR_names = TRUE,
  show_cell_LR_names = TRUE,
  cluster_on = c("PI", "LR_expr", "log2fc"),
  cor_method = c("pearson", "kendall", "spearman"),
  aggl_method = c("ward.D", "ward.D2", "single", "complete", "average", "mcquitty",
    "median", "centroid"),
  show_plot = NA,
  return_plot = NA,
  save_plot = NA,
  save_param = list(),
  default_save_name = "plotCCcomDotplot"
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{comScores}] communinication scores from \code{\LinkA{exprCellCellcom}{exprCellCellcom}} or \code{\LinkA{spatCellCellcom}{spatCellCellcom}}

\item[\code{selected\_LR}] selected ligand-receptor combinations

\item[\code{selected\_cell\_LR}] selected cell-cell combinations for ligand-receptor combinations

\item[\code{show\_LR\_names}] show ligand-receptor names

\item[\code{show\_cell\_LR\_names}] show cell-cell names

\item[\code{cluster\_on}] values to use for clustering of cell-cell and ligand-receptor pairs

\item[\code{cor\_method}] correlation method used for clustering

\item[\code{aggl\_method}] agglomeration method used by hclust

\item[\code{show\_plot}] show plots

\item[\code{return\_plot}] return plotting object

\item[\code{save\_plot}] directly save the plot [boolean]

\item[\code{save\_param}] list of saving parameters from \code{\LinkA{all\_plots\_save\_function}{all.Rul.plots.Rul.save.Rul.function}}

\item[\code{default\_save\_name}] default save name for saving, don't change, change save\_name in save\_param

\item[\code{show}] values to show on heatmap
\end{ldescription}
\end{Arguments}
%
\begin{Value}
ggplot
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    plotCCcomDotplot(CPGscores)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{plotCCcomHeatmap}{plotCCcomHeatmap}{plotCCcomHeatmap}
%
\begin{Description}\relax
Plots heatmap for ligand-receptor communication scores in cell-cell interactions
\end{Description}
%
\begin{Usage}
\begin{verbatim}
plotCCcomHeatmap(
  gobject,
  comScores,
  selected_LR = NULL,
  selected_cell_LR = NULL,
  show_LR_names = TRUE,
  show_cell_LR_names = TRUE,
  show = c("PI", "LR_expr", "log2fc"),
  cor_method = c("pearson", "kendall", "spearman"),
  aggl_method = c("ward.D", "ward.D2", "single", "complete", "average", "mcquitty",
    "median", "centroid"),
  show_plot = NA,
  return_plot = NA,
  save_plot = NA,
  save_param = list(),
  default_save_name = "plotCCcomHeatmap"
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{comScores}] communinication scores from \code{\LinkA{exprCellCellcom}{exprCellCellcom}} or \code{\LinkA{spatCellCellcom}{spatCellCellcom}}

\item[\code{selected\_LR}] selected ligand-receptor combinations

\item[\code{selected\_cell\_LR}] selected cell-cell combinations for ligand-receptor combinations

\item[\code{show\_LR\_names}] show ligand-receptor names

\item[\code{show\_cell\_LR\_names}] show cell-cell names

\item[\code{show}] values to show on heatmap

\item[\code{cor\_method}] correlation method used for clustering

\item[\code{aggl\_method}] agglomeration method used by hclust

\item[\code{show\_plot}] show plots

\item[\code{return\_plot}] return plotting object

\item[\code{save\_plot}] directly save the plot [boolean]

\item[\code{save\_param}] list of saving parameters from \code{\LinkA{all\_plots\_save\_function}{all.Rul.plots.Rul.save.Rul.function}}

\item[\code{default\_save\_name}] default save name for saving, don't change, change save\_name in save\_param
\end{ldescription}
\end{Arguments}
%
\begin{Value}
ggplot
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    plotCCcomHeatmap(CPGscores)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{plotCellProximityGenes}{plotCellProximityGenes}{plotCellProximityGenes}
%
\begin{Description}\relax
Create visualization for cell proximity gene scores
\end{Description}
%
\begin{Usage}
\begin{verbatim}
plotCellProximityGenes(
  gobject,
  cpgObject,
  method = c("volcano", "cell_barplot", "cell-cell", "cell_sankey", "heatmap",
    "dotplot"),
  min_cells = 4,
  min_cells_expr = 1,
  min_int_cells = 4,
  min_int_cells_expr = 1,
  min_fdr = 0.1,
  min_spat_diff = 0.2,
  min_log2_fc = 0.2,
  min_zscore = 2,
  zscores_column = c("cell_type", "genes"),
  direction = c("both", "up", "down"),
  cell_color_code = NULL,
  show_plot = NA,
  return_plot = NA,
  save_plot = NA,
  save_param = list(),
  default_save_name = "plotCellProximityGenes"
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{cpgObject}] cell proximity gene score object

\item[\code{method}] plotting method to use

\item[\code{min\_cells}] minimum number of source cell type

\item[\code{min\_cells\_expr}] minimum expression level for source cell type

\item[\code{min\_int\_cells}] minimum number of interacting neighbor cell type

\item[\code{min\_int\_cells\_expr}] minimum expression level for interacting neighbor cell type

\item[\code{min\_fdr}] minimum adjusted p-value

\item[\code{min\_spat\_diff}] minimum absolute spatial expression difference

\item[\code{min\_log2\_fc}] minimum log2 fold-change

\item[\code{min\_zscore}] minimum z-score change

\item[\code{zscores\_column}] calculate z-scores over cell types or genes

\item[\code{direction}] differential expression directions to keep

\item[\code{cell\_color\_code}] vector of colors with cell types as names

\item[\code{show\_plot}] show plots

\item[\code{return\_plot}] return plotting object

\item[\code{save\_plot}] directly save the plot [boolean]

\item[\code{save\_param}] list of saving parameters from \code{\LinkA{all\_plots\_save\_function}{all.Rul.plots.Rul.save.Rul.function}}

\item[\code{default\_save\_name}] default save name for saving, don't change, change save\_name in save\_param
\end{ldescription}
\end{Arguments}
%
\begin{Value}
plot
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    plotCellProximityGenes(CPGscores)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{plotCombineCCcom}{plotCombineCCcom}{plotCombineCCcom}
%
\begin{Description}\relax
Create visualization for combined (pairwise) cell proximity gene scores
\end{Description}
%
\begin{Usage}
\begin{verbatim}
plotCombineCCcom(
  gobject,
  combCCcom,
  selected_LR = NULL,
  selected_cell_LR = NULL,
  detail_plot = T,
  simple_plot = F,
  simple_plot_facet = c("interaction", "genes"),
  facet_scales = "fixed",
  facet_ncol = length(selected_LR),
  facet_nrow = length(selected_cell_LR),
  colors = c("#9932CC", "#FF8C00"),
  show_plot = NA,
  return_plot = NA,
  save_plot = NA,
  save_param = list(),
  default_save_name = "plotCombineCCcom"
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{combCCcom}] combined communcation scores, output from combCCcom()

\item[\code{selected\_LR}] selected ligand-receptor pair

\item[\code{selected\_cell\_LR}] selected cell-cell interaction pair for ligand-receptor pair

\item[\code{detail\_plot}] show detailed info in both interacting cell types

\item[\code{simple\_plot}] show a simplified plot

\item[\code{simple\_plot\_facet}] facet on interactions or genes with simple plot

\item[\code{facet\_scales}] ggplot facet scales paramter

\item[\code{facet\_ncol}] ggplot facet ncol parameter

\item[\code{facet\_nrow}] ggplot facet nrow parameter

\item[\code{colors}] vector with two colors to use

\item[\code{show\_plot}] show plots

\item[\code{return\_plot}] return plotting object

\item[\code{save\_plot}] directly save the plot [boolean]

\item[\code{save\_param}] list of saving parameters from \code{\LinkA{all\_plots\_save\_function}{all.Rul.plots.Rul.save.Rul.function}}

\item[\code{default\_save\_name}] default save name for saving, don't change, change save\_name in save\_param
\end{ldescription}
\end{Arguments}
%
\begin{Value}
ggplot
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    plotCombineCCcom(CPGscores)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{plotCombineCellCellCommunication}{plotCombineCellCellCommunication}{plotCombineCellCellCommunication}
%
\begin{Description}\relax
Create visualization for combined (pairwise) cell proximity gene scores
\end{Description}
%
\begin{Usage}
\begin{verbatim}
plotCombineCellCellCommunication(
  gobject,
  combCCcom,
  selected_LR = NULL,
  selected_cell_LR = NULL,
  detail_plot = T,
  simple_plot = F,
  simple_plot_facet = c("interaction", "genes"),
  facet_scales = "fixed",
  facet_ncol = length(selected_LR),
  facet_nrow = length(selected_cell_LR),
  colors = c("#9932CC", "#FF8C00"),
  show_plot = NA,
  return_plot = NA,
  save_plot = NA,
  save_param = list(),
  default_save_name = "plotCombineCellCellCommunication"
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{combCCcom}] combined communcation scores, output from combCCcom()

\item[\code{selected\_LR}] selected ligand-receptor pair

\item[\code{selected\_cell\_LR}] selected cell-cell interaction pair for ligand-receptor pair

\item[\code{detail\_plot}] show detailed info in both interacting cell types

\item[\code{simple\_plot}] show a simplified plot

\item[\code{simple\_plot\_facet}] facet on interactions or genes with simple plot

\item[\code{facet\_scales}] ggplot facet scales paramter

\item[\code{facet\_ncol}] ggplot facet ncol parameter

\item[\code{facet\_nrow}] ggplot facet nrow parameter

\item[\code{colors}] vector with two colors to use

\item[\code{show\_plot}] show plots

\item[\code{return\_plot}] return plotting object

\item[\code{save\_plot}] directly save the plot [boolean]

\item[\code{save\_param}] list of saving parameters from \code{\LinkA{all\_plots\_save\_function}{all.Rul.plots.Rul.save.Rul.function}}

\item[\code{default\_save\_name}] default save name for saving, don't change, change save\_name in save\_param
\end{ldescription}
\end{Arguments}
%
\begin{Value}
ggplot
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    plotCombineCellCellCommunication(CPGscores)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{plotCombineCellProximityGenes}{plotCombineCellProximityGenes}{plotCombineCellProximityGenes}
%
\begin{Description}\relax
Create visualization for combined (pairwise) cell proximity gene scores
\end{Description}
%
\begin{Usage}
\begin{verbatim}
plotCombineCellProximityGenes(
  gobject,
  combCpgObject,
  selected_interactions = NULL,
  selected_gene_to_gene = NULL,
  detail_plot = T,
  simple_plot = F,
  simple_plot_facet = c("interaction", "genes"),
  facet_scales = "fixed",
  facet_ncol = length(selected_gene_to_gene),
  facet_nrow = length(selected_interactions),
  colors = c("#9932CC", "#FF8C00"),
  show_plot = NA,
  return_plot = NA,
  save_plot = NA,
  save_param = list(),
  default_save_name = "plotCombineCPG"
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{combCpgObject}] CPGscores, output from combineCellProximityGenes()

\item[\code{selected\_interactions}] interactions to show

\item[\code{selected\_gene\_to\_gene}] pairwise gene combinations to show

\item[\code{detail\_plot}] show detailed info in both interacting cell types

\item[\code{simple\_plot}] show a simplified plot

\item[\code{simple\_plot\_facet}] facet on interactions or genes with simple plot

\item[\code{facet\_scales}] ggplot facet scales paramter

\item[\code{facet\_ncol}] ggplot facet ncol parameter

\item[\code{facet\_nrow}] ggplot facet nrow parameter

\item[\code{colors}] vector with two colors to use

\item[\code{show\_plot}] show plots

\item[\code{return\_plot}] return plotting object

\item[\code{save\_plot}] directly save the plot [boolean]

\item[\code{save\_param}] list of saving parameters from \code{\LinkA{all\_plots\_save\_function}{all.Rul.plots.Rul.save.Rul.function}}

\item[\code{default\_save\_name}] default save name for saving, don't change, change save\_name in save\_param
\end{ldescription}
\end{Arguments}
%
\begin{Value}
ggplot
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    plotCombineCellProximityGenes(CPGscores)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{plotCombineCPG}{plotCombineCPG}{plotCombineCPG}
%
\begin{Description}\relax
Create visualization for combined (pairwise) cell proximity gene scores
\end{Description}
%
\begin{Usage}
\begin{verbatim}
plotCombineCPG(
  gobject,
  combCpgObject,
  selected_interactions = NULL,
  selected_gene_to_gene = NULL,
  detail_plot = T,
  simple_plot = F,
  simple_plot_facet = c("interaction", "genes"),
  facet_scales = "fixed",
  facet_ncol = length(selected_gene_to_gene),
  facet_nrow = length(selected_interactions),
  colors = c("#9932CC", "#FF8C00"),
  show_plot = NA,
  return_plot = NA,
  save_plot = NA,
  save_param = list(),
  default_save_name = "plotCombineCPG"
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{combCpgObject}] CPGscores, output from combineCellProximityGenes()

\item[\code{selected\_interactions}] interactions to show

\item[\code{selected\_gene\_to\_gene}] pairwise gene combinations to show

\item[\code{detail\_plot}] show detailed info in both interacting cell types

\item[\code{simple\_plot}] show a simplified plot

\item[\code{simple\_plot\_facet}] facet on interactions or genes with simple plot

\item[\code{facet\_scales}] ggplot facet scales paramter

\item[\code{facet\_ncol}] ggplot facet ncol parameter

\item[\code{facet\_nrow}] ggplot facet nrow parameter

\item[\code{colors}] vector with two colors to use

\item[\code{show\_plot}] show plots

\item[\code{return\_plot}] return plotting object

\item[\code{save\_plot}] directly save the plot [boolean]

\item[\code{save\_param}] list of saving parameters from \code{\LinkA{all\_plots\_save\_function}{all.Rul.plots.Rul.save.Rul.function}}

\item[\code{default\_save\_name}] default save name for saving, don't change, change save\_name in save\_param
\end{ldescription}
\end{Arguments}
%
\begin{Value}
ggplot
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    plotCombineCPG(CPGscores)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{plotCPG}{plotCPG}{plotCPG}
%
\begin{Description}\relax
Create visualization for cell proximity gene scores
\end{Description}
%
\begin{Usage}
\begin{verbatim}
plotCPG(
  gobject,
  cpgObject,
  method = c("volcano", "cell_barplot", "cell-cell", "cell_sankey", "heatmap",
    "dotplot"),
  min_cells = 5,
  min_cells_expr = 1,
  min_int_cells = 3,
  min_int_cells_expr = 1,
  min_fdr = 0.05,
  min_spat_diff = 0.2,
  min_log2_fc = 0.2,
  min_zscore = 2,
  zscores_column = c("cell_type", "genes"),
  direction = c("both", "up", "down"),
  cell_color_code = NULL,
  show_plot = NA,
  return_plot = NA,
  save_plot = NA,
  save_param = list(),
  default_save_name = "plotCPG"
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{cpgObject}] cell proximity gene score object

\item[\code{method}] plotting method to use

\item[\code{min\_cells}] minimum number of source cell type

\item[\code{min\_cells\_expr}] minimum expression level for source cell type

\item[\code{min\_int\_cells}] minimum number of interacting neighbor cell type

\item[\code{min\_int\_cells\_expr}] minimum expression level for interacting neighbor cell type

\item[\code{min\_fdr}] minimum adjusted p-value

\item[\code{min\_spat\_diff}] minimum absolute spatial expression difference

\item[\code{min\_log2\_fc}] minimum log2 fold-change

\item[\code{min\_zscore}] minimum z-score change

\item[\code{zscores\_column}] calculate z-scores over cell types or genes

\item[\code{direction}] differential expression directions to keep

\item[\code{cell\_color\_code}] vector of colors with cell types as names

\item[\code{show\_plot}] show plots

\item[\code{return\_plot}] return plotting object

\item[\code{save\_plot}] directly save the plot [boolean]

\item[\code{save\_param}] list of saving parameters from \code{\LinkA{all\_plots\_save\_function}{all.Rul.plots.Rul.save.Rul.function}}

\item[\code{default\_save\_name}] default save name for saving, don't change, change save\_name in save\_param
\end{ldescription}
\end{Arguments}
%
\begin{Value}
plot
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    plotCPG(CPGscores)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{plotHeatmap}{plotHeatmap}{plotHeatmap}
%
\begin{Description}\relax
Creates heatmap for genes and clusters.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
plotHeatmap(
  gobject,
  expression_values = c("normalized", "scaled", "custom"),
  genes,
  cluster_column = NULL,
  cluster_order = c("size", "correlation", "custom"),
  cluster_custom_order = NULL,
  cluster_color_code = NULL,
  cluster_cor_method = "pearson",
  cluster_hclust_method = "ward.D",
  gene_order = c("correlation", "custom"),
  gene_custom_order = NULL,
  gene_cor_method = "pearson",
  gene_hclust_method = "complete",
  show_values = c("rescaled", "z-scaled", "original"),
  size_vertical_lines = 1.1,
  gradient_colors = c("blue", "yellow", "red"),
  gene_label_selection = NULL,
  axis_text_y_size = NULL,
  legend_nrows = 1,
  show_plot = NA,
  return_plot = NA,
  save_plot = NA,
  save_param = list(),
  default_save_name = "plotHeatmap"
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{expression\_values}] expression values to use

\item[\code{genes}] genes to use

\item[\code{cluster\_column}] name of column to use for clusters

\item[\code{cluster\_order}] method to determine cluster order

\item[\code{cluster\_custom\_order}] custom order for clusters

\item[\code{cluster\_color\_code}] color code for clusters

\item[\code{cluster\_cor\_method}] method for cluster correlation

\item[\code{cluster\_hclust\_method}] method for hierarchical clustering of clusters

\item[\code{gene\_order}] method to determine gene order

\item[\code{gene\_custom\_order}] custom order for genes

\item[\code{gene\_cor\_method}] method for gene correlation

\item[\code{gene\_hclust\_method}] method for hierarchical clustering of genes

\item[\code{show\_values}] which values to show on heatmap

\item[\code{size\_vertical\_lines}] sizes for vertical lines

\item[\code{gradient\_colors}] colors for heatmap gradient

\item[\code{gene\_label\_selection}] subset of genes to show on y-axis

\item[\code{axis\_text\_y\_size}] size for y-axis text

\item[\code{legend\_nrows}] number of rows for the cluster legend

\item[\code{show\_plot}] show plot

\item[\code{return\_plot}] return ggplot object

\item[\code{save\_plot}] directly save the plot [boolean]

\item[\code{save\_param}] list of saving parameters from \code{\LinkA{all\_plots\_save\_function}{all.Rul.plots.Rul.save.Rul.function}}

\item[\code{default\_save\_name}] default save name
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
If you want to display many genes there are 2 ways to proceed:
\begin{itemize}

\item{} 1. set axis\_text\_y\_size to a really small value and show all genes
\item{} 2. provide a subset of genes to display to gene\_label\_selection

\end{itemize}

\end{Details}
%
\begin{Value}
ggplot
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    plotHeatmap(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{plotICG}{plotICG}{plotICG}
%
\begin{Description}\relax
Create barplot to visualize interaction changed genes
\end{Description}
%
\begin{Usage}
\begin{verbatim}
plotICG(
  gobject,
  cpgObject,
  source_type,
  source_markers,
  ICG_genes,
  cell_color_code = NULL,
  show_plot = NA,
  return_plot = NA,
  save_plot = NA,
  save_param = list(),
  default_save_name = "plotICG"
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{cpgObject}] cell proximity gene score object

\item[\code{source\_type}] cell type of the source cell

\item[\code{source\_markers}] markers for the source cell type

\item[\code{ICG\_genes}] named character vector of ICG genes

\item[\code{cell\_color\_code}] cell color code for the interacting cell types

\item[\code{show\_plot}] show plots

\item[\code{return\_plot}] return plotting object

\item[\code{save\_plot}] directly save the plot [boolean]

\item[\code{save\_param}] list of saving parameters from \code{\LinkA{all\_plots\_save\_function}{all.Rul.plots.Rul.save.Rul.function}}

\item[\code{default\_save\_name}] default save name for saving, don't change, change save\_name in save\_param
\end{ldescription}
\end{Arguments}
%
\begin{Value}
plot
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    plotICG(CPGscores)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{plotInteractionChangedGenes}{plotInteractionChangedGenes}{plotInteractionChangedGenes}
%
\begin{Description}\relax
Create barplot to visualize interaction changed genes
\end{Description}
%
\begin{Usage}
\begin{verbatim}
plotInteractionChangedGenes(
  gobject,
  cpgObject,
  source_type,
  source_markers,
  ICG_genes,
  cell_color_code = NULL,
  show_plot = NA,
  return_plot = NA,
  save_plot = NA,
  save_param = list(),
  default_save_name = "plotInteractionChangedGenes"
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{cpgObject}] cell proximity gene score object

\item[\code{source\_type}] cell type of the source cell

\item[\code{source\_markers}] markers for the source cell type

\item[\code{ICG\_genes}] named character vector of ICG genes

\item[\code{cell\_color\_code}] cell color code for the interacting cell types

\item[\code{show\_plot}] show plots

\item[\code{return\_plot}] return plotting object

\item[\code{save\_plot}] directly save the plot [boolean]

\item[\code{save\_param}] list of saving parameters from \code{\LinkA{all\_plots\_save\_function}{all.Rul.plots.Rul.save.Rul.function}}

\item[\code{default\_save\_name}] default save name for saving, don't change, change save\_name in save\_param
\end{ldescription}
\end{Arguments}
%
\begin{Value}
plot
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    plotInteractionChangedGenes(CPGscores)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{plotly\_axis\_scale\_2D}{plotly\_axis\_scale\_2D}{plotly.Rul.axis.Rul.scale.Rul.2D}
%
\begin{Description}\relax
adjust the axis scale in 3D plotly plot
\end{Description}
%
\begin{Usage}
\begin{verbatim}
plotly_axis_scale_2D(
  cell_locations,
  sdimx = NULL,
  sdimy = NULL,
  mode = c("cube", "real", "custom"),
  custom_ratio = NULL
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{cell\_locations}] spatial\_loc in giotto object

\item[\code{sdimx}] x axis of cell spatial location

\item[\code{sdimy}] y axis of cell spatial location

\item[\code{mode}] axis adjustment mode

\item[\code{custom\_ratio}] set the ratio artificially
\end{ldescription}
\end{Arguments}
%
\begin{Value}
edges in spatial grid as data.table()
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    plotly_axis_scale_2D(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{plotly\_axis\_scale\_3D}{plotly\_axis\_scale\_3D}{plotly.Rul.axis.Rul.scale.Rul.3D}
%
\begin{Description}\relax
adjust the axis scale in 3D plotly plot
\end{Description}
%
\begin{Usage}
\begin{verbatim}
plotly_axis_scale_3D(
  cell_locations,
  sdimx = NULL,
  sdimy = NULL,
  sdimz = NULL,
  mode = c("cube", "real", "custom"),
  custom_ratio = NULL
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{cell\_locations}] spatial\_loc in giotto object

\item[\code{sdimx}] x axis of cell spatial location

\item[\code{sdimy}] y axis of cell spatial location

\item[\code{sdimz}] z axis of cell spatial location

\item[\code{mode}] axis adjustment mode

\item[\code{custom\_ratio}] set the ratio artificially
\end{ldescription}
\end{Arguments}
%
\begin{Value}
edges in spatial grid as data.table()
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    plotly_axis_scale_3D(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{plotly\_grid}{plotly\_grid}{plotly.Rul.grid}
%
\begin{Description}\relax
provide grid segment to draw in plot\_ly()
\end{Description}
%
\begin{Usage}
\begin{verbatim}
plotly_grid(
  spatial_grid,
  x_start = "x_start",
  y_start = "y_start",
  x_end = "x_end",
  y_end = "y_end"
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{spatial\_grid}] spatial\_grid in giotto object
\end{ldescription}
\end{Arguments}
%
\begin{Value}
edges in spatial grid as data.table()
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    plotly_grid(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{plotly\_network}{plotly\_network}{plotly.Rul.network}
%
\begin{Description}\relax
provide network segment to draw in 3D plot\_ly()
\end{Description}
%
\begin{Usage}
\begin{verbatim}
plotly_network(
  network,
  x = "sdimx_begin",
  y = "sdimy_begin",
  z = "sdimz_begin",
  x_end = "sdimx_end",
  y_end = "sdimy_end",
  z_end = "sdimz_end"
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] network in giotto object
\end{ldescription}
\end{Arguments}
%
\begin{Value}
edges in network as data.table()
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    plotly_network(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{plotMetaDataCellsHeatmap}{plotMetaDataCellsHeatmap}{plotMetaDataCellsHeatmap}
%
\begin{Description}\relax
Creates heatmap for numeric cell metadata within aggregated clusters.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
plotMetaDataCellsHeatmap(
  gobject,
  metadata_cols = NULL,
  spat_enr_names = NULL,
  value_cols = NULL,
  first_meta_col = NULL,
  second_meta_col = NULL,
  show_values = c("zscores", "original", "zscores_rescaled"),
  custom_cluster_order = NULL,
  clus_cor_method = "pearson",
  clus_cluster_method = "complete",
  custom_values_order = NULL,
  values_cor_method = "pearson",
  values_cluster_method = "complete",
  midpoint = 0,
  x_text_size = 8,
  x_text_angle = 45,
  y_text_size = 8,
  strip_text_size = 8,
  show_plot = NA,
  return_plot = NA,
  save_plot = NA,
  save_param = list(),
  default_save_name = "plotMetaDataCellsHeatmap"
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{metadata\_cols}] annotation columns found in pDataDT(gobject)

\item[\code{spat\_enr\_names}] spatial enrichment results to include

\item[\code{value\_cols}] value columns to use

\item[\code{first\_meta\_col}] if more than 1 metadata column, select the x-axis factor

\item[\code{second\_meta\_col}] if more than 1 metadata column, select the facetting factor

\item[\code{show\_values}] which values to show on heatmap

\item[\code{custom\_cluster\_order}] custom cluster order (default = NULL)

\item[\code{clus\_cor\_method}] correlation method for clusters

\item[\code{clus\_cluster\_method}] hierarchical cluster method for the clusters

\item[\code{midpoint}] midpoint of show\_values

\item[\code{x\_text\_size}] size of x-axis text

\item[\code{x\_text\_angle}] angle of x-axis text

\item[\code{y\_text\_size}] size of y-axis text

\item[\code{strip\_text\_size}] size of strip text

\item[\code{show\_plot}] show plot

\item[\code{return\_plot}] return ggplot object

\item[\code{save\_plot}] directly save the plot [boolean]

\item[\code{save\_param}] list of saving parameters from \code{\LinkA{all\_plots\_save\_function}{all.Rul.plots.Rul.save.Rul.function}}

\item[\code{default\_save\_name}] default save name for saving, don't change, change save\_name in save\_param

\item[\code{custom\_gene\_order}] custom gene order (default = NULL)

\item[\code{gene\_cor\_method}] correlation method for genes

\item[\code{gene\_cluster\_method}] hierarchical cluster method for the genes
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Creates heatmap for the average values of selected value columns in the different annotation groups.
\end{Details}
%
\begin{Value}
ggplot or data.table
\end{Value}
%
\begin{SeeAlso}\relax
\code{\LinkA{plotMetaDataHeatmap}{plotMetaDataHeatmap}} for gene expression instead of numeric cell annotation data.
\end{SeeAlso}
%
\begin{Examples}
\begin{ExampleCode}
    plotMetaDataCellsHeatmap(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{plotMetaDataHeatmap}{plotMetaDataHeatmap}{plotMetaDataHeatmap}
%
\begin{Description}\relax
Creates heatmap for genes within aggregated clusters.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
plotMetaDataHeatmap(
  gobject,
  expression_values = c("normalized", "scaled", "custom"),
  metadata_cols = NULL,
  selected_genes = NULL,
  first_meta_col = NULL,
  second_meta_col = NULL,
  show_values = c("zscores", "original", "zscores_rescaled"),
  custom_cluster_order = NULL,
  clus_cor_method = "pearson",
  clus_cluster_method = "complete",
  custom_gene_order = NULL,
  gene_cor_method = "pearson",
  gene_cluster_method = "complete",
  gradient_color = c("blue", "white", "red"),
  gradient_midpoint = 0,
  gradient_limits = NULL,
  x_text_size = 10,
  x_text_angle = 45,
  y_text_size = 10,
  strip_text_size = 8,
  show_plot = NA,
  return_plot = NA,
  save_plot = NA,
  save_param = list(),
  default_save_name = "plotMetaDataHeatmap"
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{expression\_values}] expression values to use

\item[\code{metadata\_cols}] annotation columns found in pDataDT(gobject)

\item[\code{selected\_genes}] subset of genes to use

\item[\code{first\_meta\_col}] if more than 1 metadata column, select the x-axis factor

\item[\code{second\_meta\_col}] if more than 1 metadata column, select the facetting factor

\item[\code{show\_values}] which values to show on heatmap

\item[\code{custom\_cluster\_order}] custom cluster order (default = NULL)

\item[\code{clus\_cor\_method}] correlation method for clusters

\item[\code{clus\_cluster\_method}] hierarchical cluster method for the clusters

\item[\code{custom\_gene\_order}] custom gene order (default = NULL)

\item[\code{gene\_cor\_method}] correlation method for genes

\item[\code{gene\_cluster\_method}] hierarchical cluster method for the genes

\item[\code{gradient\_color}] vector with 3 colors for numeric data

\item[\code{gradient\_midpoint}] midpoint for color gradient

\item[\code{gradient\_limits}] vector with lower and upper limits

\item[\code{x\_text\_size}] size of x-axis text

\item[\code{x\_text\_angle}] angle of x-axis text

\item[\code{y\_text\_size}] size of y-axis text

\item[\code{strip\_text\_size}] size of strip text

\item[\code{show\_plot}] show plot

\item[\code{return\_plot}] return ggplot object

\item[\code{save\_plot}] directly save the plot [boolean]

\item[\code{save\_param}] list of saving parameters from \code{\LinkA{all\_plots\_save\_function}{all.Rul.plots.Rul.save.Rul.function}}

\item[\code{default\_save\_name}] default save name
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Creates heatmap for the average expression of selected genes in the different annotation/cluster groups.
Calculation of cluster or gene order is done on the provided expression values, but visualization
is by default on the z-scores. Other options are the original values or z-scores rescaled per gene (-1 to 1).
\end{Details}
%
\begin{Value}
ggplot or data.table
\end{Value}
%
\begin{SeeAlso}\relax
\code{\LinkA{plotMetaDataCellsHeatmap}{plotMetaDataCellsHeatmap}} for numeric cell annotation instead of gene expression.
\end{SeeAlso}
%
\begin{Examples}
\begin{ExampleCode}
    plotMetaDataHeatmap(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{plotPCA}{plotPCA}{plotPCA}
%
\begin{Description}\relax
Short wrapper for PCA visualization
\end{Description}
%
\begin{Usage}
\begin{verbatim}
plotPCA(gobject, dim_reduction_name = "pca", default_save_name = "PCA", ...)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{dim\_reduction\_name}] dimension reduction name

\item[\code{default\_save\_name}] default save name for saving, don't change, change save\_name in save\_param

\item[\code{groub\_by}] create multiple plots based on cell annotation column

\item[\code{group\_by\_subset}] subset the group\_by factor column

\item[\code{dim1\_to\_use}] dimension to use on x-axis

\item[\code{dim2\_to\_use}] dimension to use on y-axis

\item[\code{spat\_enr\_names}] names of spatial enrichment results to include

\item[\code{show\_NN\_network}] show underlying NN network

\item[\code{nn\_network\_to\_use}] type of NN network to use (kNN vs sNN)

\item[\code{network\_name}] name of NN network to use, if show\_NN\_network = TRUE

\item[\code{cell\_color}] color for cells (see details)

\item[\code{color\_as\_factor}] convert color column to factor

\item[\code{cell\_color\_code}] named vector with colors

\item[\code{cell\_color\_gradient}] vector with 3 colors for numeric data

\item[\code{gradient\_midpoint}] midpoint for color gradient

\item[\code{gradient\_limits}] vector with lower and upper limits

\item[\code{select\_cell\_groups}] select subset of cells/clusters based on cell\_color parameter

\item[\code{select\_cells}] select subset of cells based on cell IDs

\item[\code{show\_other\_cells}] display not selected cells

\item[\code{other\_cell\_color}] color of not selected cells

\item[\code{other\_point\_size}] size of not selected cells

\item[\code{show\_cluster\_center}] plot center of selected clusters

\item[\code{show\_center\_label}] plot label of selected clusters

\item[\code{center\_point\_size}] size of center points

\item[\code{label\_size}] size of labels

\item[\code{label\_fontface}] font of labels

\item[\code{edge\_alpha}] column to use for alpha of the edges

\item[\code{point\_shape}] point with border or not (border or no\_border)

\item[\code{point\_size}] size of point (cell)

\item[\code{point\_border\_col}] color of border around points

\item[\code{point\_border\_stroke}] stroke size of border around points

\item[\code{show\_legend}] show legend

\item[\code{title}] title for plot, defaults to cell\_color parameter

\item[\code{legend\_text}] size of legend text

\item[\code{legend\_symbol\_size}] size of legend symbols

\item[\code{background\_color}] color of plot background

\item[\code{axis\_text}] size of axis text

\item[\code{axis\_title}] size of axis title

\item[\code{cow\_n\_col}] cowplot param: how many columns

\item[\code{cow\_rel\_h}] cowplot param: relative height

\item[\code{cow\_rel\_w}] cowplot param: relative width

\item[\code{cow\_align}] cowplot param: how to align

\item[\code{show\_plot}] show plot

\item[\code{return\_plot}] return ggplot object

\item[\code{save\_plot}] directly save the plot [boolean]

\item[\code{save\_param}] list of saving parameters from \code{\LinkA{all\_plots\_save\_function}{all.Rul.plots.Rul.save.Rul.function}}
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Description of parameters, see \code{\LinkA{dimPlot2D}{dimPlot2D}}. For 3D plots see \code{\LinkA{plotPCA\_3D}{plotPCA.Rul.3D}}
\end{Details}
%
\begin{Value}
ggplot
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    plotPCA(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{plotPCA\_2D}{plotPCA\_2D}{plotPCA.Rul.2D}
%
\begin{Description}\relax
Short wrapper for PCA visualization
\end{Description}
%
\begin{Usage}
\begin{verbatim}
plotPCA_2D(
  gobject,
  dim_reduction_name = "pca",
  default_save_name = "PCA_2D",
  ...
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{dim\_reduction\_name}] dimension reduction name

\item[\code{default\_save\_name}] default save name for saving, don't change, change save\_name in save\_param

\item[\code{groub\_by}] create multiple plots based on cell annotation column

\item[\code{group\_by\_subset}] subset the group\_by factor column

\item[\code{dim1\_to\_use}] dimension to use on x-axis

\item[\code{dim2\_to\_use}] dimension to use on y-axis

\item[\code{spat\_enr\_names}] names of spatial enrichment results to include

\item[\code{show\_NN\_network}] show underlying NN network

\item[\code{nn\_network\_to\_use}] type of NN network to use (kNN vs sNN)

\item[\code{network\_name}] name of NN network to use, if show\_NN\_network = TRUE

\item[\code{cell\_color}] color for cells (see details)

\item[\code{color\_as\_factor}] convert color column to factor

\item[\code{cell\_color\_code}] named vector with colors

\item[\code{cell\_color\_gradient}] vector with 3 colors for numeric data

\item[\code{gradient\_midpoint}] midpoint for color gradient

\item[\code{gradient\_limits}] vector with lower and upper limits

\item[\code{select\_cell\_groups}] select subset of cells/clusters based on cell\_color parameter

\item[\code{select\_cells}] select subset of cells based on cell IDs

\item[\code{show\_other\_cells}] display not selected cells

\item[\code{other\_cell\_color}] color of not selected cells

\item[\code{other\_point\_size}] size of not selected cells

\item[\code{show\_cluster\_center}] plot center of selected clusters

\item[\code{show\_center\_label}] plot label of selected clusters

\item[\code{center\_point\_size}] size of center points

\item[\code{label\_size}] size of labels

\item[\code{label\_fontface}] font of labels

\item[\code{edge\_alpha}] column to use for alpha of the edges

\item[\code{point\_shape}] point with border or not (border or no\_border)

\item[\code{point\_size}] size of point (cell)

\item[\code{point\_border\_col}] color of border around points

\item[\code{point\_border\_stroke}] stroke size of border around points

\item[\code{title}] title for plot, defaults to cell\_color parameter

\item[\code{show\_legend}] show legend

\item[\code{legend\_text}] size of legend text

\item[\code{legend\_symbol\_size}] size of legend symbols

\item[\code{background\_color}] color of plot background

\item[\code{axis\_text}] size of axis text

\item[\code{axis\_title}] size of axis title

\item[\code{cow\_n\_col}] cowplot param: how many columns

\item[\code{cow\_rel\_h}] cowplot param: relative height

\item[\code{cow\_rel\_w}] cowplot param: relative width

\item[\code{cow\_align}] cowplot param: how to align

\item[\code{show\_plot}] show plot

\item[\code{return\_plot}] return ggplot object

\item[\code{save\_plot}] directly save the plot [boolean]

\item[\code{save\_param}] list of saving parameters from \code{\LinkA{all\_plots\_save\_function}{all.Rul.plots.Rul.save.Rul.function}}
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Description of parameters, see \code{\LinkA{dimPlot2D}{dimPlot2D}}. For 3D plots see \code{\LinkA{plotPCA\_3D}{plotPCA.Rul.3D}}
\end{Details}
%
\begin{Value}
ggplot
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    plotPCA_2D(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{plotPCA\_3D}{plotPCA\_3D}{plotPCA.Rul.3D}
%
\begin{Description}\relax
Visualize cells according to 3D PCA dimension reduction
\end{Description}
%
\begin{Usage}
\begin{verbatim}
plotPCA_3D(
  gobject,
  dim_reduction_name = "pca",
  default_save_name = "PCA_3D",
  ...
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{dim\_reduction\_name}] pca dimension reduction name

\item[\code{default\_save\_name}] default save name for saving, ideally change save\_name in save\_param

\item[\code{dim1\_to\_use}] dimension to use on x-axis

\item[\code{dim2\_to\_use}] dimension to use on y-axis

\item[\code{dim3\_to\_use}] dimension to use on z-axis

\item[\code{show\_NN\_network}] show underlying NN network

\item[\code{nn\_network\_to\_use}] type of NN network to use (kNN vs sNN)

\item[\code{network\_name}] name of NN network to use, if show\_NN\_network = TRUE

\item[\code{cell\_color}] color for cells (see details)

\item[\code{color\_as\_factor}] convert color column to factor

\item[\code{cell\_color\_code}] named vector with colors

\item[\code{select\_cell\_groups}] select subset of cells/clusters based on cell\_color parameter

\item[\code{select\_cells}] select subset of cells based on cell IDs

\item[\code{show\_other\_cells}] display not selected cells

\item[\code{other\_cell\_color}] color of not selected cells

\item[\code{other\_point\_size}] size of not selected cells

\item[\code{show\_cluster\_center}] plot center of selected clusters

\item[\code{show\_center\_label}] plot label of selected clusters

\item[\code{center\_point\_size}] size of center points

\item[\code{label\_size}] size of labels

\item[\code{edge\_alpha}] column to use for alpha of the edges

\item[\code{point\_size}] size of point (cell)

\item[\code{show\_legend}] show legend

\item[\code{show\_plot}] show plot

\item[\code{return\_plot}] return ggplot object

\item[\code{save\_plot}] directly save the plot [boolean]

\item[\code{save\_param}] list of saving parameters from \code{\LinkA{all\_plots\_save\_function}{all.Rul.plots.Rul.save.Rul.function}}
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Description of parameters.
\end{Details}
%
\begin{Value}
plotly
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    plotPCA_3D(gobject)

\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{plotRankSpatvsExpr}{plotRankSpatvsExpr}{plotRankSpatvsExpr}
%
\begin{Description}\relax
Plots dotplot to compare ligand-receptor rankings from spatial and expression information
\end{Description}
%
\begin{Usage}
\begin{verbatim}
plotRankSpatvsExpr(
  gobject,
  combCC,
  expr_rnk_column = "LR_expr_rnk",
  spat_rnk_column = "LR_spat_rnk",
  midpoint = 10,
  size_range = c(0.01, 1.5),
  xlims = NULL,
  ylims = NULL,
  selected_ranks = c(1, 10, 20),
  show_plot = NA,
  return_plot = NA,
  save_plot = NA,
  save_param = list(),
  default_save_name = "plotRankSpatvsExpr"
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{combCC}] combined communinication scores from \code{\LinkA{combCCcom}{combCCcom}}

\item[\code{expr\_rnk\_column}] column with expression rank information to use

\item[\code{spat\_rnk\_column}] column with spatial rank information to use

\item[\code{midpoint}] midpoint of colors

\item[\code{size\_range}] size ranges of dotplot

\item[\code{xlims}] x-limits, numerical vector of 2

\item[\code{ylims}] y-limits, numerical vector of 2

\item[\code{selected\_ranks}] numerical vector, will be used to print out the percentage of top spatial ranks are recovered

\item[\code{show\_plot}] show plots

\item[\code{return\_plot}] return plotting object

\item[\code{save\_plot}] directly save the plot [boolean]

\item[\code{save\_param}] list of saving parameters from \code{\LinkA{all\_plots\_save\_function}{all.Rul.plots.Rul.save.Rul.function}}

\item[\code{default\_save\_name}] default save name for saving, don't change, change save\_name in save\_param
\end{ldescription}
\end{Arguments}
%
\begin{Value}
ggplot
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    plotRankSpatvsExpr(CPGscores)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{plotRecovery}{plotRecovery}{plotRecovery}
%
\begin{Description}\relax
Plots recovery plot to compare ligand-receptor rankings from spatial and expression information
\end{Description}
%
\begin{Usage}
\begin{verbatim}
plotRecovery(
  gobject,
  combCC,
  expr_rnk_column = "exprPI_rnk",
  spat_rnk_column = "spatPI_rnk",
  ground_truth = c("spatial", "expression"),
  show_plot = NA,
  return_plot = NA,
  save_plot = NA,
  save_param = list(),
  default_save_name = "plotRecovery"
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{combCC}] combined communinication scores from \code{\LinkA{combCCcom}{combCCcom}}

\item[\code{expr\_rnk\_column}] column with expression rank information to use

\item[\code{spat\_rnk\_column}] column with spatial rank information to use

\item[\code{ground\_truth}] what to consider as ground truth (default: spatial)

\item[\code{show\_plot}] show plots

\item[\code{return\_plot}] return plotting object

\item[\code{save\_plot}] directly save the plot [boolean]

\item[\code{save\_param}] list of saving parameters from \code{\LinkA{all\_plots\_save\_function}{all.Rul.plots.Rul.save.Rul.function}}

\item[\code{default\_save\_name}] default save name for saving, don't change, change save\_name in save\_param
\end{ldescription}
\end{Arguments}
%
\begin{Value}
ggplot
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    plotRecovery(CPGscores)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{plotRecovery\_sub}{plotRecovery\_sub}{plotRecovery.Rul.sub}
%
\begin{Description}\relax
Plots recovery plot to compare ligand-receptor rankings from spatial and expression information
\end{Description}
%
\begin{Usage}
\begin{verbatim}
plotRecovery_sub(combCC, first_col = "LR_expr_rnk", second_col = "LR_spat_rnk")
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{combCC}] combined communinication scores from \code{\LinkA{combCCcom}{combCCcom}}

\item[\code{first\_col}] first column to use

\item[\code{second\_col}] second column to use
\end{ldescription}
\end{Arguments}
%
\begin{Examples}
\begin{ExampleCode}
    plotRecovery_sub(CPGscores)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{plotStatDelaunayNetwork}{plotStatDelaunayNetwork}{plotStatDelaunayNetwork}
%
\begin{Description}\relax
Plots network statistics for a Delaunay network..
\end{Description}
%
\begin{Usage}
\begin{verbatim}
plotStatDelaunayNetwork(
  gobject,
  method = c("delaunayn_geometry", "RTriangle", "deldir"),
  dimensions = "all",
  maximum_distance = "auto",
  minimum_k = 0,
  options = "Pp",
  Y = TRUE,
  j = TRUE,
  S = 0,
  show_plot = NA,
  return_plot = NA,
  save_plot = NA,
  save_param = list(),
  default_save_name = "plotStatDelaunayNetwork",
  ...
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{dimensions}] which spatial dimensions to use (maximum 2 dimensions)

\item[\code{maximum\_distance}] distance cuttof for Delaunay neighbors to consider

\item[\code{minimum\_k}] minimum neigbhours if maximum\_distance != NULL

\item[\code{options}] (geometry) String containing extra control options for the underlying Qhull command; see the Qhull documentation (../doc/qhull/html/qdelaun.html) for the available options. (default = 'Pp', do not report precision problems)

\item[\code{Y}] (RTriangle) If TRUE prohibits the insertion of Steiner points on the mesh boundary.

\item[\code{j}] (RTriangle) If TRUE jettisons vertices that are not part of the final triangulation from the output.

\item[\code{S}] (RTriangle) Specifies the maximum number of added Steiner points.

\item[\code{show\_plot}] show plots

\item[\code{return\_plot}] return ggplot object

\item[\code{save\_plot}] directly save the plot [boolean]

\item[\code{save\_param}] list of saving parameters from \code{\LinkA{all\_plots\_save\_function}{all.Rul.plots.Rul.save.Rul.function}}

\item[\code{default\_save\_name}] default save name for saving, don't change, change save\_name in save\_param

\item[\code{...}] Other parameters of the \code{\LinkA{triangulate}{triangulate}} function

\item[\code{name}] name for spatial network (default = 'delaunay\_network')
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Plots statistics for a spatial Delaunay network as explained in \code{\LinkA{triangulate}{triangulate}}.
This can be used to further finetune the \code{\LinkA{createDelaunayNetwork}{createDelaunayNetwork}} function.
\end{Details}
%
\begin{Value}
giotto object with updated spatial network slot
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    plotStatDelaunayNetwork(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{plotTSNE}{plotTSNE}{plotTSNE}
%
\begin{Description}\relax
Short wrapper for tSNE visualization
\end{Description}
%
\begin{Usage}
\begin{verbatim}
plotTSNE(gobject, dim_reduction_name = "tsne", default_save_name = "tSNE", ...)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{dim\_reduction\_name}] dimension reduction name

\item[\code{default\_save\_name}] default save name for saving, don't change, change save\_name in save\_param

\item[\code{groub\_by}] create multiple plots based on cell annotation column

\item[\code{group\_by\_subset}] subset the group\_by factor column

\item[\code{dim1\_to\_use}] dimension to use on x-axis

\item[\code{dim2\_to\_use}] dimension to use on y-axis

\item[\code{spat\_enr\_names}] names of spatial enrichment results to include

\item[\code{show\_NN\_network}] show underlying NN network

\item[\code{nn\_network\_to\_use}] type of NN network to use (kNN vs sNN)

\item[\code{network\_name}] name of NN network to use, if show\_NN\_network = TRUE

\item[\code{cell\_color}] color for cells (see details)

\item[\code{color\_as\_factor}] convert color column to factor

\item[\code{cell\_color\_code}] named vector with colors

\item[\code{cell\_color\_gradient}] vector with 3 colors for numeric data

\item[\code{gradient\_midpoint}] midpoint for color gradient

\item[\code{gradient\_limits}] vector with lower and upper limits

\item[\code{select\_cell\_groups}] select subset of cells/clusters based on cell\_color parameter

\item[\code{select\_cells}] select subset of cells based on cell IDs

\item[\code{show\_other\_cells}] display not selected cells

\item[\code{other\_cell\_color}] color of not selected cells

\item[\code{other\_point\_size}] size of not selected cells

\item[\code{show\_cluster\_center}] plot center of selected clusters

\item[\code{show\_center\_label}] plot label of selected clusters

\item[\code{center\_point\_size}] size of center points

\item[\code{label\_size}] size of labels

\item[\code{label\_fontface}] font of labels

\item[\code{edge\_alpha}] column to use for alpha of the edges

\item[\code{point\_shape}] point with border or not (border or no\_border)

\item[\code{point\_size}] size of point (cell)

\item[\code{point\_border\_col}] color of border around points

\item[\code{point\_border\_stroke}] stroke size of border around points

\item[\code{title}] title for plot, defaults to cell\_color parameter

\item[\code{show\_legend}] show legend

\item[\code{legend\_text}] size of legend text

\item[\code{legend\_symbol\_size}] size of legend symbols

\item[\code{background\_color}] color of plot background

\item[\code{axis\_text}] size of axis text

\item[\code{axis\_title}] size of axis title

\item[\code{cow\_n\_col}] cowplot param: how many columns

\item[\code{cow\_rel\_h}] cowplot param: relative height

\item[\code{cow\_rel\_w}] cowplot param: relative width

\item[\code{cow\_align}] cowplot param: how to align

\item[\code{show\_plot}] show plot

\item[\code{return\_plot}] return ggplot object

\item[\code{save\_plot}] directly save the plot [boolean]

\item[\code{save\_param}] list of saving parameters from \code{\LinkA{all\_plots\_save\_function}{all.Rul.plots.Rul.save.Rul.function}}
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Description of parameters, see \code{\LinkA{dimPlot2D}{dimPlot2D}}. For 3D plots see \code{\LinkA{plotTSNE\_3D}{plotTSNE.Rul.3D}}
\end{Details}
%
\begin{Value}
ggplot
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    plotTSNE(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{plotTSNE\_2D}{plotTSNE\_2D}{plotTSNE.Rul.2D}
%
\begin{Description}\relax
Short wrapper for tSNE visualization
\end{Description}
%
\begin{Usage}
\begin{verbatim}
plotTSNE_2D(
  gobject,
  dim_reduction_name = "tsne",
  default_save_name = "tSNE_2D",
  ...
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{dim\_reduction\_name}] dimension reduction name

\item[\code{default\_save\_name}] default save name for saving, don't change, change save\_name in save\_param

\item[\code{groub\_by}] create multiple plots based on cell annotation column

\item[\code{group\_by\_subset}] subset the group\_by factor column

\item[\code{dim1\_to\_use}] dimension to use on x-axis

\item[\code{dim2\_to\_use}] dimension to use on y-axis

\item[\code{spat\_enr\_names}] names of spatial enrichment results to include

\item[\code{show\_NN\_network}] show underlying NN network

\item[\code{nn\_network\_to\_use}] type of NN network to use (kNN vs sNN)

\item[\code{network\_name}] name of NN network to use, if show\_NN\_network = TRUE

\item[\code{cell\_color}] color for cells (see details)

\item[\code{color\_as\_factor}] convert color column to factor

\item[\code{cell\_color\_code}] named vector with colors

\item[\code{cell\_color\_gradient}] vector with 3 colors for numeric data

\item[\code{gradient\_midpoint}] midpoint for color gradient

\item[\code{gradient\_limits}] vector with lower and upper limits

\item[\code{select\_cell\_groups}] select subset of cells/clusters based on cell\_color parameter

\item[\code{select\_cells}] select subset of cells based on cell IDs

\item[\code{show\_other\_cells}] display not selected cells

\item[\code{other\_cell\_color}] color of not selected cells

\item[\code{other\_point\_size}] size of not selected cells

\item[\code{show\_cluster\_center}] plot center of selected clusters

\item[\code{show\_center\_label}] plot label of selected clusters

\item[\code{center\_point\_size}] size of center points

\item[\code{label\_size}] size of labels

\item[\code{label\_fontface}] font of labels

\item[\code{edge\_alpha}] column to use for alpha of the edges

\item[\code{point\_shape}] point with border or not (border or no\_border)

\item[\code{point\_size}] size of point (cell)

\item[\code{point\_border\_col}] color of border around points

\item[\code{point\_border\_stroke}] stroke size of border around points

\item[\code{title}] title for plot, defaults to cell\_color parameter

\item[\code{show\_legend}] show legend

\item[\code{legend\_text}] size of legend text

\item[\code{legend\_symbol\_size}] size of legend symbols

\item[\code{background\_color}] color of plot background

\item[\code{axis\_text}] size of axis text

\item[\code{axis\_title}] size of axis title

\item[\code{cow\_n\_col}] cowplot param: how many columns

\item[\code{cow\_rel\_h}] cowplot param: relative height

\item[\code{cow\_rel\_w}] cowplot param: relative width

\item[\code{cow\_align}] cowplot param: how to align

\item[\code{show\_plot}] show plot

\item[\code{return\_plot}] return ggplot object

\item[\code{save\_plot}] directly save the plot [boolean]

\item[\code{save\_param}] list of saving parameters from \code{\LinkA{all\_plots\_save\_function}{all.Rul.plots.Rul.save.Rul.function}}
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Description of parameters, see \code{\LinkA{dimPlot2D}{dimPlot2D}}. For 3D plots see \code{\LinkA{plotTSNE\_3D}{plotTSNE.Rul.3D}}
\end{Details}
%
\begin{Value}
ggplot
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    plotTSNE_2D(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{plotTSNE\_3D}{plotTSNE\_3D}{plotTSNE.Rul.3D}
%
\begin{Description}\relax
Visualize cells according to dimension reduction coordinates
\end{Description}
%
\begin{Usage}
\begin{verbatim}
plotTSNE_3D(
  gobject,
  dim_reduction_name = "tsne",
  default_save_name = "TSNE_3D",
  ...
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{dim\_reduction\_name}] tsne dimension reduction name

\item[\code{default\_save\_name}] default save name for saving, don't change, change save\_name in save\_param

\item[\code{dim1\_to\_use}] dimension to use on x-axis

\item[\code{dim2\_to\_use}] dimension to use on y-axis

\item[\code{dim3\_to\_use}] dimension to use on z-axis

\item[\code{show\_NN\_network}] show underlying NN network

\item[\code{nn\_network\_to\_use}] type of NN network to use (kNN vs sNN)

\item[\code{network\_name}] name of NN network to use, if show\_NN\_network = TRUE

\item[\code{cell\_color}] color for cells (see details)

\item[\code{color\_as\_factor}] convert color column to factor

\item[\code{cell\_color\_code}] named vector with colors

\item[\code{select\_cell\_groups}] select subset of cells/clusters based on cell\_color parameter

\item[\code{select\_cells}] select subset of cells based on cell IDs

\item[\code{show\_other\_cells}] display not selected cells

\item[\code{other\_cell\_color}] color of not selected cells

\item[\code{other\_point\_size}] size of not selected cells

\item[\code{show\_cluster\_center}] plot center of selected clusters

\item[\code{show\_center\_label}] plot label of selected clusters

\item[\code{center\_point\_size}] size of center points

\item[\code{label\_size}] size of labels

\item[\code{edge\_alpha}] column to use for alpha of the edges

\item[\code{point\_size}] size of point (cell)

\item[\code{show\_legend}] show legend

\item[\code{show\_plot}] show plot

\item[\code{return\_plot}] return ggplot object

\item[\code{save\_plot}] directly save the plot [boolean]

\item[\code{save\_param}] list of saving parameters from \code{\LinkA{all\_plots\_save\_function}{all.Rul.plots.Rul.save.Rul.function}}
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Description of parameters.
\end{Details}
%
\begin{Value}
plotly
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    plotTSNE_3D(gobject)

\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{plotUMAP}{plotUMAP}{plotUMAP}
%
\begin{Description}\relax
Short wrapper for UMAP visualization
\end{Description}
%
\begin{Usage}
\begin{verbatim}
plotUMAP(gobject, dim_reduction_name = "umap", default_save_name = "UMAP", ...)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{dim\_reduction\_name}] dimension reduction name

\item[\code{default\_save\_name}] default save name for saving, don't change, change save\_name in save\_param

\item[\code{groub\_by}] create multiple plots based on cell annotation column

\item[\code{group\_by\_subset}] subset the group\_by factor column

\item[\code{dim1\_to\_use}] dimension to use on x-axis

\item[\code{dim2\_to\_use}] dimension to use on y-axis

\item[\code{spat\_enr\_names}] names of spatial enrichment results to include

\item[\code{show\_NN\_network}] show underlying NN network

\item[\code{nn\_network\_to\_use}] type of NN network to use (kNN vs sNN)

\item[\code{network\_name}] name of NN network to use, if show\_NN\_network = TRUE

\item[\code{cell\_color}] color for cells (see details)

\item[\code{color\_as\_factor}] convert color column to factor

\item[\code{cell\_color\_code}] named vector with colors

\item[\code{cell\_color\_gradient}] vector with 3 colors for numeric data

\item[\code{gradient\_midpoint}] midpoint for color gradient

\item[\code{gradient\_limits}] vector with lower and upper limits

\item[\code{select\_cell\_groups}] select subset of cells/clusters based on cell\_color parameter

\item[\code{select\_cells}] select subset of cells based on cell IDs

\item[\code{show\_other\_cells}] display not selected cells

\item[\code{other\_cell\_color}] color of not selected cells

\item[\code{other\_point\_size}] size of not selected cells

\item[\code{show\_cluster\_center}] plot center of selected clusters

\item[\code{show\_center\_label}] plot label of selected clusters

\item[\code{center\_point\_size}] size of center points

\item[\code{label\_size}] size of labels

\item[\code{label\_fontface}] font of labels

\item[\code{edge\_alpha}] column to use for alpha of the edges

\item[\code{point\_shape}] point with border or not (border or no\_border)

\item[\code{point\_size}] size of point (cell)

\item[\code{point\_border\_col}] color of border around points

\item[\code{point\_border\_stroke}] stroke size of border around points

\item[\code{title}] title for plot, defaults to cell\_color parameter

\item[\code{show\_legend}] show legend

\item[\code{legend\_text}] size of legend text

\item[\code{legend\_symbol\_size}] size of legend symbols

\item[\code{background\_color}] color of plot background

\item[\code{axis\_text}] size of axis text

\item[\code{axis\_title}] size of axis title

\item[\code{cow\_n\_col}] cowplot param: how many columns

\item[\code{cow\_rel\_h}] cowplot param: relative height

\item[\code{cow\_rel\_w}] cowplot param: relative width

\item[\code{cow\_align}] cowplot param: how to align

\item[\code{show\_plot}] show plot

\item[\code{return\_plot}] return ggplot object

\item[\code{save\_plot}] directly save the plot [boolean]

\item[\code{save\_param}] list of saving parameters from \code{\LinkA{all\_plots\_save\_function}{all.Rul.plots.Rul.save.Rul.function}}
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Description of parameters, see \code{\LinkA{dimPlot2D}{dimPlot2D}}. For 3D plots see \code{\LinkA{plotUMAP\_3D}{plotUMAP.Rul.3D}}
\end{Details}
%
\begin{Value}
ggplot
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    plotUMAP(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{plotUMAP\_2D}{plotUMAP\_2D}{plotUMAP.Rul.2D}
%
\begin{Description}\relax
Short wrapper for UMAP visualization
\end{Description}
%
\begin{Usage}
\begin{verbatim}
plotUMAP_2D(
  gobject,
  dim_reduction_name = "umap",
  default_save_name = "UMAP_2D",
  ...
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{dim\_reduction\_name}] dimension reduction name

\item[\code{default\_save\_name}] default save name for saving, don't change, change save\_name in save\_param

\item[\code{groub\_by}] create multiple plots based on cell annotation column

\item[\code{group\_by\_subset}] subset the group\_by factor column

\item[\code{dim1\_to\_use}] dimension to use on x-axis

\item[\code{dim2\_to\_use}] dimension to use on y-axis

\item[\code{spat\_enr\_names}] names of spatial enrichment results to include

\item[\code{show\_NN\_network}] show underlying NN network

\item[\code{nn\_network\_to\_use}] type of NN network to use (kNN vs sNN)

\item[\code{network\_name}] name of NN network to use, if show\_NN\_network = TRUE

\item[\code{cell\_color}] color for cells (see details)

\item[\code{color\_as\_factor}] convert color column to factor

\item[\code{cell\_color\_code}] named vector with colors

\item[\code{cell\_color\_gradient}] vector with 3 colors for numeric data

\item[\code{gradient\_midpoint}] midpoint for color gradient

\item[\code{gradient\_limits}] vector with lower and upper limits

\item[\code{select\_cell\_groups}] select subset of cells/clusters based on cell\_color parameter

\item[\code{select\_cells}] select subset of cells based on cell IDs

\item[\code{show\_other\_cells}] display not selected cells

\item[\code{other\_cell\_color}] color of not selected cells

\item[\code{other\_point\_size}] size of not selected cells

\item[\code{show\_cluster\_center}] plot center of selected clusters

\item[\code{show\_center\_label}] plot label of selected clusters

\item[\code{center\_point\_size}] size of center points

\item[\code{label\_size}] size of labels

\item[\code{label\_fontface}] font of labels

\item[\code{edge\_alpha}] column to use for alpha of the edges

\item[\code{point\_shape}] point with border or not (border or no\_border)

\item[\code{point\_size}] size of point (cell)

\item[\code{point\_border\_col}] color of border around points

\item[\code{point\_border\_stroke}] stroke size of border around points

\item[\code{title}] title for plot, defaults to cell\_color parameter

\item[\code{show\_legend}] show legend

\item[\code{legend\_text}] size of legend text

\item[\code{legend\_symbol\_size}] size of legend symbols

\item[\code{background\_color}] color of plot background

\item[\code{axis\_text}] size of axis text

\item[\code{axis\_title}] size of axis title

\item[\code{cow\_n\_col}] cowplot param: how many columns

\item[\code{cow\_rel\_h}] cowplot param: relative height

\item[\code{cow\_rel\_w}] cowplot param: relative width

\item[\code{cow\_align}] cowplot param: how to align

\item[\code{show\_plot}] show plot

\item[\code{return\_plot}] return ggplot object

\item[\code{save\_plot}] directly save the plot [boolean]

\item[\code{save\_param}] list of saving parameters from \code{\LinkA{all\_plots\_save\_function}{all.Rul.plots.Rul.save.Rul.function}}
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Description of parameters, see \code{\LinkA{dimPlot2D}{dimPlot2D}}. For 3D plots see \code{\LinkA{plotUMAP\_3D}{plotUMAP.Rul.3D}}
\end{Details}
%
\begin{Value}
ggplot
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    plotUMAP_2D(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{plotUMAP\_3D}{plotUMAP\_3D}{plotUMAP.Rul.3D}
%
\begin{Description}\relax
Visualize cells according to dimension reduction coordinates
\end{Description}
%
\begin{Usage}
\begin{verbatim}
plotUMAP_3D(
  gobject,
  dim_reduction_name = "umap",
  default_save_name = "UMAP_3D",
  ...
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{dim\_reduction\_name}] umap dimension reduction name

\item[\code{default\_save\_name}] default save name for saving, don't change, change save\_name in save\_param

\item[\code{dim1\_to\_use}] dimension to use on x-axis

\item[\code{dim2\_to\_use}] dimension to use on y-axis

\item[\code{dim3\_to\_use}] dimension to use on z-axis

\item[\code{show\_NN\_network}] show underlying NN network

\item[\code{nn\_network\_to\_use}] type of NN network to use (kNN vs sNN)

\item[\code{network\_name}] name of NN network to use, if show\_NN\_network = TRUE

\item[\code{cell\_color}] color for cells (see details)

\item[\code{color\_as\_factor}] convert color column to factor

\item[\code{cell\_color\_code}] named vector with colors

\item[\code{select\_cell\_groups}] select subset of cells/clusters based on cell\_color parameter

\item[\code{select\_cells}] select subset of cells based on cell IDs

\item[\code{show\_other\_cells}] display not selected cells

\item[\code{other\_cell\_color}] color of not selected cells

\item[\code{other\_point\_size}] size of not selected cells

\item[\code{show\_cluster\_center}] plot center of selected clusters

\item[\code{show\_center\_label}] plot label of selected clusters

\item[\code{center\_point\_size}] size of center points

\item[\code{label\_size}] size of labels

\item[\code{edge\_alpha}] column to use for alpha of the edges

\item[\code{point\_size}] size of point (cell)

\item[\code{show\_legend}] show legend

\item[\code{show\_plot}] show plot

\item[\code{return\_plot}] return ggplot object

\item[\code{save\_plot}] directly save the plot [boolean]

\item[\code{save\_param}] list of saving parameters from \code{\LinkA{all\_plots\_save\_function}{all.Rul.plots.Rul.save.Rul.function}}
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Description of parameters.
\end{Details}
%
\begin{Value}
plotly
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    plotUMAP_3D(gobject)

\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{plot\_network\_layer\_ggplot}{plot\_network\_layer\_ggplot}{plot.Rul.network.Rul.layer.Rul.ggplot}
%
\begin{Description}\relax
Visualize cells in network layer according to dimension reduction coordinates
\end{Description}
%
\begin{Usage}
\begin{verbatim}
plot_network_layer_ggplot(
  ggobject,
  annotated_network_DT,
  edge_alpha = NULL,
  show_legend = T
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{annotated\_network\_DT}] annotated network data.table of selected cells

\item[\code{edge\_alpha}] alpha of network edges

\item[\code{show\_legend}] show legend

\item[\code{gobject}] giotto object
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Description of parameters.
\end{Details}
%
\begin{Value}
ggplot
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    plot_network_layer_ggplot(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{plot\_point\_layer\_ggplot}{plot\_point\_layer\_ggplot}{plot.Rul.point.Rul.layer.Rul.ggplot}
%
\begin{Description}\relax
Visualize cells in point layer according to dimension reduction coordinates
\end{Description}
%
\begin{Usage}
\begin{verbatim}
plot_point_layer_ggplot(
  ggobject,
  annotated_DT_selected,
  annotated_DT_other,
  cell_color = NULL,
  color_as_factor = T,
  cell_color_code = NULL,
  cell_color_gradient = c("blue", "white", "red"),
  gradient_midpoint = 0,
  gradient_limits = NULL,
  select_cell_groups = NULL,
  select_cells = NULL,
  point_size = 1,
  point_border_col = "black",
  point_border_stroke = 0.1,
  show_cluster_center = F,
  show_center_label = T,
  center_point_size = 4,
  center_point_border_col = "black",
  center_point_border_stroke = 0.1,
  label_size = 4,
  label_fontface = "bold",
  edge_alpha = NULL,
  show_other_cells = T,
  other_cell_color = "lightgrey",
  other_point_size = 0.5,
  show_legend = T
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{annotated\_DT\_selected}] annotated data.table of selected cells

\item[\code{annotated\_DT\_other}] annotated data.table of not selected cells

\item[\code{cell\_color}] color for cells (see details)

\item[\code{color\_as\_factor}] convert color column to factor

\item[\code{cell\_color\_code}] named vector with colors

\item[\code{cell\_color\_gradient}] vector with 3 colors for numeric data

\item[\code{gradient\_midpoint}] midpoint for color gradient

\item[\code{gradient\_limits}] vector with lower and upper limits

\item[\code{select\_cell\_groups}] select subset of cells/clusters based on cell\_color parameter

\item[\code{select\_cells}] select subset of cells based on cell IDs

\item[\code{point\_size}] size of point (cell)

\item[\code{point\_border\_col}] color of border around points

\item[\code{point\_border\_stroke}] stroke size of border around points

\item[\code{show\_cluster\_center}] plot center of selected clusters

\item[\code{show\_center\_label}] plot label of selected clusters

\item[\code{center\_point\_size}] size of center points

\item[\code{label\_size}] size of labels

\item[\code{label\_fontface}] font of labels

\item[\code{edge\_alpha}] column to use for alpha of the edges

\item[\code{show\_other\_cells}] display not selected cells

\item[\code{other\_cell\_color}] color of not selected cells

\item[\code{other\_point\_size}] size of not selected cells

\item[\code{show\_legend}] show legend

\item[\code{gobject}] giotto object
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Description of parameters.
\end{Details}
%
\begin{Value}
ggplot
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    plot_point_layer_ggplot(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{plot\_point\_layer\_ggplot\_noFILL}{plot\_point\_layer\_ggplot\_noFILL}{plot.Rul.point.Rul.layer.Rul.ggplot.Rul.noFILL}
%
\begin{Description}\relax
Visualize cells in point layer according to dimension reduction coordinates without borders
\end{Description}
%
\begin{Usage}
\begin{verbatim}
plot_point_layer_ggplot_noFILL(
  ggobject,
  annotated_DT_selected,
  annotated_DT_other,
  cell_color = NULL,
  color_as_factor = T,
  cell_color_code = NULL,
  cell_color_gradient = c("blue", "white", "red"),
  gradient_midpoint = 0,
  gradient_limits = NULL,
  select_cell_groups = NULL,
  select_cells = NULL,
  point_size = 1,
  show_cluster_center = F,
  show_center_label = T,
  center_point_size = 4,
  label_size = 4,
  label_fontface = "bold",
  edge_alpha = NULL,
  show_other_cells = T,
  other_cell_color = "lightgrey",
  other_point_size = 0.5,
  show_legend = T
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{annotated\_DT\_selected}] annotated data.table of selected cells

\item[\code{annotated\_DT\_other}] annotated data.table of not selected cells

\item[\code{cell\_color}] color for cells (see details)

\item[\code{color\_as\_factor}] convert color column to factor

\item[\code{cell\_color\_code}] named vector with colors

\item[\code{cell\_color\_gradient}] vector with 3 colors for numeric data

\item[\code{gradient\_midpoint}] midpoint for color gradient

\item[\code{gradient\_limits}] vector with lower and upper limits

\item[\code{select\_cell\_groups}] select subset of cells/clusters based on cell\_color parameter

\item[\code{select\_cells}] select subset of cells based on cell IDs

\item[\code{point\_size}] size of point (cell)

\item[\code{show\_cluster\_center}] plot center of selected clusters

\item[\code{show\_center\_label}] plot label of selected clusters

\item[\code{center\_point\_size}] size of center points

\item[\code{label\_size}] size of labels

\item[\code{label\_fontface}] font of labels

\item[\code{edge\_alpha}] column to use for alpha of the edges

\item[\code{show\_other\_cells}] display not selected cells

\item[\code{other\_cell\_color}] color of not selected cells

\item[\code{other\_point\_size}] size of not selected cells

\item[\code{show\_legend}] show legend

\item[\code{gobject}] giotto object
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Description of parameters.
\end{Details}
%
\begin{Value}
ggplot
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    plot_point_layer_ggplot_noFILL(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{plot\_spat\_point\_layer\_ggplot}{plot\_spat\_point\_layer\_ggplot}{plot.Rul.spat.Rul.point.Rul.layer.Rul.ggplot}
%
\begin{Description}\relax
creat ggplot point layer for spatial coordinates
\end{Description}
%
\begin{Usage}
\begin{verbatim}
plot_spat_point_layer_ggplot(
  ggobject,
  sdimx = NULL,
  sdimy = NULL,
  cell_locations_metadata_selected,
  cell_locations_metadata_other,
  cell_color = NULL,
  color_as_factor = T,
  cell_color_code = NULL,
  cell_color_gradient = c("blue", "white", "red"),
  gradient_midpoint = NULL,
  gradient_limits = NULL,
  select_cell_groups = NULL,
  select_cells = NULL,
  point_size = 2,
  point_border_col = "lightgrey",
  point_border_stroke = 0.1,
  show_cluster_center = F,
  show_center_label = T,
  center_point_size = 4,
  center_point_border_col = "black",
  center_point_border_stroke = 0.1,
  label_size = 4,
  label_fontface = "bold",
  show_other_cells = T,
  other_cell_color = "lightgrey",
  other_point_size = 1,
  show_legend = TRUE
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{sdimx}] x-axis dimension name (default = 'sdimx')

\item[\code{sdimy}] y-axis dimension name (default = 'sdimy')

\item[\code{cell\_locations\_metadata\_selected}] annotated location from selected cells

\item[\code{cell\_locations\_metadata\_other}] annotated location from non-selected cells

\item[\code{cell\_color}] color for cells (see details)

\item[\code{color\_as\_factor}] convert color column to factor

\item[\code{cell\_color\_code}] named vector with colors

\item[\code{cell\_color\_gradient}] vector with 3 colors for numeric data

\item[\code{gradient\_midpoint}] midpoint for color gradient

\item[\code{gradient\_limits}] vector with lower and upper limits

\item[\code{select\_cell\_groups}] select subset of cells/clusters based on cell\_color parameter

\item[\code{select\_cells}] select subset of cells based on cell IDs

\item[\code{point\_size}] size of point (cell)

\item[\code{point\_border\_col}] color of border around points

\item[\code{point\_border\_stroke}] stroke size of border around points

\item[\code{show\_cluster\_center}] plot center of selected clusters

\item[\code{show\_center\_label}] plot label of selected clusters

\item[\code{center\_point\_size}] size of center points

\item[\code{label\_size}] size of labels

\item[\code{label\_fontface}] font of labels

\item[\code{show\_other\_cells}] display not selected cells

\item[\code{other\_cell\_color}] color for not selected cells

\item[\code{other\_point\_size}] point size for not selected cells

\item[\code{show\_legend}] show legend

\item[\code{gobject}] giotto object
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Description of parameters.
\end{Details}
%
\begin{Value}
ggplot
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    plot_spat_point_layer_ggplot(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{plot\_spat\_point\_layer\_ggplot\_noFILL}{plot\_spat\_point\_layer\_ggplot\_noFILL}{plot.Rul.spat.Rul.point.Rul.layer.Rul.ggplot.Rul.noFILL}
%
\begin{Description}\relax
creat ggplot point layer for spatial coordinates without borders
\end{Description}
%
\begin{Usage}
\begin{verbatim}
plot_spat_point_layer_ggplot_noFILL(
  ggobject,
  sdimx = NULL,
  sdimy = NULL,
  cell_locations_metadata_selected,
  cell_locations_metadata_other,
  cell_color = NULL,
  color_as_factor = T,
  cell_color_code = NULL,
  cell_color_gradient = c("blue", "white", "red"),
  gradient_midpoint = NULL,
  gradient_limits = NULL,
  select_cell_groups = NULL,
  select_cells = NULL,
  point_size = 2,
  show_cluster_center = F,
  show_center_label = T,
  center_point_size = 4,
  label_size = 4,
  label_fontface = "bold",
  show_other_cells = T,
  other_cell_color = "lightgrey",
  other_point_size = 1,
  show_legend = TRUE
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{sdimx}] x-axis dimension name (default = 'sdimx')

\item[\code{sdimy}] y-axis dimension name (default = 'sdimy')

\item[\code{cell\_locations\_metadata\_selected}] annotated location from selected cells

\item[\code{cell\_locations\_metadata\_other}] annotated location from non-selected cells

\item[\code{cell\_color}] color for cells (see details)

\item[\code{color\_as\_factor}] convert color column to factor

\item[\code{cell\_color\_code}] named vector with colors

\item[\code{cell\_color\_gradient}] vector with 3 colors for numeric data

\item[\code{gradient\_midpoint}] midpoint for color gradient

\item[\code{gradient\_limits}] vector with lower and upper limits

\item[\code{select\_cell\_groups}] select subset of cells/clusters based on cell\_color parameter

\item[\code{select\_cells}] select subset of cells based on cell IDs

\item[\code{point\_size}] size of point (cell)

\item[\code{show\_cluster\_center}] plot center of selected clusters

\item[\code{show\_center\_label}] plot label of selected clusters

\item[\code{center\_point\_size}] size of center points

\item[\code{label\_size}] size of labels

\item[\code{label\_fontface}] font of labels

\item[\code{show\_other\_cells}] display not selected cells

\item[\code{other\_cell\_color}] color for not selected cells

\item[\code{other\_point\_size}] point size for not selected cells

\item[\code{show\_legend}] show legend

\item[\code{gobject}] giotto object
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Description of parameters.
\end{Details}
%
\begin{Value}
ggplot
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    plot_spat_point_layer_ggplot_noFILL(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{plot\_spat\_voronoi\_layer\_ggplot}{plot\_spat\_voronoi\_layer\_ggplot}{plot.Rul.spat.Rul.voronoi.Rul.layer.Rul.ggplot}
%
\begin{Description}\relax
creat ggplot point layer for spatial coordinates without borders
\end{Description}
%
\begin{Usage}
\begin{verbatim}
plot_spat_voronoi_layer_ggplot(
  ggobject,
  sdimx = NULL,
  sdimy = NULL,
  cell_locations_metadata_selected,
  cell_locations_metadata_other,
  cell_color = NULL,
  color_as_factor = T,
  cell_color_code = NULL,
  cell_color_gradient = c("blue", "white", "red"),
  gradient_midpoint = NULL,
  gradient_limits = NULL,
  select_cell_groups = NULL,
  select_cells = NULL,
  point_size = 2,
  show_cluster_center = F,
  show_center_label = T,
  center_point_size = 4,
  label_size = 4,
  label_fontface = "bold",
  show_other_cells = T,
  other_cell_color = "lightgrey",
  other_point_size = 1,
  background_color = "white",
  vor_border_color = "white",
  vor_max_radius = 200,
  show_legend = TRUE
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{sdimx}] x-axis dimension name (default = 'sdimx')

\item[\code{sdimy}] y-axis dimension name (default = 'sdimy')

\item[\code{cell\_locations\_metadata\_selected}] annotated location from selected cells

\item[\code{cell\_locations\_metadata\_other}] annotated location from non-selected cells

\item[\code{cell\_color}] color for cells (see details)

\item[\code{color\_as\_factor}] convert color column to factor

\item[\code{cell\_color\_code}] named vector with colors

\item[\code{cell\_color\_gradient}] vector with 3 colors for numeric data

\item[\code{gradient\_midpoint}] midpoint for color gradient

\item[\code{gradient\_limits}] vector with lower and upper limits

\item[\code{select\_cell\_groups}] select subset of cells/clusters based on cell\_color parameter

\item[\code{select\_cells}] select subset of cells based on cell IDs

\item[\code{point\_size}] size of point (cell)

\item[\code{show\_cluster\_center}] plot center of selected clusters

\item[\code{show\_center\_label}] plot label of selected clusters

\item[\code{center\_point\_size}] size of center points

\item[\code{label\_size}] size of labels

\item[\code{label\_fontface}] font of labels

\item[\code{show\_other\_cells}] display not selected cells

\item[\code{other\_cell\_color}] color for not selected cells

\item[\code{other\_point\_size}] point size for not selected cells

\item[\code{background\_color}] background color

\item[\code{vor\_border\_color}] borde colorr of voronoi plot

\item[\code{vor\_max\_radius}] maximum radius for voronoi 'cells'

\item[\code{show\_legend}] show legend

\item[\code{gobject}] giotto object
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Description of parameters.
\end{Details}
%
\begin{Value}
ggplot
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    plot_spat_voronoi_layer_ggplot(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{print.giotto}{print method for giotto class}{print.giotto}
\keyword{giotto,}{print.giotto}
\keyword{object}{print.giotto}
%
\begin{Description}\relax
print method for giotto class.
Prints the chosen number of genes (rows) and cells (columns) from the raw count matrix.
Also print the spatial locations for the chosen number of cells.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
print.giotto(object, ...)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{nr\_genes}] number of genes (rows) to print

\item[\code{nr\_cells}] number of cells (columns) to print
\end{ldescription}
\end{Arguments}
\inputencoding{utf8}
\HeaderA{projection\_fun}{projection\_fun}{projection.Rul.fun}
%
\begin{Description}\relax
project a point onto a plane
\end{Description}
%
\begin{Usage}
\begin{verbatim}
projection_fun(point_to_project, plane_point, plane_norm)
\end{verbatim}
\end{Usage}
\inputencoding{utf8}
\HeaderA{rankEnrich}{rankEnrich}{rankEnrich}
%
\begin{Description}\relax
Function to calculate gene signature enrichment scores per spatial position using a rank based approach.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
rankEnrich(
  gobject,
  sign_matrix,
  expression_values = c("normalized", "scaled", "custom"),
  reverse_log_scale = TRUE,
  logbase = 2,
  output_enrichment = c("original", "zscore")
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] Giotto object

\item[\code{sign\_matrix}] Matrix of signature genes for each cell type / process

\item[\code{expression\_values}] expression values to use

\item[\code{reverse\_log\_scale}] reverse expression values from log scale

\item[\code{logbase}] log base to use if reverse\_log\_scale = TRUE

\item[\code{output\_enrichment}] how to return enrichment output
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
sign\_matrix: a rank-fold matrix with genes as row names and cell-types as column names.
Alternatively a scRNA-seq matrix and vector with clusters can be provided to makeSignMatrixRank, which will create
the matrix for you. \\{}

First a new rank is calculated as R = (R1*R2)\textasciicircum{}(1/2), where R1 is the rank of
fold-change for each gene in each spot and R2 is the rank of each marker in each cell type.
The Rank-Biased Precision is then calculated as: RBP = (1 - 0.99) * (0.99)\textasciicircum{}(R - 1)
and the final enrichment score is then calculated as the sum of top 100 RBPs.
\end{Details}
%
\begin{Value}
data.table with enrichment results
\end{Value}
%
\begin{SeeAlso}\relax
\code{\LinkA{makeSignMatrixRank}{makeSignMatrixRank}}
\end{SeeAlso}
%
\begin{Examples}
\begin{ExampleCode}
    rankEnrich(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{rankSpatialCorGroups}{rankSpatialCorGroups}{rankSpatialCorGroups}
%
\begin{Description}\relax
Rank spatial correlated clusters according to correlation structure
\end{Description}
%
\begin{Usage}
\begin{verbatim}
rankSpatialCorGroups(
  gobject,
  spatCorObject,
  use_clus_name = NULL,
  show_plot = NA,
  return_plot = FALSE,
  save_plot = NA,
  save_param = list(),
  default_save_name = "rankSpatialCorGroups"
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{spatCorObject}] spatial correlation object

\item[\code{use\_clus\_name}] name of clusters to visualize (from clusterSpatialCorGenes())

\item[\code{show\_plot}] show plot

\item[\code{return\_plot}] return ggplot object

\item[\code{save\_plot}] directly save the plot [boolean]

\item[\code{save\_param}] list of saving parameters from \code{\LinkA{all\_plots\_save\_function}{all.Rul.plots.Rul.save.Rul.function}}

\item[\code{default\_save\_name}] default save name for saving, don't change, change save\_name in save\_param
\end{ldescription}
\end{Arguments}
%
\begin{Value}
data.table with positive (within group) and negative (outside group) scores
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    rankSpatialCorGroups(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{rank\_binarize}{rank\_binarize}{rank.Rul.binarize}
%
\begin{Description}\relax
create binarized scores from a vector using arbitrary rank
\end{Description}
%
\begin{Usage}
\begin{verbatim}
rank_binarize(x, max_rank = 200)
\end{verbatim}
\end{Usage}
\inputencoding{utf8}
\HeaderA{readExprMatrix}{readExprMatrix}{readExprMatrix}
%
\begin{Description}\relax
Function to read an expression matrix into a sparse matrix.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
readExprMatrix(path, cores = NA)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{path}] path to the expression matrix
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
The expression matrix needs to have both unique column names and row names
\end{Details}
%
\begin{Value}
sparse matrix
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    readExprMatrix()
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{readGiottoInstructions}{readGiottoInstrunctions}{readGiottoInstructions}
%
\begin{Description}\relax
Retrieves the instruction associated with the provided parameter
\end{Description}
%
\begin{Usage}
\begin{verbatim}
readGiottoInstructions(giotto_instructions, param = NULL)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{giotto\_instructions}] giotto object or result from createGiottoInstructions()

\item[\code{param}] parameter to retrieve
\end{ldescription}
\end{Arguments}
%
\begin{Value}
specific parameter
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    readGiottoInstrunctions()
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{read\_crossSection}{read\_crossSection}{read.Rul.crossSection}
%
\begin{Description}\relax
read a cross section object from a giotto object
\end{Description}
%
\begin{Usage}
\begin{verbatim}
read_crossSection(gobject, name = NULL, spatial_network_name = NULL)
\end{verbatim}
\end{Usage}
\inputencoding{utf8}
\HeaderA{removeCellAnnotation}{removeCellAnnotation}{removeCellAnnotation}
%
\begin{Description}\relax
removes cell annotation of giotto object
\end{Description}
%
\begin{Usage}
\begin{verbatim}
removeCellAnnotation(gobject, columns = NULL, return_gobject = TRUE)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{columns}] names of columns to remove

\item[\code{return\_gobject}] boolean: return giotto object (default = TRUE)
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
if return\_gobject = FALSE, it will return the cell metadata
\end{Details}
%
\begin{Value}
giotto object
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    removeCellAnnotation(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{removeGeneAnnotation}{removeGeneAnnotation}{removeGeneAnnotation}
%
\begin{Description}\relax
removes gene annotation of giotto object
\end{Description}
%
\begin{Usage}
\begin{verbatim}
removeGeneAnnotation(gobject, columns = NULL, return_gobject = TRUE)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{columns}] names of columns to remove

\item[\code{return\_gobject}] boolean: return giotto object (default = TRUE)
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
if return\_gobject = FALSE, it will return the gene metadata
\end{Details}
%
\begin{Value}
giotto object
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    removeGeneAnnotation(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{replaceGiottoInstructions}{replaceGiottoInstructions}{replaceGiottoInstructions}
%
\begin{Description}\relax
Function to replace all instructions from giotto object
\end{Description}
%
\begin{Usage}
\begin{verbatim}
replaceGiottoInstructions(gobject, instructions = NULL)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{instructions}] new instructions (e.g. result from createGiottoInstructions)
\end{ldescription}
\end{Arguments}
%
\begin{Value}
named vector with giotto instructions
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    replaceGiottoInstructions()
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{reshape\_to\_data\_point}{reshape\_to\_data\_point}{reshape.Rul.to.Rul.data.Rul.point}
%
\begin{Description}\relax
reshape a mesh grid line object to data point matrix
\end{Description}
%
\begin{Usage}
\begin{verbatim}
reshape_to_data_point(mesh_grid_obj)
\end{verbatim}
\end{Usage}
\inputencoding{utf8}
\HeaderA{reshape\_to\_mesh\_grid\_obj}{reshape\_to\_mesh\_grid\_obj}{reshape.Rul.to.Rul.mesh.Rul.grid.Rul.obj}
%
\begin{Description}\relax
reshape a data point matrix to a mesh grid line object
\end{Description}
%
\begin{Usage}
\begin{verbatim}
reshape_to_mesh_grid_obj(data_points, mesh_grid_n)
\end{verbatim}
\end{Usage}
\inputencoding{utf8}
\HeaderA{rowMeans\_giotto}{rowMeans\_giotto}{rowMeans.Rul.giotto}
%
\begin{Description}\relax
rowMeans\_giotto
\end{Description}
%
\begin{Usage}
\begin{verbatim}
rowMeans_giotto(mymatrix)
\end{verbatim}
\end{Usage}
\inputencoding{utf8}
\HeaderA{rowSums\_giotto}{rowSums\_giotto}{rowSums.Rul.giotto}
%
\begin{Description}\relax
rowSums\_giotto
\end{Description}
%
\begin{Usage}
\begin{verbatim}
rowSums_giotto(mymatrix)
\end{verbatim}
\end{Usage}
\inputencoding{utf8}
\HeaderA{runPCA}{runPCA}{runPCA}
%
\begin{Description}\relax
runs a Principal Component Analysis
\end{Description}
%
\begin{Usage}
\begin{verbatim}
runPCA(
  gobject,
  expression_values = c("normalized", "scaled", "custom"),
  reduction = c("cells", "genes"),
  name = "pca",
  genes_to_use = NULL,
  return_gobject = TRUE,
  scale_unit = F,
  ncp = 200,
  ...
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{expression\_values}] expression values to use

\item[\code{reduction}] cells or genes

\item[\code{name}] arbitrary name for PCA run

\item[\code{genes\_to\_use}] subset of genes to use for PCA

\item[\code{return\_gobject}] boolean: return giotto object (default = TRUE)

\item[\code{scale\_unit}] scale features before PCA

\item[\code{ncp}] number of principal components to calculate

\item[\code{...}] additional parameters for PCA (see details)
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
See \code{\LinkA{PCA}{PCA}} for more information about other parameters.
\end{Details}
%
\begin{Value}
giotto object with updated PCA dimension recuction
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    runPCA(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{runtSNE}{runtSNE}{runtSNE}
%
\begin{Description}\relax
run tSNE
\end{Description}
%
\begin{Usage}
\begin{verbatim}
runtSNE(
  gobject,
  expression_values = c("normalized", "scaled", "custom"),
  reduction = c("cells", "genes"),
  dim_reduction_to_use = "pca",
  dim_reduction_name = "pca",
  dimensions_to_use = 1:10,
  name = "tsne",
  genes_to_use = NULL,
  return_gobject = TRUE,
  dims = 2,
  perplexity = 30,
  theta = 0.5,
  do_PCA_first = F,
  set_seed = T,
  seed_number = 1234,
  ...
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{expression\_values}] expression values to use

\item[\code{reduction}] cells or genes

\item[\code{dim\_reduction\_to\_use}] use another dimension reduction set as input

\item[\code{dim\_reduction\_name}] name of dimension reduction set to use

\item[\code{dimensions\_to\_use}] number of dimensions to use as input

\item[\code{name}] arbitrary name for tSNE run

\item[\code{genes\_to\_use}] if dim\_reduction\_to\_use = NULL, which genes to use

\item[\code{return\_gobject}] boolean: return giotto object (default = TRUE)

\item[\code{dims}] tSNE param: number of dimensions to return

\item[\code{perplexity}] tSNE param: perplexity

\item[\code{theta}] tSNE param: theta

\item[\code{do\_PCA\_first}] tSNE param: do PCA before tSNE (default = FALSE)

\item[\code{set\_seed}] use of seed

\item[\code{seed\_number}] seed number to use

\item[\code{...}] additional tSNE parameters
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
See \code{\LinkA{Rtsne}{Rtsne}} for more information about these and other parameters. \\{}
\begin{itemize}

\item{} Input for tSNE dimension reduction can be another dimension reduction (default = 'pca')
\item{} To use gene expression as input set dim\_reduction\_to\_use = NULL
\item{} multiple tSNE results can be stored by changing the \emph{name} of the analysis

\end{itemize}

\end{Details}
%
\begin{Value}
giotto object with updated tSNE dimension recuction
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    runtSNE(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{runUMAP}{runUMAP}{runUMAP}
%
\begin{Description}\relax
run UMAP
\end{Description}
%
\begin{Usage}
\begin{verbatim}
runUMAP(
  gobject,
  expression_values = c("normalized", "scaled", "custom"),
  reduction = c("cells", "genes"),
  dim_reduction_to_use = "pca",
  dim_reduction_name = "pca",
  dimensions_to_use = 1:10,
  name = "umap",
  genes_to_use = NULL,
  return_gobject = TRUE,
  n_neighbors = 40,
  n_components = 2,
  n_epochs = 400,
  min_dist = 0.01,
  n_threads = 1,
  spread = 5,
  set_seed = T,
  seed_number = 1234,
  ...
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{expression\_values}] expression values to use

\item[\code{reduction}] cells or genes

\item[\code{dim\_reduction\_to\_use}] use another dimension reduction set as input

\item[\code{dim\_reduction\_name}] name of dimension reduction set to use

\item[\code{dimensions\_to\_use}] number of dimensions to use as input

\item[\code{name}] arbitrary name for UMAP run

\item[\code{genes\_to\_use}] if dim\_reduction\_to\_use = NULL, which genes to use

\item[\code{return\_gobject}] boolean: return giotto object (default = TRUE)

\item[\code{n\_neighbors}] UMAP param: number of neighbors

\item[\code{n\_components}] UMAP param: number of components

\item[\code{n\_epochs}] UMAP param: number of epochs

\item[\code{min\_dist}] UMAP param: minimum distance

\item[\code{n\_threads}] UMAP param: threads to use

\item[\code{spread}] UMAP param: spread

\item[\code{set\_seed}] use of seed

\item[\code{seed\_number}] seed number to use

\item[\code{...}] additional UMAP parameters
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
See \code{\LinkA{umap}{umap}} for more information about these and other parameters.
\begin{itemize}

\item{} Input for UMAP dimension reduction can be another dimension reduction (default = 'pca')
\item{} To use gene expression as input set dim\_reduction\_to\_use = NULL
\item{} multiple UMAP results can be stored by changing the \emph{name} of the analysis

\end{itemize}

\end{Details}
%
\begin{Value}
giotto object with updated UMAP dimension recuction
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    runUMAP(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{selectPatternGenes}{selectPatternGenes}{selectPatternGenes}
%
\begin{Description}\relax
Select genes correlated with spatial patterns
\end{Description}
%
\begin{Usage}
\begin{verbatim}
selectPatternGenes(
  spatPatObj,
  dimensions = 1:5,
  top_pos_genes = 10,
  top_neg_genes = 10,
  min_pos_cor = 0.5,
  min_neg_cor = -0.5,
  return_top_selection = FALSE
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{spatPatObj}] Output from detectSpatialPatterns

\item[\code{dimensions}] dimensions to identify correlated genes for.

\item[\code{top\_pos\_genes}] Top positively correlated genes.

\item[\code{top\_neg\_genes}] Top negatively correlated genes.

\item[\code{min\_pos\_cor}] Minimum positive correlation score to include a gene.

\item[\code{min\_neg\_cor}] Minimum negative correlation score to include a gene.
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Description.
\end{Details}
%
\begin{Value}
Data.table with genes associated with selected dimension (PC).
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    selectPatternGenes(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{select\_expression\_values}{select\_expression\_values}{select.Rul.expression.Rul.values}
%
\begin{Description}\relax
helper function to select expression values
\end{Description}
%
\begin{Usage}
\begin{verbatim}
select_expression_values(gobject, values)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{values}] expression values to extract
\end{ldescription}
\end{Arguments}
%
\begin{Value}
expression matrix
\end{Value}
\inputencoding{utf8}
\HeaderA{select\_spatialNetwork}{select\_spatialNetwork}{select.Rul.spatialNetwork}
%
\begin{Description}\relax
function to select a spatial network
\end{Description}
%
\begin{Usage}
\begin{verbatim}
select_spatialNetwork(gobject, name = NULL, return_network_Obj = FALSE)
\end{verbatim}
\end{Usage}
\inputencoding{utf8}
\HeaderA{set\_giotto\_python\_path}{set\_giotto\_python\_path}{set.Rul.giotto.Rul.python.Rul.path}
%
\begin{Description}\relax
sets the python path and/or install miniconda and the python modules
\end{Description}
%
\begin{Usage}
\begin{verbatim}
set_giotto_python_path(
  python_path = NULL,
  packages_to_install = c("pandas", "networkx", "python-igraph", "leidenalg",
    "python-louvain")
)
\end{verbatim}
\end{Usage}
\inputencoding{utf8}
\HeaderA{show,giotto-method}{show method for giotto class}{show,giotto.Rdash.method}
\keyword{giotto,}{show,giotto-method}
\keyword{object}{show,giotto-method}
%
\begin{Description}\relax
show method for giotto class
\end{Description}
%
\begin{Usage}
\begin{verbatim}
## S4 method for signature 'giotto'
show(object)
\end{verbatim}
\end{Usage}
\inputencoding{utf8}
\HeaderA{showClusterDendrogram}{showClusterDendrogram}{showClusterDendrogram}
%
\begin{Description}\relax
Creates dendrogram for selected clusters.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
showClusterDendrogram(
  gobject,
  expression_values = c("normalized", "scaled", "custom"),
  cluster_column,
  cor = c("pearson", "spearman"),
  distance = "ward.D",
  h = NULL,
  h_color = "red",
  rotate = FALSE,
  show_plot = NA,
  return_plot = NA,
  save_plot = NA,
  save_param = list(),
  default_save_name = "showClusterDendrogram",
  ...
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{expression\_values}] expression values to use

\item[\code{cluster\_column}] name of column to use for clusters

\item[\code{cor}] correlation score to calculate distance

\item[\code{distance}] distance method to use for hierarchical clustering

\item[\code{h}] height of horizontal lines to plot

\item[\code{h\_color}] color of horizontal lines

\item[\code{rotate}] rotate dendrogram 90 degrees

\item[\code{show\_plot}] show plot

\item[\code{return\_plot}] return ggplot object

\item[\code{save\_plot}] directly save the plot [boolean]

\item[\code{save\_param}] list of saving parameters from \code{\LinkA{all\_plots\_save\_function}{all.Rul.plots.Rul.save.Rul.function}}

\item[\code{default\_save\_name}] default save name for saving, don't change, change save\_name in save\_param

\item[\code{...}] additional parameters for ggdendrogram()
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Expression correlation dendrogram for selected clusters.
\end{Details}
%
\begin{Value}
ggplot
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    showClusterDendrogram(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{showClusterHeatmap}{showClusterHeatmap}{showClusterHeatmap}
%
\begin{Description}\relax
Creates heatmap based on identified clusters
\end{Description}
%
\begin{Usage}
\begin{verbatim}
showClusterHeatmap(
  gobject,
  expression_values = c("normalized", "scaled", "custom"),
  genes = "all",
  cluster_column,
  cor = c("pearson", "spearman"),
  distance = "ward.D",
  show_plot = NA,
  return_plot = NA,
  save_plot = NA,
  save_param = list(),
  default_save_name = "showClusterHeatmap",
  ...
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{expression\_values}] expression values to use

\item[\code{genes}] vector of genes to use, default to 'all'

\item[\code{cluster\_column}] name of column to use for clusters

\item[\code{cor}] correlation score to calculate distance

\item[\code{distance}] distance method to use for hierarchical clustering

\item[\code{show\_plot}] show plot

\item[\code{return\_plot}] return ggplot object

\item[\code{save\_plot}] directly save the plot [boolean]

\item[\code{save\_param}] list of saving parameters from \code{\LinkA{all\_plots\_save\_function}{all.Rul.plots.Rul.save.Rul.function}}

\item[\code{default\_save\_name}] default save name for saving, don't change, change save\_name in save\_param

\item[\code{...}] additional parameters for the Heatmap function from ComplexHeatmap
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Correlation heatmap of selected clusters.
\end{Details}
%
\begin{Value}
ggplot
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    showClusterHeatmap(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{showGiottoInstructions}{showGiottoInstructions}{showGiottoInstructions}
%
\begin{Description}\relax
Function to display all instructions from giotto object
\end{Description}
%
\begin{Usage}
\begin{verbatim}
showGiottoInstructions(gobject)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object
\end{ldescription}
\end{Arguments}
%
\begin{Value}
named vector with giotto instructions
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    showGiottoInstructions()
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{showPattern}{showPattern}{showPattern}
%
\begin{Description}\relax
show patterns for 2D spatial data
\end{Description}
%
\begin{Usage}
\begin{verbatim}
showPattern(gobject, spatPatObj, ...)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{spatPatObj}] Output from detectSpatialPatterns

\item[\code{dimension}] dimension to plot

\item[\code{trim}] Trim ends of the PC values.

\item[\code{background\_color}] background color for plot

\item[\code{grid\_border\_color}] color for grid

\item[\code{show\_legend}] show legend of ggplot

\item[\code{show\_plot}] show plot

\item[\code{return\_plot}] return ggplot object

\item[\code{save\_plot}] directly save the plot [boolean]

\item[\code{save\_param}] list of saving parameters from \code{\LinkA{all\_plots\_save\_function}{all.Rul.plots.Rul.save.Rul.function}}

\item[\code{default\_save\_name}] default save name for saving, don't change, change save\_name in save\_param
\end{ldescription}
\end{Arguments}
%
\begin{Value}
ggplot
\end{Value}
%
\begin{SeeAlso}\relax
\code{\LinkA{showPattern2D}{showPattern2D}}
\end{SeeAlso}
%
\begin{Examples}
\begin{ExampleCode}
    showPattern(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{showPattern2D}{showPattern2D}{showPattern2D}
%
\begin{Description}\relax
show patterns for 2D spatial data
\end{Description}
%
\begin{Usage}
\begin{verbatim}
showPattern2D(
  gobject,
  spatPatObj,
  dimension = 1,
  trim = c(0.02, 0.98),
  background_color = "white",
  grid_border_color = "grey",
  show_legend = T,
  point_size = 1,
  show_plot = NA,
  return_plot = NA,
  save_plot = NA,
  save_param = list(),
  default_save_name = "showPattern2D"
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{spatPatObj}] Output from detectSpatialPatterns

\item[\code{dimension}] dimension to plot

\item[\code{trim}] Trim ends of the PC values.

\item[\code{background\_color}] background color for plot

\item[\code{grid\_border\_color}] color for grid

\item[\code{show\_legend}] show legend of ggplot

\item[\code{show\_plot}] show plot

\item[\code{return\_plot}] return ggplot object

\item[\code{save\_plot}] directly save the plot [boolean]

\item[\code{save\_param}] list of saving parameters from \code{\LinkA{all\_plots\_save\_function}{all.Rul.plots.Rul.save.Rul.function}}

\item[\code{default\_save\_name}] default save name for saving, don't change, change save\_name in save\_param
\end{ldescription}
\end{Arguments}
%
\begin{Value}
ggplot
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    showPattern2D(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{showPattern3D}{showPattern3D}{showPattern3D}
%
\begin{Description}\relax
show patterns for 3D spatial data
\end{Description}
%
\begin{Usage}
\begin{verbatim}
showPattern3D(
  gobject,
  spatPatObj,
  dimension = 1,
  trim = c(0.02, 0.98),
  background_color = "white",
  grid_border_color = "grey",
  show_legend = T,
  point_size = 1,
  axis_scale = c("cube", "real", "custom"),
  custom_ratio = NULL,
  x_ticks = NULL,
  y_ticks = NULL,
  z_ticks = NULL,
  show_plot = NA,
  return_plot = NA,
  save_plot = NA,
  save_param = list(),
  default_save_name = "showPattern3D"
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{spatPatObj}] Output from detectSpatialPatterns

\item[\code{dimension}] dimension to plot

\item[\code{trim}] Trim ends of the PC values.

\item[\code{background\_color}] background color for plot

\item[\code{grid\_border\_color}] color for grid

\item[\code{show\_legend}] show legend of plot

\item[\code{point\_size}] adjust the point size

\item[\code{axis\_scale}] scale the axis

\item[\code{custom\_ratio}] cutomize the scale of the axis

\item[\code{x\_ticks}] the tick number of x\_axis

\item[\code{y\_ticks}] the tick number of y\_axis

\item[\code{z\_ticks}] the tick number of z\_axis

\item[\code{show\_plot}] show plot

\item[\code{return\_plot}] return plot object

\item[\code{save\_plot}] directly save the plot [boolean]

\item[\code{save\_param}] list of saving parameters from \code{\LinkA{all\_plots\_save\_function}{all.Rul.plots.Rul.save.Rul.function}}

\item[\code{default\_save\_name}] default save name for saving, don't change, change save\_name in save\_param
\end{ldescription}
\end{Arguments}
%
\begin{Value}
plotly
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    showPattern3D(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{showPatternGenes}{showPatternGenes}{showPatternGenes}
%
\begin{Description}\relax
show genes correlated with spatial patterns
\end{Description}
%
\begin{Usage}
\begin{verbatim}
showPatternGenes(
  gobject,
  spatPatObj,
  dimension = 1,
  top_pos_genes = 5,
  top_neg_genes = 5,
  point_size = 1,
  return_DT = FALSE,
  show_plot = NA,
  return_plot = NA,
  save_plot = NA,
  save_param = list(),
  default_save_name = "showPatternGenes"
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{spatPatObj}] Output from detectSpatialPatterns

\item[\code{dimension}] dimension to plot genes for.

\item[\code{top\_pos\_genes}] Top positively correlated genes.

\item[\code{top\_neg\_genes}] Top negatively correlated genes.

\item[\code{point\_size}] size of points

\item[\code{return\_DT}] if TRUE, it will return the data.table used to generate the plots

\item[\code{show\_plot}] show plot

\item[\code{return\_plot}] return ggplot object

\item[\code{save\_plot}] directly save the plot [boolean]

\item[\code{save\_param}] list of saving parameters from all\_plots\_save\_function()

\item[\code{default\_save\_name}] default save name for saving, don't change, change save\_name in save\_param
\end{ldescription}
\end{Arguments}
%
\begin{Value}
ggplot
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    showPatternGenes(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{showProcessingSteps}{showProcessingSteps}{showProcessingSteps}
%
\begin{Description}\relax
shows the sequential processing steps that were performed in a summarized format
\end{Description}
%
\begin{Usage}
\begin{verbatim}
showProcessingSteps(gobject)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object
\end{ldescription}
\end{Arguments}
%
\begin{Value}
list of processing steps and names
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    showProcessingSteps(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{showSpatialCorGenes}{showSpatialCorGenes}{showSpatialCorGenes}
%
\begin{Description}\relax
Shows and filters spatially correlated genes
\end{Description}
%
\begin{Usage}
\begin{verbatim}
showSpatialCorGenes(
  spatCorObject,
  use_clus_name = NULL,
  selected_clusters = NULL,
  genes = NULL,
  min_spat_cor = 0.5,
  min_expr_cor = NULL,
  min_cor_diff = NULL,
  min_rank_diff = NULL,
  show_top_genes = NULL
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{spatCorObject}] spatial correlation object

\item[\code{use\_clus\_name}] cluster information to show

\item[\code{selected\_clusters}] subset of clusters to show

\item[\code{genes}] subset of genes to show

\item[\code{min\_spat\_cor}] filter on minimum spatial correlation

\item[\code{min\_expr\_cor}] filter on minimum single-cell expression correlation

\item[\code{min\_cor\_diff}] filter on minimum correlation difference (spatial vs expression)

\item[\code{min\_rank\_diff}] filter on minimum correlation rank difference (spatial vs expression)

\item[\code{show\_top\_genes}] show top genes per gene
\end{ldescription}
\end{Arguments}
%
\begin{Value}
data.table with filtered information
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    showSpatialCorGenes(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{signPCA}{signPCA}{signPCA}
%
\begin{Description}\relax
identify significant prinicipal components (PCs)
\end{Description}
%
\begin{Usage}
\begin{verbatim}
signPCA(
  gobject,
  method = c("screeplot", "jackstraw"),
  expression_values = c("normalized", "scaled", "custom"),
  reduction = c("cells", "genes"),
  genes_to_use = NULL,
  scale_unit = T,
  ncp = 50,
  scree_labels = T,
  scree_ylim = c(0, 10),
  jack_iter = 10,
  jack_threshold = 0.01,
  jack_verbose = T,
  show_plot = NA,
  return_plot = NA,
  save_plot = NA,
  save_param = list(),
  default_save_name = "signPCA",
  ...
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{method}] method to use to identify significant PCs

\item[\code{expression\_values}] expression values to use

\item[\code{reduction}] cells or genes

\item[\code{genes\_to\_use}] subset of genes to use for PCA

\item[\code{scale\_unit}] scale features before PCA

\item[\code{ncp}] number of principal components to calculate

\item[\code{scree\_labels}] show labels on scree plot

\item[\code{scree\_ylim}] y-axis limits on scree plot

\item[\code{jack\_iter}] number of interations for jackstraw

\item[\code{jack\_threshold}] p-value threshold to call a PC significant

\item[\code{jack\_verbose}] show progress of jackstraw method

\item[\code{show\_plot}] show plot

\item[\code{return\_plot}] return ggplot object

\item[\code{save\_plot}] directly save the plot [boolean]

\item[\code{save\_param}] list of saving parameters from all\_plots\_save\_function()

\item[\code{default\_save\_name}] default save name for saving, don't change, change save\_name in save\_param

\item[\code{...}] additional parameters for PCA
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Two different methods can be used to assess the number of relevant or significant
prinicipal components (PC's). \\{}
1. Screeplot works by plotting the explained variance of each
individual PC in a barplot allowing you to identify which PC does not show a significant
contribution anymore ( = 'elbow method'). \\{}
2. The Jackstraw method uses the \code{\LinkA{permutationPA}{permutationPA}} function. By
systematically permuting genes it identifies robust, and thus significant, PCs.
\\{} multiple PCA results can be stored by changing the \emph{name} parameter
\end{Details}
%
\begin{Value}
ggplot object for scree method and maxtrix of p-values for jackstraw
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    signPCA(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{silhouetteRank}{silhouetteRank}{silhouetteRank}
%
\begin{Description}\relax
Previously: calculate\_spatial\_genes\_python. This method computes a silhouette score per gene based on the
spatial distribution of two partitions of cells (expressed L1, and non-expressed L0).
Here, rather than L2 Euclidean norm, it uses a rank-transformed, exponentially weighted
function to represent the local physical distance between two cells.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
silhouetteRank(
  gobject,
  expression_values = c("normalized", "scaled", "custom"),
  metric = "euclidean",
  subset_genes = NULL,
  rbp_p = 0.95,
  examine_top = 0.3,
  python_path = NULL
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{expression\_values}] expression values to use

\item[\code{metric}] distance metric to use

\item[\code{subset\_genes}] only run on this subset of genes

\item[\code{rbp\_p}] fractional binarization threshold

\item[\code{examine\_top}] top fraction to evaluate with silhouette

\item[\code{python\_path}] specify specific path to python if required
\end{ldescription}
\end{Arguments}
%
\begin{Value}
data.table with spatial scores
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    silhouetteRank(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{sort\_combine\_two\_DT\_columns}{sort\_combine\_two\_DT\_columns}{sort.Rul.combine.Rul.two.Rul.DT.Rul.columns}
%
\begin{Description}\relax
fast sorting and pasting of 2 character columns
\end{Description}
%
\begin{Usage}
\begin{verbatim}
sort_combine_two_DT_columns(DT, column1, column2, myname = "unif_gene_gene")
\end{verbatim}
\end{Usage}
%
\begin{Examples}
\begin{ExampleCode}
    sort_combine_two_DT_columns()
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{spatCellCellcom}{spatCellCellcom}{spatCellCellcom}
%
\begin{Description}\relax
Spatial Cell-Cell communication scores based on spatial expression of interacting cells
\end{Description}
%
\begin{Usage}
\begin{verbatim}
spatCellCellcom(
  gobject,
  spatial_network_name = "Delaunay_network",
  cluster_column = "cell_types",
  random_iter = 1000,
  gene_set_1,
  gene_set_2,
  log2FC_addendum = 0.1,
  min_observations = 2,
  adjust_method = c("fdr", "bonferroni", "BH", "holm", "hochberg", "hommel", "BY",
    "none"),
  adjust_target = c("genes", "cells"),
  do_parallel = TRUE,
  cores = NA,
  verbose = c("a little", "a lot", "none")
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object to use

\item[\code{spatial\_network\_name}] spatial network to use for identifying interacting cells

\item[\code{cluster\_column}] cluster column with cell type information

\item[\code{random\_iter}] number of iterations

\item[\code{gene\_set\_1}] first specific gene set from gene pairs

\item[\code{gene\_set\_2}] second specific gene set from gene pairs

\item[\code{log2FC\_addendum}] addendum to add when calculating log2FC

\item[\code{min\_observations}] minimum number of interactions needed to be considered

\item[\code{adjust\_method}] which method to adjust p-values

\item[\code{adjust\_target}] adjust multiple hypotheses at the cell or gene level

\item[\code{do\_parallel}] run calculations in parallel with mclapply

\item[\code{cores}] number of cores to use if do\_parallel = TRUE

\item[\code{verbose}] verbose
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Statistical framework to identify if pairs of genes (such as ligand-receptor combinations)
are expressed at higher levels than expected based on a reshuffled null distribution
of gene expression values in cells that are spatially in proximity to eachother..
More details will follow soon.
\end{Details}
%
\begin{Value}
Cell-Cell communication scores for gene pairs based on spatial interaction
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    spatCellCellcom(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{spatCellPlot}{spatCellPlot}{spatCellPlot}
%
\begin{Description}\relax
Visualize cells according to spatial coordinates
\end{Description}
%
\begin{Usage}
\begin{verbatim}
spatCellPlot(
  gobject,
  sdimx = "sdimx",
  sdimy = "sdimy",
  spat_enr_names = NULL,
  cell_annotation_values = NULL,
  cell_color_gradient = c("blue", "white", "red"),
  gradient_midpoint = NULL,
  gradient_limits = NULL,
  select_cell_groups = NULL,
  select_cells = NULL,
  point_shape = c("border", "no_border", "voronoi"),
  point_size = 3,
  point_border_col = "black",
  point_border_stroke = 0.1,
  show_cluster_center = F,
  show_center_label = F,
  center_point_size = 4,
  center_point_border_col = "black",
  center_point_border_stroke = 0.1,
  label_size = 4,
  label_fontface = "bold",
  show_network = F,
  spatial_network_name = "Delaunay_network",
  network_color = NULL,
  network_alpha = 1,
  show_grid = F,
  spatial_grid_name = "spatial_grid",
  grid_color = NULL,
  show_other_cells = T,
  other_cell_color = "lightgrey",
  other_point_size = 1,
  other_cells_alpha = 0.1,
  coord_fix_ratio = NULL,
  show_legend = T,
  legend_text = 8,
  legend_symbol_size = 1,
  background_color = "white",
  vor_border_color = "white",
  vor_max_radius = 200,
  axis_text = 8,
  axis_title = 8,
  cow_n_col = 2,
  cow_rel_h = 1,
  cow_rel_w = 1,
  cow_align = "h",
  show_plot = NA,
  return_plot = NA,
  save_plot = NA,
  save_param = list(),
  default_save_name = "spatCellPlot"
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{sdimx}] x-axis dimension name (default = 'sdimx')

\item[\code{sdimy}] y-axis dimension name (default = 'sdimy')

\item[\code{spat\_enr\_names}] names of spatial enrichment results to include

\item[\code{cell\_annotation\_values}] numeric cell annotation columns

\item[\code{cell\_color\_gradient}] vector with 3 colors for numeric data

\item[\code{gradient\_midpoint}] midpoint for color gradient

\item[\code{gradient\_limits}] vector with lower and upper limits

\item[\code{select\_cell\_groups}] select subset of cells/clusters based on cell\_color parameter

\item[\code{select\_cells}] select subset of cells based on cell IDs

\item[\code{point\_shape}] shape of points (border, no\_border or voronoi)

\item[\code{point\_size}] size of point (cell)

\item[\code{point\_border\_col}] color of border around points

\item[\code{point\_border\_stroke}] stroke size of border around points

\item[\code{show\_cluster\_center}] plot center of selected clusters

\item[\code{show\_center\_label}] plot label of selected clusters

\item[\code{center\_point\_size}] size of center points

\item[\code{label\_size}] size of labels

\item[\code{label\_fontface}] font of labels

\item[\code{show\_network}] show underlying spatial network

\item[\code{spatial\_network\_name}] name of spatial network to use

\item[\code{network\_color}] color of spatial network

\item[\code{network\_alpha}] alpha of spatial network

\item[\code{show\_grid}] show spatial grid

\item[\code{spatial\_grid\_name}] name of spatial grid to use

\item[\code{grid\_color}] color of spatial grid

\item[\code{show\_other\_cells}] display not selected cells

\item[\code{other\_cell\_color}] color of not selected cells

\item[\code{other\_point\_size}] point size of not selected cells

\item[\code{other\_cells\_alpha}] alpha of not selected cells

\item[\code{coord\_fix\_ratio}] fix ratio between x and y-axis

\item[\code{show\_legend}] show legend

\item[\code{legend\_text}] size of legend text

\item[\code{legend\_symbol\_size}] size of legend symbols

\item[\code{background\_color}] color of plot background

\item[\code{vor\_border\_color}] border colorr for voronoi plot

\item[\code{vor\_max\_radius}] maximum radius for voronoi 'cells'

\item[\code{axis\_text}] size of axis text

\item[\code{axis\_title}] size of axis title

\item[\code{show\_plot}] show plot

\item[\code{return\_plot}] return ggplot object

\item[\code{save\_plot}] directly save the plot [boolean]

\item[\code{save\_param}] list of saving parameters from \code{\LinkA{all\_plots\_save\_function}{all.Rul.plots.Rul.save.Rul.function}}

\item[\code{default\_save\_name}] default save name for saving, don't change, change save\_name in save\_param
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Description of parameters.
\end{Details}
%
\begin{Value}
ggplot
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    spatCellPlot(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{spatCellPlot2D}{spatCellPlot2D}{spatCellPlot2D}
%
\begin{Description}\relax
Visualize cells according to spatial coordinates
\end{Description}
%
\begin{Usage}
\begin{verbatim}
spatCellPlot2D(
  gobject,
  sdimx = "sdimx",
  sdimy = "sdimy",
  spat_enr_names = NULL,
  cell_annotation_values = NULL,
  cell_color_gradient = c("blue", "white", "red"),
  gradient_midpoint = NULL,
  gradient_limits = NULL,
  select_cell_groups = NULL,
  select_cells = NULL,
  point_shape = c("border", "no_border", "voronoi"),
  point_size = 3,
  point_border_col = "black",
  point_border_stroke = 0.1,
  show_cluster_center = F,
  show_center_label = F,
  center_point_size = 4,
  center_point_border_col = "black",
  center_point_border_stroke = 0.1,
  label_size = 4,
  label_fontface = "bold",
  show_network = F,
  spatial_network_name = "Delaunay_network",
  network_color = NULL,
  network_alpha = 1,
  show_grid = F,
  spatial_grid_name = "spatial_grid",
  grid_color = NULL,
  show_other_cells = T,
  other_cell_color = "lightgrey",
  other_point_size = 1,
  other_cells_alpha = 0.1,
  coord_fix_ratio = NULL,
  show_legend = T,
  legend_text = 8,
  legend_symbol_size = 1,
  background_color = "white",
  vor_border_color = "white",
  vor_max_radius = 200,
  axis_text = 8,
  axis_title = 8,
  cow_n_col = 2,
  cow_rel_h = 1,
  cow_rel_w = 1,
  cow_align = "h",
  show_plot = NA,
  return_plot = NA,
  save_plot = NA,
  save_param = list(),
  default_save_name = "spatCellPlot2D"
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{sdimx}] x-axis dimension name (default = 'sdimx')

\item[\code{sdimy}] y-axis dimension name (default = 'sdimy')

\item[\code{spat\_enr\_names}] names of spatial enrichment results to include

\item[\code{cell\_annotation\_values}] numeric cell annotation columns

\item[\code{cell\_color\_gradient}] vector with 3 colors for numeric data

\item[\code{gradient\_midpoint}] midpoint for color gradient

\item[\code{gradient\_limits}] vector with lower and upper limits

\item[\code{select\_cell\_groups}] select subset of cells/clusters based on cell\_color parameter

\item[\code{select\_cells}] select subset of cells based on cell IDs

\item[\code{point\_shape}] shape of points (border, no\_border or voronoi)

\item[\code{point\_size}] size of point (cell)

\item[\code{point\_border\_col}] color of border around points

\item[\code{point\_border\_stroke}] stroke size of border around points

\item[\code{show\_cluster\_center}] plot center of selected clusters

\item[\code{show\_center\_label}] plot label of selected clusters

\item[\code{center\_point\_size}] size of center points

\item[\code{label\_size}] size of labels

\item[\code{label\_fontface}] font of labels

\item[\code{show\_network}] show underlying spatial network

\item[\code{spatial\_network\_name}] name of spatial network to use

\item[\code{network\_color}] color of spatial network

\item[\code{network\_alpha}] alpha of spatial network

\item[\code{show\_grid}] show spatial grid

\item[\code{spatial\_grid\_name}] name of spatial grid to use

\item[\code{grid\_color}] color of spatial grid

\item[\code{show\_other\_cells}] display not selected cells

\item[\code{other\_cell\_color}] color of not selected cells

\item[\code{other\_point\_size}] point size of not selected cells

\item[\code{other\_cells\_alpha}] alpha of not selected cells

\item[\code{coord\_fix\_ratio}] fix ratio between x and y-axis

\item[\code{show\_legend}] show legend

\item[\code{legend\_text}] size of legend text

\item[\code{legend\_symbol\_size}] size of legend symbols

\item[\code{background\_color}] color of plot background

\item[\code{vor\_border\_color}] border colorr for voronoi plot

\item[\code{vor\_max\_radius}] maximum radius for voronoi 'cells'

\item[\code{axis\_text}] size of axis text

\item[\code{axis\_title}] size of axis title

\item[\code{show\_plot}] show plot

\item[\code{return\_plot}] return ggplot object

\item[\code{save\_plot}] directly save the plot [boolean]

\item[\code{save\_param}] list of saving parameters from \code{\LinkA{all\_plots\_save\_function}{all.Rul.plots.Rul.save.Rul.function}}

\item[\code{default\_save\_name}] default save name for saving, don't change, change save\_name in save\_param
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Description of parameters.
\end{Details}
%
\begin{Value}
ggplot
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    spatCellPlot2D(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{spatDimCellPlot}{spatDimCellPlot}{spatDimCellPlot}
%
\begin{Description}\relax
Visualize numerical features of cells according to spatial AND dimension reduction coordinates in 2D
\end{Description}
%
\begin{Usage}
\begin{verbatim}
spatDimCellPlot(
  gobject,
  plot_alignment = c("vertical", "horizontal"),
  spat_enr_names = NULL,
  cell_annotation_values = NULL,
  dim_reduction_to_use = "umap",
  dim_reduction_name = "umap",
  dim1_to_use = 1,
  dim2_to_use = 2,
  sdimx = "sdimx",
  sdimy = "sdimy",
  cell_color_gradient = c("blue", "white", "red"),
  gradient_midpoint = NULL,
  gradient_limits = NULL,
  select_cell_groups = NULL,
  select_cells = NULL,
  dim_point_shape = c("border", "no_border"),
  dim_point_size = 1,
  dim_point_border_col = "black",
  dim_point_border_stroke = 0.1,
  spat_point_shape = c("border", "no_border", "voronoi"),
  spat_point_size = 1,
  spat_point_border_col = "black",
  spat_point_border_stroke = 0.1,
  dim_show_cluster_center = F,
  dim_show_center_label = T,
  dim_center_point_size = 4,
  dim_center_point_border_col = "black",
  dim_center_point_border_stroke = 0.1,
  dim_label_size = 4,
  dim_label_fontface = "bold",
  spat_show_cluster_center = F,
  spat_show_center_label = F,
  spat_center_point_size = 4,
  spat_center_point_border_col = "black",
  spat_center_point_border_stroke = 0.1,
  spat_label_size = 4,
  spat_label_fontface = "bold",
  show_NN_network = F,
  nn_network_to_use = "sNN",
  nn_network_name = "sNN.pca",
  dim_edge_alpha = 0.5,
  spat_show_network = F,
  spatial_network_name = "Delaunay_network",
  spat_network_color = "red",
  spat_network_alpha = 0.5,
  spat_show_grid = F,
  spatial_grid_name = "spatial_grid",
  spat_grid_color = "green",
  show_other_cells = TRUE,
  other_cell_color = "grey",
  dim_other_point_size = 0.5,
  spat_other_point_size = 0.5,
  spat_other_cells_alpha = 0.5,
  coord_fix_ratio = NULL,
  cow_n_col = 2,
  cow_rel_h = 1,
  cow_rel_w = 1,
  cow_align = "h",
  show_legend = T,
  legend_text = 8,
  legend_symbol_size = 1,
  dim_background_color = "white",
  spat_background_color = "white",
  vor_border_color = "white",
  vor_max_radius = 200,
  axis_text = 8,
  axis_title = 8,
  show_plot = NA,
  return_plot = NA,
  save_plot = NA,
  save_param = list(),
  default_save_name = "spatDimCellPlot"
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{plot\_alignment}] direction to align plot

\item[\code{spat\_enr\_names}] names of spatial enrichment results to include

\item[\code{cell\_annotation\_values}] numeric cell annotation columns

\item[\code{dim\_reduction\_to\_use}] dimension reduction to use

\item[\code{dim\_reduction\_name}] dimension reduction name

\item[\code{dim1\_to\_use}] dimension to use on x-axis

\item[\code{dim2\_to\_use}] dimension to use on y-axis

\item[\code{sdimx}] = spatial dimension to use on x-axis

\item[\code{sdimy}] = spatial dimension to use on y-axis

\item[\code{cell\_color\_gradient}] vector with 3 colors for numeric data

\item[\code{gradient\_midpoint}] midpoint for color gradient

\item[\code{gradient\_limits}] vector with lower and upper limits

\item[\code{select\_cell\_groups}] select subset of cells/clusters based on cell\_color parameter

\item[\code{select\_cells}] select subset of cells based on cell IDs

\item[\code{dim\_point\_shape}] spatial points with border or not (border or no\_border)

\item[\code{dim\_point\_size}] size of points in dim. reduction space

\item[\code{dim\_point\_border\_col}] border color of points in dim. reduction space

\item[\code{dim\_point\_border\_stroke}] border stroke of points in dim. reduction space

\item[\code{spat\_point\_shape}] shape of points (border, no\_border or voronoi)

\item[\code{spat\_point\_size}] size of spatial points

\item[\code{spat\_point\_border\_col}] border color of spatial points

\item[\code{spat\_point\_border\_stroke}] border stroke of spatial points

\item[\code{dim\_show\_cluster\_center}] show the center of each cluster

\item[\code{dim\_show\_center\_label}] provide a label for each cluster

\item[\code{dim\_center\_point\_size}] size of the center point

\item[\code{dim\_center\_point\_border\_col}] border color of center point

\item[\code{dim\_center\_point\_border\_stroke}] stroke size of center point

\item[\code{dim\_label\_size}] size of the center label

\item[\code{dim\_label\_fontface}] font of the center label

\item[\code{spat\_show\_cluster\_center}] show the center of each cluster

\item[\code{spat\_show\_center\_label}] provide a label for each cluster

\item[\code{spat\_center\_point\_size}] size of the center point

\item[\code{spat\_label\_size}] size of the center label

\item[\code{spat\_label\_fontface}] font of the center label

\item[\code{show\_NN\_network}] show underlying NN network

\item[\code{nn\_network\_to\_use}] type of NN network to use (kNN vs sNN)

\item[\code{nn\_network\_name}] name of NN network to use, if show\_NN\_network = TRUE

\item[\code{dim\_edge\_alpha}] column to use for alpha of the edges

\item[\code{spat\_show\_network}] show spatial network

\item[\code{spatial\_network\_name}] name of spatial network to use

\item[\code{spat\_network\_color}] color of spatial network

\item[\code{spat\_show\_grid}] show spatial grid

\item[\code{spatial\_grid\_name}] name of spatial grid to use

\item[\code{spat\_grid\_color}] color of spatial grid

\item[\code{show\_other\_cells}] display not selected cells

\item[\code{other\_cell\_color}] color of not selected cells

\item[\code{dim\_other\_point\_size}] size of not selected dim cells

\item[\code{spat\_other\_point\_size}] size of not selected spat cells

\item[\code{spat\_other\_cells\_alpha}] alpha of not selected spat cells

\item[\code{coord\_fix\_ratio}] ratio for coordinates

\item[\code{cow\_n\_col}] cowplot param: how many columns

\item[\code{cow\_rel\_h}] cowplot param: relative height

\item[\code{cow\_rel\_w}] cowplot param: relative width

\item[\code{cow\_align}] cowplot param: how to align

\item[\code{show\_legend}] show legend

\item[\code{legend\_text}] size of legend text

\item[\code{legend\_symbol\_size}] size of legend symbols

\item[\code{dim\_background\_color}] background color of points in dim. reduction space

\item[\code{spat\_background\_color}] background color of spatial points

\item[\code{vor\_border\_color}] border colorr for voronoi plot

\item[\code{vor\_max\_radius}] maximum radius for voronoi 'cells'

\item[\code{axis\_text}] size of axis text

\item[\code{axis\_title}] size of axis title

\item[\code{show\_plot}] show plot

\item[\code{return\_plot}] return ggplot object

\item[\code{save\_plot}] directly save the plot [boolean]

\item[\code{save\_param}] list of saving parameters from \code{\LinkA{all\_plots\_save\_function}{all.Rul.plots.Rul.save.Rul.function}}

\item[\code{default\_save\_name}] default save name for saving, don't change, change save\_name in save\_param
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Description of parameters.
\end{Details}
%
\begin{Value}
ggplot
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    spatDimCellPlot(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{spatDimCellPlot2D}{spatDimCellPlot2D}{spatDimCellPlot2D}
%
\begin{Description}\relax
Visualize numerical features of cells according to spatial AND dimension reduction coordinates in 2D
\end{Description}
%
\begin{Usage}
\begin{verbatim}
spatDimCellPlot2D(
  gobject,
  plot_alignment = c("vertical", "horizontal"),
  spat_enr_names = NULL,
  cell_annotation_values = NULL,
  dim_reduction_to_use = "umap",
  dim_reduction_name = "umap",
  dim1_to_use = 1,
  dim2_to_use = 2,
  sdimx = "sdimx",
  sdimy = "sdimy",
  cell_color_gradient = c("blue", "white", "red"),
  gradient_midpoint = NULL,
  gradient_limits = NULL,
  select_cell_groups = NULL,
  select_cells = NULL,
  dim_point_shape = c("border", "no_border"),
  dim_point_size = 1,
  dim_point_border_col = "black",
  dim_point_border_stroke = 0.1,
  spat_point_shape = c("border", "no_border", "voronoi"),
  spat_point_size = 1,
  spat_point_border_col = "black",
  spat_point_border_stroke = 0.1,
  dim_show_cluster_center = F,
  dim_show_center_label = T,
  dim_center_point_size = 4,
  dim_center_point_border_col = "black",
  dim_center_point_border_stroke = 0.1,
  dim_label_size = 4,
  dim_label_fontface = "bold",
  spat_show_cluster_center = F,
  spat_show_center_label = F,
  spat_center_point_size = 4,
  spat_center_point_border_col = "black",
  spat_center_point_border_stroke = 0.1,
  spat_label_size = 4,
  spat_label_fontface = "bold",
  show_NN_network = F,
  nn_network_to_use = "sNN",
  nn_network_name = "sNN.pca",
  dim_edge_alpha = 0.5,
  spat_show_network = F,
  spatial_network_name = "Delaunay_network",
  spat_network_color = "red",
  spat_network_alpha = 0.5,
  spat_show_grid = F,
  spatial_grid_name = "spatial_grid",
  spat_grid_color = "green",
  show_other_cells = TRUE,
  other_cell_color = "grey",
  dim_other_point_size = 0.5,
  spat_other_point_size = 0.5,
  spat_other_cells_alpha = 0.5,
  show_legend = T,
  legend_text = 8,
  legend_symbol_size = 1,
  dim_background_color = "white",
  spat_background_color = "white",
  vor_border_color = "white",
  vor_max_radius = 200,
  axis_text = 8,
  axis_title = 8,
  coord_fix_ratio = NULL,
  cow_n_col = 2,
  cow_rel_h = 1,
  cow_rel_w = 1,
  cow_align = "h",
  show_plot = NA,
  return_plot = NA,
  save_plot = NA,
  save_param = list(),
  default_save_name = "spatDimCellPlot2D"
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{plot\_alignment}] direction to align plot

\item[\code{spat\_enr\_names}] names of spatial enrichment results to include

\item[\code{cell\_annotation\_values}] numeric cell annotation columns

\item[\code{dim\_reduction\_to\_use}] dimension reduction to use

\item[\code{dim\_reduction\_name}] dimension reduction name

\item[\code{dim1\_to\_use}] dimension to use on x-axis

\item[\code{dim2\_to\_use}] dimension to use on y-axis

\item[\code{sdimx}] = spatial dimension to use on x-axis

\item[\code{sdimy}] = spatial dimension to use on y-axis

\item[\code{cell\_color\_gradient}] vector with 3 colors for numeric data

\item[\code{gradient\_midpoint}] midpoint for color gradient

\item[\code{gradient\_limits}] vector with lower and upper limits

\item[\code{select\_cell\_groups}] select subset of cells/clusters based on cell\_color parameter

\item[\code{select\_cells}] select subset of cells based on cell IDs

\item[\code{dim\_point\_shape}] dim reduction points with border or not (border or no\_border)

\item[\code{dim\_point\_size}] size of points in dim. reduction space

\item[\code{dim\_point\_border\_col}] border color of points in dim. reduction space

\item[\code{dim\_point\_border\_stroke}] border stroke of points in dim. reduction space

\item[\code{spat\_point\_shape}] shape of points (border, no\_border or voronoi)

\item[\code{spat\_point\_size}] size of spatial points

\item[\code{spat\_point\_border\_col}] border color of spatial points

\item[\code{spat\_point\_border\_stroke}] border stroke of spatial points

\item[\code{dim\_show\_cluster\_center}] show the center of each cluster

\item[\code{dim\_show\_center\_label}] provide a label for each cluster

\item[\code{dim\_center\_point\_size}] size of the center point

\item[\code{dim\_center\_point\_border\_col}] border color of center point

\item[\code{dim\_center\_point\_border\_stroke}] stroke size of center point

\item[\code{dim\_label\_size}] size of the center label

\item[\code{dim\_label\_fontface}] font of the center label

\item[\code{spat\_show\_cluster\_center}] show the center of each cluster

\item[\code{spat\_show\_center\_label}] provide a label for each cluster

\item[\code{spat\_center\_point\_size}] size of the center point

\item[\code{spat\_label\_size}] size of the center label

\item[\code{spat\_label\_fontface}] font of the center label

\item[\code{show\_NN\_network}] show underlying NN network

\item[\code{nn\_network\_to\_use}] type of NN network to use (kNN vs sNN)

\item[\code{nn\_network\_name}] name of NN network to use, if show\_NN\_network = TRUE

\item[\code{dim\_edge\_alpha}] column to use for alpha of the edges

\item[\code{spat\_show\_network}] show spatial network

\item[\code{spatial\_network\_name}] name of spatial network to use

\item[\code{spat\_network\_color}] color of spatial network

\item[\code{spat\_show\_grid}] show spatial grid

\item[\code{spatial\_grid\_name}] name of spatial grid to use

\item[\code{spat\_grid\_color}] color of spatial grid

\item[\code{show\_other\_cells}] display not selected cells

\item[\code{other\_cell\_color}] color of not selected cells

\item[\code{dim\_other\_point\_size}] size of not selected dim cells

\item[\code{spat\_other\_point\_size}] size of not selected spat cells

\item[\code{spat\_other\_cells\_alpha}] alpha of not selected spat cells

\item[\code{show\_legend}] show legend

\item[\code{legend\_text}] size of legend text

\item[\code{legend\_symbol\_size}] size of legend symbols

\item[\code{dim\_background\_color}] background color of points in dim. reduction space

\item[\code{spat\_background\_color}] background color of spatial points

\item[\code{vor\_border\_color}] border colorr for voronoi plot

\item[\code{vor\_max\_radius}] maximum radius for voronoi 'cells'

\item[\code{axis\_text}] size of axis text

\item[\code{axis\_title}] size of axis title

\item[\code{coord\_fix\_ratio}] ratio for coordinates

\item[\code{cow\_n\_col}] cowplot param: how many columns

\item[\code{cow\_rel\_h}] cowplot param: relative height

\item[\code{cow\_rel\_w}] cowplot param: relative width

\item[\code{cow\_align}] cowplot param: how to align

\item[\code{show\_plot}] show plot

\item[\code{return\_plot}] return ggplot object

\item[\code{save\_plot}] directly save the plot [boolean]

\item[\code{save\_param}] list of saving parameters from \code{\LinkA{all\_plots\_save\_function}{all.Rul.plots.Rul.save.Rul.function}}

\item[\code{default\_save\_name}] default save name for saving, don't change, change save\_name in save\_param
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Description of parameters.
\end{Details}
%
\begin{Value}
ggplot
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    spatDimCellPlot2D(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{spatDimGenePlot}{spatDimGenePlot}{spatDimGenePlot}
%
\begin{Description}\relax
Visualize cells according to spatial AND dimension reduction coordinates in ggplot mode
\end{Description}
%
\begin{Usage}
\begin{verbatim}
spatDimGenePlot(
  gobject,
  expression_values = c("normalized", "scaled", "custom"),
  plot_alignment = c("vertical", "horizontal"),
  genes,
  dim_reduction_to_use = "umap",
  dim_reduction_name = "umap",
  dim1_to_use = 1,
  dim2_to_use = 2,
  dim_point_shape = c("border", "no_border"),
  dim_point_size = 1,
  dim_point_border_col = "black",
  dim_point_border_stroke = 0.1,
  show_NN_network = F,
  show_spatial_network = F,
  show_spatial_grid = F,
  nn_network_to_use = "sNN",
  network_name = "sNN.pca",
  edge_alpha_dim = NULL,
  scale_alpha_with_expression = FALSE,
  sdimx = "sdimx",
  sdimy = "sdimy",
  spatial_network_name = "Delaunay_network",
  spatial_grid_name = "spatial_grid",
  spat_point_shape = c("border", "no_border", "voronoi"),
  spat_point_size = 1,
  spat_point_border_col = "black",
  spat_point_border_stroke = 0.1,
  cell_color_gradient = c("blue", "white", "red"),
  gradient_midpoint = NULL,
  gradient_limits = NULL,
  show_legend = T,
  legend_text = 8,
  dim_background_color = "white",
  spat_background_color = "white",
  vor_border_color = "white",
  vor_max_radius = 200,
  axis_text = 8,
  axis_title = 8,
  cow_n_col = 2,
  cow_rel_h = 1,
  cow_rel_w = 1,
  cow_align = "h",
  show_plot = NA,
  return_plot = NA,
  save_plot = NA,
  save_param = list(),
  default_save_name = "spatDimGenePlot"
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{expression\_values}] gene expression values to use

\item[\code{plot\_alignment}] direction to align plot

\item[\code{genes}] genes to show

\item[\code{dim\_reduction\_to\_use}] dimension reduction to use

\item[\code{dim\_reduction\_name}] dimension reduction name

\item[\code{dim1\_to\_use}] dimension to use on x-axis

\item[\code{dim2\_to\_use}] dimension to use on y-axis

\item[\code{dim\_point\_shape}] dim reduction points with border or not (border or no\_border)

\item[\code{dim\_point\_size}] dim reduction plot: point size

\item[\code{dim\_point\_border\_col}] color of border around points

\item[\code{dim\_point\_border\_stroke}] stroke size of border around points

\item[\code{show\_NN\_network}] show underlying NN network

\item[\code{nn\_network\_to\_use}] type of NN network to use (kNN vs sNN)

\item[\code{network\_name}] name of NN network to use, if show\_NN\_network = TRUE

\item[\code{edge\_alpha\_dim}] dim reduction plot: column to use for alpha of the edges

\item[\code{scale\_alpha\_with\_expression}] scale expression with ggplot alpha parameter

\item[\code{sdimx}] spatial x-axis dimension name (default = 'sdimx')

\item[\code{sdimy}] spatial y-axis dimension name (default = 'sdimy')

\item[\code{spatial\_network\_name}] name of spatial network to use

\item[\code{spatial\_grid\_name}] name of spatial grid to use

\item[\code{spat\_point\_shape}] spatial points with border or not (border or no\_border)

\item[\code{spat\_point\_size}] spatial plot: point size

\item[\code{spat\_point\_border\_col}] color of border around points

\item[\code{spat\_point\_border\_stroke}] stroke size of border around points

\item[\code{cell\_color\_gradient}] vector with 3 colors for numeric data

\item[\code{gradient\_midpoint}] midpoint for color gradient

\item[\code{gradient\_limits}] vector with lower and upper limits

\item[\code{show\_legend}] show legend

\item[\code{legend\_text}] size of legend text

\item[\code{dim\_background\_color}] color of plot background for dimension plot

\item[\code{spat\_background\_color}] color of plot background for spatial plot

\item[\code{vor\_border\_color}] border colorr for voronoi plot

\item[\code{vor\_max\_radius}] maximum radius for voronoi 'cells'

\item[\code{axis\_text}] size of axis text

\item[\code{axis\_title}] size of axis title

\item[\code{cow\_n\_col}] cowplot param: how many columns

\item[\code{cow\_rel\_h}] cowplot param: relative height

\item[\code{cow\_rel\_w}] cowplot param: relative width

\item[\code{cow\_align}] cowplot param: how to align

\item[\code{show\_plot}] show plots

\item[\code{return\_plot}] return ggplot object

\item[\code{save\_plot}] directly save the plot [boolean]

\item[\code{save\_param}] list of saving parameters from \code{\LinkA{all\_plots\_save\_function}{all.Rul.plots.Rul.save.Rul.function}}

\item[\code{default\_save\_name}] default save name for saving, don't change, change save\_name in save\_param
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Description of parameters.
\end{Details}
%
\begin{Value}
ggplot
\end{Value}
%
\begin{SeeAlso}\relax
\code{\LinkA{spatDimGenePlot3D}{spatDimGenePlot3D}}
\end{SeeAlso}
%
\begin{Examples}
\begin{ExampleCode}
    spatDimGenePlot(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{spatDimGenePlot2D}{spatDimGenePlot2D}{spatDimGenePlot2D}
%
\begin{Description}\relax
Visualize cells according to spatial AND dimension reduction coordinates in ggplot mode
\end{Description}
%
\begin{Usage}
\begin{verbatim}
spatDimGenePlot2D(
  gobject,
  expression_values = c("normalized", "scaled", "custom"),
  plot_alignment = c("vertical", "horizontal"),
  genes,
  dim_reduction_to_use = "umap",
  dim_reduction_name = "umap",
  dim1_to_use = 1,
  dim2_to_use = 2,
  dim_point_shape = c("border", "no_border"),
  dim_point_size = 1,
  dim_point_border_col = "black",
  dim_point_border_stroke = 0.1,
  show_NN_network = F,
  show_spatial_network = F,
  show_spatial_grid = F,
  nn_network_to_use = "sNN",
  network_name = "sNN.pca",
  edge_alpha_dim = NULL,
  scale_alpha_with_expression = FALSE,
  sdimx = "sdimx",
  sdimy = "sdimy",
  spatial_network_name = "Delaunay_network",
  spatial_grid_name = "spatial_grid",
  spat_point_shape = c("border", "no_border", "voronoi"),
  spat_point_size = 1,
  spat_point_border_col = "black",
  spat_point_border_stroke = 0.1,
  cell_color_gradient = c("blue", "white", "red"),
  gradient_midpoint = NULL,
  gradient_limits = NULL,
  cow_n_col = 2,
  cow_rel_h = 1,
  cow_rel_w = 1,
  cow_align = "h",
  show_legend = T,
  legend_text = 8,
  dim_background_color = "white",
  spat_background_color = "white",
  vor_border_color = "white",
  vor_max_radius = 200,
  axis_text = 8,
  axis_title = 8,
  show_plot = NA,
  return_plot = NA,
  save_plot = NA,
  save_param = list(),
  default_save_name = "spatDimGenePlot2D"
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{expression\_values}] gene expression values to use

\item[\code{plot\_alignment}] direction to align plot

\item[\code{genes}] genes to show

\item[\code{dim\_reduction\_to\_use}] dimension reduction to use

\item[\code{dim\_reduction\_name}] dimension reduction name

\item[\code{dim1\_to\_use}] dimension to use on x-axis

\item[\code{dim2\_to\_use}] dimension to use on y-axis

\item[\code{dim\_point\_shape}] dim reduction points with border or not (border or no\_border)

\item[\code{dim\_point\_size}] dim reduction plot: point size

\item[\code{dim\_point\_border\_col}] color of border around points

\item[\code{dim\_point\_border\_stroke}] stroke size of border around points

\item[\code{show\_NN\_network}] show underlying NN network

\item[\code{nn\_network\_to\_use}] type of NN network to use (kNN vs sNN)

\item[\code{network\_name}] name of NN network to use, if show\_NN\_network = TRUE

\item[\code{edge\_alpha\_dim}] dim reduction plot: column to use for alpha of the edges

\item[\code{scale\_alpha\_with\_expression}] scale expression with ggplot alpha parameter

\item[\code{sdimx}] spatial x-axis dimension name (default = 'sdimx')

\item[\code{sdimy}] spatial y-axis dimension name (default = 'sdimy')

\item[\code{spatial\_network\_name}] name of spatial network to use

\item[\code{spatial\_grid\_name}] name of spatial grid to use

\item[\code{spat\_point\_shape}] spatial points with border or not (border or no\_border)

\item[\code{spat\_point\_size}] spatial plot: point size

\item[\code{spat\_point\_border\_col}] color of border around points

\item[\code{spat\_point\_border\_stroke}] stroke size of border around points

\item[\code{cell\_color\_gradient}] vector with 3 colors for numeric data

\item[\code{gradient\_midpoint}] midpoint for color gradient

\item[\code{gradient\_limits}] vector with lower and upper limits

\item[\code{cow\_n\_col}] cowplot param: how many columns

\item[\code{cow\_rel\_h}] cowplot param: relative height

\item[\code{cow\_rel\_w}] cowplot param: relative width

\item[\code{cow\_align}] cowplot param: how to align

\item[\code{show\_legend}] show legend

\item[\code{legend\_text}] size of legend text

\item[\code{dim\_background\_color}] color of plot background for dimension plot

\item[\code{spat\_background\_color}] color of plot background for spatial plot

\item[\code{vor\_border\_color}] border colorr for voronoi plot

\item[\code{vor\_max\_radius}] maximum radius for voronoi 'cells'

\item[\code{axis\_text}] size of axis text

\item[\code{axis\_title}] size of axis title

\item[\code{show\_plot}] show plots

\item[\code{return\_plot}] return ggplot object

\item[\code{save\_plot}] directly save the plot [boolean]

\item[\code{save\_param}] list of saving parameters from \code{\LinkA{all\_plots\_save\_function}{all.Rul.plots.Rul.save.Rul.function}}

\item[\code{default\_save\_name}] default save name for saving, don't change, change save\_name in save\_param
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Description of parameters.
\end{Details}
%
\begin{Value}
ggplot
\end{Value}
%
\begin{SeeAlso}\relax
\code{\LinkA{spatDimGenePlot3D}{spatDimGenePlot3D}}
\end{SeeAlso}
%
\begin{Examples}
\begin{ExampleCode}
    spatDimGenePlot2D(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{spatDimGenePlot3D}{spatDimGenePlot3D}{spatDimGenePlot3D}
%
\begin{Description}\relax
Visualize cells according to spatial AND dimension reduction coordinates in ggplot mode
\end{Description}
%
\begin{Usage}
\begin{verbatim}
spatDimGenePlot3D(
  gobject,
  expression_values = c("normalized", "scaled", "custom"),
  plot_alignment = c("horizontal", "vertical"),
  dim_reduction_to_use = "umap",
  dim_reduction_name = "umap",
  dim1_to_use = 1,
  dim2_to_use = 2,
  dim3_to_use = NULL,
  sdimx = "sdimx",
  sdimy = "sdimy",
  sdimz = "sdimz",
  genes,
  cluster_column = NULL,
  select_cell_groups = NULL,
  select_cells = NULL,
  show_other_cells = T,
  other_cell_color = "lightgrey",
  other_point_size = 1.5,
  show_NN_network = F,
  nn_network_to_use = "sNN",
  network_name = "sNN.pca",
  label_size = 16,
  genes_low_color = "blue",
  genes_mid_color = "white",
  genes_high_color = "red",
  dim_point_size = 3,
  nn_network_alpha = 0.5,
  show_spatial_network = F,
  spatial_network_name = "Delaunay_network",
  network_color = "lightgray",
  spatial_network_alpha = 0.5,
  show_spatial_grid = F,
  spatial_grid_name = "spatial_grid",
  spatial_grid_color = NULL,
  spatial_grid_alpha = 0.5,
  spatial_point_size = 3,
  legend_text_size = 12,
  axis_scale = c("cube", "real", "custom"),
  custom_ratio = NULL,
  x_ticks = NULL,
  y_ticks = NULL,
  z_ticks = NULL,
  show_plot = NA,
  return_plot = NA,
  save_plot = NA,
  save_param = list(),
  default_save_name = "spatDimGenePlot3D"
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{expression\_values}] gene expression values to use

\item[\code{plot\_alignment}] direction to align plot

\item[\code{dim\_reduction\_to\_use}] dimension reduction to use

\item[\code{dim\_reduction\_name}] dimension reduction name

\item[\code{dim1\_to\_use}] dimension to use on x-axis

\item[\code{dim2\_to\_use}] dimension to use on y-axis

\item[\code{dim3\_to\_use}] dimension to use on z-axis

\item[\code{genes}] genes to show

\item[\code{show\_NN\_network}] show underlying NN network

\item[\code{nn\_network\_to\_use}] type of NN network to use (kNN vs sNN)

\item[\code{network\_name}] name of NN network to use, if show\_NN\_network = TRUE

\item[\code{dim\_point\_size}] dim reduction plot: point size

\item[\code{spatial\_network\_name}] name of spatial network to use

\item[\code{spatial\_grid\_name}] name of spatial grid to use

\item[\code{spatial\_point\_size}] spatial plot: point size

\item[\code{show\_plot}] show plots

\item[\code{return\_plot}] return plotly object

\item[\code{save\_plot}] directly save the plot [boolean]

\item[\code{save\_param}] list of saving parameters from \code{\LinkA{all\_plots\_save\_function}{all.Rul.plots.Rul.save.Rul.function}}

\item[\code{default\_save\_name}] default save name for saving, don't change, change save\_name in save\_param

\item[\code{edge\_alpha\_dim}] dim reduction plot: column to use for alpha of the edges

\item[\code{scale\_alpha\_with\_expression}] scale expression with ggplot alpha parameter

\item[\code{point\_size}] size of point (cell)

\item[\code{show\_legend}] show legend
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Description of parameters.
\end{Details}
%
\begin{Value}
plotly
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    spatDimGenePlot3D(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{spatDimPlot}{spatDimPlot}{spatDimPlot}
%
\begin{Description}\relax
Visualize cells according to spatial AND dimension reduction coordinates 2D
\end{Description}
%
\begin{Usage}
\begin{verbatim}
spatDimPlot(
  gobject,
  plot_alignment = c("vertical", "horizontal"),
  dim_reduction_to_use = "umap",
  dim_reduction_name = "umap",
  dim1_to_use = 1,
  dim2_to_use = 2,
  sdimx = "sdimx",
  sdimy = "sdimy",
  spat_enr_names = NULL,
  cell_color = NULL,
  color_as_factor = T,
  cell_color_code = NULL,
  cell_color_gradient = c("blue", "white", "red"),
  gradient_midpoint = NULL,
  gradient_limits = NULL,
  select_cell_groups = NULL,
  select_cells = NULL,
  dim_point_shape = c("border", "no_border"),
  dim_point_size = 1,
  dim_point_border_col = "black",
  dim_point_border_stroke = 0.1,
  spat_point_shape = c("border", "no_border", "voronoi"),
  spat_point_size = 1,
  spat_point_border_col = "black",
  spat_point_border_stroke = 0.1,
  dim_show_cluster_center = F,
  dim_show_center_label = T,
  dim_center_point_size = 4,
  dim_center_point_border_col = "black",
  dim_center_point_border_stroke = 0.1,
  dim_label_size = 4,
  dim_label_fontface = "bold",
  spat_show_cluster_center = F,
  spat_show_center_label = F,
  spat_center_point_size = 4,
  spat_label_size = 4,
  spat_label_fontface = "bold",
  show_NN_network = F,
  nn_network_to_use = "sNN",
  network_name = "sNN.pca",
  nn_network_alpha = 0.05,
  show_spatial_network = F,
  spat_network_name = "spatial_network",
  spat_network_color = "blue",
  spat_network_alpha = 0.5,
  show_spatial_grid = F,
  spat_grid_name = "spatial_grid",
  spat_grid_color = "blue",
  show_other_cells = T,
  other_cell_color = "lightgrey",
  dim_other_point_size = 1,
  spat_other_point_size = 1,
  spat_other_cells_alpha = 0.5,
  dim_show_legend = F,
  spat_show_legend = F,
  legend_text = 8,
  legend_symbol_size = 1,
  dim_background_color = "white",
  spat_background_color = "white",
  vor_border_color = "white",
  vor_max_radius = 200,
  axis_text = 8,
  axis_title = 8,
  show_plot = NA,
  return_plot = NA,
  save_plot = NA,
  save_param = list(),
  default_save_name = "spatDimPlot"
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{plot\_alignment}] direction to align plot

\item[\code{dim\_reduction\_to\_use}] dimension reduction to use

\item[\code{dim\_reduction\_name}] dimension reduction name

\item[\code{dim1\_to\_use}] dimension to use on x-axis

\item[\code{dim2\_to\_use}] dimension to use on y-axis

\item[\code{sdimx}] = spatial dimension to use on x-axis

\item[\code{sdimy}] = spatial dimension to use on y-axis

\item[\code{spat\_enr\_names}] names of spatial enrichment results to include

\item[\code{cell\_color}] color for cells (see details)

\item[\code{color\_as\_factor}] convert color column to factor

\item[\code{cell\_color\_code}] named vector with colors

\item[\code{cell\_color\_gradient}] vector with 3 colors for numeric data

\item[\code{gradient\_midpoint}] midpoint for color gradient

\item[\code{gradient\_limits}] vector with lower and upper limits

\item[\code{select\_cell\_groups}] select subset of cells/clusters based on cell\_color parameter

\item[\code{select\_cells}] select subset of cells based on cell IDs

\item[\code{dim\_point\_shape}] point with border or not (border or no\_border)

\item[\code{dim\_point\_size}] size of points in dim. reduction space

\item[\code{dim\_point\_border\_col}] border color of points in dim. reduction space

\item[\code{dim\_point\_border\_stroke}] border stroke of points in dim. reduction space

\item[\code{spat\_point\_shape}] shape of points (border, no\_border or voronoi)

\item[\code{spat\_point\_size}] size of spatial points

\item[\code{spat\_point\_border\_col}] border color of spatial points

\item[\code{spat\_point\_border\_stroke}] border stroke of spatial points

\item[\code{dim\_show\_cluster\_center}] show the center of each cluster

\item[\code{dim\_show\_center\_label}] provide a label for each cluster

\item[\code{dim\_center\_point\_size}] size of the center point

\item[\code{dim\_center\_point\_border\_col}] border color of center point

\item[\code{dim\_center\_point\_border\_stroke}] stroke size of center point

\item[\code{dim\_label\_size}] size of the center label

\item[\code{dim\_label\_fontface}] font of the center label

\item[\code{spat\_show\_cluster\_center}] show the center of each cluster

\item[\code{spat\_show\_center\_label}] provide a label for each cluster

\item[\code{spat\_center\_point\_size}] size of the center point

\item[\code{spat\_label\_size}] size of the center label

\item[\code{spat\_label\_fontface}] font of the center label

\item[\code{show\_NN\_network}] show underlying NN network

\item[\code{nn\_network\_to\_use}] type of NN network to use (kNN vs sNN)

\item[\code{network\_name}] name of NN network to use, if show\_NN\_network = TRUE

\item[\code{nn\_network\_alpha}] column to use for alpha of the edges

\item[\code{show\_spatial\_network}] show spatial network

\item[\code{spat\_network\_name}] name of spatial network to use

\item[\code{spat\_network\_color}] color of spatial network

\item[\code{show\_spatial\_grid}] show spatial grid

\item[\code{spat\_grid\_name}] name of spatial grid to use

\item[\code{spat\_grid\_color}] color of spatial grid

\item[\code{show\_other\_cells}] display not selected cells

\item[\code{other\_cell\_color}] color of not selected cells

\item[\code{dim\_other\_point\_size}] size of not selected dim cells

\item[\code{spat\_other\_point\_size}] size of not selected spat cells

\item[\code{spat\_other\_cells\_alpha}] alpha of not selected spat cells

\item[\code{dim\_show\_legend}] show legend of dimension reduction plot

\item[\code{spat\_show\_legend}] show legend of spatial plot

\item[\code{legend\_text}] size of legend text

\item[\code{legend\_symbol\_size}] size of legend symbols

\item[\code{dim\_background\_color}] background color of points in dim. reduction space

\item[\code{spat\_background\_color}] background color of spatial points

\item[\code{vor\_border\_color}] border colorr for voronoi plot

\item[\code{vor\_max\_radius}] maximum radius for voronoi 'cells'

\item[\code{axis\_text}] size of axis text

\item[\code{axis\_title}] size of axis title

\item[\code{show\_plot}] show plot

\item[\code{return\_plot}] return ggplot object

\item[\code{save\_plot}] directly save the plot [boolean]

\item[\code{save\_param}] list of saving parameters from \code{\LinkA{all\_plots\_save\_function}{all.Rul.plots.Rul.save.Rul.function}}

\item[\code{default\_save\_name}] default save name for saving, don't change, change save\_name in save\_param
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Description of parameters.
\end{Details}
%
\begin{Value}
ggplot
\end{Value}
%
\begin{SeeAlso}\relax
\code{\LinkA{spatDimPlot2D}{spatDimPlot2D}} and \code{\LinkA{spatDimPlot3D}{spatDimPlot3D}} for 3D visualization.
\end{SeeAlso}
%
\begin{Examples}
\begin{ExampleCode}
    spatDimPlot(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{spatDimPlot2D}{spatDimPlot2D}{spatDimPlot2D}
%
\begin{Description}\relax
Visualize cells according to spatial AND dimension reduction coordinates 2D
\end{Description}
%
\begin{Usage}
\begin{verbatim}
spatDimPlot2D(
  gobject,
  plot_alignment = c("vertical", "horizontal"),
  dim_reduction_to_use = "umap",
  dim_reduction_name = "umap",
  dim1_to_use = 1,
  dim2_to_use = 2,
  sdimx = "sdimx",
  sdimy = "sdimy",
  spat_enr_names = NULL,
  cell_color = NULL,
  color_as_factor = T,
  cell_color_code = NULL,
  cell_color_gradient = c("blue", "white", "red"),
  gradient_midpoint = NULL,
  gradient_limits = NULL,
  select_cell_groups = NULL,
  select_cells = NULL,
  dim_point_shape = c("border", "no_border"),
  dim_point_size = 1,
  dim_point_border_col = "black",
  dim_point_border_stroke = 0.1,
  spat_point_shape = c("border", "no_border", "voronoi"),
  spat_point_size = 1,
  spat_point_border_col = "black",
  spat_point_border_stroke = 0.1,
  dim_show_cluster_center = F,
  dim_show_center_label = T,
  dim_center_point_size = 4,
  dim_center_point_border_col = "black",
  dim_center_point_border_stroke = 0.1,
  dim_label_size = 4,
  dim_label_fontface = "bold",
  spat_show_cluster_center = F,
  spat_show_center_label = F,
  spat_center_point_size = 4,
  spat_label_size = 4,
  spat_label_fontface = "bold",
  show_NN_network = F,
  nn_network_to_use = "sNN",
  network_name = "sNN.pca",
  nn_network_alpha = 0.05,
  show_spatial_network = F,
  spat_network_name = "spatial_network",
  spat_network_color = "blue",
  spat_network_alpha = 0.5,
  show_spatial_grid = F,
  spat_grid_name = "spatial_grid",
  spat_grid_color = "blue",
  show_other_cells = T,
  other_cell_color = "lightgrey",
  dim_other_point_size = 1,
  spat_other_point_size = 1,
  spat_other_cells_alpha = 0.5,
  dim_show_legend = F,
  spat_show_legend = F,
  legend_text = 8,
  legend_symbol_size = 1,
  dim_background_color = "white",
  spat_background_color = "white",
  vor_border_color = "white",
  vor_max_radius = 200,
  axis_text = 8,
  axis_title = 8,
  show_plot = NA,
  return_plot = NA,
  save_plot = NA,
  save_param = list(),
  default_save_name = "spatDimPlot2D"
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{plot\_alignment}] direction to align plot

\item[\code{dim\_reduction\_to\_use}] dimension reduction to use

\item[\code{dim\_reduction\_name}] dimension reduction name

\item[\code{dim1\_to\_use}] dimension to use on x-axis

\item[\code{dim2\_to\_use}] dimension to use on y-axis

\item[\code{sdimx}] = spatial dimension to use on x-axis

\item[\code{sdimy}] = spatial dimension to use on y-axis

\item[\code{spat\_enr\_names}] names of spatial enrichment results to include

\item[\code{cell\_color}] color for cells (see details)

\item[\code{color\_as\_factor}] convert color column to factor

\item[\code{cell\_color\_code}] named vector with colors

\item[\code{cell\_color\_gradient}] vector with 3 colors for numeric data

\item[\code{gradient\_midpoint}] midpoint for color gradient

\item[\code{gradient\_limits}] vector with lower and upper limits

\item[\code{select\_cell\_groups}] select subset of cells/clusters based on cell\_color parameter

\item[\code{select\_cells}] select subset of cells based on cell IDs

\item[\code{dim\_point\_shape}] point with border or not (border or no\_border)

\item[\code{dim\_point\_size}] size of points in dim. reduction space

\item[\code{dim\_point\_border\_col}] border color of points in dim. reduction space

\item[\code{dim\_point\_border\_stroke}] border stroke of points in dim. reduction space

\item[\code{spat\_point\_shape}] shape of points (border, no\_border or voronoi)

\item[\code{spat\_point\_size}] size of spatial points

\item[\code{spat\_point\_border\_col}] border color of spatial points

\item[\code{spat\_point\_border\_stroke}] border stroke of spatial points

\item[\code{dim\_show\_cluster\_center}] show the center of each cluster

\item[\code{dim\_show\_center\_label}] provide a label for each cluster

\item[\code{dim\_center\_point\_size}] size of the center point

\item[\code{dim\_center\_point\_border\_col}] border color of center point

\item[\code{dim\_center\_point\_border\_stroke}] stroke size of center point

\item[\code{dim\_label\_size}] size of the center label

\item[\code{dim\_label\_fontface}] font of the center label

\item[\code{spat\_show\_cluster\_center}] show the center of each cluster

\item[\code{spat\_show\_center\_label}] provide a label for each cluster

\item[\code{spat\_center\_point\_size}] size of the center point

\item[\code{spat\_label\_size}] size of the center label

\item[\code{spat\_label\_fontface}] font of the center label

\item[\code{show\_NN\_network}] show underlying NN network

\item[\code{nn\_network\_to\_use}] type of NN network to use (kNN vs sNN)

\item[\code{network\_name}] name of NN network to use, if show\_NN\_network = TRUE

\item[\code{nn\_network\_alpha}] column to use for alpha of the edges

\item[\code{show\_spatial\_network}] show spatial network

\item[\code{spat\_network\_name}] name of spatial network to use

\item[\code{spat\_network\_color}] color of spatial network

\item[\code{show\_spatial\_grid}] show spatial grid

\item[\code{spat\_grid\_name}] name of spatial grid to use

\item[\code{spat\_grid\_color}] color of spatial grid

\item[\code{show\_other\_cells}] display not selected cells

\item[\code{other\_cell\_color}] color of not selected cells

\item[\code{dim\_other\_point\_size}] size of not selected dim cells

\item[\code{spat\_other\_point\_size}] size of not selected spat cells

\item[\code{spat\_other\_cells\_alpha}] alpha of not selected spat cells

\item[\code{dim\_show\_legend}] show legend of dimension reduction plot

\item[\code{spat\_show\_legend}] show legend of spatial plot

\item[\code{legend\_text}] size of legend text

\item[\code{legend\_symbol\_size}] size of legend symbols

\item[\code{dim\_background\_color}] background color of points in dim. reduction space

\item[\code{spat\_background\_color}] background color of spatial points

\item[\code{vor\_border\_color}] border colorr for voronoi plot

\item[\code{vor\_max\_radius}] maximum radius for voronoi 'cells'

\item[\code{axis\_text}] size of axis text

\item[\code{axis\_title}] size of axis title

\item[\code{show\_plot}] show plot

\item[\code{return\_plot}] return ggplot object

\item[\code{save\_plot}] directly save the plot [boolean]

\item[\code{save\_param}] list of saving parameters from \code{\LinkA{all\_plots\_save\_function}{all.Rul.plots.Rul.save.Rul.function}}

\item[\code{default\_save\_name}] default save name for saving, don't change, change save\_name in save\_param
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Description of parameters.
\end{Details}
%
\begin{Value}
ggplot
\end{Value}
%
\begin{SeeAlso}\relax
\code{\LinkA{spatDimPlot3D}{spatDimPlot3D}}
\end{SeeAlso}
%
\begin{Examples}
\begin{ExampleCode}
    spatDimPlot2D(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{spatDimPlot3D}{spatDimPlot3D}{spatDimPlot3D}
%
\begin{Description}\relax
Visualize cells according to spatial AND dimension reduction coordinates in plotly mode
\end{Description}
%
\begin{Usage}
\begin{verbatim}
spatDimPlot3D(
  gobject,
  plot_alignment = c("horizontal", "vertical"),
  dim_reduction_to_use = "umap",
  dim_reduction_name = "umap",
  dim1_to_use = 1,
  dim2_to_use = 2,
  dim3_to_use = 3,
  sdimx = "sdimx",
  sdimy = "sdimy",
  sdimz = "sdimz",
  show_NN_network = F,
  nn_network_to_use = "sNN",
  network_name = "sNN.pca",
  show_cluster_center = F,
  show_center_label = T,
  center_point_size = 4,
  label_size = 16,
  select_cell_groups = NULL,
  select_cells = NULL,
  show_other_cells = T,
  other_cell_color = "lightgrey",
  other_point_size = 1.5,
  cell_color = NULL,
  color_as_factor = T,
  cell_color_code = NULL,
  dim_point_size = 3,
  nn_network_alpha = 0.5,
  show_spatial_network = F,
  spatial_network_name = "Delaunay_network",
  network_color = "lightgray",
  spatial_network_alpha = 0.5,
  show_spatial_grid = F,
  spatial_grid_name = "spatial_grid",
  spatial_grid_color = NULL,
  spatial_grid_alpha = 0.5,
  spatial_point_size = 3,
  axis_scale = c("cube", "real", "custom"),
  custom_ratio = NULL,
  x_ticks = NULL,
  y_ticks = NULL,
  z_ticks = NULL,
  legend_text_size = 12,
  show_plot = NA,
  return_plot = NA,
  save_plot = NA,
  save_param = list(),
  default_save_name = "spatDimPlot3D"
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{plot\_alignment}] direction to align plot

\item[\code{dim\_reduction\_to\_use}] dimension reduction to use

\item[\code{dim\_reduction\_name}] dimension reduction name

\item[\code{dim1\_to\_use}] dimension to use on x-axis

\item[\code{dim2\_to\_use}] dimension to use on y-axis

\item[\code{dim3\_to\_use}] dimension to use on z-axis

\item[\code{sdimx}] = spatial dimension to use on x-axis

\item[\code{sdimy}] = spatial dimension to use on y-axis

\item[\code{sdimz}] = spatial dimension to use on z-axis

\item[\code{show\_NN\_network}] show underlying NN network

\item[\code{nn\_network\_to\_use}] type of NN network to use (kNN vs sNN)

\item[\code{network\_name}] name of NN network to use, if show\_NN\_network = TRUE

\item[\code{show\_cluster\_center}] show the center of each cluster

\item[\code{show\_center\_label}] provide a label for each cluster

\item[\code{center\_point\_size}] size of the center point

\item[\code{label\_size}] size of the center label

\item[\code{select\_cell\_groups}] select subset of cells/clusters based on cell\_color parameter

\item[\code{select\_cells}] select subset of cells based on cell IDs

\item[\code{show\_other\_cells}] display not selected cells

\item[\code{other\_cell\_color}] color of not selected cells

\item[\code{other\_point\_size}] size of not selected cells

\item[\code{cell\_color}] color for cells (see details)

\item[\code{color\_as\_factor}] convert color column to factor

\item[\code{cell\_color\_code}] named vector with colors

\item[\code{dim\_point\_size}] size of points in dim. reduction space

\item[\code{nn\_network\_alpha}] column to use for alpha of the edges

\item[\code{show\_spatial\_network}] show spatial network

\item[\code{spatial\_network\_name}] name of spatial network to use

\item[\code{spatial\_network\_alpha}] alpha of spatial network

\item[\code{show\_spatial\_grid}] show spatial grid

\item[\code{spatial\_grid\_name}] name of spatial grid to use

\item[\code{spatial\_grid\_color}] color of spatial grid

\item[\code{spatial\_point\_size}] size of spatial points

\item[\code{show\_plot}] show plot

\item[\code{return\_plot}] return ggplot object

\item[\code{save\_plot}] directly save the plot [boolean]

\item[\code{save\_param}] list of saving parameters from \code{\LinkA{all\_plots\_save\_function}{all.Rul.plots.Rul.save.Rul.function}}

\item[\code{default\_save\_name}] default save name for saving, don't change, change save\_name in save\_param

\item[\code{dim\_point\_border\_col}] border color of points in dim. reduction space

\item[\code{dim\_point\_border\_stroke}] border stroke of points in dim. reduction space

\item[\code{spatial\_network\_color}] color of spatial network

\item[\code{spatial\_other\_point\_size}] size of not selected spatial points

\item[\code{spatial\_other\_cells\_alpha}] alpha of not selected spatial points

\item[\code{dim\_other\_point\_size}] size of not selected dim. reduction points

\item[\code{show\_legend}] show legend
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Description of parameters.
\end{Details}
%
\begin{Value}
plotly
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    spatDimPlot3D(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{spatGenePlot}{spatGenePlot}{spatGenePlot}
%
\begin{Description}\relax
Visualize cells and gene expression according to spatial coordinates
\end{Description}
%
\begin{Usage}
\begin{verbatim}
spatGenePlot(
  gobject,
  sdimx = "sdimx",
  sdimy = "sdimy",
  expression_values = c("normalized", "scaled", "custom"),
  genes,
  cell_color_gradient = c("blue", "white", "red"),
  gradient_midpoint = NULL,
  gradient_limits = NULL,
  show_network = F,
  network_color = NULL,
  spatial_network_name = "Delaunay_network",
  edge_alpha = NULL,
  show_grid = F,
  grid_color = NULL,
  spatial_grid_name = "spatial_grid",
  midpoint = 0,
  scale_alpha_with_expression = FALSE,
  point_shape = c("border", "no_border", "voronoi"),
  point_size = 1,
  point_border_col = "black",
  point_border_stroke = 0.1,
  show_legend = T,
  legend_text = 8,
  background_color = "white",
  vor_border_color = "white",
  vor_max_radius = 200,
  axis_text = 8,
  axis_title = 8,
  cow_n_col = 2,
  cow_rel_h = 1,
  cow_rel_w = 1,
  cow_align = "h",
  show_plot = NA,
  return_plot = NA,
  save_plot = NA,
  save_param = list(),
  default_save_name = "spatGenePlot"
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{sdimx}] x-axis dimension name (default = 'sdimx')

\item[\code{sdimy}] y-axis dimension name (default = 'sdimy')

\item[\code{expression\_values}] gene expression values to use

\item[\code{genes}] genes to show

\item[\code{cell\_color\_gradient}] vector with 3 colors for numeric data

\item[\code{gradient\_midpoint}] midpoint for color gradient

\item[\code{gradient\_limits}] vector with lower and upper limits

\item[\code{show\_network}] show underlying spatial network

\item[\code{network\_color}] color of spatial network

\item[\code{spatial\_network\_name}] name of spatial network to use

\item[\code{show\_grid}] show spatial grid

\item[\code{grid\_color}] color of spatial grid

\item[\code{spatial\_grid\_name}] name of spatial grid to use

\item[\code{midpoint}] expression midpoint

\item[\code{scale\_alpha\_with\_expression}] scale expression with ggplot alpha parameter

\item[\code{point\_shape}] shape of points (border, no\_border or voronoi)

\item[\code{point\_size}] size of point (cell)

\item[\code{point\_border\_col}] color of border around points

\item[\code{point\_border\_stroke}] stroke size of border around points

\item[\code{show\_legend}] show legend

\item[\code{legend\_text}] size of legend text

\item[\code{background\_color}] color of plot background

\item[\code{vor\_border\_color}] border colorr for voronoi plot

\item[\code{vor\_max\_radius}] maximum radius for voronoi 'cells'

\item[\code{axis\_text}] size of axis text

\item[\code{axis\_title}] size of axis title

\item[\code{cow\_n\_col}] cowplot param: how many columns

\item[\code{cow\_rel\_h}] cowplot param: relative height

\item[\code{cow\_rel\_w}] cowplot param: relative width

\item[\code{cow\_align}] cowplot param: how to align

\item[\code{show\_plot}] show plots

\item[\code{return\_plot}] return ggplot object

\item[\code{save\_plot}] directly save the plot [boolean]

\item[\code{save\_param}] list of saving parameters from \code{\LinkA{all\_plots\_save\_function}{all.Rul.plots.Rul.save.Rul.function}}

\item[\code{default\_save\_name}] default save name for saving, don't change, change save\_name in save\_param

\item[\code{...}] parameters for cowplot::save\_plot()
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Description of parameters.
\end{Details}
%
\begin{Value}
ggplot
\end{Value}
%
\begin{SeeAlso}\relax
\code{\LinkA{spatGenePlot3D}{spatGenePlot3D}} and \code{\LinkA{spatGenePlot2D}{spatGenePlot2D}}
\end{SeeAlso}
%
\begin{Examples}
\begin{ExampleCode}
    spatGenePlot(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{spatGenePlot2D}{spatGenePlot2D}{spatGenePlot2D}
%
\begin{Description}\relax
Visualize cells and gene expression according to spatial coordinates
\end{Description}
%
\begin{Usage}
\begin{verbatim}
spatGenePlot2D(
  gobject,
  sdimx = "sdimx",
  sdimy = "sdimy",
  expression_values = c("normalized", "scaled", "custom"),
  genes,
  cell_color_gradient = c("blue", "white", "red"),
  gradient_midpoint = NULL,
  gradient_limits = NULL,
  show_network = F,
  network_color = NULL,
  spatial_network_name = "Delaunay_network",
  edge_alpha = NULL,
  show_grid = F,
  grid_color = NULL,
  spatial_grid_name = "spatial_grid",
  midpoint = 0,
  scale_alpha_with_expression = FALSE,
  point_shape = c("border", "no_border", "voronoi"),
  point_size = 1,
  point_border_col = "black",
  point_border_stroke = 0.1,
  show_legend = T,
  legend_text = 8,
  background_color = "white",
  vor_border_color = "white",
  vor_max_radius = 200,
  axis_text = 8,
  axis_title = 8,
  cow_n_col = 2,
  cow_rel_h = 1,
  cow_rel_w = 1,
  cow_align = "h",
  show_plot = NA,
  return_plot = NA,
  save_plot = NA,
  save_param = list(),
  default_save_name = "spatGenePlot2D"
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{sdimx}] x-axis dimension name (default = 'sdimx')

\item[\code{sdimy}] y-axis dimension name (default = 'sdimy')

\item[\code{expression\_values}] gene expression values to use

\item[\code{genes}] genes to show

\item[\code{cell\_color\_gradient}] vector with 3 colors for numeric data

\item[\code{gradient\_midpoint}] midpoint for color gradient

\item[\code{gradient\_limits}] vector with lower and upper limits

\item[\code{show\_network}] show underlying spatial network

\item[\code{network\_color}] color of spatial network

\item[\code{spatial\_network\_name}] name of spatial network to use

\item[\code{show\_grid}] show spatial grid

\item[\code{grid\_color}] color of spatial grid

\item[\code{spatial\_grid\_name}] name of spatial grid to use

\item[\code{midpoint}] expression midpoint

\item[\code{scale\_alpha\_with\_expression}] scale expression with ggplot alpha parameter

\item[\code{point\_shape}] shape of points (border, no\_border or voronoi)

\item[\code{point\_size}] size of point (cell)

\item[\code{point\_border\_col}] color of border around points

\item[\code{point\_border\_stroke}] stroke size of border around points

\item[\code{show\_legend}] show legend

\item[\code{legend\_text}] size of legend text

\item[\code{background\_color}] color of plot background

\item[\code{vor\_border\_color}] border colorr for voronoi plot

\item[\code{vor\_max\_radius}] maximum radius for voronoi 'cells'

\item[\code{axis\_text}] size of axis text

\item[\code{axis\_title}] size of axis title

\item[\code{cow\_n\_col}] cowplot param: how many columns

\item[\code{cow\_rel\_h}] cowplot param: relative height

\item[\code{cow\_rel\_w}] cowplot param: relative width

\item[\code{cow\_align}] cowplot param: how to align

\item[\code{show\_plot}] show plots

\item[\code{return\_plot}] return ggplot object

\item[\code{save\_plot}] directly save the plot [boolean]

\item[\code{save\_param}] list of saving parameters from \code{\LinkA{all\_plots\_save\_function}{all.Rul.plots.Rul.save.Rul.function}}

\item[\code{default\_save\_name}] default save name for saving, don't change, change save\_name in save\_param

\item[\code{...}] parameters for cowplot::save\_plot()
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Description of parameters.
\end{Details}
%
\begin{Value}
ggplot
\end{Value}
%
\begin{SeeAlso}\relax
\code{\LinkA{spatGenePlot3D}{spatGenePlot3D}}
\end{SeeAlso}
%
\begin{Examples}
\begin{ExampleCode}
    spatGenePlot2D(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{spatGenePlot3D}{spatGenePlot3D}{spatGenePlot3D}
%
\begin{Description}\relax
Visualize cells and gene expression according to spatial coordinates
\end{Description}
%
\begin{Usage}
\begin{verbatim}
spatGenePlot3D(
  gobject,
  expression_values = c("normalized", "scaled", "custom"),
  genes,
  show_network = F,
  network_color = NULL,
  spatial_network_name = "Delaunay_network",
  edge_alpha = NULL,
  show_grid = F,
  cluster_column = NULL,
  select_cell_groups = NULL,
  select_cells = NULL,
  show_other_cells = T,
  other_cell_color = "lightgrey",
  other_point_size = 1,
  genes_high_color = NULL,
  genes_mid_color = "white",
  genes_low_color = "blue",
  spatial_grid_name = "spatial_grid",
  point_size = 2,
  show_legend = T,
  axis_scale = c("cube", "real", "custom"),
  custom_ratio = NULL,
  x_ticks = NULL,
  y_ticks = NULL,
  z_ticks = NULL,
  show_plot = NA,
  return_plot = NA,
  save_plot = NA,
  save_param = list(),
  default_save_name = "spatGenePlot3D"
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{expression\_values}] gene expression values to use

\item[\code{genes}] genes to show

\item[\code{show\_network}] show underlying spatial network

\item[\code{network\_color}] color of spatial network

\item[\code{spatial\_network\_name}] name of spatial network to use

\item[\code{show\_grid}] show spatial grid

\item[\code{genes\_high\_color}] color represents high gene expression

\item[\code{genes\_mid\_color}] color represents middle gene expression

\item[\code{genes\_low\_color}] color represents low gene expression

\item[\code{spatial\_grid\_name}] name of spatial grid to use

\item[\code{point\_size}] size of point (cell)

\item[\code{show\_legend}] show legend

\item[\code{show\_plot}] show plots

\item[\code{return\_plot}] return ggplot object

\item[\code{save\_plot}] directly save the plot [boolean]

\item[\code{save\_param}] list of saving parameters from \code{\LinkA{all\_plots\_save\_function}{all.Rul.plots.Rul.save.Rul.function}}

\item[\code{default\_save\_name}] default save name for saving, don't change, change save\_name in save\_param

\item[\code{grid\_color}] color of spatial grid

\item[\code{midpoint}] expression midpoint

\item[\code{scale\_alpha\_with\_expression}] scale expression with ggplot alpha parameter

\item[\code{...}] parameters for cowplot::save\_plot()
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Description of parameters.
\end{Details}
%
\begin{Value}
ggplot
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    spatGenePlot3D(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{spatialAEH}{spatialAEH}{spatialAEH}
%
\begin{Description}\relax
Compute spatial variable genes with spatialDE method
\end{Description}
%
\begin{Usage}
\begin{verbatim}
spatialAEH(
  gobject = NULL,
  SpatialDE_results = NULL,
  name_pattern = "AEH_patterns",
  expression_values = c("raw", "normalized", "scaled", "custom"),
  pattern_num = 6,
  l = 1.05,
  python_path = NULL,
  return_gobject = TRUE
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] Giotto object

\item[\code{SpatialDE\_results}] results of \code{\LinkA{SpatialDE}{SpatialDE}} function

\item[\code{name\_pattern}] name for the computed spatial patterns

\item[\code{expression\_values}] gene expression values to use

\item[\code{pattern\_num}] number of spatial patterns to look for

\item[\code{l}] lengthscale

\item[\code{python\_path}] specify specific path to python if required

\item[\code{return\_gobject}] show plot
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
This function is a wrapper for the SpatialAEH method implemented in the ...
\end{Details}
%
\begin{Value}
An updated giotto object
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    spatialAEH(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{spatialDE}{spatialDE}{spatialDE}
%
\begin{Description}\relax
Compute spatial variable genes with spatialDE method
\end{Description}
%
\begin{Usage}
\begin{verbatim}
spatialDE(
  gobject = NULL,
  expression_values = c("raw", "normalized", "scaled", "custom"),
  size = c(4, 2, 1),
  color = c("blue", "green", "red"),
  sig_alpha = 0.5,
  unsig_alpha = 0.5,
  python_path = NULL,
  show_plot = NA,
  return_plot = NA,
  save_plot = NA,
  save_param = list(),
  default_save_name = "SpatialDE"
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] Giotto object

\item[\code{expression\_values}] gene expression values to use

\item[\code{size}] size of plot

\item[\code{color}] low/medium/high color scheme for plot

\item[\code{sig\_alpha}] alpha value for significance

\item[\code{unsig\_alpha}] alpha value for unsignificance

\item[\code{python\_path}] specify specific path to python if required

\item[\code{show\_plot}] show plot

\item[\code{return\_plot}] return ggplot object

\item[\code{save\_plot}] directly save the plot [boolean]

\item[\code{save\_param}] list of saving parameters from all\_plots\_save\_function()

\item[\code{default\_save\_name}] default save name for saving, don't change, change save\_name in save\_param
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
This function is a wrapper for the SpatialDE method implemented in the ...
\end{Details}
%
\begin{Value}
a list of data.frames with results and plot (optional)
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    spatialDE(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{Spatial\_AEH}{Spatial\_AEH}{Spatial.Rul.AEH}
%
\begin{Description}\relax
calculate automatic expression histology with spatialDE method
\end{Description}
%
\begin{Usage}
\begin{verbatim}
Spatial_AEH(
  gobject = NULL,
  results = NULL,
  pattern_num = 5,
  l = 1.05,
  show_AEH = T,
  sdimx = NULL,
  sdimy = NULL,
  point_size = 3,
  point_alpha = 1,
  low_color = "blue",
  mid_color = "white",
  high_color = "red",
  midpoint = 0,
  python_path = NULL
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] Giotto object

\item[\code{results}] output from spatial\_DE

\item[\code{pattern\_num}] the number of gene expression patterns

\item[\code{show\_AEH}] show AEH plot

\item[\code{python\_path}] specify specific path to python if required
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Description.
\end{Details}
%
\begin{Value}
a list or a dataframe of SVs
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    Spatial_AEH(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{Spatial\_DE}{Spatial\_DE}{Spatial.Rul.DE}
%
\begin{Description}\relax
calculate spatial varible genes with spatialDE method
\end{Description}
%
\begin{Usage}
\begin{verbatim}
Spatial_DE(
  gobject = NULL,
  show_plot = T,
  size = c(4, 2, 1),
  color = c("blue", "green", "red"),
  sig_alpha = 0.5,
  unsig_alpha = 0.5,
  python_path = NULL
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] Giotto object

\item[\code{show\_plot}] show FSV plot

\item[\code{python\_path}] specify specific path to python if required
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Description.
\end{Details}
%
\begin{Value}
a list or a dataframe of SVs
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    Spatial_DE(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{spatNetwDistributions}{spatNetwDistributionsDistance}{spatNetwDistributions}
%
\begin{Description}\relax
This function return histograms displaying the distance distribution for each spatial k-neighbor
\end{Description}
%
\begin{Usage}
\begin{verbatim}
spatNetwDistributions(
  gobject,
  spatial_network_name = "spatial_network",
  distribution = c("distance", "k_neighbors"),
  hist_bins = 30,
  test_distance_limit = NULL,
  ncol = 1,
  show_plot = NA,
  return_plot = NA,
  save_plot = NA,
  save_param = list(),
  default_save_name = "spatNetwDistributions"
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] Giotto object

\item[\code{spatial\_network\_name}] name of spatial network

\item[\code{distribution}] show the distribution of cell-to-cell distance or number of k neighbors

\item[\code{hist\_bins}] number of binds to use for the histogram

\item[\code{test\_distance\_limit}] effect of different distance threshold on k-neighbors

\item[\code{ncol}] number of columns to visualize the histograms in

\item[\code{show\_plot}] show plot

\item[\code{return\_plot}] return ggplot object

\item[\code{save\_plot}] directly save the plot [boolean]

\item[\code{save\_param}] list of saving parameters from \code{\LinkA{all\_plots\_save\_function}{all.Rul.plots.Rul.save.Rul.function}}

\item[\code{default\_save\_name}] default save name for saving, alternatively change save\_name in save\_param
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
The \strong{distance} option shows the spatial distance distribution for each nearest neighbor rank (1st, 2nd, 3th, ... neigbor).
With this option the user can also test the effect of a distance limit on the spatial network. This distance limit can be used to remove neigbor
cells that are considered to far away. \\{}
The \strong{k\_neighbors} option shows the number of k neighbors distribution over all cells.
\end{Details}
%
\begin{Value}
ggplot plot
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    spatNetwDistributionsDistance(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{spatNetwDistributionsDistance}{spatNetwDistributionsDistance}{spatNetwDistributionsDistance}
%
\begin{Description}\relax
This function return histograms displaying the distance distribution for each spatial k-neighbor
\end{Description}
%
\begin{Usage}
\begin{verbatim}
spatNetwDistributionsDistance(
  gobject,
  spatial_network_name = "spatial_network",
  hist_bins = 30,
  test_distance_limit = NULL,
  ncol = 1,
  show_plot = NA,
  return_plot = NA,
  save_plot = NA,
  save_param = list(),
  default_save_name = "spatNetwDistributionsDistance"
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] Giotto object

\item[\code{spatial\_network\_name}] name of spatial network

\item[\code{hist\_bins}] number of binds to use for the histogram

\item[\code{test\_distance\_limit}] effect of different distance threshold on k-neighbors

\item[\code{ncol}] number of columns to visualize the histograms in

\item[\code{show\_plot}] show plot

\item[\code{return\_plot}] return ggplot object

\item[\code{save\_plot}] directly save the plot [boolean]

\item[\code{save\_param}] list of saving parameters from \code{\LinkA{all\_plots\_save\_function}{all.Rul.plots.Rul.save.Rul.function}}

\item[\code{default\_save\_name}] default save name for saving, alternatively change save\_name in save\_param
\end{ldescription}
\end{Arguments}
%
\begin{Value}
ggplot plot
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    spatNetwDistributionsDistance(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{spatNetwDistributionsKneighbors}{spatNetwDistributionsKneighbors}{spatNetwDistributionsKneighbors}
%
\begin{Description}\relax
This function returns a histogram displaying the number of k-neighbors distribution for each cell
\end{Description}
%
\begin{Usage}
\begin{verbatim}
spatNetwDistributionsKneighbors(
  gobject,
  spatial_network_name = "spatial_network",
  hist_bins = 30,
  show_plot = NA,
  return_plot = NA,
  save_plot = NA,
  save_param = list(),
  default_save_name = "spatNetwDistributionsKneighbors"
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] Giotto object

\item[\code{spatial\_network\_name}] name of spatial network

\item[\code{hist\_bins}] number of binds to use for the histogram

\item[\code{show\_plot}] show plot

\item[\code{return\_plot}] return ggplot object

\item[\code{save\_plot}] directly save the plot [boolean]

\item[\code{save\_param}] list of saving parameters from \code{\LinkA{all\_plots\_save\_function}{all.Rul.plots.Rul.save.Rul.function}}

\item[\code{default\_save\_name}] default save name for saving, alternatively change save\_name in save\_param
\end{ldescription}
\end{Arguments}
%
\begin{Value}
ggplot plot
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    spatNetwDistributionsKneighbors(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{spatPlot}{spatPlot}{spatPlot}
%
\begin{Description}\relax
Visualize cells according to spatial coordinates
\end{Description}
%
\begin{Usage}
\begin{verbatim}
spatPlot(
  gobject,
  group_by = NULL,
  group_by_subset = NULL,
  sdimx = "sdimx",
  sdimy = "sdimy",
  spat_enr_names = NULL,
  cell_color = NULL,
  color_as_factor = T,
  cell_color_code = NULL,
  cell_color_gradient = c("blue", "white", "red"),
  gradient_midpoint = NULL,
  gradient_limits = NULL,
  select_cell_groups = NULL,
  select_cells = NULL,
  point_shape = c("border", "no_border", "voronoi"),
  point_size = 3,
  point_border_col = "black",
  point_border_stroke = 0.1,
  show_cluster_center = F,
  show_center_label = F,
  center_point_size = 4,
  center_point_border_col = "black",
  center_point_border_stroke = 0.1,
  label_size = 4,
  label_fontface = "bold",
  show_network = F,
  spatial_network_name = NULL,
  network_color = NULL,
  network_alpha = 1,
  show_grid = F,
  spatial_grid_name = "spatial_grid",
  grid_color = NULL,
  show_other_cells = T,
  other_cell_color = "lightgrey",
  other_point_size = 1,
  other_cells_alpha = 0.1,
  coord_fix_ratio = NULL,
  title = NULL,
  show_legend = T,
  legend_text = 8,
  legend_symbol_size = 1,
  background_color = "white",
  vor_border_color = "white",
  vor_max_radius = 200,
  axis_text = 8,
  axis_title = 8,
  cow_n_col = 2,
  cow_rel_h = 1,
  cow_rel_w = 1,
  cow_align = "h",
  show_plot = NA,
  return_plot = NA,
  save_plot = NA,
  save_param = list(),
  default_save_name = "spatPlot"
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{group\_by\_subset}] subset the group\_by factor column

\item[\code{sdimx}] x-axis dimension name (default = 'sdimx')

\item[\code{sdimy}] y-axis dimension name (default = 'sdimy')

\item[\code{spat\_enr\_names}] names of spatial enrichment results to include

\item[\code{cell\_color}] color for cells (see details)

\item[\code{color\_as\_factor}] convert color column to factor

\item[\code{cell\_color\_code}] named vector with colors

\item[\code{cell\_color\_gradient}] vector with 3 colors for numeric data

\item[\code{gradient\_midpoint}] midpoint for color gradient

\item[\code{gradient\_limits}] vector with lower and upper limits

\item[\code{select\_cell\_groups}] select subset of cells/clusters based on cell\_color parameter

\item[\code{select\_cells}] select subset of cells based on cell IDs

\item[\code{point\_shape}] shape of points (border, no\_border or voronoi)

\item[\code{point\_size}] size of point (cell)

\item[\code{point\_border\_col}] color of border around points

\item[\code{point\_border\_stroke}] stroke size of border around points

\item[\code{show\_cluster\_center}] plot center of selected clusters

\item[\code{show\_center\_label}] plot label of selected clusters

\item[\code{center\_point\_size}] size of center points

\item[\code{label\_size}] size of labels

\item[\code{label\_fontface}] font of labels

\item[\code{show\_network}] show underlying spatial network

\item[\code{spatial\_network\_name}] name of spatial network to use

\item[\code{network\_color}] color of spatial network

\item[\code{network\_alpha}] alpha of spatial network

\item[\code{show\_grid}] show spatial grid

\item[\code{spatial\_grid\_name}] name of spatial grid to use

\item[\code{grid\_color}] color of spatial grid

\item[\code{show\_other\_cells}] display not selected cells

\item[\code{other\_cell\_color}] color of not selected cells

\item[\code{other\_point\_size}] point size of not selected cells

\item[\code{other\_cells\_alpha}] alpha of not selected cells

\item[\code{coord\_fix\_ratio}] fix ratio between x and y-axis

\item[\code{title}] title of plot

\item[\code{show\_legend}] show legend

\item[\code{legend\_text}] size of legend text

\item[\code{legend\_symbol\_size}] size of legend symbols

\item[\code{background\_color}] color of plot background

\item[\code{vor\_border\_color}] border colorr for voronoi plot

\item[\code{vor\_max\_radius}] maximum radius for voronoi 'cells'

\item[\code{axis\_text}] size of axis text

\item[\code{axis\_title}] size of axis title

\item[\code{cow\_n\_col}] cowplot param: how many columns

\item[\code{cow\_rel\_h}] cowplot param: relative height

\item[\code{cow\_rel\_w}] cowplot param: relative width

\item[\code{cow\_align}] cowplot param: how to align

\item[\code{show\_plot}] show plot

\item[\code{return\_plot}] return ggplot object

\item[\code{save\_plot}] directly save the plot [boolean]

\item[\code{save\_param}] list of saving parameters from \code{\LinkA{all\_plots\_save\_function}{all.Rul.plots.Rul.save.Rul.function}}

\item[\code{default\_save\_name}] default save name for saving, don't change, change save\_name in save\_param

\item[\code{groub\_by}] create multiple plots based on cell annotation column
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Description of parameters.
\end{Details}
%
\begin{Value}
ggplot
\end{Value}
%
\begin{SeeAlso}\relax
\code{\LinkA{spatPlot3D}{spatPlot3D}}
\end{SeeAlso}
%
\begin{Examples}
\begin{ExampleCode}
    spatPlot(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{spatPlot2D}{spatPlot2D}{spatPlot2D}
%
\begin{Description}\relax
Visualize cells according to spatial coordinates
\end{Description}
%
\begin{Usage}
\begin{verbatim}
spatPlot2D(
  gobject,
  group_by = NULL,
  group_by_subset = NULL,
  sdimx = "sdimx",
  sdimy = "sdimy",
  spat_enr_names = NULL,
  cell_color = NULL,
  color_as_factor = T,
  cell_color_code = NULL,
  cell_color_gradient = c("blue", "white", "red"),
  gradient_midpoint = NULL,
  gradient_limits = NULL,
  select_cell_groups = NULL,
  select_cells = NULL,
  point_shape = c("border", "no_border", "voronoi"),
  point_size = 3,
  point_border_col = "black",
  point_border_stroke = 0.1,
  show_cluster_center = F,
  show_center_label = F,
  center_point_size = 4,
  center_point_border_col = "black",
  center_point_border_stroke = 0.1,
  label_size = 4,
  label_fontface = "bold",
  show_network = F,
  spatial_network_name = NULL,
  network_color = NULL,
  network_alpha = 1,
  show_grid = F,
  spatial_grid_name = "spatial_grid",
  grid_color = NULL,
  show_other_cells = T,
  other_cell_color = "lightgrey",
  other_point_size = 1,
  other_cells_alpha = 0.1,
  coord_fix_ratio = NULL,
  title = NULL,
  show_legend = T,
  legend_text = 8,
  legend_symbol_size = 1,
  background_color = "white",
  vor_border_color = "white",
  vor_max_radius = 200,
  axis_text = 8,
  axis_title = 8,
  cow_n_col = 2,
  cow_rel_h = 1,
  cow_rel_w = 1,
  cow_align = "h",
  show_plot = NA,
  return_plot = NA,
  save_plot = NA,
  save_param = list(),
  default_save_name = "spatPlot2D"
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{group\_by\_subset}] subset the group\_by factor column

\item[\code{sdimx}] x-axis dimension name (default = 'sdimx')

\item[\code{sdimy}] y-axis dimension name (default = 'sdimy')

\item[\code{spat\_enr\_names}] names of spatial enrichment results to include

\item[\code{cell\_color}] color for cells (see details)

\item[\code{color\_as\_factor}] convert color column to factor

\item[\code{cell\_color\_code}] named vector with colors

\item[\code{cell\_color\_gradient}] vector with 3 colors for numeric data

\item[\code{gradient\_midpoint}] midpoint for color gradient

\item[\code{gradient\_limits}] vector with lower and upper limits

\item[\code{select\_cell\_groups}] select subset of cells/clusters based on cell\_color parameter

\item[\code{select\_cells}] select subset of cells based on cell IDs

\item[\code{point\_shape}] shape of points (border, no\_border or voronoi)

\item[\code{point\_size}] size of point (cell)

\item[\code{point\_border\_col}] color of border around points

\item[\code{point\_border\_stroke}] stroke size of border around points

\item[\code{show\_cluster\_center}] plot center of selected clusters

\item[\code{show\_center\_label}] plot label of selected clusters

\item[\code{center\_point\_size}] size of center points

\item[\code{label\_size}] size of labels

\item[\code{label\_fontface}] font of labels

\item[\code{show\_network}] show underlying spatial network

\item[\code{spatial\_network\_name}] name of spatial network to use

\item[\code{network\_color}] color of spatial network

\item[\code{network\_alpha}] alpha of spatial network

\item[\code{show\_grid}] show spatial grid

\item[\code{spatial\_grid\_name}] name of spatial grid to use

\item[\code{grid\_color}] color of spatial grid

\item[\code{show\_other\_cells}] display not selected cells

\item[\code{other\_cell\_color}] color of not selected cells

\item[\code{other\_point\_size}] point size of not selected cells

\item[\code{other\_cells\_alpha}] alpha of not selected cells

\item[\code{coord\_fix\_ratio}] fix ratio between x and y-axis

\item[\code{title}] title of plot

\item[\code{show\_legend}] show legend

\item[\code{legend\_text}] size of legend text

\item[\code{legend\_symbol\_size}] size of legend symbols

\item[\code{background\_color}] color of plot background

\item[\code{vor\_border\_color}] border colorr for voronoi plot

\item[\code{vor\_max\_radius}] maximum radius for voronoi 'cells'

\item[\code{axis\_text}] size of axis text

\item[\code{axis\_title}] size of axis title

\item[\code{cow\_n\_col}] cowplot param: how many columns

\item[\code{cow\_rel\_h}] cowplot param: relative height

\item[\code{cow\_rel\_w}] cowplot param: relative width

\item[\code{cow\_align}] cowplot param: how to align

\item[\code{show\_plot}] show plot

\item[\code{return\_plot}] return ggplot object

\item[\code{save\_plot}] directly save the plot [boolean]

\item[\code{save\_param}] list of saving parameters from \code{\LinkA{all\_plots\_save\_function}{all.Rul.plots.Rul.save.Rul.function}}

\item[\code{default\_save\_name}] default save name for saving, don't change, change save\_name in save\_param

\item[\code{groub\_by}] create multiple plots based on cell annotation column
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Description of parameters.
\end{Details}
%
\begin{Value}
ggplot
\end{Value}
%
\begin{SeeAlso}\relax
\code{\LinkA{spatPlot3D}{spatPlot3D}}
\end{SeeAlso}
%
\begin{Examples}
\begin{ExampleCode}
    spatPlot2D(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{spatPlot2D\_single}{spatPlot2D\_single}{spatPlot2D.Rul.single}
%
\begin{Description}\relax
Visualize cells according to spatial coordinates
\end{Description}
%
\begin{Usage}
\begin{verbatim}
spatPlot2D_single(
  gobject,
  sdimx = "sdimx",
  sdimy = "sdimy",
  spat_enr_names = NULL,
  cell_color = NULL,
  color_as_factor = T,
  cell_color_code = NULL,
  cell_color_gradient = c("blue", "white", "red"),
  gradient_midpoint = NULL,
  gradient_limits = NULL,
  select_cell_groups = NULL,
  select_cells = NULL,
  point_shape = c("border", "no_border", "voronoi"),
  point_size = 3,
  point_border_col = "black",
  point_border_stroke = 0.1,
  show_cluster_center = F,
  show_center_label = F,
  center_point_size = 4,
  center_point_border_col = "black",
  center_point_border_stroke = 0.1,
  label_size = 4,
  label_fontface = "bold",
  show_network = F,
  spatial_network_name = NULL,
  network_color = NULL,
  network_alpha = 1,
  show_grid = F,
  spatial_grid_name = "spatial_grid",
  grid_color = NULL,
  show_other_cells = T,
  other_cell_color = "lightgrey",
  other_point_size = 1,
  other_cells_alpha = 0.1,
  coord_fix_ratio = NULL,
  title = NULL,
  show_legend = T,
  legend_text = 8,
  legend_symbol_size = 1,
  background_color = "white",
  vor_border_color = "white",
  vor_max_radius = 200,
  axis_text = 8,
  axis_title = 8,
  show_plot = NA,
  return_plot = NA,
  save_plot = NA,
  save_param = list(),
  default_save_name = "spatPlot2D_single"
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{sdimx}] x-axis dimension name (default = 'sdimx')

\item[\code{sdimy}] y-axis dimension name (default = 'sdimy')

\item[\code{spat\_enr\_names}] names of spatial enrichment results to include

\item[\code{cell\_color}] color for cells (see details)

\item[\code{color\_as\_factor}] convert color column to factor

\item[\code{cell\_color\_code}] named vector with colors

\item[\code{cell\_color\_gradient}] vector with 3 colors for numeric data

\item[\code{gradient\_midpoint}] midpoint for color gradient

\item[\code{gradient\_limits}] vector with lower and upper limits

\item[\code{select\_cell\_groups}] select subset of cells/clusters based on cell\_color parameter

\item[\code{select\_cells}] select subset of cells based on cell IDs

\item[\code{point\_shape}] shape of points (border, no\_border or voronoi)

\item[\code{point\_size}] size of point (cell)

\item[\code{point\_border\_col}] color of border around points

\item[\code{point\_border\_stroke}] stroke size of border around points

\item[\code{show\_cluster\_center}] plot center of selected clusters

\item[\code{show\_center\_label}] plot label of selected clusters

\item[\code{center\_point\_size}] size of center points

\item[\code{label\_size}] size of labels

\item[\code{label\_fontface}] font of labels

\item[\code{show\_network}] show underlying spatial network

\item[\code{spatial\_network\_name}] name of spatial network to use

\item[\code{network\_color}] color of spatial network

\item[\code{network\_alpha}] alpha of spatial network

\item[\code{show\_grid}] show spatial grid

\item[\code{spatial\_grid\_name}] name of spatial grid to use

\item[\code{grid\_color}] color of spatial grid

\item[\code{show\_other\_cells}] display not selected cells

\item[\code{other\_cell\_color}] color of not selected cells

\item[\code{other\_point\_size}] point size of not selected cells

\item[\code{other\_cells\_alpha}] alpha of not selected cells

\item[\code{coord\_fix\_ratio}] fix ratio between x and y-axis

\item[\code{title}] title of plot

\item[\code{show\_legend}] show legend

\item[\code{legend\_text}] size of legend text

\item[\code{legend\_symbol\_size}] size of legend symbols

\item[\code{background\_color}] color of plot background

\item[\code{vor\_border\_color}] border colorr for voronoi plot

\item[\code{vor\_max\_radius}] maximum radius for voronoi 'cells'

\item[\code{axis\_text}] size of axis text

\item[\code{axis\_title}] size of axis title

\item[\code{show\_plot}] show plot

\item[\code{return\_plot}] return ggplot object

\item[\code{save\_plot}] directly save the plot [boolean]

\item[\code{save\_param}] list of saving parameters from \code{\LinkA{all\_plots\_save\_function}{all.Rul.plots.Rul.save.Rul.function}}

\item[\code{default\_save\_name}] default save name for saving, don't change, change save\_name in save\_param
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Description of parameters.
\end{Details}
%
\begin{Value}
ggplot
\end{Value}
%
\begin{SeeAlso}\relax
\code{\LinkA{spatPlot3D}{spatPlot3D}}
\end{SeeAlso}
%
\begin{Examples}
\begin{ExampleCode}
    spatPlot2D_single(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{spatPlot3D}{spatPlot3D}{spatPlot3D}
%
\begin{Description}\relax
Visualize cells according to spatial coordinates
\end{Description}
%
\begin{Usage}
\begin{verbatim}
spatPlot3D(
  gobject,
  sdimx = "sdimx",
  sdimy = "sdimy",
  sdimz = "sdimz",
  point_size = 3,
  cell_color = NULL,
  cell_color_code = NULL,
  select_cell_groups = NULL,
  select_cells = NULL,
  show_other_cells = T,
  other_cell_color = "lightgrey",
  other_point_size = 0.5,
  show_network = F,
  network_color = NULL,
  network_alpha = 1,
  other_cell_alpha = 0.5,
  spatial_network_name = "Delaunay_network",
  show_grid = F,
  grid_color = NULL,
  spatial_grid_name = "spatial_grid",
  title = "",
  show_legend = T,
  axis_scale = c("cube", "real", "custom"),
  custom_ratio = NULL,
  x_ticks = NULL,
  y_ticks = NULL,
  z_ticks = NULL,
  show_plot = NA,
  return_plot = NA,
  save_plot = NA,
  save_param = list(),
  default_save_name = "spat3D"
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{sdimx}] x-axis dimension name (default = 'sdimx')

\item[\code{sdimy}] y-axis dimension name (default = 'sdimy')

\item[\code{sdimz}] z-axis dimension name (default = 'sdimy')

\item[\code{point\_size}] size of point (cell)

\item[\code{cell\_color}] color for cells (see details)

\item[\code{cell\_color\_code}] named vector with colors

\item[\code{select\_cell\_groups}] select subset of cells/clusters based on cell\_color parameter

\item[\code{select\_cells}] select subset of cells based on cell IDs

\item[\code{show\_other\_cells}] display not selected cells

\item[\code{other\_cell\_color}] color of not selected cells

\item[\code{show\_network}] show underlying spatial network

\item[\code{network\_color}] color of spatial network

\item[\code{spatial\_network\_name}] name of spatial network to use

\item[\code{show\_grid}] show spatial grid

\item[\code{grid\_color}] color of spatial grid

\item[\code{spatial\_grid\_name}] name of spatial grid to use

\item[\code{title}] title of plot

\item[\code{show\_legend}] show legend

\item[\code{axis\_scale}] the way to scale the axis

\item[\code{custom\_ratio}] customize the scale of the plot

\item[\code{x\_ticks}] set the number of ticks on the x-axis

\item[\code{y\_ticks}] set the number of ticks on the y-axis

\item[\code{z\_ticks}] set the number of ticks on the z-axis

\item[\code{show\_plot}] show plot

\item[\code{return\_plot}] return ggplot object

\item[\code{save\_plot}] directly save the plot [boolean]

\item[\code{save\_param}] list of saving parameters from \code{\LinkA{all\_plots\_save\_function}{all.Rul.plots.Rul.save.Rul.function}}

\item[\code{default\_save\_name}] default save name for saving, don't change, change save\_name in save\_param
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Description of parameters.
\end{Details}
%
\begin{Value}
ggplot
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    spatPlot3D(gobject)

\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{spat\_fish\_func}{spat\_fish\_func}{spat.Rul.fish.Rul.func}
%
\begin{Description}\relax
performs fisher exact test
\end{Description}
%
\begin{Usage}
\begin{verbatim}
spat_fish_func(gene, bin_matrix, spat_mat, calc_hub = F, hub_min_int = 3)
\end{verbatim}
\end{Usage}
\inputencoding{utf8}
\HeaderA{spat\_OR\_func}{spat\_OR\_func}{spat.Rul.OR.Rul.func}
%
\begin{Description}\relax
calculate odds-ratio
\end{Description}
%
\begin{Usage}
\begin{verbatim}
spat_OR_func(gene, bin_matrix, spat_mat, calc_hub = F, hub_min_int = 3)
\end{verbatim}
\end{Usage}
\inputencoding{utf8}
\HeaderA{specificCellCellcommunicationScores}{specificCellCellcommunicationScores}{specificCellCellcommunicationScores}
%
\begin{Description}\relax
Specific Cell-Cell communication scores based on spatial expression of interacting cells
\end{Description}
%
\begin{Usage}
\begin{verbatim}
specificCellCellcommunicationScores(
  gobject,
  spatial_network_name = "Delaunay_network",
  cluster_column = "cell_types",
  random_iter = 100,
  cell_type_1 = "astrocyte",
  cell_type_2 = "endothelial",
  gene_set_1,
  gene_set_2,
  log2FC_addendum = 0.1,
  min_observations = 2,
  adjust_method = c("fdr", "bonferroni", "BH", "holm", "hochberg", "hommel", "BY",
    "none"),
  adjust_target = c("genes", "cells"),
  verbose = T
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object to use

\item[\code{spatial\_network\_name}] spatial network to use for identifying interacting cells

\item[\code{cluster\_column}] cluster column with cell type information

\item[\code{random\_iter}] number of iterations

\item[\code{cell\_type\_1}] first cell type

\item[\code{cell\_type\_2}] second cell type

\item[\code{gene\_set\_1}] first specific gene set from gene pairs

\item[\code{gene\_set\_2}] second specific gene set from gene pairs

\item[\code{log2FC\_addendum}] addendum to add when calculating log2FC

\item[\code{min\_observations}] minimum number of interactions needed to be considered

\item[\code{adjust\_method}] which method to adjust p-values

\item[\code{adjust\_target}] adjust multiple hypotheses at the cell or gene level

\item[\code{verbose}] verbose
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Statistical framework to identify if pairs of genes (such as ligand-receptor combinations)
are expressed at higher levels than expected based on a reshuffled null distribution
of gene expression values in cells that are spatially in proximity to eachother..
More details will follow soon.
\end{Details}
%
\begin{Value}
Cell-Cell communication scores for gene pairs based on spatial interaction
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    specificCellCellcommunicationScores(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{split\_dendrogram\_in\_two}{split\_dendrogram\_in\_two}{split.Rul.dendrogram.Rul.in.Rul.two}
%
\begin{Description}\relax
Merge selected clusters based on pairwise correlation scores and size of cluster.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
split_dendrogram_in_two(dend)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{dend}] dendrogram object
\end{ldescription}
\end{Arguments}
%
\begin{Value}
list of two dendrograms and height of node
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    split_dendrogram_in_two(dend)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{stitchFieldCoordinates}{stitchFieldCoordinates}{stitchFieldCoordinates}
%
\begin{Description}\relax
Helper function to stitch field coordinates together to form one complete picture
\end{Description}
%
\begin{Usage}
\begin{verbatim}
stitchFieldCoordinates(
  location_file,
  offset_file,
  cumulate_offset_x = F,
  cumulate_offset_y = F,
  field_col = "Field of View",
  X_coord_col = "X",
  Y_coord_col = "Y",
  reverse_final_x = F,
  reverse_final_y = T
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{location\_file}] location dataframe with X and Y coordinates

\item[\code{offset\_file}] dataframe that describes the offset for each field (see details)

\item[\code{cumulate\_offset\_x}] (boolean) Do the x-axis offset values need to be cumulated?

\item[\code{cumulate\_offset\_y}] (boolean) Do the y-axis offset values need to be cumulated?

\item[\code{field\_col}] column that indicates the field within the location\_file

\item[\code{X\_coord\_col}] column that indicates the x coordinates

\item[\code{Y\_coord\_col}] column that indicates the x coordinates

\item[\code{reverse\_final\_x}] (boolean) Do the final x coordinates need to be reversed?

\item[\code{reverse\_final\_y}] (boolean) Do the final y coordinates need to be reversed?
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Stitching of fields:
\begin{itemize}

\item{} 1. have cell locations: at least 3 columns: field, X, Y
\item{} 2. create offset file: offset file has 3 columns: field, x\_offset, y\_offset
\item{} 3. create new cell location file by stitching original cell locations with stitchFieldCoordinates
\item{} 4. provide new cell location file to \code{\LinkA{createGiottoObject}{createGiottoObject}}

\end{itemize}

\end{Details}
%
\begin{Value}
Updated location dataframe with new X ['X\_final'] and Y ['Y\_final'] coordinates
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    stitchFieldCoordinates(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{stitchTileCoordinates}{stitchTileCoordinates}{stitchTileCoordinates}
%
\begin{Description}\relax
Helper function to stitch tile coordinates together to form one complete picture
\end{Description}
%
\begin{Usage}
\begin{verbatim}
stitchTileCoordinates(location_file, Xtilespan, Ytilespan)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{location\_file}] location dataframe with X and Y coordinates

\item[\code{Xtilespan}] numerical value specifying the width of each tile

\item[\code{Ytilespan}] numerical value specifying the height of each tile
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
...
\end{Details}
%
\begin{Examples}
\begin{ExampleCode}
    stitchTileCoordinates(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{subClusterCells}{subClusterCells}{subClusterCells}
%
\begin{Description}\relax
subcluster cells
\end{Description}
%
\begin{Usage}
\begin{verbatim}
subClusterCells(
  gobject,
  name = "sub_clus",
  cluster_method = c("leiden", "louvain_community", "louvain_multinet"),
  cluster_column = NULL,
  selected_clusters = NULL,
  hvg_param = list(reverse_log_scale = T, difference_in_variance = 1, expression_values
    = "normalized"),
  hvg_min_perc_cells = 5,
  hvg_mean_expr_det = 1,
  use_all_genes_as_hvg = FALSE,
  min_nr_of_hvg = 5,
  pca_param = list(expression_values = "normalized", scale_unit = T),
  nn_param = list(dimensions_to_use = 1:20),
  k_neighbors = 10,
  resolution = 1,
  gamma = 1,
  omega = 1,
  python_path = NULL,
  nn_network_to_use = "sNN",
  network_name = "sNN.pca",
  return_gobject = TRUE,
  verbose = T
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{name}] name for new clustering result

\item[\code{cluster\_method}] clustering method to use

\item[\code{cluster\_column}] cluster column to subcluster

\item[\code{selected\_clusters}] only do subclustering on these clusters

\item[\code{hvg\_param}] parameters for calculateHVG

\item[\code{hvg\_min\_perc\_cells}] threshold for detection in min percentage of cells

\item[\code{hvg\_mean\_expr\_det}] threshold for mean expression level in cells with detection

\item[\code{use\_all\_genes\_as\_hvg}] forces all genes to be HVG and to be used as input for PCA

\item[\code{min\_nr\_of\_hvg}] minimum number of HVG, or all genes will be used as input for PCA

\item[\code{pca\_param}] parameters for runPCA

\item[\code{nn\_param}] parameters for parameters for createNearestNetwork

\item[\code{k\_neighbors}] number of k for createNearestNetwork

\item[\code{resolution}] resolution

\item[\code{gamma}] gamma

\item[\code{omega}] omega

\item[\code{python\_path}] specify specific path to python if required

\item[\code{nn\_network\_to\_use}] type of NN network to use (kNN vs sNN)

\item[\code{network\_name}] name of NN network to use

\item[\code{return\_gobject}] boolean: return giotto object (default = TRUE)

\item[\code{verbose}] verbose
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
This function performs subclustering on selected clusters.
The systematic steps are:
\begin{itemize}

\item{} 1. subset Giotto object
\item{} 2. identify highly variable genes
\item{} 3. run PCA
\item{} 4. create nearest neighbouring network
\item{} 5. do clustering

\end{itemize}

\end{Details}
%
\begin{Value}
giotto object with new subclusters appended to cell metadata
\end{Value}
%
\begin{SeeAlso}\relax
\code{\LinkA{doLouvainCluster\_multinet}{doLouvainCluster.Rul.multinet}}, \code{\LinkA{doLouvainCluster\_community}{doLouvainCluster.Rul.community}}
and  @seealso \code{\LinkA{doLeidenCluster}{doLeidenCluster}}
\end{SeeAlso}
%
\begin{Examples}
\begin{ExampleCode}
    subClusterCells(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{subsetGiotto}{subsetGiotto}{subsetGiotto}
%
\begin{Description}\relax
subsets Giotto object including previous analyses.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
subsetGiotto(gobject, cell_ids = NULL, gene_ids = NULL, verbose = FALSE)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{cell\_ids}] cell IDs to keep

\item[\code{gene\_ids}] gene IDs to keep

\item[\code{verbose}] be verbose
\end{ldescription}
\end{Arguments}
%
\begin{Value}
giotto object
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    subsetGiotto(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{subsetGiottoLocs}{subsetGiottoLocs}{subsetGiottoLocs}
%
\begin{Description}\relax
subsets Giotto object based on spatial locations
\end{Description}
%
\begin{Usage}
\begin{verbatim}
subsetGiottoLocs(
  gobject,
  x_max = NULL,
  x_min = NULL,
  y_max = NULL,
  y_min = NULL,
  z_max = NULL,
  z_min = NULL,
  return_gobject = T,
  verbose = FALSE
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{x\_max}] maximum x-coordinate

\item[\code{x\_min}] minimum x-coordinate

\item[\code{y\_max}] maximum y-coordinate

\item[\code{y\_min}] minimum y-coordinate

\item[\code{z\_max}] maximum z-coordinate

\item[\code{z\_min}] minimum z-coordinate

\item[\code{return\_gobject}] return Giotto object
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
if return\_gobject = FALSE, then a filtered combined metadata data.table will be returned
\end{Details}
%
\begin{Value}
giotto object
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    subsetGiottoLocs(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{transform\_2d\_mesh\_to\_3d\_mesh}{transform\_2d\_mesh\_to\_3d\_mesh}{transform.Rul.2d.Rul.mesh.Rul.to.Rul.3d.Rul.mesh}
%
\begin{Description}\relax
transform 2d mesh to 3d mesh by reversing PCA
\end{Description}
%
\begin{Usage}
\begin{verbatim}
transform_2d_mesh_to_3d_mesh(
  mesh_line_obj_2d,
  pca_out,
  center_vec,
  mesh_grid_n
)
\end{verbatim}
\end{Usage}
\inputencoding{utf8}
\HeaderA{trendSceek}{trendSceek}{trendSceek}
%
\begin{Description}\relax
Compute spatial variable genes with trendsceek method
\end{Description}
%
\begin{Usage}
\begin{verbatim}
trendSceek(
  gobject,
  expression_values = c("normalized", "raw"),
  subset_genes = NULL,
  nrand = 100,
  ncores = 8,
  ...
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] Giotto object

\item[\code{expression\_values}] gene expression values to use

\item[\code{subset\_genes}] subset of genes to run trendsceek on

\item[\code{nrand}] An integer specifying the number of random resamplings of the mark distribution as to create the null-distribution.

\item[\code{ncores}] An integer specifying the number of cores to be used by BiocParallel

\item[\code{...}] Additional parameters to the \code{\LinkA{trendsceek\_test}{trendsceek.Rul.test}} function
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
This function is a wrapper for the trendsceek\_test method implemented in the trendsceek package
\end{Details}
%
\begin{Value}
data.frame with trendsceek spatial genes results
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    trendSceek(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{viewHMRFresults}{viewHMRFresults}{viewHMRFresults}
%
\begin{Description}\relax
View results from doHMRF.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
viewHMRFresults(
  gobject,
  HMRFoutput,
  k = NULL,
  betas_to_view = NULL,
  third_dim = NULL,
  ...
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{HMRFoutput}] HMRF output from doHMRF

\item[\code{k}] number of HMRF domains

\item[\code{betas\_to\_view}] results from different betas that you want to view

\item[\code{...}] paramters to visPlot()
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Description ...
\end{Details}
%
\begin{Value}
spatial plots with HMRF domains
\end{Value}
%
\begin{SeeAlso}\relax
\code{\LinkA{visPlot}{visPlot}}
\end{SeeAlso}
%
\begin{Examples}
\begin{ExampleCode}
    viewHMRFresults(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{viewHMRFresults2D}{viewHMRFresults2D}{viewHMRFresults2D}
%
\begin{Description}\relax
View results from doHMRF.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
viewHMRFresults2D(
  gobject,
  HMRFoutput,
  k = NULL,
  betas_to_view = NULL,
  third_dim = NULL,
  ...
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{HMRFoutput}] HMRF output from doHMRF

\item[\code{k}] number of HMRF domains

\item[\code{betas\_to\_view}] results from different betas that you want to view

\item[\code{...}] paramters to visPlot()
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Description ...
\end{Details}
%
\begin{Value}
spatial plots with HMRF domains
\end{Value}
%
\begin{SeeAlso}\relax
\code{\LinkA{spatPlot2D}{spatPlot2D}}
\end{SeeAlso}
%
\begin{Examples}
\begin{ExampleCode}
    viewHMRFresults2D(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{viewHMRFresults3D}{viewHMRFresults3D}{viewHMRFresults3D}
%
\begin{Description}\relax
View results from doHMRF.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
viewHMRFresults3D(
  gobject,
  HMRFoutput,
  k = NULL,
  betas_to_view = NULL,
  third_dim = NULL,
  ...
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{HMRFoutput}] HMRF output from doHMRF

\item[\code{k}] number of HMRF domains

\item[\code{betas\_to\_view}] results from different betas that you want to view

\item[\code{...}] paramters to visPlot()
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Description ...
\end{Details}
%
\begin{Value}
spatial plots with HMRF domains
\end{Value}
%
\begin{SeeAlso}\relax
\code{\LinkA{spatPlot3D}{spatPlot3D}}
\end{SeeAlso}
%
\begin{Examples}
\begin{ExampleCode}
    viewHMRFresults3D(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{violinPlot}{violinPlot}{violinPlot}
%
\begin{Description}\relax
Creates violinplot for selected clusters
\end{Description}
%
\begin{Usage}
\begin{verbatim}
violinPlot(
  gobject,
  expression_values = c("normalized", "scaled", "custom"),
  genes,
  cluster_column,
  cluster_custom_order = NULL,
  color_violin = c("genes", "cluster"),
  cluster_color_code = NULL,
  strip_position = c("top", "right", "left", "bottom"),
  strip_text = 7,
  axis_text_x_size = 10,
  axis_text_y_size = 6,
  show_plot = NA,
  return_plot = NA,
  save_plot = NA,
  save_param = list(),
  default_save_name = "violinPlot"
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{expression\_values}] expression values to use

\item[\code{genes}] genes to plot

\item[\code{cluster\_column}] name of column to use for clusters

\item[\code{cluster\_custom\_order}] custom order of clusters

\item[\code{color\_violin}] color violin according to genes or clusters

\item[\code{cluster\_color\_code}] color code for clusters

\item[\code{strip\_position}] position of gene labels

\item[\code{strip\_text}] size of strip text

\item[\code{axis\_text\_x\_size}] size of x-axis text

\item[\code{axis\_text\_y\_size}] size of y-axis text

\item[\code{show\_plot}] show plot

\item[\code{return\_plot}] return ggplot object

\item[\code{save\_plot}] directly save the plot [boolean]

\item[\code{save\_param}] list of saving parameters from \code{\LinkA{all\_plots\_save\_function}{all.Rul.plots.Rul.save.Rul.function}}

\item[\code{default\_save\_name}] default save name for saving, don't change, change save\_name in save\_param
\end{ldescription}
\end{Arguments}
%
\begin{Value}
ggplot
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    violinPlot(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{visDimGenePlot}{visDimGenePlot}{visDimGenePlot}
%
\begin{Description}\relax
Visualize cells and gene expression according to dimension reduction coordinates
\end{Description}
%
\begin{Usage}
\begin{verbatim}
visDimGenePlot(
  gobject,
  expression_values = c("normalized", "scaled", "custom"),
  genes = NULL,
  dim_reduction_to_use = "umap",
  dim_reduction_name = "umap",
  dim1_to_use = 1,
  dim2_to_use = 2,
  dim3_to_use = NULL,
  show_NN_network = F,
  nn_network_to_use = "sNN",
  network_name = "sNN.pca",
  network_color = "lightgray",
  edge_alpha = NULL,
  scale_alpha_with_expression = FALSE,
  point_size = 1,
  genes_high_color = NULL,
  genes_mid_color = "white",
  genes_low_color = "blue",
  point_border_col = "black",
  point_border_stroke = 0.1,
  midpoint = 0,
  cow_n_col = 2,
  cow_rel_h = 1,
  cow_rel_w = 1,
  cow_align = "h",
  show_legend = T,
  plot_method = c("ggplot", "plotly"),
  show_plots = F
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{expression\_values}] gene expression values to use

\item[\code{genes}] genes to show

\item[\code{dim\_reduction\_to\_use}] dimension reduction to use

\item[\code{dim\_reduction\_name}] dimension reduction name

\item[\code{dim1\_to\_use}] dimension to use on x-axis

\item[\code{dim2\_to\_use}] dimension to use on y-axis

\item[\code{dim3\_to\_use}] dimension to use on z-axis

\item[\code{show\_NN\_network}] show underlying NN network

\item[\code{nn\_network\_to\_use}] type of NN network to use (kNN vs sNN)

\item[\code{network\_name}] name of NN network to use, if show\_NN\_network = TRUE

\item[\code{edge\_alpha}] column to use for alpha of the edges

\item[\code{scale\_alpha\_with\_expression}] scale expression with ggplot alpha parameter

\item[\code{point\_size}] size of point (cell)

\item[\code{point\_border\_col}] color of border around points

\item[\code{point\_border\_stroke}] stroke size of border around points

\item[\code{midpoint}] size of point (cell)

\item[\code{cow\_n\_col}] cowplot param: how many columns

\item[\code{cow\_rel\_h}] cowplot param: relative height

\item[\code{cow\_rel\_w}] cowplot param: relative width

\item[\code{cow\_align}] cowplot param: how to align

\item[\code{show\_legend}] show legend

\item[\code{show\_plots}] show plots
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Description of parameters.
\end{Details}
%
\begin{Value}
ggplot
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    visDimGenePlot(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{visDimGenePlot\_2D\_ggplot}{visDimGenePlot\_2D\_ggplot}{visDimGenePlot.Rul.2D.Rul.ggplot}
%
\begin{Description}\relax
Visualize cells and gene expression according to dimension reduction coordinates
\end{Description}
%
\begin{Usage}
\begin{verbatim}
visDimGenePlot_2D_ggplot(
  gobject,
  expression_values = c("normalized", "scaled", "custom"),
  genes = NULL,
  dim_reduction_to_use = "umap",
  dim_reduction_name = "umap",
  dim1_to_use = 1,
  dim2_to_use = 2,
  show_NN_network = F,
  nn_network_to_use = "sNN",
  network_name = "sNN.pca",
  network_color = "lightgray",
  edge_alpha = NULL,
  scale_alpha_with_expression = FALSE,
  point_size = 1,
  genes_high_color = "red",
  genes_mid_color = "white",
  genes_low_color = "blue",
  point_border_col = "black",
  point_border_stroke = 0.1,
  midpoint = 0,
  cow_n_col = 2,
  cow_rel_h = 1,
  cow_rel_w = 1,
  cow_align = "h",
  show_legend = T,
  show_plots = F
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{expression\_values}] gene expression values to use

\item[\code{genes}] genes to show

\item[\code{dim\_reduction\_to\_use}] dimension reduction to use

\item[\code{dim\_reduction\_name}] dimension reduction name

\item[\code{dim1\_to\_use}] dimension to use on x-axis

\item[\code{dim2\_to\_use}] dimension to use on y-axis

\item[\code{show\_NN\_network}] show underlying NN network

\item[\code{nn\_network\_to\_use}] type of NN network to use (kNN vs sNN)

\item[\code{network\_name}] name of NN network to use, if show\_NN\_network = TRUE

\item[\code{edge\_alpha}] column to use for alpha of the edges

\item[\code{scale\_alpha\_with\_expression}] scale expression with ggplot alpha parameter

\item[\code{point\_size}] size of point (cell)

\item[\code{point\_border\_col}] color of border around points

\item[\code{point\_border\_stroke}] stroke size of border around points

\item[\code{midpoint}] size of point (cell)

\item[\code{cow\_n\_col}] cowplot param: how many columns

\item[\code{cow\_rel\_h}] cowplot param: relative height

\item[\code{cow\_rel\_w}] cowplot param: relative width

\item[\code{cow\_align}] cowplot param: how to align

\item[\code{show\_legend}] show legend

\item[\code{show\_plots}] show plots
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Description of parameters.
\end{Details}
%
\begin{Value}
ggplot
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    visDimGenePlot_2D_ggplot(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{visDimGenePlot\_3D\_plotly}{visDimGenePlot\_3D\_plotly}{visDimGenePlot.Rul.3D.Rul.plotly}
%
\begin{Description}\relax
Visualize cells and gene expression according to dimension reduction coordinates
\end{Description}
%
\begin{Usage}
\begin{verbatim}
visDimGenePlot_3D_plotly(
  gobject,
  expression_values = c("normalized", "scaled", "custom"),
  genes = NULL,
  dim_reduction_to_use = "umap",
  dim_reduction_name = "umap",
  dim1_to_use = 1,
  dim2_to_use = 2,
  dim3_to_use = 3,
  show_NN_network = F,
  nn_network_to_use = "sNN",
  network_name = "sNN.pca",
  network_color = "lightgray",
  edge_alpha = NULL,
  point_size = 1,
  genes_high_color = NULL,
  genes_mid_color = "white",
  genes_low_color = "blue",
  show_legend = T,
  show_plots = F
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{expression\_values}] gene expression values to use

\item[\code{genes}] genes to show

\item[\code{dim\_reduction\_to\_use}] dimension reduction to use

\item[\code{dim\_reduction\_name}] dimension reduction name

\item[\code{dim1\_to\_use}] dimension to use on x-axis

\item[\code{dim2\_to\_use}] dimension to use on y-axis

\item[\code{dim3\_to\_use}] dimension to use on z-axis

\item[\code{show\_NN\_network}] show underlying NN network

\item[\code{nn\_network\_to\_use}] type of NN network to use (kNN vs sNN)

\item[\code{network\_name}] name of NN network to use, if show\_NN\_network = TRUE

\item[\code{edge\_alpha}] column to use for alpha of the edges

\item[\code{point\_size}] size of point (cell)

\item[\code{show\_legend}] show legend

\item[\code{show\_plots}] show plots
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Description of parameters.
\end{Details}
%
\begin{Value}
ggplot
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    visDimGenePlot_3D_plotly(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{visDimPlot}{visDimPlot}{visDimPlot}
%
\begin{Description}\relax
Visualize cells according to dimension reduction coordinates
\end{Description}
%
\begin{Usage}
\begin{verbatim}
visDimPlot(
  gobject,
  dim_reduction_to_use = "umap",
  dim_reduction_name = "umap",
  dim1_to_use = 1,
  dim2_to_use = 2,
  dim3_to_use = NULL,
  show_NN_network = F,
  nn_network_to_use = "sNN",
  network_name = "sNN.pca",
  cell_color = NULL,
  color_as_factor = T,
  cell_color_code = NULL,
  select_cell_groups = NULL,
  select_cells = NULL,
  show_other_cells = T,
  other_cell_color = "lightgrey",
  other_point_size = 0.5,
  show_cluster_center = F,
  show_center_label = T,
  center_point_size = 4,
  center_point_border_col = "black",
  center_point_border_stroke = 0.1,
  label_size = 4,
  label_fontface = "bold",
  edge_alpha = NULL,
  point_size = 3,
  point_border_col = "black",
  point_border_stroke = 0.1,
  plot_method = c("ggplot", "plotly"),
  show_legend = T,
  show_plot = F,
  return_plot = TRUE,
  save_plot = F,
  save_dir = NULL,
  save_folder = NULL,
  save_name = NULL,
  save_format = NULL,
  show_saved_plot = F,
  ...
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{dim\_reduction\_to\_use}] dimension reduction to use

\item[\code{dim\_reduction\_name}] dimension reduction name

\item[\code{dim1\_to\_use}] dimension to use on x-axis

\item[\code{dim2\_to\_use}] dimension to use on y-axis

\item[\code{dim3\_to\_use}] dimension to use on z-axis

\item[\code{show\_NN\_network}] show underlying NN network

\item[\code{nn\_network\_to\_use}] type of NN network to use (kNN vs sNN)

\item[\code{network\_name}] name of NN network to use, if show\_NN\_network = TRUE

\item[\code{cell\_color}] color for cells (see details)

\item[\code{color\_as\_factor}] convert color column to factor

\item[\code{cell\_color\_code}] named vector with colors

\item[\code{show\_cluster\_center}] plot center of selected clusters

\item[\code{show\_center\_label}] plot label of selected clusters

\item[\code{center\_point\_size}] size of center points

\item[\code{label\_size}] size of labels

\item[\code{label\_fontface}] font of labels

\item[\code{edge\_alpha}] column to use for alpha of the edges

\item[\code{point\_size}] size of point (cell)

\item[\code{point\_border\_col}] color of border around points

\item[\code{point\_border\_stroke}] stroke size of border around points

\item[\code{show\_legend}] show legend

\item[\code{show\_plot}] show plot

\item[\code{return\_plot}] return ggplot object

\item[\code{save\_plot}] directly save the plot [boolean]

\item[\code{save\_dir}] directory to save the plot

\item[\code{save\_folder}] (optional) folder in directory to save the plot

\item[\code{save\_name}] name of plot

\item[\code{save\_format}] format of plot (e.g. tiff, png, pdf, ...)

\item[\code{show\_saved\_plot}] load \& display the saved plot
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Description of parameters.
\end{Details}
%
\begin{Value}
ggplot or plotly
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    visDimPlot(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{visDimPlot\_2D\_ggplot}{visDimPlot\_2D\_ggplot}{visDimPlot.Rul.2D.Rul.ggplot}
%
\begin{Description}\relax
Visualize cells according to dimension reduction coordinates
\end{Description}
%
\begin{Usage}
\begin{verbatim}
visDimPlot_2D_ggplot(
  gobject,
  dim_reduction_to_use = "umap",
  dim_reduction_name = "umap",
  dim1_to_use = 1,
  dim2_to_use = 2,
  show_NN_network = F,
  nn_network_to_use = "sNN",
  network_name = "sNN.pca",
  cell_color = NULL,
  color_as_factor = T,
  cell_color_code = NULL,
  select_cell_groups = NULL,
  select_cells = NULL,
  show_other_cells = T,
  other_cell_color = "lightgrey",
  other_point_size = 0.5,
  show_cluster_center = F,
  show_center_label = T,
  center_point_size = 4,
  center_point_border_col = "black",
  center_point_border_stroke = 0.1,
  label_size = 4,
  label_fontface = "bold",
  edge_alpha = NULL,
  point_size = 1,
  point_border_col = "black",
  point_border_stroke = 0.1,
  show_legend = T,
  show_plot = F,
  return_plot = TRUE,
  save_plot = F,
  save_dir = NULL,
  save_folder = NULL,
  save_name = NULL,
  save_format = NULL,
  show_saved_plot = F,
  ...
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{dim\_reduction\_to\_use}] dimension reduction to use

\item[\code{dim\_reduction\_name}] dimension reduction name

\item[\code{dim1\_to\_use}] dimension to use on x-axis

\item[\code{dim2\_to\_use}] dimension to use on y-axis

\item[\code{show\_NN\_network}] show underlying NN network

\item[\code{nn\_network\_to\_use}] type of NN network to use (kNN vs sNN)

\item[\code{network\_name}] name of NN network to use, if show\_NN\_network = TRUE

\item[\code{cell\_color}] color for cells (see details)

\item[\code{color\_as\_factor}] convert color column to factor

\item[\code{cell\_color\_code}] named vector with colors

\item[\code{select\_cell\_groups}] select subset of cells/clusters based on cell\_color parameter

\item[\code{select\_cells}] select subset of cells based on cell IDs

\item[\code{show\_other\_cells}] display not selected cells

\item[\code{other\_cell\_color}] color of not selected cells

\item[\code{other\_point\_size}] size of not selected cells

\item[\code{show\_cluster\_center}] plot center of selected clusters

\item[\code{show\_center\_label}] plot label of selected clusters

\item[\code{center\_point\_size}] size of center points

\item[\code{label\_size}] size of labels

\item[\code{label\_fontface}] font of labels

\item[\code{edge\_alpha}] column to use for alpha of the edges

\item[\code{point\_size}] size of point (cell)

\item[\code{point\_border\_col}] color of border around points

\item[\code{point\_border\_stroke}] stroke size of border around points

\item[\code{show\_legend}] show legend
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Description of parameters.
\end{Details}
%
\begin{Value}
ggplot
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    visDimPlot_2D_ggplot(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{visDimPlot\_2D\_plotly}{visDimPlot\_2D\_plotly}{visDimPlot.Rul.2D.Rul.plotly}
%
\begin{Description}\relax
Visualize cells according to dimension reduction coordinates
\end{Description}
%
\begin{Usage}
\begin{verbatim}
visDimPlot_2D_plotly(
  gobject,
  dim_reduction_to_use = "umap",
  dim_reduction_name = "umap",
  dim1_to_use = 1,
  dim2_to_use = 2,
  select_cell_groups = NULL,
  select_cells = NULL,
  show_other_cells = T,
  other_cell_color = "lightgrey",
  other_point_size = 0.5,
  show_NN_network = F,
  nn_network_to_use = "sNN",
  network_name = "sNN.pca",
  color_as_factor = T,
  cell_color = NULL,
  cell_color_code = NULL,
  show_cluster_center = F,
  show_center_label = T,
  center_point_size = 4,
  label_size = 4,
  edge_alpha = NULL,
  point_size = 5
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{dim\_reduction\_to\_use}] dimension reduction to use

\item[\code{dim\_reduction\_name}] dimension reduction name

\item[\code{dim1\_to\_use}] dimension to use on x-axis

\item[\code{dim2\_to\_use}] dimension to use on y-axis

\item[\code{show\_NN\_network}] show underlying NN network

\item[\code{nn\_network\_to\_use}] type of NN network to use (kNN vs sNN)

\item[\code{network\_name}] name of NN network to use, if show\_NN\_network = TRUE

\item[\code{color\_as\_factor}] convert color column to factor

\item[\code{cell\_color}] color for cells (see details)

\item[\code{cell\_color\_code}] named vector with colors

\item[\code{show\_cluster\_center}] plot center of selected clusters

\item[\code{show\_center\_label}] plot label of selected clusters

\item[\code{center\_point\_size}] size of center points

\item[\code{label\_size}] size of labels

\item[\code{edge\_alpha}] column to use for alpha of the edges

\item[\code{point\_size}] size of point (cell)
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Description of parameters.
\end{Details}
%
\begin{Value}
plotly
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    visDimPlot_2D_plotly(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{visDimPlot\_3D\_plotly}{visDimPlot\_3D\_plotly}{visDimPlot.Rul.3D.Rul.plotly}
%
\begin{Description}\relax
Visualize cells according to dimension reduction coordinates
\end{Description}
%
\begin{Usage}
\begin{verbatim}
visDimPlot_3D_plotly(
  gobject,
  dim_reduction_to_use = "umap",
  dim_reduction_name = "umap",
  dim1_to_use = 1,
  dim2_to_use = 2,
  dim3_to_use = 3,
  select_cell_groups = NULL,
  select_cells = NULL,
  show_other_cells = T,
  other_cell_color = "lightgrey",
  other_point_size = 0.5,
  show_NN_network = F,
  nn_network_to_use = "sNN",
  network_name = "sNN.pca",
  color_as_factor = T,
  cell_color = NULL,
  cell_color_code = NULL,
  show_cluster_center = F,
  show_center_label = T,
  center_point_size = 4,
  label_size = 4,
  edge_alpha = NULL,
  point_size = 1
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{dim\_reduction\_to\_use}] dimension reduction to use

\item[\code{dim\_reduction\_name}] dimension reduction name

\item[\code{dim1\_to\_use}] dimension to use on x-axis

\item[\code{dim2\_to\_use}] dimension to use on y-axis

\item[\code{dim3\_to\_use}] dimension to use on z-axis

\item[\code{show\_NN\_network}] show underlying NN network

\item[\code{nn\_network\_to\_use}] type of NN network to use (kNN vs sNN)

\item[\code{network\_name}] name of NN network to use, if show\_NN\_network = TRUE

\item[\code{color\_as\_factor}] convert color column to factor

\item[\code{cell\_color}] color for cells (see details)

\item[\code{cell\_color\_code}] named vector with colors

\item[\code{show\_cluster\_center}] plot center of selected clusters

\item[\code{show\_center\_label}] plot label of selected clusters

\item[\code{center\_point\_size}] size of center points

\item[\code{label\_size}] size of labels

\item[\code{edge\_alpha}] column to use for alpha of the edges

\item[\code{point\_size}] size of point (cell)
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Description of parameters.
\end{Details}
%
\begin{Value}
plotly
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    visDimPlot_3D_plotly(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{visForceLayoutPlot}{visForceLayoutPlot}{visForceLayoutPlot}
%
\begin{Description}\relax
Visualize cells according to forced layout algorithm coordinates
\end{Description}
%
\begin{Usage}
\begin{verbatim}
visForceLayoutPlot(
  gobject,
  nn_network_to_use = "sNN",
  network_name = "sNN.pca",
  layout_name = "layout",
  dim1_to_use = 1,
  dim2_to_use = 2,
  show_NN_network = T,
  cell_color = NULL,
  color_as_factor = TRUE,
  cell_color_code = NULL,
  edge_alpha = NULL,
  point_size = 1,
  point_border_col = "black",
  point_border_stroke = 0.1,
  show_legend = T,
  show_plot = F,
  return_plot = TRUE,
  save_plot = F,
  save_dir = NULL,
  save_folder = NULL,
  save_name = NULL,
  save_format = NULL,
  show_saved_plot = F,
  ...
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{nn\_network\_to\_use}] type of NN network to use (kNN vs sNN)

\item[\code{network\_name}] NN network to use

\item[\code{layout\_name}] name of layout to use

\item[\code{dim1\_to\_use}] dimension to use on x-axis

\item[\code{dim2\_to\_use}] dimension to use on y-axis

\item[\code{show\_NN\_network}] show underlying NN network

\item[\code{cell\_color}] color for cells (see details)

\item[\code{color\_as\_factor}] convert color column to factor

\item[\code{cell\_color\_code}] named vector with colors

\item[\code{edge\_alpha}] column to use for alpha of the edges

\item[\code{point\_size}] size of point (cell)

\item[\code{point\_border\_col}] color of border around points

\item[\code{point\_border\_stroke}] stroke size of border around points

\item[\code{show\_legend}] show legend

\item[\code{show\_plot}] show plot

\item[\code{return\_plot}] return ggplot object

\item[\code{save\_plot}] directly save the plot [boolean]

\item[\code{save\_dir}] directory to save the plot

\item[\code{save\_folder}] (optional) folder in directory to save the plot

\item[\code{save\_name}] name of plot

\item[\code{save\_format}] format of plot (e.g. tiff, png, pdf, ...)

\item[\code{show\_saved\_plot}] load \& display the saved plot
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Description of parameters.
\end{Details}
%
\begin{Value}
ggplot
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    visForceLayoutPlot(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{visGenePlot}{visGenePlot}{visGenePlot}
%
\begin{Description}\relax
Visualize cells and gene expression according to spatial coordinates
\end{Description}
%
\begin{Usage}
\begin{verbatim}
visGenePlot(
  gobject,
  expression_values = c("normalized", "scaled", "custom"),
  genes,
  genes_high_color = NULL,
  genes_mid_color = "white",
  genes_low_color = "blue",
  show_network = F,
  network_color = NULL,
  spatial_network_name = "spatial_network",
  edge_alpha = NULL,
  show_grid = F,
  grid_color = NULL,
  spatial_grid_name = "spatial_grid",
  midpoint = 0,
  scale_alpha_with_expression = FALSE,
  point_size = 1,
  point_border_col = "black",
  point_border_stroke = 0.1,
  show_legend = T,
  cow_n_col = 2,
  cow_rel_h = 1,
  cow_rel_w = 1,
  cow_align = "h",
  axis_scale = c("cube", "real", "custom"),
  custom_ratio = NULL,
  x_ticks = NULL,
  y_ticks = NULL,
  z_ticks = NULL,
  plot_method = c("ggplot", "plotly"),
  show_plots = F
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{expression\_values}] gene expression values to use

\item[\code{genes}] genes to show

\item[\code{genes\_high\_color}] color represents high gene expression

\item[\code{genes\_mid\_color}] color represents middle gene expression

\item[\code{genes\_low\_color}] color represents low gene expression

\item[\code{show\_network}] show underlying spatial network

\item[\code{network\_color}] color of spatial network

\item[\code{spatial\_network\_name}] name of spatial network to use

\item[\code{show\_grid}] show spatial grid

\item[\code{grid\_color}] color of spatial grid

\item[\code{spatial\_grid\_name}] name of spatial grid to use

\item[\code{midpoint}] expression midpoint

\item[\code{scale\_alpha\_with\_expression}] scale expression with ggplot alpha parameter

\item[\code{point\_size}] size of point (cell)

\item[\code{point\_border\_col}] color of border around points

\item[\code{point\_border\_stroke}] stroke size of border around points

\item[\code{show\_legend}] show legend

\item[\code{cow\_n\_col}] cowplot param: how many columns

\item[\code{cow\_rel\_h}] cowplot param: relative height

\item[\code{cow\_rel\_w}] cowplot param: relative width

\item[\code{cow\_align}] cowplot param: how to align

\item[\code{axis\_scale}] three mode to adjust axis scale

\item[\code{x\_ticks}] number of ticks on x axis

\item[\code{y\_ticks}] number of ticks on y axis

\item[\code{z\_ticks}] number of ticks on z axis

\item[\code{plot\_method}] two methods of plot

\item[\code{show\_plots}] show plots
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Description of parameters.
\end{Details}
%
\begin{Value}
ggplot or plotly
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    visGenePlot(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{visGenePlot\_2D\_ggplot}{visGenePlot\_2D\_ggplot}{visGenePlot.Rul.2D.Rul.ggplot}
%
\begin{Description}\relax
Visualize cells and gene expression according to spatial coordinates
\end{Description}
%
\begin{Usage}
\begin{verbatim}
visGenePlot_2D_ggplot(
  gobject,
  expression_values = c("normalized", "scaled", "custom"),
  genes,
  genes_high_color = "darkred",
  genes_mid_color = "white",
  genes_low_color = "darkblue",
  show_network = F,
  network_color = NULL,
  spatial_network_name = "spatial_network",
  edge_alpha = NULL,
  show_grid = F,
  grid_color = NULL,
  spatial_grid_name = "spatial_grid",
  midpoint = 0,
  scale_alpha_with_expression = FALSE,
  point_size = 1,
  point_border_col = "black",
  point_border_stroke = 0.1,
  show_legend = T,
  cow_n_col = 2,
  cow_rel_h = 1,
  cow_rel_w = 1,
  cow_align = "h",
  show_plots = F
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{expression\_values}] gene expression values to use

\item[\code{genes}] genes to show

\item[\code{genes\_high\_color}] color represents high gene expression

\item[\code{genes\_mid\_color}] color represents middle gene expression

\item[\code{genes\_low\_color}] color represents low gene expression

\item[\code{show\_network}] show underlying spatial network

\item[\code{network\_color}] color of spatial network

\item[\code{spatial\_network\_name}] name of spatial network to use

\item[\code{show\_grid}] show spatial grid

\item[\code{grid\_color}] color of spatial grid

\item[\code{spatial\_grid\_name}] name of spatial grid to use

\item[\code{midpoint}] expression midpoint

\item[\code{scale\_alpha\_with\_expression}] scale expression with ggplot alpha parameter

\item[\code{point\_size}] size of point (cell)

\item[\code{point\_border\_col}] color of border around points

\item[\code{point\_border\_stroke}] stroke size of border around points

\item[\code{show\_legend}] show legend

\item[\code{cow\_n\_col}] cowplot param: how many columns

\item[\code{cow\_rel\_h}] cowplot param: relative height

\item[\code{cow\_rel\_w}] cowplot param: relative width

\item[\code{cow\_align}] cowplot param: how to align

\item[\code{show\_plots}] show plots
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Description of parameters.
\end{Details}
%
\begin{Value}
ggplot
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    visGenePlot_2D_ggplot(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{visGenePlot\_3D\_plotly}{visGenePlot\_3D\_plotly}{visGenePlot.Rul.3D.Rul.plotly}
%
\begin{Description}\relax
Visualize cells and gene expression according to spatial coordinates
\end{Description}
%
\begin{Usage}
\begin{verbatim}
visGenePlot_3D_plotly(
  gobject,
  expression_values = c("normalized", "scaled", "custom"),
  genes,
  show_network = F,
  network_color = NULL,
  spatial_network_name = "spatial_network",
  edge_alpha = NULL,
  show_grid = F,
  genes_high_color = NULL,
  genes_mid_color = "white",
  genes_low_color = "blue",
  spatial_grid_name = "spatial_grid",
  point_size = 1,
  show_legend = T,
  axis_scale = c("cube", "real", "custom"),
  custom_ratio = NULL,
  x_ticks = NULL,
  y_ticks = NULL,
  z_ticks = NULL,
  show_plots = F
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{expression\_values}] gene expression values to use

\item[\code{genes}] genes to show

\item[\code{show\_network}] show underlying spatial network

\item[\code{network\_color}] color of spatial network

\item[\code{spatial\_network\_name}] name of spatial network to use

\item[\code{show\_grid}] show spatial grid

\item[\code{genes\_high\_color}] color represents high gene expression

\item[\code{genes\_mid\_color}] color represents middle gene expression

\item[\code{genes\_low\_color}] color represents low gene expression

\item[\code{spatial\_grid\_name}] name of spatial grid to use

\item[\code{point\_size}] size of point (cell)

\item[\code{show\_legend}] show legend

\item[\code{axis\_scale}] three mode to adjust axis scale

\item[\code{x\_ticks}] number of ticks on x axis

\item[\code{y\_ticks}] number of ticks on y axis

\item[\code{z\_ticks}] number of ticks on z axis

\item[\code{show\_plots}] show plots

\item[\code{grid\_color}] color of spatial grid

\item[\code{cow\_n\_col}] cowplot param: how many columns

\item[\code{cow\_rel\_h}] cowplot param: relative height

\item[\code{cow\_rel\_w}] cowplot param: relative width

\item[\code{cow\_align}] cowplot param: how to align
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Description of parameters.
\end{Details}
%
\begin{Value}
plotly
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    visGenePlot_3D_plotly(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{visPlot}{visPlot}{visPlot}
%
\begin{Description}\relax
Visualize cells according to spatial coordinates
\end{Description}
%
\begin{Usage}
\begin{verbatim}
visPlot(
  gobject,
  sdimx = NULL,
  sdimy = NULL,
  sdimz = NULL,
  point_size = 3,
  point_border_col = "black",
  point_border_stroke = 0.1,
  cell_color = NULL,
  cell_color_code = NULL,
  color_as_factor = T,
  select_cell_groups = NULL,
  select_cells = NULL,
  show_other_cells = T,
  other_cell_color = "lightgrey",
  show_network = F,
  network_color = NULL,
  network_alpha = 1,
  other_cell_alpha = 0.1,
  spatial_network_name = "spatial_network",
  show_grid = F,
  grid_color = NULL,
  grid_alpha = 1,
  spatial_grid_name = "spatial_grid",
  coord_fix_ratio = 0.6,
  title = "",
  show_legend = T,
  axis_scale = c("cube", "real", "custom"),
  custom_ratio = NULL,
  x_ticks = NULL,
  y_ticks = NULL,
  z_ticks = NULL,
  plot_method = c("ggplot", "plotly"),
  show_plot = F,
  return_plot = TRUE,
  save_plot = F,
  save_dir = NULL,
  save_folder = NULL,
  save_name = NULL,
  save_format = NULL,
  show_saved_plot = F,
  ...
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{sdimx}] x-axis dimension name (default = 'sdimx')

\item[\code{sdimy}] y-axis dimension name (default = 'sdimy')

\item[\code{sdimz}] z-axis dimension name (default = 'sdimz')

\item[\code{point\_size}] size of point (cell)

\item[\code{point\_border\_col}] color of border around points

\item[\code{point\_border\_stroke}] stroke size of border around points

\item[\code{cell\_color}] color for cells (see details)

\item[\code{cell\_color\_code}] named vector with colors

\item[\code{color\_as\_factor}] convert color column to factor

\item[\code{select\_cell\_groups}] select subset of cells/clusters based on cell\_color parameter

\item[\code{select\_cells}] select subset of cells based on cell IDs

\item[\code{show\_other\_cells}] display not selected cells

\item[\code{other\_cell\_color}] color of not selected cells

\item[\code{show\_network}] show underlying spatial network

\item[\code{network\_color}] color of spatial network

\item[\code{spatial\_network\_name}] name of spatial network to use

\item[\code{show\_grid}] show spatial grid

\item[\code{grid\_color}] color of spatial grid

\item[\code{spatial\_grid\_name}] name of spatial grid to use

\item[\code{coord\_fix\_ratio}] fix ratio between x and y-axis

\item[\code{title}] title of plot

\item[\code{show\_legend}] show legend

\item[\code{show\_plot}] show plot

\item[\code{return\_plot}] return ggplot object

\item[\code{save\_plot}] directly save the plot [boolean]

\item[\code{save\_dir}] directory to save the plot

\item[\code{save\_folder}] (optional) folder in directory to save the plot

\item[\code{save\_name}] name of plot

\item[\code{save\_format}] format of plot (e.g. tiff, png, pdf, ...)

\item[\code{show\_saved\_plot}] load \& display the saved plot
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Description of parameters.
\end{Details}
%
\begin{Value}
ggplot
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    visPlot(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{visPlot\_2D\_ggplot}{visPlot\_2D\_ggplot}{visPlot.Rul.2D.Rul.ggplot}
%
\begin{Description}\relax
Visualize cells according to spatial coordinates
\end{Description}
%
\begin{Usage}
\begin{verbatim}
visPlot_2D_ggplot(
  gobject,
  sdimx = NULL,
  sdimy = NULL,
  point_size = 3,
  point_border_col = "black",
  point_border_stroke = 0.1,
  cell_color = NULL,
  cell_color_code = NULL,
  color_as_factor = T,
  select_cell_groups = NULL,
  select_cells = NULL,
  show_other_cells = T,
  other_cell_color = "lightgrey",
  show_network = F,
  network_color = NULL,
  network_alpha = 1,
  other_cells_alpha = 0.1,
  spatial_network_name = "spatial_network",
  show_grid = F,
  grid_color = NULL,
  spatial_grid_name = "spatial_grid",
  coord_fix_ratio = 0.6,
  title = "",
  show_legend = T,
  axis_scale = c("cube", "real", "custom"),
  custom_ratio = NULL,
  x_ticks = NULL,
  y_ticks = NULL,
  z_ticks = NULL,
  show_plot = F,
  return_plot = TRUE,
  save_plot = F,
  save_dir = NULL,
  save_folder = NULL,
  save_name = NULL,
  save_format = NULL,
  show_saved_plot = F,
  ...
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{sdimx}] x-axis dimension name (default = 'sdimx')

\item[\code{sdimy}] y-axis dimension name (default = 'sdimy')

\item[\code{point\_size}] size of point (cell)

\item[\code{point\_border\_col}] color of border around points

\item[\code{point\_border\_stroke}] stroke size of border around points

\item[\code{cell\_color}] color for cells (see details)

\item[\code{cell\_color\_code}] named vector with colors

\item[\code{color\_as\_factor}] convert color column to factor

\item[\code{select\_cell\_groups}] select subset of cells/clusters based on cell\_color parameter

\item[\code{select\_cells}] select subset of cells based on cell IDs

\item[\code{show\_other\_cells}] display not selected cells

\item[\code{other\_cell\_color}] color of not selected cells

\item[\code{show\_network}] show underlying spatial network

\item[\code{network\_color}] color of spatial network

\item[\code{spatial\_network\_name}] name of spatial network to use

\item[\code{show\_grid}] show spatial grid

\item[\code{grid\_color}] color of spatial grid

\item[\code{spatial\_grid\_name}] name of spatial grid to use

\item[\code{coord\_fix\_ratio}] fix ratio between x and y-axis

\item[\code{title}] title of plot

\item[\code{show\_legend}] show legend

\item[\code{show\_plot}] show plot

\item[\code{return\_plot}] return ggplot object

\item[\code{save\_plot}] directly save the plot [boolean]

\item[\code{save\_dir}] directory to save the plot

\item[\code{save\_folder}] (optional) folder in directory to save the plot

\item[\code{save\_name}] name of plot

\item[\code{save\_format}] format of plot (e.g. tiff, png, pdf, ...)

\item[\code{show\_saved\_plot}] load \& display the saved plot
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Description of parameters.
\end{Details}
%
\begin{Value}
ggplot
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    visPlot_2D_ggplot(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{visPlot\_2D\_plotly}{visPlot\_2D\_plotly}{visPlot.Rul.2D.Rul.plotly}
%
\begin{Description}\relax
Visualize cells according to spatial coordinates
\end{Description}
%
\begin{Usage}
\begin{verbatim}
visPlot_2D_plotly(
  gobject,
  sdimx = NULL,
  sdimy = NULL,
  point_size = 3,
  cell_color = NULL,
  cell_color_code = NULL,
  color_as_factor = T,
  select_cell_groups = NULL,
  select_cells = NULL,
  show_other_cells = T,
  other_cell_color = "lightgrey",
  other_point_size = 0.5,
  show_network = F,
  network_color = "lightgray",
  network_alpha = 1,
  other_cell_alpha = 0.5,
  spatial_network_name = "spatial_network",
  show_grid = F,
  grid_color = NULL,
  grid_alpha = 1,
  spatial_grid_name = "spatial_grid",
  show_legend = T,
  axis_scale = c("cube", "real", "custom"),
  custom_ratio = NULL,
  x_ticks = NULL,
  y_ticks = NULL,
  show_plot = F
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{sdimx}] x-axis dimension name (default = 'sdimx')

\item[\code{sdimy}] y-axis dimension name (default = 'sdimy')

\item[\code{point\_size}] size of point (cell)

\item[\code{cell\_color}] color for cells (see details)

\item[\code{cell\_color\_code}] named vector with colors

\item[\code{color\_as\_factor}] convert color column to factor

\item[\code{select\_cell\_groups}] select a subset of the groups from cell\_color

\item[\code{show\_network}] show underlying spatial network

\item[\code{network\_color}] color of spatial network

\item[\code{spatial\_network\_name}] name of spatial network to use

\item[\code{show\_grid}] show spatial grid

\item[\code{grid\_color}] color of spatial grid

\item[\code{grid\_alpha}] alpha of spatial grid

\item[\code{spatial\_grid\_name}] name of spatial grid to use

\item[\code{show\_legend}] show legend

\item[\code{show\_plot}] show plot
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Description of parameters.
\end{Details}
%
\begin{Value}
plotly
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    visPlot_2D_plotly(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{visPlot\_3D\_plotly}{visPlot\_3D\_plotly}{visPlot.Rul.3D.Rul.plotly}
%
\begin{Description}\relax
Visualize cells according to spatial coordinates
\end{Description}
%
\begin{Usage}
\begin{verbatim}
visPlot_3D_plotly(
  gobject,
  sdimx = NULL,
  sdimy = NULL,
  sdimz = NULL,
  point_size = 3,
  cell_color = NULL,
  cell_color_code = NULL,
  select_cell_groups = NULL,
  select_cells = NULL,
  show_other_cells = T,
  other_cell_color = "lightgrey",
  other_point_size = 0.5,
  show_network = F,
  network_color = NULL,
  network_alpha = 1,
  other_cell_alpha = 0.5,
  spatial_network_name = "spatial_network",
  spatial_grid_name = "spatial_grid",
  title = "",
  show_legend = T,
  axis_scale = c("cube", "real", "custom"),
  custom_ratio = NULL,
  x_ticks = NULL,
  y_ticks = NULL,
  z_ticks = NULL,
  show_plot = F
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{sdimx}] x-axis dimension name (default = 'sdimx')

\item[\code{sdimy}] y-axis dimension name (default = 'sdimy')

\item[\code{sdimz}] z-axis dimension name (default = 'sdimz')

\item[\code{point\_size}] size of point (cell)

\item[\code{cell\_color}] color for cells (see details)

\item[\code{cell\_color\_code}] named vector with colors

\item[\code{select\_cell\_groups}] select a subset of the groups from cell\_color

\item[\code{show\_network}] show underlying spatial network

\item[\code{network\_color}] color of spatial network

\item[\code{spatial\_network\_name}] name of spatial network to use

\item[\code{spatial\_grid\_name}] name of spatial grid to use

\item[\code{title}] title of plot

\item[\code{show\_legend}] show legend

\item[\code{show\_plot}] show plot

\item[\code{point\_border\_col}] color of border around points

\item[\code{point\_border\_stroke}] stroke size of border around points

\item[\code{color\_as\_factor}] convert color column to factor

\item[\code{show\_grid}] show spatial grid

\item[\code{grid\_color}] color of spatial grid

\item[\code{coord\_fix\_ratio}] fix ratio between x and y-axis
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Description of parameters.
\end{Details}
%
\begin{Value}
ggplot
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    visPlot_3D_plotly(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{visSpatDimGenePlot}{visSpatDimGenePlot}{visSpatDimGenePlot}
%
\begin{Description}\relax
integration of visSpatDimGenePlot\_2D(ggplot) and visSpatDimGenePlot\_3D(plotly)
\end{Description}
%
\begin{Usage}
\begin{verbatim}
visSpatDimGenePlot(
  gobject,
  plot_method = c("ggplot", "plotly"),
  expression_values = c("normalized", "scaled", "custom"),
  plot_alignment = c("horizontal", "vertical"),
  dim_reduction_to_use = "umap",
  dim_reduction_name = "umap",
  dim1_to_use = 1,
  dim2_to_use = 2,
  dim3_to_use = NULL,
  sdimx = NULL,
  sdimy = NULL,
  sdimz = NULL,
  genes,
  dim_point_border_col = "black",
  dim_point_border_stroke = 0.1,
  show_NN_network = F,
  nn_network_to_use = "sNN",
  network_name = "sNN.pca",
  edge_alpha_dim = NULL,
  scale_alpha_with_expression = FALSE,
  label_size = 16,
  genes_low_color = "blue",
  genes_mid_color = "white",
  genes_high_color = "red",
  dim_point_size = 3,
  nn_network_alpha = 0.5,
  show_spatial_network = F,
  spatial_network_name = "spatial_network",
  network_color = "lightgray",
  spatial_network_alpha = 0.5,
  show_spatial_grid = F,
  spatial_grid_name = "spatial_grid",
  spatial_grid_color = NULL,
  spatial_grid_alpha = 0.5,
  spatial_point_size = 3,
  spatial_point_border_col = "black",
  spatial_point_border_stroke = 0.1,
  legend_text_size = 12,
  axis_scale = c("cube", "real", "custom"),
  custom_ratio = NULL,
  x_ticks = NULL,
  y_ticks = NULL,
  z_ticks = NULL,
  midpoint = 0,
  point_size = 1,
  cow_n_col = 2,
  cow_rel_h = 1,
  cow_rel_w = 1,
  cow_align = "h",
  show_legend = T,
  show_plots = F
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{expression\_values}] gene expression values to use

\item[\code{plot\_alignment}] direction to align plot

\item[\code{dim\_reduction\_to\_use}] dimension reduction to use

\item[\code{dim\_reduction\_name}] dimension reduction name

\item[\code{dim1\_to\_use}] dimension to use on x-axis

\item[\code{dim2\_to\_use}] dimension to use on y-axis

\item[\code{dim3\_to\_use}] dimension to use on z-axis

\item[\code{sdimx}] x-axis dimension name (default = 'sdimx')

\item[\code{sdimy}] y-axis dimension name (default = 'sdimy')

\item[\code{sdimz}] z-axis dimension name (default = 'sdimz')

\item[\code{genes}] genes to show

\item[\code{dim\_point\_border\_col}] color of border around points

\item[\code{dim\_point\_border\_stroke}] stroke size of border around points

\item[\code{show\_NN\_network}] show underlying NN network

\item[\code{nn\_network\_to\_use}] type of NN network to use (kNN vs sNN)

\item[\code{network\_name}] name of NN network to use, if show\_NN\_network = TRUE

\item[\code{edge\_alpha\_dim}] dim reduction plot: column to use for alpha of the edges

\item[\code{scale\_alpha\_with\_expression}] scale expression with ggplot alpha parameter

\item[\code{label\_size}] size for the label

\item[\code{genes\_low\_color}] color to represent low expression of gene

\item[\code{genes\_high\_color}] color to represent high expression of gene

\item[\code{dim\_point\_size}] dim reduction plot: point size

\item[\code{spatial\_network\_name}] name of spatial network to use

\item[\code{spatial\_grid\_name}] name of spatial grid to use

\item[\code{spatial\_point\_size}] spatial plot: point size

\item[\code{spatial\_point\_border\_col}] color of border around points

\item[\code{spatial\_point\_border\_stroke}] stroke size of border around points

\item[\code{legend\_text\_size}] the size of the text in legend

\item[\code{axis\_scale}] three modes to adjust axis scale ratio

\item[\code{custom\_ratio}] set the axis scale ratio on custom

\item[\code{x\_ticks}] number of ticks on x axis

\item[\code{y\_ticks}] number of ticks on y axis

\item[\code{z\_ticks}] number of ticks on z axis

\item[\code{midpoint}] size of point (cell)

\item[\code{point\_size}] size of point (cell)

\item[\code{cow\_n\_col}] cowplot param: how many columns

\item[\code{cow\_rel\_h}] cowplot param: relative height

\item[\code{cow\_rel\_w}] cowplot param: relative width

\item[\code{cow\_align}] cowplot param: how to align

\item[\code{show\_legend}] show legend

\item[\code{show\_plot}] show plot
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Description of parameters.
\end{Details}
%
\begin{Value}
ggplot or plotly
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    visSpatDimGenePlot(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{visSpatDimGenePlot\_2D}{visSpatDimGenePlot\_2D}{visSpatDimGenePlot.Rul.2D}
%
\begin{Description}\relax
Visualize cells according to spatial AND dimension reduction coordinates in ggplot mode
\end{Description}
%
\begin{Usage}
\begin{verbatim}
visSpatDimGenePlot_2D(
  gobject,
  expression_values = c("normalized", "scaled", "custom"),
  plot_alignment = c("horizontal", "vertical"),
  genes,
  dim_reduction_to_use = "umap",
  dim_reduction_name = "umap",
  dim1_to_use = 1,
  dim2_to_use = 2,
  point_size = 1,
  dim_point_border_col = "black",
  dim_point_border_stroke = 0.1,
  show_NN_network = F,
  show_spatial_network = F,
  show_spatial_grid = F,
  nn_network_to_use = "sNN",
  network_name = "sNN.pca",
  edge_alpha_dim = NULL,
  scale_alpha_with_expression = FALSE,
  spatial_network_name = "spatial_network",
  spatial_grid_name = "spatial_grid",
  spatial_point_size = 1,
  spatial_point_border_col = "black",
  spatial_point_border_stroke = 0.1,
  midpoint = 0,
  genes_high_color = "red",
  genes_mid_color = "white",
  genes_low_color = "blue",
  cow_n_col = 2,
  cow_rel_h = 1,
  cow_rel_w = 1,
  cow_align = "h",
  axis_scale = c("cube", "real", "custom"),
  custom_ratio = NULL,
  x_ticks = NULL,
  y_ticks = NULL,
  show_legend = T,
  show_plots = F
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{expression\_values}] gene expression values to use

\item[\code{plot\_alignment}] direction to align plot

\item[\code{genes}] genes to show

\item[\code{dim\_reduction\_to\_use}] dimension reduction to use

\item[\code{dim\_reduction\_name}] dimension reduction name

\item[\code{dim1\_to\_use}] dimension to use on x-axis

\item[\code{dim2\_to\_use}] dimension to use on y-axis

\item[\code{point\_size}] size of point (cell)

\item[\code{dim\_point\_border\_col}] color of border around points

\item[\code{dim\_point\_border\_stroke}] stroke size of border around points

\item[\code{show\_NN\_network}] show underlying NN network

\item[\code{nn\_network\_to\_use}] type of NN network to use (kNN vs sNN)

\item[\code{network\_name}] name of NN network to use, if show\_NN\_network = TRUE

\item[\code{edge\_alpha\_dim}] dim reduction plot: column to use for alpha of the edges

\item[\code{scale\_alpha\_with\_expression}] scale expression with ggplot alpha parameter

\item[\code{spatial\_network\_name}] name of spatial network to use

\item[\code{spatial\_grid\_name}] name of spatial grid to use

\item[\code{spatial\_point\_size}] spatial plot: point size

\item[\code{spatial\_point\_border\_col}] color of border around points

\item[\code{spatial\_point\_border\_stroke}] stroke size of border around points

\item[\code{midpoint}] size of point (cell)

\item[\code{cow\_n\_col}] cowplot param: how many columns

\item[\code{cow\_rel\_h}] cowplot param: relative height

\item[\code{cow\_rel\_w}] cowplot param: relative width

\item[\code{cow\_align}] cowplot param: how to align

\item[\code{show\_legend}] show legend

\item[\code{dim\_point\_size}] dim reduction plot: point size

\item[\code{show\_plot}] show plot
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Description of parameters.
\end{Details}
%
\begin{Value}
ggplot
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    visSpatDimGenePlot_2D(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{visSpatDimGenePlot\_3D}{visSpatDimGenePlot\_3D}{visSpatDimGenePlot.Rul.3D}
%
\begin{Description}\relax
Visualize cells according to spatial AND dimension reduction coordinates in plotly mode
\end{Description}
%
\begin{Usage}
\begin{verbatim}
visSpatDimGenePlot_3D(
  gobject,
  expression_values = c("normalized", "scaled", "custom"),
  plot_alignment = c("horizontal", "vertical"),
  dim_reduction_to_use = "umap",
  dim_reduction_name = "umap",
  dim1_to_use = 1,
  dim2_to_use = 2,
  dim3_to_use = NULL,
  sdimx = NULL,
  sdimy = NULL,
  sdimz = NULL,
  genes,
  show_NN_network = F,
  nn_network_to_use = "sNN",
  network_name = "sNN.pca",
  label_size = 16,
  genes_low_color = "blue",
  genes_mid_color = "white",
  genes_high_color = "red",
  dim_point_size = 3,
  nn_network_alpha = 0.5,
  show_spatial_network = F,
  spatial_network_name = "spatial_network",
  network_color = "lightgray",
  spatial_network_alpha = 0.5,
  show_spatial_grid = F,
  spatial_grid_name = "spatial_grid",
  spatial_grid_color = NULL,
  spatial_grid_alpha = 0.5,
  spatial_point_size = 3,
  legend_text_size = 12,
  axis_scale = c("cube", "real", "custom"),
  custom_ratio = NULL,
  x_ticks = NULL,
  y_ticks = NULL,
  z_ticks = NULL
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{plot\_alignment}] direction to align plot

\item[\code{dim\_reduction\_to\_use}] dimension reduction to use

\item[\code{dim\_reduction\_name}] dimension reduction name

\item[\code{dim1\_to\_use}] dimension to use on x-axis

\item[\code{dim2\_to\_use}] dimension to use on y-axis

\item[\code{dim3\_to\_use}] dimension to use on z-axis

\item[\code{show\_NN\_network}] show underlying NN network

\item[\code{nn\_network\_to\_use}] type of NN network to use (kNN vs sNN)

\item[\code{network\_name}] name of NN network to use, if show\_NN\_network = TRUE

\item[\code{genes\_low\_color}] color represent high gene expression (see details)

\item[\code{genes\_high\_color}] color represent high gene expression (see details)

\item[\code{nn\_network\_alpha}] column to use for alpha of the edges

\item[\code{show\_spatial\_network}] show spatial network

\item[\code{spatial\_network\_name}] name of spatial network to use

\item[\code{network\_color}] color of spatial/nn network

\item[\code{spatial\_network\_alpha}] alpha of spatial network

\item[\code{show\_spatial\_grid}] show spatial grid

\item[\code{spatial\_grid\_name}] name of spatial grid to use

\item[\code{spatial\_grid\_color}] color of spatial grid

\item[\code{spatial\_grid\_alpha}] alpha of spatial grid

\item[\code{legend\_text\_size}] text size of legend

\item[\code{show\_legend}] show legend

\item[\code{show\_plot}] show plot
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Description of parameters.
\end{Details}
%
\begin{Value}
plotly
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    visSpatDimPlot_3D(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{visSpatDimPlot}{visSpatDimPlot}{visSpatDimPlot}
%
\begin{Description}\relax
integration of visSpatDimPlot\_2D and visSpatDimPlot\_3D
\end{Description}
%
\begin{Usage}
\begin{verbatim}
visSpatDimPlot(
  gobject,
  plot_method = c("ggplot", "plotly"),
  plot_alignment = NULL,
  dim_reduction_to_use = "umap",
  dim_reduction_name = "umap",
  dim1_to_use = 1,
  dim2_to_use = 2,
  dim3_to_use = NULL,
  sdimx = NULL,
  sdimy = NULL,
  sdimz = NULL,
  show_NN_network = F,
  nn_network_to_use = "sNN",
  network_name = "sNN.pca",
  show_cluster_center = F,
  show_center_label = T,
  center_point_size = 4,
  label_size = NULL,
  label_fontface = "bold",
  cell_color = NULL,
  color_as_factor = T,
  cell_color_code = NULL,
  select_cell_groups = NULL,
  select_cells = NULL,
  show_other_cells = T,
  other_cell_color = "lightgrey",
  dim_point_size = 3,
  dim_point_border_col = "black",
  dim_point_border_stroke = 0.1,
  nn_network_alpha = NULL,
  show_spatial_network = F,
  spatial_network_name = "spatial_network",
  network_color = "lightgray",
  spatial_network_alpha = 0.5,
  show_spatial_grid = F,
  spatial_grid_name = "spatial_grid",
  spatial_grid_color = NULL,
  spatial_grid_alpha = 0.5,
  spatial_point_size = 3,
  legend_text_size = 12,
  spatial_point_border_col = "black",
  spatial_point_border_stroke = 0.1,
  show_legend = T,
  axis_scale = c("cube", "real", "custom"),
  custom_ratio = NULL,
  x_ticks = NULL,
  y_ticks = NULL,
  z_ticks = NULL,
  show_plot = F
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{plot\_alignment}] direction to align plot

\item[\code{dim\_reduction\_to\_use}] dimension reduction to use

\item[\code{dim\_reduction\_name}] dimension reduction name

\item[\code{dim1\_to\_use}] dimension to use on x-axis

\item[\code{dim2\_to\_use}] dimension to use on y-axis

\item[\code{dim3\_to\_use}] dimension to use on z-axis

\item[\code{show\_NN\_network}] show underlying NN network

\item[\code{nn\_network\_to\_use}] type of NN network to use (kNN vs sNN)

\item[\code{network\_name}] name of NN network to use, if show\_NN\_network = TRUE

\item[\code{cell\_color}] color for cells (see details)

\item[\code{color\_as\_factor}] convert color column to factor

\item[\code{cell\_color\_code}] named vector with colors

\item[\code{select\_cell\_groups}] select subset of cells/clusters based on cell\_color parameter

\item[\code{select\_cells}] select subset of cells based on cell IDs

\item[\code{show\_other\_cells}] display not selected cells

\item[\code{other\_cell\_color}] color of not selected cells

\item[\code{nn\_network\_alpha}] column to use for alpha of the edges

\item[\code{show\_spatial\_network}] show spatial network

\item[\code{spatial\_network\_name}] name of spatial network to use

\item[\code{spatial\_network\_alpha}] alpha of spatial network

\item[\code{show\_spatial\_grid}] show spatial grid

\item[\code{spatial\_grid\_name}] name of spatial grid to use

\item[\code{spatial\_grid\_color}] color of spatial grid

\item[\code{spatial\_grid\_alpha}] alpha of spatial grid

\item[\code{legend\_text\_size}] text size of legend

\item[\code{show\_legend}] show legend

\item[\code{show\_plot}] show plot

\item[\code{plot\_mode}] choose the mode to draw plot : ggplot or plotly

\item[\code{spatial\_network\_color}] color of spatial network
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Description of parameters.
\end{Details}
%
\begin{Value}
ggplot or plotly
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    visSpatDimPlot(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{visSpatDimPlot\_2D}{visSpatDimPlot\_2D}{visSpatDimPlot.Rul.2D}
%
\begin{Description}\relax
Visualize cells according to spatial AND dimension reduction coordinates in ggplot2 mode
\end{Description}
%
\begin{Usage}
\begin{verbatim}
visSpatDimPlot_2D(
  gobject,
  plot_alignment = c("vertical", "horizontal"),
  dim_reduction_to_use = "umap",
  dim_reduction_name = "umap",
  dim1_to_use = 1,
  dim2_to_use = 2,
  sdimx = NULL,
  sdimy = NULL,
  show_NN_network = F,
  nn_network_to_use = "sNN",
  network_name = "sNN.pca",
  show_cluster_center = F,
  show_center_label = T,
  center_point_size = 4,
  label_size = 4,
  label_fontface = "bold",
  cell_color = NULL,
  color_as_factor = T,
  cell_color_code = NULL,
  select_cell_groups = NULL,
  select_cells = NULL,
  show_other_cells = T,
  other_cell_color = "lightgrey",
  dim_plot_mode = NULL,
  dim_point_size = 1,
  dim_point_border_col = "black",
  dim_point_border_stroke = 0.1,
  nn_network_alpha = 0.05,
  show_spatial_network = F,
  spatial_network_name = "spatial_network",
  spatial_network_color = NULL,
  show_spatial_grid = F,
  spatial_grid_name = "spatial_grid",
  spatial_grid_color = NULL,
  spatial_point_size = 1,
  spatial_point_border_col = "black",
  spatial_point_border_stroke = 0.1,
  show_legend = T,
  show_plot = F,
  plot_method = "ggplot"
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{plot\_alignment}] direction to align plot

\item[\code{dim\_reduction\_to\_use}] dimension reduction to use

\item[\code{dim\_reduction\_name}] dimension reduction name

\item[\code{dim1\_to\_use}] dimension to use on x-axis

\item[\code{dim2\_to\_use}] dimension to use on y-axis

\item[\code{show\_NN\_network}] show underlying NN network

\item[\code{nn\_network\_to\_use}] type of NN network to use (kNN vs sNN)

\item[\code{network\_name}] name of NN network to use, if show\_NN\_network = TRUE

\item[\code{cell\_color}] color for cells (see details)

\item[\code{color\_as\_factor}] convert color column to factor

\item[\code{cell\_color\_code}] named vector with colors

\item[\code{select\_cell\_groups}] select subset of cells/clusters based on cell\_color parameter

\item[\code{select\_cells}] select subset of cells based on cell IDs

\item[\code{show\_other\_cells}] display not selected cells

\item[\code{other\_cell\_color}] color of not selected cells

\item[\code{nn\_network\_alpha}] column to use for alpha of the edges

\item[\code{show\_spatial\_network}] show spatial network

\item[\code{spatial\_network\_name}] name of spatial network to use

\item[\code{spatial\_network\_color}] color of spatial network

\item[\code{show\_spatial\_grid}] show spatial grid

\item[\code{spatial\_grid\_name}] name of spatial grid to use

\item[\code{spatial\_grid\_color}] color of spatial grid

\item[\code{show\_legend}] show legend

\item[\code{show\_plot}] show plot

\item[\code{return\_plot}] return ggplot object

\item[\code{save\_plot}] directly save the plot [boolean]

\item[\code{save\_dir}] directory to save the plot

\item[\code{save\_folder}] (optional) folder in directory to save the plot

\item[\code{save\_name}] name of plot

\item[\code{save\_format}] format of plot (e.g. tiff, png, pdf, ...)

\item[\code{show\_saved\_plot}] load \& display the saved plot
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Description of parameters.
\end{Details}
%
\begin{Value}
ggplot
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    visSpatDimPlot_2D(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{visSpatDimPlot\_3D}{visSpatDimPlot\_3D}{visSpatDimPlot.Rul.3D}
%
\begin{Description}\relax
Visualize cells according to spatial AND dimension reduction coordinates in plotly mode
\end{Description}
%
\begin{Usage}
\begin{verbatim}
visSpatDimPlot_3D(
  gobject,
  plot_alignment = c("horizontal", "vertical"),
  dim_reduction_to_use = "umap",
  dim_reduction_name = "umap",
  dim1_to_use = 1,
  dim2_to_use = 2,
  dim3_to_use = NULL,
  sdimx = NULL,
  sdimy = NULL,
  sdimz = NULL,
  show_NN_network = F,
  nn_network_to_use = "sNN",
  network_name = "sNN.pca",
  show_cluster_center = F,
  show_center_label = T,
  center_point_size = 4,
  label_size = 16,
  cell_color = NULL,
  color_as_factor = T,
  cell_color_code = NULL,
  dim_point_size = 3,
  nn_network_alpha = 0.5,
  show_spatial_network = F,
  spatial_network_name = "spatial_network",
  network_color = "lightgray",
  spatial_network_alpha = 0.5,
  show_spatial_grid = F,
  spatial_grid_name = "spatial_grid",
  spatial_grid_color = NULL,
  spatial_grid_alpha = 0.5,
  spatial_point_size = 3,
  axis_scale = c("cube", "real", "custom"),
  custom_ratio = NULL,
  x_ticks = NULL,
  y_ticks = NULL,
  z_ticks = NULL,
  legend_text_size = 12
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{plot\_alignment}] direction to align plot

\item[\code{dim\_reduction\_to\_use}] dimension reduction to use

\item[\code{dim\_reduction\_name}] dimension reduction name

\item[\code{dim1\_to\_use}] dimension to use on x-axis

\item[\code{dim2\_to\_use}] dimension to use on y-axis

\item[\code{dim3\_to\_use}] dimension to use on z-axis

\item[\code{show\_NN\_network}] show underlying NN network

\item[\code{nn\_network\_to\_use}] type of NN network to use (kNN vs sNN)

\item[\code{network\_name}] name of NN network to use, if show\_NN\_network = TRUE

\item[\code{cell\_color}] color for cells (see details)

\item[\code{color\_as\_factor}] convert color column to factor

\item[\code{cell\_color\_code}] named vector with colors

\item[\code{nn\_network\_alpha}] column to use for alpha of the edges

\item[\code{show\_spatial\_network}] show spatial network

\item[\code{spatial\_network\_name}] name of spatial network to use

\item[\code{spatial\_network\_alpha}] alpha of spatial network

\item[\code{show\_spatial\_grid}] show spatial grid

\item[\code{spatial\_grid\_name}] name of spatial grid to use

\item[\code{spatial\_grid\_color}] color of spatial grid

\item[\code{spatial\_grid\_alpha}] alpha of spatial grid

\item[\code{legend\_text\_size}] text size of legend

\item[\code{spatial\_network\_color}] color of spatial network

\item[\code{show\_legend}] show legend

\item[\code{show\_plot}] show plot
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Description of parameters.
\end{Details}
%
\begin{Value}
plotly
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    visSpatDimPlot_3D(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{writeHMRFresults}{writeHMRFresults}{writeHMRFresults}
%
\begin{Description}\relax
write results from doHMRF to a data.table.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
writeHMRFresults(
  gobject,
  HMRFoutput,
  k = NULL,
  betas_to_view = NULL,
  print_command = F
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{gobject}] giotto object

\item[\code{HMRFoutput}] HMRF output from doHMRF

\item[\code{k}] k to write results for

\item[\code{betas\_to\_view}] results from different betas that you want to view

\item[\code{print\_command}] see the python command
\end{ldescription}
\end{Arguments}
%
\begin{Value}
data.table with HMRF results for each b and the selected k
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
    writeHMRFresults(gobject)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{write\_giotto\_viewer\_annotation}{write\_giotto\_viewer\_annotation}{write.Rul.giotto.Rul.viewer.Rul.annotation}
%
\begin{Description}\relax
write out factor-like annotation data from a giotto object for the Viewer
\end{Description}
%
\begin{Usage}
\begin{verbatim}
write_giotto_viewer_annotation(
  annotation,
  annot_name = "test",
  output_directory = getwd()
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{annotation}] annotation from the data.table from giotto object

\item[\code{annot\_name}] name of the annotation

\item[\code{output\_directory}] directory where to save the files
\end{ldescription}
\end{Arguments}
%
\begin{Value}
write a .txt and .annot file for the selection annotation
\end{Value}
\inputencoding{utf8}
\HeaderA{write\_giotto\_viewer\_dim\_reduction}{write\_giotto\_viewer\_dim\_reduction}{write.Rul.giotto.Rul.viewer.Rul.dim.Rul.reduction}
%
\begin{Description}\relax
write out dimensional reduction data from a giotto object for the Viewer
\end{Description}
%
\begin{Usage}
\begin{verbatim}
write_giotto_viewer_dim_reduction(
  dim_reduction_cell,
  dim_red = NULL,
  dim_red_name = NULL,
  dim_red_rounding = NULL,
  dim_red_rescale = c(-20, 20),
  output_directory = getwd()
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{dim\_reduction\_cell}] dimension reduction slot from giotto object

\item[\code{dim\_red}] high level name of dimension reduction

\item[\code{dim\_red\_name}] specific name of dimension reduction to use

\item[\code{dim\_red\_rounding}] numerical indicating how to round the coordinates

\item[\code{dim\_red\_rescale}] numericals to rescale the coordinates

\item[\code{output\_directory}] directory where to save the files
\end{ldescription}
\end{Arguments}
%
\begin{Value}
write a .txt and .annot file for the selection annotation
\end{Value}
\inputencoding{utf8}
\HeaderA{write\_giotto\_viewer\_numeric\_annotation}{write\_giotto\_viewer\_numeric\_annotation}{write.Rul.giotto.Rul.viewer.Rul.numeric.Rul.annotation}
%
\begin{Description}\relax
write out numeric annotation data from a giotto object for the Viewer
\end{Description}
%
\begin{Usage}
\begin{verbatim}
write_giotto_viewer_numeric_annotation(
  annotation,
  annot_name = "test",
  output_directory = getwd()
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{annotation}] annotation from the data.table from giotto object

\item[\code{annot\_name}] name of the annotation

\item[\code{output\_directory}] directory where to save the files
\end{ldescription}
\end{Arguments}
%
\begin{Value}
write a .txt and .annot file for the selection annotation
\end{Value}
\printindex{}
\end{document}
