<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Spatial single-cell transcriptomics toolbox. • Giotto</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Spatial single-cell transcriptomics toolbox.">
<meta property="og:description" content="Toolbox to process, analyze and visualize spatial single-cell expression data.">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-109153024-2"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-109153024-2');
</script>
</head>
<body>
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">Giotto</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.3.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Start
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/getting_started.html#description">Description</a>
    </li>
    <li>
      <a href="articles/getting_started.html#workflow-diagram">Workflow</a>
    </li>
    <li>
      <a href="articles/getting_started.html#howtos">HOWTOs</a>
    </li>
  </ul>
</li>
<li>
  <a href="reference/index.html">Documentation</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Datasets
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/mouse_seqFISH_cortex_200319.html">mouse seqFISH cortex</a>
    </li>
    <li>
      <a href="articles/mouse_starmap_cortex.html">mouse STARmap cortex</a>
    </li>
    <li>
      <a href="articles/human_cycif_PDAC_200322.html">human CyCIF PDAC</a>
    </li>
    <li>
      <a href="articles/mouse_visium_brain_200325.html">mouse Visium brain</a>
    </li>
    <li>
      <a href="articles/mouse_CODEX_spleen.html">mouse CODEX spleen</a>
    </li>
  </ul>
</li>
<li>
  <a href="articles/faqs.html">FAQ</a>
</li>
<li>
  <a href="news/index.html">News</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/RubD/Giotto">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div id="giotto" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#giotto" class="anchor"></a>Giotto</h1></div>
<!-- badges: start -->
<!-- badges: end -->
<p>The Giotto package consists of two modules, Giotto Analyzer and Viewer (see <a href="http://spatial.rc.fas.harvard.edu/">www.spatialgiotto.com</a>), which provide tools to process, analyze and visualize <strong>single-cell spatial expression</strong> data. The underlying framework is generalizable to virtually all currently available spatial datasets. We recently demonstrated the general applicability on 10 different datasets created by 9 different state-of-the-art spatial technologies, including <em>in situ</em> hybridization (seqFISH+, merFISH, osmFISH), sequencing (Slide-seq, Visium, STARmap) and imaging-based multiplexing/proteomics (CyCIF, MIBI, CODEX). These technologies differ in terms of resolution (single cell vs multiple cells), spatial dimension (2D vs 3D), molecular modality (protein vs RNA), and throughput (number of cells and genes). More information and documentation about the latest (developmental) version of Giotto Analyzer can be found at <a href="https://rubd.github.io/Giotto/" class="uri">https://rubd.github.io/Giotto/</a>.</p>
<p><img src="inst/images/general_figs/overview_datasets.png"></p>
<div id="requirements" class="section level2">
<h2 class="hasAnchor">
<a href="#requirements" class="anchor"></a>Requirements</h2>
<ul>
<li>R (&gt;= 3.5.1)</li>
<li>Python (&gt;= 3.0)</li>
<li>Windows, MacOS, Linux</li>
</ul>
</div>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<p>See <a href="https://rubd.github.io/Giotto/articles/faqs.html">FAQs</a> for additional information.</p>
<div id="r-installation" class="section level4">
<h4 class="hasAnchor">
<a href="#r-installation" class="anchor"></a>R installation</h4>
<p>You can install (~1-5 mins) Giotto with:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" title="1"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(remotes)  <span class="co"># if not installed: install.packages('remotes')</span></a>
<a class="sourceLine" id="cb1-2" title="2"><span class="co"># to install the latest version</span></a>
<a class="sourceLine" id="cb1-3" title="3">remotes<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/remotes/man/install_github.html">install_github</a></span>(<span class="st">"RubD/Giotto"</span>)</a></code></pre></div>
</div>
<div id="python-tools-optional" class="section level4">
<h4 class="hasAnchor">
<a href="#python-tools-optional" class="anchor"></a>Python tools (optional)</h4>
<p>This is necessary to run all available analyses, including Leiden / Louvain clustering and to build and use the interactive visualization tool. An alternative, but less flexible, R version for Louvain clustering is also available. It is advisable to install everything within a specific conda environment and specify the python path at the beginning with createGiottoInstructions() or in the R function itself when required.</p>
<p>Required python modules: pandas / igraph / networkx / leidenalg</p>
<p>pip installation one-liner:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash"><code class="sourceCode bash"><a class="sourceLine" id="cb2-1" title="1"><span class="ex">pip3</span> install pandas python-igraph networkx python-louvain leidenalg</a></code></pre></div>
<p>If pip install does not work, try installing within a <a href="https://docs.conda.io/projects/conda/en/latest/user-guide/tasks/manage-environments.html#creating-an-environment-with-commands">conda environment</a>:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash"><code class="sourceCode bash"><a class="sourceLine" id="cb3-1" title="1"><span class="ex">conda</span> install -c anaconda pandas</a>
<a class="sourceLine" id="cb3-2" title="2"><span class="ex">conda</span> install -c conda-forge python-igraph</a>
<a class="sourceLine" id="cb3-3" title="3"><span class="ex">conda</span> install -c anaconda networkx</a>
<a class="sourceLine" id="cb3-4" title="4"><span class="ex">conda</span> install -c conda-forge python-louvain</a>
<a class="sourceLine" id="cb3-5" title="5"><span class="ex">conda</span> install -c conda-forge leidenalg</a></code></pre></div>
</div>
<div id="hmrf" class="section level4">
<h4 class="hasAnchor">
<a href="#hmrf" class="anchor"></a>HMRF</h4>
<p>See <a href="http://www.nature.com/articles/nbt.4260"><strong>HMRF</strong></a> installation <a href="http://spatial.rc.fas.harvard.edu/install.html">instructions</a>.</p>
</div>
<div id="giotto-viewer" class="section level4">
<h4 class="hasAnchor">
<a href="#giotto-viewer" class="anchor"></a>Giotto Viewer</h4>
<div class="sourceCode" id="cb4"><pre class="sourceCode bash"><code class="sourceCode bash"><a class="sourceLine" id="cb4-1" title="1"><span class="ex">pip3</span> install --user jsbeautifier</a>
<a class="sourceLine" id="cb4-2" title="2"><span class="ex">pip3</span> install --user giotto-viewer --no-cache <span class="co">#add --no-deps if do not wish to upgrade dependency</span></a>
<a class="sourceLine" id="cb4-3" title="3"><span class="ex">pip3</span> install --user smfish-image-processing --no-cache <span class="co">#add --no-deps if do not wish to upgrade dependency</span></a></code></pre></div>
<p> </p>
</div>
</div>
<div id="examples" class="section level2">
<h2 class="hasAnchor">
<a href="#examples" class="anchor"></a>Examples</h2>
<ul>
<li>see <a href="https://github.com/RubD/spatial-datasets" class="uri">https://github.com/RubD/spatial-datasets</a> to find raw and pre-processed input data and Giotto scripts (in progress).</li>
<li>typical run time range for the different datasets on a personal computer is around 10~45 mins.<br>
</li>
<li>click on the image and try them out yourself.</li>
</ul>
<p><a href="https://rubd.github.io/Giotto/articles/mouse_seqFISH_cortex_200319.html"><img src="./inst/images/general_figs/cortex_image_summary.png" alt="seqFISH"></a> <a href="https://rubd.github.io/Giotto/articles/mouse_starmap_cortex.html"><img src="./inst/images/general_figs/starmap_cortex_image_summary.png" alt="STARmap"></a></p>
</div>
<div id="references" class="section level2">
<h2 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h2>
<ul>
<li><p><a href="https://www.biorxiv.org/content/10.1101/701680v1">Dries, R. et al. Giotto, a pipeline for integrative analysis and visualization of single-cell spatial transcriptomic data. bioRxiv 701680 (2019).</a> <a href="doi:10.1101/701680" class="uri">doi:10.1101/701680</a></p></li>
<li><p><a href="https://www.nature.com/articles/s41586-019-1049-y">Eng, C.-H. L. et al. Transcriptome-scale super-resolved imaging in tissues by RNA seqFISH+. Nature 1 (2019).</a> <a href="doi:10.1038/s41586-019-1049-y" class="uri">doi:10.1038/s41586-019-1049-y</a></p></li>
<li><p><a href="https://www.nature.com/articles/nbt.4260">Zhu, Q., Shah, S., Dries, R., Cai, L. &amp; Yuan, G.-C. Identification of spatially associated subpopulations by combining scRNAseq and sequential fluorescence in situ hybridization data. Nature Biotechnology (2018).</a> <a href="doi:10.1038/nbt.4260" class="uri">doi:10.1038/nbt.4260</a></p></li>
</ul>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Browse source code at <br><a href="https://github.com/RubD/Giotto">https://​github.com/​RubD/​Giotto</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>
<a href="https://www.rubendries.com/">Ruben Dries</a> <br><small class="roles"> Author, maintainer </small>  </li>
<li>Huipeng Li <br><small class="roles"> Author </small>  </li>
<li>Qian Zhu <br><small class="roles"> Author </small>  </li>
<li>Rui Dong <br><small class="roles"> Author </small>  </li>
<li>Guo-Cheng Yuan <br><small class="roles"> Author </small>  </li>
</ul>
</div>

  </div>
</div>


      <footer><div class="copyright">
  <p>Developed by <a href="https://www.rubendries.com/">Ruben Dries</a>, Huipeng Li, Qian Zhu, Rui Dong, Guo-Cheng Yuan.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
