<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="Giotto">
<title>mini seqFISH â€¢ Giotto</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/JetBrains_Mono-0.4.7/font.css" rel="stylesheet">
<!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="mini seqFISH">
<meta property="og:description" content="Giotto">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">Giotto</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">4.0.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Documentation</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-tutorials">Tutorials</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-tutorials">
    <h6 class="dropdown-header" data-toc-skip>Seq-Scope</h6>
    <a class="dropdown-item" href="../articles/seqscope_mouse_liver.html">Mouse liver</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Visium</h6>
    <a class="dropdown-item" href="../articles/visium_mouse_kidney.html">Visium Mouse Kidney</a>
    <a class="dropdown-item" href="../articles/visium_mouse_brain.html">Visium Mouse Brain</a>
    <a class="dropdown-item" href="../articles/visium_prostate_integration.html">Visium Prostate Integration</a>
    <a class="dropdown-item" href="../articles/visium_cytassist_lungcancer.html">CytAssist Human Lung Cancer</a>
    <a class="dropdown-item" href="../articles/visium_cytassist_human_glioblastoma.html">CytAssist Multi-omics Human Glioblastoma</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Spatial Genomics</h6>
    <a class="dropdown-item" href="../articles/spatial_genomics_mouse_kidney.html">Mouse Kidney</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Xenium</h6>
    <a class="dropdown-item" href="../articles/xenium_breast_cancer.html">Xenium Human breast cancer</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Nanostring CosMx</h6>
    <a class="dropdown-item" href="../articles/nanostring_cosmx_lung_cancer.html">Nanostring CosMx Subcellular Lung Cancer</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>MERFISH</h6>
    <a class="dropdown-item" href="../articles/merfish_mouse_hypothalamic.html">MERFISH Mouse Hypothalamic Preoptic Region</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Vizgen</h6>
    <a class="dropdown-item" href="../articles/vizgen_mouse_brain.html">Vizgen Mouse Brain Receptor Map</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>seqFISH</h6>
    <a class="dropdown-item" href="../articles/mini_seqfish.html">mini seqFISH</a>
    <a class="dropdown-item" href="../articles/seqfish_cortex.html">SeqFish+ Mouse Cortex Example</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>osmFISH</h6>
    <a class="dropdown-item" href="../articles/osmfish_mouse_ss_cortex.html">osmFISH Mouse SS Cortex</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Resolve Bioscience</h6>
    <a class="dropdown-item" href="../articles/resolve_biosciences.html">Resolve Bioscience Breast Cancer Subcellular</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>CODEX</h6>
    <a class="dropdown-item" href="../articles/codex_mouse_spleen.html">Codex Mouse Spleen</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Single Cell RNA Sequencing</h6>
    <a class="dropdown-item" href="../articles/singlecell_prostate_standard.html">Prostate</a>
    <a class="dropdown-item" href="../articles/singlecell_prostate_integration.html">Prostate Integration</a>
  </div>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-tools">Tools</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-tools">
    <h6 class="dropdown-header" data-toc-skip>Spatial manipulation</h6>
    <a class="dropdown-item" href="../articles/spatial_transformations.html">Spatial transformations</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Image alignment</h6>
    <a class="dropdown-item" href="../articles/image_alignment.html">Image alignment</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Interactive selection</h6>
    <a class="dropdown-item" href="../articles/interactive_selection.html">Interactive selection</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Visualizations</h6>
    <a class="dropdown-item" href="../articles/visualizations.html">Visualizations</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Saving and Loading</h6>
    <a class="dropdown-item" href="../articles/saving_object.html">Saving a Giotto object</a>
    <a class="dropdown-item" href="../articles/saving_plots.html">Saving plots</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Interoperability</h6>
    <a class="dropdown-item" href="../articles/interoperability_anndata.html">AnnData</a>
    <div class="dropdown-divider"></div>
  </div>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-try-giotto-in-the-cloud">Try Giotto in the Cloud</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-try-giotto-in-the-cloud">
    <a class="dropdown-item" href="../articles/terra.html">Terra.bio</a>
    <a class="dropdown-item" href="../articles/docker.html">Docker</a>
    <div class="dropdown-divider"></div>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">News</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/drieslab/Giotto/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>mini seqFISH</h1>
            
            <h4 data-toc-skip class="date">2022-09-16</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/drieslab/Giotto/blob/HEAD/vignettes/mini_seqfish.Rmd" class="external-link"><code>vignettes/mini_seqfish.Rmd</code></a></small>
      <div class="d-none name"><code>mini_seqfish.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="start-giotto">Start Giotto<a class="anchor" aria-label="anchor" href="#start-giotto"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Ensure Giotto Suite is installed.</span></span>
<span><span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="st">"Giotto"</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/installed.packages.html" class="external-link">installed.packages</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"drieslab/Giotto@suite"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co"># Ensure GiottoData, a small, helper module for tutorials, is installed.</span></span>
<span><span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="st">"GiottoData"</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/installed.packages.html" class="external-link">installed.packages</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"drieslab/GiottoData"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://drieslab.github.io/Giotto/">Giotto</a></span><span class="op">)</span></span>
<span><span class="co"># Ensure the Python environment for Giotto has been installed.</span></span>
<span><span class="va">genv_exists</span> <span class="op">=</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/checkGiottoEnvironment.html">checkGiottoEnvironment</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="va">genv_exists</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="co"># The following command need only be run once to install the Giotto environment.</span></span>
<span>  <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/installGiottoEnvironment.html">installGiottoEnvironment</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<div class="section level3">
<h3 id="set-giotto-instructions-optional">Set Giotto instructions (optional)<a class="anchor" aria-label="anchor" href="#set-giotto-instructions-optional"></a>
</h3>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://drieslab.github.io/Giotto/">Giotto</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://drieslab.github.io/GiottoData/">GiottoData</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># to automatically save figures in save_dir set save_plot to TRUE</span></span>
<span><span class="va">temp_dir</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/getwd.html" class="external-link">getwd</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">myinstructions</span> <span class="op">=</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/createGiottoInstructions.html">createGiottoInstructions</a></span><span class="op">(</span>save_dir <span class="op">=</span> <span class="va">temp_dir</span>, </span>
<span>                                          save_plot <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>                                          show_plot <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="create-a-giotto-object">1. Create a Giotto object<a class="anchor" aria-label="anchor" href="#create-a-giotto-object"></a>
</h2>
<p>The minimum requirements are:</p>
<ul>
<li>Matrix with expression information (or path to)</li>
<li>x,y(,z) coordinates for cells or spots (or path to)</li>
</ul>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># download data</span></span>
<span><span class="va">data_directory</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">temp_dir</span>, <span class="st">'/data/'</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/GiottoData/reference/getSpatialDataset.html">getSpatialDataset</a></span><span class="op">(</span>dataset <span class="op">=</span> <span class="st">'mini_seqFISH'</span>, directory <span class="op">=</span> <span class="va">data_directory</span>, method <span class="op">=</span> <span class="st">'wget'</span><span class="op">)</span></span>
<span><span class="co"># giotto object </span></span>
<span><span class="va">expr_path</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">data_directory</span>, <span class="st">"seqfish_field_expr.txt.gz"</span><span class="op">)</span></span>
<span><span class="va">loc_path</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">data_directory</span>, <span class="st">"seqfish_field_locs.txt"</span><span class="op">)</span></span>
<span><span class="va">seqfish_mini</span> <span class="op">=</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/createGiottoObject.html">createGiottoObject</a></span><span class="op">(</span>expression <span class="op">=</span> <span class="va">expr_path</span>, </span>
<span>                                  spatial_locs <span class="op">=</span> <span class="va">loc_path</span>, </span>
<span>                                  instructions <span class="op">=</span> <span class="va">myinstructions</span><span class="op">)</span></span></code></pre></div>
<p>How to work with Giotto instructions that are part of your Giotto
object:</p>
<ul>
<li>Show the instructions associated with your Giotto object with
<strong>showGiottoInstructions()</strong>
</li>
<li>Change one or more instructions with
<strong>changeGiottoInstructions()</strong>
</li>
<li>Replace all instructions at once with
<strong>replaceGiottoInstructions()</strong>
</li>
<li>Read or get a specific Giotto instruction with
<strong>readGiottoInstructions()</strong>
</li>
</ul>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># show instructions associated with giotto object (seqfish_mini)</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/showGiottoInstructions.html">showGiottoInstructions</a></span><span class="op">(</span><span class="va">seqfish_mini</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="processing-steps">2. Processing steps<a class="anchor" aria-label="anchor" href="#processing-steps"></a>
</h2>
<ul>
<li>Filter genes and cells based on detection frequencies</li>
<li>Normalize expression matrix (log transformation, scaling factor
and/or z-scores)</li>
<li>Add cell and gene statistics (optional)</li>
<li>Adjust expression matrix for technical covariates or batches
(optional). These results will be stored in the <em>custom</em>
slot.</li>
</ul>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seqfish_mini</span> <span class="op">=</span> <span class="fu"><a href="../reference/filterGiotto.html">filterGiotto</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">seqfish_mini</span>, </span>
<span>                            expression_threshold <span class="op">=</span> <span class="fl">0.5</span>, </span>
<span>                            feat_det_in_min_cells <span class="op">=</span> <span class="fl">20</span>, </span>
<span>                            min_det_feats_per_cell <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span>
<span><span class="va">seqfish_mini</span> <span class="op">=</span> <span class="fu"><a href="../reference/normalizeGiotto.html">normalizeGiotto</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">seqfish_mini</span>, scalefactor <span class="op">=</span> <span class="fl">6000</span>, verbose <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span><span class="va">seqfish_mini</span> <span class="op">=</span> <span class="fu"><a href="../reference/addStatistics.html">addStatistics</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">seqfish_mini</span><span class="op">)</span></span>
<span><span class="va">seqfish_mini</span> <span class="op">=</span> <span class="fu"><a href="../reference/adjustGiottoMatrix.html">adjustGiottoMatrix</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">seqfish_mini</span>, </span>
<span>                                  expression_values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'normalized'</span><span class="op">)</span>, </span>
<span>                                  covariate_columns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'nr_feats'</span>, <span class="st">'total_expr'</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="dimension-reduction">3. Dimension reduction<a class="anchor" aria-label="anchor" href="#dimension-reduction"></a>
</h2>
<ul>
<li>Identify highly variable features (HVF)</li>
</ul>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seqfish_mini</span> <span class="op">=</span> <span class="fu"><a href="../reference/calculateHVF.html">calculateHVF</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">seqfish_mini</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/mini_seqfish/0-HVFplot.png"></p>
<ul>
<li>Perform PCA</li>
<li>Identify number of significant principal components (PCs)</li>
</ul>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seqfish_mini</span> <span class="op">=</span> <span class="fu"><a href="../reference/runPCA.html">runPCA</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">seqfish_mini</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/screePlot.html">screePlot</a></span><span class="op">(</span><span class="va">seqfish_mini</span>, ncp <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/mini_seqfish/1-screePlot.png"></p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/plotPCA.html">plotPCA</a></span><span class="op">(</span><span class="va">seqfish_mini</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/mini_seqfish/2-PCA.png"></p>
<ul>
<li>Run UMAP and/or t-SNE on PCs (or directly on matrix)</li>
</ul>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seqfish_mini</span> <span class="op">=</span> <span class="fu"><a href="../reference/runUMAP.html">runUMAP</a></span><span class="op">(</span><span class="va">seqfish_mini</span>, dimensions_to_use <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, n_threads <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/plotUMAP.html">plotUMAP</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">seqfish_mini</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/mini_seqfish/3-UMAP.png"></p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seqfish_mini</span> <span class="op">=</span> <span class="fu"><a href="../reference/runtSNE.html">runtSNE</a></span><span class="op">(</span><span class="va">seqfish_mini</span>, dimensions_to_use <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/plotTSNE.html">plotTSNE</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">seqfish_mini</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/mini_seqfish/4-tSNE.png"></p>
</div>
<div class="section level2">
<h2 id="clustering">4. Clustering<a class="anchor" aria-label="anchor" href="#clustering"></a>
</h2>
<ul>
<li>Create a shared (default) nearest network in PCA space (or directly
on matrix)</li>
<li>Cluster on nearest network with Leiden or Louvain (k-means and
hclust are alternatives)</li>
</ul>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seqfish_mini</span> <span class="op">=</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/createNearestNetwork.html">createNearestNetwork</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">seqfish_mini</span>, dimensions_to_use <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, k <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span><span class="va">seqfish_mini</span> <span class="op">=</span> <span class="fu"><a href="../reference/doLeidenCluster.html">doLeidenCluster</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">seqfish_mini</span>, resolution <span class="op">=</span> <span class="fl">0.4</span>, n_iterations <span class="op">=</span> <span class="fl">1000</span><span class="op">)</span></span>
<span><span class="co"># visualize UMAP cluster results</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/plotUMAP.html">plotUMAP</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">seqfish_mini</span>, cell_color <span class="op">=</span> <span class="st">'leiden_clus'</span>, show_NN_network <span class="op">=</span> <span class="cn">T</span>, point_size <span class="op">=</span> <span class="fl">2.5</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/mini_seqfish/5-UMAP.png"></p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># visualize UMAP and spatial results</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/spatDimPlot.html">spatDimPlot</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">seqfish_mini</span>, cell_color <span class="op">=</span> <span class="st">'leiden_clus'</span>, spat_point_shape <span class="op">=</span> <span class="st">'voronoi'</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/mini_seqfish/6-spatDimPlot2D.png"></p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># heatmap and dendrogram</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/showClusterHeatmap.html">showClusterHeatmap</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">seqfish_mini</span>, cluster_column <span class="op">=</span> <span class="st">'leiden_clus'</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/mini_seqfish/7-showClusterHeatmap.png"></p>
<p>The following step requires the installation of {ggdendro}.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># install.packages('ggdendro')</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/andrie/ggdendro" class="external-link">ggdendro</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/showClusterDendrogram.html">showClusterDendrogram</a></span><span class="op">(</span><span class="va">seqfish_mini</span>, h <span class="op">=</span> <span class="fl">0.5</span>, rotate <span class="op">=</span> <span class="cn">T</span>, cluster_column <span class="op">=</span> <span class="st">'leiden_clus'</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/mini_seqfish/8-showClusterDendrogram.png"></p>
</div>
<div class="section level2">
<h2 id="differential-expression">5. Differential expression<a class="anchor" aria-label="anchor" href="#differential-expression"></a>
</h2>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gini_markers</span> <span class="op">=</span> <span class="fu"><a href="../reference/findMarkers_one_vs_all.html">findMarkers_one_vs_all</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">seqfish_mini</span>,</span>
<span>                                      method <span class="op">=</span> <span class="st">'gini'</span>,</span>
<span>                                      expression_values <span class="op">=</span> <span class="st">'normalized'</span>,</span>
<span>                                      cluster_column <span class="op">=</span> <span class="st">'leiden_clus'</span>,</span>
<span>                                      min_feats <span class="op">=</span> <span class="fl">20</span>,</span>
<span>                                      min_expr_gini_score <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>                                      min_det_gini_score <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span>
<span><span class="co"># get top 2 genes per cluster and visualize with violin plot</span></span>
<span><span class="va">topgenes_gini</span> <span class="op">=</span> <span class="va">gini_markers</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">.SD</span>, <span class="fl">2</span><span class="op">)</span>, by <span class="op">=</span> <span class="st">'cluster'</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/violinPlot.html">violinPlot</a></span><span class="op">(</span><span class="va">seqfish_mini</span>, feats <span class="op">=</span> <span class="va">topgenes_gini</span><span class="op">$</span><span class="va">feats</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span>, cluster_column <span class="op">=</span> <span class="st">'leiden_clus'</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/mini_seqfish/9-violinPlot.png"></p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># get top 6 genes per cluster and visualize with heatmap</span></span>
<span><span class="va">topgenes_gini2</span> <span class="op">=</span> <span class="va">gini_markers</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">.SD</span>, <span class="fl">6</span><span class="op">)</span>, by <span class="op">=</span> <span class="st">'cluster'</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/plotMetaDataHeatmap.html">plotMetaDataHeatmap</a></span><span class="op">(</span><span class="va">seqfish_mini</span>, selected_feats <span class="op">=</span> <span class="va">topgenes_gini2</span><span class="op">$</span><span class="va">feats</span>,</span>
<span>                    metadata_cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'leiden_clus'</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/mini_seqfish/10-plotMetaDataHeatmap.png"></p>
</div>
<div class="section level2">
<h2 id="cell-type-annotation">6. Cell type annotation<a class="anchor" aria-label="anchor" href="#cell-type-annotation"></a>
</h2>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">clusters_cell_types</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'cell A'</span>, <span class="st">'cell B'</span>, <span class="st">'cell C'</span>, <span class="st">'cell D'</span>,</span>
<span>                        <span class="st">'cell E'</span>, <span class="st">'cell F'</span>, <span class="st">'cell G'</span>, <span class="st">'cell H'</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">clusters_cell_types</span><span class="op">)</span> <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">8</span></span>
<span><span class="va">seqfish_mini</span> <span class="op">=</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/annotateGiotto.html">annotateGiotto</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">seqfish_mini</span>, </span>
<span>                              annotation_vector <span class="op">=</span> <span class="va">clusters_cell_types</span>, </span>
<span>                              cluster_column <span class="op">=</span> <span class="st">'leiden_clus'</span>, </span>
<span>                              name <span class="op">=</span> <span class="st">'cell_types'</span><span class="op">)</span></span>
<span><span class="co"># check new cell metadata</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/pDataDT.html">pDataDT</a></span><span class="op">(</span><span class="va">seqfish_mini</span><span class="op">)</span></span>
<span><span class="co"># visualize annotations</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/spatDimPlot.html">spatDimPlot</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">seqfish_mini</span>, cell_color <span class="op">=</span> <span class="st">'cell_types'</span>, </span>
<span>            spat_point_size <span class="op">=</span> <span class="fl">3</span>, dim_point_size <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/mini_seqfish/11-spatDimPlot2D.png"></p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># heatmap</span></span>
<span><span class="va">topgenes_heatmap</span> <span class="op">=</span> <span class="va">gini_markers</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">.SD</span>, <span class="fl">4</span><span class="op">)</span>, by <span class="op">=</span> <span class="st">'cluster'</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/plotHeatmap.html">plotHeatmap</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">seqfish_mini</span>,</span>
<span>            feats <span class="op">=</span> <span class="va">topgenes_heatmap</span><span class="op">$</span><span class="va">feats</span>, </span>
<span>            feat_order <span class="op">=</span> <span class="st">'custom'</span>,</span>
<span>            feat_custom_order <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">topgenes_heatmap</span><span class="op">$</span><span class="va">feats</span><span class="op">)</span>,</span>
<span>            cluster_column <span class="op">=</span> <span class="st">'cell_types'</span>,</span>
<span>            legend_nrows <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/mini_seqfish/12-plotHeatmap.png"></p>
</div>
<div class="section level2">
<h2 id="spatial-grid">7. Spatial grid<a class="anchor" aria-label="anchor" href="#spatial-grid"></a>
</h2>
<ul>
<li>Create a grid based on defined step sizes in the x,y(,z) axes.</li>
</ul>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seqfish_mini</span> <span class="op">=</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/createSpatialGrid.html">createSpatialGrid</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">seqfish_mini</span>,</span>
<span>                                 sdimx_stepsize <span class="op">=</span> <span class="fl">300</span>,</span>
<span>                                 sdimy_stepsize <span class="op">=</span> <span class="fl">300</span>,</span>
<span>                                 minimum_padding <span class="op">=</span> <span class="fl">50</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/showGiottoSpatGrids.html">showGiottoSpatGrids</a></span><span class="op">(</span><span class="va">seqfish_mini</span><span class="op">)</span></span>
<span><span class="co"># visualize grid</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/spatPlot.html">spatPlot</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">seqfish_mini</span>, show_grid <span class="op">=</span> <span class="cn">T</span>, point_size <span class="op">=</span> <span class="fl">1.5</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/mini_seqfish/13-spatPlot2D.png"></p>
</div>
<div class="section level2">
<h2 id="spatial-network">8. Spatial network<a class="anchor" aria-label="anchor" href="#spatial-network"></a>
</h2>
<ul>
<li>Visualize information about the default Delaunay network</li>
<li>Create a spatial Delaunay network (default)</li>
<li>Create a spatial kNN network</li>
</ul>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/plotStatDelaunayNetwork.html">plotStatDelaunayNetwork</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">seqfish_mini</span>, maximum_distance <span class="op">=</span> <span class="fl">400</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/mini_seqfish/14-plotStatDelaunayNetwork.png"></p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seqfish_mini</span> <span class="op">=</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/createSpatialNetwork.html">createSpatialNetwork</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">seqfish_mini</span>, minimum_k <span class="op">=</span> <span class="fl">2</span>, </span>
<span>                                    maximum_distance_delaunay <span class="op">=</span> <span class="fl">400</span><span class="op">)</span></span>
<span><span class="va">seqfish_mini</span> <span class="op">=</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/createSpatialNetwork.html">createSpatialNetwork</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">seqfish_mini</span>, minimum_k <span class="op">=</span> <span class="fl">2</span>, </span>
<span>                                    method <span class="op">=</span> <span class="st">'kNN'</span>, k <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/showGiottoSpatNetworks.html">showGiottoSpatNetworks</a></span><span class="op">(</span><span class="va">seqfish_mini</span><span class="op">)</span></span>
<span><span class="co"># visualize the two different spatial networks  </span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/spatPlot.html">spatPlot</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">seqfish_mini</span>, show_network <span class="op">=</span> <span class="cn">T</span>,</span>
<span>         network_color <span class="op">=</span> <span class="st">'blue'</span>, spatial_network_name <span class="op">=</span> <span class="st">'Delaunay_network'</span>,</span>
<span>         point_size <span class="op">=</span> <span class="fl">2.5</span>, cell_color <span class="op">=</span> <span class="st">'leiden_clus'</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/mini_seqfish/15-spatPlot2D.png"></p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/spatPlot.html">spatPlot</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">seqfish_mini</span>, show_network <span class="op">=</span> <span class="cn">T</span>,</span>
<span>         network_color <span class="op">=</span> <span class="st">'blue'</span>, spatial_network_name <span class="op">=</span> <span class="st">'kNN_network'</span>,</span>
<span>         point_size <span class="op">=</span> <span class="fl">2.5</span>, cell_color <span class="op">=</span> <span class="st">'leiden_clus'</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/mini_seqfish/16-spatPlot2D.png"></p>
</div>
<div class="section level2">
<h2 id="spatial-genes">9. Spatial genes<a class="anchor" aria-label="anchor" href="#spatial-genes"></a>
</h2>
<p>Identify spatial genes with 3 different methods:</p>
<ul>
<li>binSpect with k-means binarization (default)</li>
<li>binSpect with rank binarization</li>
<li>silhouetteRank</li>
</ul>
<p>Visualize top 4 genes per method.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">km_spatialgenes</span> <span class="op">=</span> <span class="fu"><a href="../reference/binSpect.html">binSpect</a></span><span class="op">(</span><span class="va">seqfish_mini</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/spatFeatPlot2D.html">spatFeatPlot2D</a></span><span class="op">(</span><span class="va">seqfish_mini</span>, expression_values <span class="op">=</span> <span class="st">'scaled'</span>, </span>
<span>               feats <span class="op">=</span> <span class="va">km_spatialgenes</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span><span class="op">$</span><span class="va">feats</span>,</span>
<span>               point_shape <span class="op">=</span> <span class="st">'border'</span>, point_border_stroke <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>               show_network <span class="op">=</span> <span class="cn">F</span>, network_color <span class="op">=</span> <span class="st">'lightgrey'</span>, point_size <span class="op">=</span> <span class="fl">2.5</span>,</span>
<span>               cow_n_col <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/mini_seqfish/17-spatFeatPlot2D.png"></p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rank_spatialgenes</span> <span class="op">=</span> <span class="fu"><a href="../reference/binSpect.html">binSpect</a></span><span class="op">(</span><span class="va">seqfish_mini</span>, bin_method <span class="op">=</span> <span class="st">'rank'</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/spatFeatPlot2D.html">spatFeatPlot2D</a></span><span class="op">(</span><span class="va">seqfish_mini</span>, expression_values <span class="op">=</span> <span class="st">'scaled'</span>, </span>
<span>               feats <span class="op">=</span> <span class="va">rank_spatialgenes</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span><span class="op">$</span><span class="va">feats</span>,</span>
<span>               point_shape <span class="op">=</span> <span class="st">'border'</span>, point_border_stroke <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>               show_network <span class="op">=</span> <span class="cn">F</span>, network_color <span class="op">=</span> <span class="st">'lightgrey'</span>, point_size <span class="op">=</span> <span class="fl">2.5</span>,</span>
<span>               cow_n_col <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/mini_seqfish/18-spatFeatPlot2D.png"></p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">silh_spatialgenes</span> <span class="op">=</span> <span class="fu"><a href="../reference/silhouetteRank.html">silhouetteRank</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">seqfish_mini</span><span class="op">)</span> <span class="co"># TODO: suppress print output</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/spatFeatPlot2D.html">spatFeatPlot2D</a></span><span class="op">(</span><span class="va">seqfish_mini</span>, expression_values <span class="op">=</span> <span class="st">'scaled'</span>, </span>
<span>               feats <span class="op">=</span> <span class="va">silh_spatialgenes</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span><span class="op">$</span><span class="va">genes</span>,</span>
<span>               point_shape <span class="op">=</span> <span class="st">'border'</span>, point_border_stroke <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>               show_network <span class="op">=</span> <span class="cn">F</span>, network_color <span class="op">=</span> <span class="st">'lightgrey'</span>, point_size <span class="op">=</span> <span class="fl">2.5</span>,</span>
<span>               cow_n_col <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/mini_seqfish/19-spatFeatPlot2D.png"></p>
</div>
<div class="section level2">
<h2 id="spatial-co-expression-patterns">10. Spatial co-expression patterns<a class="anchor" aria-label="anchor" href="#spatial-co-expression-patterns"></a>
</h2>
<p>Identify robust spatial co-expression patterns using the spatial
network or grid and a subset of individual spatial genes.</p>
<ol style="list-style-type: decimal">
<li>Calculate spatial correlation scores</li>
<li>Cluster correlation scores</li>
</ol>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># 1. calculate spatial correlation scores </span></span>
<span><span class="va">ext_spatial_genes</span> <span class="op">=</span> <span class="va">km_spatialgenes</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">500</span><span class="op">]</span><span class="op">$</span><span class="va">feats</span></span>
<span><span class="va">spat_cor_netw_DT</span> <span class="op">=</span> <span class="fu"><a href="../reference/detectSpatialCorFeats.html">detectSpatialCorFeats</a></span><span class="op">(</span><span class="va">seqfish_mini</span>,</span>
<span>                                         method <span class="op">=</span> <span class="st">'network'</span>, </span>
<span>                                         spatial_network_name <span class="op">=</span> <span class="st">'Delaunay_network'</span>,</span>
<span>                                         subset_feats <span class="op">=</span> <span class="va">ext_spatial_genes</span><span class="op">)</span></span>
<span><span class="co"># 2. cluster correlation scores</span></span>
<span><span class="va">spat_cor_netw_DT</span> <span class="op">=</span> <span class="fu"><a href="../reference/clusterSpatialCorFeats.html">clusterSpatialCorFeats</a></span><span class="op">(</span><span class="va">spat_cor_netw_DT</span>, </span>
<span>                                          name <span class="op">=</span> <span class="st">'spat_netw_clus'</span>, k <span class="op">=</span> <span class="fl">8</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/heatmSpatialCorFeats.html">heatmSpatialCorFeats</a></span><span class="op">(</span><span class="va">seqfish_mini</span>, spatCorObject <span class="op">=</span> <span class="va">spat_cor_netw_DT</span>, </span>
<span>                     use_clus_name <span class="op">=</span> <span class="st">'spat_netw_clus'</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/mini_seqfish/20-heatmSpatialCorFeats.png"></p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">netw_ranks</span> <span class="op">=</span> <span class="fu"><a href="../reference/rankSpatialCorGroups.html">rankSpatialCorGroups</a></span><span class="op">(</span><span class="va">seqfish_mini</span>, </span>
<span>                                  spatCorObject <span class="op">=</span> <span class="va">spat_cor_netw_DT</span>, </span>
<span>                                  use_clus_name <span class="op">=</span> <span class="st">'spat_netw_clus'</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/mini_seqfish/21-rankSpatialCorGroups.png"></p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">top_netw_spat_cluster</span> <span class="op">=</span> <span class="fu"><a href="../reference/showSpatialCorFeats.html">showSpatialCorFeats</a></span><span class="op">(</span><span class="va">spat_cor_netw_DT</span>, </span>
<span>                                            use_clus_name <span class="op">=</span> <span class="st">'spat_netw_clus'</span>,</span>
<span>                                            selected_clusters <span class="op">=</span> <span class="fl">6</span>, </span>
<span>                                            show_top_feats <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">cluster_genes_DT</span> <span class="op">=</span> <span class="fu"><a href="../reference/showSpatialCorFeats.html">showSpatialCorFeats</a></span><span class="op">(</span><span class="va">spat_cor_netw_DT</span>, </span>
<span>                                       use_clus_name <span class="op">=</span> <span class="st">'spat_netw_clus'</span>,</span>
<span>                                       show_top_feats <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">cluster_genes</span> <span class="op">=</span> <span class="va">cluster_genes_DT</span><span class="op">$</span><span class="va">clus</span>; <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">cluster_genes</span><span class="op">)</span> <span class="op">=</span> <span class="va">cluster_genes_DT</span><span class="op">$</span><span class="va">feat_ID</span></span>
<span><span class="va">seqfish_mini</span> <span class="op">=</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/createMetafeats.html">createMetafeats</a></span><span class="op">(</span><span class="va">seqfish_mini</span>,</span>
<span>                               feat_clusters <span class="op">=</span> <span class="va">cluster_genes</span>,</span>
<span>                               name <span class="op">=</span> <span class="st">'cluster_metagene'</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/spatCellPlot.html">spatCellPlot</a></span><span class="op">(</span><span class="va">seqfish_mini</span>,</span>
<span>             spat_enr_names <span class="op">=</span> <span class="st">'cluster_metagene'</span>,</span>
<span>             cell_annotation_values <span class="op">=</span> <span class="va">netw_ranks</span><span class="op">$</span><span class="va">clusters</span>,</span>
<span>             point_size <span class="op">=</span> <span class="fl">1.5</span>, cow_n_col <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/mini_seqfish/22-spatCellPlot2D.png"></p>
</div>
<div class="section level2">
<h2 id="spatial-hmrf-domains">11. Spatial HMRF domains<a class="anchor" aria-label="anchor" href="#spatial-hmrf-domains"></a>
</h2>
<p>The following HMRF function requires {smfishHmrf} .</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># remotes::install_bitbucket(repo = 'qzhudfci/smfishhmrf-r', ref='master')</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bitbucket.org/qzhudfci/smfishhmrf-r/src/master/" class="external-link">smfishHmrf</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">hmrf_folder</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">temp_dir</span>,<span class="st">'/'</span>,<span class="st">'11_HMRF/'</span><span class="op">)</span></span>
<span><span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.exists</a></span><span class="op">(</span><span class="va">hmrf_folder</span><span class="op">)</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="va">hmrf_folder</span>, recursive <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span><span class="co"># perform hmrf</span></span>
<span><span class="va">my_spatial_genes</span> <span class="op">=</span> <span class="va">km_spatialgenes</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">100</span><span class="op">]</span><span class="op">$</span><span class="va">feats</span></span>
<span><span class="va">HMRF_spatial_genes</span> <span class="op">=</span> <span class="fu"><a href="../reference/doHMRF.html">doHMRF</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">seqfish_mini</span>,</span>
<span>                            expression_values <span class="op">=</span> <span class="st">'scaled'</span>,</span>
<span>                            spatial_genes <span class="op">=</span> <span class="va">my_spatial_genes</span>,</span>
<span>                            spatial_network_name <span class="op">=</span> <span class="st">'Delaunay_network'</span>,</span>
<span>                            k <span class="op">=</span> <span class="fl">9</span>,</span>
<span>                            betas <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">28</span>,<span class="fl">2</span>,<span class="fl">2</span><span class="op">)</span>,</span>
<span>                            output_folder <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">hmrf_folder</span>, <span class="st">'/'</span>, <span class="st">'Spatial_genes/SG_top100_k9_scaled'</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># check and select hmrf</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">28</span>, <span class="fl">30</span>, by <span class="op">=</span> <span class="fl">2</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="../reference/viewHMRFresults2D.html">viewHMRFresults2D</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">seqfish_mini</span>,</span>
<span>                    HMRFoutput <span class="op">=</span> <span class="va">HMRF_spatial_genes</span>,</span>
<span>                    k <span class="op">=</span> <span class="fl">9</span>, betas_to_view <span class="op">=</span> <span class="va">i</span>,</span>
<span>                    point_size <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="va">seqfish_mini</span> <span class="op">=</span> <span class="fu"><a href="../reference/addHMRF.html">addHMRF</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">seqfish_mini</span>,</span>
<span>                       HMRFoutput <span class="op">=</span> <span class="va">HMRF_spatial_genes</span>,</span>
<span>                       k <span class="op">=</span> <span class="fl">9</span>, betas_to_add <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">28</span><span class="op">)</span>,</span>
<span>                       hmrf_name <span class="op">=</span> <span class="st">'HMRF'</span><span class="op">)</span></span>
<span><span class="co"># visualize selected hmrf result</span></span>
<span><span class="va">giotto_colors</span> <span class="op">=</span> <span class="fu">GiottogetDistinctColors</span><span class="op">(</span><span class="fl">9</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">giotto_colors</span><span class="op">)</span> <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">9</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/spatPlot.html">spatPlot</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">seqfish_mini</span>, cell_color <span class="op">=</span> <span class="st">'HMRF_k9_b.28'</span>,</span>
<span>         point_size <span class="op">=</span> <span class="fl">3</span>, coord_fix_ratio <span class="op">=</span> <span class="fl">1</span>, cell_color_code <span class="op">=</span> <span class="va">giotto_colors</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/mini_seqfish/23-spatPlot2D.png"></p>
</div>
<div class="section level2">
<h2 id="cell-neighborhood-cell-typecell-type-interactions">12. Cell neighborhood: cell-type/cell-type interactions<a class="anchor" aria-label="anchor" href="#cell-neighborhood-cell-typecell-type-interactions"></a>
</h2>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span>seed <span class="op">=</span> <span class="fl">2841</span><span class="op">)</span></span>
<span><span class="va">cell_proximities</span> <span class="op">=</span> <span class="fu"><a href="../reference/cellProximityEnrichment.html">cellProximityEnrichment</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">seqfish_mini</span>,</span>
<span>                                           cluster_column <span class="op">=</span> <span class="st">'cell_types'</span>,</span>
<span>                                           spatial_network_name <span class="op">=</span> <span class="st">'Delaunay_network'</span>,</span>
<span>                                           adjust_method <span class="op">=</span> <span class="st">'fdr'</span>,</span>
<span>                                           number_of_simulations <span class="op">=</span> <span class="fl">1000</span><span class="op">)</span></span>
<span><span class="co"># barplot</span></span>
<span><span class="fu"><a href="../reference/cellProximityBarplot.html">cellProximityBarplot</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">seqfish_mini</span>, </span>
<span>                     CPscore <span class="op">=</span> <span class="va">cell_proximities</span>, </span>
<span>                     min_orig_ints <span class="op">=</span> <span class="fl">5</span>, min_sim_ints <span class="op">=</span> <span class="fl">5</span>, p_val <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/mini_seqfish/24-cellProximityBarplot.png"></p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## heatmap</span></span>
<span><span class="fu"><a href="../reference/cellProximityHeatmap.html">cellProximityHeatmap</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">seqfish_mini</span>, CPscore <span class="op">=</span> <span class="va">cell_proximities</span>, </span>
<span>                     order_cell_types <span class="op">=</span> <span class="cn">T</span>, scale <span class="op">=</span> <span class="cn">T</span>,</span>
<span>                     color_breaks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1.5</span>, <span class="fl">0</span>, <span class="fl">1.5</span><span class="op">)</span>, </span>
<span>                     color_names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'blue'</span>, <span class="st">'white'</span>, <span class="st">'red'</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/mini_seqfish/25-cellProximityHeatmap.png"></p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># network</span></span>
<span><span class="fu"><a href="../reference/cellProximityNetwork.html">cellProximityNetwork</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">seqfish_mini</span>, CPscore <span class="op">=</span> <span class="va">cell_proximities</span>, </span>
<span>                     remove_self_edges <span class="op">=</span> <span class="cn">T</span>, only_show_enrichment_edges <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/mini_seqfish/26-cellProximityNetwork.png"></p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># network with self-edges</span></span>
<span><span class="fu"><a href="../reference/cellProximityNetwork.html">cellProximityNetwork</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">seqfish_mini</span>, CPscore <span class="op">=</span> <span class="va">cell_proximities</span>,</span>
<span>                     remove_self_edges <span class="op">=</span> <span class="cn">F</span>, self_loop_strength <span class="op">=</span> <span class="fl">0.3</span>,</span>
<span>                     only_show_enrichment_edges <span class="op">=</span> <span class="cn">F</span>,</span>
<span>                     rescale_edge_weights <span class="op">=</span> <span class="cn">T</span>,</span>
<span>                     node_size <span class="op">=</span> <span class="fl">8</span>,</span>
<span>                     edge_weight_range_depletion <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span>,</span>
<span>                     edge_weight_range_enrichment <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fl">5</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/mini_seqfish/27-cellProximityNetwork.png"></p>
<div class="section level3">
<h3 id="visualization-of-specific-cell-types">Visualization of specific cell types<a class="anchor" aria-label="anchor" href="#visualization-of-specific-cell-types"></a>
</h3>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Option 1</span></span>
<span><span class="va">spec_interaction</span> <span class="op">=</span> <span class="st">"cell D--cell F"</span></span>
<span><span class="fu"><a href="../reference/cellProximitySpatPlot2D.html">cellProximitySpatPlot2D</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">seqfish_mini</span>,</span>
<span>                        interaction_name <span class="op">=</span> <span class="va">spec_interaction</span>,</span>
<span>                        show_network <span class="op">=</span> <span class="cn">T</span>,</span>
<span>                        cluster_column <span class="op">=</span> <span class="st">'cell_types'</span>,</span>
<span>                        cell_color <span class="op">=</span> <span class="st">'cell_types'</span>,</span>
<span>                        cell_color_code <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'cell D'</span> <span class="op">=</span> <span class="st">'lightblue'</span>, <span class="st">'cell F'</span> <span class="op">=</span> <span class="st">'red'</span><span class="op">)</span>,</span>
<span>                        point_size_select <span class="op">=</span> <span class="fl">4</span>, point_size_other <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/mini_seqfish/28-cellProximitySpatPlot2D.png"></p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Option 2: create additional metadata</span></span>
<span><span class="va">seqfish_mini</span> <span class="op">=</span> <span class="fu"><a href="../reference/addCellIntMetadata.html">addCellIntMetadata</a></span><span class="op">(</span><span class="va">seqfish_mini</span>,</span>
<span>                                  spat_unit <span class="op">=</span> <span class="st">"cell"</span>,</span>
<span>                                  spatial_network <span class="op">=</span> <span class="st">'Delaunay_network'</span>,</span>
<span>                                  cluster_column <span class="op">=</span> <span class="st">'cell_types'</span>,</span>
<span>                                  cell_interaction <span class="op">=</span> <span class="va">spec_interaction</span>,</span>
<span>                                  name <span class="op">=</span> <span class="st">'D_F_interactions'</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/spatPlot.html">spatPlot</a></span><span class="op">(</span><span class="va">seqfish_mini</span>, cell_color <span class="op">=</span> <span class="st">'D_F_interactions'</span>, legend_symbol_size <span class="op">=</span> <span class="fl">3</span>,</span>
<span>         select_cell_groups <span class="op">=</span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'other_cell D'</span>, <span class="st">'other_cell F'</span>, <span class="st">'select_cell D'</span>, <span class="st">'select_cell F'</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/mini_seqfish/29-spatPlot2D.png"></p>
</div>
</div>
<div class="section level2">
<h2 id="cell-neighborhood-interaction-changed-features">13. Cell neighborhood: interaction changed features<a class="anchor" aria-label="anchor" href="#cell-neighborhood-interaction-changed-features"></a>
</h2>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## select top 25 highest expressing genes</span></span>
<span><span class="va">gene_metadata</span> <span class="op">=</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/fDataDT.html">fDataDT</a></span><span class="op">(</span><span class="va">seqfish_mini</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">gene_metadata</span><span class="op">$</span><span class="va">nr_cells</span>, <span class="va">gene_metadata</span><span class="op">$</span><span class="va">mean_expr</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">gene_metadata</span><span class="op">$</span><span class="va">nr_cells</span>, <span class="va">gene_metadata</span><span class="op">$</span><span class="va">mean_expr_det</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/quantile.html" class="external-link">quantile</a></span><span class="op">(</span><span class="va">gene_metadata</span><span class="op">$</span><span class="va">mean_expr_det</span><span class="op">)</span></span>
<span><span class="va">high_expressed_genes</span> <span class="op">=</span> <span class="va">gene_metadata</span><span class="op">[</span><span class="va">mean_expr_det</span> <span class="op">&gt;</span> <span class="fl">4</span><span class="op">]</span><span class="op">$</span><span class="va">feat_ID</span></span>
<span></span>
<span><span class="co">## identify features (genes) that are associated with proximity to other cell types</span></span>
<span><span class="va">ICFscoresHighGenes</span> <span class="op">=</span> <span class="fu"><a href="../reference/findICF.html">findICF</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">seqfish_mini</span>,</span>
<span>                             selected_feats <span class="op">=</span> <span class="va">high_expressed_genes</span>,</span>
<span>                             spatial_network_name <span class="op">=</span> <span class="st">'Delaunay_network'</span>,</span>
<span>                             cluster_column <span class="op">=</span> <span class="st">'cell_types'</span>,</span>
<span>                             diff_test <span class="op">=</span> <span class="st">'permutation'</span>,</span>
<span>                             adjust_method <span class="op">=</span> <span class="st">'fdr'</span>,</span>
<span>                             nr_permutations <span class="op">=</span> <span class="fl">500</span>,</span>
<span>                             do_parallel <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span><span class="co">## visualize all genes</span></span>
<span><span class="fu"><a href="../reference/plotCellProximityFeats.html">plotCellProximityFeats</a></span><span class="op">(</span><span class="va">seqfish_mini</span>, icfObject <span class="op">=</span> <span class="va">ICFscoresHighGenes</span>, method <span class="op">=</span> <span class="st">'dotplot'</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/mini_seqfish/30-plotCellProximityGenes.png"></p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## filter genes</span></span>
<span><span class="va">ICFscoresFilt</span> <span class="op">=</span> <span class="fu"><a href="../reference/filterICF.html">filterICF</a></span><span class="op">(</span><span class="va">ICFscoresHighGenes</span>, min_cells <span class="op">=</span> <span class="fl">2</span>, min_int_cells <span class="op">=</span> <span class="fl">2</span>, min_fdr <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>                          min_spat_diff <span class="op">=</span> <span class="fl">0.1</span>, min_log2_fc <span class="op">=</span> <span class="fl">0.1</span>, min_zscore <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="co">## visualize subset of interaction changed genes (ICGs)</span></span>
<span><span class="va">ICF_genes</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'Cpne2'</span>, <span class="st">'Scg3'</span>, <span class="st">'Cmtm3'</span>, <span class="st">'Cplx1'</span>, <span class="st">'Lingo1'</span><span class="op">)</span></span>
<span><span class="va">ICF_genes_types</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'cell E'</span>, <span class="st">'cell D'</span>, <span class="st">'cell D'</span>, <span class="st">'cell G'</span>, <span class="st">'cell E'</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">ICF_genes</span><span class="op">)</span> <span class="op">=</span> <span class="va">ICF_genes_types</span></span>
<span><span class="fu"><a href="../reference/plotICF.html">plotICF</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">seqfish_mini</span>,</span>
<span>        icfObject <span class="op">=</span> <span class="va">ICFscoresHighGenes</span>,</span>
<span>        source_type <span class="op">=</span> <span class="st">'cell A'</span>,</span>
<span>        source_markers <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'Csf1r'</span>, <span class="st">'Laptm5'</span><span class="op">)</span>,</span>
<span>        ICF_feats <span class="op">=</span> <span class="va">ICF_genes</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/mini_seqfish/31-plotICF.png"></p>
</div>
<div class="section level2">
<h2 id="cell-neighborhood-ligand-receptor-cell-cell-communication">14. Cell neighborhood: ligand-receptor cell-cell communication<a class="anchor" aria-label="anchor" href="#cell-neighborhood-ligand-receptor-cell-cell-communication"></a>
</h2>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">LR_data</span> <span class="op">=</span> <span class="fu">data.table</span><span class="fu">::</span><span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">data_directory</span>, <span class="st">"mouse_ligand_receptors.txt"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">LR_data</span><span class="op">[</span>, <span class="va">ligand_det</span> <span class="op">:=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">mouseLigand</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">seqfish_mini</span><span class="op">@</span><span class="va">feat_ID</span><span class="op">[[</span><span class="st">'rna'</span><span class="op">]</span><span class="op">]</span>, <span class="cn">T</span>, <span class="cn">F</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">LR_data</span><span class="op">[</span>, <span class="va">receptor_det</span> <span class="op">:=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">mouseReceptor</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">seqfish_mini</span><span class="op">@</span><span class="va">feat_ID</span><span class="op">[[</span><span class="st">'rna'</span><span class="op">]</span><span class="op">]</span>, <span class="cn">T</span>, <span class="cn">F</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">LR_data_det</span> <span class="op">=</span> <span class="va">LR_data</span><span class="op">[</span><span class="va">ligand_det</span> <span class="op">==</span> <span class="cn">T</span> <span class="op">&amp;</span> <span class="va">receptor_det</span> <span class="op">==</span> <span class="cn">T</span><span class="op">]</span></span>
<span><span class="va">select_ligands</span> <span class="op">=</span> <span class="va">LR_data_det</span><span class="op">$</span><span class="va">mouseLigand</span></span>
<span><span class="va">select_receptors</span> <span class="op">=</span> <span class="va">LR_data_det</span><span class="op">$</span><span class="va">mouseReceptor</span></span>
<span></span>
<span><span class="co">## get statistical significance of gene pair expression changes based on expression ##</span></span>
<span><span class="va">expr_only_scores</span> <span class="op">=</span> <span class="fu"><a href="../reference/exprCellCellcom.html">exprCellCellcom</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">seqfish_mini</span>,</span>
<span>                                   cluster_column <span class="op">=</span> <span class="st">'cell_types'</span>,</span>
<span>                                   random_iter <span class="op">=</span> <span class="fl">50</span>,</span>
<span>                                   feat_set_1 <span class="op">=</span> <span class="va">select_ligands</span>,</span>
<span>                                   feat_set_2 <span class="op">=</span> <span class="va">select_receptors</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## get statistical significance of gene pair expression changes upon cell-cell interaction</span></span>
<span><span class="va">spatial_all_scores</span> <span class="op">=</span> <span class="fu"><a href="../reference/spatCellCellcom.html">spatCellCellcom</a></span><span class="op">(</span><span class="va">seqfish_mini</span>,</span>
<span>                                     spat_unit <span class="op">=</span> <span class="st">'cell'</span>,</span>
<span>                                     feat_type <span class="op">=</span> <span class="st">'rna'</span>,</span>
<span>                                     spatial_network_name <span class="op">=</span> <span class="st">'Delaunay_network'</span>,</span>
<span>                                     cluster_column <span class="op">=</span> <span class="st">'cell_types'</span>,</span>
<span>                                     random_iter <span class="op">=</span> <span class="fl">50</span>,</span>
<span>                                     feat_set_1 <span class="op">=</span> <span class="va">select_ligands</span>,</span>
<span>                                     feat_set_2 <span class="op">=</span> <span class="va">select_receptors</span>,</span>
<span>                                     adjust_method <span class="op">=</span> <span class="st">'fdr'</span>,</span>
<span>                                     do_parallel <span class="op">=</span> <span class="cn">T</span>,</span>
<span>                                     cores <span class="op">=</span> <span class="fl">4</span>,</span>
<span>                                     verbose <span class="op">=</span> <span class="st">'none'</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## * plot communication scores ####</span></span>
<span><span class="co">## select top LR ##</span></span>
<span><span class="va">selected_spat</span> <span class="op">=</span> <span class="va">spatial_all_scores</span><span class="op">[</span><span class="va">p.adj</span> <span class="op">&lt;=</span> <span class="fl">0.5</span> <span class="op">&amp;</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">log2fc</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0.1</span> <span class="op">&amp;</span> <span class="va">lig_nr</span> <span class="op">&gt;=</span> <span class="fl">2</span> <span class="op">&amp;</span> <span class="va">rec_nr</span> <span class="op">&gt;=</span> <span class="fl">2</span><span class="op">]</span></span>
<span><span class="fu">data.table</span><span class="fu">::</span><span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/setorder.html" class="external-link">setorder</a></span><span class="op">(</span><span class="va">selected_spat</span>, <span class="op">-</span><span class="va">PI</span><span class="op">)</span></span>
<span><span class="va">top_LR_ints</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">selected_spat</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">PI</span><span class="op">)</span><span class="op">)</span><span class="op">]</span><span class="op">$</span><span class="va">LR_comb</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">33</span><span class="op">]</span></span>
<span><span class="va">top_LR_cell_ints</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">selected_spat</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">PI</span><span class="op">)</span><span class="op">)</span><span class="op">]</span><span class="op">$</span><span class="va">LR_cell_comb</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">33</span><span class="op">]</span></span>
<span><span class="fu"><a href="../reference/plotCCcomHeatmap.html">plotCCcomHeatmap</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">seqfish_mini</span>,</span>
<span>                 comScores <span class="op">=</span> <span class="va">spatial_all_scores</span>,</span>
<span>                 selected_LR <span class="op">=</span> <span class="va">top_LR_ints</span>,</span>
<span>                 selected_cell_LR <span class="op">=</span> <span class="va">top_LR_cell_ints</span>,</span>
<span>                 show <span class="op">=</span> <span class="st">'LR_expr'</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/mini_seqfish/32-plotCCcomHeatmap.png"></p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotCCcomDotplot.html">plotCCcomDotplot</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">seqfish_mini</span>,</span>
<span>                 comScores <span class="op">=</span> <span class="va">spatial_all_scores</span>,</span>
<span>                 selected_LR <span class="op">=</span> <span class="va">top_LR_ints</span>,</span>
<span>                 selected_cell_LR <span class="op">=</span> <span class="va">top_LR_cell_ints</span>,</span>
<span>                 cluster_on <span class="op">=</span> <span class="st">'PI'</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/mini_seqfish/33-plotCCcomDotplot.png"></p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## * spatial vs rank ####</span></span>
<span><span class="va">comb_comm</span> <span class="op">=</span> <span class="fu"><a href="../reference/combCCcom.html">combCCcom</a></span><span class="op">(</span>spatialCC <span class="op">=</span> <span class="va">spatial_all_scores</span>,</span>
<span>                      exprCC <span class="op">=</span> <span class="va">expr_only_scores</span><span class="op">)</span></span>
<span><span class="co"># top differential activity levels for ligand receptor pairs</span></span>
<span><span class="fu"><a href="../reference/plotRankSpatvsExpr.html">plotRankSpatvsExpr</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">seqfish_mini</span>,</span>
<span>                   <span class="va">comb_comm</span>,</span>
<span>                   expr_rnk_column <span class="op">=</span> <span class="st">'exprPI_rnk'</span>,</span>
<span>                   spat_rnk_column <span class="op">=</span> <span class="st">'spatPI_rnk'</span>,</span>
<span>                   midpoint <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/mini_seqfish/34-plotRankSpatvsExpr.png"></p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## * recovery ####</span></span>
<span><span class="co">## predict maximum differential activity</span></span>
<span><span class="fu"><a href="../reference/plotRecovery.html">plotRecovery</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">seqfish_mini</span>,</span>
<span>             <span class="va">comb_comm</span>,</span>
<span>             expr_rnk_column <span class="op">=</span> <span class="st">'exprPI_rnk'</span>,</span>
<span>             spat_rnk_column <span class="op">=</span> <span class="st">'spatPI_rnk'</span>,</span>
<span>             ground_truth <span class="op">=</span> <span class="st">'spatial'</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/mini_seqfish/35-plotRecovery.png"></p>
</div>
<div class="section level2">
<h2 id="export-giotto-analyzer-to-viewer">15. Export Giotto Analyzer to Viewer<a class="anchor" aria-label="anchor" href="#export-giotto-analyzer-to-viewer"></a>
</h2>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">viewer_folder</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">temp_dir</span>, <span class="st">'/'</span>, <span class="st">'Mouse_cortex_viewer'</span><span class="op">)</span></span>
<span><span class="co"># select annotations, reductions and expression values to view in Giotto Viewer</span></span>
<span><span class="fu"><a href="../reference/exportGiottoViewer.html">exportGiottoViewer</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">seqfish_mini</span>,</span>
<span>                   output_directory <span class="op">=</span> <span class="va">viewer_folder</span>,</span>
<span>                   factor_annotations <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'cell_types'</span>,</span>
<span>                                          <span class="st">'leiden_clus'</span>,</span>
<span>                                          <span class="st">'HMRF_k9_b.28'</span><span class="op">)</span>,</span>
<span>                   numeric_annotations <span class="op">=</span> <span class="st">'total_expr'</span>,</span>
<span>                   dim_reductions <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'umap'</span><span class="op">)</span>,</span>
<span>                   dim_reduction_names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'umap'</span><span class="op">)</span>,</span>
<span>                   expression_values <span class="op">=</span> <span class="st">'scaled'</span>,</span>
<span>                   expression_rounding <span class="op">=</span> <span class="fl">3</span>,</span>
<span>                   overwrite_dir <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by <a href="https://www.drieslab.com" class="external-link">Ruben Dries</a>, Jiaji George Chen, Natalie Del Rossi, <a href="https://labs.icahn.mssm.edu/yuanlab/" class="external-link">Guo-Cheng Yuan</a>, Matthew Oâ€™Brien, <a href="https://josschavezf.github.io" class="external-link">Joselyn C. ChÃ¡vez-Fuentes</a>, Edward Ruiz.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
