<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="Giotto">
<title>Resolve Bioscience Breast Cancer Subcellular â€¢ Giotto</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/JetBrains_Mono-0.4.7/font.css" rel="stylesheet">
<!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Resolve Bioscience Breast Cancer Subcellular">
<meta property="og:description" content="Giotto">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">Giotto</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">4.0.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-get-started">Get started</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-get-started">
    <a class="dropdown-item" href="../articles/installation.html">Giotto Installation</a>
    <a class="dropdown-item" href="../articles/ecosystem.html">Giotto Ecosystem</a>
    <a class="dropdown-item" href="../articles/configuration.html">Giotto Configuration</a>
    <a class="dropdown-item" href="../articles/object_creation.html">Giotto Object Creation</a>
    <a class="dropdown-item" href="../articles/structure.html">Giotto Class Structure</a>
    <a class="dropdown-item" href="../articles/branches.html">Branches</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Documentation</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-examples">Examples</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-examples">
    <h6 class="dropdown-header" data-toc-skip>Seq-Scope</h6>
    <a class="dropdown-item" href="../articles/seqscope_mouse_liver.html">Mouse liver</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Visium</h6>
    <a class="dropdown-item" href="../articles/visium_mouse_kidney.html">Visium Mouse Kidney</a>
    <a class="dropdown-item" href="../articles/visium_mouse_brain.html">Visium Mouse Brain</a>
    <a class="dropdown-item" href="../articles/visium_prostate_integration.html">Visium Prostate Integration</a>
    <a class="dropdown-item" href="../articles/visium_cytassist_lungcancer.html">CytAssist Human Lung Cancer</a>
    <a class="dropdown-item" href="../articles/visium_cytassist_human_glioblastoma.html">CytAssist Multi-omics Human Glioblastoma</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Spatial Genomics</h6>
    <a class="dropdown-item" href="../articles/spatial_genomics_mouse_kidney.html">Mouse Kidney</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Xenium</h6>
    <a class="dropdown-item" href="../articles/xenium_breast_cancer.html">Xenium Human breast cancer</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Nanostring CosMx</h6>
    <a class="dropdown-item" href="../articles/nanostring_cosmx_lung_cancer.html">Nanostring CosMx Subcellular Lung Cancer</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>MERFISH</h6>
    <a class="dropdown-item" href="../articles/merfish_mouse_hypothalamic.html">MERFISH Mouse Hypothalamic Preoptic Region</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Vizgen</h6>
    <a class="dropdown-item" href="../articles/vizgen_mouse_brain.html">Vizgen Mouse Brain Receptor Map</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>seqFISH</h6>
    <a class="dropdown-item" href="../articles/mini_seqfish.html">mini seqFISH</a>
    <a class="dropdown-item" href="../articles/seqfish_cortex.html">SeqFish+ Mouse Cortex Example</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>osmFISH</h6>
    <a class="dropdown-item" href="../articles/osmfish_mouse_ss_cortex.html">osmFISH Mouse SS Cortex</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Resolve Bioscience</h6>
    <a class="dropdown-item" href="../articles/resolve_biosciences.html">Resolve Bioscience Breast Cancer Subcellular</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>CODEX</h6>
    <a class="dropdown-item" href="../articles/codex_mouse_spleen.html">Codex Mouse Spleen</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Single Cell RNA Sequencing</h6>
    <a class="dropdown-item" href="../articles/singlecell_prostate_standard.html">Prostate</a>
    <a class="dropdown-item" href="../articles/singlecell_prostate_integration.html">Prostate Integration</a>
  </div>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-tutorials">Tutorials</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-tutorials">
    <h6 class="dropdown-header" data-toc-skip>Spatial manipulation</h6>
    <a class="dropdown-item" href="../articles/spatial_transformations.html">Spatial transformations</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Image alignment</h6>
    <a class="dropdown-item" href="../articles/image_alignment.html">Image alignment</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Interactive selection</h6>
    <a class="dropdown-item" href="../articles/interactive_selection.html">Interactive selection</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Visualizations</h6>
    <a class="dropdown-item" href="../articles/visualizations.html">Visualizations</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Saving and Loading</h6>
    <a class="dropdown-item" href="../articles/saving_object.html">Saving a Giotto object</a>
    <a class="dropdown-item" href="../articles/saving_plots.html">Saving plots</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Interoperability</h6>
    <a class="dropdown-item" href="../articles/interoperability_anndata.html">AnnData</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Try Giotto in the Cloud</h6>
    <a class="dropdown-item" href="../articles/terra.html">Terra.bio</a>
    <a class="dropdown-item" href="../articles/docker.html">Docker</a>
  </div>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-news">News</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-news">
    <a class="dropdown-item" href="../news/index.html">Changes</a>
    <a class="dropdown-item" href="../articles/presentations.html">Presentations</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/drieslab/Giotto/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Resolve Bioscience Breast Cancer Subcellular</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/drieslab/Giotto/blob/HEAD/vignettes/resolve_biosciences.Rmd" class="external-link"><code>vignettes/resolve_biosciences.Rmd</code></a></small>
      <div class="d-none name"><code>resolve_biosciences.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Ensure Giotto Suite is installed.</span></span>
<span><span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="st">"Giotto"</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/installed.packages.html" class="external-link">installed.packages</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"drieslab/Giotto@suite"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co"># Ensure GiottoData, a small, helper module for tutorials, is installed.</span></span>
<span><span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="st">"GiottoData"</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/installed.packages.html" class="external-link">installed.packages</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"drieslab/GiottoData"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://drieslab.github.io/Giotto/">Giotto</a></span><span class="op">)</span></span>
<span><span class="co"># Ensure the Python environment for Giotto has been installed.</span></span>
<span><span class="va">genv_exists</span> <span class="op">=</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/checkGiottoEnvironment.html">checkGiottoEnvironment</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="va">genv_exists</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="co"># The following command need only be run once to install the Giotto environment.</span></span>
<span>  <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/installGiottoEnvironment.html">installGiottoEnvironment</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<div class="section level2">
<h2 id="start-giotto">Start Giotto<a class="anchor" aria-label="anchor" href="#start-giotto"></a>
</h2>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://drieslab.github.io/Giotto/">Giotto</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://drieslab.github.io/GiottoData/">GiottoData</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 1. set working directory</span></span>
<span><span class="va">my_working_dir</span> <span class="op">=</span> <span class="st">'/path/to/directory'</span></span>
<span></span>
<span><span class="co"># Optional: Specify a path to a Python executable within a conda or miniconda </span></span>
<span><span class="co"># environment. If set to NULL (default), the Python executable within the previously</span></span>
<span><span class="co"># installed Giotto environment will be used.</span></span>
<span><span class="va">my_python_path</span> <span class="op">=</span> <span class="cn">NULL</span> <span class="co"># alternatively, "/local/python/path/python" if desired.</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="input-files">Input Files<a class="anchor" aria-label="anchor" href="#input-files"></a>
</h2>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## provide path to resolve bioscience folder</span></span>
<span><span class="va">data_path</span> <span class="op">=</span> <span class="st">'/path/to/Resolve_bioscience_cancer'</span></span>
<span></span>
<span></span>
<span><span class="co"># 1. original image as png</span></span>
<span><span class="va">original_DAPI_image</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">data_path</span>, <span class="st">'/'</span>, <span class="st">'sample2_DAPI.jpg'</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 2. input cell segmentation as mask file</span></span>
<span><span class="co"># can also be provided as a 3-column polygon file</span></span>
<span><span class="co"># to be used as image background AND to store segmentations as polygons</span></span>
<span><span class="co"># can be obtained through Fiji / QuPath / Ilastik / Cellpose / ...</span></span>
<span><span class="va">segmentation_mask</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">data_path</span>, <span class="st">'/'</span>, <span class="st">'Mask3.png'</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 3. input features coordinates</span></span>
<span><span class="va">tx_coord</span> <span class="op">=</span> <span class="fu">fread</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">data_path</span>, <span class="st">'/'</span>, <span class="st">'data_sample2.txt'</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/row-plus-colnames-generic.html">colnames</a></span><span class="op">(</span><span class="va">tx_coord</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'x'</span>, <span class="st">'y'</span>, <span class="st">'z'</span>, <span class="st">'gene_id'</span><span class="op">)</span></span>
<span><span class="va">tx_coord</span> <span class="op">=</span> <span class="va">tx_coord</span><span class="op">[</span>,<span class="fu">.</span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span>, <span class="va">gene_id</span><span class="op">)</span><span class="op">]</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="create-subcellular-giotto-object">1. Create Subcellular Giotto Object<a class="anchor" aria-label="anchor" href="#create-subcellular-giotto-object"></a>
</h2>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">testobj</span> <span class="op">=</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/createGiottoObjectSubcellular.html">createGiottoObjectSubcellular</a></span><span class="op">(</span>gpoints <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">'rna'</span> <span class="op">=</span> <span class="va">tx_coord</span><span class="op">)</span>,</span>
<span>                                        gpolygons <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">'cell'</span> <span class="op">=</span> <span class="va">segmentation_mask</span><span class="op">)</span>,</span>
<span>                                        instructions <span class="op">=</span> <span class="va">instrs</span>,</span>
<span>                                        verbose <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                                        cores <span class="op">=</span> <span class="fl">32</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="create-spatial-locations">2. Create Spatial Locations<a class="anchor" aria-label="anchor" href="#create-spatial-locations"></a>
</h2>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># centroids are now used to provide the spatial locations (centroid of each cell)</span></span>
<span><span class="co"># needed for certain downstream spatial analyses</span></span>
<span><span class="va">testobj</span> <span class="op">=</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/addSpatialCentroidLocations.html">addSpatialCentroidLocations</a></span><span class="op">(</span><span class="va">testobj</span>,</span>
<span>                                      poly_info <span class="op">=</span> <span class="st">'cell'</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="add-image-information">3. Add Image Information<a class="anchor" aria-label="anchor" href="#add-image-information"></a>
</h2>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># create Giotto images</span></span>
<span><span class="va">DAPI_image</span> <span class="op">=</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/createGiottoImage.html">createGiottoImage</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">testobj</span>,</span>
<span>                               name <span class="op">=</span> <span class="st">'DAPI'</span>,</span>
<span>                               do_manual_adj <span class="op">=</span> <span class="cn">T</span>,</span>
<span>                               xmax_adj <span class="op">=</span> <span class="fl">0</span>,ymax_adj <span class="op">=</span> <span class="fl">0</span>,</span>
<span>                               xmin_adj <span class="op">=</span> <span class="fl">0</span>,ymin_adj <span class="op">=</span> <span class="fl">0</span>,</span>
<span>                               image_transformations <span class="op">=</span> <span class="st">'flip_x_axis'</span>,</span>
<span>                               mg_object <span class="op">=</span> <span class="va">original_DAPI_image</span><span class="op">)</span></span>
<span></span>
<span><span class="va">segm_image</span> <span class="op">=</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/createGiottoImage.html">createGiottoImage</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">testobj</span>,</span>
<span>                               name <span class="op">=</span> <span class="st">'segmentation'</span>,</span>
<span>                               do_manual_adj <span class="op">=</span> <span class="cn">T</span>,</span>
<span>                               xmax_adj <span class="op">=</span> <span class="fl">0</span>,ymax_adj <span class="op">=</span> <span class="fl">0</span>,</span>
<span>                               xmin_adj <span class="op">=</span> <span class="fl">0</span>,ymin_adj <span class="op">=</span> <span class="fl">0</span>,</span>
<span>                               image_transformations <span class="op">=</span> <span class="st">'flip_x_axis'</span>,</span>
<span>                               mg_object <span class="op">=</span> <span class="va">segmentation_mask</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># add images to Giotto object</span></span>
<span><span class="va">testobj</span> <span class="op">=</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/addGiottoImage.html">addGiottoImage</a></span><span class="op">(</span><span class="va">testobj</span>,</span>
<span>                         images <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">DAPI_image</span>, <span class="va">segm_image</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># provides an overview of available images</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/showGiottoImageNames.html">showGiottoImageNames</a></span><span class="op">(</span><span class="va">testobj</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="visualize-original-images">4. Visualize Original Images<a class="anchor" aria-label="anchor" href="#visualize-original-images"></a>
</h2>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># visualize overlay of calculated cell centroid with original image and segmentation mask file</span></span>
<span><span class="co"># by setting show_plot to FALSE and save_plot to TRUE you can save quite some time when creating plots</span></span>
<span><span class="co"># with big images it sometimes takes quite long for R/Rstudio to render them</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/spatPlot.html">spatPlot2D</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">testobj</span>, </span>
<span>           image_name <span class="op">=</span> <span class="st">'DAPI'</span>, </span>
<span>           point_size <span class="op">=</span> <span class="fl">1.5</span><span class="op">)</span> </span></code></pre></div>
<p><img src="images/resolve_biosciences/0-spatPlot2D.png"></p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/spatPlot.html">spatPlot2D</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">testobj</span>, </span>
<span>           image_name <span class="op">=</span> <span class="st">'segmentation'</span>, </span>
<span>           point_size <span class="op">=</span> <span class="fl">1.5</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/resolve_biosciences/1-spatPlot2D.png"></p>
</div>
<div class="section level2">
<h2 id="calculate-cell-shape-overlap">5. Calculate Cell Shape Overlap<a class="anchor" aria-label="anchor" href="#calculate-cell-shape-overlap"></a>
</h2>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">tictoc</span><span class="fu">::</span><span class="fu">tic</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">testobj</span> <span class="op">=</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/calculateOverlap.html">calculateOverlap</a></span><span class="op">(</span><span class="va">testobj</span>,</span>
<span>                           method <span class="op">=</span> <span class="st">'parallel'</span>,</span>
<span>                           x_step <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>                           y_step <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>                           poly_info <span class="op">=</span> <span class="st">'cell'</span>,</span>
<span>                           feat_info <span class="op">=</span> <span class="st">'rna'</span><span class="op">)</span></span>
<span><span class="fu">tictoc</span><span class="fu">::</span><span class="fu">toc</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#convert overlap to matrix </span></span>
<span><span class="va">testobj</span> <span class="op">=</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/overlapToMatrix.html">overlapToMatrix</a></span><span class="op">(</span><span class="va">testobj</span>,</span>
<span>                          poly_info <span class="op">=</span> <span class="st">'cell'</span>,</span>
<span>                          feat_info <span class="op">=</span> <span class="st">'rna'</span>,</span>
<span>                          name <span class="op">=</span> <span class="st">'raw'</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="filter-data">6. Filter Data<a class="anchor" aria-label="anchor" href="#filter-data"></a>
</h2>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># features can be filtered individually</span></span>
<span><span class="co"># cells will be filtered across features</span></span>
<span></span>
<span><span class="co"># first filter on rna</span></span>
<span><span class="va">subc_test</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/filterGiotto.html">filterGiotto</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">testobj</span>,</span>
<span>                          expression_threshold <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                          feat_det_in_min_cells <span class="op">=</span> <span class="fl">20</span>,</span>
<span>                          min_det_feats_per_cell <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/spatPlot.html">spatPlot2D</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">subc_test</span>,</span>
<span>           image_name <span class="op">=</span> <span class="st">'segmentation'</span>, show_image <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>           point_size <span class="op">=</span> <span class="fl">1.5</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/resolve_biosciences/69-spatPlot2D.png"></p>
</div>
<div class="section level2">
<h2 id="process-giotto-object">7. Process Giotto Object<a class="anchor" aria-label="anchor" href="#process-giotto-object"></a>
</h2>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># rna data, default.</span></span>
<span><span class="co"># other feature modalities can be processed and filtered in an anologous manner</span></span>
<span><span class="va">subc_test</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/normalizeGiotto.html">normalizeGiotto</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">subc_test</span>, </span>
<span>                             scalefactor <span class="op">=</span> <span class="fl">6000</span>, </span>
<span>                             verbose <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span><span class="va">subc_test</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/addStatistics.html">addStatistics</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">subc_test</span><span class="op">)</span></span>
<span><span class="va">subc_test</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/adjustGiottoMatrix.html">adjustGiottoMatrix</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">subc_test</span>,</span>
<span>                                expression_values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'normalized'</span><span class="op">)</span>,</span>
<span>                                covariate_columns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'nr_feats'</span>, <span class="st">'total_expr'</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">subc_test</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/normalizeGiotto.html">normalizeGiotto</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">subc_test</span>, norm_methods <span class="op">=</span> <span class="st">'pearson_resid'</span>, update_slot <span class="op">=</span> <span class="st">'pearson'</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/showGiottoExpression.html">showGiottoExpression</a></span><span class="op">(</span><span class="va">subc_test</span><span class="op">)</span> </span></code></pre></div>
</div>
<div class="section level2">
<h2 id="dimension-reduction">8. Dimension Reduction<a class="anchor" aria-label="anchor" href="#dimension-reduction"></a>
</h2>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Find highly valuable Features</span></span>
<span></span>
<span><span class="co"># typical way of calculating HVF</span></span>
<span><span class="va">subc_test</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateHVF.html">calculateHVF</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">subc_test</span>, HVFname<span class="op">=</span> <span class="st">'hvg_orig'</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/resolve_biosciences/4-HVFplot.png"></p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># new method based on variance of pearson residuals for each gene</span></span>
<span><span class="va">subc_test</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateHVF.html">calculateHVF</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">subc_test</span>,</span>
<span>                          method <span class="op">=</span> <span class="st">'var_p_resid'</span>, </span>
<span>                          expression_values <span class="op">=</span> <span class="st">'pearson'</span>,</span>
<span>                          show_plot <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/resolve_biosciences/5-HVFplot.png"></p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#run PCA</span></span>
<span><span class="va">subc_test</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runPCA.html">runPCA</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">subc_test</span>,</span>
<span>                    expression_values <span class="op">=</span> <span class="st">'pearson'</span>,</span>
<span>                    scale_unit <span class="op">=</span> <span class="cn">F</span>, center <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/screePlot.html">screePlot</a></span><span class="op">(</span><span class="va">subc_test</span>, ncp <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/resolve_biosciences/6-screePlot.png"></p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/plotPCA.html">plotPCA</a></span><span class="op">(</span><span class="va">subc_test</span>,</span>
<span>        dim1_to_use <span class="op">=</span> <span class="fl">1</span>,</span>
<span>        dim2_to_use <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/resolve_biosciences/7-PCA.png"></p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># run UMAP</span></span>
<span><span class="va">subc_test</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runUMAP.html">runUMAP</a></span><span class="op">(</span><span class="va">subc_test</span>, </span>
<span>                     dimensions_to_use <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, </span>
<span>                     n_threads <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/plotUMAP.html">plotUMAP</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">subc_test</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/resolve_biosciences/8-UMAP.png"></p>
</div>
<div class="section level2">
<h2 id="cluster">9. Cluster<a class="anchor" aria-label="anchor" href="#cluster"></a>
</h2>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">subc_test</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/createNearestNetwork.html">createNearestNetwork</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">subc_test</span>, </span>
<span>                                  dimensions_to_use <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, </span>
<span>                                  k <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span><span class="va">subc_test</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/doLeidenCluster.html">doLeidenCluster</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">subc_test</span>, </span>
<span>                             resolution <span class="op">=</span> <span class="fl">0.05</span>, </span>
<span>                             n_iterations <span class="op">=</span> <span class="fl">1000</span>, </span>
<span>                             name <span class="op">=</span> <span class="st">'leiden_0.05'</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Create color palettes, or proceed with Giotto defaults</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"alyssafrazee/RSkittleBrewer"</span><span class="op">)</span></span>
<span><span class="va">colorcode</span> <span class="op">=</span> <span class="fu">lacroix_palette</span><span class="op">(</span>type <span class="op">=</span> <span class="st">"paired"</span><span class="op">)</span></span>
<span><span class="va">featcolor</span> <span class="op">=</span> <span class="fu">lacroix_palette</span><span class="op">(</span><span class="st">"KeyLime"</span>, type <span class="op">=</span> <span class="st">"discrete"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># visualize UMAP cluster results</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/plotUMAP.html">plotUMAP</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">subc_test</span>, </span>
<span>         cell_color <span class="op">=</span> <span class="st">'leiden_clus'</span>,</span>
<span>         show_NN_network <span class="op">=</span> <span class="cn">T</span>, </span>
<span>         point_size <span class="op">=</span> <span class="fl">2.5</span>, </span>
<span>         cell_color_code <span class="op">=</span> <span class="va">colorcode</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/resolve_biosciences/39-UMAP.png"></p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># visualize UMAP and spatial results</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/spatDimPlot.html">spatDimPlot2D</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">subc_test</span>,</span>
<span>              show_image <span class="op">=</span> <span class="cn">T</span>, </span>
<span>              image_name <span class="op">=</span> <span class="st">'segmentation'</span>,</span>
<span>              cell_color <span class="op">=</span> <span class="st">'leiden_clus'</span>,</span>
<span>              spat_point_size <span class="op">=</span> <span class="fl">2</span>, </span>
<span>              cell_color_code <span class="op">=</span> <span class="va">colorcode</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/resolve_biosciences/51-spatDimPlot2D.png"></p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Plot a cluster heatmap</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/showClusterHeatmap.html">showClusterHeatmap</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">subc_test</span>, </span>
<span>                   cluster_column <span class="op">=</span> <span class="st">'leiden_clus'</span>,</span>
<span>                   save_param <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>save_format <span class="op">=</span> <span class="st">'pdf'</span>,base_height <span class="op">=</span> <span class="fl">6</span>, base_width <span class="op">=</span> <span class="fl">8</span>, units <span class="op">=</span> <span class="st">'cm'</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/resolve_biosciences/72-Heatmap.png"></p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># See cluster relationships in a dendogram</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/showClusterDendrogram.html">showClusterDendrogram</a></span><span class="op">(</span><span class="va">subc_test</span>, </span>
<span>                      h <span class="op">=</span> <span class="fl">0.5</span>, </span>
<span>                      rotate <span class="op">=</span> <span class="cn">T</span>, </span>
<span>                      cluster_column <span class="op">=</span> <span class="st">'leiden_clus'</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/resolve_biosciences/12-showClusterDendrogram.png"></p>
</div>
<div class="section level2">
<h2 id="create-a-spatial-network">10. Create a Spatial Network<a class="anchor" aria-label="anchor" href="#create-a-spatial-network"></a>
</h2>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">subc_test</span> <span class="op">=</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/createSpatialNetwork.html">createSpatialNetwork</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">subc_test</span>,</span>
<span>                                 spat_loc_name <span class="op">=</span> <span class="st">'cell'</span>,</span>
<span>                                 minimum_k <span class="op">=</span> <span class="fl">3</span>,</span>
<span>                                 maximum_distance_delaunay <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/spatPlot.html">spatPlot2D</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">subc_test</span>,</span>
<span>           image_name <span class="op">=</span> <span class="st">'segmentation'</span>, </span>
<span>           show_image <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>           point_size <span class="op">=</span> <span class="fl">1.5</span>, </span>
<span>           show_network <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/resolve_biosciences/13-spatPlot2D.png"></p>
</div>
<div class="section level2">
<h2 id="visualize-subcellular-data">11. Visualize SubCellular Data<a class="anchor" aria-label="anchor" href="#visualize-subcellular-data"></a>
</h2>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Visualize clustered cells</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/spatInSituPlotPoints.html">spatInSituPlotPoints</a></span><span class="op">(</span><span class="va">subc_test</span>,</span>
<span>                     show_polygon <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                     polygon_feat_type <span class="op">=</span> <span class="st">'cell'</span>,</span>
<span>                     polygon_color <span class="op">=</span> <span class="st">'white'</span>,</span>
<span>                     polygon_line_size <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>                     polygon_fill <span class="op">=</span> <span class="st">'leiden_clus'</span>,</span>
<span>                     polygon_fill_as_factor <span class="op">=</span> <span class="cn">T</span> , </span>
<span>                     polygon_fill_code <span class="op">=</span> <span class="va">colorcode</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/resolve_biosciences/54-spatInSituPlotPoints.png"></p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># individual plotting of transcripts and polygon information</span></span>
<span></span>
<span><span class="co"># all cells</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/spatInSituPlotPoints.html">spatInSituPlotPoints</a></span><span class="op">(</span><span class="va">testobj</span>,</span>
<span>                     feats <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">'rna'</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"MMP2"</span>, <span class="st">"VEGFA"</span>, <span class="st">"IGF1R"</span>, <span class="st">'CDH2'</span>, <span class="st">'MKI67'</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                     point_size <span class="op">=</span> <span class="fl">0.2</span>,</span>
<span>                     show_polygon <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                     polygon_feat_type <span class="op">=</span> <span class="st">'cell'</span>,</span>
<span>                     polygon_color <span class="op">=</span> <span class="st">'white'</span>,</span>
<span>                     polygon_line_size <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/resolve_biosciences/16-spatInSituPlotPoints.png"></p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># filtered cells</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/spatInSituPlotPoints.html">spatInSituPlotPoints</a></span><span class="op">(</span><span class="va">subc_test</span>,</span>
<span>                     feats <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">'rna'</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"MMP2"</span>, <span class="st">"VEGFA"</span>, <span class="st">"IGF1R"</span>, <span class="st">'CDH2'</span>, <span class="st">'MKI67'</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                     point_size <span class="op">=</span> <span class="fl">0.2</span>,</span>
<span>                     show_polygon <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                     polygon_feat_type <span class="op">=</span> <span class="st">'cell'</span>,</span>
<span>                     polygon_color <span class="op">=</span> <span class="st">'white'</span>,</span>
<span>                     polygon_line_size <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/resolve_biosciences/17-spatInSituPlotPoints.png"></p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># faster plotting method if you have many points</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/spatInSituPlotPoints.html">spatInSituPlotPoints</a></span><span class="op">(</span><span class="va">subc_test</span>,</span>
<span>                     plot_method <span class="op">=</span> <span class="st">'scattermore'</span>,</span>
<span>                     feats <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">'rna'</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"MMP2"</span>, <span class="st">"VEGFA"</span>, <span class="st">"IGF1R"</span>, <span class="st">'CDH2'</span>, <span class="st">'MKI67'</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                     point_size <span class="op">=</span> <span class="fl">0.2</span>,</span>
<span>                     show_polygon <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                     polygon_feat_type <span class="op">=</span> <span class="st">'cell'</span>,</span>
<span>                     polygon_color <span class="op">=</span> <span class="st">'white'</span>,</span>
<span>                     polygon_line_size <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/resolve_biosciences/18-spatInSituPlotPoints.png"></p>
<div class="section level3">
<h3 id="subset-by-location">11.1 Subset by Location<a class="anchor" aria-label="anchor" href="#subset-by-location"></a>
</h3>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># can be used to focus on specific spatial structures</span></span>
<span><span class="co"># to zoom in on niche environments</span></span>
<span></span>
<span><span class="va">subloc</span> <span class="op">=</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/subsetGiottoLocs.html">subsetGiottoLocs</a></span><span class="op">(</span><span class="va">subc_test</span>,</span>
<span>                          x_min <span class="op">=</span> <span class="fl">0</span>, x_max <span class="op">=</span> <span class="fl">2000</span>,</span>
<span>                          y_min <span class="op">=</span> <span class="fl">0</span>, y_max <span class="op">=</span> <span class="fl">2000</span>,</span>
<span>                          poly_info <span class="op">=</span> <span class="st">'cell'</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># show subset of genes</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/spatInSituPlotPoints.html">spatInSituPlotPoints</a></span><span class="op">(</span><span class="va">subloc</span>,</span>
<span>                     feats <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">'rna'</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"MMP2"</span>, <span class="st">"VEGFA"</span>, <span class="st">"IGF1R"</span>, <span class="st">'CDH2'</span>, <span class="st">'MKI67'</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                     point_size <span class="op">=</span> <span class="fl">0.6</span>,</span>
<span>                     show_polygon <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                     polygon_feat_type <span class="op">=</span> <span class="st">'cell'</span>,</span>
<span>                     polygon_color <span class="op">=</span> <span class="st">'white'</span>,</span>
<span>                     polygon_line_size <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/resolve_biosciences/19-spatInSituPlotPoints.png"></p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># show subset of genes and color cells according to clusters</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/spatInSituPlotPoints.html">spatInSituPlotPoints</a></span><span class="op">(</span><span class="va">subloc</span>,</span>
<span>                     feats <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">'rna'</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"MMP2"</span>, <span class="st">"VEGFA"</span>, <span class="st">"IGF1R"</span>, <span class="st">'CDH2'</span>, <span class="st">'MKI67'</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                     point_size <span class="op">=</span> <span class="fl">0.6</span>,</span>
<span>                     show_polygon <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                     polygon_feat_type <span class="op">=</span> <span class="st">'cell'</span>,</span>
<span>                     polygon_color <span class="op">=</span> <span class="st">'white'</span>,</span>
<span>                     polygon_line_size <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>                     polygon_fill <span class="op">=</span> <span class="st">'leiden_clus'</span>,</span>
<span>                     polygon_fill_as_factor <span class="op">=</span> <span class="cn">T</span>, </span>
<span>                     polygon_fill_code <span class="op">=</span> <span class="va">colorcode</span>,</span>
<span>                     feats_color_code <span class="op">=</span> <span class="va">featcolor</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/resolve_biosciences/21-spatInSituPlotPoints.png"></p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># show subset of genes and color cells according to total expression</span></span>
<span><span class="co"># use a faster and more efficient point plotting method = scattermore</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/spatInSituPlotPoints.html">spatInSituPlotPoints</a></span><span class="op">(</span><span class="va">subloc</span>,</span>
<span>                     plot_method <span class="op">=</span> <span class="st">'scattermore'</span>,</span>
<span>                     feats <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">'rna'</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"MMP2"</span>, <span class="st">"VEGFA"</span>, <span class="st">"IGF1R"</span>, <span class="st">'CDH2'</span>, <span class="st">'MKI67'</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                     point_size <span class="op">=</span> <span class="fl">0.6</span>,</span>
<span>                     show_polygon <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                     polygon_feat_type <span class="op">=</span> <span class="st">'cell'</span>,</span>
<span>                     polygon_color <span class="op">=</span> <span class="st">'white'</span>,</span>
<span>                     polygon_line_size <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>                     polygon_fill <span class="op">=</span> <span class="st">'total_expr'</span>,</span>
<span>                     polygon_fill_as_factor <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/resolve_biosciences/61-spatInSituPlotPoints.png"></p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># show cells and color them according to total expression</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/spatInSituPlotPoints.html">spatInSituPlotPoints</a></span><span class="op">(</span><span class="va">subloc</span>,</span>
<span>                     show_polygon <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                     polygon_feat_type <span class="op">=</span> <span class="st">'cell'</span>,</span>
<span>                     polygon_color <span class="op">=</span> <span class="st">'white'</span>,</span>
<span>                     polygon_line_size <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>                     polygon_fill <span class="op">=</span> <span class="st">'total_expr'</span>,</span>
<span>                     polygon_fill_as_factor <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/resolve_biosciences/22-spatInSituPlotPoints.png"></p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># show cells and color them according to total cluster information</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/spatInSituPlotPoints.html">spatInSituPlotPoints</a></span><span class="op">(</span><span class="va">subloc</span>,</span>
<span>                     show_polygon <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                     polygon_feat_type <span class="op">=</span> <span class="st">'cell'</span>,</span>
<span>                     polygon_color <span class="op">=</span> <span class="st">'white'</span>,</span>
<span>                     polygon_line_size <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>                     polygon_fill <span class="op">=</span> <span class="st">'leiden_clus'</span>,</span>
<span>                     polygon_fill_as_factor <span class="op">=</span> <span class="cn">T</span>, </span>
<span>                     polygon_fill_code <span class="op">=</span> <span class="va">colorcode</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/resolve_biosciences/66-spatInSituPlotPoints.png"></p>
</div>
</div>
<div class="section level2">
<h2 id="find-interaction-changed-features">12. Find Interaction Changed Features<a class="anchor" aria-label="anchor" href="#find-interaction-changed-features"></a>
</h2>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># find interaction changed Features</span></span>
<span><span class="co"># In this case, features are genes whose expression difference is associated with a neighboring cell type</span></span>
<span><span class="fu">future</span><span class="fu">::</span><span class="fu"><a href="https://future.futureverse.org/reference/plan.html" class="external-link">plan</a></span><span class="op">(</span><span class="st">'multisession'</span>, workers <span class="op">=</span> <span class="fl">4</span><span class="op">)</span> <span class="co"># sometimes unstable, restart R session</span></span>
<span></span>
<span><span class="va">test</span> <span class="op">=</span> <span class="fu"><a href="../reference/findInteractionChangedFeats.html">findInteractionChangedFeats</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">subc_test</span>,</span>
<span>                                   cluster_column <span class="op">=</span> <span class="st">'leiden_clus'</span><span class="op">)</span></span>
<span></span>
<span><span class="va">test</span><span class="op">$</span><span class="va">ICFscores</span><span class="op">[</span><span class="va">type_int</span> <span class="op">==</span> <span class="st">'hetero'</span><span class="op">]</span></span>
<span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/spatInSituPlotPoints.html">spatInSituPlotPoints</a></span><span class="op">(</span><span class="va">subc_test</span>,</span>
<span>                     feats <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">'rna'</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CTSD"</span>, <span class="st">"BMP1"</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                     point_size <span class="op">=</span> <span class="fl">0.6</span>,</span>
<span>                     show_polygon <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                     polygon_feat_type <span class="op">=</span> <span class="st">'cell'</span>,</span>
<span>                     polygon_color <span class="op">=</span> <span class="st">'black'</span>,</span>
<span>                     polygon_line_size <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>                     polygon_fill <span class="op">=</span> <span class="st">'leiden_clus'</span>,</span>
<span>                     polygon_fill_as_factor <span class="op">=</span> <span class="cn">T</span>,</span>
<span>                     polygon_fill_code <span class="op">=</span> <span class="va">colorcode</span>,</span>
<span>                     feats_color_code <span class="op">=</span> <span class="va">featcolor</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/resolve_biosciences/43-spatInSituPlotPoints.png"></p>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by <a href="https://www.drieslab.com" class="external-link">Ruben Dries</a>, Jiaji George Chen, Natalie Del Rossi, <a href="https://labs.icahn.mssm.edu/yuanlab/" class="external-link">Guo-Cheng Yuan</a>, Matthew Oâ€™Brien, <a href="https://josschavezf.github.io" class="external-link">Joselyn C. ChÃ¡vez-Fuentes</a>, Edward Ruiz.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
