<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="Giotto">
<title>Saving your analysis â€¢ Giotto</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/JetBrains_Mono-0.4.7/font.css" rel="stylesheet">
<!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Saving your analysis">
<meta property="og:description" content="Giotto">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">Giotto</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">4.0.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Documentation</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-tutorials">Tutorials</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-tutorials">
    <h6 class="dropdown-header" data-toc-skip>Seq-Scope</h6>
    <a class="dropdown-item" href="../articles/seqscope_mouse_liver.html">Mouse liver</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Visium</h6>
    <a class="dropdown-item" href="../articles/visium_mouse_kidney.html">Visium Mouse Kidney</a>
    <a class="dropdown-item" href="../articles/visium_mouse_brain.html">Visium Mouse Brain</a>
    <a class="dropdown-item" href="../articles/visium_prostate_integration.html">Visium Prostate Integration</a>
    <a class="dropdown-item" href="../articles/visium_cytassist_lungcancer.html">CytAssist Human Lung Cancer</a>
    <a class="dropdown-item" href="../articles/visium_cytassist_human_glioblastoma.html">CytAssist Multi-omics Human Glioblastoma</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Spatial Genomics</h6>
    <a class="dropdown-item" href="../articles/spatial_genomics_mouse_kidney.html">Mouse Kidney</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Xenium</h6>
    <a class="dropdown-item" href="../articles/xenium_breast_cancer.html">Xenium Human breast cancer</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Nanostring CosMx</h6>
    <a class="dropdown-item" href="../articles/nanostring_cosmx_lung_cancer.html">Nanostring CosMx Subcellular Lung Cancer</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>MERFISH</h6>
    <a class="dropdown-item" href="../articles/merfish_mouse_hypothalamic.html">MERFISH Mouse Hypothalamic Preoptic Region</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Vizgen</h6>
    <a class="dropdown-item" href="../articles/vizgen_mouse_brain.html">Vizgen Mouse Brain Receptor Map</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>seqFISH</h6>
    <a class="dropdown-item" href="../articles/mini_seqfish.html">mini seqFISH</a>
    <a class="dropdown-item" href="../articles/seqfish_cortex.html">SeqFish+ Mouse Cortex Example</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>osmFISH</h6>
    <a class="dropdown-item" href="../articles/osmfish_mouse_ss_cortex.html">osmFISH Mouse SS Cortex</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Resolve Bioscience</h6>
    <a class="dropdown-item" href="../articles/resolve_biosciences.html">Resolve Bioscience Breast Cancer Subcellular</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>CODEX</h6>
    <a class="dropdown-item" href="../articles/codex_mouse_spleen.html">Codex Mouse Spleen</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Single Cell RNA Sequencing</h6>
    <a class="dropdown-item" href="../articles/singlecell_prostate_standard.html">Prostate</a>
    <a class="dropdown-item" href="../articles/singlecell_prostate_integration.html">Prostate Integration</a>
  </div>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-tools">Tools</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-tools">
    <h6 class="dropdown-header" data-toc-skip>Spatial manipulation</h6>
    <a class="dropdown-item" href="../articles/spatial_transformations.html">Spatial transformations</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Image alignment</h6>
    <a class="dropdown-item" href="../articles/image_alignment.html">Image alignment</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Interactive selection</h6>
    <a class="dropdown-item" href="../articles/interactive_selection.html">Interactive selection</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Visualizations</h6>
    <a class="dropdown-item" href="../articles/visualizations.html">Visualizations</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Saving and Loading</h6>
    <a class="dropdown-item" href="../articles/saving_object.html">Saving a Giotto object</a>
    <a class="dropdown-item" href="../articles/saving_plots.html">Saving plots</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Interoperability</h6>
    <a class="dropdown-item" href="../articles/interoperability_anndata.html">AnnData</a>
    <div class="dropdown-divider"></div>
  </div>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-try-giotto-in-the-cloud">Try Giotto in the Cloud</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-try-giotto-in-the-cloud">
    <a class="dropdown-item" href="../articles/terra.html">Terra.bio</a>
    <a class="dropdown-item" href="../articles/docker.html">Docker</a>
    <div class="dropdown-divider"></div>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">News</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/drieslab/Giotto/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Saving your analysis</h1>
            
            <h4 data-toc-skip class="date">4/14/23</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/drieslab/Giotto/blob/HEAD/vignettes/saving_object.Rmd" class="external-link"><code>vignettes/saving_object.Rmd</code></a></small>
      <div class="d-none name"><code>saving_object.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="tutorial-explanation">1. Tutorial explanation<a class="anchor" aria-label="anchor" href="#tutorial-explanation"></a>
</h2>
<p>Giotto objects contain certain objects that act as pointers to the
data that they reference or pull from. While this can help with lowering
the amount of memory needed and decrease load times, pointer objects do
not save and reload properly when using <code><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS()</a></code> and
<code><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS()</a></code>.</p>
<p>Giotto instead implements the <code><a href="https://drieslab.github.io/GiottoClass/reference/saveGiotto.html">saveGiotto()</a></code> and
<code><a href="https://drieslab.github.io/GiottoClass/reference/loadGiotto.html">loadGiotto()</a></code> functions that collects all the data attached
to the object and saves it in a standalone manner. These functions will
be demonstrated using a small subset of the Vizgen mouse brain dataset
that is easily available through the <strong>GiottoData</strong>
package.</p>
</div>
<div class="section level2">
<h2 id="start-giotto">2. Start Giotto<a class="anchor" aria-label="anchor" href="#start-giotto"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Ensure Giotto Suite is installed</span></span>
<span><span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="st">"Giotto"</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/installed.packages.html" class="external-link">installed.packages</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"drieslab/Giotto@suite"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://drieslab.github.io/Giotto/">Giotto</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Ensure Giotto Data is installed</span></span>
<span><span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="st">"GiottoData"</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/installed.packages.html" class="external-link">installed.packages</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"drieslab/GiottoData"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://drieslab.github.io/GiottoData/">GiottoData</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Ensure the Python environment for Giotto has been installed</span></span>
<span><span class="va">genv_exists</span> <span class="op">=</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/checkGiottoEnvironment.html">checkGiottoEnvironment</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="va">genv_exists</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="co"># The following command need only be run once to install the Giotto environment</span></span>
<span>  <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/installGiottoEnvironment.html">installGiottoEnvironment</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="create-a-giotto-object">3. Create a Giotto object<a class="anchor" aria-label="anchor" href="#create-a-giotto-object"></a>
</h2>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vizgen</span> <span class="op">=</span> <span class="fu"><a href="https://drieslab.github.io/GiottoData/reference/loadGiottoMini.html">loadGiottoMini</a></span><span class="op">(</span>dataset <span class="op">=</span> <span class="st">'vizgen'</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a>    [<span class="dv">1</span>] <span class="st">"/Library/Frameworks/R.framework/Versions/4.2/Resources/library/GiottoData//Mini_datasets/Vizgen/VizgenObject//Features/rna_feature_spatVector.shp"</span></span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a>    [<span class="dv">1</span>] <span class="st">"rna"</span></span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a>    [<span class="dv">1</span>] <span class="st">"aggregate_spatInfo_spatVector.shp"</span> <span class="st">"z0_spatInfo_spatVector.shp"</span>       </span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a>    [<span class="dv">3</span>] <span class="st">"z1_spatInfo_spatVector.shp"</span>       </span>
<span id="cb3-5"><a href="#cb3-5" tabindex="-1"></a>    [<span class="dv">1</span>] <span class="st">"aggregate_spatInfo_spatVectorCentroids.shp"</span></span>
<span id="cb3-6"><a href="#cb3-6" tabindex="-1"></a>    [<span class="dv">2</span>] <span class="st">"z0_spatInfo_spatVectorCentroids.shp"</span>       </span>
<span id="cb3-7"><a href="#cb3-7" tabindex="-1"></a>    [<span class="dv">3</span>] <span class="st">"z1_spatInfo_spatVectorCentroids.shp"</span>       </span>
<span id="cb3-8"><a href="#cb3-8" tabindex="-1"></a>    [<span class="dv">1</span>] <span class="st">"rna_aggregate_spatInfo_spatVectorOverlaps.shp"</span></span>
<span id="cb3-9"><a href="#cb3-9" tabindex="-1"></a>    [<span class="dv">2</span>] <span class="st">"rna_z0_spatInfo_spatVectorOverlaps.shp"</span>       </span>
<span id="cb3-10"><a href="#cb3-10" tabindex="-1"></a>    [<span class="dv">3</span>] <span class="st">"rna_z1_spatInfo_spatVectorOverlaps.shp"</span>       </span>
<span id="cb3-11"><a href="#cb3-11" tabindex="-1"></a>    [<span class="dv">1</span>] <span class="st">"poly_ID"</span>      <span class="st">"feat_ID"</span>      <span class="st">"feat_ID_uniq"</span> <span class="st">"stack"</span>       </span>
<span id="cb3-12"><a href="#cb3-12" tabindex="-1"></a>    [<span class="dv">1</span>] <span class="st">"poly_ID"</span>      <span class="st">"feat_ID"</span>      <span class="st">"feat_ID_uniq"</span></span>
<span id="cb3-13"><a href="#cb3-13" tabindex="-1"></a>    [<span class="dv">1</span>] <span class="st">"poly_ID"</span>      <span class="st">"feat_ID"</span>      <span class="st">"feat_ID_uniq"</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="show-existing-info">4. Show Existing Info<a class="anchor" aria-label="anchor" href="#show-existing-info"></a>
</h2>
<p>Many analyses were first performed on this mini object to fill it out
before it was saved using <code><a href="https://drieslab.github.io/GiottoClass/reference/saveGiotto.html">saveGiotto()</a></code>. Returning the
object gives an overview of its contents.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vizgen</span></span></code></pre></div>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a>    An object of class giotto </span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a>    <span class="sc">&gt;</span>Active spat_unit<span class="sc">:</span>  z0 </span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a>    <span class="sc">&gt;</span>Active feat_type<span class="sc">:</span>  rna </span>
<span id="cb5-4"><a href="#cb5-4" tabindex="-1"></a>    [SUBCELLULAR INFO]</span>
<span id="cb5-5"><a href="#cb5-5" tabindex="-1"></a>    polygons      <span class="sc">:</span> z0 z1 aggregate </span>
<span id="cb5-6"><a href="#cb5-6" tabindex="-1"></a>    features      <span class="sc">:</span> rna </span>
<span id="cb5-7"><a href="#cb5-7" tabindex="-1"></a>    [AGGREGATE INFO]</span>
<span id="cb5-8"><a href="#cb5-8" tabindex="-1"></a>    expression <span class="sc">-----------------------</span></span>
<span id="cb5-9"><a href="#cb5-9" tabindex="-1"></a>      [z0][rna] raw</span>
<span id="cb5-10"><a href="#cb5-10" tabindex="-1"></a>      [z1][rna] raw</span>
<span id="cb5-11"><a href="#cb5-11" tabindex="-1"></a>      [aggregate][rna] raw normalized scaled pearson</span>
<span id="cb5-12"><a href="#cb5-12" tabindex="-1"></a>    spatial locations <span class="sc">----------------</span></span>
<span id="cb5-13"><a href="#cb5-13" tabindex="-1"></a>      [z0] raw</span>
<span id="cb5-14"><a href="#cb5-14" tabindex="-1"></a>      [z1] raw</span>
<span id="cb5-15"><a href="#cb5-15" tabindex="-1"></a>      [aggregate] raw</span>
<span id="cb5-16"><a href="#cb5-16" tabindex="-1"></a>    spatial networks <span class="sc">-----------------</span></span>
<span id="cb5-17"><a href="#cb5-17" tabindex="-1"></a>      [aggregate] Delaunay_network kNN_network</span>
<span id="cb5-18"><a href="#cb5-18" tabindex="-1"></a>    dim reduction <span class="sc">--------------------</span></span>
<span id="cb5-19"><a href="#cb5-19" tabindex="-1"></a>      [aggregate][rna] pca umap tsne</span>
<span id="cb5-20"><a href="#cb5-20" tabindex="-1"></a>    nearest neighbor networks <span class="sc">--------</span></span>
<span id="cb5-21"><a href="#cb5-21" tabindex="-1"></a>      [aggregate][rna] sNN.pca</span>
<span id="cb5-22"><a href="#cb5-22" tabindex="-1"></a>    attached images <span class="sc">------------------</span></span>
<span id="cb5-23"><a href="#cb5-23" tabindex="-1"></a>    giottoLargeImage <span class="sc">:</span> <span class="dv">4</span> items...</span>
<span id="cb5-24"><a href="#cb5-24" tabindex="-1"></a></span>
<span id="cb5-25"><a href="#cb5-25" tabindex="-1"></a>    Use <span class="fu">objHistory</span>() to see steps and params used</span></code></pre></div>
<p>The object also functions as expected. Here we first set the active
spatial unit as â€˜aggregateâ€™ and then plot the saved leiden clustering
results.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/activeSpatUnit-generic.html">activeSpatUnit</a></span><span class="op">(</span><span class="va">vizgen</span><span class="op">)</span> <span class="op">=</span> <span class="st">'aggregate'</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/spatPlot.html">spatPlot2D</a></span><span class="op">(</span><span class="va">vizgen</span>,</span>
<span>           show_image <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>           cell_color <span class="op">=</span> <span class="st">'leiden_clus'</span>,</span>
<span>           largeImage_name <span class="op">=</span> <span class="st">'dapi_z0'</span>,</span>
<span>           point_alpha <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/saving_object/unnamed-chunk-4-1.png"></p>
</div>
<div class="section level2">
<h2 id="saving-the-object">5. Saving the Object<a class="anchor" aria-label="anchor" href="#saving-the-object"></a>
</h2>
<p>To save the object, most important inputs are the object to be saved,
a directory in which to place the saved object, and the actual folder
within that directory to contain Giottoâ€™s files.</p>
<p>The dir param defaults to the working directory, but here I am
specifying for the object to be saved to a temporary directory.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/saveGiotto.html">saveGiotto</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">vizgen</span>,</span>
<span>           dir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>           foldername <span class="op">=</span> <span class="st">'vizgen'</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="faster-saving-with-qs">6. Faster Saving with <code>qs</code><a class="anchor" aria-label="anchor" href="#faster-saving-with-qs"></a>
</h2>
<p><code><a href="https://drieslab.github.io/GiottoClass/reference/saveGiotto.html">saveGiotto()</a></code> defaults to using <code><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS()</a></code>
for its main files, but <code>qs</code> implements faster read and write
functions for serialization. Saving using <code>qs</code> is done by
passing the param <code>method = 'qs'</code>.</p>
<p>Note that <code>qs</code> is an additional package that is not
installed by default with Giotto but it can be installed with
<code>install.packages(qs)</code>.</p>
</div>
<div class="section level2">
<h2 id="loading-the-object">7. Loading the Object<a class="anchor" aria-label="anchor" href="#loading-the-object"></a>
</h2>
<p>To load the object, only the path to the folder is needed. Additional
params of interest are:</p>
<ul>
<li>
<code>reconnect_giottoImage</code> -instead of reading from the
saved images in the folder- attempts to reconnect the pointer objects
based on filepaths that were recorded. This requires that the Giotto
object is being loaded on the same system it was generated on and that
those images have not been moved.</li>
<li>
<code>python_path</code> allows a new python path to be set during
loading.</li>
</ul>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">test_vizgen</span> <span class="op">=</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/loadGiotto.html">loadGiotto</a></span><span class="op">(</span>path_to_folder <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">'vizgen'</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">test_vizgen</span></span></code></pre></div>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a>    An object of class giotto </span>
<span id="cb10-2"><a href="#cb10-2" tabindex="-1"></a>    <span class="sc">&gt;</span>Active spat_unit<span class="sc">:</span>  aggregate </span>
<span id="cb10-3"><a href="#cb10-3" tabindex="-1"></a>    <span class="sc">&gt;</span>Active feat_type<span class="sc">:</span>  rna </span>
<span id="cb10-4"><a href="#cb10-4" tabindex="-1"></a>    [SUBCELLULAR INFO]</span>
<span id="cb10-5"><a href="#cb10-5" tabindex="-1"></a>    polygons      <span class="sc">:</span> z0 z1 aggregate </span>
<span id="cb10-6"><a href="#cb10-6" tabindex="-1"></a>    features      <span class="sc">:</span> rna </span>
<span id="cb10-7"><a href="#cb10-7" tabindex="-1"></a>    [AGGREGATE INFO]</span>
<span id="cb10-8"><a href="#cb10-8" tabindex="-1"></a>    expression <span class="sc">-----------------------</span></span>
<span id="cb10-9"><a href="#cb10-9" tabindex="-1"></a>      [z0][rna] raw</span>
<span id="cb10-10"><a href="#cb10-10" tabindex="-1"></a>      [z1][rna] raw</span>
<span id="cb10-11"><a href="#cb10-11" tabindex="-1"></a>      [aggregate][rna] raw normalized scaled pearson</span>
<span id="cb10-12"><a href="#cb10-12" tabindex="-1"></a>    spatial locations <span class="sc">----------------</span></span>
<span id="cb10-13"><a href="#cb10-13" tabindex="-1"></a>      [z0] raw</span>
<span id="cb10-14"><a href="#cb10-14" tabindex="-1"></a>      [z1] raw</span>
<span id="cb10-15"><a href="#cb10-15" tabindex="-1"></a>      [aggregate] raw</span>
<span id="cb10-16"><a href="#cb10-16" tabindex="-1"></a>    spatial networks <span class="sc">-----------------</span></span>
<span id="cb10-17"><a href="#cb10-17" tabindex="-1"></a>      [aggregate] Delaunay_network kNN_network</span>
<span id="cb10-18"><a href="#cb10-18" tabindex="-1"></a>    dim reduction <span class="sc">--------------------</span></span>
<span id="cb10-19"><a href="#cb10-19" tabindex="-1"></a>      [aggregate][rna] pca umap tsne</span>
<span id="cb10-20"><a href="#cb10-20" tabindex="-1"></a>    nearest neighbor networks <span class="sc">--------</span></span>
<span id="cb10-21"><a href="#cb10-21" tabindex="-1"></a>      [aggregate][rna] sNN.pca</span>
<span id="cb10-22"><a href="#cb10-22" tabindex="-1"></a>    attached images <span class="sc">------------------</span></span>
<span id="cb10-23"><a href="#cb10-23" tabindex="-1"></a>    giottoLargeImage <span class="sc">:</span> <span class="dv">4</span> items...</span>
<span id="cb10-24"><a href="#cb10-24" tabindex="-1"></a></span>
<span id="cb10-25"><a href="#cb10-25" tabindex="-1"></a>    Use <span class="fu">objHistory</span>() to see steps and params used</span></code></pre></div>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/activeSpatUnit-generic.html">activeSpatUnit</a></span><span class="op">(</span><span class="va">test_vizgen</span><span class="op">)</span> <span class="op">=</span> <span class="st">'aggregate'</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/spatPlot.html">spatPlot2D</a></span><span class="op">(</span><span class="va">test_vizgen</span>,</span>
<span>           show_image <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>           cell_color <span class="op">=</span> <span class="st">'leiden_clus'</span>,</span>
<span>           largeImage_name <span class="op">=</span> <span class="st">'dapi_z0'</span>,</span>
<span>           point_alpha <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/saving_object/unnamed-chunk-8-1.png"></p>
</div>
<div class="section level2">
<h2 id="session-info">8. Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" tabindex="-1"></a>    R version <span class="dv">4</span>.<span class="fl">2.1</span> (<span class="dv">2022-06-23</span>)</span>
<span id="cb13-2"><a href="#cb13-2" tabindex="-1"></a>    Platform<span class="sc">:</span> x86_64<span class="sc">-</span>apple<span class="sc">-</span><span class="fu">darwin17.0</span> (<span class="dv">64</span><span class="sc">-</span>bit)</span>
<span id="cb13-3"><a href="#cb13-3" tabindex="-1"></a>    Running under<span class="sc">:</span> macOS Big Sur ... <span class="fl">10.16</span></span>
<span id="cb13-4"><a href="#cb13-4" tabindex="-1"></a></span>
<span id="cb13-5"><a href="#cb13-5" tabindex="-1"></a>    Matrix products<span class="sc">:</span> default</span>
<span id="cb13-6"><a href="#cb13-6" tabindex="-1"></a>    BLAS<span class="sc">:</span>   <span class="er">/</span>Library<span class="sc">/</span>Frameworks<span class="sc">/</span>R.framework<span class="sc">/</span>Versions<span class="sc">/</span><span class="fl">4.2</span><span class="sc">/</span>Resources<span class="sc">/</span>lib<span class="sc">/</span>libRblas.<span class="fl">0.</span>dylib</span>
<span id="cb13-7"><a href="#cb13-7" tabindex="-1"></a>    LAPACK<span class="sc">:</span> <span class="er">/</span>Library<span class="sc">/</span>Frameworks<span class="sc">/</span>R.framework<span class="sc">/</span>Versions<span class="sc">/</span><span class="fl">4.2</span><span class="sc">/</span>Resources<span class="sc">/</span>lib<span class="sc">/</span>libRlapack.dylib</span>
<span id="cb13-8"><a href="#cb13-8" tabindex="-1"></a></span>
<span id="cb13-9"><a href="#cb13-9" tabindex="-1"></a>    locale<span class="sc">:</span></span>
<span id="cb13-10"><a href="#cb13-10" tabindex="-1"></a>    [<span class="dv">1</span>] en_US.UTF<span class="dv">-8</span><span class="sc">/</span>en_US.UTF<span class="dv">-8</span><span class="sc">/</span>en_US.UTF<span class="dv">-8</span><span class="sc">/</span>C<span class="sc">/</span>en_US.UTF<span class="dv">-8</span><span class="sc">/</span>en_US.UTF<span class="dv">-8</span></span>
<span id="cb13-11"><a href="#cb13-11" tabindex="-1"></a></span>
<span id="cb13-12"><a href="#cb13-12" tabindex="-1"></a>    attached base packages<span class="sc">:</span></span>
<span id="cb13-13"><a href="#cb13-13" tabindex="-1"></a>    [<span class="dv">1</span>] stats     graphics  grDevices utils     datasets  methods   base     </span>
<span id="cb13-14"><a href="#cb13-14" tabindex="-1"></a></span>
<span id="cb13-15"><a href="#cb13-15" tabindex="-1"></a>    other attached packages<span class="sc">:</span></span>
<span id="cb13-16"><a href="#cb13-16" tabindex="-1"></a>    [<span class="dv">1</span>] GiottoData_0.<span class="fl">2.1</span> Giotto_3.<span class="fl">2.1</span>    </span>
<span id="cb13-17"><a href="#cb13-17" tabindex="-1"></a></span>
<span id="cb13-18"><a href="#cb13-18" tabindex="-1"></a>    loaded via a <span class="fu">namespace</span> (and not attached)<span class="sc">:</span></span>
<span id="cb13-19"><a href="#cb13-19" tabindex="-1"></a>     [<span class="dv">1</span>] Rcpp_1.<span class="fl">0.10</span>        RColorBrewer_1<span class="fl">.1</span><span class="dv">-3</span> pillar_1.<span class="fl">9.0</span>       compiler_4.<span class="fl">2.1</span>    </span>
<span id="cb13-20"><a href="#cb13-20" tabindex="-1"></a>     [<span class="dv">5</span>] tools_4.<span class="fl">2.1</span>        digest_0.<span class="fl">6.31</span>      jsonlite_1.<span class="fl">8.4</span>     evaluate_0<span class="fl">.20</span>     </span>
<span id="cb13-21"><a href="#cb13-21" tabindex="-1"></a>     [<span class="dv">9</span>] lifecycle_1.<span class="fl">0.3</span>    tibble_3.<span class="fl">2.1</span>       gtable_0.<span class="fl">3.3</span>       lattice_0<span class="fl">.20</span><span class="dv">-45</span>   </span>
<span id="cb13-22"><a href="#cb13-22" tabindex="-1"></a>    [<span class="dv">13</span>] png_0<span class="fl">.1</span><span class="dv">-8</span>          pkgconfig_2.<span class="fl">0.3</span>    rlang_1.<span class="fl">1.0</span>        igraph_1.<span class="fl">4.1</span>      </span>
<span id="cb13-23"><a href="#cb13-23" tabindex="-1"></a>    [<span class="dv">17</span>] Matrix_1<span class="fl">.5</span><span class="dv">-3</span>       cli_3.<span class="fl">6.1</span>          rstudioapi_0<span class="fl">.14</span>    parallel_4.<span class="fl">2.1</span>    </span>
<span id="cb13-24"><a href="#cb13-24" tabindex="-1"></a>    [<span class="dv">21</span>] yaml_2.<span class="fl">3.7</span>         xfun_0<span class="fl">.38</span>          fastmap_1.<span class="fl">1.1</span>      terra_1<span class="fl">.7</span><span class="dv">-22</span>      </span>
<span id="cb13-25"><a href="#cb13-25" tabindex="-1"></a>    [<span class="dv">25</span>] withr_2.<span class="fl">5.0</span>        dplyr_1.<span class="fl">1.1</span>        knitr_1<span class="fl">.42</span>         generics_0.<span class="fl">1.3</span>    </span>
<span id="cb13-26"><a href="#cb13-26" tabindex="-1"></a>    [<span class="dv">29</span>] vctrs_0.<span class="fl">6.1</span>        rprojroot_2.<span class="fl">0.3</span>    grid_4.<span class="fl">2.1</span>         tidyselect_1.<span class="fl">2.0</span>  </span>
<span id="cb13-27"><a href="#cb13-27" tabindex="-1"></a>    [<span class="dv">33</span>] here_1.<span class="fl">0.1</span>         reticulate_1<span class="fl">.28</span>    glue_1.<span class="fl">6.2</span>         data.table_1.<span class="fl">14.8</span> </span>
<span id="cb13-28"><a href="#cb13-28" tabindex="-1"></a>    [<span class="dv">37</span>] R6_2.<span class="fl">5.1</span>           fansi_1.<span class="fl">0.4</span>        rmarkdown_2<span class="fl">.21</span>     farver_2.<span class="fl">1.1</span>      </span>
<span id="cb13-29"><a href="#cb13-29" tabindex="-1"></a>    [<span class="dv">41</span>] ggplot2_3.<span class="fl">4.2</span>      magrittr_2.<span class="fl">0.3</span>     scales_1.<span class="fl">2.1</span>       codetools_0<span class="fl">.2</span><span class="dv">-18</span>  </span>
<span id="cb13-30"><a href="#cb13-30" tabindex="-1"></a>    [<span class="dv">45</span>] htmltools_0.<span class="fl">5.5</span>    colorspace_2<span class="fl">.1</span><span class="dv">-0</span>   labeling_0.<span class="fl">4.2</span>     utf8_1.<span class="fl">2.3</span>        </span>
<span id="cb13-31"><a href="#cb13-31" tabindex="-1"></a>    [<span class="dv">49</span>] munsell_0.<span class="fl">5.0</span>     </span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by <a href="https://www.drieslab.com" class="external-link">Ruben Dries</a>, Jiaji George Chen, Natalie Del Rossi, <a href="https://labs.icahn.mssm.edu/yuanlab/" class="external-link">Guo-Cheng Yuan</a>, Matthew Oâ€™Brien, <a href="https://josschavezf.github.io" class="external-link">Joselyn C. ChÃ¡vez-Fuentes</a>, Edward Ruiz.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
