<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="Giotto">
<title>SeqFish+ Mouse Cortex Example â€¢ Giotto</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/JetBrains_Mono-0.4.7/font.css" rel="stylesheet">
<!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="SeqFish+ Mouse Cortex Example">
<meta property="og:description" content="Giotto">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">Giotto</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">4.0.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-get-started">Get started</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-get-started">
    <a class="dropdown-item" href="../articles/installation.html">Giotto Installation</a>
    <a class="dropdown-item" href="../articles/ecosystem.html">Giotto Ecosystem</a>
    <a class="dropdown-item" href="../articles/configuration.html">Giotto Configuration</a>
    <a class="dropdown-item" href="../articles/object_creation.html">Giotto Object Creation</a>
    <a class="dropdown-item" href="../articles/structure.html">Giotto Class Structure</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Documentation</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-tutorials">Tutorials</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-tutorials">
    <h6 class="dropdown-header" data-toc-skip>Seq-Scope</h6>
    <a class="dropdown-item" href="../articles/seqscope_mouse_liver.html">Mouse liver</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Visium</h6>
    <a class="dropdown-item" href="../articles/visium_mouse_kidney.html">Visium Mouse Kidney</a>
    <a class="dropdown-item" href="../articles/visium_mouse_brain.html">Visium Mouse Brain</a>
    <a class="dropdown-item" href="../articles/visium_prostate_integration.html">Visium Prostate Integration</a>
    <a class="dropdown-item" href="../articles/visium_cytassist_lungcancer.html">CytAssist Human Lung Cancer</a>
    <a class="dropdown-item" href="../articles/visium_cytassist_human_glioblastoma.html">CytAssist Multi-omics Human Glioblastoma</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Spatial Genomics</h6>
    <a class="dropdown-item" href="../articles/spatial_genomics_mouse_kidney.html">Mouse Kidney</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Xenium</h6>
    <a class="dropdown-item" href="../articles/xenium_breast_cancer.html">Xenium Human breast cancer</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Nanostring CosMx</h6>
    <a class="dropdown-item" href="../articles/nanostring_cosmx_lung_cancer.html">Nanostring CosMx Subcellular Lung Cancer</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>MERFISH</h6>
    <a class="dropdown-item" href="../articles/merfish_mouse_hypothalamic.html">MERFISH Mouse Hypothalamic Preoptic Region</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Vizgen</h6>
    <a class="dropdown-item" href="../articles/vizgen_mouse_brain.html">Vizgen Mouse Brain Receptor Map</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>seqFISH</h6>
    <a class="dropdown-item" href="../articles/mini_seqfish.html">mini seqFISH</a>
    <a class="dropdown-item" href="../articles/seqfish_cortex.html">SeqFish+ Mouse Cortex Example</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>osmFISH</h6>
    <a class="dropdown-item" href="../articles/osmfish_mouse_ss_cortex.html">osmFISH Mouse SS Cortex</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Resolve Bioscience</h6>
    <a class="dropdown-item" href="../articles/resolve_biosciences.html">Resolve Bioscience Breast Cancer Subcellular</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>CODEX</h6>
    <a class="dropdown-item" href="../articles/codex_mouse_spleen.html">Codex Mouse Spleen</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Single Cell RNA Sequencing</h6>
    <a class="dropdown-item" href="../articles/singlecell_prostate_standard.html">Prostate</a>
    <a class="dropdown-item" href="../articles/singlecell_prostate_integration.html">Prostate Integration</a>
  </div>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-tools">Tools</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-tools">
    <h6 class="dropdown-header" data-toc-skip>Spatial manipulation</h6>
    <a class="dropdown-item" href="../articles/spatial_transformations.html">Spatial transformations</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Image alignment</h6>
    <a class="dropdown-item" href="../articles/image_alignment.html">Image alignment</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Interactive selection</h6>
    <a class="dropdown-item" href="../articles/interactive_selection.html">Interactive selection</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Visualizations</h6>
    <a class="dropdown-item" href="../articles/visualizations.html">Visualizations</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Saving and Loading</h6>
    <a class="dropdown-item" href="../articles/saving_object.html">Saving a Giotto object</a>
    <a class="dropdown-item" href="../articles/saving_plots.html">Saving plots</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Interoperability</h6>
    <a class="dropdown-item" href="../articles/interoperability_anndata.html">AnnData</a>
  </div>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-try-giotto-in-the-cloud">Try Giotto in the Cloud</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-try-giotto-in-the-cloud">
    <a class="dropdown-item" href="../articles/terra.html">Terra.bio</a>
    <a class="dropdown-item" href="../articles/docker.html">Docker</a>
    <div class="dropdown-divider"></div>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/presentations.html">Presentations</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/branches.html">Branches</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">News</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/drieslab/Giotto/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>SeqFish+ Mouse Cortex Example</h1>
            
            <h4 data-toc-skip class="date">2022-09-16</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/drieslab/Giotto/blob/HEAD/vignettes/seqfish_cortex.Rmd" class="external-link"><code>vignettes/seqfish_cortex.Rmd</code></a></small>
      <div class="d-none name"><code>seqfish_cortex.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="start-giotto">Start Giotto<a class="anchor" aria-label="anchor" href="#start-giotto"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Ensure Giotto Suite is installed.</span></span>
<span><span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="st">"Giotto"</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/installed.packages.html" class="external-link">installed.packages</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"drieslab/Giotto@suite"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co"># Ensure GiottoData, a small, helper module for tutorials, is installed.</span></span>
<span><span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="st">"GiottoData"</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/installed.packages.html" class="external-link">installed.packages</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"drieslab/GiottoData"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://drieslab.github.io/Giotto/">Giotto</a></span><span class="op">)</span></span>
<span><span class="co"># Ensure the Python environment for Giotto has been installed.</span></span>
<span><span class="va">genv_exists</span> <span class="op">=</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/checkGiottoEnvironment.html">checkGiottoEnvironment</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="va">genv_exists</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="co"># The following command need only be run once to install the Giotto environment.</span></span>
<span>  <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/installGiottoEnvironment.html">installGiottoEnvironment</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="dataset-explanation">Dataset Explanation<a class="anchor" aria-label="anchor" href="#dataset-explanation"></a>
</h2>
<p>Several fields - containing 100â€™s of cells - in the mouse cortex and
subventricular zone were imaged for seqFISH+. The coordinates of the
cells within each field are independent of each other, so in order to
visualize and process all cells together imaging fields will be stitched
together by providing x and y-offset values specific to each field.</p>
<p>These offset values are known or estimates based on the original raw
image:</p>
<p><img src="images/seqfish_cortex/cortex_svz_location_fields.png"></p>
</div>
<div class="section level2">
<h2 id="download-data">Download Data<a class="anchor" aria-label="anchor" href="#download-data"></a>
</h2>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://drieslab.github.io/GiottoData/">GiottoData</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">my_working_dir</span> <span class="op">=</span> <span class="st">'/path/to/directory/'</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/GiottoData/reference/getSpatialDataset.html">getSpatialDataset</a></span><span class="op">(</span>dataset <span class="op">=</span> <span class="st">'seqfish_SS_cortex'</span>, directory <span class="op">=</span> <span class="va">my_working_dir</span>, method <span class="op">=</span> <span class="st">'wget'</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="giotto-instructions-and-preparation">1. Giotto Instructions and Preparation<a class="anchor" aria-label="anchor" href="#giotto-instructions-and-preparation"></a>
</h2>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#  set Giotto instructions</span></span>
<span><span class="va">instrs</span> <span class="op">=</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/createGiottoInstructions.html">createGiottoInstructions</a></span><span class="op">(</span>save_plot <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>                                  show_plot <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                                  save_dir <span class="op">=</span> <span class="va">my_working_dir</span>, </span>
<span>                                  python_path <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># create giotto object from provided paths ####</span></span>
<span><span class="va">expr_path</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">my_working_dir</span>, <span class="st">"cortex_svz_expression.txt"</span><span class="op">)</span></span>
<span><span class="va">loc_path</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">my_working_dir</span>, <span class="st">"cortex_svz_centroids_coord.txt"</span><span class="op">)</span></span>
<span><span class="va">meta_path</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">my_working_dir</span>, <span class="st">"cortex_svz_centroids_annot.txt"</span><span class="op">)</span></span>
<span><span class="co">#This dataset contains multiple field of views which need to be stitched together</span></span>
<span><span class="co"># first merge location and additional metadata</span></span>
<span><span class="va">SS_locations</span> <span class="op">=</span> <span class="fu">data.table</span><span class="fu">::</span><span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="va">loc_path</span><span class="op">)</span></span>
<span><span class="va">cortex_fields</span> <span class="op">=</span> <span class="fu">data.table</span><span class="fu">::</span><span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="va">meta_path</span><span class="op">)</span></span>
<span><span class="va">SS_loc_annot</span> <span class="op">=</span> <span class="fu">data.table</span><span class="fu">::</span><span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/merge.html" class="external-link">merge.data.table</a></span><span class="op">(</span><span class="va">SS_locations</span>, <span class="va">cortex_fields</span>, by <span class="op">=</span> <span class="st">'ID'</span><span class="op">)</span></span>
<span><span class="va">SS_loc_annot</span><span class="op">[</span>, <span class="va">ID</span> <span class="op">:=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">ID</span>, levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">'cell_'</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">913</span><span class="op">)</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu">data.table</span><span class="fu">::</span><span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/setorder.html" class="external-link">setorder</a></span><span class="op">(</span><span class="va">SS_loc_annot</span>, <span class="va">ID</span><span class="op">)</span></span>
<span><span class="co"># create file with offset information</span></span>
<span><span class="va">my_offset_file</span> <span class="op">=</span> <span class="fu">data.table</span><span class="fu">::</span><span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/data.table.html" class="external-link">data.table</a></span><span class="op">(</span>field <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">3</span>, <span class="fl">4</span>, <span class="fl">5</span>, <span class="fl">6</span><span class="op">)</span>,</span>
<span>                                        x_offset <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1654.97</span>, <span class="fl">1750.75</span>, <span class="fl">1674.35</span>, <span class="fl">675.5</span>, <span class="fl">2048</span>, <span class="fl">675</span><span class="op">)</span>,</span>
<span>                                        y_offset <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="op">-</span><span class="fl">1438.02</span>, <span class="op">-</span><span class="fl">1438.02</span>, <span class="fl">0</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># create a stitch file</span></span>
<span><span class="va">stitch_file</span> <span class="op">=</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/stitchFieldCoordinates.html">stitchFieldCoordinates</a></span><span class="op">(</span>location_file <span class="op">=</span> <span class="va">SS_loc_annot</span>,</span>
<span>                                     offset_file <span class="op">=</span> <span class="va">my_offset_file</span>,</span>
<span>                                     cumulate_offset_x <span class="op">=</span> <span class="cn">T</span>,</span>
<span>                                     cumulate_offset_y <span class="op">=</span> <span class="cn">F</span>,</span>
<span>                                     field_col <span class="op">=</span> <span class="st">'FOV'</span>,</span>
<span>                                     reverse_final_x <span class="op">=</span> <span class="cn">F</span>,</span>
<span>                                     reverse_final_y <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span><span class="va">stitch_file</span> <span class="op">=</span> <span class="va">stitch_file</span><span class="op">[</span>,<span class="fu">.</span><span class="op">(</span><span class="va">ID</span>, <span class="va">X_final</span>, <span class="va">Y_final</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">stitch_file</span><span class="op">$</span><span class="va">ID</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">stitch_file</span><span class="op">$</span><span class="va">ID</span><span class="op">)</span></span>
<span><span class="va">my_offset_file</span> <span class="op">=</span> <span class="va">my_offset_file</span><span class="op">[</span>,<span class="fu">.</span><span class="op">(</span><span class="va">field</span>, <span class="va">x_offset_final</span>, <span class="va">y_offset_final</span><span class="op">)</span><span class="op">]</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="create-giotto-object-process-data">2. Create Giotto object &amp; process data<a class="anchor" aria-label="anchor" href="#create-giotto-object-process-data"></a>
</h2>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># create Giotto object</span></span>
<span><span class="va">SS_seqfish</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/createGiottoObject.html">createGiottoObject</a></span><span class="op">(</span>expression <span class="op">=</span> <span class="va">expr_path</span>,</span>
<span>                                 spatial_locs <span class="op">=</span> <span class="va">stitch_file</span>,</span>
<span>                                 offset_file <span class="op">=</span> <span class="va">my_offset_file</span>,</span>
<span>                                 instructions <span class="op">=</span> <span class="va">instrs</span><span class="op">)</span></span>
<span><span class="co"># add additional annotation if wanted</span></span>
<span><span class="va">SS_seqfish</span> <span class="op">=</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/addCellMetadata.html">addCellMetadata</a></span><span class="op">(</span><span class="va">SS_seqfish</span>,</span>
<span>                             new_metadata <span class="op">=</span> <span class="va">cortex_fields</span>,</span>
<span>                             by_column <span class="op">=</span> <span class="cn">T</span>,</span>
<span>                             column_cell_ID <span class="op">=</span> <span class="st">'ID'</span><span class="op">)</span></span>
<span><span class="co"># subset data to the cortex field of views</span></span>
<span><span class="va">cell_metadata</span> <span class="op">=</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/pDataDT.html">pDataDT</a></span><span class="op">(</span><span class="va">SS_seqfish</span><span class="op">)</span></span>
<span><span class="va">cortex_cell_ids</span> <span class="op">=</span> <span class="va">cell_metadata</span><span class="op">[</span><span class="va">FOV</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fl">0</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span><span class="op">$</span><span class="va">cell_ID</span></span>
<span><span class="va">SS_seqfish</span> <span class="op">=</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/subsetGiotto.html">subsetGiotto</a></span><span class="op">(</span><span class="va">SS_seqfish</span>, cell_ids <span class="op">=</span> <span class="va">cortex_cell_ids</span><span class="op">)</span></span>
<span><span class="co"># filter</span></span>
<span><span class="va">SS_seqfish</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/filterGiotto.html">filterGiotto</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">SS_seqfish</span>,</span>
<span>                           expression_threshold <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                           feat_det_in_min_cells <span class="op">=</span> <span class="fl">10</span>,</span>
<span>                           min_det_feats_per_cell <span class="op">=</span> <span class="fl">10</span>,</span>
<span>                           expression_values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'raw'</span><span class="op">)</span>,</span>
<span>                           verbose <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span><span class="co"># normalize</span></span>
<span><span class="va">SS_seqfish</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/normalizeGiotto.html">normalizeGiotto</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">SS_seqfish</span>, scalefactor <span class="op">=</span> <span class="fl">6000</span>, verbose <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span><span class="co"># add gene &amp; cell statistics</span></span>
<span><span class="va">SS_seqfish</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/addStatistics.html">addStatistics</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">SS_seqfish</span><span class="op">)</span></span>
<span><span class="co"># adjust expression matrix for technical or known variables</span></span>
<span><span class="va">SS_seqfish</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/adjustGiottoMatrix.html">adjustGiottoMatrix</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">SS_seqfish</span>, expression_values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'normalized'</span><span class="op">)</span>,</span>
<span>                                 covariate_columns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'nr_feats'</span>, <span class="st">'total_expr'</span><span class="op">)</span>,</span>
<span>                                 return_gobject <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                                 update_slot <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'custom'</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># visualize</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/spatPlot.html">spatPlot</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">SS_seqfish</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/seqfish_cortex/0-spatPlot2D.png"></p>
</div>
<div class="section level2">
<h2 id="dimension-reduction">3. Dimension Reduction<a class="anchor" aria-label="anchor" href="#dimension-reduction"></a>
</h2>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## highly variable features (HVF)</span></span>
<span><span class="va">SS_seqfish</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateHVF.html">calculateHVF</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">SS_seqfish</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/seqfish_cortex/1-HVFplot.png"></p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## select genes based on highly variable features and gene statistics, both found in feature (gene) metadata</span></span>
<span><span class="va">gene_metadata</span> <span class="op">=</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/fDataDT.html">fDataDT</a></span><span class="op">(</span><span class="va">SS_seqfish</span><span class="op">)</span></span>
<span><span class="va">featgenes</span> <span class="op">=</span> <span class="va">gene_metadata</span><span class="op">[</span><span class="va">hvf</span> <span class="op">==</span> <span class="st">'yes'</span> <span class="op">&amp;</span> <span class="va">perc_cells</span> <span class="op">&gt;</span> <span class="fl">4</span> <span class="op">&amp;</span> <span class="va">mean_expr_det</span> <span class="op">&gt;</span> <span class="fl">0.5</span><span class="op">]</span><span class="op">$</span><span class="va">feat_ID</span></span>
<span><span class="co">## run PCA on expression values (default)</span></span>
<span><span class="va">SS_seqfish</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runPCA.html">runPCA</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">SS_seqfish</span>, genes_to_use <span class="op">=</span> <span class="va">featgenes</span>, scale_unit <span class="op">=</span> <span class="cn">F</span>, center <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/screePlot.html">screePlot</a></span><span class="op">(</span><span class="va">SS_seqfish</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/seqfish_cortex/2-screePlot.png"></p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/plotPCA.html">plotPCA</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">SS_seqfish</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/seqfish_cortex/3-PCA.png"></p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">SS_seqfish</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runUMAP.html">runUMAP</a></span><span class="op">(</span><span class="va">SS_seqfish</span>, dimensions_to_use <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">15</span>, n_threads <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/plotUMAP.html">plotUMAP</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">SS_seqfish</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/seqfish_cortex/4-UMAP.png"></p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">SS_seqfish</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runtSNE.html">runtSNE</a></span><span class="op">(</span><span class="va">SS_seqfish</span>, dimensions_to_use <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">15</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/plotTSNE.html">plotTSNE</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">SS_seqfish</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/seqfish_cortex/5-tSNE.png"></p>
</div>
<div class="section level2">
<h2 id="cluster">4. Cluster<a class="anchor" aria-label="anchor" href="#cluster"></a>
</h2>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## sNN network (default)</span></span>
<span><span class="va">SS_seqfish</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/createNearestNetwork.html">createNearestNetwork</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">SS_seqfish</span>,</span>
<span>                                   dimensions_to_use <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">15</span>,</span>
<span>                                   k <span class="op">=</span> <span class="fl">15</span><span class="op">)</span></span>
<span><span class="co">## Leiden clustering</span></span>
<span><span class="va">SS_seqfish</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/doLeidenCluster.html">doLeidenCluster</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">SS_seqfish</span>,</span>
<span>                              resolution <span class="op">=</span> <span class="fl">0.4</span>,</span>
<span>                              n_iterations <span class="op">=</span> <span class="fl">1000</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/plotUMAP.html">plotUMAP</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">SS_seqfish</span>,</span>
<span>         cell_color <span class="op">=</span> <span class="st">'leiden_clus'</span>,</span>
<span>         show_NN_network <span class="op">=</span> <span class="cn">T</span>,</span>
<span>         point_size <span class="op">=</span> <span class="fl">2.5</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/seqfish_cortex/6-UMAP.png"></p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Leiden subclustering for specified clusters</span></span>
<span><span class="va">SS_seqfish</span> <span class="op">=</span> <span class="fu"><a href="../reference/doLeidenSubCluster.html">doLeidenSubCluster</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">SS_seqfish</span>, </span>
<span>                                cluster_column <span class="op">=</span> <span class="st">'leiden_clus'</span>,</span>
<span>                                resolution <span class="op">=</span> <span class="fl">0.2</span>, k_neighbors <span class="op">=</span> <span class="fl">10</span>,</span>
<span>                                pca_param <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>expression_values <span class="op">=</span> <span class="st">'normalized'</span>, scale_unit <span class="op">=</span> <span class="cn">F</span><span class="op">)</span>,</span>
<span>                                nn_param <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>dimensions_to_use <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">)</span>,</span>
<span>                                selected_clusters <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">5</span>, <span class="fl">6</span>, <span class="fl">7</span><span class="op">)</span>,</span>
<span>                                name <span class="op">=</span> <span class="st">'sub_leiden_clus_select'</span><span class="op">)</span></span>
<span><span class="co">## set colors for clusters</span></span>
<span><span class="va">subleiden_order</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span> <span class="fl">1.1</span>, <span class="fl">2.1</span>, <span class="fl">3.1</span>, <span class="fl">4.1</span>, <span class="fl">5.1</span>, <span class="fl">5.2</span>, </span>
<span>                     <span class="fl">6.1</span>, <span class="fl">6.2</span>, <span class="fl">7.1</span>,  <span class="fl">7.2</span>, <span class="fl">8.1</span>, <span class="fl">9.1</span><span class="op">)</span></span>
<span><span class="va">subleiden_colors</span> <span class="op">=</span> <span class="fu">GiottogetDistinctColors</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">subleiden_order</span><span class="op">)</span><span class="op">)</span> </span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">subleiden_colors</span><span class="op">)</span> <span class="op">=</span> <span class="va">subleiden_order</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/plotUMAP.html">plotUMAP</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">SS_seqfish</span>,</span>
<span>         cell_color <span class="op">=</span> <span class="st">'sub_leiden_clus_select'</span>, cell_color_code <span class="op">=</span> <span class="va">subleiden_colors</span>,</span>
<span>         show_NN_network <span class="op">=</span> <span class="cn">T</span>, point_size <span class="op">=</span> <span class="fl">2.5</span>, show_center_label <span class="op">=</span> <span class="cn">F</span>, </span>
<span>         legend_text <span class="op">=</span> <span class="fl">12</span>, legend_symbol_size <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/seqfish_cortex/10-UMAP.png"></p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## show cluster relationships</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/showClusterHeatmap.html">showClusterHeatmap</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">SS_seqfish</span>, cluster_column <span class="op">=</span> <span class="st">'sub_leiden_clus_select'</span>,</span>
<span>                   row_names_gp <span class="op">=</span> <span class="fu">grid</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/grid/gpar.html" class="external-link">gpar</a></span><span class="op">(</span>fontsize <span class="op">=</span> <span class="fl">9</span><span class="op">)</span>, column_names_gp <span class="op">=</span> <span class="fu">grid</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/grid/gpar.html" class="external-link">gpar</a></span><span class="op">(</span>fontsize <span class="op">=</span> <span class="fl">9</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/seqfish_cortex/11-showClusterHeatmap.png"></p>
<p>The following step requires the installation of {ggdendro}.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># install.packages('ggdendro')</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/andrie/ggdendro" class="external-link">ggdendro</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/showClusterDendrogram.html">showClusterDendrogram</a></span><span class="op">(</span><span class="va">SS_seqfish</span>, h <span class="op">=</span> <span class="fl">0.5</span>, rotate <span class="op">=</span> <span class="cn">T</span>, cluster_column <span class="op">=</span> <span class="st">'sub_leiden_clus_select'</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/seqfish_cortex/12-showClusterDendrogram.png"></p>
</div>
<div class="section level2">
<h2 id="visualize-spatial-and-expression-space">5. Visualize Spatial and Expression Space<a class="anchor" aria-label="anchor" href="#visualize-spatial-and-expression-space"></a>
</h2>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># expression and spatial</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/spatDimPlot.html">spatDimPlot</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">SS_seqfish</span>, cell_color <span class="op">=</span> <span class="st">'sub_leiden_clus_select'</span>, </span>
<span>            cell_color_code <span class="op">=</span> <span class="va">subleiden_colors</span>,</span>
<span>            dim_point_size <span class="op">=</span> <span class="fl">2</span>, spat_point_size <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/seqfish_cortex/13-spatDimPlot2D.png"></p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># selected groups and provide new colors</span></span>
<span><span class="va">groups_of_interest</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">6.1</span>, <span class="fl">6.2</span>, <span class="fl">7.1</span>, <span class="fl">7.2</span><span class="op">)</span></span>
<span><span class="va">group_colors</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'red'</span>, <span class="st">'green'</span>, <span class="st">'blue'</span>, <span class="st">'purple'</span><span class="op">)</span>; <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">group_colors</span><span class="op">)</span> <span class="op">=</span> <span class="va">groups_of_interest</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/spatDimPlot.html">spatDimPlot</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">SS_seqfish</span>, cell_color <span class="op">=</span> <span class="st">'sub_leiden_clus_select'</span>, </span>
<span>            dim_point_size <span class="op">=</span> <span class="fl">2</span>, spat_point_size <span class="op">=</span> <span class="fl">2</span>,</span>
<span>            select_cell_groups <span class="op">=</span> <span class="va">groups_of_interest</span>, cell_color_code <span class="op">=</span> <span class="va">group_colors</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/seqfish_cortex/14-spatDimPlot2D.png"></p>
</div>
<div class="section level2">
<h2 id="cell-type-marker-gene-detection">6. Cell Type Marker Gene Detection<a class="anchor" aria-label="anchor" href="#cell-type-marker-gene-detection"></a>
</h2>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## gini </span></span>
<span><span class="va">gini_markers_subclusters</span> <span class="op">=</span> <span class="fu"><a href="../reference/findMarkers_one_vs_all.html">findMarkers_one_vs_all</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">SS_seqfish</span>,</span>
<span>                                                  method <span class="op">=</span> <span class="st">'gini'</span>,</span>
<span>                                                  expression_values <span class="op">=</span> <span class="st">'normalized'</span>,</span>
<span>                                                  cluster_column <span class="op">=</span> <span class="st">'sub_leiden_clus_select'</span>,</span>
<span>                                                  min_feats <span class="op">=</span> <span class="fl">20</span>,</span>
<span>                                                  min_expr_gini_score <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>                                                  min_det_gini_score <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span>
<span><span class="va">topgenes_gini</span> <span class="op">=</span> <span class="va">gini_markers_subclusters</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">.SD</span>, <span class="fl">2</span><span class="op">)</span>, by <span class="op">=</span> <span class="st">'cluster'</span><span class="op">]</span></span>
<span><span class="co">## violin plot</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/violinPlot.html">violinPlot</a></span><span class="op">(</span><span class="va">SS_seqfish</span>, feats <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">topgenes_gini</span><span class="op">$</span><span class="va">feats</span><span class="op">)</span>, cluster_column <span class="op">=</span> <span class="st">'sub_leiden_clus_select'</span>,</span>
<span>           strip_text <span class="op">=</span> <span class="fl">8</span>, strip_position <span class="op">=</span> <span class="st">'right'</span>, cluster_custom_order <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">topgenes_gini</span><span class="op">$</span><span class="va">cluster</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/seqfish_cortex/15-violinPlot.png"></p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># cluster heatmap</span></span>
<span><span class="va">topgenes_gini2</span> <span class="op">=</span> <span class="va">gini_markers_subclusters</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">.SD</span>, <span class="fl">6</span><span class="op">)</span>, by <span class="op">=</span> <span class="st">'cluster'</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/plotMetaDataHeatmap.html">plotMetaDataHeatmap</a></span><span class="op">(</span><span class="va">SS_seqfish</span>, selected_feats <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">topgenes_gini2</span><span class="op">$</span><span class="va">feats</span><span class="op">)</span>, </span>
<span>                    custom_feat_order <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">topgenes_gini2</span><span class="op">$</span><span class="va">feats</span><span class="op">)</span>,</span>
<span>                    custom_cluster_order <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">topgenes_gini2</span><span class="op">$</span><span class="va">cluster</span><span class="op">)</span>,</span>
<span>                    metadata_cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'sub_leiden_clus_select'</span><span class="op">)</span>, x_text_size <span class="op">=</span> <span class="fl">10</span>, y_text_size <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/seqfish_cortex/16-plotMetaDataHeatmap.png"></p>
</div>
<div class="section level2">
<h2 id="cell-type-annotation">7. Cell Type Annotation<a class="anchor" aria-label="anchor" href="#cell-type-annotation"></a>
</h2>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## general cell types</span></span>
<span><span class="co">## create vector with names</span></span>
<span><span class="va">clusters_cell_types_cortex</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'L6 eNeuron'</span>, <span class="st">'L4 eNeuron'</span>, <span class="st">'L2/3 eNeuron'</span>, <span class="st">'L5 eNeuron'</span>, </span>
<span>                               <span class="st">'Lhx6 iNeuron'</span>, <span class="st">'Adarb2 iNeuron'</span>, </span>
<span>                               <span class="st">'endothelial'</span>, <span class="st">'mural'</span>,</span>
<span>                               <span class="st">'OPC'</span>,<span class="st">'Olig'</span>,</span>
<span>                               <span class="st">'astrocytes'</span>, <span class="st">'microglia'</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">clusters_cell_types_cortex</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1.1</span>, <span class="fl">2.1</span>, <span class="fl">3.1</span>, <span class="fl">4.1</span>,</span>
<span>                                      <span class="fl">5.1</span>, <span class="fl">5.2</span>,</span>
<span>                                      <span class="fl">6.1</span>, <span class="fl">6.2</span>, </span>
<span>                                      <span class="fl">7.1</span>, <span class="fl">7.2</span>,</span>
<span>                                      <span class="fl">8.1</span>, <span class="fl">9.1</span><span class="op">)</span></span>
<span><span class="va">SS_seqfish</span> <span class="op">=</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/annotateGiotto.html">annotateGiotto</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">SS_seqfish</span>, annotation_vector <span class="op">=</span> <span class="va">clusters_cell_types_cortex</span>,</span>
<span>                            cluster_column <span class="op">=</span> <span class="st">'sub_leiden_clus_select'</span>, name <span class="op">=</span> <span class="st">'cell_types'</span><span class="op">)</span></span>
<span><span class="co"># cell type order and colors</span></span>
<span><span class="va">cell_type_order</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'L6 eNeuron'</span>, <span class="st">'L5 eNeuron'</span>, <span class="st">'L4 eNeuron'</span>, <span class="st">'L2/3 eNeuron'</span>,</span>
<span>                    <span class="st">'astrocytes'</span>, <span class="st">'Olig'</span>, <span class="st">'OPC'</span>,<span class="st">'Adarb2 iNeuron'</span>, <span class="st">'Lhx6 iNeuron'</span>,</span>
<span>                    <span class="st">'endothelial'</span>, <span class="st">'mural'</span>, <span class="st">'microglia'</span><span class="op">)</span></span>
<span><span class="va">cell_type_colors</span> <span class="op">=</span> <span class="va">subleiden_colors</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">cell_type_colors</span><span class="op">)</span> <span class="op">=</span> <span class="va">clusters_cell_types_cortex</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">subleiden_colors</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">cell_type_colors</span> <span class="op">=</span> <span class="va">cell_type_colors</span><span class="op">[</span><span class="va">cell_type_order</span><span class="op">]</span></span>
<span><span class="co">## violin plot</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/violinPlot.html">violinPlot</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">SS_seqfish</span>, feats <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">topgenes_gini</span><span class="op">$</span><span class="va">feats</span><span class="op">)</span>,</span>
<span>           strip_text <span class="op">=</span> <span class="fl">7</span>, strip_position <span class="op">=</span> <span class="st">'right'</span>, </span>
<span>           cluster_custom_order <span class="op">=</span> <span class="va">cell_type_order</span>,</span>
<span>           cluster_column <span class="op">=</span> <span class="st">'cell_types'</span>, color_violin <span class="op">=</span> <span class="st">'cluster'</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/seqfish_cortex/17-violinPlot.png"></p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># co-visualization</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/spatDimPlot.html">spatDimPlot</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">SS_seqfish</span>, cell_color <span class="op">=</span> <span class="st">'cell_types'</span>,</span>
<span>            dim_point_size <span class="op">=</span> <span class="fl">2</span>, spat_point_size <span class="op">=</span> <span class="fl">2</span>, dim_show_cluster_center <span class="op">=</span> <span class="cn">F</span>, dim_show_center_label <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/seqfish_cortex/18-spatDimPlot2D.png"></p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## heatmap genes vs cells</span></span>
<span><span class="va">gini_markers_subclusters</span><span class="op">[</span>, <span class="va">cell_types</span> <span class="op">:=</span> <span class="va">clusters_cell_types_cortex</span><span class="op">[</span><span class="va">cluster</span><span class="op">]</span> <span class="op">]</span></span>
<span><span class="va">gini_markers_subclusters</span><span class="op">[</span>, <span class="va">cell_types</span> <span class="op">:=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">cell_types</span>, <span class="va">cell_type_order</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu">data.table</span><span class="fu">::</span><span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/setorder.html" class="external-link">setorder</a></span><span class="op">(</span><span class="va">gini_markers_subclusters</span>, <span class="va">cell_types</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/plotHeatmap.html">plotHeatmap</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">SS_seqfish</span>,</span>
<span>            feats <span class="op">=</span> <span class="va">gini_markers_subclusters</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">.SD</span>, <span class="fl">3</span><span class="op">)</span>, by <span class="op">=</span> <span class="st">'cell_types'</span><span class="op">]</span><span class="op">$</span><span class="va">feats</span>, </span>
<span>            feat_order <span class="op">=</span> <span class="st">'custom'</span>,</span>
<span>            feat_custom_order <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">gini_markers_subclusters</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">.SD</span>, <span class="fl">3</span><span class="op">)</span>, by <span class="op">=</span> <span class="st">'cluster'</span><span class="op">]</span><span class="op">$</span><span class="va">feats</span><span class="op">)</span>,</span>
<span>            cluster_column <span class="op">=</span> <span class="st">'cell_types'</span>, cluster_order <span class="op">=</span> <span class="st">'custom'</span>,</span>
<span>            cluster_custom_order <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">gini_markers_subclusters</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">.SD</span>, <span class="fl">3</span><span class="op">)</span>, by <span class="op">=</span> <span class="st">'cell_types'</span><span class="op">]</span><span class="op">$</span><span class="va">cell_types</span><span class="op">)</span>, </span>
<span>            legend_nrows <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/seqfish_cortex/19-plotHeatmap.png"></p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/plotHeatmap.html">plotHeatmap</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">SS_seqfish</span>,</span>
<span>            cluster_color_code <span class="op">=</span> <span class="va">cell_type_colors</span>,</span>
<span>            feats <span class="op">=</span> <span class="va">gini_markers_subclusters</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">.SD</span>, <span class="fl">6</span><span class="op">)</span>, by <span class="op">=</span> <span class="st">'cell_types'</span><span class="op">]</span><span class="op">$</span><span class="va">feats</span>,</span>
<span>            feat_order <span class="op">=</span> <span class="st">'custom'</span>,</span>
<span>            feat_label_selection <span class="op">=</span> <span class="va">gini_markers_subclusters</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">.SD</span>, <span class="fl">2</span><span class="op">)</span>, by <span class="op">=</span> <span class="st">'cluster'</span><span class="op">]</span><span class="op">$</span><span class="va">feats</span>,</span>
<span>            feat_custom_order <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">gini_markers_subclusters</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">.SD</span>, <span class="fl">6</span><span class="op">)</span>, by <span class="op">=</span> <span class="st">'cluster'</span><span class="op">]</span><span class="op">$</span><span class="va">feats</span><span class="op">)</span>,</span>
<span>            cluster_column <span class="op">=</span> <span class="st">'cell_types'</span>, cluster_order <span class="op">=</span> <span class="st">'custom'</span>,</span>
<span>            cluster_custom_order <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">gini_markers_subclusters</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">.SD</span>, <span class="fl">3</span><span class="op">)</span>, by <span class="op">=</span> <span class="st">'cell_types'</span><span class="op">]</span><span class="op">$</span><span class="va">cell_types</span><span class="op">)</span>, </span>
<span>            legend_nrows <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/seqfish_cortex/20-plotHeatmap.png"></p>
</div>
<div class="section level2">
<h2 id="spatial-grid">8. Spatial Grid<a class="anchor" aria-label="anchor" href="#spatial-grid"></a>
</h2>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">SS_seqfish</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/createSpatialGrid.html">createSpatialGrid</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">SS_seqfish</span>,</span>
<span>                                sdimx_stepsize <span class="op">=</span> <span class="fl">500</span>,</span>
<span>                                sdimy_stepsize <span class="op">=</span> <span class="fl">500</span>,</span>
<span>                                minimum_padding <span class="op">=</span> <span class="fl">50</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/spatPlot.html">spatPlot</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">SS_seqfish</span>, show_grid <span class="op">=</span> <span class="cn">T</span>, point_size <span class="op">=</span> <span class="fl">1.5</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/seqfish_cortex/21-spatPlot2D.png"></p>
</div>
<div class="section level2">
<h2 id="spatial-network">9. Spatial Network<a class="anchor" aria-label="anchor" href="#spatial-network"></a>
</h2>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## delaunay network: stats + creation</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/plotStatDelaunayNetwork.html">plotStatDelaunayNetwork</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">SS_seqfish</span>, maximum_distance <span class="op">=</span> <span class="fl">400</span>, save_plot <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span>
<span><span class="va">SS_seqfish</span> <span class="op">=</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/createSpatialNetwork.html">createSpatialNetwork</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">SS_seqfish</span>, minimum_k <span class="op">=</span> <span class="fl">2</span>, maximum_distance_delaunay <span class="op">=</span> <span class="fl">400</span><span class="op">)</span></span>
<span><span class="co">## create spatial networks based on k and/or distance from centroid</span></span>
<span><span class="va">SS_seqfish</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/createSpatialNetwork.html">createSpatialNetwork</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">SS_seqfish</span>, method <span class="op">=</span> <span class="st">'kNN'</span>, k <span class="op">=</span> <span class="fl">5</span>, name <span class="op">=</span> <span class="st">'spatial_network'</span><span class="op">)</span></span>
<span><span class="va">SS_seqfish</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/createSpatialNetwork.html">createSpatialNetwork</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">SS_seqfish</span>, method <span class="op">=</span> <span class="st">'kNN'</span>, k <span class="op">=</span> <span class="fl">10</span>, name <span class="op">=</span> <span class="st">'large_network'</span><span class="op">)</span></span>
<span><span class="va">SS_seqfish</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/createSpatialNetwork.html">createSpatialNetwork</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">SS_seqfish</span>, method <span class="op">=</span> <span class="st">'kNN'</span>, k <span class="op">=</span> <span class="fl">100</span>,</span>
<span>                                   maximum_distance_knn <span class="op">=</span> <span class="fl">200</span>, minimum_k <span class="op">=</span> <span class="fl">2</span>, name <span class="op">=</span> <span class="st">'distance_network'</span><span class="op">)</span></span>
<span><span class="co">## visualize different spatial networks on first field (~ layer 1)</span></span>
<span><span class="va">cell_metadata</span> <span class="op">=</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/pDataDT.html">pDataDT</a></span><span class="op">(</span><span class="va">SS_seqfish</span><span class="op">)</span></span>
<span><span class="va">field1_ids</span> <span class="op">=</span> <span class="va">cell_metadata</span><span class="op">[</span><span class="va">FOV</span> <span class="op">==</span> <span class="fl">0</span><span class="op">]</span><span class="op">$</span><span class="va">cell_ID</span></span>
<span><span class="va">subSS_seqfish</span> <span class="op">=</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/subsetGiotto.html">subsetGiotto</a></span><span class="op">(</span><span class="va">SS_seqfish</span>, cell_ids <span class="op">=</span> <span class="va">field1_ids</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/spatPlot.html">spatPlot</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">subSS_seqfish</span>, show_network <span class="op">=</span> <span class="cn">T</span>,</span>
<span>         network_color <span class="op">=</span> <span class="st">'blue'</span>, spatial_network_name <span class="op">=</span> <span class="st">'Delaunay_network'</span>,</span>
<span>         point_size <span class="op">=</span> <span class="fl">2.5</span>, cell_color <span class="op">=</span> <span class="st">'cell_types'</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/seqfish_cortex/22-spatPlot2D.png"></p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/spatPlot.html">spatPlot</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">subSS_seqfish</span>, show_network <span class="op">=</span> <span class="cn">T</span>,</span>
<span>         network_color <span class="op">=</span> <span class="st">'blue'</span>, spatial_network_name <span class="op">=</span> <span class="st">'spatial_network'</span>,</span>
<span>         point_size <span class="op">=</span> <span class="fl">2.5</span>, cell_color <span class="op">=</span> <span class="st">'cell_types'</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/seqfish_cortex/23-spatPlot2D.png"></p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/spatPlot.html">spatPlot</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">subSS_seqfish</span>, show_network <span class="op">=</span> <span class="cn">T</span>,</span>
<span>         network_color <span class="op">=</span> <span class="st">'blue'</span>, spatial_network_name <span class="op">=</span> <span class="st">'large_network'</span>,</span>
<span>         point_size <span class="op">=</span> <span class="fl">2.5</span>, cell_color <span class="op">=</span> <span class="st">'cell_types'</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/seqfish_cortex/24-spatPlot2D.png"></p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/spatPlot.html">spatPlot</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">subSS_seqfish</span>, show_network <span class="op">=</span> <span class="cn">T</span>,</span>
<span>         network_color <span class="op">=</span> <span class="st">'blue'</span>, spatial_network_name <span class="op">=</span> <span class="st">'distance_network'</span>,</span>
<span>         point_size <span class="op">=</span> <span class="fl">2.5</span>, cell_color <span class="op">=</span> <span class="st">'cell_types'</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/seqfish_cortex/25-spatPlot2D.png"></p>
</div>
<div class="section level2">
<h2 id="spatial-genes">10. Spatial Genes<a class="anchor" aria-label="anchor" href="#spatial-genes"></a>
</h2>
<div class="section level3">
<h3 id="individual-spatial-genes">Individual spatial genes<a class="anchor" aria-label="anchor" href="#individual-spatial-genes"></a>
</h3>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## 3 new methods to identify spatial genes</span></span>
<span><span class="va">km_spatialfeats</span> <span class="op">=</span> <span class="fu"><a href="../reference/binSpect.html">binSpect</a></span><span class="op">(</span><span class="va">SS_seqfish</span><span class="op">)</span></span>
<span><span class="fu">spatGenePlot</span><span class="op">(</span><span class="va">SS_seqfish</span>, expression_values <span class="op">=</span> <span class="st">'scaled'</span>, genes <span class="op">=</span> <span class="va">km_spatialfeats</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span><span class="op">$</span><span class="va">feats</span>,</span>
<span>             point_shape <span class="op">=</span> <span class="st">'border'</span>, point_border_stroke <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>             show_network <span class="op">=</span> <span class="cn">F</span>, network_color <span class="op">=</span> <span class="st">'lightgrey'</span>, point_size <span class="op">=</span> <span class="fl">2.5</span>, </span>
<span>             cow_n_col <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/seqfish_cortex/26-spatGenePlot2D.png"></p>
</div>
<div class="section level3">
<h3 id="spatial-genes-co-expression-modules">Spatial Genes Co-Expression Modules<a class="anchor" aria-label="anchor" href="#spatial-genes-co-expression-modules"></a>
</h3>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## spatial co-expression patterns ##</span></span>
<span><span class="va">ext_spatial_genes</span> <span class="op">=</span> <span class="va">km_spatialfeats</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">500</span><span class="op">]</span><span class="op">$</span><span class="va">feats</span></span>
<span><span class="co">## 1. calculate gene spatial correlation and single-cell correlation </span></span>
<span><span class="co">## create spatial correlation object</span></span>
<span><span class="va">spat_cor_netw_DT</span> <span class="op">=</span> <span class="fu"><a href="../reference/detectSpatialCorFeats.html">detectSpatialCorFeats</a></span><span class="op">(</span><span class="va">SS_seqfish</span>, </span>
<span>                                         method <span class="op">=</span> <span class="st">'network'</span>,</span>
<span>                                         spatial_network_name <span class="op">=</span> <span class="st">'Delaunay_network'</span>,</span>
<span>                                         subset_feats <span class="op">=</span> <span class="va">ext_spatial_genes</span><span class="op">)</span></span>
<span><span class="co">## 2. cluster correlated genes &amp; visualize</span></span>
<span><span class="va">spat_cor_netw_DT</span> <span class="op">=</span> <span class="fu"><a href="../reference/clusterSpatialCorFeats.html">clusterSpatialCorFeats</a></span><span class="op">(</span><span class="va">spat_cor_netw_DT</span>, </span>
<span>                                          name <span class="op">=</span> <span class="st">'spat_netw_clus'</span>, </span>
<span>                                          k <span class="op">=</span> <span class="fl">8</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/heatmSpatialCorFeats.html">heatmSpatialCorFeats</a></span><span class="op">(</span><span class="va">SS_seqfish</span>, spatCorObject <span class="op">=</span> <span class="va">spat_cor_netw_DT</span>, use_clus_name <span class="op">=</span> <span class="st">'spat_netw_clus'</span>, </span>
<span>                     heatmap_legend_param <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>title <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/seqfish_cortex/27-heatmSpatialCorFeats.png"></p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># 3. rank spatial correlated clusters and show genes for selected clusters</span></span>
<span><span class="va">netw_ranks</span> <span class="op">=</span> <span class="fu"><a href="../reference/rankSpatialCorGroups.html">rankSpatialCorGroups</a></span><span class="op">(</span><span class="va">SS_seqfish</span>, </span>
<span>                                  spatCorObject <span class="op">=</span> <span class="va">spat_cor_netw_DT</span>,</span>
<span>                                  use_clus_name <span class="op">=</span> <span class="st">'spat_netw_clus'</span><span class="op">)</span></span>
<span><span class="va">top_netw_spat_cluster</span> <span class="op">=</span> <span class="fu"><a href="../reference/showSpatialCorFeats.html">showSpatialCorFeats</a></span><span class="op">(</span><span class="va">spat_cor_netw_DT</span>, </span>
<span>                                            use_clus_name <span class="op">=</span> <span class="st">'spat_netw_clus'</span>,</span>
<span>                                            selected_clusters <span class="op">=</span> <span class="fl">6</span>,</span>
<span>                                            show_top_feats <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/seqfish_cortex/28-rankSpatialCorGroups.png"></p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># 4. create metagene enrichment score for clusters</span></span>
<span><span class="va">cluster_genes_DT</span> <span class="op">=</span> <span class="fu"><a href="../reference/showSpatialCorFeats.html">showSpatialCorFeats</a></span><span class="op">(</span><span class="va">spat_cor_netw_DT</span>, </span>
<span>                                       use_clus_name <span class="op">=</span> <span class="st">'spat_netw_clus'</span>,</span>
<span>                                       show_top_feats <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">cluster_genes</span> <span class="op">=</span> <span class="va">cluster_genes_DT</span><span class="op">$</span><span class="va">clus</span>; <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">cluster_genes</span><span class="op">)</span> <span class="op">=</span> <span class="va">cluster_genes_DT</span><span class="op">$</span><span class="va">feat_ID</span></span>
<span><span class="va">SS_seqfish</span> <span class="op">=</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/createMetafeats.html">createMetafeats</a></span><span class="op">(</span><span class="va">SS_seqfish</span>, </span>
<span>                             feat_clusters <span class="op">=</span> <span class="va">cluster_genes</span>, </span>
<span>                             name <span class="op">=</span> <span class="st">'cluster_metagene'</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/spatCellPlot.html">spatCellPlot</a></span><span class="op">(</span><span class="va">SS_seqfish</span>,</span>
<span>             spat_enr_names <span class="op">=</span> <span class="st">'cluster_metagene'</span>,</span>
<span>             cell_annotation_values <span class="op">=</span> <span class="va">netw_ranks</span><span class="op">$</span><span class="va">clusters</span>,</span>
<span>             point_size <span class="op">=</span> <span class="fl">1.5</span>, cow_n_col <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/seqfish_cortex/29-spatCellPlot2D.png"></p>
</div>
</div>
<div class="section level2">
<h2 id="hmrf-spatial-domains">11. HMRF Spatial Domains<a class="anchor" aria-label="anchor" href="#hmrf-spatial-domains"></a>
</h2>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hmrf_folder</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">my_working_dir</span>,<span class="st">'/'</span>,<span class="st">'11_HMRF/'</span><span class="op">)</span></span>
<span><span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.exists</a></span><span class="op">(</span><span class="va">hmrf_folder</span><span class="op">)</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="va">hmrf_folder</span>, recursive <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span><span class="va">my_spatial_genes</span> <span class="op">=</span> <span class="va">km_spatialfeats</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">100</span><span class="op">]</span><span class="op">$</span><span class="va">feats</span></span>
<span><span class="co"># do HMRF with different betas</span></span>
<span><span class="va">HMRF_spatial_genes</span> <span class="op">=</span> <span class="fu"><a href="../reference/doHMRF.html">doHMRF</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">SS_seqfish</span>, </span>
<span>                            expression_values <span class="op">=</span> <span class="st">'scaled'</span>,</span>
<span>                            spatial_genes <span class="op">=</span> <span class="va">my_spatial_genes</span>,</span>
<span>                            spatial_network_name <span class="op">=</span> <span class="st">'Delaunay_network'</span>,</span>
<span>                            k <span class="op">=</span> <span class="fl">9</span>,</span>
<span>                            betas <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">28</span>,<span class="fl">2</span>,<span class="fl">3</span><span class="op">)</span>, </span>
<span>                            output_folder <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">hmrf_folder</span>, <span class="st">'/'</span>, <span class="st">'Spatial_genes/SG_top100_k9_scaled'</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">## view results of HMRF</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">28</span>, <span class="fl">32</span>, by <span class="op">=</span> <span class="fl">2</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="../reference/viewHMRFresults2D.html">viewHMRFresults2D</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">SS_seqfish</span>,</span>
<span>                    HMRFoutput <span class="op">=</span> <span class="va">HMRF_spatial_genes</span>,</span>
<span>                    k <span class="op">=</span> <span class="fl">9</span>, betas_to_view <span class="op">=</span> <span class="va">i</span>,</span>
<span>                    point_size <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="co">## add HMRF of interest to giotto object</span></span>
<span><span class="va">SS_seqfish</span> <span class="op">=</span> <span class="fu"><a href="../reference/addHMRF.html">addHMRF</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">SS_seqfish</span>,</span>
<span>                     HMRFoutput <span class="op">=</span> <span class="va">HMRF_spatial_genes</span>,</span>
<span>                     k <span class="op">=</span> <span class="fl">9</span>, betas_to_add <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">28</span><span class="op">)</span>,</span>
<span>                     hmrf_name <span class="op">=</span> <span class="st">'HMRF_2'</span><span class="op">)</span></span>
<span><span class="co">## visualize</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/spatPlot.html">spatPlot</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">SS_seqfish</span>, </span>
<span>         cell_color <span class="op">=</span> <span class="st">'HMRF_2_k9_b.28'</span>, </span>
<span>         point_size <span class="op">=</span> <span class="fl">3</span>, </span>
<span>         coord_fix_ratio <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/seqfish_cortex/30-spatPlot2D.png"></p>
</div>
<div class="section level2">
<h2 id="cell-neighborhood-cell-typecell-type-interactions">12. Cell Neighborhood: Cell-Type/Cell-Type Interactions<a class="anchor" aria-label="anchor" href="#cell-neighborhood-cell-typecell-type-interactions"></a>
</h2>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cell_proximities</span> <span class="op">=</span> <span class="fu"><a href="../reference/cellProximityEnrichment.html">cellProximityEnrichment</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">SS_seqfish</span>,</span>
<span>                                           cluster_column <span class="op">=</span> <span class="st">'cell_types'</span>,</span>
<span>                                           spatial_network_name <span class="op">=</span> <span class="st">'Delaunay_network'</span>,</span>
<span>                                           adjust_method <span class="op">=</span> <span class="st">'fdr'</span>,</span>
<span>                                           number_of_simulations <span class="op">=</span> <span class="fl">2000</span><span class="op">)</span></span>
<span><span class="co">## barplot</span></span>
<span><span class="fu"><a href="../reference/cellProximityBarplot.html">cellProximityBarplot</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">SS_seqfish</span>,</span>
<span>                     CPscore <span class="op">=</span> <span class="va">cell_proximities</span>, </span>
<span>                     min_orig_ints <span class="op">=</span> <span class="fl">5</span>, min_sim_ints <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/seqfish_cortex/31-cellProximityBarplot.png"></p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## heatmap</span></span>
<span><span class="fu"><a href="../reference/cellProximityHeatmap.html">cellProximityHeatmap</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">SS_seqfish</span>, </span>
<span>                     CPscore <span class="op">=</span> <span class="va">cell_proximities</span>, </span>
<span>                     order_cell_types <span class="op">=</span> <span class="cn">T</span>, scale <span class="op">=</span> <span class="cn">T</span>,</span>
<span>                     color_breaks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1.5</span>, <span class="fl">0</span>, <span class="fl">1.5</span><span class="op">)</span>, </span>
<span>                     color_names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'blue'</span>, <span class="st">'white'</span>, <span class="st">'red'</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/seqfish_cortex/32-cellProximityHeatmap.png"></p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## network</span></span>
<span><span class="fu"><a href="../reference/cellProximityNetwork.html">cellProximityNetwork</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">SS_seqfish</span>, </span>
<span>                     CPscore <span class="op">=</span> <span class="va">cell_proximities</span>, remove_self_edges <span class="op">=</span> <span class="cn">T</span>,</span>
<span>                     only_show_enrichment_edges <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/seqfish_cortex/33-cellProximityNetwork.png"></p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## network with self-edges</span></span>
<span><span class="fu"><a href="../reference/cellProximityNetwork.html">cellProximityNetwork</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">SS_seqfish</span>, CPscore <span class="op">=</span> <span class="va">cell_proximities</span>,</span>
<span>                     remove_self_edges <span class="op">=</span> <span class="cn">F</span>, self_loop_strength <span class="op">=</span> <span class="fl">0.3</span>,</span>
<span>                     only_show_enrichment_edges <span class="op">=</span> <span class="cn">F</span>,</span>
<span>                     rescale_edge_weights <span class="op">=</span> <span class="cn">T</span>,</span>
<span>                     node_size <span class="op">=</span> <span class="fl">8</span>,</span>
<span>                     edge_weight_range_depletion <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span>,</span>
<span>                     edge_weight_range_enrichment <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fl">5</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/seqfish_cortex/12_d_network_cell_cell_enrichment_self.png"></p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## visualization of specific cell types</span></span>
<span><span class="co"># Option 1</span></span>
<span><span class="va">spec_interaction</span> <span class="op">=</span> <span class="st">"astrocytes--Olig"</span></span>
<span><span class="fu"><a href="../reference/cellProximitySpatPlot2D.html">cellProximitySpatPlot2D</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">SS_seqfish</span>,</span>
<span>                        interaction_name <span class="op">=</span> <span class="va">spec_interaction</span>,</span>
<span>                        show_network <span class="op">=</span> <span class="cn">T</span>,</span>
<span>                        cluster_column <span class="op">=</span> <span class="st">'cell_types'</span>,</span>
<span>                        cell_color <span class="op">=</span> <span class="st">'cell_types'</span>,</span>
<span>                        cell_color_code <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>astrocytes <span class="op">=</span> <span class="st">'lightblue'</span>, Olig <span class="op">=</span> <span class="st">'red'</span><span class="op">)</span>,</span>
<span>                        point_size_select <span class="op">=</span> <span class="fl">4</span>, point_size_other <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/seqfish_cortex/34-cellProximitySpatPlot2D.png"></p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Option 2: create additional metadata</span></span>
<span><span class="va">SS_seqfish</span> <span class="op">=</span> <span class="fu"><a href="../reference/addCellIntMetadata.html">addCellIntMetadata</a></span><span class="op">(</span><span class="va">SS_seqfish</span>, </span>
<span>                                spatial_network <span class="op">=</span> <span class="st">'spatial_network'</span>,</span>
<span>                                cluster_column <span class="op">=</span> <span class="st">'cell_types'</span>,</span>
<span>                                cell_interaction <span class="op">=</span> <span class="va">spec_interaction</span>,</span>
<span>                                name <span class="op">=</span> <span class="st">'astro_olig_ints'</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/spatPlot.html">spatPlot</a></span><span class="op">(</span><span class="va">SS_seqfish</span>, cell_color <span class="op">=</span> <span class="st">'astro_olig_ints'</span>,</span>
<span>         select_cell_groups <span class="op">=</span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'other_astrocytes'</span>, <span class="st">'other_Olig'</span>, <span class="st">'select_astrocytes'</span>, <span class="st">'select_Olig'</span><span class="op">)</span>,</span>
<span>         legend_symbol_size <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/seqfish_cortex/35-spatPlot2D.png"></p>
</div>
<div class="section level2">
<h2 id="part-13--cell-neighborhood-interaction-changed-features">Part 13. Cell Neighborhood: Interaction Changed Features<a class="anchor" aria-label="anchor" href="#part-13--cell-neighborhood-interaction-changed-features"></a>
</h2>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://future.futureverse.org" class="external-link">future</a></span><span class="op">)</span></span>
<span><span class="co">## select top 25th highest expressing genes</span></span>
<span><span class="va">gene_metadata</span> <span class="op">=</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/fDataDT.html">fDataDT</a></span><span class="op">(</span><span class="va">SS_seqfish</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">gene_metadata</span><span class="op">$</span><span class="va">nr_cells</span>, <span class="va">gene_metadata</span><span class="op">$</span><span class="va">mean_expr</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">gene_metadata</span><span class="op">$</span><span class="va">nr_cells</span>, <span class="va">gene_metadata</span><span class="op">$</span><span class="va">mean_expr_det</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/quantile.html" class="external-link">quantile</a></span><span class="op">(</span><span class="va">gene_metadata</span><span class="op">$</span><span class="va">mean_expr_det</span><span class="op">)</span></span>
<span><span class="va">high_expressed_genes</span> <span class="op">=</span> <span class="va">gene_metadata</span><span class="op">[</span><span class="va">mean_expr_det</span> <span class="op">&gt;</span> <span class="fl">3.5</span><span class="op">]</span><span class="op">$</span><span class="va">feat_ID</span></span>
<span><span class="co">## identify genes that are associated with proximity to other cell types</span></span>
<span><span class="fu"><a href="https://future.futureverse.org/reference/plan.html" class="external-link">plan</a></span><span class="op">(</span><span class="st">'multisession'</span>, workers <span class="op">=</span> <span class="fl">6</span><span class="op">)</span></span>
<span><span class="va">ICFsForesHighGenes</span> <span class="op">=</span>  <span class="fu"><a href="../reference/findInteractionChangedFeats.html">findInteractionChangedFeats</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">SS_seqfish</span>,</span>
<span>                              selected_feats <span class="op">=</span> <span class="va">high_expressed_genes</span>,</span>
<span>                              spatial_network_name <span class="op">=</span> <span class="st">'Delaunay_network'</span>,</span>
<span>                              cluster_column <span class="op">=</span> <span class="st">'cell_types'</span>,</span>
<span>                              diff_test <span class="op">=</span> <span class="st">'permutation'</span>,</span>
<span>                              adjust_method <span class="op">=</span> <span class="st">'fdr'</span>,</span>
<span>                              nr_permutations <span class="op">=</span> <span class="fl">2000</span>, </span>
<span>                              do_parallel <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span><span class="co">## visualize all genes</span></span>
<span><span class="fu"><a href="../reference/plotCellProximityFeats.html">plotCellProximityFeats</a></span><span class="op">(</span><span class="va">SS_seqfish</span>, icfObject <span class="op">=</span> <span class="va">ICFscoresHighGenes</span>, </span>
<span>                       method <span class="op">=</span> <span class="st">'dotplot'</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/seqfish_cortex/36-plotCellProximityGenes.png"></p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## filter genes</span></span>
<span><span class="va">ICFscoresFilt</span> <span class="op">=</span> <span class="fu"><a href="../reference/filterICF.html">filterICF</a></span><span class="op">(</span><span class="va">ICFscoresHighGenes</span><span class="op">)</span></span>
<span><span class="co">## visualize subset of interaction changed genes (ICFs)</span></span>
<span><span class="va">ICF_genes</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'Jakmip1'</span>, <span class="st">'Golgb1'</span>, <span class="st">'Dact2'</span>, <span class="st">'Ddx27'</span>, <span class="st">'Abl1'</span>, <span class="st">'Zswim8'</span><span class="op">)</span></span>
<span><span class="va">ICF_genes_types</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'Lhx6 iNeuron'</span>, <span class="st">'Lhx6 iNeuron'</span>, <span class="st">'L4 eNeuron'</span>, <span class="st">'L4 eNeuron'</span>, <span class="st">'astrocytes'</span>, <span class="st">'astrocytes'</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">ICF_genes</span><span class="op">)</span> <span class="op">=</span> <span class="va">ICF_genes_types</span></span>
<span><span class="fu"><a href="../reference/plotICF.html">plotICF</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">SS_seqfish</span>,</span>
<span>        icfObject <span class="op">=</span> <span class="va">ICFscoresHighGenes</span>,</span>
<span>        source_type <span class="op">=</span> <span class="st">'endothelial'</span>,</span>
<span>        source_markers <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'Pltp'</span>, <span class="st">'Cldn5'</span>, <span class="st">'Apcdd1'</span><span class="op">)</span>,</span>
<span>        ICF_feats <span class="op">=</span> <span class="va">ICF_genes</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/seqfish_cortex/37-plotICF.png"></p>
</div>
<div class="section level2">
<h2 id="cell-neighborhood-ligand-receptor-cell-cell-communication">14. Cell Neighborhood: Ligand-Receptor Cell-Cell Communication<a class="anchor" aria-label="anchor" href="#cell-neighborhood-ligand-receptor-cell-cell-communication"></a>
</h2>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## LR expression</span></span>
<span><span class="co">## LR activity changes</span></span>
<span><span class="va">LR_data</span> <span class="op">=</span> <span class="fu">data.table</span><span class="fu">::</span><span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"mouse_ligand_receptors.txt"</span>, package <span class="op">=</span> <span class="st">'Giotto'</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">LR_data</span><span class="op">[</span>, <span class="va">ligand_det</span> <span class="op">:=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">LR_data</span><span class="op">$</span><span class="va">mouseLigand</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">SS_seqfish</span><span class="op">@</span><span class="va">feat_ID</span><span class="op">$</span><span class="va">rna</span>, <span class="cn">T</span>, <span class="cn">F</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">LR_data</span><span class="op">[</span>, <span class="va">receptor_det</span> <span class="op">:=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">LR_data</span><span class="op">$</span><span class="va">mouseReceptor</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">SS_seqfish</span><span class="op">@</span><span class="va">feat_ID</span><span class="op">$</span><span class="va">rna</span>, <span class="cn">T</span>, <span class="cn">F</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">LR_data_det</span> <span class="op">=</span> <span class="va">LR_data</span><span class="op">[</span><span class="va">ligand_det</span> <span class="op">==</span> <span class="cn">T</span> <span class="op">&amp;</span> <span class="va">receptor_det</span> <span class="op">==</span> <span class="cn">T</span><span class="op">]</span></span>
<span><span class="va">select_ligands</span> <span class="op">=</span> <span class="va">LR_data_det</span><span class="op">$</span><span class="va">mouseLigand</span></span>
<span><span class="va">select_receptors</span> <span class="op">=</span> <span class="va">LR_data_det</span><span class="op">$</span><span class="va">mouseReceptor</span></span>
<span><span class="co">## get statistical significance of gene pair expression changes based on expression</span></span>
<span><span class="va">expr_only_scores</span> <span class="op">=</span> <span class="fu"><a href="../reference/exprCellCellcom.html">exprCellCellcom</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">SS_seqfish</span>,</span>
<span>                                   cluster_column <span class="op">=</span> <span class="st">'cell_types'</span>, </span>
<span>                                   random_iter <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>                                   feat_set_1 <span class="op">=</span> <span class="va">select_ligands</span>,</span>
<span>                                   feat_set_2 <span class="op">=</span> <span class="va">select_receptors</span>, </span>
<span>                                   verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="co">## get statistical significance of gene pair expression changes upon cell-cell interaction</span></span>
<span><span class="va">spatial_all_scores</span> <span class="op">=</span> <span class="fu"><a href="../reference/spatCellCellcom.html">spatCellCellcom</a></span><span class="op">(</span><span class="va">SS_seqfish</span>,</span>
<span>                                     spatial_network_name <span class="op">=</span> <span class="st">'spatial_network'</span>,</span>
<span>                                     cluster_column <span class="op">=</span> <span class="st">'cell_types'</span>, </span>
<span>                                     random_iter <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>                                     feat_set_1 <span class="op">=</span> <span class="va">select_ligands</span>,</span>
<span>                                     feat_set_2 <span class="op">=</span> <span class="va">select_receptors</span>,</span>
<span>                                     adjust_method <span class="op">=</span> <span class="st">'fdr'</span>,</span>
<span>                                     do_parallel <span class="op">=</span> <span class="cn">T</span>,</span>
<span>                                     cores <span class="op">=</span> <span class="fl">4</span>,</span>
<span>                                     verbose <span class="op">=</span> <span class="st">'a little'</span><span class="op">)</span></span>
<span><span class="co">## select top LR ##</span></span>
<span><span class="va">selected_spat</span> <span class="op">=</span> <span class="va">spatial_all_scores</span><span class="op">[</span><span class="va">p.adj</span> <span class="op">&lt;=</span> <span class="fl">0.01</span> <span class="op">&amp;</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">log2fc</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0.25</span> <span class="op">&amp;</span> <span class="va">lig_nr</span> <span class="op">&gt;=</span> <span class="fl">4</span> <span class="op">&amp;</span> <span class="va">rec_nr</span> <span class="op">&gt;=</span> <span class="fl">4</span><span class="op">]</span></span>
<span><span class="fu">data.table</span><span class="fu">::</span><span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/setorder.html" class="external-link">setorder</a></span><span class="op">(</span><span class="va">selected_spat</span>, <span class="op">-</span><span class="va">PI</span><span class="op">)</span></span>
<span><span class="va">top_LR_ints</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">selected_spat</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">PI</span><span class="op">)</span><span class="op">)</span><span class="op">]</span><span class="op">$</span><span class="va">LR_comb</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">33</span><span class="op">]</span></span>
<span><span class="va">top_LR_cell_ints</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">selected_spat</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">PI</span><span class="op">)</span><span class="op">)</span><span class="op">]</span><span class="op">$</span><span class="va">LR_cell_comb</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">33</span><span class="op">]</span></span>
<span><span class="fu"><a href="../reference/plotCCcomDotplot.html">plotCCcomDotplot</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">SS_seqfish</span>,</span>
<span>                 comScores <span class="op">=</span> <span class="va">spatial_all_scores</span>,</span>
<span>                 selected_LR <span class="op">=</span> <span class="va">top_LR_ints</span>,</span>
<span>                 selected_cell_LR <span class="op">=</span> <span class="va">top_LR_cell_ints</span>,</span>
<span>                 cluster_on <span class="op">=</span> <span class="st">'PI'</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/seqfish_cortex/38-plotCCcomDotplot.png"></p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## spatial vs rank ####</span></span>
<span><span class="va">comb_comm</span> <span class="op">=</span> <span class="fu"><a href="../reference/combCCcom.html">combCCcom</a></span><span class="op">(</span>spatialCC <span class="op">=</span> <span class="va">spatial_all_scores</span>,</span>
<span>                      exprCC <span class="op">=</span> <span class="va">expr_only_scores</span><span class="op">)</span></span>
<span><span class="co">## highest levels of ligand and receptor prediction</span></span>
<span><span class="co">## top differential activity levels for ligand receptor pairs</span></span>
<span><span class="fu"><a href="../reference/plotRankSpatvsExpr.html">plotRankSpatvsExpr</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">SS_seqfish</span>,</span>
<span>                   <span class="va">comb_comm</span>,</span>
<span>                   expr_rnk_column <span class="op">=</span> <span class="st">'LR_expr_rnk'</span>,</span>
<span>                   spat_rnk_column <span class="op">=</span> <span class="st">'LR_spat_rnk'</span>,</span>
<span>                   midpoint <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/seqfish_cortex/39-plotRankSpatvsExpr.png"></p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## recovery</span></span>
<span><span class="fu"><a href="../reference/plotRecovery.html">plotRecovery</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">SS_seqfish</span>,</span>
<span>             <span class="va">comb_comm</span>,</span>
<span>             expr_rnk_column <span class="op">=</span> <span class="st">'LR_expr_rnk'</span>,</span>
<span>             spat_rnk_column <span class="op">=</span> <span class="st">'LR_spat_rnk'</span>,</span>
<span>             ground_truth <span class="op">=</span> <span class="st">'spatial'</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/seqfish_cortex/40-plotRecovery.png"></p>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by <a href="https://www.drieslab.com" class="external-link">Ruben Dries</a>, Jiaji George Chen, Natalie Del Rossi, <a href="https://labs.icahn.mssm.edu/yuanlab/" class="external-link">Guo-Cheng Yuan</a>, Matthew Oâ€™Brien, <a href="https://josschavezf.github.io" class="external-link">Joselyn C. ChÃ¡vez-Fuentes</a>, Edward Ruiz.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
