<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="Giotto">
<title>Stereo-Seq MOSTA E12.5_E1S3 • Giotto</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><link href="../deps/JetBrains_Mono-0.4.7/font.css" rel="stylesheet">
<!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Stereo-Seq MOSTA E12.5_E1S3">
<meta property="og:description" content="Giotto">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-dark"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">Giotto</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">3.3.2</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Documentation</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-tutorials">Tutorials</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-tutorials">
    <h6 class="dropdown-header" data-toc-skip>Stereo-seq</h6>
    <a class="dropdown-item" href="../articles/stereoseq_mosta.html">MOSTA E12.5_E1S3</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Seq-Scope</h6>
    <a class="dropdown-item" href="../articles/seqscope_mouse_liver.html">Mouse liver</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Visium</h6>
    <a class="dropdown-item" href="../articles/visium_mouse_kidney.html">Visium Mouse Kidney</a>
    <a class="dropdown-item" href="../articles/visium_mouse_brain.html">Visium Mouse Brain</a>
    <a class="dropdown-item" href="../articles/visium_prostate_integration.html">Visium Prostate Integration</a>
    <a class="dropdown-item" href="../articles/visium_cytassist_lungcancer.html">CytAssist Human Lung Cancer</a>
    <a class="dropdown-item" href="../articles/visium_cytassist_human_glioblastoma.html">CytAssist Multi-omics Human Glioblastoma</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Single Cell RNA Sequencing</h6>
    <a class="dropdown-item" href="../articles/singlecell_prostate_standard.html">Prostate</a>
    <a class="dropdown-item" href="../articles/singlecell_prostate_integration.html">Prostate Integration</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">News</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/drieslab/Giotto/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Stereo-Seq MOSTA E12.5_E1S3</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/drieslab/Giotto/blob/HEAD/vignettes/stereoseq_mosta.Rmd" class="external-link"><code>vignettes/stereoseq_mosta.Rmd</code></a></small>
      <div class="d-none name"><code>stereoseq_mosta.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="dataset-explanation">Dataset explanation<a class="anchor" aria-label="anchor" href="#dataset-explanation"></a>
</h2>
<p>Stereo-seq is a sequencing-based spatial transcriptomics technology
that was developed and used by <a href="https://doi.org/10.1016/j.cell.2022.04.003" class="external-link">Chen et al. in
2022</a> to generate the Mouse Organogenesis Spatiotemporal
Transcriptomic Atlas (MOSTA). This tutorial demonstrates how to
interactively filter and analyze the third sagittal section of a mouse
embryo at embryonic day 12.5 (“E12.5_E1S3”) from MOSTA.</p>
</div>
<div class="section level2">
<h2 id="pre-processing">Pre-processing<a class="anchor" aria-label="anchor" href="#pre-processing"></a>
</h2>
<p>The necessary input files for sample E12.5_E1S3 are publicly
available and were downloaded from the following original sources:</p>
<ol style="list-style-type: decimal">
<li>FASTQ files (<a href="https://db.cngb.org/search/sample/?q=CNP0001543" class="external-link uri">https://db.cngb.org/search/sample/?q=CNP0001543</a>)</li>
<li>barcodeToPos.h5 file (<a href="https://db.cngb.org/stomics/mosta/download.html" class="external-link uri">https://db.cngb.org/stomics/mosta/download.html</a>)</li>
</ol>
<p>The STOMICS Analysis Workflow <a href="https://github.com/BGIResearch/SAW" class="external-link">SAW pipeline</a> was used to
process these files. The output of the SAW pipeline is an .h5ad file at
a specific bin size. One bin of size n represents an n x n square of
aggregated spatial barcodes. In this tutorial, a sample with a bin size
of 200 was used.</p>
</div>
<div class="section level2">
<h2 id="start-giotto">Start Giotto<a class="anchor" aria-label="anchor" href="#start-giotto"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Ensure Giotto Suite is installed</span></span>
<span><span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="st">"Giotto"</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/installed.packages.html" class="external-link">installed.packages</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"drieslab/Giotto@suite"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://drieslab.github.io/Giotto/">Giotto</a></span><span class="op">)</span></span>
<span><span class="co"># Ensure the Python environment for Giotto has been installed</span></span>
<span><span class="va">genv_exists</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/GiottoClass/man/checkGiottoEnvironment.html" class="external-link">checkGiottoEnvironment</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="va">genv_exists</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="co"># The following command need only be run once to install the Giotto environment.</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/GiottoClass/man/installGiottoEnvironment.html" class="external-link">installGiottoEnvironment</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="create-a-giotto-object">1. Create a Giotto object<a class="anchor" aria-label="anchor" href="#create-a-giotto-object"></a>
</h2>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># download E12.5_E1S3_bin200.h5ad output from SAW pipeline (509.4 MB)</span></span>
<span><span class="co"># alternatively, specify path to *.h5ad output of SAW pipeline </span></span>
<span><span class="va">anndata_download</span> <span class="op">=</span> <span class="st">"https://zenodo.org/record/7323947/files/E12.5_E1S3_bin200.h5ad?download=1"</span></span>
<span><span class="va">anndata_file</span> <span class="op">=</span> <span class="st">"E12.5_E1S3_bin_200.h5ad"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span><span class="va">anndata_download</span>, <span class="va">anndata_file</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># convert anndata file to giotto object</span></span>
<span><span class="va">stereo_go</span> <span class="op">&lt;-</span> <span class="fu">Giotto</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/GiottoClass/man/anndataToGiotto.html" class="external-link">anndataToGiotto</a></span><span class="op">(</span><span class="va">anndata_file</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># alternatively, specify path to *.gef output of SAW pipeline (requires Giotto v3.2.0 or higher)</span></span>
<span><span class="co"># gef_file = "E12.5_E1S3.gef"</span></span>
<span><span class="co"># stereo_go &lt;- Giotto::gefToGiotto(gef_file, bin_size = "bin200")</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="process-giotto-object">2. Process Giotto object<a class="anchor" aria-label="anchor" href="#process-giotto-object"></a>
</h2>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># filter number of genes</span></span>
<span><span class="co"># important to discard bins (aggregated barcodes) outside of embryo</span></span>
<span><span class="va">stereo_go</span> <span class="op">&lt;-</span> <span class="va">stereo_go</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="../reference/filterGiotto.html">filterGiotto</a></span><span class="op">(</span>expression_threshold <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                                        feat_det_in_min_cells <span class="op">=</span> <span class="fl">5</span>,</span>
<span>                                        min_det_feats_per_cell <span class="op">=</span> <span class="fl">750</span><span class="op">)</span></span>
<span><span class="co"># normalize</span></span>
<span><span class="va">stereo_go</span> <span class="op">&lt;-</span> <span class="va">stereo_go</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="../reference/normalizeGiotto.html">normalizeGiotto</a></span><span class="op">(</span>scalefactor <span class="op">=</span> <span class="fl">5000</span>, verbose <span class="op">=</span> <span class="cn">T</span><span class="op">)</span> </span>
<span></span>
<span><span class="co"># add statistics</span></span>
<span><span class="va">stereo_go</span> <span class="op">&lt;-</span> <span class="va">stereo_go</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="../reference/addStatistics.html">addStatistics</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># make plot</span></span>
<span><span class="co"># each dot here represents a 200x200 aggregation of spatial barcodes (bin size 200)</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/spatPlot.html">spatPlot2D</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">stereo_go</span>, cell_color <span class="op">=</span> <span class="st">"nr_feats"</span>, color_as_factor <span class="op">=</span> <span class="cn">F</span>, point_size <span class="op">=</span> <span class="fl">1.5</span>, show_plot <span class="op">=</span> <span class="cn">T</span>, save_plot <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/stereoseq_mosta/1.png"></p>
</div>
<div class="section level2">
<h2 id="dimension-reduction">3. Dimension reduction<a class="anchor" aria-label="anchor" href="#dimension-reduction"></a>
</h2>
<ul>
<li>Identify highly variable features (HVF)</li>
</ul>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">stereo_go</span> <span class="op">&lt;-</span> <span class="va">stereo_go</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="../reference/calculateHVF.html">calculateHVF</a></span><span class="op">(</span>zscore_threshold <span class="op">=</span> <span class="fl">1</span>, show_plot <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span></code></pre></div>
<ul>
<li>Perform PCA</li>
<li>Identify the number of significant principal components (PCs)</li>
</ul>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">stereo_go</span> <span class="op">&lt;-</span> <span class="va">stereo_go</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="../reference/runPCA.html">runPCA</a></span><span class="op">(</span>expression_values <span class="op">=</span> <span class="st">'scaled'</span>, feats_to_use <span class="op">=</span> <span class="st">'hvf'</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/screePlot.html">screePlot</a></span><span class="op">(</span><span class="va">stereo_go</span>, ncp <span class="op">=</span> <span class="fl">30</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/plotPCA.html">plotPCA</a></span><span class="op">(</span><span class="va">stereo_go</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/stereoseq_mosta/2.png"></p>
<p><img src="images/stereoseq_mosta/3.png"></p>
<ul>
<li>Run UMAP and TSNE on PCs (or directly on matrix)</li>
</ul>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">stereo_go</span> <span class="op">&lt;-</span> <span class="va">stereo_go</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="../reference/runUMAP.html">runUMAP</a></span><span class="op">(</span>dimensions_to_use <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span>, n_threads <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># plot UMAP, coloring cells/points based on nr_feats</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/plotUMAP.html">plotUMAP</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">stereo_go</span>,</span>
<span>         cell_color <span class="op">=</span> <span class="st">'nr_feats'</span>, color_as_factor <span class="op">=</span> <span class="cn">F</span>, point_size <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/stereoseq_mosta/4.png"></p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">stereo_go</span> <span class="op">=</span> <span class="va">stereo_go</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="../reference/runtSNE.html">runtSNE</a></span><span class="op">(</span>dimensions_to_use <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/plotTSNE.html">plotTSNE</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">stereo_go</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/stereoseq_mosta/5.png"></p>
</div>
<div class="section level2">
<h2 id="clustering">4. Clustering<a class="anchor" aria-label="anchor" href="#clustering"></a>
</h2>
<ul>
<li>Create a shared (default) nearest network in PCA space (or directly
on matrix)</li>
<li>Cluster on nearest network with Leiden or Louvan (kmeans and hclust
are alternatives)</li>
</ul>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># sNN network (default)</span></span>
<span><span class="va">stereo_go</span> <span class="op">&lt;-</span> <span class="va">stereo_go</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/pkg/GiottoClass/man/createNearestNetwork.html" class="external-link">createNearestNetwork</a></span><span class="op">(</span>dimensions_to_use <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span>, k <span class="op">=</span> <span class="fl">12</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># leiden clustering</span></span>
<span><span class="va">stereo_go</span> <span class="op">&lt;-</span> <span class="va">stereo_go</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="../reference/doLeidenCluster.html">doLeidenCluster</a></span><span class="op">(</span>resolution <span class="op">=</span> <span class="fl">1</span>, n_iterations <span class="op">=</span> <span class="fl">1000</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/plotUMAP.html">plotUMAP</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">stereo_go</span>, cell_color <span class="op">=</span> <span class="st">'leiden_clus'</span>, point_size <span class="op">=</span> <span class="fl">2.5</span>,</span>
<span>         show_NN_network <span class="op">=</span> <span class="cn">F</span>, edge_alpha <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># merge small groups based on similarity</span></span>
<span><span class="va">leiden_similarities</span> <span class="op">=</span> <span class="va">stereo_go</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="../reference/getClusterSimilarity.html">getClusterSimilarity</a></span><span class="op">(</span>expression_values <span class="op">=</span> <span class="st">'scaled'</span>,</span>
<span>                                                         cluster_column <span class="op">=</span> <span class="st">'leiden_clus'</span><span class="op">)</span></span>
<span></span>
<span><span class="va">stereo_go</span> <span class="op">=</span> <span class="va">stereo_go</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="../reference/mergeClusters.html">mergeClusters</a></span><span class="op">(</span>expression_values <span class="op">=</span> <span class="st">'scaled'</span>,</span>
<span>                                        cluster_column <span class="op">=</span> <span class="st">'leiden_clus'</span>,</span>
<span>                                        new_cluster_name <span class="op">=</span> <span class="st">'leiden_clus_m'</span>,</span>
<span>                                        max_group_size <span class="op">=</span> <span class="fl">100</span>,</span>
<span>                                        force_min_group_size <span class="op">=</span> <span class="fl">25</span>,</span>
<span>                                        max_sim_clusters <span class="op">=</span> <span class="fl">10</span>,</span>
<span>                                        min_cor_score <span class="op">=</span> <span class="fl">0.7</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/plotUMAP.html">plotUMAP</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">stereo_go</span>, cell_color <span class="op">=</span> <span class="st">'leiden_clus_m'</span>, point_size <span class="op">=</span> <span class="fl">2.5</span>,</span>
<span>         show_NN_network <span class="op">=</span> <span class="cn">F</span>, edge_alpha <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/stereoseq_mosta/6.png"></p>
<p><img src="images/stereoseq_mosta/7.png"></p>
</div>
<div class="section level2">
<h2 id="co-visualization">5. Co-visualization<a class="anchor" aria-label="anchor" href="#co-visualization"></a>
</h2>
<ul>
<li>Co-visualize expression UMAP and spatial data clusters</li>
</ul>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/spatDimPlot.html">spatDimPlot2D</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">stereo_go</span>, cell_color <span class="op">=</span> <span class="st">'leiden_clus_m'</span>,</span>
<span>              dim_point_size <span class="op">=</span> <span class="fl">1.5</span>, spat_point_size <span class="op">=</span> <span class="fl">1.5</span>,</span>
<span>              show_plot <span class="op">=</span> <span class="cn">T</span>, return_plot <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/stereoseq_mosta/8.png"></p>
</div>
<div class="section level2">
<h2 id="spatial-genes">6. Spatial Genes<a class="anchor" aria-label="anchor" href="#spatial-genes"></a>
</h2>
<ul>
<li>Find genes with spatially coherent expression patterns</li>
</ul>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># create knn</span></span>
<span><span class="va">stereo_go</span> <span class="op">&lt;-</span> <span class="va">stereo_go</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/pkg/GiottoClass/man/createSpatialNetwork.html" class="external-link">createSpatialNetwork</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"kNN"</span>, k <span class="op">=</span> <span class="fl">8</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># select 100 random genes</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">144</span><span class="op">)</span></span>
<span><span class="va">featureMetadata</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/GiottoClass/man/fDataDT.html" class="external-link">fDataDT</a></span><span class="op">(</span><span class="va">stereo_go</span><span class="op">)</span> </span>
<span><span class="va">gene_list</span> <span class="op">=</span> <span class="va">featureMetadata</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">featureMetadata</span><span class="op">$</span><span class="va">feat_ID</span><span class="op">)</span>, <span class="fl">100</span><span class="op">)</span>, <span class="st">"feat_ID"</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># use binSpect method to find spatial genes</span></span>
<span><span class="va">spat_genes</span> <span class="op">&lt;-</span> <span class="va">stereo_go</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="../reference/binSpect.html">binSpect</a></span><span class="op">(</span>expression_values <span class="op">=</span> <span class="st">"scaled"</span>, </span>
<span>                                     subset_feats <span class="op">=</span> <span class="va">gene_list</span><span class="op">$</span><span class="va">feat_ID</span>,</span>
<span>                                     spatial_network_name <span class="op">=</span> <span class="st">"kNN_network"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="subsettingfiltering">7. Subsetting/Filtering<a class="anchor" aria-label="anchor" href="#subsettingfiltering"></a>
</h2>
<ul>
<li>Perform these steps to select an ROI using an interactive polygon
selection tool.</li>
<li>To draw a polygon on the interactive plot, click the mouse to start
a line segment. Click again to draw the endpoint of the segment, which
becomes the startpoint of the following segment. Click “Done” to close
the app and save the polygon coordinates.</li>
<li>See our tutorial on interactive selection/filtering in “Getting
started” to learn more.</li>
</ul>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">my_spatPlot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://drieslab.github.io/Giotto/reference/spatPlot.html">spatPlot2D</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">stereo_go</span>,</span>
<span>                          cell_color <span class="op">=</span> <span class="st">'leiden_clus'</span>,</span>
<span>                          color_as_factor <span class="op">=</span> <span class="cn">T</span>,</span>
<span>                          show_plot <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                          point_size <span class="op">=</span> <span class="fl">2</span>,</span>
<span>                          save_plot <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># create a polygon mask around a ROI, coordinates will be saved after clicking 'Done'</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://shiny.posit.co/" class="external-link">shiny</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">miniUI</span><span class="op">)</span></span>
<span><span class="va">my_polygon_coordinates</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotInteractivePolygons.html">plotInteractivePolygons</a></span><span class="op">(</span><span class="va">my_spatPlot</span>, height <span class="op">=</span> <span class="fl">500</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># create new giotto object from polygon coordinates</span></span>
<span><span class="va">lasso_polygons</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GiottoClass/man/createGiottoPolygonsFromDfr.html" class="external-link">createGiottoPolygonsFromDfr</a></span><span class="op">(</span><span class="va">my_polygon_coordinates</span>, </span>
<span>                                              name <span class="op">=</span> <span class="st">"cell"</span>, </span>
<span>                                              calc_centroids <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># store the polygons info within the giotto object</span></span>
<span><span class="va">stereo_go</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GiottoClass/man/addGiottoPolygons.html" class="external-link">addGiottoPolygons</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">stereo_go</span>,</span>
<span>                               gpolygons <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">lasso_polygons</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># find intersection between original giotto object and polygon subset </span></span>
<span><span class="va">my_intersect</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getCellsFromPolygon.html">getCellsFromPolygon</a></span><span class="op">(</span><span class="va">stereo_go</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># create new giotto roi subset</span></span>
<span><span class="va">stereo_go_subset</span> <span class="op">&lt;-</span> <span class="va">stereo_go</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/pkg/GiottoClass/man/subsetGiotto.html" class="external-link">subsetGiotto</a></span><span class="op">(</span>cell_ids <span class="op">=</span> <span class="va">my_intersect</span><span class="op">$</span><span class="va">cell_ID</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># visualize filtered ROI</span></span>
<span><span class="co"># Your plot below will reflect the polygon(s) you constructed above in my_polygon_coordinates</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/spatPlot.html">spatPlot2D</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">stereo_go_subset</span>, cell_color <span class="op">=</span> <span class="st">'leiden_clus'</span>,</span>
<span>           color_as_factor <span class="op">=</span> <span class="cn">T</span>, show_plot <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>           point_size <span class="op">=</span> <span class="fl">2</span>,save_plot <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/stereoseq_mosta/9.png"></p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span>
<span id="cb12-2"><a href="#cb12-2" tabindex="-1"></a></span>
<span id="cb12-3"><a href="#cb12-3" tabindex="-1"></a>R version <span class="dv">4</span>.<span class="fl">2.1</span> (<span class="dv">2022-06-23</span>)</span>
<span id="cb12-4"><a href="#cb12-4" tabindex="-1"></a>Platform<span class="sc">:</span> x86_64<span class="sc">-</span>pc<span class="sc">-</span>linux<span class="sc">-</span><span class="fu">gnu</span> (<span class="dv">64</span><span class="sc">-</span>bit)</span>
<span id="cb12-5"><a href="#cb12-5" tabindex="-1"></a>Running under<span class="sc">:</span> CentOS Linux <span class="dv">7</span> (Core)</span>
<span id="cb12-6"><a href="#cb12-6" tabindex="-1"></a></span>
<span id="cb12-7"><a href="#cb12-7" tabindex="-1"></a>Matrix products<span class="sc">:</span> default</span>
<span id="cb12-8"><a href="#cb12-8" tabindex="-1"></a>BLAS<span class="sc">:</span>   <span class="er">/</span>share<span class="sc">/</span>pkg<span class="fl">.7</span><span class="sc">/</span>r<span class="sc">/</span><span class="dv">4</span>.<span class="fl">2.1</span><span class="sc">/</span>install<span class="sc">/</span>lib64<span class="sc">/</span>R<span class="sc">/</span>lib<span class="sc">/</span>libRblas.so</span>
<span id="cb12-9"><a href="#cb12-9" tabindex="-1"></a>LAPACK<span class="sc">:</span> <span class="er">/</span>share<span class="sc">/</span>pkg<span class="fl">.7</span><span class="sc">/</span>r<span class="sc">/</span><span class="dv">4</span>.<span class="fl">2.1</span><span class="sc">/</span>install<span class="sc">/</span>lib64<span class="sc">/</span>R<span class="sc">/</span>lib<span class="sc">/</span>libRlapack.so</span>
<span id="cb12-10"><a href="#cb12-10" tabindex="-1"></a></span>
<span id="cb12-11"><a href="#cb12-11" tabindex="-1"></a>locale<span class="sc">:</span></span>
<span id="cb12-12"><a href="#cb12-12" tabindex="-1"></a>  [<span class="dv">1</span>] LC_CTYPE<span class="ot">=</span>en_US.UTF<span class="dv">-8</span>       LC_NUMERIC<span class="ot">=</span>C               LC_TIME<span class="ot">=</span>en_US.UTF<span class="dv">-8</span>       </span>
<span id="cb12-13"><a href="#cb12-13" tabindex="-1"></a>[<span class="dv">4</span>] LC_COLLATE<span class="ot">=</span>en_US.UTF<span class="dv">-8</span>     LC_MONETARY<span class="ot">=</span>en_US.UTF<span class="dv">-8</span>    LC_MESSAGES<span class="ot">=</span>en_US.UTF<span class="dv">-8</span>   </span>
<span id="cb12-14"><a href="#cb12-14" tabindex="-1"></a>[<span class="dv">7</span>] LC_PAPER<span class="ot">=</span>en_US.UTF<span class="dv">-8</span>       LC_NAME<span class="ot">=</span>C                  LC_ADDRESS<span class="ot">=</span>C              </span>
<span id="cb12-15"><a href="#cb12-15" tabindex="-1"></a>[<span class="dv">10</span>] LC_TELEPHONE<span class="ot">=</span>C             LC_MEASUREMENT<span class="ot">=</span>en_US.UTF<span class="dv">-8</span> LC_IDENTIFICATION<span class="ot">=</span>C       </span>
<span id="cb12-16"><a href="#cb12-16" tabindex="-1"></a></span>
<span id="cb12-17"><a href="#cb12-17" tabindex="-1"></a>attached base packages<span class="sc">:</span></span>
<span id="cb12-18"><a href="#cb12-18" tabindex="-1"></a>  [<span class="dv">1</span>] stats     graphics  grDevices utils     datasets  methods   base     </span>
<span id="cb12-19"><a href="#cb12-19" tabindex="-1"></a></span>
<span id="cb12-20"><a href="#cb12-20" tabindex="-1"></a>other attached packages<span class="sc">:</span></span>
<span id="cb12-21"><a href="#cb12-21" tabindex="-1"></a>  [<span class="dv">1</span>] miniUI_0.<span class="dv">1</span>.<span class="fl">1.1</span> shiny_1.<span class="fl">7.2</span>    Giotto_2<span class="fl">.1</span>    </span>
<span id="cb12-22"><a href="#cb12-22" tabindex="-1"></a></span>
<span id="cb12-23"><a href="#cb12-23" tabindex="-1"></a>loaded via a <span class="fu">namespace</span> (and not attached)<span class="sc">:</span></span>
<span id="cb12-24"><a href="#cb12-24" tabindex="-1"></a>  [<span class="dv">1</span>] systemfonts_1.<span class="fl">0.4</span>     plyr_1.<span class="fl">8.8</span>            igraph_1.<span class="fl">3.5</span>         </span>
<span id="cb12-25"><a href="#cb12-25" tabindex="-1"></a>[<span class="dv">4</span>] lazyeval_0.<span class="fl">2.2</span>        sp_1<span class="fl">.5</span><span class="dv">-1</span>              splines_4.<span class="fl">2.1</span>        </span>
<span id="cb12-26"><a href="#cb12-26" tabindex="-1"></a>[<span class="dv">7</span>] BiocParallel_1.<span class="fl">32.1</span>   listenv_0.<span class="fl">8.0</span>         scattermore_0<span class="fl">.8</span>      </span>
<span id="cb12-27"><a href="#cb12-27" tabindex="-1"></a>[<span class="dv">10</span>] ggplot2_3.<span class="fl">4.0</span>         digest_0.<span class="fl">6.30</span>         htmltools_0.<span class="fl">5.3</span>      </span>
<span id="cb12-28"><a href="#cb12-28" tabindex="-1"></a>[<span class="dv">13</span>] fansi_1.<span class="fl">0.3</span>           memoise_2.<span class="fl">0.1</span>         magrittr_2.<span class="fl">0.3</span>       </span>
<span id="cb12-29"><a href="#cb12-29" tabindex="-1"></a>[<span class="dv">16</span>] ScaledMatrix_1.<span class="fl">6.0</span>    tensor_1<span class="fl">.5</span>            cluster_2.<span class="fl">1.3</span>        </span>
<span id="cb12-30"><a href="#cb12-30" tabindex="-1"></a>[<span class="dv">19</span>] ROCR_1<span class="fl">.0</span><span class="dv">-11</span>           tzdb_0.<span class="fl">3.0</span>            remotes_2.<span class="fl">4.2</span>        </span>
<span id="cb12-31"><a href="#cb12-31" tabindex="-1"></a>[<span class="dv">22</span>] globals_0.<span class="fl">16.1</span>        readr_2.<span class="fl">1.2</span>           matrixStats_0.<span class="fl">62.0</span>   </span>
<span id="cb12-32"><a href="#cb12-32" tabindex="-1"></a>[<span class="dv">25</span>] spatstat.sparse_2<span class="fl">.1</span><span class="dv">-1</span> colorspace_2<span class="fl">.1</span><span class="dv">-0</span>      rappdirs_0.<span class="fl">3.3</span>       </span>
<span id="cb12-33"><a href="#cb12-33" tabindex="-1"></a>[<span class="dv">28</span>] ggrepel_0.<span class="fl">9.1</span>         textshaping_0.<span class="fl">3.6</span>     xfun_0<span class="fl">.34</span>            </span>
<span id="cb12-34"><a href="#cb12-34" tabindex="-1"></a>[<span class="dv">31</span>] dplyr_1.<span class="fl">0.10</span>          crayon_1.<span class="fl">5.2</span>          jsonlite_1.<span class="fl">8.3</span>       </span>
<span id="cb12-35"><a href="#cb12-35" tabindex="-1"></a>[<span class="dv">34</span>] progressr_0.<span class="fl">10.1</span>      spatstat.data_2<span class="fl">.2</span><span class="dv">-0</span>   survival_3<span class="fl">.3</span><span class="dv">-1</span>       </span>
<span id="cb12-36"><a href="#cb12-36" tabindex="-1"></a>[<span class="dv">37</span>] zoo_1<span class="fl">.8</span><span class="dv">-10</span>            glue_1.<span class="fl">6.2</span>            polyclip_1<span class="fl">.10</span><span class="dv">-0</span>      </span>
<span id="cb12-37"><a href="#cb12-37" tabindex="-1"></a>[<span class="dv">40</span>] gtable_0.<span class="fl">3.1</span>          leiden_0.<span class="fl">4.2</span>          DelayedArray_0.<span class="fl">24.0</span>  </span>
<span id="cb12-38"><a href="#cb12-38" tabindex="-1"></a>[<span class="dv">43</span>] BiocSingular_1.<span class="fl">14.0</span>   future.apply_1.<span class="fl">10.0</span>   BiocGenerics_0.<span class="fl">44.0</span>  </span>
<span id="cb12-39"><a href="#cb12-39" tabindex="-1"></a>[<span class="dv">46</span>] abind_1<span class="fl">.4</span><span class="dv">-7</span>           scales_1.<span class="fl">2.1</span>          DBI_1.<span class="fl">1.3</span>            </span>
<span id="cb12-40"><a href="#cb12-40" tabindex="-1"></a>[<span class="dv">49</span>] spatstat.random_2<span class="fl">.2</span><span class="dv">-0</span> Rcpp_1.<span class="fl">0.9</span>            viridisLite_0.<span class="fl">4.1</span>    </span>
<span id="cb12-41"><a href="#cb12-41" tabindex="-1"></a>[<span class="dv">52</span>] xtable_1<span class="fl">.8</span><span class="dv">-6</span>          rsthemes_0.<span class="fl">3.1</span>        reticulate_1<span class="fl">.26</span>      </span>
<span id="cb12-42"><a href="#cb12-42" tabindex="-1"></a>[<span class="dv">55</span>] spatstat.core_2<span class="fl">.4</span><span class="dv">-4</span>   rsvd_1.<span class="fl">0.5</span>            bit_4.<span class="fl">0.4</span>            </span>
<span id="cb12-43"><a href="#cb12-43" tabindex="-1"></a>[<span class="dv">58</span>] stats4_4.<span class="fl">2.1</span>          htmlwidgets_1.<span class="fl">5.4</span>     httr_1.<span class="fl">4.4</span>           </span>
<span id="cb12-44"><a href="#cb12-44" tabindex="-1"></a>[<span class="dv">61</span>] FNN_1.<span class="dv">1</span>.<span class="fl">3.1</span>           RColorBrewer_1<span class="fl">.1</span><span class="dv">-3</span>    ellipsis_0.<span class="fl">3.2</span>       </span>
<span id="cb12-45"><a href="#cb12-45" tabindex="-1"></a>[<span class="dv">64</span>] Seurat_4.<span class="fl">1.1</span>          ica_1<span class="fl">.0</span><span class="dv">-3</span>             pkgconfig_2.<span class="fl">0.3</span>      </span>
<span id="cb12-46"><a href="#cb12-46" tabindex="-1"></a>[<span class="dv">67</span>] farver_2.<span class="fl">1.1</span>          sass_0.<span class="dv">4</span>.<span class="fl">2.9000</span>       uwot_0.<span class="fl">1.14</span>          </span>
<span id="cb12-47"><a href="#cb12-47" tabindex="-1"></a>[<span class="dv">70</span>] deldir_1<span class="fl">.0</span><span class="dv">-6</span>          utf8_1.<span class="fl">2.2</span>            here_1.<span class="fl">0.1</span>           </span>
<span id="cb12-48"><a href="#cb12-48" tabindex="-1"></a>[<span class="dv">73</span>] tidyselect_1.<span class="fl">2.0</span>      labeling_0.<span class="fl">4.2</span>        rlang_1.<span class="fl">0.6</span>          </span>
<span id="cb12-49"><a href="#cb12-49" tabindex="-1"></a>[<span class="dv">76</span>] reshape2_1.<span class="fl">4.4</span>        later_1.<span class="fl">3.0</span>           cachem_1.<span class="fl">0.6</span>         </span>
<span id="cb12-50"><a href="#cb12-50" tabindex="-1"></a>[<span class="dv">79</span>] munsell_0.<span class="fl">5.0</span>         tools_4.<span class="fl">2.1</span>           cli_3.<span class="fl">4.1</span>            </span>
<span id="cb12-51"><a href="#cb12-51" tabindex="-1"></a>[<span class="dv">82</span>] dbscan_1<span class="fl">.1</span><span class="dv">-11</span>         generics_0.<span class="fl">1.3</span>        ggridges_0.<span class="fl">5.3</span>       </span>
<span id="cb12-52"><a href="#cb12-52" tabindex="-1"></a>[<span class="dv">85</span>] evaluate_0<span class="fl">.18</span>         stringr_1.<span class="fl">4.1</span>         fastmap_1.<span class="fl">1.0</span>        </span>
<span id="cb12-53"><a href="#cb12-53" tabindex="-1"></a>[<span class="dv">88</span>] ragg_1.<span class="fl">2.2</span>            yaml_2.<span class="fl">3.6</span>            goftest_1<span class="fl">.2</span><span class="dv">-3</span>        </span>
<span id="cb12-54"><a href="#cb12-54" tabindex="-1"></a>[<span class="dv">91</span>] knitr_1<span class="fl">.40</span>            bit64_4.<span class="fl">0.5</span>           fitdistrplus_1<span class="fl">.1</span><span class="dv">-8</span>   </span>
<span id="cb12-55"><a href="#cb12-55" tabindex="-1"></a>[<span class="dv">94</span>] purrr_0.<span class="fl">3.5</span>           RANN_2.<span class="fl">6.1</span>            pbapply_1<span class="fl">.5</span><span class="dv">-0</span>        </span>
<span id="cb12-56"><a href="#cb12-56" tabindex="-1"></a>[<span class="dv">97</span>] future_1.<span class="fl">29.0</span>         nlme_3<span class="fl">.1</span><span class="dv">-158</span>          mime_0<span class="fl">.12</span>            </span>
<span id="cb12-57"><a href="#cb12-57" tabindex="-1"></a>[<span class="dv">100</span>] arrow_9.<span class="fl">0.0</span>           hdf5r_1.<span class="fl">3.5</span>           compiler_4.<span class="fl">2.1</span>       </span>
<span id="cb12-58"><a href="#cb12-58" tabindex="-1"></a>[<span class="dv">103</span>] rstudioapi_0<span class="fl">.14</span>       plotly_4.<span class="fl">10.1</span>         png_0<span class="fl">.1</span><span class="dv">-7</span>            </span>
<span id="cb12-59"><a href="#cb12-59" tabindex="-1"></a>[<span class="dv">106</span>] spatstat.utils_2<span class="fl">.3</span><span class="dv">-1</span>  tibble_3.<span class="fl">1.8</span>          bslib_0.<span class="fl">4.1</span>          </span>
<span id="cb12-60"><a href="#cb12-60" tabindex="-1"></a>[<span class="dv">109</span>] stringi_1.<span class="fl">7.8</span>         rgeos_0<span class="fl">.5</span><span class="dv">-9</span>           lattice_0<span class="fl">.20</span><span class="dv">-45</span>      </span>
<span id="cb12-61"><a href="#cb12-61" tabindex="-1"></a>[<span class="dv">112</span>] Matrix_1<span class="fl">.5</span><span class="dv">-1</span>          SeuratDisk_0.<span class="dv">0</span>.<span class="fl">0.9020</span> vctrs_0.<span class="fl">5.0</span>          </span>
<span id="cb12-62"><a href="#cb12-62" tabindex="-1"></a>[<span class="dv">115</span>] pillar_1.<span class="fl">8.1</span>          lifecycle_1.<span class="fl">0.3</span>       jquerylib_0.<span class="fl">1.4</span>      </span>
<span id="cb12-63"><a href="#cb12-63" tabindex="-1"></a>[<span class="dv">118</span>] spatstat.geom_2<span class="fl">.4</span><span class="dv">-0</span>   lmtest_0<span class="fl">.9</span><span class="dv">-40</span>         RcppAnnoy_0.<span class="fl">0.20</span>     </span>
<span id="cb12-64"><a href="#cb12-64" tabindex="-1"></a>[<span class="dv">121</span>] data.table_1.<span class="fl">14.4</span>     cowplot_1.<span class="fl">1.1</span>         irlba_2.<span class="dv">3</span>.<span class="fl">5.1</span>        </span>
<span id="cb12-65"><a href="#cb12-65" tabindex="-1"></a>[<span class="dv">124</span>] httpuv_1.<span class="fl">6.6</span>          patchwork_1.<span class="dv">1</span>.<span class="fl">0.9000</span>  R6_2.<span class="fl">5.1</span>             </span>
<span id="cb12-66"><a href="#cb12-66" tabindex="-1"></a>[<span class="dv">127</span>] promises_1.<span class="dv">2</span>.<span class="fl">0.1</span>      KernSmooth_2<span class="fl">.23</span><span class="dv">-20</span>    gridExtra_2<span class="fl">.3</span>        </span>
<span id="cb12-67"><a href="#cb12-67" tabindex="-1"></a>[<span class="dv">130</span>] IRanges_2.<span class="fl">32.0</span>        parallelly_1.<span class="fl">32.1</span>     codetools_0<span class="fl">.2</span><span class="dv">-18</span>     </span>
<span id="cb12-68"><a href="#cb12-68" tabindex="-1"></a>[<span class="dv">133</span>] MASS_7<span class="fl">.3</span><span class="dv">-57</span>           gtools_3.<span class="fl">9.3</span>          assertthat_0.<span class="fl">2.1</span>     </span>
<span id="cb12-69"><a href="#cb12-69" tabindex="-1"></a>[<span class="dv">136</span>] rprojroot_2.<span class="fl">0.3</span>       withr_2.<span class="fl">5.0</span>           SeuratObject_4.<span class="fl">1.0</span>   </span>
<span id="cb12-70"><a href="#cb12-70" tabindex="-1"></a>[<span class="dv">139</span>] sctransform_0.<span class="fl">3.3</span>     S4Vectors_0.<span class="fl">36.0</span>      mgcv_1<span class="fl">.8</span><span class="dv">-40</span>          </span>
<span id="cb12-71"><a href="#cb12-71" tabindex="-1"></a>[<span class="dv">142</span>] parallel_4.<span class="fl">2.1</span>        hms_1.<span class="fl">1.1</span>             terra_1<span class="fl">.5</span><span class="dv">-34</span>         </span>
<span id="cb12-72"><a href="#cb12-72" tabindex="-1"></a>[<span class="dv">145</span>] beachmat_2.<span class="fl">14.0</span>       grid_4.<span class="fl">2.1</span>            rpart_4.<span class="fl">1.16</span>         </span>
<span id="cb12-73"><a href="#cb12-73" tabindex="-1"></a>[<span class="dv">148</span>] tidyr_1.<span class="fl">2.1</span>           rmarkdown_2<span class="fl">.18</span>        MatrixGenerics_1.<span class="fl">10.0</span></span>
<span id="cb12-74"><a href="#cb12-74" tabindex="-1"></a>[<span class="dv">151</span>] Rtsne_0<span class="fl">.16</span>       </span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Ruben Dries, Jiaji George Chen, Natalie Del Rossi, Guo-Cheng Yuan, Matthew O’Brien, <a href="https://josschavezf.github.io" class="external-link">Joselyn Chávez</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
