<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="Giotto">
<title>Visium Prostate Integration â€¢ Giotto</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><link href="../deps/JetBrains_Mono-0.4.7/font.css" rel="stylesheet">
<!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Visium Prostate Integration">
<meta property="og:description" content="Giotto">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-dark"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">Giotto</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">3.3.2</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Documentation</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-tutorials">Tutorials</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-tutorials">
    <h6 class="dropdown-header" data-toc-skip>Seq-Scope</h6>
    <a class="dropdown-item" href="../articles/seqscope_mouse_liver.html">Mouse liver</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Visium</h6>
    <a class="dropdown-item" href="../articles/visium_mouse_kidney.html">Visium Mouse Kidney</a>
    <a class="dropdown-item" href="../articles/visium_mouse_brain.html">Visium Mouse Brain</a>
    <a class="dropdown-item" href="../articles/visium_prostate_integration.html">Visium Prostate Integration</a>
    <a class="dropdown-item" href="../articles/visium_cytassist_lungcancer.html">CytAssist Human Lung Cancer</a>
    <a class="dropdown-item" href="../articles/visium_cytassist_human_glioblastoma.html">CytAssist Multi-omics Human Glioblastoma</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Spatial Genomics</h6>
    <a class="dropdown-item" href="../articles/spatial_genomics_mouse_kidney.html">Mouse Kidney</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Xenium</h6>
    <a class="dropdown-item" href="../articles/xenium_breast_cancer.html">Xenium Human breast cancer</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Nanostring CosMx</h6>
    <a class="dropdown-item" href="../articles/nanostring_cosmx_lung_cancer.html">Nanostring CosMx Subcellular Lung Cancer</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>MERFISH</h6>
    <a class="dropdown-item" href="../articles/merfish_mouse_hypothalamic.html">MERFISH Mouse Hypothalamic Preoptic Region</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Vizgen</h6>
    <a class="dropdown-item" href="../articles/vizgen_mouse_brain.html">Vizgen Mouse Brain Receptor Map</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>seqFISH</h6>
    <a class="dropdown-item" href="../articles/mini_seqfish.html">mini seqFISH</a>
    <a class="dropdown-item" href="../articles/seqfish_cortex.html">SeqFish+ Mouse Cortex Example</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>osmFISH</h6>
    <a class="dropdown-item" href="../articles/osmfish_mouse_ss_cortex.html">osmFISH Mouse SS Cortex</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Resolve Bioscience</h6>
    <a class="dropdown-item" href="../articles/resolve_biosciences.html">Resolve Bioscience Breast Cancer Subcellular</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>CODEX</h6>
    <a class="dropdown-item" href="../articles/codex_mouse_spleen.html">Codex Mouse Spleen</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Single Cell RNA Sequencing</h6>
    <a class="dropdown-item" href="../articles/singlecell_prostate_standard.html">Prostate</a>
    <a class="dropdown-item" href="../articles/singlecell_prostate_integration.html">Prostate Integration</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">News</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/drieslab/Giotto/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Visium Prostate Integration</h1>
            
            <h4 data-toc-skip class="date">5/5/23</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/drieslab/Giotto/blob/HEAD/vignettes/visium_prostate_integration.Rmd" class="external-link"><code>vignettes/visium_prostate_integration.Rmd</code></a></small>
      <div class="d-none name"><code>visium_prostate_integration.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="start-giotto">1 Start Giotto<a class="anchor" aria-label="anchor" href="#start-giotto"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Ensure Giotto Suite is installed</span></span>
<span><span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="st">"Giotto"</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/installed.packages.html" class="external-link">installed.packages</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"drieslab/Giotto@suite"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://drieslab.github.io/Giotto/">Giotto</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Ensure Giotto Data is installed</span></span>
<span><span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="st">"GiottoData"</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/installed.packages.html" class="external-link">installed.packages</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"drieslab/GiottoData"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://drieslab.github.io/GiottoData/">GiottoData</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Ensure the Python environment for Giotto has been installed</span></span>
<span><span class="va">genv_exists</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/GiottoClass/man/checkGiottoEnvironment.html" class="external-link">checkGiottoEnvironment</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="va">genv_exists</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="co"># The following command need only be run once to install the Giotto environment</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/GiottoClass/man/installGiottoEnvironment.html" class="external-link">installGiottoEnvironment</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># 1. set working directory</span></span>
<span><span class="va">results_directory</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/getwd.html" class="external-link">getwd</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 2. set giotto python path</span></span>
<span><span class="co"># set python path to your preferred python version path</span></span>
<span><span class="co"># set python path to NULL if you want to automatically install (only the 1st time) and use the giotto miniconda environment</span></span>
<span><span class="va">python_path</span> <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="kw">if</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NULL.html" class="external-link">is.null</a></span><span class="op">(</span><span class="va">python_path</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/GiottoClass/man/installGiottoEnvironment.html" class="external-link">installGiottoEnvironment</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># 3. create giotto instructions</span></span>
<span><span class="va">instrs</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/GiottoClass/man/createGiottoInstructions.html" class="external-link">createGiottoInstructions</a></span><span class="op">(</span>save_dir <span class="op">=</span> <span class="va">results_directory</span>,</span>
<span>                                  save_plot <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                                  show_plot <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                                  python_path <span class="op">=</span> <span class="va">python_path</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="dataset-explanation">2 Dataset explanation<a class="anchor" aria-label="anchor" href="#dataset-explanation"></a>
</h2>
<p><a href="https://www.10xgenomics.com/spatial-transcriptomics/" class="external-link">10X
genomics</a> recently launched a new platform to obtain spatial
expression data using a Visium Spatial Gene Expression slide.</p>
<p>The Visium Cancer Prostate data to run this tutorial can be found <a href="https://www.10xgenomics.com/resources/datasets/human-prostate-cancer-adenocarcinoma-with-invasive-carcinoma-ffpe-1-standard-1-3-0" class="external-link">here</a>
The Visium Normal Prostate data to run this tutorial can be found <a href="https://www.10xgenomics.com/resources/datasets/normal-human-prostate-ffpe-1-standard-1-3-0" class="external-link">here</a></p>
<p>Visium technology:</p>
<p><img src="images/general_figs/visium_technology.png" style="width:50.0%"></p>
<p>High resolution png from original tissue:</p>
<p><img src="images/general_figs/Visium_FFPE_Human_Normal_Prostate_image.png" style="width:50.0%"></p>
<p><img src="images/general_figs/Visium_FFPE_Human_Prostate_Cancer_image.png" style="width:50.0%"></p>
</div>
<div class="section level2">
<h2 id="create-giotto-objects-and-join">3 Create Giotto objects and join<a class="anchor" aria-label="anchor" href="#create-giotto-objects-and-join"></a>
</h2>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># This dataset must be downlaoded manually; please do so and change the path below as appropriate</span></span>
<span><span class="va">data_directory</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/getwd.html" class="external-link">getwd</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## obese upper</span></span>
<span><span class="va">N_pros</span> <span class="op">=</span> <span class="fu"><a href="../reference/createGiottoVisiumObject.html">createGiottoVisiumObject</a></span><span class="op">(</span></span>
<span>    visium_dir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">data_directory</span>,<span class="st">'/Visium_FFPE_Human_Normal_Prostate'</span><span class="op">)</span>,</span>
<span>    expr_data <span class="op">=</span> <span class="st">'raw'</span>,</span>
<span>    png_name <span class="op">=</span> <span class="st">'tissue_lowres_image.png'</span>,</span>
<span>    gene_column_index <span class="op">=</span> <span class="fl">2</span>,</span>
<span>    instructions <span class="op">=</span> <span class="va">instrs</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## obese lower</span></span>
<span><span class="va">C_pros</span> <span class="op">=</span> <span class="fu"><a href="../reference/createGiottoVisiumObject.html">createGiottoVisiumObject</a></span><span class="op">(</span></span>
<span>    visium_dir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">data_directory</span>,<span class="st">'/Visium_FFPE_Human_Prostate_Cancer/'</span><span class="op">)</span>,</span>
<span>    expr_data <span class="op">=</span> <span class="st">'raw'</span>,</span>
<span>    png_name <span class="op">=</span> <span class="st">'tissue_lowres_image.png'</span>,</span>
<span>    gene_column_index <span class="op">=</span> <span class="fl">2</span>,</span>
<span>    instructions <span class="op">=</span> <span class="va">instrs</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># join giotto objects</span></span>
<span><span class="co"># joining with x_shift has the advantage that you can join both 2D and 3D data</span></span>
<span><span class="co"># x_padding determines how much distance is between each dataset</span></span>
<span><span class="co"># if x_shift = NULL, then the total shift will be guessed from the giotto image</span></span>
<span><span class="va">testcombo</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/GiottoClass/man/joinGiottoObjects.html" class="external-link">joinGiottoObjects</a></span><span class="op">(</span>gobject_list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">N_pros</span>, <span class="va">C_pros</span><span class="op">)</span>,</span>
<span>    gobject_names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'NP'</span>, <span class="st">'CP'</span><span class="op">)</span>,</span>
<span>    join_method <span class="op">=</span> <span class="st">'shift'</span>, x_padding <span class="op">=</span> <span class="fl">1000</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># join info is stored in this slot</span></span>
<span><span class="co"># simple list for now</span></span>
<span><span class="va">testcombo</span><span class="op">@</span><span class="va">join_info</span></span></code></pre></div>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># check joined Giotto object</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/GiottoClass/man/fDataDT.html" class="external-link">fDataDT</a></span><span class="op">(</span><span class="va">testcombo</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/GiottoClass/man/pDataDT.html" class="external-link">pDataDT</a></span><span class="op">(</span><span class="va">testcombo</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/GiottoClass/man/showGiottoImageNames.html" class="external-link">showGiottoImageNames</a></span><span class="op">(</span><span class="va">testcombo</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/GiottoClass/man/showGiottoSpatLocs.html" class="external-link">showGiottoSpatLocs</a></span><span class="op">(</span><span class="va">testcombo</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/GiottoClass/man/showGiottoExpression.html" class="external-link">showGiottoExpression</a></span><span class="op">(</span><span class="va">testcombo</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># this plots all the images by list_ID</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/spatPlot.html">spatPlot2D</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">testcombo</span>, cell_color <span class="op">=</span> <span class="st">'in_tissue'</span>,</span>
<span>    show_image <span class="op">=</span> <span class="cn">T</span>, image_name <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"NP-image"</span>, <span class="st">"CP-image"</span><span class="op">)</span>,</span>
<span>    group_by <span class="op">=</span> <span class="st">'list_ID'</span>, point_alpha <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>    save_param <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>save_name <span class="op">=</span> <span class="st">"1a_plot"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/visium_prostate_integration/unnamed-chunk-5-1.png"></p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># this plots one selected image</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/spatPlot.html">spatPlot2D</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">testcombo</span>, cell_color <span class="op">=</span> <span class="st">'in_tissue'</span>,</span>
<span>    show_image <span class="op">=</span> <span class="cn">T</span>, image_name <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"NP-image"</span><span class="op">)</span>, point_alpha <span class="op">=</span> <span class="fl">0.3</span>,</span>
<span>    save_param <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>save_name <span class="op">=</span> <span class="st">"1b_plot"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/visium_prostate_integration/unnamed-chunk-6-1.png"></p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># this plots two selected images</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/spatPlot.html">spatPlot2D</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">testcombo</span>, cell_color <span class="op">=</span> <span class="st">'in_tissue'</span>,</span>
<span>    show_image <span class="op">=</span> <span class="cn">T</span>, image_name <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span> <span class="st">"NP-image"</span>, <span class="st">"CP-image"</span><span class="op">)</span>,</span>
<span>    point_alpha <span class="op">=</span> <span class="fl">0.3</span>,</span>
<span>    save_param <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>save_name <span class="op">=</span> <span class="st">"1c_plot"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/visium_prostate_integration/unnamed-chunk-7-1.png"></p>
</div>
<div class="section level2">
<h2 id="process-giotto-objects">4 Process Giotto Objects<a class="anchor" aria-label="anchor" href="#process-giotto-objects"></a>
</h2>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># subset on in-tissue spots</span></span>
<span><span class="va">metadata</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/GiottoClass/man/pDataDT.html" class="external-link">pDataDT</a></span><span class="op">(</span><span class="va">testcombo</span><span class="op">)</span></span>
<span><span class="va">in_tissue_barcodes</span> <span class="op">=</span> <span class="va">metadata</span><span class="op">[</span><span class="va">in_tissue</span> <span class="op">==</span> <span class="fl">1</span><span class="op">]</span><span class="op">$</span><span class="va">cell_ID</span></span>
<span><span class="va">testcombo</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/GiottoClass/man/subsetGiotto.html" class="external-link">subsetGiotto</a></span><span class="op">(</span><span class="va">testcombo</span>, cell_ids <span class="op">=</span> <span class="va">in_tissue_barcodes</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## filter</span></span>
<span><span class="va">testcombo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/filterGiotto.html">filterGiotto</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">testcombo</span>,</span>
<span>    expression_threshold <span class="op">=</span> <span class="fl">1</span>,</span>
<span>    feat_det_in_min_cells <span class="op">=</span> <span class="fl">50</span>,</span>
<span>    min_det_feats_per_cell <span class="op">=</span> <span class="fl">500</span>,</span>
<span>    expression_values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'raw'</span><span class="op">)</span>,</span>
<span>    verbose <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## normalize</span></span>
<span><span class="va">testcombo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/normalizeGiotto.html">normalizeGiotto</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">testcombo</span>, scalefactor <span class="op">=</span> <span class="fl">6000</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## add gene &amp; cell statistics</span></span>
<span><span class="va">testcombo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/addStatistics.html">addStatistics</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">testcombo</span>, expression_values <span class="op">=</span> <span class="st">'raw'</span><span class="op">)</span></span>
<span></span>
<span><span class="va">fmeta</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/GiottoClass/man/fDataDT.html" class="external-link">fDataDT</a></span><span class="op">(</span><span class="va">testcombo</span><span class="op">)</span></span>
<span><span class="va">testfeats</span> <span class="op">=</span> <span class="va">fmeta</span><span class="op">[</span><span class="va">perc_cells</span> <span class="op">&gt;</span> <span class="fl">20</span> <span class="op">&amp;</span> <span class="va">perc_cells</span> <span class="op">&lt;</span> <span class="fl">50</span><span class="op">]</span><span class="op">[</span><span class="fl">100</span><span class="op">:</span><span class="fl">110</span><span class="op">]</span><span class="op">$</span><span class="va">feat_ID</span></span>
<span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/violinPlot.html">violinPlot</a></span><span class="op">(</span><span class="va">testcombo</span>, feats <span class="op">=</span> <span class="va">testfeats</span>, cluster_column <span class="op">=</span> <span class="st">'list_ID'</span>, save_param <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>save_name <span class="op">=</span> <span class="st">"2a_plot"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/visium_prostate_integration/unnamed-chunk-9-1.png"></p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/plotMetaDataHeatmap.html">plotMetaDataHeatmap</a></span><span class="op">(</span><span class="va">testcombo</span>, selected_feats <span class="op">=</span> <span class="va">testfeats</span>, metadata_cols <span class="op">=</span> <span class="st">'list_ID'</span>, save_param <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>save_name <span class="op">=</span> <span class="st">"2b_plot"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/visium_prostate_integration/unnamed-chunk-9-3.png"></p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## visualize</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/spatPlot.html">spatPlot2D</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">testcombo</span>, group_by <span class="op">=</span> <span class="st">'list_ID'</span>, cell_color <span class="op">=</span> <span class="st">'nr_feats'</span>, color_as_factor <span class="op">=</span> <span class="cn">F</span>, point_size <span class="op">=</span> <span class="fl">0.75</span>, save_param <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>save_name <span class="op">=</span> <span class="st">"2c_plot"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/visium_prostate_integration/unnamed-chunk-10-1.png"></p>
</div>
<div class="section level2">
<h2 id="dimention-reduction">5 Dimention Reduction<a class="anchor" aria-label="anchor" href="#dimention-reduction"></a>
</h2>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## PCA ##</span></span>
<span><span class="va">testcombo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateHVF.html">calculateHVF</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">testcombo</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/visium_prostate_integration/unnamed-chunk-11-1.png"></p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">testcombo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runPCA.html">runPCA</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">testcombo</span>, center <span class="op">=</span> <span class="cn">TRUE</span>, scale_unit <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/screePlot.html">screePlot</a></span><span class="op">(</span><span class="va">testcombo</span>, ncp <span class="op">=</span> <span class="fl">30</span>, save_param <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>save_name <span class="op">=</span> <span class="st">"3a_screeplot"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/visium_prostate_integration/unnamed-chunk-11-2.png"></p>
</div>
<div class="section level2">
<h2 id="clustering">6 Clustering<a class="anchor" aria-label="anchor" href="#clustering"></a>
</h2>
<div class="section level3">
<h3 id="without-integration">6.1 Without Integration<a class="anchor" aria-label="anchor" href="#without-integration"></a>
</h3>
<p>Integration is usually needed for dataset of different conditions to
minimize batch effects. Without integration means without using any
integration methods.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## cluster and run UMAP ##</span></span>
<span><span class="co"># sNN network (default)</span></span>
<span><span class="va">testcombo</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GiottoClass/man/createNearestNetwork.html" class="external-link">createNearestNetwork</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">testcombo</span>,</span>
<span>    dim_reduction_to_use <span class="op">=</span> <span class="st">'pca'</span>, dim_reduction_name <span class="op">=</span> <span class="st">'pca'</span>,</span>
<span>    dimensions_to_use <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, k <span class="op">=</span> <span class="fl">15</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Leiden clustering</span></span>
<span><span class="va">testcombo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/doLeidenCluster.html">doLeidenCluster</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">testcombo</span>, resolution <span class="op">=</span> <span class="fl">0.2</span>, n_iterations <span class="op">=</span> <span class="fl">1000</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># UMAP</span></span>
<span><span class="va">testcombo</span> <span class="op">=</span> <span class="fu"><a href="../reference/runUMAP.html">runUMAP</a></span><span class="op">(</span><span class="va">testcombo</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/plotUMAP.html">plotUMAP</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">testcombo</span>,</span>
<span>    cell_color <span class="op">=</span> <span class="st">'leiden_clus'</span>, show_NN_network <span class="op">=</span> <span class="cn">T</span>, point_size <span class="op">=</span> <span class="fl">1.5</span>,</span>
<span>    save_param <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>save_name <span class="op">=</span> <span class="st">"4.1a_plot"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/visium_prostate_integration/unnamed-chunk-12-1.png"></p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/spatPlot.html">spatPlot2D</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">testcombo</span>, group_by <span class="op">=</span> <span class="st">'list_ID'</span>,</span>
<span>    cell_color <span class="op">=</span> <span class="st">'leiden_clus'</span>,</span>
<span>    point_size <span class="op">=</span> <span class="fl">1.5</span>,</span>
<span>    save_param <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>save_name <span class="op">=</span> <span class="st">"4.1b_plot"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/visium_prostate_integration/unnamed-chunk-13-1.png"></p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/spatDimPlot.html">spatDimPlot2D</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">testcombo</span>,</span>
<span>    cell_color <span class="op">=</span> <span class="st">'leiden_clus'</span>,</span>
<span>    save_param <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>save_name <span class="op">=</span> <span class="st">"4.1c_plot"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/visium_prostate_integration/unnamed-chunk-14-1.png"></p>
</div>
<div class="section level3">
<h3 id="with-harmony-integration">6.2 With Harmony integration<a class="anchor" aria-label="anchor" href="#with-harmony-integration"></a>
</h3>
<p>Harmony is a integration algorithm developed by <a href="https://www.nature.com/articles/s41592-019-0619-0" class="external-link">Korsunsky, I.
et al.</a>. It was designed for integration of single cell data but also
work well on spatial datasets.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## data integration, cluster and run UMAP ##</span></span>
<span></span>
<span><span class="co"># harmony</span></span>
<span><span class="co">#library(devtools)</span></span>
<span><span class="co">#install_github("immunogenomics/harmony")</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">harmony</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## run harmony integration</span></span>
<span><span class="va">testcombo</span> <span class="op">=</span> <span class="fu"><a href="../reference/runGiottoHarmony.html">runGiottoHarmony</a></span><span class="op">(</span><span class="va">testcombo</span>, vars_use <span class="op">=</span> <span class="st">'list_ID'</span>, do_pca <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## sNN network (default)</span></span>
<span><span class="va">testcombo</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GiottoClass/man/createNearestNetwork.html" class="external-link">createNearestNetwork</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">testcombo</span>,</span>
<span>    dim_reduction_to_use <span class="op">=</span> <span class="st">'harmony'</span>, dim_reduction_name <span class="op">=</span> <span class="st">'harmony'</span>, name <span class="op">=</span> <span class="st">'NN.harmony'</span>,</span>
<span>    dimensions_to_use <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, k <span class="op">=</span> <span class="fl">15</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Leiden clustering</span></span>
<span><span class="va">testcombo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/doLeidenCluster.html">doLeidenCluster</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">testcombo</span>,</span>
<span>    network_name <span class="op">=</span> <span class="st">'NN.harmony'</span>, resolution <span class="op">=</span> <span class="fl">0.2</span>, n_iterations <span class="op">=</span> <span class="fl">1000</span>, name <span class="op">=</span> <span class="st">'leiden_harmony'</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># UMAP dimension reduction</span></span>
<span><span class="va">testcombo</span> <span class="op">=</span> <span class="fu"><a href="../reference/runUMAP.html">runUMAP</a></span><span class="op">(</span><span class="va">testcombo</span>, dim_reduction_name <span class="op">=</span> <span class="st">'harmony'</span>, dim_reduction_to_use <span class="op">=</span> <span class="st">'harmony'</span>, name <span class="op">=</span> <span class="st">'umap_harmony'</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/plotUMAP.html">plotUMAP</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">testcombo</span>,</span>
<span>    dim_reduction_name <span class="op">=</span> <span class="st">'umap_harmony'</span>,</span>
<span>    cell_color <span class="op">=</span> <span class="st">'leiden_harmony'</span>, </span>
<span>    show_NN_network <span class="op">=</span> <span class="cn">F</span>, </span>
<span>    point_size <span class="op">=</span> <span class="fl">1.5</span>,</span>
<span>    save_param <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>save_name <span class="op">=</span> <span class="st">"4.2a_plot"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/visium_prostate_integration/unnamed-chunk-15-1.png"></p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># If you want to show NN network information, you will need to specify these arguments in the plotUMAP function</span></span>
<span><span class="co"># show_NN_network = T, nn_network_to_use = 'sNN' , network_name = 'NN.harmony'</span></span></code></pre></div>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/spatPlot.html">spatPlot2D</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">testcombo</span>, group_by <span class="op">=</span> <span class="st">'list_ID'</span>,</span>
<span>    cell_color <span class="op">=</span> <span class="st">'leiden_harmony'</span>,</span>
<span>    point_size <span class="op">=</span> <span class="fl">1.5</span>,</span>
<span>    save_param <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>save_name <span class="op">=</span> <span class="st">"4.2b_plot"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/visium_prostate_integration/unnamed-chunk-16-1.png"></p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/spatDimPlot.html">spatDimPlot2D</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">testcombo</span>,</span>
<span>    dim_reduction_to_use <span class="op">=</span> <span class="st">'umap'</span>, dim_reduction_name <span class="op">=</span> <span class="st">'umap_harmony'</span>,</span>
<span>    cell_color <span class="op">=</span> <span class="st">'leiden_harmony'</span>,</span>
<span>    save_param <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>save_name <span class="op">=</span> <span class="st">"4.2c_plot"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/visium_prostate_integration/unnamed-chunk-17-1.png"></p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># compare to previous results</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/spatPlot.html">spatPlot2D</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">testcombo</span>,</span>
<span>    cell_color <span class="op">=</span> <span class="st">'leiden_clus'</span>, </span>
<span>    save_param <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>save_name <span class="op">=</span> <span class="st">"4_w_o_integration_plot"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/visium_prostate_integration/unnamed-chunk-18-1.png"></p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/spatPlot.html">spatPlot2D</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">testcombo</span>,</span>
<span>    cell_color <span class="op">=</span> <span class="st">'leiden_harmony'</span>,</span>
<span>    save_param <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>save_name <span class="op">=</span> <span class="st">"4_w_integration_plot"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/visium_prostate_integration/unnamed-chunk-18-3.png"></p>
</div>
</div>
<div class="section level2">
<h2 id="cell-type-annotation">7 Cell type annotation<a class="anchor" aria-label="anchor" href="#cell-type-annotation"></a>
</h2>
<p>Visium spatial transcriptomics does not provide single-cell
resolution, making cell type annotation a harder problem. Giotto
provides several ways to calculate enrichment of specific cell-type
signature gene list:</p>
<ul>
<li>PAGE</li>
<li>hypergeometric test</li>
<li>Rank</li>
<li><a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-021-02362-7" class="external-link">DWLS
Deconvolution</a></li>
</ul>
<p>This is also the easiest way to integrate Visium datasets with single
cell data. Example shown here is from <a href="https://pubmed.ncbi.nlm.nih.gov/33032611/" class="external-link">Ma et al.</a> from two
prostate cancer patients. The raw dataset can be found <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE157703" class="external-link">here</a>
Giotto_SC is processed variable in the <a href="https://giottosuite.com/articles/singlecell_prostate_standard_Sep29_21.html" class="external-link">single
cell RNAseq tutorial</a>. You can also get access to the processed files
of this dataset using getSpatialDataset</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># download data to results directory ####</span></span>
<span><span class="co"># if wget is installed, set method = 'wget'</span></span>
<span><span class="co"># if you run into authentication issues with wget, then add " extra = '--no-check-certificate' "</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/GiottoData/reference/getSpatialDataset.html">getSpatialDataset</a></span><span class="op">(</span>dataset <span class="op">=</span> <span class="st">'scRNA_prostate'</span>, directory <span class="op">=</span> <span class="va">data_directory</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sc_expression</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">data_directory</span>, <span class="st">"/prostate_sc_expression_matrix.csv.gz"</span><span class="op">)</span></span>
<span><span class="va">sc_metadata</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">data_directory</span>, <span class="st">"/prostate_sc_metadata.csv"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">giotto_SC</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GiottoClass/man/createGiottoObject.html" class="external-link">createGiottoObject</a></span><span class="op">(</span></span>
<span>  expression <span class="op">=</span> <span class="va">sc_expression</span>,</span>
<span>  instructions <span class="op">=</span> <span class="va">instrs</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">giotto_SC</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GiottoClass/man/addCellMetadata.html" class="external-link">addCellMetadata</a></span><span class="op">(</span><span class="va">giotto_SC</span>, </span>
<span>                             new_metadata <span class="op">=</span> <span class="fu">data.table</span><span class="fu">::</span><span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="va">sc_metadata</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">giotto_SC</span><span class="op">&lt;-</span> <span class="fu"><a href="../reference/normalizeGiotto.html">normalizeGiotto</a></span><span class="op">(</span><span class="va">giotto_SC</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="page-enrichment">7.1 PAGE enrichment<a class="anchor" aria-label="anchor" href="#page-enrichment"></a>
</h3>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Create PAGE matrix</span></span>
<span><span class="co"># PAGE matrix should be a binary matrix with each row represent a gene marker and each column represent a cell type</span></span>
<span><span class="co"># markers_scran is generated from single cell analysis ()</span></span>
<span><span class="va">markers_scran</span> <span class="op">=</span> <span class="fu"><a href="../reference/findMarkers_one_vs_all.html">findMarkers_one_vs_all</a></span><span class="op">(</span>gobject<span class="op">=</span><span class="va">giotto_SC</span>, </span>
<span>                                       method<span class="op">=</span><span class="st">"scran"</span>,</span>
<span>                                       expression_values<span class="op">=</span><span class="st">"normalized"</span>, </span>
<span>                                       cluster_column<span class="op">=</span><span class="st">'prostate_labels'</span>,</span>
<span>                                       min_feats<span class="op">=</span><span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">top_markers</span> <span class="op">&lt;-</span> <span class="va">markers_scran</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">.SD</span>, <span class="fl">10</span><span class="op">)</span>, by<span class="op">=</span><span class="st">"cluster"</span><span class="op">]</span></span>
<span><span class="va">celltypes</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/levels.html" class="external-link">levels</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">markers_scran</span><span class="op">$</span><span class="va">cluster</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">sign_list</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">celltypes</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">sign_list</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">&lt;-</span><span class="va">top_markers</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">top_markers</span><span class="op">$</span><span class="va">cluster</span> <span class="op">==</span> <span class="va">celltypes</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span>,<span class="op">]</span><span class="op">$</span><span class="va">feats</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">PAGE_matrix</span> <span class="op">=</span> <span class="fu"><a href="../reference/makeSignMatrixPAGE.html">makeSignMatrixPAGE</a></span><span class="op">(</span>sign_names <span class="op">=</span> <span class="va">celltypes</span>,</span>
<span>                                 sign_list <span class="op">=</span> <span class="va">sign_list</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">testcombo</span> <span class="op">=</span> <span class="fu"><a href="../reference/runPAGEEnrich.html">runPAGEEnrich</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">testcombo</span>,</span>
<span>                          sign_matrix <span class="op">=</span> <span class="va">PAGE_matrix</span>,</span>
<span>                          min_overlap_genes <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span></span>
<span><span class="va">cell_types_subset</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/GiottoClass/man/row-plus-colnames-generic.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">PAGE_matrix</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot PAGE enrichment result</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/spatCellPlot.html">spatCellPlot</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">testcombo</span>,</span>
<span>             spat_enr_names <span class="op">=</span> <span class="st">'PAGE'</span>,</span>
<span>             cell_annotation_values <span class="op">=</span> <span class="va">cell_types_subset</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span>,</span>
<span>             cow_n_col <span class="op">=</span> <span class="fl">2</span>,coord_fix_ratio <span class="op">=</span> <span class="cn">NULL</span>, point_size <span class="op">=</span> <span class="fl">1.25</span>,</span>
<span>             save_param <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>save_name <span class="op">=</span> <span class="st">"5a_PAGE_plot"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/visium_prostate_integration/unnamed-chunk-21-1.png"></p>
</div>
<div class="section level3">
<h3 id="hypergeometric-test">7.2 Hypergeometric test<a class="anchor" aria-label="anchor" href="#hypergeometric-test"></a>
</h3>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">testcombo</span> <span class="op">=</span> <span class="fu"><a href="../reference/runHyperGeometricEnrich.html">runHyperGeometricEnrich</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">testcombo</span>,</span>
<span>                                    expression_values <span class="op">=</span> <span class="st">"normalized"</span>,</span>
<span>                                    sign_matrix <span class="op">=</span> <span class="va">PAGE_matrix</span><span class="op">)</span></span>
<span><span class="va">cell_types_subset</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/GiottoClass/man/row-plus-colnames-generic.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">PAGE_matrix</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/spatCellPlot.html">spatCellPlot</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">testcombo</span>, </span>
<span>             spat_enr_names <span class="op">=</span> <span class="st">'hypergeometric'</span>,</span>
<span>             cell_annotation_values <span class="op">=</span> <span class="va">cell_types_subset</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span>,</span>
<span>             cow_n_col <span class="op">=</span> <span class="fl">2</span>,coord_fix_ratio <span class="op">=</span> <span class="cn">NULL</span>, point_size <span class="op">=</span> <span class="fl">1.75</span>,</span>
<span>             save_param <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>save_name <span class="op">=</span> <span class="st">"5b_HyperGeometric_plot"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/visium_prostate_integration/unnamed-chunk-22-1.png"></p>
</div>
<div class="section level3">
<h3 id="rank-enrichment">7.3 Rank Enrichment<a class="anchor" aria-label="anchor" href="#rank-enrichment"></a>
</h3>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Create rank matrix, not that rank matrix is different from PAGE</span></span>
<span><span class="co"># A count matrix and a vector for all cell labels will be needed</span></span>
<span><span class="va">sc_expression_norm</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/GiottoClass/man/getExpression.html" class="external-link">getExpression</a></span><span class="op">(</span><span class="va">giotto_SC</span>,</span>
<span>                                  values <span class="op">=</span> <span class="st">"normalized"</span>,</span>
<span>                                  output <span class="op">=</span> <span class="st">"matrix"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">prostate_feats</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/GiottoClass/man/pDataDT.html" class="external-link">pDataDT</a></span><span class="op">(</span><span class="va">giotto_SC</span><span class="op">)</span><span class="op">$</span><span class="va">prostate_label</span></span>
<span></span>
<span><span class="va">rank_matrix</span> <span class="op">=</span> <span class="fu"><a href="../reference/makeSignMatrixRank.html">makeSignMatrixRank</a></span><span class="op">(</span>sc_matrix <span class="op">=</span> <span class="va">sc_expression_norm</span>,</span>
<span>                                 sc_cluster_ids <span class="op">=</span> <span class="va">prostate_feats</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/GiottoClass/man/row-plus-colnames-generic.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">rank_matrix</span><span class="op">)</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/levels.html" class="external-link">levels</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">prostate_feats</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">testcombo</span> <span class="op">=</span> <span class="fu"><a href="../reference/runRankEnrich.html">runRankEnrich</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">testcombo</span>, sign_matrix <span class="op">=</span> <span class="va">rank_matrix</span>,expression_values <span class="op">=</span> <span class="st">"normalized"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot Rank enrichment result</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/spatCellPlot.html">spatCellPlot2D</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">testcombo</span>,</span>
<span>               spat_enr_names <span class="op">=</span> <span class="st">'rank'</span>,</span>
<span>               cell_annotation_values <span class="op">=</span> <span class="va">cell_types_subset</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span>,</span>
<span>               cow_n_col <span class="op">=</span> <span class="fl">2</span>,coord_fix_ratio <span class="op">=</span> <span class="cn">NULL</span>, point_size <span class="op">=</span> <span class="fl">1</span>,</span>
<span>               save_param <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>save_name <span class="op">=</span> <span class="st">"5c_Rank_plot"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/visium_prostate_integration/unnamed-chunk-23-1.png"></p>
</div>
<div class="section level3">
<h3 id="dwls-deconvolution">7.4 DWLS Deconvolution<a class="anchor" aria-label="anchor" href="#dwls-deconvolution"></a>
</h3>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Create DWLS matrix, not that DWLS matrix is different from PAGE and rank</span></span>
<span><span class="co"># A count matrix a vector for a list of gene signatures and a vector for all cell labels will be needed</span></span>
<span><span class="va">DWLS_matrix</span><span class="op">&lt;-</span><span class="fu"><a href="../reference/makeSignMatrixDWLSfromMatrix.html">makeSignMatrixDWLSfromMatrix</a></span><span class="op">(</span>matrix <span class="op">=</span> <span class="va">sc_expression_norm</span>,</span>
<span>                                          cell_type <span class="op">=</span> <span class="va">prostate_feats</span>,</span>
<span>                                          sign_gene <span class="op">=</span> <span class="va">top_markers</span><span class="op">$</span><span class="va">feats</span><span class="op">)</span></span>
<span><span class="va">testcombo</span> <span class="op">=</span> <span class="fu"><a href="../reference/runDWLSDeconv.html">runDWLSDeconv</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">testcombo</span>, sign_matrix <span class="op">=</span> <span class="va">DWLS_matrix</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># Plot DWLS deconvolution result</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/spatCellPlot.html">spatCellPlot2D</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">testcombo</span>,</span>
<span>               spat_enr_names <span class="op">=</span> <span class="st">'DWLS'</span>,</span>
<span>               cell_annotation_values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/levels.html" class="external-link">levels</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">prostate_feats</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span>,</span>
<span>               cow_n_col <span class="op">=</span> <span class="fl">2</span>,coord_fix_ratio <span class="op">=</span> <span class="cn">NULL</span>, point_size <span class="op">=</span> <span class="fl">1</span>,</span>
<span>               save_param <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>save_name <span class="op">=</span> <span class="st">"5d_DWLS_plot"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/visium_prostate_integration/unnamed-chunk-24-1.png"></p>
</div>
</div>
<div class="section level2">
<h2 id="session-info">8 Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb45"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span>
<span id="cb45-2"><a href="#cb45-2" tabindex="-1"></a></span>
<span id="cb45-3"><a href="#cb45-3" tabindex="-1"></a>R version <span class="dv">4</span>.<span class="fl">2.2</span> (<span class="dv">2022-10-31</span> ucrt)</span>
<span id="cb45-4"><a href="#cb45-4" tabindex="-1"></a>    Platform<span class="sc">:</span> x86_64<span class="sc">-</span>w64<span class="sc">-</span>mingw32<span class="sc">/</span><span class="fu">x64</span> (<span class="dv">64</span><span class="sc">-</span>bit)</span>
<span id="cb45-5"><a href="#cb45-5" tabindex="-1"></a>    Running under<span class="sc">:</span> Windows <span class="dv">10</span> <span class="fu">x64</span> (build <span class="dv">22621</span>)</span>
<span id="cb45-6"><a href="#cb45-6" tabindex="-1"></a></span>
<span id="cb45-7"><a href="#cb45-7" tabindex="-1"></a>    Matrix products<span class="sc">:</span> default</span>
<span id="cb45-8"><a href="#cb45-8" tabindex="-1"></a></span>
<span id="cb45-9"><a href="#cb45-9" tabindex="-1"></a>    locale<span class="sc">:</span></span>
<span id="cb45-10"><a href="#cb45-10" tabindex="-1"></a>    [<span class="dv">1</span>] LC_COLLATE<span class="ot">=</span>English_United States.utf8 </span>
<span id="cb45-11"><a href="#cb45-11" tabindex="-1"></a>    [<span class="dv">2</span>] LC_CTYPE<span class="ot">=</span>English_United States.utf8   </span>
<span id="cb45-12"><a href="#cb45-12" tabindex="-1"></a>    [<span class="dv">3</span>] LC_MONETARY<span class="ot">=</span>English_United States.utf8</span>
<span id="cb45-13"><a href="#cb45-13" tabindex="-1"></a>    [<span class="dv">4</span>] LC_NUMERIC<span class="ot">=</span>C                          </span>
<span id="cb45-14"><a href="#cb45-14" tabindex="-1"></a>    [<span class="dv">5</span>] LC_TIME<span class="ot">=</span>English_United States.utf8    </span>
<span id="cb45-15"><a href="#cb45-15" tabindex="-1"></a></span>
<span id="cb45-16"><a href="#cb45-16" tabindex="-1"></a>    attached base packages<span class="sc">:</span></span>
<span id="cb45-17"><a href="#cb45-17" tabindex="-1"></a>    [<span class="dv">1</span>] stats     graphics  grDevices utils     datasets  methods   base     </span>
<span id="cb45-18"><a href="#cb45-18" tabindex="-1"></a></span>
<span id="cb45-19"><a href="#cb45-19" tabindex="-1"></a>    other attached packages<span class="sc">:</span></span>
<span id="cb45-20"><a href="#cb45-20" tabindex="-1"></a>    [<span class="dv">1</span>] harmony_0.<span class="fl">1.1</span>    Rcpp_1.<span class="fl">0.10</span>      GiottoData_0.<span class="fl">2.1</span> Giotto_3.<span class="fl">3.0</span>    </span>
<span id="cb45-21"><a href="#cb45-21" tabindex="-1"></a>    [<span class="dv">5</span>] testthat_3.<span class="fl">1.5</span>  </span>
<span id="cb45-22"><a href="#cb45-22" tabindex="-1"></a></span>
<span id="cb45-23"><a href="#cb45-23" tabindex="-1"></a>    loaded via a <span class="fu">namespace</span> (and not attached)<span class="sc">:</span></span>
<span id="cb45-24"><a href="#cb45-24" tabindex="-1"></a>      [<span class="dv">1</span>] systemfonts_1.<span class="fl">0.4</span>           plyr_1.<span class="fl">8.8</span>                 </span>
<span id="cb45-25"><a href="#cb45-25" tabindex="-1"></a>      [<span class="dv">3</span>] igraph_1.<span class="fl">4.1</span>                lazyeval_0.<span class="fl">2.2</span>             </span>
<span id="cb45-26"><a href="#cb45-26" tabindex="-1"></a>      [<span class="dv">5</span>] sp_1<span class="fl">.6</span><span class="dv">-0</span>                    BiocParallel_1.<span class="fl">32.6</span>        </span>
<span id="cb45-27"><a href="#cb45-27" tabindex="-1"></a>      [<span class="dv">7</span>] listenv_0.<span class="fl">9.0</span>               usethis_2.<span class="fl">1.6</span>              </span>
<span id="cb45-28"><a href="#cb45-28" tabindex="-1"></a>      [<span class="dv">9</span>] GenomeInfoDb_1.<span class="fl">34.6</span>         ggplot2_3.<span class="fl">4.2</span>              </span>
<span id="cb45-29"><a href="#cb45-29" tabindex="-1"></a>     [<span class="dv">11</span>] digest_0.<span class="fl">6.30</span>               htmltools_0.<span class="fl">5.4</span>            </span>
<span id="cb45-30"><a href="#cb45-30" tabindex="-1"></a>     [<span class="dv">13</span>] magick_2.<span class="fl">7.4</span>                fansi_1.<span class="fl">0.4</span>                </span>
<span id="cb45-31"><a href="#cb45-31" tabindex="-1"></a>     [<span class="dv">15</span>] magrittr_2.<span class="fl">0.3</span>              memoise_2.<span class="fl">0.1</span>              </span>
<span id="cb45-32"><a href="#cb45-32" tabindex="-1"></a>     [<span class="dv">17</span>] ScaledMatrix_1.<span class="fl">6.0</span>          cluster_2.<span class="fl">1.4</span>              </span>
<span id="cb45-33"><a href="#cb45-33" tabindex="-1"></a>     [<span class="dv">19</span>] limma_3.<span class="fl">54.2</span>                remotes_2.<span class="fl">4.2</span>              </span>
<span id="cb45-34"><a href="#cb45-34" tabindex="-1"></a>     [<span class="dv">21</span>] globals_0.<span class="fl">16.2</span>              matrixStats_0.<span class="fl">63.0</span>         </span>
<span id="cb45-35"><a href="#cb45-35" tabindex="-1"></a>     [<span class="dv">23</span>] R.utils_2.<span class="fl">12.2</span>              prettyunits_1.<span class="fl">1.1</span>          </span>
<span id="cb45-36"><a href="#cb45-36" tabindex="-1"></a>     [<span class="dv">25</span>] colorspace_2<span class="fl">.1</span><span class="dv">-0</span>            rappdirs_0.<span class="fl">3.3</span>             </span>
<span id="cb45-37"><a href="#cb45-37" tabindex="-1"></a>     [<span class="dv">27</span>] ggrepel_0.<span class="fl">9.2</span>               textshaping_0.<span class="fl">3.6</span>          </span>
<span id="cb45-38"><a href="#cb45-38" tabindex="-1"></a>     [<span class="dv">29</span>] xfun_0<span class="fl">.38</span>                   dplyr_1.<span class="fl">1.1</span>                </span>
<span id="cb45-39"><a href="#cb45-39" tabindex="-1"></a>     [<span class="dv">31</span>] callr_3.<span class="fl">7.3</span>                 crayon_1.<span class="fl">5.2</span>               </span>
<span id="cb45-40"><a href="#cb45-40" tabindex="-1"></a>     [<span class="dv">33</span>] RCurl_1<span class="fl">.98-1.9</span>              jsonlite_1.<span class="fl">8.3</span>             </span>
<span id="cb45-41"><a href="#cb45-41" tabindex="-1"></a>     [<span class="dv">35</span>] progressr_0.<span class="fl">13.0</span>            glue_1.<span class="fl">6.2</span>                 </span>
<span id="cb45-42"><a href="#cb45-42" tabindex="-1"></a>     [<span class="dv">37</span>] gtable_0.<span class="fl">3.3</span>                zlibbioc_1.<span class="fl">44.0</span>            </span>
<span id="cb45-43"><a href="#cb45-43" tabindex="-1"></a>     [<span class="dv">39</span>] XVector_0.<span class="fl">38.0</span>              DelayedArray_0.<span class="fl">24.0</span>        </span>
<span id="cb45-44"><a href="#cb45-44" tabindex="-1"></a>     [<span class="dv">41</span>] pkgbuild_1.<span class="fl">4.0</span>              BiocSingular_1.<span class="fl">14.0</span>        </span>
<span id="cb45-45"><a href="#cb45-45" tabindex="-1"></a>     [<span class="dv">43</span>] RcppZiggurat_0.<span class="fl">1.6</span>          future.apply_1.<span class="fl">10.0</span>        </span>
<span id="cb45-46"><a href="#cb45-46" tabindex="-1"></a>     [<span class="dv">45</span>] SingleCellExperiment_1.<span class="fl">20.0</span> BiocGenerics_0.<span class="fl">44.0</span>        </span>
<span id="cb45-47"><a href="#cb45-47" tabindex="-1"></a>     [<span class="dv">47</span>] scales_1.<span class="fl">2.1</span>                edgeR_3.<span class="fl">40.1</span>               </span>
<span id="cb45-48"><a href="#cb45-48" tabindex="-1"></a>     [<span class="dv">49</span>] miniUI_0.<span class="dv">1</span>.<span class="fl">1.1</span>              viridisLite_0.<span class="fl">4.2</span>          </span>
<span id="cb45-49"><a href="#cb45-49" tabindex="-1"></a>     [<span class="dv">51</span>] xtable_1<span class="fl">.8</span><span class="dv">-4</span>                dqrng_0.<span class="fl">3.0</span>                </span>
<span id="cb45-50"><a href="#cb45-50" tabindex="-1"></a>     [<span class="dv">53</span>] reticulate_1<span class="fl">.26</span>             rsvd_1.<span class="fl">0.5</span>                 </span>
<span id="cb45-51"><a href="#cb45-51" tabindex="-1"></a>     [<span class="dv">55</span>] stats4_4.<span class="fl">2.2</span>                profvis_0.<span class="fl">3.7</span>              </span>
<span id="cb45-52"><a href="#cb45-52" tabindex="-1"></a>     [<span class="dv">57</span>] metapod_1.<span class="fl">6.0</span>               htmlwidgets_1.<span class="fl">6.2</span>          </span>
<span id="cb45-53"><a href="#cb45-53" tabindex="-1"></a>     [<span class="dv">59</span>] httr_1.<span class="fl">4.5</span>                  RColorBrewer_1<span class="fl">.1</span><span class="dv">-3</span>         </span>
<span id="cb45-54"><a href="#cb45-54" tabindex="-1"></a>     [<span class="dv">61</span>] ellipsis_0.<span class="fl">3.2</span>              scuttle_1.<span class="fl">8.3</span>              </span>
<span id="cb45-55"><a href="#cb45-55" tabindex="-1"></a>     [<span class="dv">63</span>] urlchecker_1.<span class="fl">0.1</span>            pkgconfig_2.<span class="fl">0.3</span>            </span>
<span id="cb45-56"><a href="#cb45-56" tabindex="-1"></a>     [<span class="dv">65</span>] R.methodsS3_1.<span class="fl">8.2</span>           farver_2.<span class="fl">1.1</span>               </span>
<span id="cb45-57"><a href="#cb45-57" tabindex="-1"></a>     [<span class="dv">67</span>] uwot_0.<span class="fl">1.14</span>                 deldir_1<span class="fl">.0</span><span class="dv">-6</span>               </span>
<span id="cb45-58"><a href="#cb45-58" tabindex="-1"></a>     [<span class="dv">69</span>] locfit_1<span class="fl">.5-9.7</span>              utf8_1.<span class="fl">2.3</span>                 </span>
<span id="cb45-59"><a href="#cb45-59" tabindex="-1"></a>     [<span class="dv">71</span>] here_1.<span class="fl">0.1</span>                  tidyselect_1.<span class="fl">2.0</span>           </span>
<span id="cb45-60"><a href="#cb45-60" tabindex="-1"></a>     [<span class="dv">73</span>] labeling_0.<span class="fl">4.2</span>              rlang_1.<span class="fl">1.0</span>                </span>
<span id="cb45-61"><a href="#cb45-61" tabindex="-1"></a>     [<span class="dv">75</span>] reshape2_1.<span class="fl">4.4</span>              later_1.<span class="fl">3.0</span>                </span>
<span id="cb45-62"><a href="#cb45-62" tabindex="-1"></a>     [<span class="dv">77</span>] munsell_0.<span class="fl">5.0</span>               tools_4.<span class="fl">2.2</span>                </span>
<span id="cb45-63"><a href="#cb45-63" tabindex="-1"></a>     [<span class="dv">79</span>] cachem_1.<span class="fl">0.6</span>                cli_3.<span class="fl">4.1</span>                  </span>
<span id="cb45-64"><a href="#cb45-64" tabindex="-1"></a>     [<span class="dv">81</span>] dbscan_1<span class="fl">.1</span><span class="dv">-11</span>               generics_0.<span class="fl">1.3</span>             </span>
<span id="cb45-65"><a href="#cb45-65" tabindex="-1"></a>     [<span class="dv">83</span>] devtools_2.<span class="fl">4.5</span>              evaluate_0<span class="fl">.20</span>              </span>
<span id="cb45-66"><a href="#cb45-66" tabindex="-1"></a>     [<span class="dv">85</span>] stringr_1.<span class="fl">5.0</span>               fastmap_1.<span class="fl">1.0</span>              </span>
<span id="cb45-67"><a href="#cb45-67" tabindex="-1"></a>     [<span class="dv">87</span>] yaml_2.<span class="fl">3.7</span>                  ragg_1.<span class="fl">2.4</span>                 </span>
<span id="cb45-68"><a href="#cb45-68" tabindex="-1"></a>     [<span class="dv">89</span>] processx_3.<span class="fl">8.0</span>              knitr_1<span class="fl">.42</span>                 </span>
<span id="cb45-69"><a href="#cb45-69" tabindex="-1"></a>     [<span class="dv">91</span>] fs_1.<span class="fl">5.2</span>                    purrr_1.<span class="fl">0.1</span>                </span>
<span id="cb45-70"><a href="#cb45-70" tabindex="-1"></a>     [<span class="dv">93</span>] future_1.<span class="fl">32.0</span>               sparseMatrixStats_1.<span class="fl">10.0</span>   </span>
<span id="cb45-71"><a href="#cb45-71" tabindex="-1"></a>     [<span class="dv">95</span>] mime_0<span class="fl">.12</span>                   scran_1.<span class="fl">26.1</span>               </span>
<span id="cb45-72"><a href="#cb45-72" tabindex="-1"></a>     [<span class="dv">97</span>] R.oo_1.<span class="fl">25.0</span>                 brio_1.<span class="fl">1.3</span>                 </span>
<span id="cb45-73"><a href="#cb45-73" tabindex="-1"></a>     [<span class="dv">99</span>] compiler_4.<span class="fl">2.2</span>              rstudioapi_0<span class="fl">.14</span>            </span>
<span id="cb45-74"><a href="#cb45-74" tabindex="-1"></a>    [<span class="dv">101</span>] plotly_4.<span class="fl">10.1</span>               png_0<span class="fl">.1</span><span class="dv">-7</span>                  </span>
<span id="cb45-75"><a href="#cb45-75" tabindex="-1"></a>    [<span class="dv">103</span>] statmod_1.<span class="fl">4.37</span>              tibble_3.<span class="fl">2.1</span>               </span>
<span id="cb45-76"><a href="#cb45-76" tabindex="-1"></a>    [<span class="dv">105</span>] stringi_1.<span class="fl">7.8</span>               ps_1.<span class="fl">7.2</span>                   </span>
<span id="cb45-77"><a href="#cb45-77" tabindex="-1"></a>    [<span class="dv">107</span>] desc_1.<span class="fl">4.2</span>                  bluster_1.<span class="fl">8.0</span>              </span>
<span id="cb45-78"><a href="#cb45-78" tabindex="-1"></a>    [<span class="dv">109</span>] lattice_0<span class="fl">.20</span><span class="dv">-45</span>             Matrix_1<span class="fl">.5</span><span class="dv">-1</span>               </span>
<span id="cb45-79"><a href="#cb45-79" tabindex="-1"></a>    [<span class="dv">111</span>] vctrs_0.<span class="fl">6.1</span>                 pillar_1.<span class="fl">9.0</span>               </span>
<span id="cb45-80"><a href="#cb45-80" tabindex="-1"></a>    [<span class="dv">113</span>] lifecycle_1.<span class="fl">0.3</span>             BiocNeighbors_1.<span class="fl">16.0</span>       </span>
<span id="cb45-81"><a href="#cb45-81" tabindex="-1"></a>    [<span class="dv">115</span>] RcppAnnoy_0.<span class="fl">0.20</span>            data.table_1.<span class="fl">14.6</span>          </span>
<span id="cb45-82"><a href="#cb45-82" tabindex="-1"></a>    [<span class="dv">117</span>] cowplot_1.<span class="fl">1.1</span>               bitops_1<span class="fl">.0</span><span class="dv">-7</span>               </span>
<span id="cb45-83"><a href="#cb45-83" tabindex="-1"></a>    [<span class="dv">119</span>] irlba_2.<span class="dv">3</span>.<span class="fl">5.1</span>               httpuv_1.<span class="fl">6.6</span>               </span>
<span id="cb45-84"><a href="#cb45-84" tabindex="-1"></a>    [<span class="dv">121</span>] GenomicRanges_1.<span class="fl">50.2</span>        R6_2.<span class="fl">5.1</span>                   </span>
<span id="cb45-85"><a href="#cb45-85" tabindex="-1"></a>    [<span class="dv">123</span>] promises_1.<span class="dv">2</span>.<span class="fl">0.1</span>            IRanges_2.<span class="fl">32.0</span>             </span>
<span id="cb45-86"><a href="#cb45-86" tabindex="-1"></a>    [<span class="dv">125</span>] parallelly_1.<span class="fl">35.0</span>           sessioninfo_1.<span class="fl">2.2</span>          </span>
<span id="cb45-87"><a href="#cb45-87" tabindex="-1"></a>    [<span class="dv">127</span>] codetools_0<span class="fl">.2</span><span class="dv">-18</span>            pkgload_1.<span class="fl">3.2</span>              </span>
<span id="cb45-88"><a href="#cb45-88" tabindex="-1"></a>    [<span class="dv">129</span>] SummarizedExperiment_1.<span class="fl">28.0</span> rprojroot_2.<span class="fl">0.3</span>            </span>
<span id="cb45-89"><a href="#cb45-89" tabindex="-1"></a>    [<span class="dv">131</span>] withr_2.<span class="fl">5.0</span>                 S4Vectors_0.<span class="fl">36.2</span>           </span>
<span id="cb45-90"><a href="#cb45-90" tabindex="-1"></a>    [<span class="dv">133</span>] GenomeInfoDbData_1.<span class="fl">2.9</span>      parallel_4.<span class="fl">2.2</span>             </span>
<span id="cb45-91"><a href="#cb45-91" tabindex="-1"></a>    [<span class="dv">135</span>] terra_1<span class="fl">.7</span><span class="dv">-18</span>                quadprog_1<span class="fl">.5</span><span class="dv">-8</span>             </span>
<span id="cb45-92"><a href="#cb45-92" tabindex="-1"></a>    [<span class="dv">137</span>] grid_4.<span class="fl">2.2</span>                  beachmat_2.<span class="fl">14.0</span>            </span>
<span id="cb45-93"><a href="#cb45-93" tabindex="-1"></a>    [<span class="dv">139</span>] tidyr_1.<span class="fl">3.0</span>                 Rfast_2.<span class="fl">0.6</span>                </span>
<span id="cb45-94"><a href="#cb45-94" tabindex="-1"></a>    [<span class="dv">141</span>] DelayedMatrixStats_1.<span class="fl">20.0</span>   rmarkdown_2<span class="fl">.21</span>             </span>
<span id="cb45-95"><a href="#cb45-95" tabindex="-1"></a>    [<span class="dv">143</span>] MatrixGenerics_1.<span class="fl">10.0</span>       Rtsne_0<span class="fl">.16</span>                 </span>
<span id="cb45-96"><a href="#cb45-96" tabindex="-1"></a>    [<span class="dv">145</span>] Biobase_2.<span class="fl">58.0</span>              shiny_1.<span class="fl">7.4</span>                </span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Ruben Dries, Jiaji George Chen, Natalie Del Rossi, Guo-Cheng Yuan, Matthew Oâ€™Brien, <a href="https://josschavezf.github.io" class="external-link">Joselyn ChÃ¡vez</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
