<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="Giotto">
<title>Visium CytAssist Multi-omics Human Glioblastoma • Giotto</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/JetBrains_Mono-0.4.7/font.css" rel="stylesheet">
<!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Visium CytAssist Multi-omics Human Glioblastoma">
<meta property="og:description" content="Giotto">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">Giotto</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">4.0.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-get-started">Get started</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-get-started">
    <a class="dropdown-item" href="../articles/installation.html">Giotto Installation</a>
    <a class="dropdown-item" href="../articles/ecosystem.html">Giotto Ecosystem</a>
    <a class="dropdown-item" href="../articles/configuration.html">Giotto Configuration</a>
    <a class="dropdown-item" href="../articles/object_creation.html">Giotto Object Creation</a>
    <a class="dropdown-item" href="../articles/structure.html">Giotto Class Structure</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Documentation</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-tutorials">Tutorials</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-tutorials">
    <h6 class="dropdown-header" data-toc-skip>Seq-Scope</h6>
    <a class="dropdown-item" href="../articles/seqscope_mouse_liver.html">Mouse liver</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Visium</h6>
    <a class="dropdown-item" href="../articles/visium_mouse_kidney.html">Visium Mouse Kidney</a>
    <a class="dropdown-item" href="../articles/visium_mouse_brain.html">Visium Mouse Brain</a>
    <a class="dropdown-item" href="../articles/visium_prostate_integration.html">Visium Prostate Integration</a>
    <a class="dropdown-item" href="../articles/visium_cytassist_lungcancer.html">CytAssist Human Lung Cancer</a>
    <a class="dropdown-item" href="../articles/visium_cytassist_human_glioblastoma.html">CytAssist Multi-omics Human Glioblastoma</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Spatial Genomics</h6>
    <a class="dropdown-item" href="../articles/spatial_genomics_mouse_kidney.html">Mouse Kidney</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Xenium</h6>
    <a class="dropdown-item" href="../articles/xenium_breast_cancer.html">Xenium Human breast cancer</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Nanostring CosMx</h6>
    <a class="dropdown-item" href="../articles/nanostring_cosmx_lung_cancer.html">Nanostring CosMx Subcellular Lung Cancer</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>MERFISH</h6>
    <a class="dropdown-item" href="../articles/merfish_mouse_hypothalamic.html">MERFISH Mouse Hypothalamic Preoptic Region</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Vizgen</h6>
    <a class="dropdown-item" href="../articles/vizgen_mouse_brain.html">Vizgen Mouse Brain Receptor Map</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>seqFISH</h6>
    <a class="dropdown-item" href="../articles/mini_seqfish.html">mini seqFISH</a>
    <a class="dropdown-item" href="../articles/seqfish_cortex.html">SeqFish+ Mouse Cortex Example</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>osmFISH</h6>
    <a class="dropdown-item" href="../articles/osmfish_mouse_ss_cortex.html">osmFISH Mouse SS Cortex</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Resolve Bioscience</h6>
    <a class="dropdown-item" href="../articles/resolve_biosciences.html">Resolve Bioscience Breast Cancer Subcellular</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>CODEX</h6>
    <a class="dropdown-item" href="../articles/codex_mouse_spleen.html">Codex Mouse Spleen</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Single Cell RNA Sequencing</h6>
    <a class="dropdown-item" href="../articles/singlecell_prostate_standard.html">Prostate</a>
    <a class="dropdown-item" href="../articles/singlecell_prostate_integration.html">Prostate Integration</a>
  </div>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-tools">Tools</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-tools">
    <h6 class="dropdown-header" data-toc-skip>Spatial manipulation</h6>
    <a class="dropdown-item" href="../articles/spatial_transformations.html">Spatial transformations</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Image alignment</h6>
    <a class="dropdown-item" href="../articles/image_alignment.html">Image alignment</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Interactive selection</h6>
    <a class="dropdown-item" href="../articles/interactive_selection.html">Interactive selection</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Visualizations</h6>
    <a class="dropdown-item" href="../articles/visualizations.html">Visualizations</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Saving and Loading</h6>
    <a class="dropdown-item" href="../articles/saving_object.html">Saving a Giotto object</a>
    <a class="dropdown-item" href="../articles/saving_plots.html">Saving plots</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Interoperability</h6>
    <a class="dropdown-item" href="../articles/interoperability_anndata.html">AnnData</a>
  </div>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-try-giotto-in-the-cloud">Try Giotto in the Cloud</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-try-giotto-in-the-cloud">
    <a class="dropdown-item" href="../articles/terra.html">Terra.bio</a>
    <a class="dropdown-item" href="../articles/docker.html">Docker</a>
    <div class="dropdown-divider"></div>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/presentations.html">Presentations</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/branches.html">Branches</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">News</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/drieslab/Giotto/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Visium CytAssist Multi-omics Human Glioblastoma</h1>
            
            <h4 data-toc-skip class="date">9/7/23</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/drieslab/Giotto/blob/HEAD/vignettes/visium_cytassist_human_glioblastoma.Rmd" class="external-link"><code>vignettes/visium_cytassist_human_glioblastoma.Rmd</code></a></small>
      <div class="d-none name"><code>visium_cytassist_human_glioblastoma.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="dataset-explanation">1 Dataset explanation<a class="anchor" aria-label="anchor" href="#dataset-explanation"></a>
</h2>
<p>The <a href="https://www.10xgenomics.com/resources/datasets/gene-and-protein-expression-library-of-human-glioblastoma-cytassist-ffpe-2-standard/" class="external-link">Human
glioblastoma (FFPE) dataset</a> was obtained from 10x Genomics. The
tissue was sectioned as described in Visium CytAssist Spatial Gene
Expression for FFPE – Tissue Preparation Guide Demonstrated Protocol
(CG000518). 5 µm tissue sections were placed on Superfrost glass slides,
then IF stained following deparaffinization, then hard coverslipped.
Sections were imaged, decoverslipped, followed by Demonstrated Protocol
(CG000494).</p>
<p>More information about this dataset can be found <a href="https://www.10xgenomics.com/resources/datasets/gene-and-protein-expression-library-of-human-glioblastoma-cytassist-ffpe-2-standard" class="external-link">here</a>.</p>
</div>
<div class="section level2">
<h2 id="start-giotto">2 Start Giotto<a class="anchor" aria-label="anchor" href="#start-giotto"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Ensure Giotto Suite is installed</span></span>
<span><span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="st">"Giotto"</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/installed.packages.html" class="external-link">installed.packages</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"drieslab/Giotto"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://drieslab.github.io/Giotto/">Giotto</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Ensure the Python environment for Giotto has been installed</span></span>
<span><span class="va">genv_exists</span> <span class="op">=</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/checkGiottoEnvironment.html">checkGiottoEnvironment</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="va">genv_exists</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="co"># The following command need only be run once to install the Giotto environment</span></span>
<span>  <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/installGiottoEnvironment.html">installGiottoEnvironment</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># 1. set results directory</span></span>
<span><span class="va">results_directory</span> <span class="op">=</span> <span class="st">'results'</span></span>
<span></span>
<span><span class="co"># 2. set giotto python path</span></span>
<span><span class="co"># set python path to your preferred python version path</span></span>
<span><span class="co"># set python path to NULL if you want to automatically install (only the 1st time) and use the giotto miniconda environment</span></span>
<span><span class="va">python_path</span> <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="kw">if</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NULL.html" class="external-link">is.null</a></span><span class="op">(</span><span class="va">python_path</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/installGiottoEnvironment.html">installGiottoEnvironment</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co"># 3. create giotto instructions</span></span>
<span><span class="va">instrs</span> <span class="op">=</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/createGiottoInstructions.html">createGiottoInstructions</a></span><span class="op">(</span>save_dir <span class="op">=</span> <span class="va">results_directory</span>,</span>
<span>                                  save_plot <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                                  show_plot <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                                  python_path <span class="op">=</span> <span class="va">python_path</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="create-giotto-object">3 Create Giotto object<a class="anchor" aria-label="anchor" href="#create-giotto-object"></a>
</h2>
<p>The minimum requirements are</p>
<ul>
<li>matrix with expression information (or path to)</li>
<li>x,y(,z) coordinates for cells or spots (or path to)</li>
</ul>
<p>createGiottoVisiumObject() will automatically detect both rna and
protein modalities within the expression matrix creating a multi-omics
Giotto object.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Provide path to visium folder</span></span>
<span><span class="va">data_directory</span> <span class="op">&lt;-</span> <span class="st">'path/to/visium/data'</span></span>
<span></span>
<span><span class="co"># Create Giotto object</span></span>
<span><span class="va">visium</span> <span class="op">=</span> <span class="fu"><a href="../reference/createGiottoVisiumObject.html">createGiottoVisiumObject</a></span><span class="op">(</span>visium_dir <span class="op">=</span> <span class="va">data_directory</span>,</span>
<span>                                  expr_data <span class="op">=</span> <span class="st">'raw'</span>,</span>
<span>                                  png_name <span class="op">=</span> <span class="st">'tissue_lowres_image.png'</span>,</span>
<span>                                  gene_column_index <span class="op">=</span> <span class="fl">2</span>,</span>
<span>                                  instructions <span class="op">=</span> <span class="va">instrs</span><span class="op">)</span></span></code></pre></div>
<p>How to work with Giotto instructions that are part of your Giotto
object:</p>
<ul>
<li>Show the instructions associated with your Giotto object with
showGiottoInstructions()</li>
<li>Change one or more instructions with changeGiottoInstructions()</li>
<li>Replace all instructions at once with
replaceGiottoInstructions()</li>
<li>Read or get a specific Giotto instruction with
readGiottoInstructions()</li>
</ul>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># show instructions associated with the giotto object</span></span>
<span></span>
<span><span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/showGiottoInstructions.html">showGiottoInstructions</a></span><span class="op">(</span><span class="va">visium</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="processing">4 Processing<a class="anchor" aria-label="anchor" href="#processing"></a>
</h2>
<ul>
<li>Filter features and cells based on detection frequencies</li>
<li>Normalize expression matrix (log transformation, scaling factor
and/or z-scores)</li>
<li>Add cell and feature statistics (optional)</li>
<li>Adjust expression matrix for technical covariates or batches
(optional).</li>
</ul>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Subset on spots that were covered by tissue</span></span>
<span><span class="va">metadata</span> <span class="op">=</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/pDataDT.html">pDataDT</a></span><span class="op">(</span><span class="va">visium</span><span class="op">)</span></span>
<span><span class="va">in_tissue_barcodes</span> <span class="op">=</span> <span class="va">metadata</span><span class="op">[</span><span class="va">in_tissue</span> <span class="op">==</span> <span class="fl">1</span><span class="op">]</span><span class="op">$</span><span class="va">cell_ID</span></span>
<span><span class="va">visium</span> <span class="op">=</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/subsetGiotto.html">subsetGiotto</a></span><span class="op">(</span><span class="va">visium</span>, cell_ids <span class="op">=</span> <span class="va">in_tissue_barcodes</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Visualize aligned tissue</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/spatPlot.html">spatPlot2D</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">visium</span>,</span>
<span>           point_alpha <span class="op">=</span> <span class="fl">0.7</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/visium_cytassist_human_glioblastoma/1-spatPlot2D.png"></p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Filtering, normalization, and statistics</span></span>
<span></span>
<span><span class="co">## RNA feature</span></span>
<span><span class="va">visium</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/filterGiotto.html">filterGiotto</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">visium</span>,</span>
<span>                       expression_threshold <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                       feat_det_in_min_cells <span class="op">=</span> <span class="fl">50</span>,</span>
<span>                       min_det_feats_per_cell <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>                       expression_values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'raw'</span><span class="op">)</span>,</span>
<span>                       verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">visium</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/normalizeGiotto.html">normalizeGiotto</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">visium</span>,</span>
<span>                          scalefactor <span class="op">=</span> <span class="fl">6000</span>,</span>
<span>                          verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">visium</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/addStatistics.html">addStatistics</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">visium</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Visualize number of features after processing</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/spatPlot.html">spatPlot2D</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">visium</span>,</span>
<span>           point_alpha <span class="op">=</span> <span class="fl">0.7</span>,</span>
<span>           cell_color <span class="op">=</span> <span class="st">'nr_feats'</span>,</span>
<span>           color_as_factor <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/visium_cytassist_human_glioblastoma/2-spatPlot2D.png"></p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Protein feature </span></span>
<span><span class="va">visium</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/filterGiotto.html">filterGiotto</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">visium</span>, </span>
<span>                       spat_unit <span class="op">=</span> <span class="st">'cell'</span>,</span>
<span>                       feat_type <span class="op">=</span> <span class="st">'protein'</span>,</span>
<span>                       expression_threshold <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                       feat_det_in_min_cells <span class="op">=</span> <span class="fl">50</span>, </span>
<span>                       min_det_feats_per_cell <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                       expression_values <span class="op">=</span> <span class="st">'raw'</span>, </span>
<span>                       verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">visium</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/normalizeGiotto.html">normalizeGiotto</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">visium</span>,</span>
<span>                          spat_unit <span class="op">=</span> <span class="st">'cell'</span>, </span>
<span>                          feat_type <span class="op">=</span> <span class="st">'protein'</span>, </span>
<span>                          scalefactor <span class="op">=</span> <span class="fl">6000</span>, </span>
<span>                          verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">visium</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/addStatistics.html">addStatistics</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">visium</span>,</span>
<span>                        spat_unit <span class="op">=</span> <span class="st">'cell'</span>, </span>
<span>                        feat_type <span class="op">=</span> <span class="st">'protein'</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Visualize number of features after processing </span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/spatPlot.html">spatPlot2D</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">visium</span>,</span>
<span>           spat_unit <span class="op">=</span> <span class="st">'cell'</span>, </span>
<span>           feat_type <span class="op">=</span> <span class="st">'protein'</span>, </span>
<span>           point_alpha <span class="op">=</span> <span class="fl">0.7</span>, </span>
<span>           cell_color <span class="op">=</span> <span class="st">'nr_feats'</span>, </span>
<span>           color_as_factor <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/visium_cytassist_human_glioblastoma/3-spatPlot2D.png"></p>
</div>
<div class="section level2">
<h2 id="dimention-reduction">5 Dimention Reduction<a class="anchor" aria-label="anchor" href="#dimention-reduction"></a>
</h2>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Identify highly variable features (HVF)</span></span>
<span><span class="va">visium</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateHVF.html">calculateHVF</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">visium</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/visium_cytassist_human_glioblastoma/4-HVFplot.png"></p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># PCA</span></span>
<span></span>
<span><span class="co">## RNA</span></span>
<span><span class="va">visium</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runPCA.html">runPCA</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">visium</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/screePlot.html">screePlot</a></span><span class="op">(</span><span class="va">visium</span>, ncp <span class="op">=</span> <span class="fl">30</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/visium_cytassist_human_glioblastoma/5-screePlot.png"></p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">### Visualize RNA PCA</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/plotPCA.html">plotPCA</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">visium</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/visium_cytassist_human_glioblastoma/6-PCA.png"></p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Protein</span></span>
<span><span class="va">visium</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runPCA.html">runPCA</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">visium</span>,</span>
<span>                 spat_unit <span class="op">=</span> <span class="st">'cell'</span>,</span>
<span>                 feat_type <span class="op">=</span> <span class="st">'protein'</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/screePlot.html">screePlot</a></span><span class="op">(</span><span class="va">visium</span>,</span>
<span>          spat_unit <span class="op">=</span> <span class="st">'cell'</span>,</span>
<span>          feat_type <span class="op">=</span> <span class="st">'protein'</span>,</span>
<span>          ncp <span class="op">=</span> <span class="fl">30</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/visium_cytassist_human_glioblastoma/7-screePlot.png"></p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">### Visualize Protein PCA</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/plotPCA.html">plotPCA</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">visium</span>,</span>
<span>        spat_unit <span class="op">=</span> <span class="st">'cell'</span>,</span>
<span>        feat_type <span class="op">=</span> <span class="st">'protein'</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/visium_cytassist_human_glioblastoma/8-PCA.png"></p>
</div>
<div class="section level2">
<h2 id="clustering">6 Clustering<a class="anchor" aria-label="anchor" href="#clustering"></a>
</h2>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># cluster and run UMAP</span></span>
<span><span class="co"># sNN network (default)</span></span>
<span></span>
<span><span class="co">## RNA feature</span></span>
<span><span class="va">visium</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/createNearestNetwork.html">createNearestNetwork</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">visium</span>,</span>
<span>                               dimensions_to_use <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span>,</span>
<span>                               k <span class="op">=</span> <span class="fl">30</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Protein feature</span></span>
<span><span class="va">visium</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/createNearestNetwork.html">createNearestNetwork</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">visium</span>,</span>
<span>                               spat_unit <span class="op">=</span> <span class="st">'cell'</span>,</span>
<span>                               feat_type <span class="op">=</span> <span class="st">'protein'</span>,</span>
<span>                               dimensions_to_use <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span>,</span>
<span>                               k <span class="op">=</span> <span class="fl">30</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Leiden clustering</span></span>
<span></span>
<span><span class="co">## RNA feature</span></span>
<span><span class="va">visium</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/doLeidenCluster.html">doLeidenCluster</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">visium</span>,</span>
<span>                          resolution <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                          n_iterations <span class="op">=</span> <span class="fl">1000</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Protein feature</span></span>
<span><span class="va">visium</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/doLeidenCluster.html">doLeidenCluster</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">visium</span>,</span>
<span>                          spat_unit <span class="op">=</span> <span class="st">'cell'</span>,</span>
<span>                          feat_type <span class="op">=</span> <span class="st">'protein'</span>,</span>
<span>                          resolution <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                          n_iterations <span class="op">=</span> <span class="fl">1000</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># UMAP</span></span>
<span></span>
<span><span class="co">## RNA feature</span></span>
<span><span class="va">visium</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runUMAP.html">runUMAP</a></span><span class="op">(</span><span class="va">visium</span>,</span>
<span>                  dimensions_to_use <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/plotUMAP.html">plotUMAP</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">visium</span>,</span>
<span>         cell_color <span class="op">=</span> <span class="st">'leiden_clus'</span>,</span>
<span>         show_NN_network <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>         point_size <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/visium_cytassist_human_glioblastoma/9-UMAP.png"></p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Protein feature </span></span>
<span><span class="va">visium</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runUMAP.html">runUMAP</a></span><span class="op">(</span><span class="va">visium</span>,</span>
<span>                  spat_unit <span class="op">=</span> <span class="st">'cell'</span>, </span>
<span>                  feat_type <span class="op">=</span> <span class="st">'protein'</span>, </span>
<span>                  dimensions_to_use <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/plotUMAP.html">plotUMAP</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">visium</span>,</span>
<span>         spat_unit <span class="op">=</span> <span class="st">'cell'</span>, </span>
<span>         feat_type <span class="op">=</span> <span class="st">'protein'</span>, </span>
<span>         cell_color <span class="op">=</span> <span class="st">'leiden_clus'</span>, </span>
<span>         show_NN_network <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>         point_size <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/visium_cytassist_human_glioblastoma/10-UMAP.png"></p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Visualize spatial plot</span></span>
<span></span>
<span><span class="co">## RNA feature</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/spatPlot.html">spatPlot2D</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">visium</span>,</span>
<span>           show_image <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>           cell_color <span class="op">=</span> <span class="st">'leiden_clus'</span>,</span>
<span>           point_size <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/visium_cytassist_human_glioblastoma/11-rna_spatPlot2D.png"></p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Protein feature</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/spatPlot.html">spatPlot2D</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">visium</span>,</span>
<span>           spat_unit <span class="op">=</span> <span class="st">'cell'</span>,</span>
<span>           feat_type <span class="op">=</span> <span class="st">'protein'</span>,</span>
<span>           show_image <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>           cell_color <span class="op">=</span> <span class="st">'leiden_clus'</span>,</span>
<span>           point_size <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/visium_cytassist_human_glioblastoma/12-protein_spatPlot2D.png"></p>
</div>
<div class="section level2">
<h2 id="multi-omics-integration">7 Multi-omics integration<a class="anchor" aria-label="anchor" href="#multi-omics-integration"></a>
</h2>
<p>The Weighted Nearest Neighbors allows to integrate two or more
modalities acquired from the same sample. WNN will re-calculate the
clustering to provide an integrated umap and leiden clustering. For
running WNN, the Giotto object must contain the results of running PCA
calculation for each modality.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Calculate kNN</span></span>
<span></span>
<span><span class="co">## RNA modality</span></span>
<span><span class="va">visium</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/createNearestNetwork.html">createNearestNetwork</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">visium</span>,</span>
<span>                               type <span class="op">=</span> <span class="st">'kNN'</span>,</span>
<span>                               dimensions_to_use <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span>,</span>
<span>                               k <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Protein modality</span></span>
<span><span class="va">visium</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/createNearestNetwork.html">createNearestNetwork</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">visium</span>,</span>
<span>                               spat_unit <span class="op">=</span> <span class="st">'cell'</span>,</span>
<span>                               feat_type <span class="op">=</span> <span class="st">'protein'</span>,</span>
<span>                               type <span class="op">=</span> <span class="st">'kNN'</span>,</span>
<span>                               dimensions_to_use <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span>,</span>
<span>                               k <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># Run WNN</span></span>
<span><span class="va">visium</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runWNN.html">runWNN</a></span><span class="op">(</span><span class="va">visium</span>,</span>
<span>                 spat_unit <span class="op">=</span> <span class="st">"cell"</span>,</span>
<span>                 modality_1 <span class="op">=</span> <span class="st">"rna"</span>,</span>
<span>                 modality_2 <span class="op">=</span> <span class="st">"protein"</span>,</span>
<span>                 pca_name_modality_1 <span class="op">=</span> <span class="st">"pca"</span>,</span>
<span>                 pca_name_modality_2 <span class="op">=</span> <span class="st">"protein.pca"</span>,</span>
<span>                 k <span class="op">=</span> <span class="fl">20</span>,</span>
<span>                 integrated_feat_type <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>                 matrix_result_name <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>                 w_name_modality_1 <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>                 w_name_modality_2 <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>                 verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Run Integrated umap</span></span>
<span><span class="va">visium</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runIntegratedUMAP.html">runIntegratedUMAP</a></span><span class="op">(</span><span class="va">visium</span>,</span>
<span>                            modality1 <span class="op">=</span> <span class="st">"rna"</span>,</span>
<span>                            modality2 <span class="op">=</span> <span class="st">"protein"</span>,</span>
<span>                            spread <span class="op">=</span> <span class="fl">5</span>,</span>
<span>                            min_dist <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>                            force <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Calculate integrated clusters</span></span>
<span><span class="va">visium</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/doLeidenCluster.html">doLeidenCluster</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">visium</span>,</span>
<span>                          spat_unit <span class="op">=</span> <span class="st">"cell"</span>,</span>
<span>                          feat_type <span class="op">=</span> <span class="st">"rna"</span>,</span>
<span>                          nn_network_to_use <span class="op">=</span> <span class="st">"kNN"</span>,</span>
<span>                          network_name <span class="op">=</span> <span class="st">"integrated_kNN"</span>,</span>
<span>                          name <span class="op">=</span> <span class="st">"integrated_leiden_clus"</span>,</span>
<span>                          resolution <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Visualize integrated umap</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/plotUMAP.html">plotUMAP</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">visium</span>,</span>
<span>         spat_unit <span class="op">=</span> <span class="st">"cell"</span>,</span>
<span>         feat_type <span class="op">=</span> <span class="st">"rna"</span>,</span>
<span>         cell_color <span class="op">=</span> <span class="st">'integrated_leiden_clus'</span>,</span>
<span>         dim_reduction_name <span class="op">=</span> <span class="st">"integrated.umap"</span>,</span>
<span>         point_size <span class="op">=</span> <span class="fl">1.5</span>,</span>
<span>         title <span class="op">=</span> <span class="st">"Integrated UMAP using Integrated Leiden clusters"</span>,</span>
<span>         axis_title <span class="op">=</span> <span class="fl">12</span>,</span>
<span>         axis_text <span class="op">=</span> <span class="fl">10</span> <span class="op">)</span></span></code></pre></div>
<p><img src="images/visium_cytassist_human_glioblastoma/13-UMAP.png"></p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Visualize spatial plot with integrated clusters</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/spatPlot.html">spatPlot2D</a></span><span class="op">(</span><span class="va">visium</span>,</span>
<span>           spat_unit <span class="op">=</span> <span class="st">"cell"</span>,</span>
<span>           feat_type <span class="op">=</span> <span class="st">"rna"</span>,</span>
<span>           cell_color <span class="op">=</span> <span class="st">"integrated_leiden_clus"</span>,</span>
<span>           point_size <span class="op">=</span> <span class="fl">2</span>,</span>
<span>           show_image <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>           title <span class="op">=</span> <span class="st">"Integrated Leiden clustering"</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/visium_cytassist_human_glioblastoma/14-integrated_spatPlot2D.png"></p>
</div>
<div class="section level2">
<h2 id="session-info">8 Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span>
<span id="cb19-2"><a href="#cb19-2" tabindex="-1"></a></span>
<span id="cb19-3"><a href="#cb19-3" tabindex="-1"></a>R version <span class="dv">4</span>.<span class="fl">3.1</span> (<span class="dv">2023-06-16</span>)</span>
<span id="cb19-4"><a href="#cb19-4" tabindex="-1"></a>Platform<span class="sc">:</span> x86_64<span class="sc">-</span>apple<span class="sc">-</span><span class="fu">darwin20</span> (<span class="dv">64</span><span class="sc">-</span>bit)</span>
<span id="cb19-5"><a href="#cb19-5" tabindex="-1"></a>Running under<span class="sc">:</span> macOS Ventura <span class="dv">13</span>.<span class="fl">5.1</span></span>
<span id="cb19-6"><a href="#cb19-6" tabindex="-1"></a></span>
<span id="cb19-7"><a href="#cb19-7" tabindex="-1"></a>Matrix products<span class="sc">:</span> default</span>
<span id="cb19-8"><a href="#cb19-8" tabindex="-1"></a>BLAS<span class="sc">:</span>   <span class="er">/</span>System<span class="sc">/</span>Library<span class="sc">/</span>Frameworks<span class="sc">/</span>Accelerate.framework<span class="sc">/</span>Versions<span class="sc">/</span>A                <span class="sc">/</span>Frameworks<span class="sc">/</span>vecLib.framework<span class="sc">/</span>Versions<span class="sc">/</span>A<span class="sc">/</span>libBLAS.dylib</span>
<span id="cb19-9"><a href="#cb19-9" tabindex="-1"></a>LAPACK<span class="sc">:</span> <span class="er">/</span>Library<span class="sc">/</span>Frameworks<span class="sc">/</span>R.framework<span class="sc">/</span>Versions<span class="sc">/</span><span class="fl">4.3</span><span class="sc">-</span>x86_64<span class="sc">/</span>Resources<span class="sc">/</span>lib          <span class="sc">/</span>libRlapack.dylib;  LAPACK version <span class="dv">3</span>.<span class="fl">11.0</span></span>
<span id="cb19-10"><a href="#cb19-10" tabindex="-1"></a></span>
<span id="cb19-11"><a href="#cb19-11" tabindex="-1"></a>locale<span class="sc">:</span></span>
<span id="cb19-12"><a href="#cb19-12" tabindex="-1"></a>[<span class="dv">1</span>] en_US.UTF<span class="dv">-8</span><span class="sc">/</span>en_US.UTF<span class="dv">-8</span><span class="sc">/</span>en_US.UTF<span class="dv">-8</span><span class="sc">/</span>C<span class="sc">/</span>en_US.UTF<span class="dv">-8</span><span class="sc">/</span>en_US.UTF<span class="dv">-8</span></span>
<span id="cb19-13"><a href="#cb19-13" tabindex="-1"></a></span>
<span id="cb19-14"><a href="#cb19-14" tabindex="-1"></a>time zone<span class="sc">:</span> America<span class="sc">/</span>New_York</span>
<span id="cb19-15"><a href="#cb19-15" tabindex="-1"></a>tzcode source<span class="sc">:</span> internal</span>
<span id="cb19-16"><a href="#cb19-16" tabindex="-1"></a></span>
<span id="cb19-17"><a href="#cb19-17" tabindex="-1"></a>attached base packages<span class="sc">:</span></span>
<span id="cb19-18"><a href="#cb19-18" tabindex="-1"></a>[<span class="dv">1</span>] stats     graphics  grDevices utils     datasets  methods   base</span>
<span id="cb19-19"><a href="#cb19-19" tabindex="-1"></a></span>
<span id="cb19-20"><a href="#cb19-20" tabindex="-1"></a>other attached packages<span class="sc">:</span></span>
<span id="cb19-21"><a href="#cb19-21" tabindex="-1"></a>[<span class="dv">1</span>] Giotto_3.<span class="fl">3.2</span>             GiottoVisuals_0.<span class="dv">0</span>.<span class="fl">0.9002</span></span>
<span id="cb19-22"><a href="#cb19-22" tabindex="-1"></a>[<span class="dv">3</span>] GiottoClass_0.<span class="dv">0</span>.<span class="fl">0.9003</span>   GiottoUtils_0.<span class="dv">0</span>.<span class="fl">0.9002</span></span>
<span id="cb19-23"><a href="#cb19-23" tabindex="-1"></a></span>
<span id="cb19-24"><a href="#cb19-24" tabindex="-1"></a>loaded via a <span class="fu">namespace</span> (and not attached)<span class="sc">:</span></span>
<span id="cb19-25"><a href="#cb19-25" tabindex="-1"></a>[<span class="dv">1</span>] gtable_0.<span class="fl">3.4</span>       xfun_0<span class="fl">.40</span>          ggplot2_3.<span class="fl">4.3</span></span>
<span id="cb19-26"><a href="#cb19-26" tabindex="-1"></a>[<span class="dv">4</span>] htmlwidgets_1.<span class="fl">6.2</span>  devtools_2.<span class="fl">4.5</span>     remotes_2.<span class="dv">4</span>.<span class="fl">2.1</span></span>
<span id="cb19-27"><a href="#cb19-27" tabindex="-1"></a>[<span class="dv">7</span>] processx_3.<span class="fl">8.2</span>     lattice_0<span class="fl">.21</span><span class="dv">-8</span>     callr_3.<span class="fl">7.3</span></span>
<span id="cb19-28"><a href="#cb19-28" tabindex="-1"></a>[<span class="dv">10</span>] vctrs_0.<span class="fl">6.3</span>        tools_4.<span class="fl">3.1</span>        ps_1.<span class="fl">7.5</span></span>
<span id="cb19-29"><a href="#cb19-29" tabindex="-1"></a>[<span class="dv">13</span>] generics_0.<span class="fl">1.3</span>     parallel_4.<span class="fl">3.1</span>     tibble_3.<span class="fl">2.1</span></span>
<span id="cb19-30"><a href="#cb19-30" tabindex="-1"></a>[<span class="dv">16</span>] fansi_1.<span class="fl">0.4</span>        colorRamp2_0.<span class="fl">1.0</span>   pkgconfig_2.<span class="fl">0.3</span></span>
<span id="cb19-31"><a href="#cb19-31" tabindex="-1"></a>[<span class="dv">19</span>] Matrix_1<span class="fl">.6</span><span class="dv">-1</span>       data.table_1.<span class="fl">14.8</span>  checkmate_2.<span class="fl">2.0</span></span>
<span id="cb19-32"><a href="#cb19-32" tabindex="-1"></a>[<span class="dv">22</span>] RColorBrewer_1<span class="fl">.1</span><span class="dv">-3</span> lifecycle_1.<span class="fl">0.3</span>    farver_2.<span class="fl">1.1</span></span>
<span id="cb19-33"><a href="#cb19-33" tabindex="-1"></a>[<span class="dv">25</span>] compiler_4.<span class="fl">3.1</span>     stringr_1.<span class="fl">5.0</span>      textshaping_0.<span class="fl">3.6</span></span>
<span id="cb19-34"><a href="#cb19-34" tabindex="-1"></a>[<span class="dv">28</span>] munsell_0.<span class="fl">5.0</span>      terra_1<span class="fl">.7</span><span class="dv">-39</span>       codetools_0<span class="fl">.2</span><span class="dv">-19</span></span>
<span id="cb19-35"><a href="#cb19-35" tabindex="-1"></a>[<span class="dv">31</span>] httpuv_1.<span class="fl">6.11</span>      htmltools_0.<span class="fl">5.6</span>    usethis_2.<span class="fl">2.2</span></span>
<span id="cb19-36"><a href="#cb19-36" tabindex="-1"></a>[<span class="dv">34</span>] yaml_2.<span class="fl">3.7</span>         later_1.<span class="fl">3.1</span>        pillar_1.<span class="fl">9.0</span></span>
<span id="cb19-37"><a href="#cb19-37" tabindex="-1"></a>[<span class="dv">37</span>] crayon_1.<span class="fl">5.2</span>       urlchecker_1.<span class="fl">0.1</span>   ellipsis_0.<span class="fl">3.2</span></span>
<span id="cb19-38"><a href="#cb19-38" tabindex="-1"></a>[<span class="dv">40</span>] cachem_1.<span class="fl">0.8</span>       magick_2.<span class="fl">7.5</span>       sessioninfo_1.<span class="fl">2.2</span></span>
<span id="cb19-39"><a href="#cb19-39" tabindex="-1"></a>[<span class="dv">43</span>] mime_0<span class="fl">.12</span>          tidyselect_1.<span class="fl">2.0</span>   digest_0.<span class="fl">6.33</span></span>
<span id="cb19-40"><a href="#cb19-40" tabindex="-1"></a>[<span class="dv">46</span>] stringi_1.<span class="fl">7.12</span>     dplyr_1.<span class="fl">1.2</span>        purrr_1.<span class="fl">0.2</span></span>
<span id="cb19-41"><a href="#cb19-41" tabindex="-1"></a>[<span class="dv">49</span>] labeling_0.<span class="fl">4.2</span>     cowplot_1.<span class="fl">1.1</span>      fastmap_1.<span class="fl">1.1</span></span>
<span id="cb19-42"><a href="#cb19-42" tabindex="-1"></a>[<span class="dv">52</span>] grid_4.<span class="fl">3.1</span>         colorspace_2<span class="fl">.1</span><span class="dv">-0</span>   cli_3.<span class="fl">6.1</span></span>
<span id="cb19-43"><a href="#cb19-43" tabindex="-1"></a>[<span class="dv">55</span>] magrittr_2.<span class="fl">0.3</span>     pkgbuild_1.<span class="fl">4.2</span>     utf8_1.<span class="fl">2.3</span></span>
<span id="cb19-44"><a href="#cb19-44" tabindex="-1"></a>[<span class="dv">58</span>] withr_2.<span class="fl">5.0</span>        prettyunits_1.<span class="fl">1.1</span>  scales_1.<span class="fl">2.1</span></span>
<span id="cb19-45"><a href="#cb19-45" tabindex="-1"></a>[<span class="dv">61</span>] promises_1.<span class="fl">2.1</span>     backports_1.<span class="fl">4.1</span>    rmarkdown_2<span class="fl">.24</span></span>
<span id="cb19-46"><a href="#cb19-46" tabindex="-1"></a>[<span class="dv">64</span>] igraph_1.<span class="fl">5.1</span>       reticulate_1<span class="fl">.31</span>    ragg_1.<span class="fl">2.5</span></span>
<span id="cb19-47"><a href="#cb19-47" tabindex="-1"></a>[<span class="dv">67</span>] png_0<span class="fl">.1</span><span class="dv">-8</span>          memoise_2.<span class="fl">0.1</span>      shiny_1.<span class="fl">7.5</span></span>
<span id="cb19-48"><a href="#cb19-48" tabindex="-1"></a>[<span class="dv">70</span>] evaluate_0<span class="fl">.21</span>      knitr_1<span class="fl">.43</span>         miniUI_0.<span class="dv">1</span>.<span class="fl">1.1</span></span>
<span id="cb19-49"><a href="#cb19-49" tabindex="-1"></a>[<span class="dv">73</span>] profvis_0.<span class="fl">3.8</span>      rlang_1.<span class="fl">1.1</span>        Rcpp_1.<span class="fl">0.11</span></span>
<span id="cb19-50"><a href="#cb19-50" tabindex="-1"></a>[<span class="dv">76</span>] xtable_1<span class="fl">.8</span><span class="dv">-4</span>       glue_1.<span class="fl">6.2</span>         pkgload_1.<span class="dv">3</span>.<span class="fl">2.1</span></span>
<span id="cb19-51"><a href="#cb19-51" tabindex="-1"></a>[<span class="dv">79</span>] jsonlite_1.<span class="fl">8.7</span>     rstudioapi_0.<span class="fl">15.0</span>  R6_2.<span class="fl">5.1</span></span>
<span id="cb19-52"><a href="#cb19-52" tabindex="-1"></a>[<span class="dv">82</span>] systemfonts_1.<span class="fl">0.4</span>  fs_1.<span class="fl">6.3</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by <a href="https://www.drieslab.com" class="external-link">Ruben Dries</a>, Jiaji George Chen, Natalie Del Rossi, <a href="https://labs.icahn.mssm.edu/yuanlab/" class="external-link">Guo-Cheng Yuan</a>, Matthew O’Brien, <a href="https://josschavezf.github.io" class="external-link">Joselyn C. Chávez-Fuentes</a>, Edward Ruiz.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
