<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="Previously: binGetSpatialGenes. BinSpect (Binary Spatial Extraction of genes) is a fast computational method
that identifies genes with a spatially coherent expression pattern."><title>binSpect — binSpect • Giotto</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><link href="../deps/JetBrains_Mono-0.4.7/font.css" rel="stylesheet"><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="binSpect — binSpect"><meta property="og:description" content="Previously: binGetSpatialGenes. BinSpect (Binary Spatial Extraction of genes) is a fast computational method
that identifies genes with a spatially coherent expression pattern."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-dark"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">Giotto</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">3.3.2</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Documentation</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-tutorials">Tutorials</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-tutorials">
    <h6 class="dropdown-header" data-toc-skip>Seq-Scope</h6>
    <a class="dropdown-item" href="../articles/seqscope_mouse_liver.html">Mouse liver</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Visium</h6>
    <a class="dropdown-item" href="../articles/visium_mouse_kidney.html">Visium Mouse Kidney</a>
    <a class="dropdown-item" href="../articles/visium_mouse_brain.html">Visium Mouse Brain</a>
    <a class="dropdown-item" href="../articles/visium_prostate_integration.html">Visium Prostate Integration</a>
    <a class="dropdown-item" href="../articles/visium_cytassist_lungcancer.html">CytAssist Human Lung Cancer</a>
    <a class="dropdown-item" href="../articles/visium_cytassist_human_glioblastoma.html">CytAssist Multi-omics Human Glioblastoma</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Spatial Genomics</h6>
    <a class="dropdown-item" href="../articles/spatial_genomics_mouse_kidney.html">Mouse Kidney</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Xenium</h6>
    <a class="dropdown-item" href="../articles/xenium_breast_cancer.html">Xenium Human breast cancer</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Nanostring CosMx</h6>
    <a class="dropdown-item" href="../articles/nanostring_cosmx_lung_cancer.html">Nanostring CosMx Subcellular Lung Cancer</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>MERFISH</h6>
    <a class="dropdown-item" href="../articles/merfish_mouse_hypothalamic.html">MERFISH Mouse Hypothalamic Preoptic Region</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Vizgen</h6>
    <a class="dropdown-item" href="../articles/vizgen_mouse_brain.html">Vizgen Mouse Brain Receptor Map</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>seqFISH</h6>
    <a class="dropdown-item" href="../articles/mini_seqfish.html">mini seqFISH</a>
    <a class="dropdown-item" href="../articles/seqfish_cortex.html">SeqFish+ Mouse Cortex Example</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>osmFISH</h6>
    <a class="dropdown-item" href="../articles/osmfish_mouse_ss_cortex.html">osmFISH Mouse SS Cortex</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Resolve Bioscience</h6>
    <a class="dropdown-item" href="../articles/resolve_biosciences.html">Resolve Bioscience Breast Cancer Subcellular</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>CODEX</h6>
    <a class="dropdown-item" href="../articles/codex_mouse_spleen.html">Codex Mouse Spleen</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Single Cell RNA Sequencing</h6>
    <a class="dropdown-item" href="../articles/singlecell_prostate_standard.html">Prostate</a>
    <a class="dropdown-item" href="../articles/singlecell_prostate_integration.html">Prostate Integration</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">News</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/drieslab/Giotto/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>binSpect</h1>
      <small class="dont-index">Source: <a href="https://github.com/drieslab/Giotto/blob/HEAD/R/spatial_genes.R" class="external-link"><code>R/spatial_genes.R</code></a></small>
      <div class="d-none name"><code>binSpect.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Previously: binGetSpatialGenes. BinSpect (Binary Spatial Extraction of genes) is a fast computational method
that identifies genes with a spatially coherent expression pattern.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">binSpect</span><span class="op">(</span></span>
<span>  <span class="va">gobject</span>,</span>
<span>  spat_unit <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  feat_type <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  bin_method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"kmeans"</span>, <span class="st">"rank"</span><span class="op">)</span>,</span>
<span>  expression_values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"normalized"</span>, <span class="st">"scaled"</span>, <span class="st">"custom"</span><span class="op">)</span>,</span>
<span>  subset_feats <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  subset_genes <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  spatial_network_name <span class="op">=</span> <span class="st">"Delaunay_network"</span>,</span>
<span>  spatial_network_k <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  reduce_network <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  kmeans_algo <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"kmeans"</span>, <span class="st">"kmeans_arma"</span>, <span class="st">"kmeans_arma_subset"</span><span class="op">)</span>,</span>
<span>  nstart <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  iter_max <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  extreme_nr <span class="op">=</span> <span class="fl">50</span>,</span>
<span>  sample_nr <span class="op">=</span> <span class="fl">50</span>,</span>
<span>  percentage_rank <span class="op">=</span> <span class="fl">30</span>,</span>
<span>  do_fisher_test <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  adjust_method <span class="op">=</span> <span class="st">"fdr"</span>,</span>
<span>  calc_hub <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  hub_min_int <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  get_av_expr <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  get_high_expr <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  implementation <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"data.table"</span>, <span class="st">"simple"</span>, <span class="st">"matrix"</span><span class="op">)</span>,</span>
<span>  group_size <span class="op">=</span> <span class="st">"automatic"</span>,</span>
<span>  do_parallel <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  cores <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">T</span>,</span>
<span>  knn_params <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  set.seed <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  bin_matrix <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  summarize <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"p.value"</span>, <span class="st">"adj.p.value"</span><span class="op">)</span>,</span>
<span>  return_gobject <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>gobject</dt>
<dd><p>giotto object</p></dd>


<dt>spat_unit</dt>
<dd><p>spatial unit</p></dd>


<dt>feat_type</dt>
<dd><p>feature type</p></dd>


<dt>bin_method</dt>
<dd><p>method to binarize gene expression</p></dd>


<dt>expression_values</dt>
<dd><p>expression values to use</p></dd>


<dt>subset_feats</dt>
<dd><p>only select a subset of features to test</p></dd>


<dt>subset_genes</dt>
<dd><p>deprecated, use subset_feats</p></dd>


<dt>spatial_network_name</dt>
<dd><p>name of spatial network to use (default = 'spatial_network')</p></dd>


<dt>spatial_network_k</dt>
<dd><p>different k's for a spatial kNN to evaluate</p></dd>


<dt>reduce_network</dt>
<dd><p>default uses the full network</p></dd>


<dt>kmeans_algo</dt>
<dd><p>kmeans algorithm to use (kmeans, kmeans_arma, kmeans_arma_subset)</p></dd>


<dt>nstart</dt>
<dd><p>kmeans: nstart parameter</p></dd>


<dt>iter_max</dt>
<dd><p>kmeans: iter.max parameter</p></dd>


<dt>extreme_nr</dt>
<dd><p>number of top and bottom cells (see details)</p></dd>


<dt>sample_nr</dt>
<dd><p>total number of cells to sample (see details)</p></dd>


<dt>percentage_rank</dt>
<dd><p>percentage of top cells for binarization</p></dd>


<dt>do_fisher_test</dt>
<dd><p>perform fisher test</p></dd>


<dt>adjust_method</dt>
<dd><p>p-value adjusted method to use (see <code><a href="https://rdrr.io/r/stats/p.adjust.html" class="external-link">p.adjust</a></code>)</p></dd>


<dt>calc_hub</dt>
<dd><p>calculate the number of hub cells</p></dd>


<dt>hub_min_int</dt>
<dd><p>minimum number of cell-cell interactions for a hub cell</p></dd>


<dt>get_av_expr</dt>
<dd><p>calculate the average expression per gene of the high expressing cells</p></dd>


<dt>get_high_expr</dt>
<dd><p>calculate the number of high expressing cells  per gene</p></dd>


<dt>implementation</dt>
<dd><p>enrichment implementation (data.table, simple, matrix)</p></dd>


<dt>group_size</dt>
<dd><p>number of genes to process together with data.table implementation (default = automatic)</p></dd>


<dt>do_parallel</dt>
<dd><p>run calculations in parallel with mclapply</p></dd>


<dt>cores</dt>
<dd><p>number of cores to use if do_parallel = TRUE</p></dd>


<dt>verbose</dt>
<dd><p>be verbose</p></dd>


<dt>knn_params</dt>
<dd><p>list of parameters to create spatial kNN network</p></dd>


<dt>set.seed</dt>
<dd><p>set a seed before kmeans binarization</p></dd>


<dt>bin_matrix</dt>
<dd><p>a binarized matrix, when provided it will skip the binarization process</p></dd>


<dt>summarize</dt>
<dd><p>summarize the p-values or adjusted p-values</p></dd>


<dt>return_gobject</dt>
<dd><p>whether to return values attached to the gobject or
separately (default)</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>data.table with results (see details)</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>We provide two ways to identify spatial genes based on gene expression binarization.
Both methods are identicial except for how binarization is performed.</p><ul><li><p>1. binarize: Each gene is binarized (0 or 1) in each cell with <b>kmeans</b> (k = 2) or based on <b>rank</b> percentile</p></li>
<li><p>2. network: Alll cells are connected through a spatial network based on the physical coordinates</p></li>
<li><p>3. contingency table: A contingency table is calculated based on all edges of neighboring cells and the binarized expression (0-0, 0-1, 1-0 or 1-1)</p></li>
<li><p>4. For each gene an odds-ratio (OR) and fisher.test (optional) is calculated</p></li>
</ul><p>Three different kmeans algorithmes have been implemented:</p><ul><li><p>1. kmeans: default, see <code><a href="https://rdrr.io/r/stats/kmeans.html" class="external-link">kmeans</a></code></p></li>
<li><p>2. kmeans_arma: from ClusterR, see <code>KMeans_arma</code></p></li>
<li><p>3. kmeans_arma_subst: from ClusterR, see <code>KMeans_arma</code>,
   but random subsetting the vector for each gene to increase speed. Change extreme_nr and sample_nr for control.</p></li>
</ul><p>Other statistics are provided (optional):</p><ul><li><p>Number of cells with high expression (binary = 1)</p></li>
<li><p>Average expression of each gene within high expressing cells</p></li>
<li><p>Number of hub cells, these are high expressing cells that have a user defined number of high expressing neighbors</p></li>
</ul><p>By selecting a subset of likely spatial genes (e.g. soft thresholding highly variable genes) can accelerate the speed.
The simple implementation is usually faster, but lacks the possibility to run in parallel and to calculate hub cells.
The data.table implementation might be more appropriate for large datasets by setting the group_size (number of genes) parameter to divide the workload.</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Ruben Dries, Jiaji George Chen, Natalie Del Rossi, Guo-Cheng Yuan, Matthew O’Brien, <a href="https://josschavezf.github.io" class="external-link">Joselyn Chávez</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

