<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="Giotto">
<title>Seq-Scope Mouse liver â€¢ Giotto</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/JetBrains_Mono-0.4.7/font.css" rel="stylesheet">
<!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Seq-Scope Mouse liver">
<meta property="og:description" content="Giotto">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">Giotto</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">4.0.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-get-started">Get started</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-get-started">
    <a class="dropdown-item" href="../articles/installation.html">Giotto Installation</a>
    <a class="dropdown-item" href="../articles/ecosystem.html">Giotto Ecosystem</a>
    <a class="dropdown-item" href="../articles/configuration.html">Giotto Configuration</a>
    <a class="dropdown-item" href="../articles/object_creation.html">Giotto Object Creation</a>
    <a class="dropdown-item" href="../articles/structure.html">Giotto Class Structure</a>
    <a class="dropdown-item" href="../articles/branches.html">Branches</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Documentation</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-examples">Examples</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-examples">
    <h6 class="dropdown-header" data-toc-skip>Seq-Scope</h6>
    <a class="dropdown-item" href="../articles/seqscope_mouse_liver.html">Mouse liver</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Visium</h6>
    <a class="dropdown-item" href="../articles/visium_mouse_kidney.html">Visium Mouse Kidney</a>
    <a class="dropdown-item" href="../articles/visium_mouse_brain.html">Visium Mouse Brain</a>
    <a class="dropdown-item" href="../articles/visium_prostate_integration.html">Visium Prostate Integration</a>
    <a class="dropdown-item" href="../articles/visium_cytassist_lungcancer.html">CytAssist Human Lung Cancer</a>
    <a class="dropdown-item" href="../articles/visium_cytassist_human_glioblastoma.html">CytAssist Multi-omics Human Glioblastoma</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Spatial Genomics</h6>
    <a class="dropdown-item" href="../articles/spatial_genomics_mouse_kidney.html">Mouse Kidney</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Xenium</h6>
    <a class="dropdown-item" href="../articles/xenium_breast_cancer.html">Xenium Human breast cancer</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Nanostring CosMx</h6>
    <a class="dropdown-item" href="../articles/nanostring_cosmx_lung_cancer.html">Nanostring CosMx Subcellular Lung Cancer</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>MERFISH</h6>
    <a class="dropdown-item" href="../articles/merfish_mouse_hypothalamic.html">MERFISH Mouse Hypothalamic Preoptic Region</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Vizgen</h6>
    <a class="dropdown-item" href="../articles/vizgen_mouse_brain.html">Vizgen Mouse Brain Receptor Map</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>seqFISH</h6>
    <a class="dropdown-item" href="../articles/mini_seqfish.html">mini seqFISH</a>
    <a class="dropdown-item" href="../articles/seqfish_cortex.html">SeqFish+ Mouse Cortex Example</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>osmFISH</h6>
    <a class="dropdown-item" href="../articles/osmfish_mouse_ss_cortex.html">osmFISH Mouse SS Cortex</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Resolve Bioscience</h6>
    <a class="dropdown-item" href="../articles/resolve_biosciences.html">Resolve Bioscience Breast Cancer Subcellular</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>CODEX</h6>
    <a class="dropdown-item" href="../articles/codex_mouse_spleen.html">Codex Mouse Spleen</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Single Cell RNA Sequencing</h6>
    <a class="dropdown-item" href="../articles/singlecell_prostate_standard.html">Prostate</a>
    <a class="dropdown-item" href="../articles/singlecell_prostate_integration.html">Prostate Integration</a>
  </div>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-tutorials">Tutorials</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-tutorials">
    <h6 class="dropdown-header" data-toc-skip>Spatial manipulation</h6>
    <a class="dropdown-item" href="../articles/spatial_transformations.html">Spatial transformations</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Image alignment</h6>
    <a class="dropdown-item" href="../articles/image_alignment.html">Image alignment</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Interactive selection</h6>
    <a class="dropdown-item" href="../articles/interactive_selection.html">Interactive selection</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Visualizations</h6>
    <a class="dropdown-item" href="../articles/visualizations.html">Visualizations</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Saving and Loading</h6>
    <a class="dropdown-item" href="../articles/saving_object.html">Saving a Giotto object</a>
    <a class="dropdown-item" href="../articles/saving_plots.html">Saving plots</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Interoperability</h6>
    <a class="dropdown-item" href="../articles/interoperability_anndata.html">AnnData</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Try Giotto in the Cloud</h6>
    <a class="dropdown-item" href="../articles/terra.html">Terra.bio</a>
    <a class="dropdown-item" href="../articles/docker.html">Docker</a>
  </div>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-news">News</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-news">
    <a class="dropdown-item" href="../news/index.html">Changes</a>
    <a class="dropdown-item" href="../articles/presentations.html">Presentations</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/drieslab/Giotto/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Seq-Scope Mouse liver</h1>
            
            <h4 data-toc-skip class="date">2022-11-15</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/drieslab/Giotto/blob/HEAD/vignettes/seqscope_mouse_liver.rmd" class="external-link"><code>vignettes/seqscope_mouse_liver.rmd</code></a></small>
      <div class="d-none name"><code>seqscope_mouse_liver.rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="dataset-explanation">Dataset explanation<a class="anchor" aria-label="anchor" href="#dataset-explanation"></a>
</h2>
<p><a href="https://pubmed.ncbi.nlm.nih.gov/34115981/" class="external-link">Seqscope</a> is a
illumina sequencing based spatial sequencing platform developed by <a href="https://lee.lab.medicine.umich.edu/seq-scope" class="external-link">Jun Hee Lee Lab</a>.
The basic strategy is to use illumina sequencing by synthesis to
generate the spatial barcodes and use the barcodes to capture mRNAs in
tissue.</p>
<div class="section level3">
<h3 id="example-raw-data-needed-for-seqscope">Example Raw Data needed for seqscope:<a class="anchor" aria-label="anchor" href="#example-raw-data-needed-for-seqscope"></a>
</h3>
<p>1st-seq data (single-ended, for generating spatial barcodes)</p>
<ul>
<li>SeqScope_1st.fastq.gz 2nd-seq data (pair-ended, for generating count
matrix)</li>
<li>SeqScope_2nd_R1.fastq.gz</li>
<li>SeqScope_2nd_R2.fastq.gz</li>
</ul>
<p>Image (for seqmentation)</p>
<ul>
<li>Tile_No_Segmented.png</li>
</ul>
</div>
<div class="section level3">
<h3 id="preprocessing-to-generate-count-matrix">Preprocessing to generate count matrix<a class="anchor" aria-label="anchor" href="#preprocessing-to-generate-count-matrix"></a>
</h3>
<p>Seqscope has its own pipeline to generate the count matrix(Gene
barcode). Please refer to their <a href="https://github.com/leeju-umich/Cho_Xi_Seqscope" class="external-link">github page</a> or
use customized methods to proceed.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="co"># Note you need to install seqtk, STARsolo and clone their github page first. All bash scripts are stored in script directory of their github.</span></span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a></span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a><span class="do">## First generate the whitelist of spatial barcodes:</span></span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a>bash extractCoord.sh [SeqScope_1st.fastq.gz] [SeqScope_2nd_R1.fastq.gz] [HDMI_length]</span>
<span id="cb1-5"><a href="#cb1-5" tabindex="-1"></a><span class="do">## Second, generating the count matrix via STAR.Solo:</span></span>
<span id="cb1-6"><a href="#cb1-6" tabindex="-1"></a>bash align.sh [SeqScope_2nd_R1.fastq.gz] [abc_SeqScope_2nd_R2.fastq.gz] [HDMI_length] [whitelists.txt] [outprefix] [starpath] [seqtkpath] [geneIndex]</span></code></pre></div>
<p>You should now have the count matrix for Giotto Object. Note the
Spatial Barcodes are typically at 0.6 um resolution. One way to analyze
is to use the getSimpleGrid.R or getSlidingGrid.R provided by Lee Lab to
bin the counts and simply follow the Giotto analysis steps of <a href="https://giottosuite.readthedocs.io/en/latest/subsections/datasets/ouse_visium_brain.html" class="external-link">Visium</a>
and <a href="https://giottosuite.readthedocs.io/en/latest/subsections/datasets/visium_prostate_integration.html" class="external-link">Integration
of Visium</a></p>
</div>
</div>
<div class="section level2">
<h2 id="start-giotto">Start Giotto<a class="anchor" aria-label="anchor" href="#start-giotto"></a>
</h2>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Ensure Giotto Suite is installed.</span></span>
<span><span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="st">"Giotto"</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/installed.packages.html" class="external-link">installed.packages</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"drieslab/Giotto@suite"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://drieslab.github.io/Giotto/">Giotto</a></span><span class="op">)</span></span>
<span><span class="co"># Ensure the Python environment for Giotto has been installed.</span></span>
<span><span class="va">genv_exists</span> <span class="op">=</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/checkGiottoEnvironment.html">checkGiottoEnvironment</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="va">genv_exists</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="co"># The following command need only be run once to install the Giotto environment.</span></span>
<span>  <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/installGiottoEnvironment.html">installGiottoEnvironment</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="set-up-giotto-environment">Set up Giotto Environment<a class="anchor" aria-label="anchor" href="#set-up-giotto-environment"></a>
</h2>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># 1. set working directory</span></span>
<span><span class="va">results_folder</span> <span class="op">=</span> <span class="st">'path/to/result'</span></span>
<span></span>
<span><span class="co"># 2. set giotto python path</span></span>
<span><span class="co"># set python path to your preferred python version path</span></span>
<span><span class="co"># set python path to conda env/bin/ directory if manually installed Giotto python dependencies by conda</span></span>
<span><span class="co"># python_path = '/path_to_conda/.conda/envs/giotto/bin/python'</span></span>
<span><span class="co"># set python path to NULL if you want to automatically install (only the 1st time) and use the giotto miniconda environment</span></span>
<span><span class="co"># python_path = NULL</span></span>
<span><span class="co">#if(is.null(python_path)) {</span></span>
<span><span class="co">#  installGiottoEnvironment()</span></span>
<span><span class="co">#}</span></span>
<span></span>
<span><span class="co"># 3. create giotto instructions</span></span>
<span><span class="va">instrs</span> <span class="op">=</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/createGiottoInstructions.html">createGiottoInstructions</a></span><span class="op">(</span>save_dir <span class="op">=</span> <span class="va">results_folder</span>,</span>
<span>                                  save_plot <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                                  show_plot <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="create-giotto-subcellular-object">1. Create Giotto Subcellular Object<a class="anchor" aria-label="anchor" href="#create-giotto-subcellular-object"></a>
</h2>
<p>In order to do single cell and subcellular analysis with Giotto, we
suggest using the GiottoSubcellular Object. Giotto Subcellular objects
take cell polygons and a Giotto Point file (a data.table contains Gene
Name, Expression, x, y, etc). Therefore, first we need to do some
transformation of the count matrix.</p>
<p><strong>Giotto Point file formatting</strong></p>
<p>Unlike normal cell by gene matrix, a Giotto Point file is usually a
data.table contains Gene Name, Expression, x, y, where each row
represent a subcellular point. For Seqscope data, one HDMI typically
have more than one subcellular point.</p>
<p>Example Giotto Point file for <strong>Seqscope</strong></p>
<table class="table">
<thead><tr class="header">
<th align="left">HDMI</th>
<th align="left">Feat_ID</th>
<th align="left">Count</th>
<th align="left">sdimX</th>
<th align="left">sdimY</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">HDMI1</td>
<td align="left">GeneA</td>
<td align="left"></td>
<td align="left"></td>
<td align="left"></td>
</tr>
<tr class="even">
<td align="left">HDMI1</td>
<td align="left">GeneB</td>
<td align="left"></td>
<td align="left"></td>
<td align="left"></td>
</tr>
<tr class="odd">
<td align="left">HDMI2</td>
<td align="left">GeneA</td>
<td align="left"></td>
<td align="left"></td>
<td align="left"></td>
</tr>
<tr class="even">
<td align="left">HDMI3</td>
<td align="left">GeneC</td>
<td align="left"></td>
<td align="left"></td>
<td align="left"></td>
</tr>
</tbody>
</table>
<div class="section level3">
<h3 id="process-giotto-point-file-per-tile">1.1 Process Giotto Point file per tile<a class="anchor" aria-label="anchor" href="#process-giotto-point-file-per-tile"></a>
</h3>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">##expression matrix</span></span>
<span><span class="va">countDir</span> <span class="op">=</span> <span class="st">"/path/to/Solo.out/GeneFull/raw"</span></span>
<span><span class="va">expr_matrix</span> <span class="op">=</span> <span class="fu">Giotto</span><span class="fu">::</span><span class="fu"><a href="../reference/get10Xmatrix.html">get10Xmatrix</a></span><span class="op">(</span>path_to_data <span class="op">=</span> <span class="va">countDir</span>, gene_column_index <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span></span>
<span><span class="co">##Spatial coordinates</span></span>
<span><span class="va">spatial_coords_Dir</span> <span class="op">=</span> <span class="st">"/path/to/extractCoord.sh/results/spatialcoordinates.txt"</span></span>
<span><span class="va">spatial_coords</span> <span class="op">=</span> <span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="st">"spatial_coords_Dir"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/row-plus-colnames-generic.html">colnames</a></span><span class="op">(</span><span class="va">spatial_coords</span><span class="op">)</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"HDMI"</span>,<span class="st">"Lane"</span>,<span class="st">"Tile"</span>,<span class="st">"X"</span>,<span class="st">"Y"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">##Prepare Giotto Point</span></span>
<span><span class="co"># Subset expression and spatial info by tile</span></span>
<span><span class="va">spatial_coords_tile</span> <span class="op">=</span> <span class="va">spatial_coords</span><span class="op">[</span><span class="va">Tile</span> <span class="op">==</span> <span class="st">'2104'</span><span class="op">]</span></span>
<span><span class="va">expr_matrix_tile</span> <span class="op">=</span> <span class="va">expr_matrix</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/row-plus-colnames-generic.html">colnames</a></span><span class="op">(</span><span class="va">expr_matrix</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">spatial_coords_tile</span><span class="op">$</span><span class="va">HDMI</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># convert expression matrix to minimal data.table object</span></span>
<span><span class="va">matrix_tile_dt</span> <span class="op">=</span> <span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/as.data.table.html" class="external-link">as.data.table</a></span><span class="op">(</span><span class="fu">Matrix</span><span class="fu">::</span><span class="fu">summary</span><span class="op">(</span><span class="va">expr_matrix_tile</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">genes</span> <span class="op">=</span> <span class="va">expr_matrix_tile</span><span class="op">@</span><span class="va">Dimnames</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">samples</span> <span class="op">=</span> <span class="va">expr_matrix_tile</span><span class="op">@</span><span class="va">Dimnames</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">matrix_tile_dt</span><span class="op">[</span>, <span class="va">gene</span> <span class="op">:=</span> <span class="va">genes</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">matrix_tile_dt</span><span class="op">[</span>, <span class="va">hdmi</span> <span class="op">:=</span> <span class="va">samples</span><span class="op">[</span><span class="va">j</span><span class="op">]</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># merge data.table matrix and spatial coordinates to create input for Giotto Polygons</span></span>
<span><span class="va">gpoints</span> <span class="op">=</span> <span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/merge.html" class="external-link">merge.data.table</a></span><span class="op">(</span><span class="va">matrix_tile_dt</span>, <span class="va">spatial_coords_tile</span>, by.x <span class="op">=</span> <span class="st">'hdmi'</span>, by.y <span class="op">=</span> <span class="st">'HDMI'</span><span class="op">)</span></span>
<span><span class="va">gpoints</span> <span class="op">=</span> <span class="va">gpoints</span><span class="op">[</span>,<span class="fu">.</span><span class="op">(</span><span class="va">hdmi</span>, <span class="va">X</span>, <span class="va">Y</span>, <span class="va">gene</span>, <span class="va">x</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/row-plus-colnames-generic.html">colnames</a></span><span class="op">(</span><span class="va">gpoints</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'hdmi'</span>, <span class="st">'x'</span>, <span class="st">'y'</span>, <span class="st">'gene'</span>, <span class="st">'counts'</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="prepare-the-polygon-mask-file">1.2 Prepare the polygon mask file<a class="anchor" aria-label="anchor" href="#prepare-the-polygon-mask-file"></a>
</h3>
<p>Giotto can read in a variety of different mask files provided by
common segmentation tools. But first we need to check if we need to filp
the x and y axis.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># check total counts per hdmi</span></span>
<span><span class="va">gpoints_aggr</span> <span class="op">=</span> <span class="va">gpoints</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span>, by <span class="op">=</span> <span class="fu">.</span><span class="op">(</span><span class="va">hdmi</span>, <span class="va">x</span>, <span class="va">y</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/row-plus-colnames-generic.html">colnames</a></span><span class="op">(</span><span class="va">gpoints_aggr</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"hdmi"</span>,<span class="st">"x"</span>,<span class="st">"y"</span>,<span class="st">"total_counts"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/setorder.html" class="external-link">setorder</a></span><span class="op">(</span><span class="va">gpoints_aggr</span>, <span class="op">-</span><span class="va">total_counts</span><span class="op">)</span></span>
<span></span>
<span><span class="va">pl</span> <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">pl</span> <span class="op">=</span> <span class="va">pl</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">gpoints_aggr</span><span class="op">[</span><span class="va">total_counts</span> <span class="op">&lt;</span> <span class="fl">1000</span> <span class="op">&amp;</span> <span class="va">total_counts</span> <span class="op">&gt;</span> <span class="fl">8</span><span class="op">]</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, y <span class="op">=</span> <span class="va">y</span>, color <span class="op">=</span> <span class="va">total_counts</span><span class="op">)</span>, size <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span></span>
<span><span class="va">pl</span> <span class="op">=</span> <span class="va">pl</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html" class="external-link">scale_color_gradient2</a></span><span class="op">(</span>midpoint <span class="op">=</span> <span class="fl">15</span>, low <span class="op">=</span> <span class="st">'blue'</span>, mid <span class="op">=</span> <span class="st">'yellow'</span>, high <span class="op">=</span> <span class="st">'red'</span><span class="op">)</span></span>
<span><span class="va">pl</span></span></code></pre></div>
<p><img src="images/seqscope_mouse_liver/HDMI_scatter.png"></p>
<p>Read polygon mask file</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">segmentation_mask</span> <span class="op">=</span> <span class="st">"/path/to/segmentation.tif"</span></span>
<span><span class="va">final_polygons</span> <span class="op">=</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/createGiottoPolygon.html">createGiottoPolygonsFromMask</a></span><span class="op">(</span><span class="va">segmentation_mask</span>,</span>
<span>                               flip_vertical <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>                               flip_horizontal <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">final_polygons</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/seqscope_mouse_liver/polygon.png"></p>
<p><strong>Polygon mask file manual formatting</strong></p>
<p>However, sometimes Giotto does not read in the provided mask file and
we will need a manual processing for the mask file and we will do that
using terra and createGiottoPolygonsFromDfr.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">raster</span> <span class="op">=</span> <span class="fu">terra</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/terra/man/rast.html" class="external-link">rast</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">segmentation_mask</span><span class="op">)</span></span>
<span><span class="va">terra_polygon</span> <span class="op">=</span> <span class="fu">terra</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/terra/man/as.polygons.html" class="external-link">as.polygons</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">raster</span>, value <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># convert polygon to data.table and remove unwantedly detected polygons (e.g. canvas etc)</span></span>
<span><span class="va">dt</span> <span class="op">=</span> <span class="fu">GiottospatVector_to_dt</span><span class="op">(</span><span class="va">terra_polygon</span><span class="op">)</span></span>
<span><span class="va">npolygons</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/levels.html" class="external-link">levels</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">dt</span><span class="op">$</span><span class="va">part</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">-</span> <span class="fl">1</span></span>
<span><span class="va">filter_dt</span> <span class="op">=</span> <span class="va">dt</span><span class="op">[</span><span class="va">geom</span> <span class="op">==</span> <span class="fl">1</span> <span class="op">&amp;</span> <span class="va">hole</span> <span class="op">==</span> <span class="fl">0</span> <span class="op">&amp;</span> <span class="va">part</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span><span class="op">:</span><span class="va">npolygons</span><span class="op">)</span>, <span class="fu">.</span><span class="op">(</span><span class="va">x</span>,<span class="va">y</span>,<span class="va">part</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">filter_dt</span><span class="op">[</span>, <span class="va">part</span> <span class="op">:=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="va">part</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># create new polygons from filtered data.table</span></span>
<span><span class="va">final_polygons</span> <span class="op">=</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/createGiottoPolygon.html">createGiottoPolygonsFromDfr</a></span><span class="op">(</span>segmdfr <span class="op">=</span> <span class="va">filter_dt</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># create giotto points first to get the extent of the points (hdmi)</span></span>
<span><span class="va">original_points</span> <span class="op">=</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/createGiottoPoints.html">createGiottoPoints</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">gpoints</span><span class="op">[</span>,<span class="fu">.</span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span>, <span class="va">gene</span>, <span class="va">hdmi</span>, <span class="va">counts</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">original_feat_ext</span> <span class="op">=</span> <span class="fu">ext</span><span class="op">(</span><span class="va">original_points</span><span class="op">@</span><span class="va">spatVector</span><span class="op">)</span></span>
<span></span>
<span></span>
<span></span>
<span><span class="co"># convert polygon to spatRaster to change extent to that of original points</span></span>
<span><span class="va">final_spatraster</span> <span class="op">=</span> <span class="fu">Giottopolygon_to_raster</span><span class="op">(</span>polygon <span class="op">=</span> <span class="va">final_polygons</span><span class="op">@</span><span class="va">spatVector</span><span class="op">)</span></span>
<span><span class="fu">ext</span><span class="op">(</span><span class="va">final_spatraster</span><span class="op">$</span><span class="va">raster</span><span class="op">)</span> <span class="op">=</span> <span class="va">original_feat_ext</span></span>
<span><span class="va">final_polygons</span><span class="op">@</span><span class="va">spatVector</span> <span class="op">=</span> <span class="fu">as.polygons</span><span class="op">(</span><span class="va">final_spatraster</span><span class="op">$</span><span class="va">raster</span><span class="op">)</span></span>
<span><span class="va">final_polygons</span><span class="op">@</span><span class="va">spatVector</span><span class="op">$</span><span class="va">poly_ID</span> <span class="op">=</span> <span class="va">final_spatraster</span><span class="op">$</span><span class="va">ID_vector</span><span class="op">[</span><span class="va">final_polygons</span><span class="op">@</span><span class="va">spatVector</span><span class="op">$</span><span class="va">poly_i</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># flip and shift, if needed</span></span>
<span><span class="co">#final_polygons@spatVector = flip(final_polygons@spatVector)</span></span>
<span><span class="co">#yshift = ymin(original_feat_ext) - ymax(original_feat_ext)</span></span>
<span><span class="co">#final_polygons@spatVector = terra::shift(final_polygons@spatVector, dy = -yshift)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">final_polygons</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="create-giotto-object">1.3 Create Giotto Object<a class="anchor" aria-label="anchor" href="#create-giotto-object"></a>
</h3>
<p>Add a random jitter to the HDMI location to make a pseudo-in situ
transcript file.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># add giotto points class</span></span>
<span><span class="va">gpoints_subset</span> <span class="op">=</span> <span class="va">gpoints</span><span class="op">[</span><span class="va">hdmi</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">gpoints_aggr</span><span class="op">[</span><span class="va">total_counts</span> <span class="op">&gt;</span> <span class="fl">5</span><span class="op">]</span><span class="op">$</span><span class="va">hdmi</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># multiply rows with multiple counts and add jitter</span></span>
<span><span class="va">gpoints_extra</span> <span class="op">=</span> <span class="va">gpoints_subset</span><span class="op">[</span><span class="va">counts</span> <span class="op">&gt;</span> <span class="fl">1</span><span class="op">]</span></span>
<span><span class="va">gpoints_extra</span> <span class="op">=</span> <span class="va">gpoints_extra</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="va">counts</span>, <span class="va">counts</span><span class="op">)</span>, by <span class="op">=</span> <span class="fu">.</span><span class="op">(</span><span class="va">hdmi</span>, <span class="va">gene</span>, <span class="va">x</span>, <span class="va">y</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">gpoints_extra</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind</a></span><span class="op">(</span><span class="va">gpoints_extra</span><span class="op">[</span>,<span class="fu">.</span><span class="op">(</span><span class="va">hdmi</span>, <span class="va">gene</span>, <span class="va">x</span>, <span class="va">y</span><span class="op">)</span><span class="op">]</span>, <span class="va">gpoints_subset</span><span class="op">[</span><span class="va">counts</span> <span class="op">==</span> <span class="fl">1</span> ,<span class="fu">.</span><span class="op">(</span><span class="va">hdmi</span>, <span class="va">gene</span>, <span class="va">x</span>, <span class="va">y</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">jitter_x</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, size <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">gpoints_extra</span><span class="op">)</span>, replace <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span><span class="va">jitter_y</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, size <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">gpoints_extra</span><span class="op">)</span>, replace <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span><span class="va">gpoints_extra</span><span class="op">[</span>, <span class="va">x</span> <span class="op">:=</span> <span class="va">x</span> <span class="op">+</span> <span class="va">jitter_x</span><span class="op">]</span></span>
<span><span class="va">gpoints_extra</span><span class="op">[</span>, <span class="va">y</span> <span class="op">:=</span> <span class="va">y</span> <span class="op">+</span> <span class="va">jitter_y</span><span class="op">]</span></span></code></pre></div>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># add subcellular information</span></span>
<span><span class="va">seqscope</span> <span class="op">=</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/createGiottoObjectSubcellular.html">createGiottoObjectSubcellular</a></span><span class="op">(</span>gpoints <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">gpoints_extra</span><span class="op">[</span>,<span class="fu">.</span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span>, <span class="va">gene</span>, <span class="va">hdmi</span><span class="op">)</span><span class="op">]</span><span class="op">)</span>,</span>
<span>                                         gpolygons <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">final_polygons</span><span class="op">)</span>,</span>
<span>                                         instructions <span class="op">=</span> <span class="va">instrs</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># add centroids</span></span>
<span><span class="va">seqscope</span> <span class="op">=</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/addSpatialCentroidLocations.html">addSpatialCentroidLocations</a></span><span class="op">(</span><span class="va">seqscope</span>,</span>
<span>                                       poly_info <span class="op">=</span> <span class="st">'cell'</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#Overlap to Polygon information</span></span>
<span><span class="va">seqscope</span> <span class="op">=</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/calculateOverlapRaster.html">calculateOverlapRaster</a></span><span class="op">(</span><span class="va">seqscope</span><span class="op">)</span></span>
<span><span class="va">seqscope</span> <span class="op">=</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/overlapToMatrix.html">overlapToMatrix</a></span><span class="op">(</span><span class="va">seqscope</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Visualize top 200 expressed genes in situ</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/spatInSituPlotPoints.html">spatInSituPlotPoints</a></span><span class="op">(</span><span class="va">seqscope</span>, show_legend <span class="op">=</span> <span class="cn">F</span>,</span>
<span>                     show_image <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                     feats <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">'rna'</span> <span class="op">=</span> <span class="va">seqscope</span><span class="op">@</span><span class="va">feat_info</span><span class="op">$</span><span class="va">rna</span><span class="op">@</span><span class="va">spatVector</span><span class="op">$</span><span class="va">feat_ID</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">200</span><span class="op">]</span><span class="op">)</span>,</span>
<span>                     spat_unit <span class="op">=</span> <span class="st">'cell'</span>,</span>
<span>                     point_size <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                     show_polygon <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                     use_overlap <span class="op">=</span> <span class="cn">F</span>,</span>
<span>                     polygon_feat_type <span class="op">=</span> <span class="st">'cell'</span>,</span>
<span>                     polygon_color <span class="op">=</span> <span class="st">'red'</span>,</span>
<span>                     polygon_bg_color <span class="op">=</span> <span class="st">'white'</span>,</span>
<span>                     polygon_line_size <span class="op">=</span> <span class="fl">0.2</span>,</span>
<span>                     coord_fix_ratio <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                     background_color <span class="op">=</span> <span class="st">'white'</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/seqscope_mouse_liver/top200.png"></p>
</div>
</div>
<div class="section level2">
<h2 id="process-giotto-and-quality-control">2. Process Giotto and Quality Control<a class="anchor" aria-label="anchor" href="#process-giotto-and-quality-control"></a>
</h2>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># filter</span></span>
<span><span class="va">seqscope</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/filterGiotto.html">filterGiotto</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">seqscope</span>,</span>
<span>                         expression_threshold <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                         feat_det_in_min_cells <span class="op">=</span> <span class="fl">5</span>,</span>
<span>                         min_det_feats_per_cell <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#normalize</span></span>
<span><span class="va">seqscope</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/normalizeGiotto.html">normalizeGiotto</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">seqscope</span>, scalefactor <span class="op">=</span> <span class="fl">5000</span>, verbose <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span><span class="co"># add statistics</span></span>
<span><span class="va">seqscope</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/addStatistics.html">addStatistics</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">seqscope</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># View cellular data</span></span>
<span><span class="co"># pDataDT(seqscope)</span></span>
<span><span class="co"># View rna data</span></span>
<span><span class="co"># fDataDT(seqscope)</span></span>
<span></span>
<span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/spatPlot.html">spatPlot2D</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">seqscope</span>,</span>
<span>           cell_color <span class="op">=</span> <span class="st">'total_expr'</span>, color_as_factor <span class="op">=</span> <span class="cn">F</span>,</span>
<span>           show_image <span class="op">=</span> <span class="cn">F</span>,</span>
<span>           point_size <span class="op">=</span> <span class="fl">2.5</span>, point_alpha <span class="op">=</span> <span class="fl">0.75</span>, coord_fix_ratio <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/seqscope_mouse_liver/total_expr.png"></p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cellmeta</span> <span class="op">=</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/pDataDT.html">pDataDT</a></span><span class="op">(</span><span class="va">seqscope</span>, feat_type <span class="op">=</span> <span class="st">'rna'</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html" class="external-link">hist</a></span><span class="op">(</span><span class="va">cellmeta</span><span class="op">$</span><span class="va">nr_feats</span>, <span class="fl">100</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/seqscope_mouse_liver/feature_distribution.png"></p>
</div>
<div class="section level2">
<h2 id="dimention-reduction">3. Dimention Reduction<a class="anchor" aria-label="anchor" href="#dimention-reduction"></a>
</h2>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># cluster cells</span></span>
<span><span class="va">seqscope</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateHVF.html">calculateHVF</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">seqscope</span>, HVFname <span class="op">=</span> <span class="st">'hvg_orig'</span><span class="op">)</span></span>
<span></span>
<span><span class="va">seqscope</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runPCA.html">runPCA</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">seqscope</span>,</span>
<span>                   expression_values <span class="op">=</span> <span class="st">'normalized'</span>,</span>
<span>                   scale_unit <span class="op">=</span> <span class="cn">T</span>, center <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span></span>
<span><span class="va">seqscope</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runUMAP.html">runUMAP</a></span><span class="op">(</span><span class="va">seqscope</span>, dimensions_to_use <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">100</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="cluster">4. Cluster<a class="anchor" aria-label="anchor" href="#cluster"></a>
</h2>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seqscope</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/createNearestNetwork.html">createNearestNetwork</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">seqscope</span>, dimensions_to_use <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">100</span>, k <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span><span class="va">seqscope</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/doLeidenCluster.html">doLeidenCluster</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">seqscope</span>, resolution <span class="op">=</span> <span class="fl">0.9</span>, n_iterations <span class="op">=</span> <span class="fl">1000</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># visualize UMAP cluster results</span></span>
<span></span>
<span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/plotUMAP.html">plotUMAP</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">seqscope</span>, cell_color <span class="op">=</span> <span class="st">'leiden_clus'</span>,</span>
<span>         show_NN_network <span class="op">=</span> <span class="cn">F</span>, point_size <span class="op">=</span> <span class="fl">3.5</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/seqscope_mouse_liver/leiden_umap.png"></p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/spatInSituPlotPoints.html">spatInSituPlotPoints</a></span><span class="op">(</span><span class="va">seqscope</span>,</span>
<span>                     show_polygon <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                     polygon_color <span class="op">=</span> <span class="st">'white'</span>,</span>
<span>                     polygon_line_size <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>                     polygon_fill <span class="op">=</span> <span class="st">'leiden_clus'</span>,</span>
<span>                     polygon_fill_as_factor <span class="op">=</span> <span class="cn">T</span>,</span>
<span>                     coord_fix_ratio <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/seqscope_mouse_liver/leiden_spat.png"></p>
</div>
<div class="section level2">
<h2 id="find-spatial-genes">5. Find spatial genes<a class="anchor" aria-label="anchor" href="#find-spatial-genes"></a>
</h2>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seqscope</span><span class="op">&lt;-</span><span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/createSpatialNetwork.html">createSpatialNetwork</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">seqscope</span>, minimum_k <span class="op">=</span> <span class="fl">2</span>, maximum_distance_delaunay <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></span>
<span></span>
<span><span class="va">km_spatialgenes</span> <span class="op">=</span> <span class="fu"><a href="../reference/binSpect.html">binSpect</a></span><span class="op">(</span><span class="va">seqscope</span>, subset_feats <span class="op">=</span> <span class="va">seqscope</span><span class="op">@</span><span class="va">feat_ID</span><span class="op">$</span><span class="va">rna</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/spatFeatPlot2D.html">spatFeatPlot2D</a></span><span class="op">(</span><span class="va">seqscope</span>, expression_values <span class="op">=</span> <span class="st">'scaled'</span>,</span>
<span>               feats <span class="op">=</span> <span class="va">km_spatialgenes</span><span class="op">$</span><span class="va">feats</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span>,</span>
<span>               cell_color_gradient <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'blue'</span>, <span class="st">'white'</span>, <span class="st">'red'</span><span class="op">)</span>,</span>
<span>               point_shape <span class="op">=</span> <span class="st">'border'</span>, point_border_stroke <span class="op">=</span> <span class="fl">0.01</span>,</span>
<span>               show_network <span class="op">=</span> <span class="cn">T</span>, network_color <span class="op">=</span> <span class="st">'lightgrey'</span>, point_size <span class="op">=</span> <span class="fl">1.2</span>,</span>
<span>               cow_n_col <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/seqscope_mouse_liver/spatgenes.png"></p>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by <a href="https://www.drieslab.com" class="external-link">Ruben Dries</a>, Jiaji George Chen, Natalie Del Rossi, <a href="https://labs.icahn.mssm.edu/yuanlab/" class="external-link">Guo-Cheng Yuan</a>, Matthew Oâ€™Brien, <a href="https://josschavezf.github.io" class="external-link">Joselyn C. ChÃ¡vez-Fuentes</a>, Edward Ruiz.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
