<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="Giotto">
<title>Integration of single cell datasets • Giotto</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/JetBrains_Mono-0.4.7/font.css" rel="stylesheet">
<!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Integration of single cell datasets">
<meta property="og:description" content="Giotto">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">Giotto</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">4.0.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-get-started">Get started</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-get-started">
    <a class="dropdown-item" href="../articles/installation.html">Giotto Installation</a>
    <a class="dropdown-item" href="../articles/ecosystem.html">Giotto Ecosystem</a>
    <a class="dropdown-item" href="../articles/configuration.html">Giotto Configuration</a>
    <a class="dropdown-item" href="../articles/object_creation.html">Giotto Object Creation</a>
    <a class="dropdown-item" href="../articles/structure.html">Giotto Class Structure</a>
    <a class="dropdown-item" href="../articles/branches.html">Branches</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Documentation</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-examples">Examples</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-examples">
    <h6 class="dropdown-header" data-toc-skip>Seq-Scope</h6>
    <a class="dropdown-item" href="../articles/seqscope_mouse_liver.html">Mouse liver</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Visium</h6>
    <a class="dropdown-item" href="../articles/visium_mouse_kidney.html">Visium Mouse Kidney</a>
    <a class="dropdown-item" href="../articles/visium_mouse_brain.html">Visium Mouse Brain</a>
    <a class="dropdown-item" href="../articles/visium_prostate_integration.html">Visium Prostate Integration</a>
    <a class="dropdown-item" href="../articles/visium_cytassist_lungcancer.html">CytAssist Human Lung Cancer</a>
    <a class="dropdown-item" href="../articles/visium_cytassist_human_glioblastoma.html">CytAssist Multi-omics Human Glioblastoma</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Spatial Genomics</h6>
    <a class="dropdown-item" href="../articles/spatial_genomics_mouse_kidney.html">Mouse Kidney</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Xenium</h6>
    <a class="dropdown-item" href="../articles/xenium_breast_cancer.html">Xenium Human breast cancer</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Nanostring CosMx</h6>
    <a class="dropdown-item" href="../articles/nanostring_cosmx_lung_cancer.html">Nanostring CosMx Subcellular Lung Cancer</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>MERFISH</h6>
    <a class="dropdown-item" href="../articles/merfish_mouse_hypothalamic.html">MERFISH Mouse Hypothalamic Preoptic Region</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Vizgen</h6>
    <a class="dropdown-item" href="../articles/vizgen_mouse_brain.html">Vizgen Mouse Brain Receptor Map</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>seqFISH</h6>
    <a class="dropdown-item" href="../articles/mini_seqfish.html">mini seqFISH</a>
    <a class="dropdown-item" href="../articles/seqfish_cortex.html">SeqFish+ Mouse Cortex Example</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>osmFISH</h6>
    <a class="dropdown-item" href="../articles/osmfish_mouse_ss_cortex.html">osmFISH Mouse SS Cortex</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Resolve Bioscience</h6>
    <a class="dropdown-item" href="../articles/resolve_biosciences.html">Resolve Bioscience Breast Cancer Subcellular</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>CODEX</h6>
    <a class="dropdown-item" href="../articles/codex_mouse_spleen.html">Codex Mouse Spleen</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Single Cell RNA Sequencing</h6>
    <a class="dropdown-item" href="../articles/singlecell_prostate_standard.html">Prostate</a>
    <a class="dropdown-item" href="../articles/singlecell_prostate_integration.html">Prostate Integration</a>
  </div>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-tutorials">Tutorials</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-tutorials">
    <h6 class="dropdown-header" data-toc-skip>Spatial manipulation</h6>
    <a class="dropdown-item" href="../articles/spatial_transformations.html">Spatial transformations</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Image alignment</h6>
    <a class="dropdown-item" href="../articles/image_alignment.html">Image alignment</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Interactive selection</h6>
    <a class="dropdown-item" href="../articles/interactive_selection.html">Interactive selection</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Visualizations</h6>
    <a class="dropdown-item" href="../articles/visualizations.html">Visualizations</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Saving and Loading</h6>
    <a class="dropdown-item" href="../articles/saving_object.html">Saving a Giotto object</a>
    <a class="dropdown-item" href="../articles/saving_plots.html">Saving plots</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Interoperability</h6>
    <a class="dropdown-item" href="../articles/interoperability_anndata.html">AnnData</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Try Giotto in the Cloud</h6>
    <a class="dropdown-item" href="../articles/terra.html">Terra.bio</a>
    <a class="dropdown-item" href="../articles/docker.html">Docker</a>
  </div>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-news">News</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-news">
    <a class="dropdown-item" href="../news/index.html">Changes</a>
    <a class="dropdown-item" href="../articles/presentations.html">Presentations</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/drieslab/Giotto/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Integration of single cell datasets</h1>
            
            <h4 data-toc-skip class="date">2022-09-16</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/drieslab/Giotto/blob/HEAD/vignettes/singlecell_prostate_integration.Rmd" class="external-link"><code>vignettes/singlecell_prostate_integration.Rmd</code></a></small>
      <div class="d-none name"><code>singlecell_prostate_integration.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Ensure Giotto Suite is installed.</span></span>
<span><span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="st">"Giotto"</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/installed.packages.html" class="external-link">installed.packages</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"drieslab/Giotto@suite"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co"># Ensure GiottoData, a small, helper module for tutorials, is installed.</span></span>
<span><span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="st">"GiottoData"</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/installed.packages.html" class="external-link">installed.packages</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"drieslab/GiottoData"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://drieslab.github.io/Giotto/">Giotto</a></span><span class="op">)</span></span>
<span><span class="co"># Ensure the Python environment for Giotto has been installed.</span></span>
<span><span class="va">genv_exists</span> <span class="op">=</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/checkGiottoEnvironment.html">checkGiottoEnvironment</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="va">genv_exists</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="co"># The following command need only be run once to install the Giotto environment.</span></span>
<span>  <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/installGiottoEnvironment.html">installGiottoEnvironment</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<div class="section level2">
<h2 id="set-up-giotto-environment">Set up Giotto Environment<a class="anchor" aria-label="anchor" href="#set-up-giotto-environment"></a>
</h2>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://drieslab.github.io/Giotto/">Giotto</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://drieslab.github.io/GiottoData/">GiottoData</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 1. set working directory</span></span>
<span><span class="va">results_folder</span> <span class="op">=</span> <span class="st">'path/to/result'</span></span>
<span></span>
<span><span class="co"># Optional: Specify a path to a Python executable within a conda or miniconda </span></span>
<span><span class="co"># environment. If set to NULL (default), the Python executable within the previously</span></span>
<span><span class="co"># installed Giotto environment will be used.</span></span>
<span><span class="va">my_python_path</span> <span class="op">=</span> <span class="cn">NULL</span> <span class="co"># alternatively, "/local/python/path/python" if desired.</span></span>
<span></span>
<span><span class="co"># 2. create giotto instructions</span></span>
<span><span class="va">instrs</span> <span class="op">=</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/createGiottoInstructions.html">createGiottoInstructions</a></span><span class="op">(</span>save_dir <span class="op">=</span> <span class="va">results_folder</span>,</span>
<span>                                  save_plot <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                                  show_plot <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                                  python_path <span class="op">=</span> <span class="va">my_python_path</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="dataset-explanation">Dataset explanation<a class="anchor" aria-label="anchor" href="#dataset-explanation"></a>
</h2>
<p>This is a tutorial for Harmony integration of different single cell
RNAseq datasets using two prostate cancer patient datasets. <a href="https:/pubmed.ncbi.nlm.nih.gov/33032611/">Ma et al.</a> Processed
10X Single Cell RNAseq from two prostate cancer patients. The raw
dataset can be found <a href="https:/www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE157703">here</a></p>
</div>
<div class="section level2">
<h2 id="part-1-create-giotto-object-from-10x-dataset-and-join">Part 1: Create Giotto object from 10X dataset and join<a class="anchor" aria-label="anchor" href="#part-1-create-giotto-object-from-10x-dataset-and-join"></a>
</h2>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">giotto_P1</span><span class="op">&lt;-</span><span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/createGiottoObject.html">createGiottoObject</a></span><span class="op">(</span>expression <span class="op">=</span> <span class="fu"><a href="../reference/get10Xmatrix.html">get10Xmatrix</a></span><span class="op">(</span><span class="st">"path/to/P1_result/outs/filtered_feature_bc_matrix"</span>, </span>
<span>    gene_column_index <span class="op">=</span> <span class="fl">2</span>,</span>
<span>    remove_zero_rows <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>    instructions <span class="op">=</span> <span class="va">instrs</span><span class="op">)</span> </span>
<span></span>
<span><span class="va">giotto_P2</span><span class="op">&lt;-</span><span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/createGiottoObject.html">createGiottoObject</a></span><span class="op">(</span>expression <span class="op">=</span> <span class="fu"><a href="../reference/get10Xmatrix.html">get10Xmatrix</a></span><span class="op">(</span><span class="st">"path/to/P2_result/outs/filtered_feature_bc_matrix"</span>, </span>
<span>    gene_column_index <span class="op">=</span> <span class="fl">2</span>,</span>
<span>    remove_zero_rows <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>    instructions <span class="op">=</span> <span class="va">instrs</span><span class="op">)</span> </span>
<span></span>
<span><span class="va">giotto_SC_join</span> <span class="op">=</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/joinGiottoObjects.html">joinGiottoObjects</a></span><span class="op">(</span>gobject_list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">giotto_P1</span>, <span class="va">giotto_P2</span><span class="op">)</span>,</span>
<span>                                   gobject_names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'P1'</span>, <span class="st">'P2'</span><span class="op">)</span>,</span>
<span>                                   join_method <span class="op">=</span> <span class="st">"z_stack"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="part-2-process-joined-object">Part 2: Process Joined object<a class="anchor" aria-label="anchor" href="#part-2-process-joined-object"></a>
</h2>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">giotto_SC_join</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/filterGiotto.html">filterGiotto</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">giotto_SC_join</span>,</span>
<span>                               expression_threshold <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                               feat_det_in_min_cells <span class="op">=</span> <span class="fl">50</span>,</span>
<span>                               min_det_feats_per_cell <span class="op">=</span> <span class="fl">500</span>,</span>
<span>                               expression_values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'raw'</span><span class="op">)</span>,</span>
<span>                               verbose <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## normalize</span></span>
<span><span class="va">giotto_SC_join</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/normalizeGiotto.html">normalizeGiotto</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">giotto_SC_join</span>,</span>
<span>                                  scalefactor <span class="op">=</span> <span class="fl">6000</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## add gene &amp; cell statistics</span></span>
<span><span class="va">giotto_SC_join</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/addStatistics.html">addStatistics</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">giotto_SC_join</span>,</span>
<span>                                expression_values <span class="op">=</span> <span class="st">'raw'</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="part-3-dimension-reduction-and-clustering">Part 3: Dimension reduction and clustering<a class="anchor" aria-label="anchor" href="#part-3-dimension-reduction-and-clustering"></a>
</h2>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## PCA ##</span></span>
<span><span class="va">giotto_SC_join</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateHVF.html">calculateHVF</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">giotto_SC_join</span><span class="op">)</span></span>
<span><span class="va">giotto_SC_join</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runPCA.html">runPCA</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">giotto_SC_join</span>,</span>
<span>                         center <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                         scale_unit <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co"># Check screeplot to select number of PCs for clustering</span></span>
<span><span class="co"># screePlot(giotto_SC_join, ncp = 30, save_param = list(save_name = '3_scree_plot'))</span></span>
<span></span>
<span><span class="co">## WITHOUT INTEGRATION ##</span></span>
<span><span class="co"># --------------------- #</span></span>
<span></span>
<span><span class="co">## cluster and run UMAP ##</span></span>
<span><span class="co"># sNN network (default)</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/showGiottoDimRed.html">showGiottoDimRed</a></span><span class="op">(</span><span class="va">giotto_SC_join</span><span class="op">)</span></span>
<span><span class="va">giotto_SC_join</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/createNearestNetwork.html">createNearestNetwork</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">giotto_SC_join</span>,</span>
<span>                                       dim_reduction_to_use <span class="op">=</span> <span class="st">'pca'</span>,</span>
<span>                                       dim_reduction_name <span class="op">=</span> <span class="st">'pca'</span>,</span>
<span>                                       dimensions_to_use <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span>,</span>
<span>                                       k <span class="op">=</span> <span class="fl">15</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Leiden clustering</span></span>
<span><span class="va">giotto_SC_join</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/doLeidenCluster.html">doLeidenCluster</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">giotto_SC_join</span>,</span>
<span>                                  resolution <span class="op">=</span> <span class="fl">0.2</span>,</span>
<span>                                  n_iterations <span class="op">=</span> <span class="fl">1000</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># UMAP</span></span>
<span><span class="va">giotto_SC_join</span> <span class="op">=</span> <span class="fu"><a href="../reference/runUMAP.html">runUMAP</a></span><span class="op">(</span><span class="va">giotto_SC_join</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/plotUMAP.html">plotUMAP</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">giotto_SC_join</span>,</span>
<span>         cell_color <span class="op">=</span> <span class="st">'leiden_clus'</span>,</span>
<span>         show_NN_network <span class="op">=</span> <span class="cn">T</span>,</span>
<span>         point_size <span class="op">=</span> <span class="fl">1.5</span>,</span>
<span>         save_param <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>save_name <span class="op">=</span> <span class="st">"4_cluster_without_integration"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/singlecell_prostate_integration/4_cluster_without_integration.png" style="width:50.0%"></p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/dimPlot.html">dimPlot2D</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">giotto_SC_join</span>,</span>
<span>          dim_reduction_name <span class="op">=</span> <span class="st">'umap'</span>,</span>
<span>          point_shape <span class="op">=</span> <span class="st">'no_border'</span>,</span>
<span>          cell_color <span class="op">=</span> <span class="st">"leiden_clus"</span>,</span>
<span>          group_by <span class="op">=</span> <span class="st">"list_ID"</span>,</span>
<span>          show_NN_network <span class="op">=</span> <span class="cn">F</span>,</span>
<span>          point_size <span class="op">=</span> <span class="fl">0.5</span>, </span>
<span>          show_center_label <span class="op">=</span> <span class="cn">F</span>,</span>
<span>          show_legend <span class="op">=</span><span class="cn">F</span>,</span>
<span>          save_param <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>save_name <span class="op">=</span> <span class="st">"4_list_without_integration"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/singlecell_prostate_integration/4_list_without_integration.png" style="width:50.0%"></p>
<p>Harmony is a integration algorithm developed by <a href="https:/www.nature.com/articles/s41592-019-0619-0">Korsunsky, I. et
al.</a>. It was designed for integration of single cell data but also
work well on spatial datasets.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## WITH INTEGRATION ##</span></span>
<span><span class="co"># --------------------- #</span></span>
<span></span>
<span><span class="co">## data integration, cluster and run UMAP ##</span></span>
<span></span>
<span><span class="co"># harmony</span></span>
<span><span class="co">#library(devtools)</span></span>
<span><span class="co">#install_github("immunogenomics/harmony")</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">harmony</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#pDataDT(giotto_SC_join)</span></span>
<span><span class="va">giotto_SC_join</span> <span class="op">=</span> <span class="fu"><a href="../reference/runGiottoHarmony.html">runGiottoHarmony</a></span><span class="op">(</span><span class="va">giotto_SC_join</span>,</span>
<span>                                  vars_use <span class="op">=</span> <span class="st">'list_ID'</span>,</span>
<span>                                  do_pca <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co">## sNN network (default)</span></span>
<span><span class="co">#showGiottoDimRed(giotto_SC_join)</span></span>
<span><span class="va">giotto_SC_join</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/createNearestNetwork.html">createNearestNetwork</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">giotto_SC_join</span>,</span>
<span>                                       dim_reduction_to_use <span class="op">=</span> <span class="st">'harmony'</span>,</span>
<span>                                       dim_reduction_name <span class="op">=</span> <span class="st">'harmony'</span>,</span>
<span>                                       name <span class="op">=</span> <span class="st">'NN.harmony'</span>,</span>
<span>                                       dimensions_to_use <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span>,</span>
<span>                                       k <span class="op">=</span> <span class="fl">15</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Leiden clustering</span></span>
<span><span class="va">giotto_SC_join</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/doLeidenCluster.html">doLeidenCluster</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">giotto_SC_join</span>,</span>
<span>                                  network_name <span class="op">=</span> <span class="st">'NN.harmony'</span>,</span>
<span>                                  resolution <span class="op">=</span> <span class="fl">0.2</span>,</span>
<span>                                  n_iterations <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>                                  name <span class="op">=</span> <span class="st">'leiden_harmony'</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># UMAP dimension reduction</span></span>
<span><span class="co">#showGiottoDimRed(giotto_SC_join)</span></span>
<span><span class="va">giotto_SC_join</span> <span class="op">=</span> <span class="fu"><a href="../reference/runUMAP.html">runUMAP</a></span><span class="op">(</span><span class="va">giotto_SC_join</span>,</span>
<span>                         dim_reduction_name <span class="op">=</span> <span class="st">'harmony'</span>,</span>
<span>                         dim_reduction_to_use <span class="op">=</span> <span class="st">'harmony'</span>,</span>
<span>                         name <span class="op">=</span> <span class="st">'umap_harmony'</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/plotUMAP.html">plotUMAP</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">giotto_SC_join</span>,</span>
<span>         dim_reduction_name <span class="op">=</span> <span class="st">'umap_harmony'</span>,</span>
<span>         cell_color <span class="op">=</span> <span class="st">'leiden_harmony'</span>,</span>
<span>         show_NN_network <span class="op">=</span> <span class="cn">T</span>,</span>
<span>         point_size <span class="op">=</span> <span class="fl">1.5</span>,</span>
<span>         save_param <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>save_name <span class="op">=</span> <span class="st">"4_cluster_with_integration"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/singlecell_prostate_integration/4_cluster_with_integration.png" style="width:50.0%"></p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/dimPlot.html">dimPlot2D</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">giotto_SC_join</span>,</span>
<span>          dim_reduction_name <span class="op">=</span> <span class="st">'umap_harmony'</span>,</span>
<span>          point_shape <span class="op">=</span> <span class="st">'no_border'</span>,</span>
<span>          cell_color <span class="op">=</span> <span class="st">"leiden_harmony"</span>,</span>
<span>          group_by <span class="op">=</span> <span class="st">"list_ID"</span>,</span>
<span>          show_NN_network <span class="op">=</span> <span class="cn">F</span>,</span>
<span>          point_size <span class="op">=</span> <span class="fl">0.5</span>, </span>
<span>          show_center_label <span class="op">=</span> <span class="cn">F</span>,</span>
<span>          show_legend <span class="op">=</span><span class="cn">F</span> ,</span>
<span>          save_param <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>save_name <span class="op">=</span> <span class="st">"4_list_with_integration"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/singlecell_prostate_integration/4_list_with_integration.png" style="width:50.0%"></p>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by <a href="https://www.drieslab.com" class="external-link">Ruben Dries</a>, Jiaji George Chen, Natalie Del Rossi, <a href="https://labs.icahn.mssm.edu/yuanlab/" class="external-link">Guo-Cheng Yuan</a>, Matthew O’Brien, <a href="https://josschavezf.github.io" class="external-link">Joselyn C. Chávez-Fuentes</a>, Edward Ruiz.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
