<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Visium CytAssist Human Lung Cancer â€¢ Giotto</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Visium CytAssist Human Lung Cancer">
<meta property="og:description" content="Giotto">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">Giotto</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">4.0.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Get started
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/installation.html">Giotto Installation</a>
    </li>
    <li>
      <a href="../articles/ecosystem.html">Giotto Ecosystem</a>
    </li>
    <li>
      <a href="../articles/structure.html">Giotto Class Structure</a>
    </li>
    <li>
      <a href="../articles/errorsfaqsandtips.html">Solutions to Installation Errors</a>
    </li>
    <li>
      <a href="../articles/faqs.html">FAQs</a>
    </li>
    <li>
      <a href="../articles/branches.html">Branches</a>
    </li>
    <li>
      <a href="../articles/github_issues.html">How to report a bug</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Documentation</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Examples
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Seq-Scope</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/seqscope_mouse_liver.html">Mouse liver</a>
        </li>
      </ul>
</li>
    <li class="divider">
    </li>
<li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Visium</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/visium_mouse_kidney.html">Visium Mouse Kidney</a>
        </li>
        <li>
          <a href="../articles/visium_mouse_brain.html">Visium Mouse Brain</a>
        </li>
        <li>
          <a href="../articles/visium_prostate_integration.html">Visium Prostate Integration</a>
        </li>
        <li>
          <a href="../articles/visium_cytassist_lungcancer.html">CytAssist Human Lung Cancer</a>
        </li>
        <li>
          <a href="../articles/visium_cytassist_human_glioblastoma.html">CytAssist Multi-omics Human Glioblastoma</a>
        </li>
      </ul>
</li>
    <li class="divider">
    </li>
<li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Spatial Genomics</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/spatial_genomics_mouse_kidney.html">Mouse Kidney</a>
        </li>
      </ul>
</li>
    <li class="divider">
    </li>
<li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Xenium</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/xenium_breast_cancer.html">Xenium Human breast cancer</a>
        </li>
      </ul>
</li>
    <li class="divider">
    </li>
<li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Nanostring CosMx</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/nanostring_cosmx_lung_cancer.html">Nanostring CosMx Subcellular Lung Cancer</a>
        </li>
      </ul>
</li>
    <li class="divider">
    </li>
<li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">MERFISH</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/merfish_mouse_hypothalamic.html">MERFISH Mouse Hypothalamic Preoptic Region</a>
        </li>
      </ul>
</li>
    <li class="divider">
    </li>
<li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Vizgen</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/vizgen_mouse_brain.html">Vizgen Mouse Brain Receptor Map</a>
        </li>
      </ul>
</li>
    <li class="divider">
    </li>
<li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">seqFISH</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/mini_seqfish.html">mini seqFISH</a>
        </li>
        <li>
          <a href="../articles/seqfish_cortex.html">SeqFish+ Mouse Cortex Example</a>
        </li>
      </ul>
</li>
    <li class="divider">
    </li>
<li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">osmFISH</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/osmfish_mouse_ss_cortex.html">osmFISH Mouse SS Cortex</a>
        </li>
      </ul>
</li>
    <li class="divider">
    </li>
<li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Resolve Bioscience</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/resolve_biosciences.html">Resolve Bioscience Breast Cancer Subcellular</a>
        </li>
      </ul>
</li>
    <li class="divider">
    </li>
<li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">CODEX</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/codex_mouse_spleen.html">Codex Mouse Spleen</a>
        </li>
      </ul>
</li>
    <li class="divider">
    </li>
<li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Single Cell RNA Sequencing</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/singlecell_prostate_standard.html">Prostate</a>
        </li>
        <li>
          <a href="../articles/singlecell_prostate_integration.html">Prostate Integration</a>
        </li>
      </ul>
</li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/object_creation.html">Giotto Object Creation</a>
    </li>
    <li class="divider">
    </li>
<li>
      <a href="../articles/configuration.html">Giotto Configuration</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Spatial manipulation</li>
    <li>
      <a href="../articles/spatial_transformations.html">Spatial transformations</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Image alignment</li>
    <li>
      <a href="../articles/image_alignment.html">Image alignment</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Interactive selection</li>
    <li>
      <a href="../articles/interactive_selection.html">Interactive selection</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Visualizations</li>
    <li>
      <a href="../articles/visualizations.html">Visualizations</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Saving and Loading</li>
    <li>
      <a href="../articles/saving_object.html">Saving a Giotto object</a>
    </li>
    <li>
      <a href="../articles/saving_plots.html">Saving plots</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Interoperability</li>
    <li>
      <a href="../articles/interoperability_anndata.html">AnnData</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Try Giotto in the Cloud</li>
    <li>
      <a href="../articles/terra.html">Terra.bio</a>
    </li>
    <li>
      <a href="../articles/docker.html">Docker</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    News
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../news/index.html">Changes</a>
    </li>
    <li>
      <a href="../articles/presentations.html">Presentations</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/drieslab/Giotto/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Visium CytAssist Human Lung Cancer</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/drieslab/Giotto/blob/HEAD/vignettes/visium_cytassist_lungcancer.Rmd" class="external-link"><code>vignettes/visium_cytassist_lungcancer.Rmd</code></a></small>
      <div class="hidden name"><code>visium_cytassist_lungcancer.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="dataset-explanation">Dataset Explanation<a class="anchor" aria-label="anchor" href="#dataset-explanation"></a>
</h2>
<p>The human lung cancer (FFPE) dataset was obtained from 10x Genomics
using their CytAssist Visium technology that has been recently developed
to allow users to perform standard histology workflows on two standard
glass slides before transferring the transcriptional probes on the
two-area capture visium slide.</p>
<p>More information about this dataset can be found <a href="https://www.10xgenomics.com/resources/datasets/human-lung-cancer-ffpe-2-standard/" class="external-link">here</a>.</p>
<p><img src="images/visium_cytassist_lungcancer/CytAssist_Visium_workflow.png" style="width:90.0%"></p>
</div>
<div class="section level2">
<h2 id="start-giotto">Start Giotto<a class="anchor" aria-label="anchor" href="#start-giotto"></a>
</h2>
<p>To run the current vignette you need to install the Giotto Suite
branch.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Ensure Giotto Suite and GiottoData packages are installed</span></span>
<span><span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="st">"Giotto"</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/installed.packages.html" class="external-link">installed.packages</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>   <span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"drieslab/Giotto@suite"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="st">"Giotto"</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/installed.packages.html" class="external-link">installed.packages</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>   <span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"drieslab/GiottoData"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://drieslab.github.io/Giotto/">Giotto</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://drieslab.github.io/GiottoData/">GiottoData</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Ensure the Python environment for Giotto has been installed</span></span>
<span><span class="va">genv_exists</span> <span class="op">=</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/checkGiottoEnvironment.html">checkGiottoEnvironment</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="va">genv_exists</span><span class="op">)</span><span class="op">{</span></span>
<span>   <span class="co"># The following command need only be run once to install the Giotto environment.</span></span>
<span>   <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/installGiottoEnvironment.html">installGiottoEnvironment</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co"># to automatically save figures in save_dir set save_plot to TRUE</span></span>
<span><span class="va">temp_dir</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/getwd.html" class="external-link">getwd</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">myinstructions</span> <span class="op">=</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/createGiottoInstructions.html">createGiottoInstructions</a></span><span class="op">(</span>save_dir <span class="op">=</span> <span class="va">temp_dir</span>,</span>
<span>                                          save_plot <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                                          show_plot <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="create-a-giotto-object">1. Create a Giotto object<a class="anchor" aria-label="anchor" href="#create-a-giotto-object"></a>
</h2>
<p>The minimum requirements are</p>
<ul>
<li>matrix with expression information (or path to)</li>
<li>x,y(,z) coordinates for cells or spots (or path to)</li>
</ul>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Provide path to visium folder</span></span>
<span><span class="va">data_path</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">temp_dir</span>, <span class="st">'/data/'</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/GiottoData/reference/getSpatialDataset.html">getSpatialDataset</a></span><span class="op">(</span>dataset <span class="op">=</span> <span class="st">'cytassist_lungcancer'</span>, directory <span class="op">=</span> <span class="va">data_path</span>, method <span class="op">=</span> <span class="st">'wget'</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Create Giotto object</span></span>
<span>  <span class="va">visium_lungcancer</span> <span class="op">=</span> <span class="fu"><a href="../reference/createGiottoVisiumObject.html">createGiottoVisiumObject</a></span><span class="op">(</span>visium_dir <span class="op">=</span> <span class="va">data_path</span>,</span>
<span>                                               expr_data <span class="op">=</span> <span class="st">'raw'</span>,</span>
<span>                                               png_name <span class="op">=</span> <span class="st">'tissue_lowres_image.png'</span>,</span>
<span>                                               gene_column_index <span class="op">=</span> <span class="fl">2</span>,</span>
<span>                                               instructions <span class="op">=</span> <span class="va">myinstructions</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># check metadata</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/pDataDT.html">pDataDT</a></span><span class="op">(</span><span class="va">visium_lungcancer</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># check available image names</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/showGiottoImageNames.html">showGiottoImageNames</a></span><span class="op">(</span><span class="va">visium_lungcancer</span><span class="op">)</span> <span class="co"># "image" is the default name</span></span>
<span></span>
<span><span class="co"># show aligned image</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/GiottoVisuals/reference/spatPlot.html">spatPlot</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">visium_lungcancer</span>, cell_color <span class="op">=</span> <span class="st">'in_tissue'</span>, show_image <span class="op">=</span> <span class="cn">T</span>, point_alpha <span class="op">=</span> <span class="fl">0.7</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/visium_cytassist_lungcancer/0-spatPlot2D.png" style="width:50.0%"></p>
<p>How to work with Giotto instructions that are part of your Giotto
object:</p>
<ul>
<li>show the instructions associated with your Giotto object with
<strong>showGiottoInstructions()</strong>
</li>
<li>change one or more instructions with
<strong>changeGiottoInstructions()</strong>
</li>
<li>replace all instructions at once with
<strong>replaceGiottoInstructions()</strong>
</li>
<li>read or get a specific Giotto instruction with
<strong>readGiottoInstructions()</strong>
</li>
</ul>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># show instructions associated with giotto object (visium_lungcancer)</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/showGiottoInstructions.html">showGiottoInstructions</a></span><span class="op">(</span><span class="va">visium_lungcancer</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="processing-steps">2. Processing steps<a class="anchor" aria-label="anchor" href="#processing-steps"></a>
</h2>
<ul>
<li>filter genes and cells based on detection frequencies</li>
<li>normalize expression matrix (log transformation, scaling factor
and/or z-scores)</li>
<li>add cell and gene statistics (optional)</li>
<li>adjust expression matrix for technical covariates or batches
(optional). These results will be stored in the <em>custom</em>
slot.</li>
</ul>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">visium_lungcancer</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/filterGiotto.html">filterGiotto</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">visium_lungcancer</span>,</span>
<span>                                  expression_threshold <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                                  feat_det_in_min_cells <span class="op">=</span> <span class="fl">50</span>,</span>
<span>                                  min_det_feats_per_cell <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>                                  expression_values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'raw'</span><span class="op">)</span>,</span>
<span>                                  verbose <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span><span class="va">visium_lungcancer</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/normalizeGiotto.html">normalizeGiotto</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">visium_lungcancer</span>, scalefactor <span class="op">=</span> <span class="fl">6000</span>, verbose <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span><span class="va">visium_lungcancer</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/addStatistics.html">addStatistics</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">visium_lungcancer</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="visualize-aligned-tissue-with-number-of-features-after-processing">Visualize aligned tissue with number of features after
processing<a class="anchor" aria-label="anchor" href="#visualize-aligned-tissue-with-number-of-features-after-processing"></a>
</h3>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://drieslab.github.io/GiottoVisuals/reference/spatPlot.html">spatPlot2D</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">visium_lungcancer</span>, show_image <span class="op">=</span> <span class="cn">T</span>, point_alpha <span class="op">=</span> <span class="fl">0.7</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/visium_cytassist_lungcancer/1-spatPlot2D.png" style="width:50.0%"></p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://drieslab.github.io/GiottoVisuals/reference/spatPlot.html">spatPlot2D</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">visium_lungcancer</span>, show_image <span class="op">=</span> <span class="cn">T</span>, point_alpha <span class="op">=</span> <span class="fl">0.7</span>,</span>
<span>           cell_color <span class="op">=</span> <span class="st">'nr_feats'</span>, color_as_factor <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/visium_cytassist_lungcancer/2-spatPlot2D.png" style="width:50.0%"></p>
</div>
</div>
<div class="section level2">
<h2 id="dimension-reduction">3. Dimension Reduction<a class="anchor" aria-label="anchor" href="#dimension-reduction"></a>
</h2>
<ul>
<li>identify highly variable features (HVF)</li>
</ul>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">visium_lungcancer</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateHVF.html">calculateHVF</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">visium_lungcancer</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/visium_cytassist_lungcancer/3-HVFplot.png" style="width:50.0%"></p>
<ul>
<li>perform PCA</li>
<li>identify number of significant principal components (PCs)</li>
</ul>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">visium_lungcancer</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runPCA.html">runPCA</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">visium_lungcancer</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/screePlot.html">screePlot</a></span><span class="op">(</span><span class="va">visium_lungcancer</span>, ncp <span class="op">=</span> <span class="fl">30</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/visium_cytassist_lungcancer/4-screePlot.png" style="width:50.0%"></p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://drieslab.github.io/GiottoVisuals/reference/plotPCA.html">plotPCA</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">visium_lungcancer</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/visium_cytassist_lungcancer/5-PCA.png" style="width:50.0%"></p>
<ul>
<li>run UMAP and/or t-SNE on PCs (or directly on matrix)</li>
</ul>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">visium_lungcancer</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runUMAP.html">runUMAP</a></span><span class="op">(</span><span class="va">visium_lungcancer</span>, dimensions_to_use <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/GiottoVisuals/reference/plotUMAP.html">plotUMAP</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">visium_lungcancer</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/visium_cytassist_lungcancer/6-UMAP.png" style="width:50.0%"></p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">visium_lungcancer</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runtSNE.html">runtSNE</a></span><span class="op">(</span><span class="va">visium_lungcancer</span>, dimensions_to_use <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/GiottoVisuals/reference/plotTSNE.html">plotTSNE</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">visium_lungcancer</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/visium_cytassist_lungcancer/7-tSNE.png" style="width:50.0%"></p>
</div>
<div class="section level2">
<h2 id="clustering">4. Clustering<a class="anchor" aria-label="anchor" href="#clustering"></a>
</h2>
<ul>
<li>create a shared (default) nearest network in PCA space (or directly
on matrix)</li>
<li>cluster on nearest network with Leiden or Louvain (k-means and
hclust are alternatives)</li>
</ul>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Create shared nearest network (SNN) and perform leiden clustering</span></span>
<span><span class="va">visium_lungcancer</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/createNearestNetwork.html">createNearestNetwork</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">visium_lungcancer</span>, dimensions_to_use <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, k <span class="op">=</span> <span class="fl">30</span><span class="op">)</span></span>
<span><span class="va">visium_lungcancer</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/doLeidenCluster.html">doLeidenCluster</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">visium_lungcancer</span>, spat_unit <span class="op">=</span> <span class="st">'cell'</span>, feat_type <span class="op">=</span> <span class="st">'rna'</span>, resolution <span class="op">=</span> <span class="fl">0.4</span>, n_iterations <span class="op">=</span> <span class="fl">1000</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># visualize UMAP cluster results</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/GiottoVisuals/reference/plotUMAP.html">plotUMAP</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">visium_lungcancer</span>, cell_color <span class="op">=</span> <span class="st">'leiden_clus'</span>, show_NN_network <span class="op">=</span> <span class="cn">T</span>, point_size <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/visium_cytassist_lungcancer/8-UMAP.png" style="width:50.0%"></p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># visualize tSNE cluster results</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/GiottoVisuals/reference/plotTSNE.html">plotTSNE</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">visium_lungcancer</span>, cell_color <span class="op">=</span> <span class="st">'leiden_clus'</span>, show_NN_network <span class="op">=</span> <span class="cn">T</span>, point_size <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/visium_cytassist_lungcancer/9-tSNE.png" style="width:50.0%"></p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># visualize expression and spatial results</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/GiottoVisuals/reference/spatDimPlot.html">spatDimPlot</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">visium_lungcancer</span>, cell_color <span class="op">=</span> <span class="st">'leiden_clus'</span>,</span>
<span>            dim_point_size <span class="op">=</span> <span class="fl">2</span>, spat_point_size <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/visium_cytassist_lungcancer/10-spatDimPlot2D.png" style="width:50.0%"></p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://drieslab.github.io/GiottoVisuals/reference/spatDimPlot.html">spatDimPlot</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">visium_lungcancer</span>, cell_color <span class="op">=</span> <span class="st">'nr_feats'</span>, color_as_factor <span class="op">=</span> <span class="cn">F</span>,</span>
<span>            dim_point_size <span class="op">=</span> <span class="fl">2</span>, dim_show_legend <span class="op">=</span> <span class="cn">T</span>, spat_show_legend <span class="op">=</span> <span class="cn">T</span>, spat_point_size <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/visium_cytassist_lungcancer/11-spatDimPlot2D.png" style="width:50.0%"></p>
</div>
<div class="section level2">
<h2 id="differential-expression">5. Differential expression<a class="anchor" aria-label="anchor" href="#differential-expression"></a>
</h2>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Cell type marker detection</span></span>
<span><span class="co"># Gini markers</span></span>
<span><span class="va">gini_markers_subclusters</span> <span class="op">=</span> <span class="fu"><a href="../reference/findMarkers_one_vs_all.html">findMarkers_one_vs_all</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">visium_lungcancer</span>,</span>
<span>                                                  method <span class="op">=</span> <span class="st">'gini'</span>,</span>
<span>                                                  expression_values <span class="op">=</span> <span class="st">'normalized'</span>,</span>
<span>                                                  cluster_column <span class="op">=</span> <span class="st">'leiden_clus'</span>,</span>
<span>                                                  min_featss <span class="op">=</span> <span class="fl">20</span>,</span>
<span>                                                  min_expr_gini_score <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>                                                  min_det_gini_score <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># get top 2 genes per cluster and visualize with violin plot</span></span>
<span><span class="va">topgenes_gini</span> <span class="op">=</span> <span class="va">gini_markers_subclusters</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">.SD</span>, <span class="fl">2</span><span class="op">)</span>, by <span class="op">=</span> <span class="st">'cluster'</span><span class="op">]</span><span class="op">$</span><span class="va">feats</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/GiottoVisuals/reference/violinPlot.html">violinPlot</a></span><span class="op">(</span><span class="va">visium_lungcancer</span>, feats <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">topgenes_gini</span><span class="op">)</span>, cluster_column <span class="op">=</span> <span class="st">'leiden_clus'</span>,</span>
<span>           strip_text <span class="op">=</span> <span class="fl">8</span>, strip_position <span class="op">=</span> <span class="st">'right'</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/visium_cytassist_lungcancer/12-violinPlot.png" style="width:50.0%"></p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># cluster heatmap</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/GiottoVisuals/reference/plotMetaDataHeatmap.html">plotMetaDataHeatmap</a></span><span class="op">(</span><span class="va">visium_lungcancer</span>,</span>
<span>                    selected_feats <span class="op">=</span> <span class="va">topgenes_gini</span>,</span>
<span>                    metadata_cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'leiden_clus'</span><span class="op">)</span>,</span>
<span>                    x_text_size <span class="op">=</span> <span class="fl">10</span>, y_text_size <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/visium_cytassist_lungcancer/13-plotMetaDataHeatmap.png" style="width:50.0%"></p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># umap plots</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/GiottoVisuals/reference/dimFeatPlot2D.html">dimFeatPlot2D</a></span><span class="op">(</span><span class="va">visium_lungcancer</span>,</span>
<span>              expression_values <span class="op">=</span> <span class="st">'scaled'</span>,</span>
<span>              feats <span class="op">=</span> <span class="va">gini_markers_subclusters</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">.SD</span>, <span class="fl">1</span><span class="op">)</span>, by <span class="op">=</span> <span class="st">'cluster'</span><span class="op">]</span><span class="op">$</span><span class="va">feats</span>,</span>
<span>              cow_n_col <span class="op">=</span> <span class="fl">3</span>, point_size <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/visium_cytassist_lungcancer/14-dimFeatPlot2D.png" style="width:50.0%"></p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Cell type marker detection</span></span>
<span><span class="co"># Scran markers</span></span>
<span><span class="va">scran_markers_subclusters</span> <span class="op">=</span> <span class="fu"><a href="../reference/findMarkers_one_vs_all.html">findMarkers_one_vs_all</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">visium_lungcancer</span>,</span>
<span>                                                   method <span class="op">=</span> <span class="st">'scran'</span>,</span>
<span>                                                   expression_values <span class="op">=</span> <span class="st">'normalized'</span>,</span>
<span>                                                   cluster_column <span class="op">=</span> <span class="st">'leiden_clus'</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># get top 2 genes per cluster and visualize with violin plot</span></span>
<span><span class="va">topgenes_scran</span> <span class="op">=</span> <span class="va">scran_markers_subclusters</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">.SD</span>, <span class="fl">2</span><span class="op">)</span>, by <span class="op">=</span> <span class="st">'cluster'</span><span class="op">]</span><span class="op">$</span><span class="va">feats</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/GiottoVisuals/reference/violinPlot.html">violinPlot</a></span><span class="op">(</span><span class="va">visium_lungcancer</span>, feats <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">topgenes_scran</span><span class="op">)</span>,</span>
<span>           cluster_column <span class="op">=</span> <span class="st">'leiden_clus'</span>,</span>
<span>           strip_text <span class="op">=</span> <span class="fl">10</span>, strip_position <span class="op">=</span> <span class="st">'right'</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/visium_cytassist_lungcancer/15-violinPlot.png" style="width:50.0%"></p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># cluster heatmap</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/GiottoVisuals/reference/plotMetaDataHeatmap.html">plotMetaDataHeatmap</a></span><span class="op">(</span><span class="va">visium_lungcancer</span>,</span>
<span>                    selected_feats <span class="op">=</span> <span class="va">topgenes_scran</span>,</span>
<span>                    metadata_cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'leiden_clus'</span><span class="op">)</span>,</span>
<span>                    x_text_size <span class="op">=</span> <span class="fl">10</span>, y_text_size <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/visium_cytassist_lungcancer/16-plotMetaDataHeatmap.png" style="width:50.0%"></p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># umap plots</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/GiottoVisuals/reference/dimFeatPlot2D.html">dimFeatPlot2D</a></span><span class="op">(</span><span class="va">visium_lungcancer</span>,</span>
<span>              expression_values <span class="op">=</span> <span class="st">'scaled'</span>,</span>
<span>              feats <span class="op">=</span> <span class="va">scran_markers_subclusters</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">.SD</span>, <span class="fl">1</span><span class="op">)</span>, by <span class="op">=</span> <span class="st">'cluster'</span><span class="op">]</span><span class="op">$</span><span class="va">feats</span>,</span>
<span>              cow_n_col <span class="op">=</span> <span class="fl">3</span>, point_size <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="cell-type-enrichment">6. Cell Type Enrichment<a class="anchor" aria-label="anchor" href="#cell-type-enrichment"></a>
</h2>
<p>Visium spatial transcriptomics does not provide single-cell
resolution, making cell type annotation a harder problem. Giotto
provides several ways to calculate enrichment of specific cell-type
signature gene lists:</p>
<ul>
<li>PAGE</li>
<li>hypergeometric test</li>
<li>Rank</li>
<li>DWLS</li>
</ul>
<p><a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-021-02362-7" class="external-link">Deconvolution</a>
Corresponded Single cell dataset can be generated from <a href="http://mousebrain.org/" class="external-link">here</a>. Giotto_SC is processed from the
downsampled <a href="https://satijalab.org/loomr/loomr_tutorial" class="external-link">Loom</a> file and can
also be downloaded from getSpatialDataset.</p>
<div class="section level3">
<h3 id="page">PAGE<a class="anchor" aria-label="anchor" href="#page"></a>
</h3>
<p>The cell-type specific signature gene list was obtained from a
previous study focused on investigating the therapy-induced evolution of
lung cancer revealed by single-cell RNA sequencing. More information
about the paper and scRNA-seq dataset can be found <a href="https://doi.org/10.1016/j.cell.2020.07.017" class="external-link">here</a>.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># umap plots</span></span>
<span><span class="co"># Create PAGE matrix</span></span>
<span><span class="co"># PAGE matrix should be a binary matrix with each row represent a gene marker and each column represent a cell type</span></span>
<span><span class="co"># There are several ways to create PAGE matrix</span></span>
<span><span class="co"># 1.1 create binary matrix of cell signature genes</span></span>
<span><span class="co"># small example #</span></span>
<span><span class="va">Tcells_markers</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD2"</span>, <span class="st">"CD3D"</span>, <span class="st">"CD3E"</span>, <span class="st">"CD3G"</span><span class="op">)</span></span>
<span><span class="va">macrophage_markers</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"MARCO"</span>, <span class="st">"CSF1R"</span>, <span class="st">"CD68"</span>, <span class="st">"GLDN"</span>, <span class="st">"APOE"</span>, <span class="st">"CCL3L1"</span>, <span class="st">"TREM2"</span>, <span class="st">"C1QB"</span>, <span class="st">"NUPR1"</span>, <span class="st">"FOLR2"</span>, <span class="st">"RNASE1"</span>, <span class="st">"C1QA"</span><span class="op">)</span></span>
<span><span class="va">dendritic_markers</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD1E"</span>, <span class="st">"CD1C"</span>, <span class="st">"FCER1A"</span>, <span class="st">"PKIB"</span>, <span class="st">"CYP2S1"</span>, <span class="st">"NDRG2"</span><span class="op">)</span></span>
<span><span class="va">mast_markers</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CMA1"</span>, <span class="st">"TPSAB1"</span>, <span class="st">"TPSB2"</span><span class="op">)</span></span>
<span><span class="va">Bcell_markers</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"IGLL5"</span>, <span class="st">"MZB1"</span>, <span class="st">"JCHAIN"</span>, <span class="st">"DERL3"</span>, <span class="st">"SDC1"</span>, <span class="st">"MS$A1"</span>, <span class="st">"BANK1"</span>, <span class="st">"PAX5"</span>, <span class="st">"CD79A"</span><span class="op">)</span></span>
<span><span class="va">Bcell_PB_markers</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"PRDM1"</span>, <span class="st">"XSP1"</span>, <span class="st">"IRF4"</span><span class="op">)</span></span>
<span><span class="va">Bcell_mem_markers</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"MS4A1"</span>, <span class="st">"IRF8"</span><span class="op">)</span></span>
<span><span class="va">housekeeping_markers</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ACTB"</span>, <span class="st">"GAPDH"</span>, <span class="st">"MALAT1"</span><span class="op">)</span></span>
<span><span class="va">neutrophils_markers</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"FCGR3B"</span>, <span class="st">"ALPL"</span>, <span class="st">"CXCR1"</span>, <span class="st">"CXCR2"</span>, <span class="st">"ADGRG3"</span>, <span class="st">"CMTM2"</span>, <span class="st">"PROK2"</span>, <span class="st">"MME"</span>, <span class="st">"MMP25"</span>, <span class="st">"TNFRSF10C"</span><span class="op">)</span></span>
<span><span class="va">pdcs_markers</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"SLC32A1"</span>, <span class="st">"SHD"</span>, <span class="st">"LRRC26"</span>, <span class="st">"PACSIN1"</span>, <span class="st">"LILRA4"</span>, <span class="st">"CLEC4C"</span>, <span class="st">"DNASE1L3"</span>, <span class="st">"SCT"</span>, <span class="st">"LAMP5"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">signature_matrix</span> <span class="op">=</span> <span class="fu"><a href="../reference/makeSignMatrixPAGE.html">makeSignMatrixPAGE</a></span><span class="op">(</span>sign_names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'T_Cells'</span>, <span class="st">'Macrophage'</span>, <span class="st">'Dendritic'</span>, <span class="st">'Mast'</span>, <span class="st">'B_cell'</span>, <span class="st">'Bcell_PB'</span>, <span class="st">'Bcells_memory'</span>,</span>
<span><span class="st">'Housekeeping'</span>, <span class="st">'Neutrophils'</span>, <span class="st">'pDCs'</span><span class="op">)</span>,</span>
<span> sign_list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">Tcells_markers</span>,</span>
<span>           <span class="va">macrophage_markers</span>,</span>
<span>           <span class="va">dendritic_markers</span>,</span>
<span>           <span class="va">mast_markers</span>,</span>
<span>           <span class="va">Bcell_markers</span>,</span>
<span>           <span class="va">Bcell_PB_markers</span>,</span>
<span>           <span class="va">Bcell_mem_markers</span>,</span>
<span>           <span class="va">housekeeping_markers</span>,</span>
<span>           <span class="va">neutrophils_markers</span>,</span>
<span>           <span class="va">pdcs_markers</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 1.3 enrichment test with PAGE</span></span>
<span></span>
<span><span class="va">markers_scran</span> <span class="op">=</span> <span class="fu"><a href="../reference/findMarkers_one_vs_all.html">findMarkers_one_vs_all</a></span><span class="op">(</span>gobject<span class="op">=</span><span class="va">giotto_SC</span>, method<span class="op">=</span><span class="st">"scran"</span>,</span>
<span>                                       expression_values<span class="op">=</span><span class="st">"normalized"</span>, cluster_column <span class="op">=</span> <span class="st">"Class"</span>, min_feats<span class="op">=</span><span class="fl">3</span><span class="op">)</span></span>
<span></span>
<span><span class="va">top_markers</span> <span class="op">&lt;-</span> <span class="va">markers_scran</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">.SD</span>, <span class="fl">10</span><span class="op">)</span>, by<span class="op">=</span><span class="st">"cluster"</span><span class="op">]</span></span>
<span><span class="va">celltypes</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/levels.html" class="external-link">levels</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">markers_scran</span><span class="op">$</span><span class="va">cluster</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">sign_list</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">celltypes</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span> <span class="va">sign_list</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">&lt;-</span><span class="va">top_markers</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">top_markers</span><span class="op">$</span><span class="va">cluster</span> <span class="op">==</span> <span class="va">celltypes</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span>,<span class="op">]</span><span class="op">$</span><span class="va">feats</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">PAGE_matrix_3</span> <span class="op">=</span> <span class="fu"><a href="../reference/makeSignMatrixPAGE.html">makeSignMatrixPAGE</a></span><span class="op">(</span>sign_names <span class="op">=</span> <span class="va">celltypes</span>,</span>
<span>                                   sign_list <span class="op">=</span> <span class="va">sign_list</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#  runSpatialEnrich() can also be used as a wrapper for all currently provided enrichment options</span></span>
<span><span class="va">visium_lungcancer</span> <span class="op">=</span> <span class="fu"><a href="../reference/runPAGEEnrich.html">runPAGEEnrich</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">visium_lungcancer</span>, sign_matrix <span class="op">=</span> <span class="va">signature_matrix</span>, min_overlap_genes <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 1.4 heatmap of enrichment versus annotation (e.g. clustering result)</span></span>
<span><span class="va">cell_types</span> <span class="op">=</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/row-plus-colnames-generic.html">colnames</a></span><span class="op">(</span><span class="va">signature_matrix</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/GiottoVisuals/reference/plotMetaDataCellsHeatmap.html">plotMetaDataCellsHeatmap</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">visium_lungcancer</span>,</span>
<span>                         metadata_cols <span class="op">=</span> <span class="st">'leiden_clus'</span>,</span>
<span>                         value_cols <span class="op">=</span> <span class="va">cell_types</span>,</span>
<span>                         spat_enr_names <span class="op">=</span> <span class="st">'PAGE'</span>,</span>
<span>                         x_text_size <span class="op">=</span> <span class="fl">8</span>,</span>
<span>                         y_text_size <span class="op">=</span> <span class="fl">8</span>,</span>
<span>                         show_plot <span class="op">=</span> <span class="cn">T</span>,</span>
<span>                         save_param <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>save_name<span class="op">=</span><span class="st">"7_a_metaheatmap"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/visium_cytassist_lungcancer/17-metaheatmap.png" style="width:50.0%"></p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cell_types_subset</span> <span class="op">=</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/row-plus-colnames-generic.html">colnames</a></span><span class="op">(</span><span class="va">signature_matrix</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/GiottoVisuals/reference/spatCellPlot.html">spatCellPlot</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">visium_lungcancer</span>,</span>
<span>             spat_enr_names <span class="op">=</span> <span class="st">'PAGE'</span>,</span>
<span>             cell_annotation_values <span class="op">=</span> <span class="va">cell_types_subset</span>,</span>
<span>             cow_n_col <span class="op">=</span> <span class="fl">4</span>, coord_fix_ratio <span class="op">=</span> <span class="cn">NULL</span>, point_size <span class="op">=</span> <span class="fl">0.75</span>,</span>
<span>             save_param <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>save_name<span class="op">=</span><span class="st">"7_b_spatcellplot_1"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/visium_cytassist_lungcancer/18-spatcellplot_1.png" style="width:80.0%"></p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://drieslab.github.io/GiottoVisuals/reference/spatDimCellPlot.html">spatDimCellPlot</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">visium_lungcancer</span>,</span>
<span>                spat_enr_names <span class="op">=</span> <span class="st">'PAGE'</span>,</span>
<span>                cell_annotation_values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'B_cell'</span>,<span class="st">'Macrophage'</span><span class="op">)</span>,</span>
<span>                cow_n_col <span class="op">=</span> <span class="fl">1</span>, spat_point_size <span class="op">=</span> <span class="fl">1.2</span>,</span>
<span>                plot_alignment <span class="op">=</span> <span class="st">'horizontal'</span>,</span>
<span>                save_param <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>save_name<span class="op">=</span><span class="st">"7_d_spatDimCellPlot"</span>, base_width<span class="op">=</span><span class="fl">7</span>, base_height<span class="op">=</span><span class="fl">10</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/visium_cytassist_lungcancer/19-spatDimCellPlot.png" style="width:80.0%"></p>
</div>
</div>
<div class="section level2">
<h2 id="spatial-grids">7. Spatial Grids<a class="anchor" aria-label="anchor" href="#spatial-grids"></a>
</h2>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">visium_lungcancer</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/createSpatialGrid.html">createSpatialGrid</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">visium_lungcancer</span>,</span>
<span>                                       sdimx_stepsize <span class="op">=</span> <span class="fl">400</span>,</span>
<span>                                       sdimy_stepsize <span class="op">=</span> <span class="fl">400</span>,</span>
<span>                                       minimum_padding <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://drieslab.github.io/GiottoVisuals/reference/spatPlot.html">spatPlot</a></span><span class="op">(</span><span class="va">visium_lungcancer</span>, cell_color <span class="op">=</span> <span class="st">'leiden_clus'</span>, point_size <span class="op">=</span> <span class="fl">2.5</span>, show_grid <span class="op">=</span> <span class="cn">T</span>,</span>
<span>grid_color <span class="op">=</span> <span class="st">'red'</span>, spatial_grid_name <span class="op">=</span> <span class="st">'spatial_grid'</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/visium_cytassist_lungcancer/20-spatPlot2D.png" style="width:50.0%"></p>
</div>
<div class="section level2">
<h2 id="spatial-network">8. Spatial Network<a class="anchor" aria-label="anchor" href="#spatial-network"></a>
</h2>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Delaunay network: stats + creation</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/GiottoVisuals/reference/plotStatDelaunayNetwork.html">plotStatDelaunayNetwork</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">visium_lungcancer</span>, maximum_distance <span class="op">=</span> <span class="fl">400</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/visium_cytassist_lungcancer/21-plotStatDelaunayNetwork.png" style="width:50.0%"></p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">visium_lungcancer</span> <span class="op">=</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/createSpatialNetwork.html">createSpatialNetwork</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">visium_lungcancer</span>, minimum_k <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/showNetworks.html">showNetworks</a></span><span class="op">(</span><span class="va">visium_lungcancer</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/GiottoVisuals/reference/spatPlot.html">spatPlot</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">visium_lungcancer</span>, show_network <span class="op">=</span> <span class="cn">T</span>,</span>
<span>network_color <span class="op">=</span> <span class="st">'blue'</span>, spatial_network_name <span class="op">=</span> <span class="st">'Delaunay_network'</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/visium_cytassist_lungcancer/22-spatPlot2D.png" style="width:50.0%"></p>
</div>
<div class="section level2">
<h2 id="spatial-genes">9. Spatial Genes<a class="anchor" aria-label="anchor" href="#spatial-genes"></a>
</h2>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># kmeans binarization</span></span>
<span><span class="va">kmtest</span> <span class="op">=</span> <span class="fu"><a href="../reference/binSpect.html">binSpect</a></span><span class="op">(</span><span class="va">visium_lungcancer</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/GiottoVisuals/reference/spatFeatPlot2D.html">spatFeatPlot2D</a></span><span class="op">(</span><span class="va">visium_lungcancer</span>, expression_values <span class="op">=</span> <span class="st">'scaled'</span>,</span>
<span>               feats <span class="op">=</span> <span class="va">kmtest</span><span class="op">$</span><span class="va">feats</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span>, cow_n_col <span class="op">=</span> <span class="fl">2</span>, point_size <span class="op">=</span> <span class="fl">1.5</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/visium_cytassist_lungcancer/23-spatFeatPlot2D.png" style="width:50.0%"></p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## rank binarization</span></span>
<span><span class="va">ranktest</span> <span class="op">=</span> <span class="fu"><a href="../reference/binSpect.html">binSpect</a></span><span class="op">(</span><span class="va">visium_lungcancer</span>, bin_method <span class="op">=</span> <span class="st">'rank'</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/GiottoVisuals/reference/spatFeatPlot2D.html">spatFeatPlot2D</a></span><span class="op">(</span><span class="va">visium_lungcancer</span>, expression_values <span class="op">=</span> <span class="st">'scaled'</span>,</span>
<span>               feats <span class="op">=</span> <span class="va">ranktest</span><span class="op">$</span><span class="va">feats</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span>, cow_n_col <span class="op">=</span> <span class="fl">2</span>, point_size <span class="op">=</span> <span class="fl">1.5</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## spatially correlated genes ##</span></span>
<span><span class="va">ext_spatial_genes</span> <span class="op">=</span> <span class="va">kmtest</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">500</span><span class="op">]</span><span class="op">$</span><span class="va">feats</span></span>
<span></span>
<span><span class="co"># 1. calculate gene spatial correlation and single-cell correlation</span></span>
<span><span class="co"># create spatial correlation object</span></span>
<span><span class="va">spat_cor_netw_DT</span> <span class="op">=</span> <span class="fu"><a href="../reference/detectSpatialCorFeats.html">detectSpatialCorFeats</a></span><span class="op">(</span><span class="va">visium_lungcancer</span>,</span>
<span>                                         method <span class="op">=</span> <span class="st">'network'</span>,</span>
<span>                                         spatial_network_name <span class="op">=</span> <span class="st">'Delaunay_network'</span>,</span>
<span>                                         subset_feats <span class="op">=</span> <span class="va">ext_spatial_genes</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 2. identify most similar spatially correlated genes for one gene</span></span>
<span><span class="va">DNAI1_top10_genes</span> <span class="op">=</span> <span class="fu"><a href="../reference/showSpatialCorFeats.html">showSpatialCorFeats</a></span><span class="op">(</span><span class="va">spat_cor_netw_DT</span>, feats <span class="op">=</span> <span class="st">'DNAI1'</span>, show_top_feats <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://drieslab.github.io/GiottoVisuals/reference/spatFeatPlot2D.html">spatFeatPlot2D</a></span><span class="op">(</span><span class="va">visium_lungcancer</span>, expression_values <span class="op">=</span> <span class="st">'scaled'</span>,</span>
<span>               feats <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'RSPH1'</span>, <span class="st">'C20orf85'</span>, <span class="st">'DNAAF1'</span>,<span class="st">'TEKT2'</span><span class="op">)</span>, point_size <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/visium_cytassist_lungcancer/24-spatFeatPlot2D.png" style="width:50.0%"></p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://drieslab.github.io/GiottoVisuals/reference/spatFeatPlot2D.html">spatFeatPlot2D</a></span><span class="op">(</span><span class="va">visium_lungcancer</span>, expression_values <span class="op">=</span> <span class="st">'scaled'</span>,</span>
<span>               feats <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'TEKT2'</span>, <span class="st">'CFAP157'</span>, <span class="st">'MAPK15'</span>, <span class="st">'MS4A8'</span>, <span class="st">'CDHR3'</span>, <span class="st">'C9orf24'</span><span class="op">)</span>, point_size <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/visium_cytassist_lungcancer/25-spatFeatPlot2D.png" style="width:50.0%"></p>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" tabindex="-1"></a><span class="co"># 3. cluster correlated genes &amp; visualize</span></span>
<span id="cb32-2"><a href="#cb32-2" tabindex="-1"></a>spat_cor_netw_DT <span class="ot">=</span> <span class="fu">clusterSpatialCorFeats</span>(spat_cor_netw_DT, <span class="at">name =</span> â€˜spat_netw_clusâ€™, <span class="at">k =</span> <span class="dv">10</span>)</span>
<span id="cb32-3"><a href="#cb32-3" tabindex="-1"></a></span>
<span id="cb32-4"><a href="#cb32-4" tabindex="-1"></a><span class="fu">heatmSpatialCorFeats</span>(visium_lungcancer, <span class="at">spatCorObject =</span> spat_cor_netw_DT, <span class="at">use_clus_name =</span> â€˜spat_netw_clusâ€™,</span>
<span id="cb32-5"><a href="#cb32-5" tabindex="-1"></a><span class="at">save_param =</span> <span class="fu">c</span>(<span class="at">save_name =</span> â€˜<span class="dv">22</span><span class="sc">-</span>z1<span class="sc">-</span>heatmap_correlated_genesâ€™, <span class="at">save_format =</span> â€˜pdfâ€™, <span class="at">base_height =</span> <span class="dv">6</span>, <span class="at">base_width =</span> <span class="dv">8</span>, <span class="at">units =</span> â€˜cmâ€™), <span class="at">heatmap_legend_param =</span> <span class="fu">list</span>(<span class="at">title =</span> <span class="cn">NULL</span>))</span></code></pre></div>
<p><img src="images/visium_cytassist_lungcancer/26-heatmap_correlated_genes.png" style="width:50.0%"></p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># 4. rank spatial correlated clusters and show genes for selected clusters</span></span>
<span><span class="va">netw_ranks</span> <span class="op">=</span> <span class="fu"><a href="../reference/rankSpatialCorGroups.html">rankSpatialCorGroups</a></span><span class="op">(</span><span class="va">visium_lungcancer</span>,</span>
<span>                                  spatCorObject <span class="op">=</span> <span class="va">spat_cor_netw_DT</span>,</span>
<span>                                  use_clus_name <span class="op">=</span> <span class="st">'spat_netw_clus'</span>,</span>
<span>                                  save_param <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>save_name <span class="op">=</span> <span class="st">'22-z2-rank_correlated_groups'</span>,</span>
<span>                                                 base_height <span class="op">=</span> <span class="fl">3</span>, base_width <span class="op">=</span> <span class="fl">5</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">top_netw_spat_cluster</span> <span class="op">=</span> <span class="fu"><a href="../reference/showSpatialCorFeats.html">showSpatialCorFeats</a></span><span class="op">(</span><span class="va">spat_cor_netw_DT</span>, use_clus_name <span class="op">=</span> <span class="st">'spat_netw_clus'</span>,</span>
<span>                                            selected_clusters <span class="op">=</span> <span class="fl">6</span>, show_top_feats <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/visium_cytassist_lungcancer/27-rank_correlated_groups.png" style="width:50.0%"></p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># 5. create metagene enrichment score for clusters</span></span>
<span><span class="va">cluster_genes_DT</span> <span class="op">=</span> <span class="fu"><a href="../reference/showSpatialCorFeats.html">showSpatialCorFeats</a></span><span class="op">(</span><span class="va">spat_cor_netw_DT</span>, use_clus_name <span class="op">=</span> <span class="st">'spat_netw_clus'</span>, show_top_feats <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">cluster_genes</span> <span class="op">=</span> <span class="va">cluster_genes_DT</span><span class="op">$</span><span class="va">clus</span>; <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">cluster_genes</span><span class="op">)</span> <span class="op">=</span> <span class="va">cluster_genes_DT</span><span class="op">$</span><span class="va">feat_ID</span></span>
<span></span>
<span><span class="va">visium_lungcancer</span> <span class="op">=</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/createMetafeats.html">createMetafeats</a></span><span class="op">(</span><span class="va">visium_lungcancer</span>, feat_clusters <span class="op">=</span> <span class="va">cluster_genes</span>, name <span class="op">=</span> <span class="st">'cluster_metagene'</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/showGiottoSpatEnrichments.html">showGiottoSpatEnrichments</a></span><span class="op">(</span><span class="va">visium_lungcancer</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://drieslab.github.io/GiottoVisuals/reference/spatCellPlot.html">spatCellPlot</a></span><span class="op">(</span><span class="va">visium_lungcancer</span>,</span>
<span>             spat_enr_names <span class="op">=</span> <span class="st">'cluster_metagene'</span>,</span>
<span>             cell_annotation_values <span class="op">=</span> <span class="va">netw_ranks</span><span class="op">$</span><span class="va">clusters</span>,</span>
<span>             point_size <span class="op">=</span> <span class="fl">1.5</span>, cow_n_col <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/visium_cytassist_lungcancer/28-spatCellPlot2D.png" style="width:80.0%"></p>
</div>
<div class="section level2">
<h2 id="hmrf-domains">10. HMRF Domains<a class="anchor" aria-label="anchor" href="#hmrf-domains"></a>
</h2>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># HMRF requires a fully connected network!</span></span>
<span><span class="va">visium_lungcancer</span> <span class="op">=</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/createSpatialNetwork.html">createSpatialNetwork</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">visium_lungcancer</span>, minimum_k <span class="op">=</span> <span class="fl">2</span>, name <span class="op">=</span> <span class="st">'Delaunay_full'</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># spatial genes</span></span>
<span><span class="va">my_spatial_genes</span> <span class="op">&lt;-</span> <span class="va">kmtest</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">100</span><span class="op">]</span><span class="op">$</span><span class="va">feats</span></span>
<span></span>
<span><span class="co"># do HMRF with different betas</span></span>
<span><span class="va">hmrf_folder</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">results_folder</span>,<span class="st">'/'</span>,<span class="st">'HMRF_results/'</span><span class="op">)</span></span>
<span><span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.exists</a></span><span class="op">(</span><span class="va">hmrf_folder</span><span class="op">)</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="va">hmrf_folder</span>, recursive <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># if Rscript is not found, you might have to create a symbolic link, e.g.</span></span>
<span><span class="co"># cd /usr/local/bin</span></span>
<span><span class="co"># sudo ln -s /Library/Frameworks/R.framework/Resources/Rscript Rscript</span></span>
<span><span class="va">HMRF_spatial_genes</span> <span class="op">=</span> <span class="fu"><a href="../reference/doHMRF.html">doHMRF</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">visium_lungcancer</span>,</span>
<span>                            expression_values <span class="op">=</span> <span class="st">'scaled'</span>,</span>
<span>                            spatial_network_name <span class="op">=</span> <span class="st">'Delaunay_full'</span>,</span>
<span>                            spatial_genes <span class="op">=</span> <span class="va">my_spatial_genes</span>,</span>
<span>                            k <span class="op">=</span> <span class="fl">5</span>,</span>
<span>                            betas <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">10</span>, <span class="fl">3</span><span class="op">)</span>,</span>
<span>                            output_folder <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">hmrf_folder</span>, <span class="st">'/'</span>, <span class="st">'Spatial_genes/SG_topgenes_k5_scaled'</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## alternative way to view HMRF results</span></span>
<span><span class="co"># results = writeHMRFresults(gobject = ST_test,</span></span>
<span><span class="co">#                            HMRFoutput = HMRF_spatial_genes,</span></span>
<span><span class="co">#                            k = 5, betas_to_view = seq(0, 25, by = 5))</span></span>
<span><span class="co"># ST_test = addCellMetadata(ST_test, new_metadata = results, by_column = T, column_cell_ID = 'cell_ID')</span></span>
<span></span>
<span><span class="co">## add HMRF of interest to giotto object</span></span>
<span><span class="va">visium_lungcancer</span> <span class="op">=</span> <span class="fu"><a href="../reference/addHMRF.html">addHMRF</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">visium_lungcancer</span>,</span>
<span>                            HMRFoutput <span class="op">=</span> <span class="va">HMRF_spatial_genes</span>,</span>
<span>                            k <span class="op">=</span> <span class="fl">5</span>, betas_to_add <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">10</span>,<span class="fl">20</span><span class="op">)</span>,</span>
<span>                            hmrf_name <span class="op">=</span> <span class="st">'HMRF'</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/showGiottoSpatEnrichments.html">showGiottoSpatEnrichments</a></span><span class="op">(</span><span class="va">visium_lungcancer</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## visualize</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/GiottoVisuals/reference/spatPlot.html">spatPlot</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">visium_lungcancer</span>, cell_color <span class="op">=</span> <span class="st">'HMRF_k5_b.0'</span>, point_size <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/visium_cytassist_lungcancer/29-spatPlot2D.png" style="width:50.0%"></p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://drieslab.github.io/GiottoVisuals/reference/spatPlot.html">spatPlot</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">visium_lungcancer</span>, cell_color <span class="op">=</span> <span class="st">'HMRF_k5_b.10'</span>, point_size <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/visium_cytassist_lungcancer/30-spatPlot2D.png" style="width:50.0%"></p>
<div class="sourceCode" id="cb37"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span>
<span id="cb37-2"><a href="#cb37-2" tabindex="-1"></a></span>
<span id="cb37-3"><a href="#cb37-3" tabindex="-1"></a>R version <span class="dv">4</span>.<span class="fl">2.2</span> (<span class="dv">2022-10-31</span>)</span>
<span id="cb37-4"><a href="#cb37-4" tabindex="-1"></a>Platform<span class="sc">:</span> aarch64<span class="sc">-</span>apple<span class="sc">-</span><span class="fu">darwin20</span> (<span class="dv">64</span><span class="sc">-</span>bit)</span>
<span id="cb37-5"><a href="#cb37-5" tabindex="-1"></a>Running under<span class="sc">:</span> macOS Monterey <span class="dv">12</span>.<span class="fl">5.1</span></span>
<span id="cb37-6"><a href="#cb37-6" tabindex="-1"></a></span>
<span id="cb37-7"><a href="#cb37-7" tabindex="-1"></a>Matrix products<span class="sc">:</span> default</span>
<span id="cb37-8"><a href="#cb37-8" tabindex="-1"></a>LAPACK<span class="sc">:</span> <span class="er">/</span>Library<span class="sc">/</span>Frameworks<span class="sc">/</span>R.framework<span class="sc">/</span>Versions<span class="sc">/</span><span class="fl">4.2</span><span class="sc">-</span>arm64<span class="sc">/</span>Resources<span class="sc">/</span>lib<span class="sc">/</span>libRlapack.dylib</span>
<span id="cb37-9"><a href="#cb37-9" tabindex="-1"></a></span>
<span id="cb37-10"><a href="#cb37-10" tabindex="-1"></a>locale<span class="sc">:</span></span>
<span id="cb37-11"><a href="#cb37-11" tabindex="-1"></a>[<span class="dv">1</span>] en_US.UTF<span class="dv">-8</span><span class="sc">/</span>en_US.UTF<span class="dv">-8</span><span class="sc">/</span>en_US.UTF<span class="dv">-8</span><span class="sc">/</span>C<span class="sc">/</span>en_US.UTF<span class="dv">-8</span><span class="sc">/</span>en_US.UTF<span class="dv">-8</span></span>
<span id="cb37-12"><a href="#cb37-12" tabindex="-1"></a></span>
<span id="cb37-13"><a href="#cb37-13" tabindex="-1"></a>attached base packages<span class="sc">:</span></span>
<span id="cb37-14"><a href="#cb37-14" tabindex="-1"></a>[<span class="dv">1</span>] stats     graphics  grDevices utils     datasets  methods   base</span>
<span id="cb37-15"><a href="#cb37-15" tabindex="-1"></a></span>
<span id="cb37-16"><a href="#cb37-16" tabindex="-1"></a>other attached packages<span class="sc">:</span></span>
<span id="cb37-17"><a href="#cb37-17" tabindex="-1"></a>[<span class="dv">1</span>] Giotto_3.<span class="fl">0.1</span></span>
<span id="cb37-18"><a href="#cb37-18" tabindex="-1"></a></span>
<span id="cb37-19"><a href="#cb37-19" tabindex="-1"></a>loaded via a <span class="fu">namespace</span> (and not attached)<span class="sc">:</span></span>
<span id="cb37-20"><a href="#cb37-20" tabindex="-1"></a> [<span class="dv">1</span>] bitops_1<span class="fl">.0</span><span class="dv">-7</span>                matrixStats_0.<span class="fl">63.0</span>          fs_1.<span class="fl">5.2</span></span>
<span id="cb37-21"><a href="#cb37-21" tabindex="-1"></a> [<span class="dv">4</span>] usethis_2.<span class="fl">1.6</span>               devtools_2.<span class="fl">4.5</span>              GenomeInfoDb_1.<span class="fl">34.3</span></span>
<span id="cb37-22"><a href="#cb37-22" tabindex="-1"></a> [<span class="dv">7</span>] tools_4.<span class="fl">2.2</span>                 profvis_0.<span class="fl">3.7</span>               utf8_1.<span class="fl">2.2</span></span>
<span id="cb37-23"><a href="#cb37-23" tabindex="-1"></a>[<span class="dv">10</span>] R6_2.<span class="fl">5.1</span>                    irlba_2.<span class="dv">3</span>.<span class="fl">5.1</span>               BiocGenerics_0.<span class="fl">44.0</span></span>
<span id="cb37-24"><a href="#cb37-24" tabindex="-1"></a>[<span class="dv">13</span>] colorspace_2<span class="fl">.0</span><span class="dv">-3</span>            urlchecker_1.<span class="fl">0.1</span>            tidyselect_1.<span class="fl">2.0</span></span>
<span id="cb37-25"><a href="#cb37-25" tabindex="-1"></a>[<span class="dv">16</span>] prettyunits_1.<span class="fl">1.1</span>           processx_3.<span class="fl">8.0</span>              compiler_4.<span class="fl">2.2</span></span>
<span id="cb37-26"><a href="#cb37-26" tabindex="-1"></a>[<span class="dv">19</span>] cli_3.<span class="fl">4.1</span>                   Biobase_2.<span class="fl">58.0</span>              BiocNeighbors_1.<span class="fl">16.0</span></span>
<span id="cb37-27"><a href="#cb37-27" tabindex="-1"></a>[<span class="dv">22</span>] DelayedArray_0.<span class="fl">24.0</span>         scales_1.<span class="fl">2.1</span>                callr_3.<span class="fl">7.3</span></span>
<span id="cb37-28"><a href="#cb37-28" tabindex="-1"></a>[<span class="dv">25</span>] stringr_1.<span class="fl">4.1</span>               digest_0.<span class="fl">6.30</span>               XVector_0.<span class="fl">38.0</span></span>
<span id="cb37-29"><a href="#cb37-29" tabindex="-1"></a>[<span class="dv">28</span>] pkgconfig_2.<span class="fl">0.3</span>             htmltools_0.<span class="fl">5.3</span>             sessioninfo_1.<span class="fl">2.2</span></span>
<span id="cb37-30"><a href="#cb37-30" tabindex="-1"></a>[<span class="dv">31</span>] sparseMatrixStats_1.<span class="fl">10.0</span>    MatrixGenerics_1.<span class="fl">10.0</span>       fastmap_1.<span class="fl">1.0</span></span>
<span id="cb37-31"><a href="#cb37-31" tabindex="-1"></a>[<span class="dv">34</span>] limma_3.<span class="fl">54.0</span>                htmlwidgets_1.<span class="fl">5.4</span>           rlang_1.<span class="fl">0.6</span></span>
<span id="cb37-32"><a href="#cb37-32" tabindex="-1"></a>[<span class="dv">37</span>] rstudioapi_0<span class="fl">.14</span>             shiny_1.<span class="fl">7.3</span>                 DelayedMatrixStats_1.<span class="fl">20.0</span></span>
<span id="cb37-33"><a href="#cb37-33" tabindex="-1"></a>[<span class="dv">40</span>] generics_0.<span class="fl">1.3</span>              BiocParallel_1.<span class="fl">32.1</span>         dplyr_1.<span class="fl">0.10</span></span>
<span id="cb37-34"><a href="#cb37-34" tabindex="-1"></a>[<span class="dv">43</span>] RCurl_1<span class="fl">.98-1.9</span>              magrittr_2.<span class="fl">0.3</span>              BiocSingular_1.<span class="fl">14.0</span></span>
<span id="cb37-35"><a href="#cb37-35" tabindex="-1"></a>[<span class="dv">46</span>] GenomeInfoDbData_1.<span class="fl">2.9</span>      Matrix_1<span class="fl">.5</span><span class="dv">-3</span>                Rcpp_1.<span class="fl">0.9</span></span>
<span id="cb37-36"><a href="#cb37-36" tabindex="-1"></a>[<span class="dv">49</span>] munsell_0.<span class="fl">5.0</span>               S4Vectors_0.<span class="fl">36.0</span>            fansi_1.<span class="fl">0.3</span></span>
<span id="cb37-37"><a href="#cb37-37" tabindex="-1"></a>[<span class="dv">52</span>] lifecycle_1.<span class="fl">0.3</span>             stringi_1.<span class="fl">7.8</span>               SummarizedExperiment_1.<span class="fl">28.0</span></span>
<span id="cb37-38"><a href="#cb37-38" tabindex="-1"></a>[<span class="dv">55</span>] zlibbioc_1.<span class="fl">44.0</span>             pkgbuild_1.<span class="fl">3.1</span>              grid_4.<span class="fl">2.2</span></span>
<span id="cb37-39"><a href="#cb37-39" tabindex="-1"></a>[<span class="dv">58</span>] parallel_4.<span class="fl">2.2</span>              promises_1.<span class="dv">2</span>.<span class="fl">0.1</span>            dqrng_0.<span class="fl">3.0</span></span>
<span id="cb37-40"><a href="#cb37-40" tabindex="-1"></a>[<span class="dv">61</span>] crayon_1.<span class="fl">5.2</span>                miniUI_0.<span class="dv">1</span>.<span class="fl">1.1</span>              lattice_0<span class="fl">.20</span><span class="dv">-45</span></span>
<span id="cb37-41"><a href="#cb37-41" tabindex="-1"></a>[<span class="dv">64</span>] beachmat_2.<span class="fl">14.0</span>             locfit_1<span class="fl">.5-9.6</span>              ps_1.<span class="fl">7.2</span></span>
<span id="cb37-42"><a href="#cb37-42" tabindex="-1"></a>[<span class="dv">67</span>] pillar_1.<span class="fl">8.1</span>                igraph_1.<span class="fl">3.5</span>                GenomicRanges_1.<span class="fl">50.1</span></span>
<span id="cb37-43"><a href="#cb37-43" tabindex="-1"></a>[<span class="dv">70</span>] codetools_0<span class="fl">.2</span><span class="dv">-18</span>            ScaledMatrix_1.<span class="fl">6.0</span>          stats4_4.<span class="fl">2.2</span></span>
<span id="cb37-44"><a href="#cb37-44" tabindex="-1"></a>[<span class="dv">73</span>] pkgload_1.<span class="fl">3.2</span>               glue_1.<span class="fl">6.2</span>                  data.table_1.<span class="fl">14.6</span></span>
<span id="cb37-45"><a href="#cb37-45" tabindex="-1"></a>[<span class="dv">76</span>] remotes_2.<span class="fl">4.2</span>               BiocManager_1.<span class="fl">30.19</span>         vctrs_0.<span class="fl">5.1</span></span>
<span id="cb37-46"><a href="#cb37-46" tabindex="-1"></a>[<span class="dv">79</span>] httpuv_1.<span class="fl">6.6</span>                gtable_0.<span class="fl">3.1</span>                purrr_0.<span class="fl">3.5</span></span>
<span id="cb37-47"><a href="#cb37-47" tabindex="-1"></a>[<span class="dv">82</span>] cachem_1.<span class="fl">0.6</span>                ggplot2_3.<span class="fl">4.0</span>               rsvd_1.<span class="fl">0.5</span></span>
<span id="cb37-48"><a href="#cb37-48" tabindex="-1"></a>[<span class="dv">85</span>] mime_0<span class="fl">.12</span>                   xtable_1<span class="fl">.8</span><span class="dv">-4</span>                later_1.<span class="fl">3.0</span></span>
<span id="cb37-49"><a href="#cb37-49" tabindex="-1"></a>[<span class="dv">88</span>] SingleCellExperiment_1.<span class="fl">20.0</span> tibble_3.<span class="fl">1.8</span>                memoise_2.<span class="fl">0.1</span></span>
<span id="cb37-50"><a href="#cb37-50" tabindex="-1"></a>[<span class="dv">91</span>] IRanges_2.<span class="fl">32.0</span>              cluster_2.<span class="fl">1.4</span>               bluster_1.<span class="fl">8.0</span></span>
<span id="cb37-51"><a href="#cb37-51" tabindex="-1"></a>[<span class="dv">94</span>] ellipsis_0.<span class="fl">3.2</span></span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by <a href="https://www.drieslab.com" class="external-link">Ruben Dries</a>, Jiaji George Chen, <a href="https://josschavezf.github.io" class="external-link">Joselyn C. ChÃ¡vez-Fuentes</a>, <a href="https://labs.icahn.mssm.edu/yuanlab/" class="external-link">Guo-Cheng Yuan</a>, Matthew Oâ€™Brien, Edward Ruiz, Wen Wang, Natalie Del Rossi.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
