<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Vizgen Mouse Brain Receptor Map â€¢ Giotto</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Vizgen Mouse Brain Receptor Map">
<meta property="og:description" content="Giotto">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">Giotto</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">4.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Get started
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/installation.html">Giotto Installation</a>
    </li>
    <li>
      <a href="../articles/ecosystem.html">Giotto Ecosystem</a>
    </li>
    <li>
      <a href="../articles/configuration.html">Giotto Configuration</a>
    </li>
    <li>
      <a href="../articles/object_creation.html">Giotto Object Creation</a>
    </li>
    <li>
      <a href="../articles/structure.html">Giotto Class Structure</a>
    </li>
    <li>
      <a href="../articles/structure.html">Giotto Class Structure</a>
    </li>
    <li>
      <a href="../articles/errorsfaqsandtips.html">Solutions to Installation Errors</a>
    </li>
    <li>
      <a href="../articles/faqs.html">FAQs</a>
    </li>
    <li>
      <a href="../articles/branches.html">Branches</a>
    </li>
    <li>
      <a href="../articles/github_issues.html">How to report a bug</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Documentation</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Examples
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Seq-Scope</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/seqscope_mouse_liver.html">Mouse liver</a>
        </li>
      </ul>
</li>
    <li class="divider">
    </li>
<li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Visium</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/visium_mouse_kidney.html">Visium Mouse Kidney</a>
        </li>
        <li>
          <a href="../articles/visium_mouse_brain.html">Visium Mouse Brain</a>
        </li>
        <li>
          <a href="../articles/visium_prostate_integration.html">Visium Prostate Integration</a>
        </li>
        <li>
          <a href="../articles/visium_cytassist_lungcancer.html">CytAssist Human Lung Cancer</a>
        </li>
        <li>
          <a href="../articles/visium_cytassist_human_glioblastoma.html">CytAssist Multi-omics Human Glioblastoma</a>
        </li>
      </ul>
</li>
    <li class="divider">
    </li>
<li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Spatial Genomics</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/spatial_genomics_mouse_kidney.html">Mouse Kidney</a>
        </li>
      </ul>
</li>
    <li class="divider">
    </li>
<li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Xenium</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/xenium_breast_cancer.html">Xenium Human breast cancer</a>
        </li>
      </ul>
</li>
    <li class="divider">
    </li>
<li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Nanostring CosMx</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/nanostring_cosmx_lung_cancer.html">Nanostring CosMx Subcellular Lung Cancer</a>
        </li>
      </ul>
</li>
    <li class="divider">
    </li>
<li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">MERFISH</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/merfish_mouse_hypothalamic.html">MERFISH Mouse Hypothalamic Preoptic Region</a>
        </li>
      </ul>
</li>
    <li class="divider">
    </li>
<li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Vizgen</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/vizgen_mouse_brain.html">Vizgen Mouse Brain Receptor Map</a>
        </li>
      </ul>
</li>
    <li class="divider">
    </li>
<li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">seqFISH</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/mini_seqfish.html">mini seqFISH</a>
        </li>
        <li>
          <a href="../articles/seqfish_cortex.html">SeqFish+ Mouse Cortex Example</a>
        </li>
      </ul>
</li>
    <li class="divider">
    </li>
<li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">osmFISH</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/osmfish_mouse_ss_cortex.html">osmFISH Mouse SS Cortex</a>
        </li>
      </ul>
</li>
    <li class="divider">
    </li>
<li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Resolve Bioscience</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/resolve_biosciences.html">Resolve Bioscience Breast Cancer Subcellular</a>
        </li>
      </ul>
</li>
    <li class="divider">
    </li>
<li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">CODEX</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/codex_mouse_spleen.html">Codex Mouse Spleen</a>
        </li>
      </ul>
</li>
    <li class="divider">
    </li>
<li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Single Cell RNA Sequencing</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/singlecell_prostate_standard.html">Prostate</a>
        </li>
        <li>
          <a href="../articles/singlecell_prostate_integration.html">Prostate Integration</a>
        </li>
      </ul>
</li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">Spatial manipulation</li>
    <li>
      <a href="../articles/spatial_transformations.html">Spatial transformations</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Image alignment</li>
    <li>
      <a href="../articles/image_alignment.html">Image alignment</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Interactive selection</li>
    <li>
      <a href="../articles/interactive_selection.html">Interactive selection</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Visualizations</li>
    <li>
      <a href="../articles/visualizations.html">Visualizations</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Saving and Loading</li>
    <li>
      <a href="../articles/saving_object.html">Saving a Giotto object</a>
    </li>
    <li>
      <a href="../articles/saving_plots.html">Saving plots</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Interoperability</li>
    <li>
      <a href="../articles/interoperability_anndata.html">AnnData</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Try Giotto in the Cloud</li>
    <li>
      <a href="../articles/terra.html">Terra.bio</a>
    </li>
    <li>
      <a href="../articles/docker.html">Docker</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    News
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../news/index.html">Changes</a>
    </li>
    <li>
      <a href="../articles/presentations.html">Presentations</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/drieslab/Giotto/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Vizgen Mouse Brain Receptor Map</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/drieslab/Giotto/blob/HEAD/vignettes/vizgen_mouse_brain.Rmd" class="external-link"><code>vignettes/vizgen_mouse_brain.Rmd</code></a></small>
      <div class="hidden name"><code>vizgen_mouse_brain.Rmd</code></div>

    </div>

    
    
<p><em>Giotto</em> version <code>2.0.0.998</code>. Please check the
version you are using to get the same results.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Ensure Giotto Suite is installed.</span></span>
<span><span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="st">"Giotto"</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/installed.packages.html" class="external-link">installed.packages</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"drieslab/Giotto@suite"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co"># Ensure GiottoData, a small, helper module for tutorials, is installed.</span></span>
<span><span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="st">"GiottoData"</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/installed.packages.html" class="external-link">installed.packages</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"drieslab/GiottoData"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://drieslab.github.io/Giotto/">Giotto</a></span><span class="op">)</span></span>
<span><span class="co"># Ensure the Python environment for Giotto has been installed.</span></span>
<span><span class="va">genv_exists</span> <span class="op">=</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/checkGiottoEnvironment.html">checkGiottoEnvironment</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="va">genv_exists</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="co"># The following command need only be run once to install the Giotto environment.</span></span>
<span>  <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/installGiottoEnvironment.html">installGiottoEnvironment</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<div class="section level2">
<h2 id="set-up-giotto">1. Set up Giotto<a class="anchor" aria-label="anchor" href="#set-up-giotto"></a>
</h2>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://drieslab.github.io/Giotto/">Giotto</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://drieslab.github.io/GiottoData/">GiottoData</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 1. set working directory where project outputs will be saved to</span></span>
<span><span class="va">results_folder</span> <span class="op">=</span> <span class="st">'/path/to/save/directory/'</span></span>
<span></span>
<span><span class="co"># Optional: Specify a path to a Python executable within a conda or miniconda </span></span>
<span><span class="co"># environment. If set to NULL (default), the Python executable within the previously</span></span>
<span><span class="co"># installed Giotto environment will be used.</span></span>
<span><span class="va">my_python_path</span> <span class="op">=</span> <span class="cn">NULL</span> <span class="co"># alternatively, "/local/python/path/python" if desired.</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="dataset-explanation">2. Dataset explanation<a class="anchor" aria-label="anchor" href="#dataset-explanation"></a>
</h2>
<p>This vignette covers Giotto object creation and simple exploratory
analysis with the subcellular <a href="https://vizgen.com/applications/neuroscience-showcase/" class="external-link">MERFISH
Mouse Brain Receptor Map dataset</a> provided by Vizgen with their
MERSCOPE platform.</p>
<p>Transcripts are captured at the single-molecule level with
subcellular spatial resolution (â‰¤100nm). This dataset includes
information from 9 full coronal mouse brain slices (3 positions with 3
biological replicates) that were profiled for 483 genes. This vignette
works with slice 1, replicate 1.</p>
<p><strong>Provided Outputs:</strong></p>
<ul>
<li>List of all <strong>detected transcripts</strong> and their spatial
locations in three dimensions (CSV)</li>
</ul>
<p>Show first 4 rows</p>
<table class="table">
<thead><tr>
<th style="text-align:right;">
V1
</th>
<th style="text-align:right;">
barcode_id
</th>
<th style="text-align:right;">
global_x
</th>
<th style="text-align:right;">
global_y
</th>
<th style="text-align:right;">
global_z
</th>
<th style="text-align:right;">
x
</th>
<th style="text-align:right;">
y
</th>
<th style="text-align:right;">
fov
</th>
<th style="text-align:left;">
gene
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:right;">
159.9778
</td>
<td style="text-align:right;">
4208.468
</td>
<td style="text-align:right;">
4
</td>
<td style="text-align:right;">
1762.116
</td>
<td style="text-align:right;">
159.9814
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:left;">
Htr1b
</td>
</tr>
<tr>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:right;">
165.9403
</td>
<td style="text-align:right;">
4321.805
</td>
<td style="text-align:right;">
4
</td>
<td style="text-align:right;">
1817.324
</td>
<td style="text-align:right;">
1209.3922
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:left;">
Htr1b
</td>
</tr>
<tr>
<td style="text-align:right;">
2
</td>
<td style="text-align:right;">
11
</td>
<td style="text-align:right;">
158.4767
</td>
<td style="text-align:right;">
4320.901
</td>
<td style="text-align:right;">
5
</td>
<td style="text-align:right;">
1748.217
</td>
<td style="text-align:right;">
1201.0283
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:left;">
Htr6
</td>
</tr>
<tr>
<td style="text-align:right;">
3
</td>
<td style="text-align:right;">
13
</td>
<td style="text-align:right;">
171.2179
</td>
<td style="text-align:right;">
4283.950
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
1866.191
</td>
<td style="text-align:right;">
858.8855
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:left;">
Adora1
</td>
</tr>
</tbody>
</table>
<ul>
<li><p>Output from the cell segmentation analysis:</p></li>
<li><p><strong>transcripts</strong> (cols) per <strong>cell</strong>
(rows) aggregated count matrix (CSV)</p></li>
</ul>
<p>Show first 4 rows and columns</p>
<table class="table">
<thead><tr>
<th style="text-align:right;">
V1
</th>
<th style="text-align:right;">
Oxgr1
</th>
<th style="text-align:right;">
Htr1a
</th>
<th style="text-align:right;">
Htr1b
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:right;">
1.108834e+38
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:right;">
1.351882e+38
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:right;">
1.647670e+38
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:right;">
1.657479e+38
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
1
</td>
</tr>
</tbody>
</table>
<ul>
<li>
<strong>cell metadata</strong> (CSV)</li>
</ul>
<p>Show first 4 rows</p>
<table class="table">
<thead><tr>
<th style="text-align:right;">
V1
</th>
<th style="text-align:right;">
fov
</th>
<th style="text-align:right;">
volume
</th>
<th style="text-align:right;">
center_x
</th>
<th style="text-align:right;">
center_y
</th>
<th style="text-align:right;">
min_x
</th>
<th style="text-align:right;">
max_x
</th>
<th style="text-align:right;">
min_y
</th>
<th style="text-align:right;">
max_y
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:right;">
1.108834e+38
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
432.1414
</td>
<td style="text-align:right;">
156.5633
</td>
<td style="text-align:right;">
4271.326
</td>
<td style="text-align:right;">
151.5305
</td>
<td style="text-align:right;">
161.5961
</td>
<td style="text-align:right;">
4264.620
</td>
<td style="text-align:right;">
4278.033
</td>
</tr>
<tr>
<td style="text-align:right;">
1.351882e+38
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
1351.8026
</td>
<td style="text-align:right;">
156.5093
</td>
<td style="text-align:right;">
4256.962
</td>
<td style="text-align:right;">
148.2905
</td>
<td style="text-align:right;">
164.7281
</td>
<td style="text-align:right;">
4247.664
</td>
<td style="text-align:right;">
4266.261
</td>
</tr>
<tr>
<td style="text-align:right;">
1.647670e+38
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
1080.6533
</td>
<td style="text-align:right;">
159.9653
</td>
<td style="text-align:right;">
4228.180
</td>
<td style="text-align:right;">
152.1785
</td>
<td style="text-align:right;">
167.7521
</td>
<td style="text-align:right;">
4220.556
</td>
<td style="text-align:right;">
4235.805
</td>
</tr>
<tr>
<td style="text-align:right;">
1.657479e+38
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
1652.0007
</td>
<td style="text-align:right;">
167.5793
</td>
<td style="text-align:right;">
4323.868
</td>
<td style="text-align:right;">
158.2265
</td>
<td style="text-align:right;">
176.9321
</td>
<td style="text-align:right;">
4314.192
</td>
<td style="text-align:right;">
4333.545
</td>
</tr>
</tbody>
</table>
<ul>
<li><p><strong>cell boundaries</strong> (HDF5)</p></li>
<li><p>The <strong>DAPI</strong> and <strong>Poly T mosaic
images</strong> (TIFF)</p></li>
</ul>
<p><code>Vizgen Data Release V1.0. May 2021</code></p>
</div>
<div class="section level2">
<h2 id="giotto-global-instructions-and-preparations">3. Giotto global instructions and preparations<a class="anchor" aria-label="anchor" href="#giotto-global-instructions-and-preparations"></a>
</h2>
<p>Define plot saving behavior and project data paths</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Directly saving plots to the working directory without rendering them in the editor saves time.</span></span>
<span><span class="va">instrs</span> <span class="op">=</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/createGiottoInstructions.html">createGiottoInstructions</a></span><span class="op">(</span>save_dir <span class="op">=</span> <span class="va">results_folder</span>,</span>
<span>                                  save_plot <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                                  show_plot <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                                  return_plot <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                                  python_path <span class="op">=</span> <span class="va">my_python_path</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Add Needed paths below:</span></span>
<span><span class="co"># provide path to pre-aggregated information</span></span>
<span><span class="va">expr_path</span> <span class="op">=</span> <span class="st">'/path/to/datasets_mouse_brain_map_BrainReceptorShowcase_Slice1_Replicate1_cell_by_gene_S1R1.csv'</span></span>
<span><span class="co"># provide path to metadata (includes spatial locations of aggregated expression)</span></span>
<span><span class="va">meta_path</span> <span class="op">=</span> <span class="st">'/path/to/datasets_mouse_brain_map_BrainReceptorShowcase_Slice1_Replicate1_cell_metadata_S1R1.csv'</span></span>
<span><span class="co"># provide path to the detected transcripts (single molecule level transcript spatial information)</span></span>
<span><span class="va">tx_path</span> <span class="op">=</span> <span class="st">'/path/to/datasets_mouse_brain_map_BrainReceptorShowcase_Slice1_Replicate1_detected_transcripts_S1R1.csv'</span></span>
<span><span class="co"># define path to cell boundaries folder</span></span>
<span><span class="va">bound_path</span> <span class="op">=</span> <span class="st">'/path/to/cell_boundaries'</span></span>
<span><span class="co"># path to image scale conversion values</span></span>
<span><span class="va">img_scale_path</span> <span class="op">=</span> <span class="st">'path/to/micron_to_mosaic_pixel_transform.csv'</span></span>
<span><span class="co"># provide path to the dapi image of slice 1 replicate 1</span></span>
<span><span class="va">img_path</span> <span class="op">=</span> <span class="st">'path/to/mosaic_DAPI_z0.tif'</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="create-giotto-object-from-aggregated-data">4. Create Giotto object from aggregated data<a class="anchor" aria-label="anchor" href="#create-giotto-object-from-aggregated-data"></a>
</h2>
<p>Vizgen provides a cell by transcript output matrix
(<code>cell_by_gene.csv</code>) with the subcellular spatial transcript
information already aggregated by the provided polygon cell annotations
into a count matrix.</p>
<p>Along with the count matrix, metadata information about the field of
view (FOV), spatial location, and volume of the cell (annotation
polygons) is also provided through the
<code>cell_metadata.csv</code>.</p>
<p><strong>Pre-aggregated</strong> information can be loaded into Giotto
with the usual generic <code><a href="https://drieslab.github.io/GiottoClass/reference/createGiottoObject.html">createGiottoObject()</a></code> function. For
starting from the raw subcellular information, skip to Section 10. To
create the Giotto object, the <code>cell_by_gene</code> expression
matrix and the <code>cell_metadata</code> information are first read
into R. Since Giotto accepts the expression information with features
(in this case genes/transcript counts) as rows and cells as columns, the
expression matrix must first be transposed to create the object.</p>
<p>Additionally for this dataset, y values should be inverted when
loaded to match the included images. For more information consult ths
standard workflow at getting_started_images section.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># read expression matrix and metadata</span></span>
<span><span class="va">expr_matrix</span> <span class="op">=</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/readExprMatrix.html">readExprMatrix</a></span><span class="op">(</span><span class="va">expr_path</span><span class="op">)</span></span>
<span><span class="va">meta_dt</span> <span class="op">=</span> <span class="fu">data.table</span><span class="fu">::</span><span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="va">meta_path</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># create giotto object</span></span>
<span><span class="va">vizgen</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/createGiottoObject.html">createGiottoObject</a></span><span class="op">(</span>expression <span class="op">=</span> <span class="fu">Giottot_flex</span><span class="op">(</span><span class="va">expr_matrix</span><span class="op">)</span>,</span>
<span>                             spatial_locs <span class="op">=</span> <span class="va">meta_dt</span><span class="op">[</span>,<span class="fu">.</span><span class="op">(</span><span class="va">center_x</span>, <span class="op">-</span><span class="va">center_y</span>, <span class="va">V1</span><span class="op">)</span><span class="op">]</span>,</span>
<span>                             instructions <span class="op">=</span> <span class="va">instrs</span><span class="op">)</span></span>
<span><span class="co"># add metadata of fov and volume</span></span>
<span><span class="va">vizgen</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/addCellMetadata.html">addCellMetadata</a></span><span class="op">(</span><span class="va">vizgen</span>,</span>
<span>                          new_metadata <span class="op">=</span> <span class="va">meta_dt</span><span class="op">[</span>,<span class="fu">.</span><span class="op">(</span><span class="va">fov</span>, <span class="va">volume</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="visualize-cells-in-space-">4.1 Visualize cells in space.<a class="anchor" aria-label="anchor" href="#visualize-cells-in-space-"></a>
</h3>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/spatPlot.html">spatPlot2D</a></span><span class="op">(</span><span class="va">vizgen</span>,</span>
<span>           point_size <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/vizgen_mouse_brain/0-overview_spatplot.png"></p>
</div>
<div class="section level3">
<h3 id="visualize-cells-by-fov-">4.2 Visualize cells by FOV.<a class="anchor" aria-label="anchor" href="#visualize-cells-by-fov-"></a>
</h3>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># FOVs as a factor</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/spatPlot.html">spatPlot2D</a></span><span class="op">(</span><span class="va">vizgen</span>, point_size <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>           cell_color <span class="op">=</span> <span class="st">'fov'</span>,</span>
<span>           show_legend <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># FOVs sequentially</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/spatPlot.html">spatPlot2D</a></span><span class="op">(</span><span class="va">vizgen</span>, point_size <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>           cell_color <span class="op">=</span> <span class="st">'fov'</span>,</span>
<span>           color_as_factor <span class="op">=</span> <span class="cn">F</span>,</span>
<span>           cell_color_gradient <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'white'</span>, <span class="st">'pink'</span>, <span class="st">'red'</span><span class="op">)</span>,</span>
<span>           show_legend <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/vizgen_mouse_brain/1-FOVs.png"><img src="images/vizgen_mouse_brain/1b-FOVs.png"></p>
</div>
</div>
<div class="section level2">
<h2 id="attaching-images">5. Attaching images<a class="anchor" aria-label="anchor" href="#attaching-images"></a>
</h2>
<p>Images for confocal planes z0 to z6 are provided for both
<strong>DAPI</strong> (cell nucleus staining) and <strong>polyT</strong>
for all datasets. A <code>micron_to_mosaic_pixel_transform.csv</code> is
included within the <code>images</code> folder that provides scaling
factors to map the image to the spatial coordinates. For this
dataset:</p>
<p>micron_to_mosaic_pixel_transform.csv</p>
<table class="table">
<thead><tr>
<th style="text-align:right;">
V1
</th>
<th style="text-align:right;">
V2
</th>
<th style="text-align:right;">
V3
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:right;">
9.205861
</td>
<td style="text-align:right;">
0.00000
</td>
<td style="text-align:right;">
279.2204
</td>
</tr>
<tr>
<td style="text-align:right;">
0.000000
</td>
<td style="text-align:right;">
9.20585
</td>
<td style="text-align:right;">
349.8105
</td>
</tr>
<tr>
<td style="text-align:right;">
0.000000
</td>
<td style="text-align:right;">
0.00000
</td>
<td style="text-align:right;">
1.0000
</td>
</tr>
</tbody>
</table>
<p>Here we will attach the z0 dapi image to the Giotto object.
<strong>Note:</strong> It is recommended for the image files to be
local. Placing the images on the cloud or network may be very slow.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load in image as a giottoLargeImage object that maps spatial coordinates 1:1 with pixel coordinates</span></span>
<span><span class="va">dapi0</span> <span class="op">=</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/createGiottoLargeImage.html">createGiottoLargeImage</a></span><span class="op">(</span>raster_object <span class="op">=</span> <span class="va">img_path</span>,</span>
<span>                               name <span class="op">=</span> <span class="st">'image'</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Preview image</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">dapi0</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/vizgen_mouse_brain/2-dapi0_preview.png"></p>
<p>Attaching the <code>giottoLargeImage</code> to our Giotto object
(provided as a list of 1) and then updating it to map the image to the
spatial coordinates which are in microns.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Adds the giottoLargeImage object to giotto object while also shifting values into the negatives</span></span>
<span><span class="va">vizgen</span> <span class="op">=</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/addGiottoImage.html">addGiottoImage</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">vizgen</span>,</span>
<span>                        largeImages <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">dapi0</span><span class="op">)</span>,</span>
<span>                        negative_y <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Read in image scale transform values</span></span>
<span><span class="va">img_scale_DT</span> <span class="op">=</span> <span class="fu">data.table</span><span class="fu">::</span><span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="va">img_scale_path</span><span class="op">)</span></span>
<span><span class="va">x_scale</span> <span class="op">=</span> <span class="va">img_scale_DT</span><span class="op">$</span><span class="va">V1</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">y_scale</span> <span class="op">=</span> <span class="va">img_scale_DT</span><span class="op">$</span><span class="va">V2</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">x_shift</span> <span class="op">=</span> <span class="va">img_scale_DT</span><span class="op">$</span><span class="va">V3</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">y_shift</span> <span class="op">=</span> <span class="op">-</span><span class="va">img_scale_DT</span><span class="op">$</span><span class="va">V3</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Update image to reverse the above transformations to convert mosaic pixel to micron</span></span>
<span><span class="co"># 'first_adj' means that the xy shifts are applied before the subsequent scaling</span></span>
<span><span class="va">vizgen</span> <span class="op">=</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/updateGiottoLargeImage.html">updateGiottoLargeImage</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">vizgen</span>,</span>
<span>                                largeImage_name <span class="op">=</span> <span class="st">'image'</span>,</span>
<span>                                x_shift <span class="op">=</span> <span class="op">-</span><span class="va">x_shift</span>,</span>
<span>                                y_shift <span class="op">=</span> <span class="op">-</span><span class="va">y_shift</span>,</span>
<span>                                scale_x <span class="op">=</span> <span class="fl">1</span><span class="op">/</span><span class="va">x_scale</span>,</span>
<span>                                scale_y <span class="op">=</span> <span class="fl">1</span><span class="op">/</span><span class="va">y_scale</span>,</span>
<span>                                order <span class="op">=</span> <span class="st">'first_adj'</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="check-image-alignment">5.1 Check image alignment<a class="anchor" aria-label="anchor" href="#check-image-alignment"></a>
</h3>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/spatPlot.html">spatPlot2D</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">vizgen</span>,</span>
<span>           largeImage_name <span class="op">=</span> <span class="st">'image'</span>,</span>
<span>           point_size <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>           show_image <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/vizgen_mouse_brain/3-img_overlay.png"></p>
</div>
<div class="section level3">
<h3 id="zooming-in-by-subsetting-the-dataset">5.2 Zooming in by subsetting the dataset<a class="anchor" aria-label="anchor" href="#zooming-in-by-subsetting-the-dataset"></a>
</h3>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">zoom</span> <span class="op">=</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/subsetGiottoLocs.html">subsetGiottoLocs</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">vizgen</span>,</span>
<span>                        x_min <span class="op">=</span> <span class="fl">2000</span>,</span>
<span>                        x_max <span class="op">=</span> <span class="fl">2500</span>,</span>
<span>                        y_min <span class="op">=</span> <span class="op">-</span><span class="fl">2500</span>,</span>
<span>                        y_max <span class="op">=</span> <span class="op">-</span><span class="fl">2000</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/spatPlot.html">spatPlot2D</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">zoom</span>,</span>
<span>           largeImage_name <span class="op">=</span> <span class="st">'image'</span>,</span>
<span>           point_size <span class="op">=</span> <span class="fl">1</span>,</span>
<span>           show_image <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/vizgen_mouse_brain/4-img_overlay_zoom.png"></p>
</div>
</div>
<div class="section level2">
<h2 id="data-processing">6. Data processing<a class="anchor" aria-label="anchor" href="#data-processing"></a>
</h2>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vizgen</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/filterGiotto.html">filterGiotto</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">vizgen</span>,</span>
<span>                       expression_threshold <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                       feat_det_in_min_cells <span class="op">=</span> <span class="fl">100</span>,</span>
<span>                       min_det_feats_per_cell <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></span>
<span></span>
<span><span class="va">vizgen</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/normalizeGiotto.html">normalizeGiotto</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">vizgen</span>,</span>
<span>                          scalefactor <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>                          verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># add gene and cell statistics</span></span>
<span><span class="va">vizgen</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/addStatistics.html">addStatistics</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">vizgen</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="visualize-the-number-of-features-per-cell-">6.1 Visualize the number of features per cell.<a class="anchor" aria-label="anchor" href="#visualize-the-number-of-features-per-cell-"></a>
</h3>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/spatPlot.html">spatPlot2D</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">vizgen</span>,</span>
<span>           show_image <span class="op">=</span> <span class="cn">F</span>,</span>
<span>           point_alpha <span class="op">=</span> <span class="fl">0.7</span>,</span>
<span>           cell_color <span class="op">=</span> <span class="st">'nr_feats'</span>,</span>
<span>           color_as_factor <span class="op">=</span> <span class="cn">F</span>,</span>
<span>           point_border_col <span class="op">=</span> <span class="st">'grey'</span>,</span>
<span>           point_border_stroke <span class="op">=</span> <span class="fl">0.01</span>,</span>
<span>           point_size <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/vizgen_mouse_brain/5-nr_feats.png"></p>
</div>
</div>
<div class="section level2">
<h2 id="dimension-reduction">7. Dimension reduction<a class="anchor" aria-label="anchor" href="#dimension-reduction"></a>
</h2>
<p>Skipping highly variable feature (HVF) detection. PCA will be
calculated based on all available genes.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vizgen</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runPCA.html">runPCA</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">vizgen</span>,</span>
<span>                 center <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                 scale_unit <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co"># visualize variance explained per component </span></span>
<span><span class="fu"><a href="../reference/screePlot.html">screePlot</a></span><span class="op">(</span><span class="va">vizgen</span>,</span>
<span>          ncp <span class="op">=</span> <span class="fl">30</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/vizgen_mouse_brain/6-screePlot.png"></p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/plotPCA.html">plotPCA</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">vizgen</span>,</span>
<span>        point_size <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/vizgen_mouse_brain/7-PCA.png"></p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vizgen</span> <span class="op">=</span> <span class="fu"><a href="../reference/runUMAP.html">runUMAP</a></span><span class="op">(</span><span class="va">vizgen</span>,</span>
<span>                 dimensions_to_use <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/plotUMAP.html">plotUMAP</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">vizgen</span>,</span>
<span>         point_size <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/vizgen_mouse_brain/8-UMAP.png"></p>
</div>
<div class="section level2">
<h2 id="leiden-clustering">8. Leiden Clustering<a class="anchor" aria-label="anchor" href="#leiden-clustering"></a>
</h2>
<p>Calculate nearest neighbor network and perform Leiden clustering.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vizgen</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/createNearestNetwork.html">createNearestNetwork</a></span><span class="op">(</span><span class="va">vizgen</span>,</span>
<span>                               dimensions_to_use <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span>,</span>
<span>                               k <span class="op">=</span> <span class="fl">15</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Default name for the results is 'leiden_clus' which is then appended to the cell metadata</span></span>
<span><span class="va">vizgen</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/doLeidenCluster.html">doLeidenCluster</a></span><span class="op">(</span><span class="va">vizgen</span>,</span>
<span>                          resolution <span class="op">=</span> <span class="fl">0.2</span>,</span>
<span>                          n_iterations <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></span></code></pre></div>
<p>Cell Metadata Preview</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/pDataDT.html">pDataDT</a></span><span class="op">(</span><span class="va">vizgen</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" tabindex="-1"></a>cell_ID  fov    volume leiden_clus</span>
<span id="cb18-2"><a href="#cb18-2" tabindex="-1"></a><span class="dv">1</span><span class="sc">:</span> <span class="dv">110883424764611924400221639916314253469</span>    <span class="dv">0</span>  <span class="fl">432.1414</span>           <span class="dv">9</span></span>
<span id="cb18-3"><a href="#cb18-3" tabindex="-1"></a><span class="dv">2</span><span class="sc">:</span> <span class="dv">135188247894899244046039873973964001182</span>    <span class="dv">0</span> <span class="fl">1351.8026</span>           <span class="dv">9</span></span>
<span id="cb18-4"><a href="#cb18-4" tabindex="-1"></a><span class="dv">3</span><span class="sc">:</span> <span class="dv">164766962839370328502017156371562646881</span>    <span class="dv">0</span> <span class="fl">1080.6533</span>           <span class="dv">9</span></span>
<span id="cb18-5"><a href="#cb18-5" tabindex="-1"></a><span class="dv">4</span><span class="sc">:</span> <span class="dv">165747897693809971960756442245389760838</span>    <span class="dv">0</span> <span class="fl">1652.0007</span>           <span class="dv">9</span></span>
<span id="cb18-6"><a href="#cb18-6" tabindex="-1"></a><span class="dv">5</span><span class="sc">:</span> <span class="dv">260943245639750847364278545493286724628</span>    <span class="dv">0</span> <span class="fl">1343.3786</span>           <span class="dv">9</span></span>
<span id="cb18-7"><a href="#cb18-7" tabindex="-1"></a><span class="sc">---</span>                                                                   </span>
<span id="cb18-8"><a href="#cb18-8" tabindex="-1"></a>  <span class="dv">78258</span><span class="sc">:</span> <span class="dv">165273009496786595275688065919008183969</span> <span class="dv">1225</span> <span class="fl">1159.6232</span>           <span class="dv">9</span></span>
<span id="cb18-9"><a href="#cb18-9" tabindex="-1"></a><span class="dv">78259</span><span class="sc">:</span> <span class="dv">250474226357477911702383283537224741401</span> <span class="dv">1225</span> <span class="fl">1058.0623</span>           <span class="dv">9</span></span>
<span id="cb18-10"><a href="#cb18-10" tabindex="-1"></a><span class="dv">78260</span><span class="sc">:</span>  <span class="dv">66106840181174834341279408890707577820</span> <span class="dv">1225</span> <span class="fl">1155.3068</span>           <span class="dv">9</span></span>
<span id="cb18-11"><a href="#cb18-11" tabindex="-1"></a><span class="dv">78261</span><span class="sc">:</span>  <span class="dv">66165211106933093510165165316573672348</span> <span class="dv">1225</span>  <span class="fl">394.8081</span>           <span class="dv">9</span></span>
<span id="cb18-12"><a href="#cb18-12" tabindex="-1"></a><span class="dv">78262</span><span class="sc">:</span>  <span class="dv">71051447268015582817266088343399517927</span> <span class="dv">1225</span>  <span class="fl">798.6088</span>           <span class="dv">9</span></span></code></pre></div>
<p>Visualize the Leiden clustering results (â€˜leiden_clusâ€™) mapped onto
the UMAP dimension reduction.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/plotUMAP.html">plotUMAP</a></span><span class="op">(</span><span class="va">vizgen</span>,</span>
<span>         cell_color <span class="op">=</span> <span class="st">'leiden_clus'</span>,</span>
<span>         point_size <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/vizgen_mouse_brain/9-UMAP.png"></p>
<p>Visualize the leiden clustering mapped onto the spatial data.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/spatPlot.html">spatPlot2D</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">vizgen</span>,</span>
<span>           cell_color <span class="op">=</span> <span class="st">'leiden_clus'</span>,</span>
<span>           point_size <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>           background_color <span class="op">=</span> <span class="st">'black'</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/vizgen_mouse_brain/10-spatPlot2D.png"></p>
</div>
<div class="section level2">
<h2 id="spatial-expression-patterns">9. Spatial expression patterns<a class="anchor" aria-label="anchor" href="#spatial-expression-patterns"></a>
</h2>
<p>Spatially interesting gene expression can be detected by first
generating a spatial network then performing Binary Spatial Extraction
of genes.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># create spatial network based on physical distance of cell centroids</span></span>
<span><span class="va">vizgen</span> <span class="op">=</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/createSpatialNetwork.html">createSpatialNetwork</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">vizgen</span>,</span>
<span>                              minimum_k <span class="op">=</span> <span class="fl">2</span>,</span>
<span>                              maximum_distance_delaunay <span class="op">=</span> <span class="fl">50</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># perform Binary Spatial Extraction of genes</span></span>
<span><span class="va">km_spatialgenes</span> <span class="op">=</span> <span class="fu"><a href="../reference/binSpect.html">binSpect</a></span><span class="op">(</span><span class="va">vizgen</span><span class="op">)</span></span></code></pre></div>
<p>Preview km_spatialgenes</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">km_spatialgenes</span><span class="op">$</span><span class="va">feats</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">30</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" tabindex="-1"></a>[<span class="dv">1</span>] <span class="st">"Slc47a1"</span>   <span class="st">"Chat"</span>      <span class="st">"Th"</span>        <span class="st">"Insrr"</span>     <span class="st">"Slc17a7"</span>   <span class="st">"Pln"</span>      </span>
<span id="cb23-2"><a href="#cb23-2" tabindex="-1"></a>[<span class="dv">7</span>] <span class="st">"Lmod1"</span>     <span class="st">"Blank-119"</span> <span class="st">"Hcar1"</span>     <span class="st">"Glp1r"</span>     <span class="st">"Ptgdr"</span>     <span class="st">"Avpr2"</span>    </span>
<span id="cb23-3"><a href="#cb23-3" tabindex="-1"></a>[<span class="dv">13</span>] <span class="st">"Gpr20"</span>     <span class="st">"Myh11"</span>     <span class="st">"Glp2r"</span>     <span class="st">"Npy2r"</span>     <span class="st">"Gpr182"</span>    <span class="st">"Chrm1"</span>    </span>
<span id="cb23-4"><a href="#cb23-4" tabindex="-1"></a>[<span class="dv">19</span>] <span class="st">"Adgrd1"</span>    <span class="st">"Mrgprf"</span>    <span class="st">"Trhr"</span>      <span class="st">"Gfap"</span>      <span class="st">"Slc17a8"</span>   <span class="st">"Nmbr"</span>     </span>
<span id="cb23-5"><a href="#cb23-5" tabindex="-1"></a>[<span class="dv">25</span>] <span class="st">"Pth2r"</span>     <span class="st">"Rxfp1"</span>     <span class="st">"Musk"</span>      <span class="st">"F2rl1"</span>     <span class="st">"Dgkk"</span>      <span class="st">"Chrm5"</span>    </span></code></pre></div>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># visualize spatial expression of select genes obtained from binSpect</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/spatFeatPlot2D.html">spatFeatPlot2D</a></span><span class="op">(</span><span class="va">vizgen</span>,</span>
<span>               expression_values <span class="op">=</span> <span class="st">'scaled'</span>,</span>
<span>               feats <span class="op">=</span> <span class="va">km_spatialgenes</span><span class="op">$</span><span class="va">feats</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span>,<span class="fl">3</span>,<span class="fl">5</span>,<span class="fl">16</span>,<span class="fl">22</span><span class="op">)</span><span class="op">]</span>,</span>
<span>               cell_color_gradient <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'blue'</span>, <span class="st">'white'</span>, <span class="st">'red'</span><span class="op">)</span>,</span>
<span>               point_shape <span class="op">=</span> <span class="st">'border'</span>,</span>
<span>               point_border_col <span class="op">=</span> <span class="st">'grey'</span>,</span>
<span>               point_border_stroke <span class="op">=</span> <span class="fl">0.01</span>,</span>
<span>               point_size <span class="op">=</span> <span class="fl">0.2</span>,</span>
<span>               cow_n_col <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/vizgen_mouse_brain/11-spatFeatPlot2D.png"></p>
</div>
<div class="section level2">
<h2 id="sec-subcellular">10. Working with subcellular information<a class="anchor" aria-label="anchor" href="#sec-subcellular"></a>
</h2>
<p><em>These steps may require a strong computer.</em></p>
<p>Vizgen provides the raw information used to generate the aggregated
data through the detected_transcripts.csv and cell_boundaries hdf5
files. Giotto can also work directly with this information.</p>
<div class="section level3">
<h3 id="optional-define-region-of-interest-and-find-fovs-needed">10.1 (Optional) Define region of interest and find FOVs needed<a class="anchor" aria-label="anchor" href="#optional-define-region-of-interest-and-find-fovs-needed"></a>
</h3>
<p>Loading information by only grabbing the needed FOVs can cut down on
computational requirements.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">subsetFOVs</span> <span class="op">=</span> <span class="va">meta_dt</span><span class="op">[</span><span class="va">center_x</span> <span class="op">&gt;</span> <span class="fl">2000</span> <span class="op">&amp;</span> <span class="va">center_x</span> <span class="op">&lt;</span> <span class="fl">3100</span> <span class="op">&amp;</span></span>
<span>                       <span class="va">center_y</span> <span class="op">&gt;</span> <span class="fl">2500</span> <span class="op">&amp;</span> <span class="va">center_y</span> <span class="op">&lt;</span> <span class="fl">3500</span><span class="op">]</span><span class="op">$</span><span class="va">fov</span></span>
<span><span class="va">subsetFOVs</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">subsetFOVs</span><span class="op">)</span></span></code></pre></div>
<p>FOVs needed</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">subsetFOVs</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" tabindex="-1"></a>[<span class="dv">1</span>] <span class="dv">220</span> <span class="dv">221</span> <span class="dv">222</span> <span class="dv">223</span> <span class="dv">224</span> <span class="dv">225</span> <span class="dv">245</span> <span class="dv">246</span> <span class="dv">247</span> <span class="dv">248</span> <span class="dv">249</span> <span class="dv">250</span> <span class="dv">275</span> <span class="dv">276</span> <span class="dv">277</span> <span class="dv">278</span> <span class="dv">279</span> <span class="dv">280</span> <span class="dv">302</span></span>
<span id="cb27-2"><a href="#cb27-2" tabindex="-1"></a>[<span class="dv">20</span>] <span class="dv">303</span> <span class="dv">304</span> <span class="dv">305</span> <span class="dv">306</span> <span class="dv">307</span> <span class="dv">330</span> <span class="dv">331</span> <span class="dv">332</span> <span class="dv">333</span> <span class="dv">334</span> <span class="dv">335</span> <span class="dv">358</span> <span class="dv">359</span> <span class="dv">360</span> <span class="dv">361</span> <span class="dv">362</span> <span class="dv">363</span></span></code></pre></div>
<p><strong>Note:</strong> The following steps will be assuming that this
step was run.</p>
</div>
<div class="section level3">
<h3 id="creating-a-giottopolygon-object">10.2 Creating a giottoPolygon object<a class="anchor" aria-label="anchor" href="#creating-a-giottopolygon-object"></a>
</h3>
<p>Cell boundary annotations are represented in Giotto as
<code>giottoPolygon</code> objects which can be previewed by directly
plotting them.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># read polygons and add them to Giotto</span></span>
<span><span class="co"># fovs param is optional</span></span>
<span><span class="co"># polygon_feat_types determines which Vizgen polygon z slices are loaded (There are 0 - 6)</span></span>
<span><span class="va">polys</span> <span class="op">=</span> <span class="fu"><a href="../reference/readPolygonFilesVizgenHDF5.html">readPolygonFilesVizgenHDF5</a></span><span class="op">(</span>boundaries_path <span class="op">=</span> <span class="va">bound_path</span>,</span>
<span>                                   polygon_feat_types <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">6</span><span class="op">)</span>,</span>
<span>                                   flip_y_axis <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                                   fovs <span class="op">=</span> <span class="va">subsetFOVs</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># polys is produced as a list of 2 giottoPolygon objects (z0 and z6)</span></span>
<span><span class="co"># previewing the first one...</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">polys</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/vizgen_mouse_brain/13-poly.png"></p>
</div>
<div class="section level3">
<h3 id="creating-a-giottopoints-object">10.3 Creating a giottoPoints object<a class="anchor" aria-label="anchor" href="#creating-a-giottopoints-object"></a>
</h3>
<p>Giotto represents single-molecule transcript level spatial data as
<code>giottoPoints</code> objects.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tx_dt</span> <span class="op">=</span> <span class="fu">data.table</span><span class="fu">::</span><span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="va">tx_path</span><span class="op">)</span></span>
<span><span class="co"># select transcripts in FOVs</span></span>
<span><span class="va">tx_dt_selected</span> <span class="op">=</span> <span class="va">tx_dt</span><span class="op">[</span><span class="va">fov</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">subsetFOVs</span><span class="op">]</span></span>
<span><span class="va">tx_dt_selected</span><span class="op">$</span><span class="va">global_y</span> <span class="op">=</span> <span class="op">-</span><span class="va">tx_dt_selected</span><span class="op">$</span><span class="va">global_y</span></span>
<span><span class="co"># (note the inverted y is the same as when spatial locations were loaded)</span></span>
<span></span>
<span><span class="co"># create Giotto points from transcripts </span></span>
<span><span class="va">gpoints</span> <span class="op">=</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/createGiottoPoints.html">createGiottoPoints</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">tx_dt_selected</span><span class="op">[</span>,<span class="fu">.</span><span class="op">(</span><span class="va">global_x</span>, <span class="va">global_y</span>, <span class="va">gene</span>, <span class="va">global_z</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># preview the giottoPoints object (Including specific feats to plot is highly recommended)</span></span>
<span><span class="co"># Not providing the feats param will plot ALL features detected</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">gpoints</span>,</span>
<span>     feats <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'Gfap'</span>, <span class="st">'Ackr1'</span><span class="op">)</span>,</span>
<span>     point_size <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/vizgen_mouse_brain/14-point.png"></p>
</div>
<div class="section level3">
<h3 id="creating-a-subcellular-giotto-object">10.4 Creating a subcellular Giotto object<a class="anchor" aria-label="anchor" href="#creating-a-subcellular-giotto-object"></a>
</h3>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Create new giotto instructions to set different save directory if desired</span></span>
<span><span class="co"># instrs_sub = createGiottoInstructions(show_plot = FALSE,</span></span>
<span><span class="co">#                                       return_plot = FALSE,</span></span>
<span><span class="co">#                                       save_plot = TRUE,</span></span>
<span><span class="co">#                                       save_dir = "path/to/subcellular/save/folder")</span></span>
<span></span>
<span><span class="va">vizgen_subcellular</span> <span class="op">=</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/createGiottoObjectSubcellular.html">createGiottoObjectSubcellular</a></span><span class="op">(</span>gpoints <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>rna <span class="op">=</span> <span class="va">gpoints</span><span class="op">)</span>,</span>
<span>                                                   gpolygons <span class="op">=</span> <span class="va">polys</span>,</span>
<span>                                                   <span class="co"># instructions = instrs_sub</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Find polygon centroids and generate associated spatial locations</span></span>
<span><span class="va">vizgen_subcellular</span> <span class="op">=</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/addSpatialCentroidLocations.html">addSpatialCentroidLocations</a></span><span class="op">(</span><span class="va">vizgen_subcellular</span>,</span>
<span>                                                 poly_info <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'z0'</span>, <span class="st">'z6'</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># Append images **(Details about these functions can be found in step 5)**</span></span>
<span><span class="co"># Run following two commented lines if skipping to this step</span></span>
<span><span class="co"># dapi0 = createGiottoLargeImage(raster_object = img_path, name = 'image')</span></span>
<span><span class="co"># img_scale_DT = data.table::fread(img_scale_path)</span></span>
<span><span class="va">vizgen_subcellular</span> <span class="op">=</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/addGiottoImage.html">addGiottoImage</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">vizgen_subcellular</span>,</span>
<span>                                    largeImages <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">dapi0</span><span class="op">)</span>, negative_y <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">x_scale</span> <span class="op">=</span> <span class="va">img_scale_DT</span><span class="op">$</span><span class="va">V1</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> ; <span class="va">y_scale</span> <span class="op">=</span> <span class="va">img_scale_DT</span><span class="op">$</span><span class="va">V2</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">x_shift</span> <span class="op">=</span> <span class="va">img_scale_DT</span><span class="op">$</span><span class="va">V3</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> ; <span class="va">y_shift</span> <span class="op">=</span> <span class="op">-</span><span class="va">img_scale_DT</span><span class="op">$</span><span class="va">V3</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">vizgen_subcellular</span> <span class="op">=</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/updateGiottoLargeImage.html">updateGiottoLargeImage</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">vizgen_subcellular</span>,</span>
<span>                                            largeImage_name <span class="op">=</span> <span class="st">'image'</span>, x_shift <span class="op">=</span> <span class="op">-</span><span class="va">x_shift</span>, y_shift <span class="op">=</span> <span class="op">-</span><span class="va">y_shift</span>,</span>
<span>                                            scale_x <span class="op">=</span> <span class="fl">1</span><span class="op">/</span><span class="va">x_scale</span>, scale_y <span class="op">=</span> <span class="fl">1</span><span class="op">/</span><span class="va">y_scale</span>, order <span class="op">=</span> <span class="st">'first_adj'</span><span class="op">)</span></span></code></pre></div>
<p>Alternatively to append subcellular data to existing giotto
object.</p>
<p>The subcellular information can also be attached to giotto objects
built with the pre-aggregated information.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Subset dataset to work with smaller area</span></span>
<span><span class="va">vizgen_subset</span> <span class="op">=</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/subsetGiottoLocs.html">subsetGiottoLocs</a></span><span class="op">(</span><span class="va">vizgen</span>,</span>
<span>                                 x_min <span class="op">=</span> <span class="fl">2000</span>, x_max <span class="op">=</span> <span class="fl">3000</span>,</span>
<span>                                 y_min <span class="op">=</span> <span class="op">-</span><span class="fl">3500</span>, y_max <span class="op">=</span> <span class="op">-</span><span class="fl">2500</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># add points to Giotto object</span></span>
<span><span class="va">vizgen_subset</span> <span class="op">=</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/addGiottoPoints.html">addGiottoPoints</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">vizgen_subset</span>,</span>
<span>                                gpoints <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>rna <span class="op">=</span> <span class="va">gpoints</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># There will be a warning that 14 features are not present in vizgen_subset</span></span>
<span><span class="co"># (They were previously removed during data processing in step 6)</span></span>
<span></span>
<span><span class="co"># add polygons to Giotto object</span></span>
<span><span class="va">vizgen_subset</span> <span class="op">=</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/addGiottoPolygons.html">addGiottoPolygons</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">vizgen_subset</span>,</span>
<span>                                  gpolygons <span class="op">=</span> <span class="va">polys</span><span class="op">)</span></span>
<span><span class="co"># If the polygons had not already been read readPolygonFilesVizgen() should be used instead</span></span>
<span><span class="co"># vizgen_subset = readPolygonFilesVizgen(gobject = vizgen_subset,</span></span>
<span><span class="co"># boundaries_path = bound_path,</span></span>
<span><span class="co"># polygon_feat_types = c(0,6)) # Defines which z slices (polys) are read in</span></span>
<span></span>
<span><span class="co"># Calculate centroid locations</span></span>
<span><span class="va">vizgen_subset</span> <span class="op">=</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/addSpatialCentroidLocations.html">addSpatialCentroidLocations</a></span><span class="op">(</span><span class="va">vizgen_subset</span>,</span>
<span>                                            poly_info <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'z0'</span>, <span class="st">'z6'</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><strong>Visualize subset</strong></p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/spatPlot.html">spatPlot2D</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">vizgen_subset</span>,</span>
<span>           show_image <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>           largeImage_name <span class="op">=</span> <span class="st">'image'</span>,</span>
<span>           cell_color <span class="op">=</span> <span class="st">'leiden_clus'</span>,</span>
<span>           point_size <span class="op">=</span> <span class="fl">2.5</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/vizgen_mouse_brain/12-spatPlot2D.png"></p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># identify genes for visualization</span></span>
<span><span class="va">gene_meta</span> <span class="op">=</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/fDataDT.html">fDataDT</a></span><span class="op">(</span><span class="va">vizgen_subset</span><span class="op">)</span></span>
<span><span class="fu">data.table</span><span class="fu">::</span><span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/setorder.html" class="external-link">setorder</a></span><span class="op">(</span><span class="va">gene_meta</span>, <span class="va">perc_cells</span><span class="op">)</span></span>
<span><span class="va">gene_meta</span><span class="op">[</span><span class="va">perc_cells</span> <span class="op">&gt;</span> <span class="fl">25</span> <span class="op">&amp;</span> <span class="va">perc_cells</span> <span class="op">&lt;</span> <span class="fl">50</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># visualize points with z0 polygons (confocal plane)</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/spatInSituPlotPoints.html">spatInSituPlotPoints</a></span><span class="op">(</span><span class="va">vizgen_subset</span>,</span>
<span>                     feats <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">'rna'</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Oxgr1"</span>, <span class="st">"Htr1a"</span>, <span class="st">"Gjc3"</span>, <span class="st">"Axl"</span>,</span>
<span>                                            <span class="st">'Gfap'</span>, <span class="st">"Olig1"</span>, <span class="st">"Epha7"</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                     polygon_feat_type <span class="op">=</span> <span class="st">'z0'</span>,</span>
<span>                     use_overlap <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                     point_size <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>                     polygon_line_size <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                     show_polygon <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                     polygon_bg_color <span class="op">=</span> <span class="st">'white'</span>,</span>
<span>                     polygon_color <span class="op">=</span> <span class="st">'white'</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/vizgen_mouse_brain/12b-spatInSituPlotPoints.png"></p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># visualize select genes with z0 polygons (confocal plane)</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/spatInSituPlotPoints.html">spatInSituPlotPoints</a></span><span class="op">(</span><span class="va">vizgen_subcellular</span>,</span>
<span>                     feats <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">'rna'</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Oxgr1"</span>, <span class="st">"Htr1a"</span>, <span class="st">"Gjc3"</span>, <span class="st">"Axl"</span>,</span>
<span>                                            <span class="st">'Gfap'</span>, <span class="st">"Olig1"</span>, <span class="st">"Epha7"</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                     polygon_feat_type <span class="op">=</span> <span class="st">'z0'</span>,</span>
<span>                     use_overlap <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                     point_size <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>                     polygon_line_size <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>                     show_polygon <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                     polygon_bg_color <span class="op">=</span> <span class="st">'white'</span>,</span>
<span>                     polygon_color <span class="op">=</span> <span class="st">'white'</span>,</span>
<span>                     coord_fix_ratio <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                     save_param <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>base_height <span class="op">=</span> <span class="fl">10</span>,</span>
<span>                                       base_width <span class="op">=</span> <span class="fl">10</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/vizgen_mouse_brain/15-spatInSituPlotPoints.png"></p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Zoom in further and visualize with image</span></span>
<span><span class="va">vizgen_subcellular_zoom</span> <span class="op">=</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/subsetGiottoLocs.html">subsetGiottoLocs</a></span><span class="op">(</span><span class="va">vizgen_subcellular</span>,</span>
<span>                                           poly_info <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'z0'</span>,<span class="st">'z6'</span><span class="op">)</span>,</span>
<span>                                           x_min <span class="op">=</span> <span class="fl">2400</span>, x_max <span class="op">=</span> <span class="fl">2600</span>,</span>
<span>                                           y_min <span class="op">=</span> <span class="op">-</span><span class="fl">3200</span>, y_max <span class="op">=</span> <span class="op">-</span><span class="fl">3000</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/spatInSituPlotPoints.html">spatInSituPlotPoints</a></span><span class="op">(</span><span class="va">vizgen_subcellular_zoom</span>,</span>
<span>                     feats <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">'rna'</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Oxgr1"</span>, <span class="st">"Htr1a"</span>, <span class="st">"Gjc3"</span>, <span class="st">"Axl"</span>,</span>
<span>                                            <span class="st">'Gfap'</span>, <span class="st">"Olig1"</span>, <span class="st">"Epha7"</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                     polygon_feat_type <span class="op">=</span> <span class="st">'z0'</span>,</span>
<span>                     use_overlap <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                     point_size <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>                     polygon_line_size <span class="op">=</span> <span class="fl">0.2</span>,</span>
<span>                     show_polygon <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                     polygon_bg_color <span class="op">=</span> <span class="st">'white'</span>,</span>
<span>                     polygon_color <span class="op">=</span> <span class="st">'white'</span>,</span>
<span>                     polygon_alpha <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>                     show_image <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                     largeImage_name <span class="op">=</span> <span class="st">'image'</span>,</span>
<span>                     coord_fix_ratio <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                     save_param <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>base_height <span class="op">=</span> <span class="fl">10</span>,</span>
<span>                                       base_width <span class="op">=</span> <span class="fl">10</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/vizgen_mouse_brain/16-spatInSituPlotPoints.png"></p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://drieslab.github.io/Giotto/reference/spatInSituPlotPoints.html">spatInSituPlotPoints</a></span><span class="op">(</span><span class="va">vizgen_subcellular_zoom</span>,</span>
<span>                     feats <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">'rna'</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Oxgr1"</span>, <span class="st">"Htr1a"</span>, <span class="st">"Gjc3"</span>, <span class="st">"Axl"</span>,</span>
<span>                                            <span class="st">'Gfap'</span>, <span class="st">"Olig1"</span>, <span class="st">"Epha7"</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                     polygon_feat_type <span class="op">=</span> <span class="st">'z6'</span>, <span class="co"># Use polygon from z slice 6 this time</span></span>
<span>                     use_overlap <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                     point_size <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>                     polygon_line_size <span class="op">=</span> <span class="fl">0.2</span>,</span>
<span>                     show_polygon <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                     polygon_bg_color <span class="op">=</span> <span class="st">'white'</span>,</span>
<span>                     polygon_color <span class="op">=</span> <span class="st">'white'</span>,</span>
<span>                     polygon_alpha <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>                     show_image <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                     largeImage_name <span class="op">=</span> <span class="st">'image'</span>,</span>
<span>                     coord_fix_ratio <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                     save_param <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>base_height <span class="op">=</span> <span class="fl">10</span>,</span>
<span>                                       base_width <span class="op">=</span> <span class="fl">10</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/vizgen_mouse_brain/17-spatInSituPlotPoints.png"></p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by <a href="https://www.drieslab.com" class="external-link">Ruben Dries</a>, Jiaji George Chen, <a href="https://josschavezf.github.io" class="external-link">Joselyn C. ChÃ¡vez-Fuentes</a>, <a href="https://labs.icahn.mssm.edu/yuanlab/" class="external-link">Guo-Cheng Yuan</a>, Matthew Oâ€™Brien, Edward Ruiz, Wen Wang, Natalie Del Rossi.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
