<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Xenium Human Breast Cancer Pre-Release â€¢ Giotto</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Xenium Human Breast Cancer Pre-Release">
<meta property="og:description" content="Giotto">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">Giotto</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">4.0.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Get started
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/installation.html">Giotto Installation</a>
    </li>
    <li>
      <a href="../articles/ecosystem.html">Giotto Ecosystem</a>
    </li>
    <li>
      <a href="../articles/structure.html">Giotto Class Structure</a>
    </li>
    <li>
      <a href="../articles/errorsfaqsandtips.html">Solutions to Installation Errors</a>
    </li>
    <li>
      <a href="../articles/faqs.html">FAQs</a>
    </li>
    <li>
      <a href="../articles/branches.html">Branches</a>
    </li>
    <li>
      <a href="../articles/github_issues.html">How to report a bug</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Documentation</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Examples
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Seq-Scope</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/seqscope_mouse_liver.html">Mouse liver</a>
        </li>
      </ul>
</li>
    <li class="divider">
    </li>
<li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Visium</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/visium_mouse_kidney.html">Visium Mouse Kidney</a>
        </li>
        <li>
          <a href="../articles/visium_mouse_brain.html">Visium Mouse Brain</a>
        </li>
        <li>
          <a href="../articles/visium_prostate_integration.html">Visium Prostate Integration</a>
        </li>
        <li>
          <a href="../articles/visium_cytassist_lungcancer.html">CytAssist Human Lung Cancer</a>
        </li>
        <li>
          <a href="../articles/visium_cytassist_human_glioblastoma.html">CytAssist Multi-omics Human Glioblastoma</a>
        </li>
      </ul>
</li>
    <li class="divider">
    </li>
<li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Spatial Genomics</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/spatial_genomics_mouse_kidney.html">Mouse Kidney</a>
        </li>
      </ul>
</li>
    <li class="divider">
    </li>
<li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Xenium</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/xenium_breast_cancer.html">Xenium Human breast cancer</a>
        </li>
      </ul>
</li>
    <li class="divider">
    </li>
<li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Nanostring CosMx</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/nanostring_cosmx_lung_cancer.html">Nanostring CosMx Subcellular Lung Cancer</a>
        </li>
      </ul>
</li>
    <li class="divider">
    </li>
<li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">MERFISH</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/merfish_mouse_hypothalamic.html">MERFISH Mouse Hypothalamic Preoptic Region</a>
        </li>
      </ul>
</li>
    <li class="divider">
    </li>
<li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Vizgen</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/vizgen_mouse_brain.html">Vizgen Mouse Brain Receptor Map</a>
        </li>
      </ul>
</li>
    <li class="divider">
    </li>
<li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">seqFISH</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/mini_seqfish.html">mini seqFISH</a>
        </li>
        <li>
          <a href="../articles/seqfish_cortex.html">SeqFish+ Mouse Cortex Example</a>
        </li>
      </ul>
</li>
    <li class="divider">
    </li>
<li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">osmFISH</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/osmfish_mouse_ss_cortex.html">osmFISH Mouse SS Cortex</a>
        </li>
      </ul>
</li>
    <li class="divider">
    </li>
<li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Resolve Bioscience</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/resolve_biosciences.html">Resolve Bioscience Breast Cancer Subcellular</a>
        </li>
      </ul>
</li>
    <li class="divider">
    </li>
<li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">CODEX</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/codex_mouse_spleen.html">Codex Mouse Spleen</a>
        </li>
      </ul>
</li>
    <li class="divider">
    </li>
<li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Single Cell RNA Sequencing</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/singlecell_prostate_standard.html">Prostate</a>
        </li>
        <li>
          <a href="../articles/singlecell_prostate_integration.html">Prostate Integration</a>
        </li>
      </ul>
</li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/object_creation.html">Giotto Object Creation</a>
    </li>
    <li class="divider">
    </li>
<li>
      <a href="../articles/configuration.html">Giotto Configuration</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Spatial manipulation</li>
    <li>
      <a href="../articles/spatial_transformations.html">Spatial transformations</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Image alignment</li>
    <li>
      <a href="../articles/image_alignment.html">Image alignment</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Interactive selection</li>
    <li>
      <a href="../articles/interactive_selection.html">Interactive selection</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Visualizations</li>
    <li>
      <a href="../articles/visualizations.html">Visualizations</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Saving and Loading</li>
    <li>
      <a href="../articles/saving_object.html">Saving a Giotto object</a>
    </li>
    <li>
      <a href="../articles/saving_plots.html">Saving plots</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Interoperability</li>
    <li>
      <a href="../articles/interoperability_anndata.html">AnnData</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Try Giotto in the Cloud</li>
    <li>
      <a href="../articles/terra.html">Terra.bio</a>
    </li>
    <li>
      <a href="../articles/docker.html">Docker</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    News
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../news/index.html">Changes</a>
    </li>
    <li>
      <a href="../articles/presentations.html">Presentations</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/drieslab/Giotto/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Xenium Human Breast Cancer Pre-Release</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/drieslab/Giotto/blob/HEAD/vignettes/xenium_breast_cancer.Rmd" class="external-link"><code>vignettes/xenium_breast_cancer.Rmd</code></a></small>
      <div class="hidden name"><code>xenium_breast_cancer.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Ensure Giotto Suite is installed.</span></span>
<span><span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="st">"Giotto"</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/installed.packages.html" class="external-link">installed.packages</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"drieslab/Giotto@suite"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://drieslab.github.io/Giotto/">Giotto</a></span><span class="op">)</span></span>
<span><span class="co"># Ensure the Python environment for Giotto has been installed.</span></span>
<span><span class="va">genv_exists</span> <span class="op">=</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/checkGiottoEnvironment.html">checkGiottoEnvironment</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="va">genv_exists</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="co"># The following command need only be run once to install the Giotto environment.</span></span>
<span>  <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/installGiottoEnvironment.html">installGiottoEnvironment</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<div class="section level2">
<h2 id="set-up-giotto-environment">1. Set up Giotto environment<a class="anchor" aria-label="anchor" href="#set-up-giotto-environment"></a>
</h2>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># 1. ** SET WORKING DIRECTORY WHERE PROJECT OUPUTS WILL SAVE TO **</span></span>
<span><span class="va">results_folder</span> <span class="op">=</span> <span class="st">'/path/to/save/directory/'</span></span>
<span></span>
<span><span class="co"># 2. set giotto python path</span></span>
<span><span class="co"># set python path to your preferred python version path</span></span>
<span><span class="co"># set python path to NULL if you want to automatically install (only the 1st time) and use the giotto miniconda environment</span></span>
<span><span class="va">python_path</span> <span class="op">=</span> <span class="cn">NULL</span> </span>
<span><span class="kw">if</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NULL.html" class="external-link">is.null</a></span><span class="op">(</span><span class="va">python_path</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/installGiottoEnvironment.html">installGiottoEnvironment</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co"># 3. Create Giotto instructions</span></span>
<span><span class="co"># Directly saving plots to the working directory without rendering them in the editor saves time.</span></span>
<span><span class="va">instrs</span> <span class="op">=</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/createGiottoInstructions.html">createGiottoInstructions</a></span><span class="op">(</span>save_dir <span class="op">=</span> <span class="va">results_folder</span>,</span>
<span>                                  save_plot <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                                  show_plot <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                                  return_plot <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="dataset-explanation">2. Dataset explanation<a class="anchor" aria-label="anchor" href="#dataset-explanation"></a>
</h2>
<p>This vignette covers Giotto object creation and simple exploratory
analysis with 10x Genomicsâ€™ subcellular <em>Xenium In Situ</em> platform
data using their <a href="https://www.10xgenomics.com/products/xenium-in-situ/preview-dataset-human-breast" class="external-link">Human
Breast Cancer Dataset</a> provided with their recent <a href="https://www.biorxiv.org/content/10.1101/2022.10.06.510405v1" class="external-link">bioRxiv
pre-print</a>.</p>
<p>The data from the first tissue replicate will be worked with:</p>
<p><img src="images/xenium_breast_cancer/large_preview.png"></p>
</div>
<div class="section level2">
<h2 id="project-data-paths">3. Project data paths<a class="anchor" aria-label="anchor" href="#project-data-paths"></a>
</h2>
<p>Xeniumâ€™s exact output folder structure and which file/formats will be
available in the final release are still subject to change. This
tutorial will be updated for any differences at release.</p>
<p><strong>Note:</strong> The cell_feature_matrix.tar.gz must be
unpacked into a subdirectory within the xenium working directory before
starting.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># ** SET PATH TO FOLDER CONTAINING XENIUM DATA **</span></span>
<span><span class="va">xenium_folder</span> <span class="op">=</span> <span class="st">'/path/to/xenium/data/outputs'</span></span>
<span></span>
<span><span class="co"># general files (some are supplemental files)</span></span>
<span><span class="va">settings_path</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">xenium_folder</span>, <span class="st">'experiment.xenium'</span><span class="op">)</span></span>
<span><span class="va">he_img_path</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">xenium_folder</span>, <span class="st">'Xenium_FFPE_Human_Breast_Cancer_Rep1_he_image.tif'</span><span class="op">)</span></span>
<span><span class="va">if_img_path</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">xenium_folder</span>, <span class="st">'Xenium_FFPE_Human_Breast_Cancer_Rep1_if_image.tif'</span><span class="op">)</span></span>
<span><span class="va">panel_meta_path</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">xenium_folder</span>, <span class="st">'Xenium_FFPE_Human_Breast_Cancer_Rep1_panel.tsv'</span><span class="op">)</span> <span class="co"># (optional)</span></span>
<span></span>
<span><span class="co"># files (SUBCELLULAR): (tutorial focuses on working with these files)</span></span>
<span><span class="va">cell_bound_path</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">xenium_folder</span>, <span class="st">'cell_boundaries.csv.gz'</span><span class="op">)</span></span>
<span><span class="va">nuc_bound_path</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">xenium_folder</span>, <span class="st">'nucleus_boundaries.csv.gz'</span><span class="op">)</span></span>
<span><span class="va">tx_path</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">xenium_folder</span>, <span class="st">'transcripts.csv.gz'</span><span class="op">)</span></span>
<span><span class="va">feat_meta_path</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">xenium_folder</span>, <span class="st">'cell_feature_matrix/features.tsv.gz'</span><span class="op">)</span> <span class="co"># (also used in aggregate)</span></span>
<span></span>
<span><span class="co"># files (AGGREGATE):</span></span>
<span><span class="va">expr_mat_path</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">xenium_folder</span>, <span class="st">'cell_feature_matrix'</span><span class="op">)</span></span>
<span><span class="va">cell_meta_path</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">xenium_folder</span>, <span class="st">'cells.csv.gz'</span><span class="op">)</span> <span class="co"># contains spatlocs</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="xenium-feature-types-exploration">4. Xenium feature types exploration<a class="anchor" aria-label="anchor" href="#xenium-feature-types-exploration"></a>
</h2>
<p><code>features.tsv.gz</code> within
<code>cell_feature_matrix.tar.gz</code> provides information on the
different feature types available within Xeniumâ€™s two types of
expression outputs:</p>
<ol style="list-style-type: decimal">
<li>aggregated cell by feature matrix</li>
<li>transcript detections in <code>transcripts.csv.gz</code>
</li>
</ol>
<p>There are four types of probes used, each of which is represented in
both exports:</p>
<ul>
<li>
<code>gene expression</code>: Gene expression detection</li>
<li>
<code>blank codeword</code>: Unused codeword - there are no probes
that will generate the codeword</li>
<li>
<code>negative control codeword</code>: Valid codewords that do not
have any probes that should yield that code, so they can be used to
assess the specificity of the decoding algorithm</li>
<li>
<code>negative control probe</code>: Probes that exist in the panel,
but target ERCC or other non-biological sequences, which can be used to
assess the specificity of the assay</li>
</ul>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># load features metadata</span></span>
<span><span class="co"># (make sure cell_feature_matrix folder is unpacked)</span></span>
<span><span class="va">feature_dt</span> <span class="op">=</span> <span class="fu">data.table</span><span class="fu">::</span><span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="va">feat_meta_path</span>, header <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/row-plus-colnames-generic.html">colnames</a></span><span class="op">(</span><span class="va">feature_dt</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'ensembl_ID'</span>,<span class="st">'feat_name'</span>,<span class="st">'feat_type'</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># find the feature IDs that belong to each feature type</span></span>
<span><span class="va">feature_dt</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">feat_type</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">feat_types</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">feature_dt</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">feat_type</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="va">feat_types_IDs</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span></span>
<span>  <span class="va">feat_types</span>, <span class="kw">function</span><span class="op">(</span><span class="va">type</span><span class="op">)</span> <span class="va">feature_dt</span><span class="op">[</span><span class="va">feat_type</span> <span class="op">==</span> <span class="va">type</span>, <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">feat_name</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">feat_types_IDs</span><span class="op">)</span> <span class="op">=</span> <span class="va">feat_types</span></span></code></pre></div>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># feat_type</span></span>
<span><span class="co">#            Blank Codeword           Gene Expression</span></span>
<span><span class="co">#                       159                       313</span></span>
<span><span class="co">#            Negative Control Codeword    Negative Control Probe </span></span>
<span><span class="co">#                                   41                        28 </span></span></code></pre></div>
<p>This dataset has 313 probes that are dedicated for gene expression
transcript detection.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#   [1] "ABCC11"   "ACTA2"    "ACTG2"    "ADAM9"    "ADGRE5"   "ADH1B"   </span></span>
<span><span class="co">#   [7] "ADIPOQ"   "AGR3"     "AHSP"     "AIF1"     "AKR1C1"   "AKR1C3"  </span></span>
<span><span class="co">#  [13] "ALDH1A3"  "ANGPT2"   "ANKRD28"  "ANKRD29"  "ANKRD30A" "APOBEC3A"</span></span>
<span><span class="co">#  [19] "APOBEC3B" "APOC1"    "AQP1"     "AQP3"     "AR"       "AVPR1A"  </span></span>
<span><span class="co">#  [25] "BACE2"    "BANK1"    "BASP1"    "BTNL9"    "C15orf48" "C1QA"    </span></span>
<span><span class="co">#  [31] "C1QC"     "C2orf42"  "C5orf46"  "C6orf132" "CAV1"     "CAVIN2"  </span></span>
<span><span class="co">#  [37] "CCDC6"    "CCDC80"   "CCL20"    "CCL5"     "CCL8"     "CCND1"   </span></span>
<span><span class="co">#  [43] "CCPG1"    "CCR7"     "CD14"     "CD163"    "CD19"     "CD1C"    </span></span>
<span><span class="co">#  [49] "CD247"    "CD27"     "CD274"    "CD3D"     "CD3E"     "CD3G"    </span></span>
<span><span class="co">#  [55] "CD4"      "CD68"     "CD69"     "CD79A"    "CD79B"    "CD80"    </span></span>
<span><span class="co">#  [61] "CD83"     "CD86"     "CD8A"     "CD8B"     "CD9"      "CD93"    </span></span>
<span><span class="co">#  [67] "CDC42EP1" "CDH1"     "CEACAM6"  "CEACAM8"  "CENPF"    "CLCA2"   </span></span>
<span><span class="co">#  [73] "CLDN4"    "CLDN5"    "CLEC14A"  "CLEC9A"   "CLECL1"   "CLIC6"   </span></span>
<span><span class="co">#  [79] "CPA3"     "CRHBP"    "CRISPLD2" "CSF3"     "CTH"      "CTLA4"   </span></span>
<span><span class="co">#  [85] "CTSG"     "CTTN"     "CX3CR1"   "CXCL12"   "CXCL16"   "CXCL5"   </span></span>
<span><span class="co">#  [91] "CXCR4"    "CYP1A1"   "CYTIP"    "DAPK3"    "DERL3"    "DMKN"    </span></span>
<span><span class="co">#  [97] "DNAAF1"   "DNTTIP1"  "DPT"      "DSC2"     "DSP"      "DST"     </span></span>
<span><span class="co"># [103] "DUSP2"    "DUSP5"    "EDN1"     "EDNRB"    "EGFL7"    "EGFR"    </span></span>
<span><span class="co"># [109] "EIF4EBP1" "ELF3"     "ELF5"     "ENAH"     "EPCAM"    "ERBB2"   </span></span>
<span><span class="co"># [115] "ERN1"     "ESM1"     "ESR1"     "FAM107B"  "FAM49A"   "FASN"    </span></span>
<span><span class="co"># [121] "FBLIM1"   "FBLN1"    "FCER1A"   "FCER1G"   "FCGR3A"   "FGL2"    </span></span>
<span><span class="co"># [127] "FLNB"     "FOXA1"    "FOXC2"    "FOXP3"    "FSTL3"    "GATA3"   </span></span>
<span><span class="co"># [133] "GJB2"     "GLIPR1"   "GNLY"     "GPR183"   "GZMA"     "GZMB"    </span></span>
<span><span class="co"># [139] "GZMK"     "HAVCR2"   "HDC"      "HMGA1"    "HOOK2"    "HOXD8"   </span></span>
<span><span class="co"># [145] "HOXD9"    "HPX"      "IGF1"     "IGSF6"    "IL2RA"    "IL2RG"   </span></span>
<span><span class="co"># [151] "IL3RA"    "IL7R"     "ITGAM"    "ITGAX"    "ITM2C"    "JUP"     </span></span>
<span><span class="co"># [157] "KARS"     "KDR"      "KIT"      "KLF5"     "KLRB1"    "KLRC1"   </span></span>
<span><span class="co"># [163] "KLRD1"    "KLRF1"    "KRT14"    "KRT15"    "KRT16"    "KRT23"   </span></span>
<span><span class="co"># [169] "KRT5"     "KRT6B"    "KRT7"     "KRT8"     "LAG3"     "LARS"    </span></span>
<span><span class="co"># [175] "LDHB"     "LEP"      "LGALSL"   "LIF"      "LILRA4"   "LPL"     </span></span>
<span><span class="co"># [181] "LPXN"     "LRRC15"   "LTB"      "LUM"      "LY86"     "LYPD3"   </span></span>
<span><span class="co"># [187] "LYZ"      "MAP3K8"   "MDM2"     "MEDAG"    "MKI67"    "MLPH"    </span></span>
<span><span class="co"># [193] "MMP1"     "MMP12"    "MMP2"     "MMRN2"    "MNDA"     "MPO"     </span></span>
<span><span class="co"># [199] "MRC1"     "MS4A1"    "MUC6"     "MYBPC1"   "MYH11"    "MYLK"    </span></span>
<span><span class="co"># [205] "MYO5B"    "MZB1"     "NARS"     "NCAM1"    "NDUFA4L2" "NKG7"    </span></span>
<span><span class="co"># [211] "NOSTRIN"  "NPM3"     "OCIAD2"   "OPRPN"    "OXTR"     "PCLAF"   </span></span>
<span><span class="co"># [217] "PCOLCE"   "PDCD1"    "PDCD1LG2" "PDE4A"    "PDGFRA"   "PDGFRB"  </span></span>
<span><span class="co"># [223] "PDK4"     "PECAM1"   "PELI1"    "PGR"      "PIGR"     "PIM1"    </span></span>
<span><span class="co"># [229] "PLD4"     "POLR2J3"  "POSTN"    "PPARG"    "PRDM1"    "PRF1"    </span></span>
<span><span class="co"># [235] "PTGDS"    "PTN"      "PTPRC"    "PTRHD1"   "QARS"     "RAB30"   </span></span>
<span><span class="co"># [241] "RAMP2"    "RAPGEF3"  "REXO4"    "RHOH"     "RORC"     "RTKN2"   </span></span>
<span><span class="co"># [247] "RUNX1"    "S100A14"  "S100A4"   "S100A8"   "SCD"      "SCGB2A1" </span></span>
<span><span class="co"># [253] "SDC4"     "SEC11C"   "SEC24A"   "SELL"     "SERHL2"   "SERPINA3"</span></span>
<span><span class="co"># [259] "SERPINB9" "SFRP1"    "SFRP4"    "SH3YL1"   "SLAMF1"   "SLAMF7"  </span></span>
<span><span class="co"># [265] "SLC25A37" "SLC4A1"   "SLC5A6"   "SMAP2"    "SMS"      "SNAI1"   </span></span>
<span><span class="co"># [271] "SOX17"    "SOX18"    "SPIB"     "SQLE"     "SRPK1"    "SSTR2"   </span></span>
<span><span class="co"># [277] "STC1"     "SVIL"     "TAC1"     "TACSTD2"  "TCEAL7"   "TCF15"   </span></span>
<span><span class="co"># [283] "TCF4"     "TCF7"     "TCIM"     "TCL1A"    "TENT5C"   "TFAP2A"  </span></span>
<span><span class="co"># [289] "THAP2"    "TIFA"     "TIGIT"    "TIMP4"    "TMEM147"  "TNFRSF17"</span></span>
<span><span class="co"># [295] "TOMM7"    "TOP2A"    "TPD52"    "TPSAB1"   "TRAC"     "TRAF4"   </span></span>
<span><span class="co"># [301] "TRAPPC3"  "TRIB1"    "TUBA4A"   "TUBB2B"   "TYROBP"   "UCP1"    </span></span>
<span><span class="co"># [307] "USP53"    "VOPP1"    "VWF"      "WARS"     "ZEB1"     "ZEB2"    </span></span>
<span><span class="co"># [313] "ZNF562"  </span></span></code></pre></div>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#   [1] "BLANK_0006" "BLANK_0013" "BLANK_0037" "BLANK_0069" "BLANK_0072"</span></span>
<span><span class="co">#   [6] "BLANK_0087" "BLANK_0110" "BLANK_0114" "BLANK_0120" "BLANK_0147"</span></span>
<span><span class="co">#  [11] "BLANK_0180" "BLANK_0186" "BLANK_0272" "BLANK_0278" "BLANK_0319"</span></span>
<span><span class="co">#  [16] "BLANK_0321" "BLANK_0337" "BLANK_0350" "BLANK_0351" "BLANK_0352"</span></span>
<span><span class="co">#  [21] "BLANK_0353" "BLANK_0354" "BLANK_0355" "BLANK_0356" "BLANK_0357"</span></span>
<span><span class="co">#  [26] "BLANK_0358" "BLANK_0359" "BLANK_0360" "BLANK_0361" "BLANK_0362"</span></span>
<span><span class="co">#  [31] "BLANK_0363" "BLANK_0364" "BLANK_0365" "BLANK_0366" "BLANK_0367"</span></span>
<span><span class="co">#  [36] "BLANK_0368" "BLANK_0369" "BLANK_0370" "BLANK_0371" "BLANK_0372"</span></span>
<span><span class="co">#  [41] "BLANK_0373" "BLANK_0374" "BLANK_0375" "BLANK_0376" "BLANK_0377"</span></span>
<span><span class="co">#  [46] "BLANK_0378" "BLANK_0379" "BLANK_0380" "BLANK_0381" "BLANK_0382"</span></span>
<span><span class="co">#  [51] "BLANK_0383" "BLANK_0384" "BLANK_0385" "BLANK_0386" "BLANK_0387"</span></span>
<span><span class="co">#  [56] "BLANK_0388" "BLANK_0389" "BLANK_0390" "BLANK_0391" "BLANK_0392"</span></span>
<span><span class="co">#  [61] "BLANK_0393" "BLANK_0394" "BLANK_0395" "BLANK_0396" "BLANK_0397"</span></span>
<span><span class="co">#  [66] "BLANK_0398" "BLANK_0399" "BLANK_0400" "BLANK_0401" "BLANK_0402"</span></span>
<span><span class="co">#  [71] "BLANK_0403" "BLANK_0404" "BLANK_0405" "BLANK_0406" "BLANK_0407"</span></span>
<span><span class="co">#  [76] "BLANK_0408" "BLANK_0409" "BLANK_0410" "BLANK_0411" "BLANK_0412"</span></span>
<span><span class="co">#  [81] "BLANK_0413" "BLANK_0414" "BLANK_0415" "BLANK_0416" "BLANK_0417"</span></span>
<span><span class="co">#  [86] "BLANK_0418" "BLANK_0419" "BLANK_0420" "BLANK_0421" "BLANK_0422"</span></span>
<span><span class="co">#  [91] "BLANK_0423" "BLANK_0424" "BLANK_0425" "BLANK_0426" "BLANK_0427"</span></span>
<span><span class="co">#  [96] "BLANK_0428" "BLANK_0429" "BLANK_0430" "BLANK_0431" "BLANK_0432"</span></span>
<span><span class="co"># [101] "BLANK_0433" "BLANK_0434" "BLANK_0435" "BLANK_0436" "BLANK_0437"</span></span>
<span><span class="co"># [106] "BLANK_0438" "BLANK_0439" "BLANK_0440" "BLANK_0441" "BLANK_0442"</span></span>
<span><span class="co"># [111] "BLANK_0443" "BLANK_0444" "BLANK_0445" "BLANK_0446" "BLANK_0447"</span></span>
<span><span class="co"># [116] "BLANK_0448" "BLANK_0449" "BLANK_0450" "BLANK_0451" "BLANK_0452"</span></span>
<span><span class="co"># [121] "BLANK_0453" "BLANK_0454" "BLANK_0455" "BLANK_0456" "BLANK_0457"</span></span>
<span><span class="co"># [126] "BLANK_0458" "BLANK_0459" "BLANK_0460" "BLANK_0461" "BLANK_0462"</span></span>
<span><span class="co"># [131] "BLANK_0463" "BLANK_0464" "BLANK_0465" "BLANK_0466" "BLANK_0467"</span></span>
<span><span class="co"># [136] "BLANK_0468" "BLANK_0469" "BLANK_0470" "BLANK_0471" "BLANK_0472"</span></span>
<span><span class="co"># [141] "BLANK_0473" "BLANK_0474" "BLANK_0475" "BLANK_0476" "BLANK_0477"</span></span>
<span><span class="co"># [146] "BLANK_0478" "BLANK_0479" "BLANK_0480" "BLANK_0481" "BLANK_0482"</span></span>
<span><span class="co"># [151] "BLANK_0483" "BLANK_0484" "BLANK_0485" "BLANK_0486" "BLANK_0487"</span></span>
<span><span class="co"># [156] "BLANK_0488" "BLANK_0489" "BLANK_0497" "BLANK_0499"</span></span></code></pre></div>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#  [1] "NegControlCodeword_0500" "NegControlCodeword_0501"</span></span>
<span><span class="co">#  [3] "NegControlCodeword_0502" "NegControlCodeword_0503"</span></span>
<span><span class="co">#  [5] "NegControlCodeword_0504" "NegControlCodeword_0505"</span></span>
<span><span class="co">#  [7] "NegControlCodeword_0506" "NegControlCodeword_0507"</span></span>
<span><span class="co">#  [9] "NegControlCodeword_0508" "NegControlCodeword_0509"</span></span>
<span><span class="co"># [11] "NegControlCodeword_0510" "NegControlCodeword_0511"</span></span>
<span><span class="co"># [13] "NegControlCodeword_0512" "NegControlCodeword_0513"</span></span>
<span><span class="co"># [15] "NegControlCodeword_0514" "NegControlCodeword_0515"</span></span>
<span><span class="co"># [17] "NegControlCodeword_0516" "NegControlCodeword_0517"</span></span>
<span><span class="co"># [19] "NegControlCodeword_0518" "NegControlCodeword_0519"</span></span>
<span><span class="co"># [21] "NegControlCodeword_0520" "NegControlCodeword_0521"</span></span>
<span><span class="co"># [23] "NegControlCodeword_0522" "NegControlCodeword_0523"</span></span>
<span><span class="co"># [25] "NegControlCodeword_0524" "NegControlCodeword_0525"</span></span>
<span><span class="co"># [27] "NegControlCodeword_0526" "NegControlCodeword_0527"</span></span>
<span><span class="co"># [29] "NegControlCodeword_0528" "NegControlCodeword_0529"</span></span>
<span><span class="co"># [31] "NegControlCodeword_0530" "NegControlCodeword_0531"</span></span>
<span><span class="co"># [33] "NegControlCodeword_0532" "NegControlCodeword_0533"</span></span>
<span><span class="co"># [35] "NegControlCodeword_0534" "NegControlCodeword_0535"</span></span>
<span><span class="co"># [37] "NegControlCodeword_0536" "NegControlCodeword_0537"</span></span>
<span><span class="co"># [39] "NegControlCodeword_0538" "NegControlCodeword_0539"</span></span>
<span><span class="co"># [41] "NegControlCodeword_0540"</span></span></code></pre></div>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#  [1] "NegControlProbe_00042" "NegControlProbe_00041" "NegControlProbe_00039"</span></span>
<span><span class="co">#  [4] "NegControlProbe_00035" "NegControlProbe_00034" "NegControlProbe_00033"</span></span>
<span><span class="co">#  [7] "NegControlProbe_00031" "NegControlProbe_00025" "NegControlProbe_00024"</span></span>
<span><span class="co"># [10] "NegControlProbe_00022" "NegControlProbe_00019" "NegControlProbe_00017"</span></span>
<span><span class="co"># [13] "NegControlProbe_00016" "NegControlProbe_00014" "NegControlProbe_00013"</span></span>
<span><span class="co"># [16] "NegControlProbe_00012" "NegControlProbe_00009" "NegControlProbe_00004"</span></span>
<span><span class="co"># [19] "NegControlProbe_00003" "NegControlProbe_00002" "antisense_PROKR2"     </span></span>
<span><span class="co"># [22] "antisense_ULK3"        "antisense_SCRIB"       "antisense_TRMU"       </span></span>
<span><span class="co"># [25] "antisense_MYLIP"       "antisense_LGI3"        "antisense_BCL2L15"    </span></span>
<span><span class="co"># [28] "antisense_ADCY4"   </span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="loading-xenium-data">5. Loading Xenium data<a class="anchor" aria-label="anchor" href="#loading-xenium-data"></a>
</h2>
<div class="section level3">
<h3 id="manual-method">5.1 Manual Method<a class="anchor" aria-label="anchor" href="#manual-method"></a>
</h3>
<p>Giotto objects can be manually assembled feeding data and subobjects
into a creation function. A convenience function for automatically
loading the xenium data from the directory and generating a giotto
object is also available. (See Section 5.2])</p>
<p>Xenium outputs can be analyzed as either the subcellular information
or as aggregated data where each detected cellâ€™s subcellular data has
been spatially assigned to a cell centroid. This tutorial will work
mainly with the subcellular data and how to work with it, however a
workflow to load in just the aggregated data is also available through
the convenience function.</p>
<div class="section level4">
<h4 id="load-transcript-level-data">5.1.1 Load transcript-level data<a class="anchor" aria-label="anchor" href="#load-transcript-level-data"></a>
</h4>
<p><code>transcripts.csv.gz</code> is a file containing x, y, z
coordinates for individual transcript molecules detected during the
Xenium run. It also contains a QC Phred score for which this tutorial
will set a cutoff at 20, the same as what 10x uses.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tx_dt</span> <span class="op">=</span> <span class="fu">data.table</span><span class="fu">::</span><span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="va">tx_path</span><span class="op">)</span></span>
<span><span class="fu">data.table</span><span class="fu">::</span><span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/setattr.html" class="external-link">setnames</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">tx_dt</span>,</span>
<span>                     old <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'feature_name'</span>, <span class="st">'x_location'</span>, <span class="st">'y_location'</span><span class="op">)</span>,</span>
<span>                     new <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'feat_ID'</span>, <span class="st">'x'</span>, <span class="st">'y'</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="st">'Transcripts info available:\n '</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">'"'</span>, <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/row-plus-colnames-generic.html">colnames</a></span><span class="op">(</span><span class="va">tx_dt</span><span class="op">)</span>, <span class="st">'"'</span><span class="op">)</span>, <span class="st">'\n'</span>,</span>
<span>    <span class="st">'with'</span>, <span class="va">tx_dt</span><span class="op">[</span>,<span class="va">.N</span><span class="op">]</span>, <span class="st">'unfiltered detections\n'</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># filter by qv (Phred score)</span></span>
<span><span class="va">tx_dt_filtered</span> <span class="op">=</span> <span class="va">tx_dt</span><span class="op">[</span><span class="va">qv</span> <span class="op">&gt;=</span> <span class="fl">20</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="st">'and'</span>, <span class="va">tx_dt_filtered</span><span class="op">[</span>,<span class="va">.N</span><span class="op">]</span>, <span class="st">'filtered detections\n\n'</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># separate detections by feature type</span></span>
<span><span class="va">tx_dt_types</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span></span>
<span>  <span class="va">feat_types_IDs</span>, <span class="kw">function</span><span class="op">(</span><span class="va">types</span><span class="op">)</span> <span class="va">tx_dt_filtered</span><span class="op">[</span><span class="va">feat_ID</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">types</span><span class="op">]</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/invisible.html" class="external-link">invisible</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_along</a></span><span class="op">(</span><span class="va">tx_dt_types</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">tx_dt_types</span><span class="op">)</span><span class="op">[[</span><span class="va">x</span><span class="op">]</span><span class="op">]</span>, <span class="st">'detections: '</span>, <span class="va">tx_dt_types</span><span class="op">[[</span><span class="va">x</span><span class="op">]</span><span class="op">]</span><span class="op">[</span>,<span class="va">.N</span><span class="op">]</span>, <span class="st">'\n'</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Transcripts info available:</span></span>
<span><span class="co">#   "transcript_id" "cell_id" "overlaps_nucleus" "feat_ID" "x" "y" "z_location" "qv" </span></span>
<span><span class="co">#  with 43664530 unfiltered detections</span></span>
<span><span class="co">#  and 34813341 filtered detections</span></span>
<span><span class="co"># </span></span>
<span><span class="co"># Blank Codeword detections: 8805 </span></span>
<span><span class="co"># Gene Expression detections: 34764833 </span></span>
<span><span class="co"># Negative Control Codeword detections: 1855 </span></span>
<span><span class="co"># Negative Control Probe detections: 37848 </span></span></code></pre></div>
<p>Giotto loads these filtered subcellular detections in as a
<code>giottoPoints</code> object and determines the correct columns by
looking for columns named <code>'feat_ID',</code> <code>'x'</code>, and
<code>'y'</code>.</p>
<p>Here, we use the list of <code>data.table</code>s generated in the
previous step to create a list of <code>giottoPoints</code> objects.
When previewing these objects using <code><a href="https://rdrr.io/r/base/plot.html" class="external-link">plot()</a></code>, the default
behavior is to plot ALL points within the object. For objects that
contain many feature points, it is highly recommended to specify a
subset of features to plot using the <code>feats</code> param.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gpoints_list</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span></span>
<span>  <span class="va">tx_dt_types</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/createGiottoPoints.html">createGiottoPoints</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span><span class="op">)</span></span>
<span><span class="op">)</span> <span class="co"># 208.499 sec elapsed</span></span>
<span></span>
<span><span class="co"># preview QC probe detections</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">gpoints_list</span><span class="op">$</span><span class="va">`Blank Codeword`</span>,</span>
<span>     point_size <span class="op">=</span> <span class="fl">0.3</span>,</span>
<span>     main <span class="op">=</span> <span class="st">'Blank Codeword'</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">gpoints_list</span><span class="op">$</span><span class="va">`Negative Control Codeword`</span>,</span>
<span>     point_size <span class="op">=</span> <span class="fl">0.3</span>,</span>
<span>     main <span class="op">=</span> <span class="st">'Negative Control Codeword'</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">gpoints_list</span><span class="op">$</span><span class="va">`Negative Control Probe`</span>,</span>
<span>     point_size <span class="op">=</span> <span class="fl">0.3</span>,</span>
<span>     main <span class="op">=</span> <span class="st">'Negative Control Probe'</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># preview two genes (slower)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">gpoints_list</span><span class="op">$</span><span class="va">`Gene Expression`</span>,  <span class="co"># 77.843 sec elapsed</span></span>
<span>     feats <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'KRT8'</span>, <span class="st">'MS4A1'</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">tx_dt_types</span><span class="op">$</span><span class="va">`Gene Expression`</span><span class="op">[</span><span class="va">feat_ID</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'KRT8'</span>, <span class="st">'MS4A1'</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">feat_ID</span><span class="op">)</span><span class="op">]</span></span></code></pre></div>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># feat_ID</span></span>
<span><span class="co">#   KRT8  MS4A1 </span></span>
<span><span class="co"># 530190  20926 </span></span></code></pre></div>
<p><img src="images/xenium_breast_cancer/gpoints_blnk.png"><img src="images/xenium_breast_cancer/gpoints_ngcode.png"><img src="images/xenium_breast_cancer/gpoints_ngprbe.png"><img src="images/xenium_breast_cancer/gpoints_expr.png"></p>
</div>
<div class="section level4">
<h4 id="load-polygon-data">5.1.2 Load polygon data<a class="anchor" aria-label="anchor" href="#load-polygon-data"></a>
</h4>
<p>Xenium output provides segmentation/cell boundary information in
.csv.gz files. These are represented within Giotto as
<code>giottoPolygon</code> objects and can also be directly plotted.
This function also determines the correct columns to use by looking for
columns named <code>'poly_ID'</code>, <code>'x'</code>, and
<code>'y'</code>.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cellPoly_dt</span> <span class="op">=</span> <span class="fu">data.table</span><span class="fu">::</span><span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="va">cell_bound_path</span><span class="op">)</span></span>
<span><span class="va">nucPoly_dt</span> <span class="op">=</span> <span class="fu">data.table</span><span class="fu">::</span><span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="va">nuc_bound_path</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">data.table</span><span class="fu">::</span><span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/setattr.html" class="external-link">setnames</a></span><span class="op">(</span><span class="va">cellPoly_dt</span>,</span>
<span>                     old <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'cell_id'</span>, <span class="st">'vertex_x'</span>, <span class="st">'vertex_y'</span><span class="op">)</span>,</span>
<span>                     new <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'poly_ID'</span>, <span class="st">'x'</span>, <span class="st">'y'</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">data.table</span><span class="fu">::</span><span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/setattr.html" class="external-link">setnames</a></span><span class="op">(</span><span class="va">nucPoly_dt</span>,</span>
<span>                     old <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'cell_id'</span>, <span class="st">'vertex_x'</span>, <span class="st">'vertex_y'</span><span class="op">)</span>,</span>
<span>                     new <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'poly_ID'</span>, <span class="st">'x'</span>, <span class="st">'y'</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">gpoly_cells</span> <span class="op">=</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/createGiottoPolygon.html">createGiottoPolygonsFromDfr</a></span><span class="op">(</span>segmdfr <span class="op">=</span> <span class="va">cellPoly_dt</span>,</span>
<span>                                          name <span class="op">=</span> <span class="st">'cell'</span>,</span>
<span>                                          calc_centroids <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">gpoly_nucs</span> <span class="op">=</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/createGiottoPolygon.html">createGiottoPolygonsFromDfr</a></span><span class="op">(</span>segmdfr <span class="op">=</span> <span class="va">nucPoly_dt</span>,</span>
<span>                                         name <span class="op">=</span> <span class="st">'nucleus'</span>,</span>
<span>                                         calc_centroids <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><code>giottoPolygon</code> objects can be directly plotted with
<code><a href="https://rdrr.io/r/base/plot.html" class="external-link">plot()</a></code>, but the field of view here is so large that it
would take a long time and the details would be lost. Here, we will only
plot the polygon centroids for the cell nucleus polygons by accessing
the calculated results within the <code>giottoPolygon</code>â€™s
<code>spatVectorCentroids</code> slot.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/plot.html" class="external-link">plot</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">gpoly_nucs</span>, point_size <span class="op">=</span> <span class="fl">0.1</span>, type <span class="op">=</span> <span class="st">'centroid'</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/xenium_breast_cancer/gpolys_centroids.png"></p>
</div>
<div class="section level4">
<h4 id="create-giotto-object">5.1.3 Create Giotto Object<a class="anchor" aria-label="anchor" href="#create-giotto-object"></a>
</h4>
<p>Now that both the feature data and the boundaries are loaded in, a
subcellular Giotto object can be created.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">xenium_gobj</span> <span class="op">=</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/createGiottoObjectSubcellular.html">createGiottoObjectSubcellular</a></span><span class="op">(</span></span>
<span>  gpoints <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>rna <span class="op">=</span> <span class="va">gpoints_list</span><span class="op">$</span><span class="va">`Gene Expression`</span>,</span>
<span>                 blank_code <span class="op">=</span> <span class="va">gpoints_list</span><span class="op">$</span><span class="va">`Blank Codeword`</span>,</span>
<span>                 neg_code <span class="op">=</span> <span class="va">gpoints_list</span><span class="op">$</span><span class="va">`Negative Control Codeword`</span>,</span>
<span>                 neg_probe <span class="op">=</span> <span class="va">gpoints_list</span><span class="op">$</span><span class="va">`Negative Control Probe`</span><span class="op">)</span>,</span>
<span>  gpolygons <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>cell <span class="op">=</span> <span class="va">gpoly_cells</span>,</span>
<span>                   nucleus <span class="op">=</span> <span class="va">gpoly_nucs</span><span class="op">)</span>,</span>
<span>  instructions <span class="op">=</span> <span class="va">instrs</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="sec-autoload">5.2 Load dataset - Convenience Function<a class="anchor" aria-label="anchor" href="#sec-autoload"></a>
</h3>
<p>The dataset can also be loaded through a convenience function.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">subcellular</span> <span class="op">=</span> <span class="fu"><a href="../reference/createGiottoXeniumObject.html">createGiottoXeniumObject</a></span><span class="op">(</span>xenium_dir <span class="op">=</span> <span class="va">xenium_folder</span>,</span>
<span>                                       data_to_use <span class="op">=</span> <span class="st">'subcellular'</span>,</span>
<span>                                       bounds_to_load <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'cell'</span>, <span class="st">'nucleus'</span><span class="op">)</span>,</span>
<span>                                       qv_threshold <span class="op">=</span> <span class="fl">20</span>,</span>
<span>                                       h5_expression <span class="op">=</span> <span class="cn">F</span>,</span>
<span>                                       instructions <span class="op">=</span> <span class="va">instrs</span>,</span>
<span>                                       cores <span class="op">=</span> <span class="cn">NA</span><span class="op">)</span> <span class="co"># set number of cores to use</span></span></code></pre></div>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># A structured Xenium directory will be used</span></span>
<span><span class="co"># </span></span>
<span><span class="co"># Checking directory contents...</span></span>
<span><span class="co">#    &gt; analysis info found</span></span>
<span><span class="co">#       â””â”€â”€Xenium_FFPE_Human_Breast_Cancer_Rep1_analysis.tar.gz</span></span>
<span><span class="co">#       â””â”€â”€Xenium_FFPE_Human_Breast_Cancer_Rep1_analysis.zarr.zip</span></span>
<span><span class="co">#    &gt; boundary info found</span></span>
<span><span class="co">#       â””â”€â”€Xenium_FFPE_Human_Breast_Cancer_Rep1_cell_boundaries.csv.gz</span></span>
<span><span class="co">#       â””â”€â”€Xenium_FFPE_Human_Breast_Cancer_Rep1_cell_boundaries.parquet</span></span>
<span><span class="co">#       â””â”€â”€Xenium_FFPE_Human_Breast_Cancer_Rep1_nucleus_boundaries.csv.gz</span></span>
<span><span class="co">#       â””â”€â”€Xenium_FFPE_Human_Breast_Cancer_Rep1_nucleus_boundaries.parquet</span></span>
<span><span class="co">#    &gt; cell feature matrix found</span></span>
<span><span class="co">#       â””â”€â”€cell_feature_matrix</span></span>
<span><span class="co">#       â””â”€â”€Xenium_FFPE_Human_Breast_Cancer_Rep1_cell_feature_matrix.h5</span></span>
<span><span class="co">#       â””â”€â”€Xenium_FFPE_Human_Breast_Cancer_Rep1_cell_feature_matrix.zarr.zip</span></span>
<span><span class="co">#    &gt; cell metadata found</span></span>
<span><span class="co">#       â””â”€â”€Xenium_FFPE_Human_Breast_Cancer_Rep1_cells.csv.gz</span></span>
<span><span class="co">#       â””â”€â”€Xenium_FFPE_Human_Breast_Cancer_Rep1_cells.parquet</span></span>
<span><span class="co">#       â””â”€â”€Xenium_FFPE_Human_Breast_Cancer_Rep1_cells.zarr.zip</span></span>
<span><span class="co">#    &gt; image info found</span></span>
<span><span class="co">#       â””â”€â”€Xenium_FFPE_Human_Breast_Cancer_Rep1_he_image.tif</span></span>
<span><span class="co">#       â””â”€â”€Xenium_FFPE_Human_Breast_Cancer_Rep1_morphology_focus.ome.tif</span></span>
<span><span class="co">#       â””â”€â”€Xenium_FFPE_Human_Breast_Cancer_Rep1_morphology_mip.ome.tif</span></span>
<span><span class="co">#       â””â”€â”€Xenium_FFPE_Human_Breast_Cancer_Rep1_morphology.ome.tif</span></span>
<span><span class="co">#    &gt; panel metadata found</span></span>
<span><span class="co">#       â””â”€â”€Xenium_FFPE_Human_Breast_Cancer_Rep1_panel.tsv</span></span>
<span><span class="co">#    &gt; raw transcript info found</span></span>
<span><span class="co">#       â””â”€â”€Xenium_FFPE_Human_Breast_Cancer_Rep1_transcripts.csv.gz</span></span>
<span><span class="co">#       â””â”€â”€Xenium_FFPE_Human_Breast_Cancer_Rep1_transcripts.parquet</span></span>
<span><span class="co">#       â””â”€â”€Xenium_FFPE_Human_Breast_Cancer_Rep1_transcripts.zarr.zip</span></span>
<span><span class="co">#    &gt; experiment info found</span></span>
<span><span class="co">#       â””â”€â”€Xenium_FFPE_Human_Breast_Cancer_Rep1_experiment.xenium</span></span>
<span><span class="co"># Directory check done</span></span>
<span><span class="co"># Loading feature metadata...</span></span>
<span><span class="co"># Loading transcript level info...</span></span>
<span><span class="co"># |--------------------------------------------------|</span></span>
<span><span class="co"># |==================================================|</span></span>
<span><span class="co"># Loading boundary info...</span></span>
<span><span class="co"># Loading cell metadata...</span></span>
<span><span class="co"># Building subcellular giotto object...</span></span>
<span><span class="co"># &gt; points data...</span></span>
<span><span class="co">#   Selecting col "feature_name" as feat_ID column</span></span>
<span><span class="co">#   Selecting cols "x_location" and "y_location" as x and y respectively</span></span>
<span><span class="co"># &gt; polygons data...</span></span>
<span><span class="co">#   [cell] bounds...</span></span>
<span><span class="co">#   Selecting col "cell_id" as poly_ID column</span></span>
<span><span class="co">#   Selecting cols "vertex_x" and "vertex_y" as x and y respectively</span></span>
<span><span class="co">#   [nucleus] bounds...</span></span>
<span><span class="co">#   Selecting col "cell_id" as poly_ID column</span></span>
<span><span class="co">#   Selecting cols "vertex_x" and "vertex_y" as x and y respectively</span></span>
<span><span class="co"># 1. Start extracting polygon information </span></span>
<span><span class="co"># 2. Finished extracting polygon information </span></span>
<span><span class="co"># 3. Add centroid / spatial locations if available </span></span>
<span><span class="co"># 3. Finish adding centroid / spatial locations </span></span>
<span><span class="co"># 3. Start extracting spatial feature information </span></span>
<span><span class="co"># 4. Finished extracting spatial feature information </span></span>
<span><span class="co"># Calculating polygon centroids...</span></span>
<span><span class="co"># Start centroid calculation for polygon information layer:  cell </span></span>
<span><span class="co"># Start centroid calculation for polygon information layer:  nucleus </span></span></code></pre></div>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">aggregate</span> <span class="op">=</span> <span class="fu"><a href="../reference/createGiottoXeniumObject.html">createGiottoXeniumObject</a></span><span class="op">(</span>xenium_dir <span class="op">=</span> <span class="va">xenium_folder</span>,</span>
<span>                                     data_to_use <span class="op">=</span> <span class="st">'aggregate'</span>,</span>
<span>                                     h5_expression <span class="op">=</span> <span class="cn">F</span>,</span>
<span>                                     instructions <span class="op">=</span> <span class="va">instrs</span>,</span>
<span>                                     cores <span class="op">=</span> <span class="cn">NA</span><span class="op">)</span> <span class="co"># set number of cores to use</span></span></code></pre></div>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># A structured Xenium directory will be used</span></span>
<span><span class="co"># </span></span>
<span><span class="co"># Checking directory contents...</span></span>
<span><span class="co">#    &gt; analysis info found</span></span>
<span><span class="co">#       â””â”€â”€Xenium_FFPE_Human_Breast_Cancer_Rep1_analysis.tar.gz</span></span>
<span><span class="co">#       â””â”€â”€Xenium_FFPE_Human_Breast_Cancer_Rep1_analysis.zarr.zip</span></span>
<span><span class="co">#    &gt; boundary info found</span></span>
<span><span class="co">#       â””â”€â”€Xenium_FFPE_Human_Breast_Cancer_Rep1_cell_boundaries.csv.gz</span></span>
<span><span class="co">#       â””â”€â”€Xenium_FFPE_Human_Breast_Cancer_Rep1_cell_boundaries.parquet</span></span>
<span><span class="co">#       â””â”€â”€Xenium_FFPE_Human_Breast_Cancer_Rep1_nucleus_boundaries.csv.gz</span></span>
<span><span class="co">#       â””â”€â”€Xenium_FFPE_Human_Breast_Cancer_Rep1_nucleus_boundaries.parquet</span></span>
<span><span class="co">#    &gt; cell feature matrix found</span></span>
<span><span class="co">#       â””â”€â”€cell_feature_matrix</span></span>
<span><span class="co">#       â””â”€â”€Xenium_FFPE_Human_Breast_Cancer_Rep1_cell_feature_matrix.h5</span></span>
<span><span class="co">#       â””â”€â”€Xenium_FFPE_Human_Breast_Cancer_Rep1_cell_feature_matrix.zarr.zip</span></span>
<span><span class="co">#    &gt; cell metadata found</span></span>
<span><span class="co">#       â””â”€â”€Xenium_FFPE_Human_Breast_Cancer_Rep1_cells.csv.gz</span></span>
<span><span class="co">#       â””â”€â”€Xenium_FFPE_Human_Breast_Cancer_Rep1_cells.parquet</span></span>
<span><span class="co">#       â””â”€â”€Xenium_FFPE_Human_Breast_Cancer_Rep1_cells.zarr.zip</span></span>
<span><span class="co">#    &gt; image info found</span></span>
<span><span class="co">#       â””â”€â”€Xenium_FFPE_Human_Breast_Cancer_Rep1_he_image.tif</span></span>
<span><span class="co">#       â””â”€â”€Xenium_FFPE_Human_Breast_Cancer_Rep1_morphology_focus.ome.tif</span></span>
<span><span class="co">#       â””â”€â”€Xenium_FFPE_Human_Breast_Cancer_Rep1_morphology_mip.ome.tif</span></span>
<span><span class="co">#       â””â”€â”€Xenium_FFPE_Human_Breast_Cancer_Rep1_morphology.ome.tif</span></span>
<span><span class="co">#    &gt; panel metadata found</span></span>
<span><span class="co">#       â””â”€â”€Xenium_FFPE_Human_Breast_Cancer_Rep1_panel.tsv</span></span>
<span><span class="co">#    &gt; raw transcript info found</span></span>
<span><span class="co">#       â””â”€â”€Xenium_FFPE_Human_Breast_Cancer_Rep1_transcripts.csv.gz</span></span>
<span><span class="co">#       â””â”€â”€Xenium_FFPE_Human_Breast_Cancer_Rep1_transcripts.parquet</span></span>
<span><span class="co">#       â””â”€â”€Xenium_FFPE_Human_Breast_Cancer_Rep1_transcripts.zarr.zip</span></span>
<span><span class="co">#    &gt; experiment info found</span></span>
<span><span class="co">#       â””â”€â”€Xenium_FFPE_Human_Breast_Cancer_Rep1_experiment.xenium</span></span>
<span><span class="co"># Directory check done</span></span>
<span><span class="co"># Loading feature metadata...</span></span>
<span><span class="co"># Loading cell metadata...</span></span>
<span><span class="co"># Loading aggregated expression...</span></span>
<span><span class="co"># Building aggregate giotto object...</span></span>
<span><span class="co"># Consider to install these (optional) packages to run all possible</span></span>
<span><span class="co"># Giotto commands for spatial analyses: trendsceek multinet RTriangle</span></span>
<span><span class="co"># Giotto does not automatically install all these packages as they are</span></span>
<span><span class="co"># not absolutely required and this reduces the number of dependencies</span></span>
<span><span class="co"># List of 4</span></span>
<span><span class="co">#  $ raw                      :Formal class 'dgTMatrix' [package "Matrix"] with 6 slots</span></span>
<span><span class="co">#   .. ..@ i       : int [1:10663545] 1 2 29 30 36 37 39 42 44 47 ...</span></span>
<span><span class="co">#   .. ..@ j       : int [1:10663545] 0 0 0 0 0 0 0 0 0 0 ...</span></span>
<span><span class="co">#   .. ..@ Dim     : int [1:2] 313 167782</span></span>
<span><span class="co">#   .. ..@ Dimnames:List of 2</span></span>
<span><span class="co">#   .. .. ..$ : Named chr [1:313] "ABCC11" "ACTA2" "ACTG2" "ADAM9" ...</span></span>
<span><span class="co">#   .. .. .. ..- attr(*, "names")= chr [1:313] "1" "2" "3" "4" ...</span></span>
<span><span class="co">#   .. .. ..$ : chr [1:167782] "1" "2" "3" "4" ...</span></span>
<span><span class="co">#   .. ..@ x       : num [1:10663545] 3 1 1 3 1 1 1 1 1 1 ...</span></span>
<span><span class="co">#   .. ..@ factors : list()</span></span>
<span><span class="co">#  $ Negative_Control_Probe   :Formal class 'dgTMatrix' [package "Matrix"] with 6 slots</span></span>
<span><span class="co">#   .. ..@ i       : int [1:24312] 18 20 22 22 22 22 20 25 26 20 ...</span></span>
<span><span class="co">#   .. ..@ j       : int [1:24312] 48 79 79 130 175 205 214 215 223 248 ...</span></span>
<span><span class="co">#   .. ..@ Dim     : int [1:2] 28 167782</span></span>
<span><span class="co">#   .. ..@ Dimnames:List of 2</span></span>
<span><span class="co">#   .. .. ..$ : Named chr [1:28] "NegControlProbe_00042" "NegControlProbe_00041" "NegControlProbe_00039" "NegControlProbe_00035" ...</span></span>
<span><span class="co">#   .. .. .. ..- attr(*, "names")= chr [1:28] "314" "315" "316" "317" ...</span></span>
<span><span class="co">#   .. .. ..$ : chr [1:167782] "1" "2" "3" "4" ...</span></span>
<span><span class="co">#   .. ..@ x       : num [1:24312] 1 1 1 1 1 1 1 1 1 1 ...</span></span>
<span><span class="co">#   .. ..@ factors : list()</span></span>
<span><span class="co">#  $ Negative_Control_Codeword:Formal class 'dgTMatrix' [package "Matrix"] with 6 slots</span></span>
<span><span class="co">#   .. ..@ i       : int [1:1777] 33 14 3 31 3 8 24 10 22 27 ...</span></span>
<span><span class="co">#   .. ..@ j       : int [1:1777] 177 373 381 583 605 673 733 850 924 1033 ...</span></span>
<span><span class="co">#   .. ..@ Dim     : int [1:2] 41 167782</span></span>
<span><span class="co">#   .. ..@ Dimnames:List of 2</span></span>
<span><span class="co">#   .. .. ..$ : Named chr [1:41] "NegControlCodeword_0500" "NegControlCodeword_0501" "NegControlCodeword_0502" "NegControlCodeword_0503" ...</span></span>
<span><span class="co">#   .. .. .. ..- attr(*, "names")= chr [1:41] "342" "343" "344" "345" ...</span></span>
<span><span class="co">#   .. .. ..$ : chr [1:167782] "1" "2" "3" "4" ...</span></span>
<span><span class="co">#   .. ..@ x       : num [1:1777] 1 1 1 1 1 1 1 1 1 1 ...</span></span>
<span><span class="co">#   .. ..@ factors : list()</span></span>
<span><span class="co">#  $ Blank_Codeword           :Formal class 'dgTMatrix' [package "Matrix"] with 6 slots</span></span>
<span><span class="co">#   .. ..@ i       : int [1:8404] 114 76 33 7 34 8 119 91 6 116 ...</span></span>
<span><span class="co">#   .. ..@ j       : int [1:8404] 83 111 114 138 183 245 312 322 338 354 ...</span></span>
<span><span class="co">#   .. ..@ Dim     : int [1:2] 159 167782</span></span>
<span><span class="co">#   .. ..@ Dimnames:List of 2</span></span>
<span><span class="co">#   .. .. ..$ : Named chr [1:159] "BLANK_0006" "BLANK_0013" "BLANK_0037" "BLANK_0069" ...</span></span>
<span><span class="co">#   .. .. .. ..- attr(*, "names")= chr [1:159] "383" "384" "385" "386" ...</span></span>
<span><span class="co">#   .. .. ..$ : chr [1:167782] "1" "2" "3" "4" ...</span></span>
<span><span class="co">#   .. ..@ x       : num [1:8404] 1 1 1 1 1 1 1 1 1 1 ...</span></span>
<span><span class="co">#   .. ..@ factors : list()</span></span>
<span><span class="co"># NULL</span></span>
<span><span class="co"># list depth of 1 </span></span>
<span><span class="co"># finished expression data</span></span>
<span><span class="co"># List of 1</span></span>
<span><span class="co">#  $ raw:Classes â€˜data.tableâ€™ and 'data.frame': 167782 obs. of  3 variables:</span></span>
<span><span class="co">#   ..$ x_centroid: num [1:167782] 378 382 320 259 371 ...</span></span>
<span><span class="co">#   ..$ y_centroid: num [1:167782] 844 859 869 852 865 ...</span></span>
<span><span class="co">#   ..$ cell_ID   : chr [1:167782] "1" "2" "3" "4" ...</span></span>
<span><span class="co">#   ..- attr(*, ".internal.selfref")=&lt;externalptr&gt; </span></span>
<span><span class="co"># NULL</span></span>
<span><span class="co"># list depth of 1 </span></span>
<span><span class="co"># There are non numeric or integer columns for the spatial location input at column position(s): 3</span></span>
<span><span class="co">#  The first non-numeric column will be considered as a cell ID to test for consistency with the expression matrix</span></span>
<span><span class="co">#  Other non numeric columns will be removed</span></span>
<span><span class="co"># finished spatial location data</span></span>
<span><span class="co"># finished cell metadata</span></span>
<span><span class="co"># No spatial networks are provided</span></span>
<span><span class="co"># No spatial enrichment results are provided</span></span>
<span><span class="co"># No dimension reduction results are provided</span></span>
<span><span class="co"># No nearest network results are provided</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="visualize-giotto-object-and-cells">6. Visualize Giotto object and cells<a class="anchor" aria-label="anchor" href="#visualize-giotto-object-and-cells"></a>
</h2>
<div class="section level3">
<h3 id="spatial-info">6.1 Spatial info<a class="anchor" aria-label="anchor" href="#spatial-info"></a>
</h3>
<p>Print the available spatial cell and nucleus boundary information
(polygons) within the Giotto object <code>spatial_info</code> slot.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/showGiottoSpatialInfo.html">showGiottoSpatialInfo</a></span><span class="op">(</span><span class="va">xenium_gobj</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># For Spatial info:  cell </span></span>
<span><span class="co"># </span></span>
<span><span class="co"># An object of class "giottoPolygon"</span></span>
<span><span class="co"># Slot "name":</span></span>
<span><span class="co"># [1] "cell"</span></span>
<span><span class="co"># </span></span>
<span><span class="co"># Slot "spatVector":</span></span>
<span><span class="co">#  class       : SpatVector </span></span>
<span><span class="co">#  geometry    : polygons </span></span>
<span><span class="co">#  dimensions  : 167782, 1  (geometries, attributes)</span></span>
<span><span class="co">#  extent      : 0, 7525.9, 0, 5478.038  (xmin, xmax, ymin, ymax)</span></span>
<span><span class="co">#  coord. ref. :  </span></span>
<span><span class="co"># </span></span>
<span><span class="co"># Slot "spatVectorCentroids":</span></span>
<span><span class="co">#  class       : SpatVector </span></span>
<span><span class="co">#  geometry    : points </span></span>
<span><span class="co">#  dimensions  : 167782, 1  (geometries, attributes)</span></span>
<span><span class="co">#  extent      : 2.189156, 7523.163, 1.406448, 5476.467  (xmin, xmax, ymin, ymax)</span></span>
<span><span class="co">#  coord. ref. :  </span></span>
<span><span class="co"># </span></span>
<span><span class="co"># Slot "overlaps":</span></span>
<span><span class="co"># NULL</span></span>
<span><span class="co"># </span></span>
<span><span class="co"># -----------------------------</span></span>
<span><span class="co">#  </span></span>
<span><span class="co"># For Spatial info:  nucleus </span></span>
<span><span class="co"># </span></span>
<span><span class="co"># An object of class "giottoPolygon"</span></span>
<span><span class="co"># Slot "name":</span></span>
<span><span class="co"># [1] "nucleus"</span></span>
<span><span class="co"># </span></span>
<span><span class="co"># Slot "spatVector":</span></span>
<span><span class="co">#  class       : SpatVector </span></span>
<span><span class="co">#  geometry    : polygons </span></span>
<span><span class="co">#  dimensions  : 167782, 1  (geometries, attributes)</span></span>
<span><span class="co">#  extent      : 1.4875, 7524.413, 0, 5478.038  (xmin, xmax, ymin, ymax)</span></span>
<span><span class="co">#  coord. ref. :  </span></span>
<span><span class="co"># </span></span>
<span><span class="co"># Slot "spatVectorCentroids":</span></span>
<span><span class="co">#  class       : SpatVector </span></span>
<span><span class="co">#  geometry    : points </span></span>
<span><span class="co">#  dimensions  : 167782, 1  (geometries, attributes)</span></span>
<span><span class="co">#  extent      : 2.596845, 7523.503, 0.8111559, 5477.374  (xmin, xmax, ymin, ymax)</span></span>
<span><span class="co">#  coord. ref. :  </span></span>
<span><span class="co"># </span></span>
<span><span class="co"># Slot "overlaps":</span></span>
<span><span class="co"># NULL</span></span>
<span><span class="co"># </span></span>
<span><span class="co"># -----------------------------</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="spatial-locations">6.2 Spatial locations<a class="anchor" aria-label="anchor" href="#spatial-locations"></a>
</h3>
<p>Print the available spatial locations within the Giotto objectâ€™s
<code>spatial_locs</code> slot. These are generated from the centroids
calculation for the polygons, and will be used for any generated
aggregate information.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/showGiottoSpatLocs.html">showGiottoSpatLocs</a></span><span class="op">(</span><span class="va">xenium_gobj</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># â”œâ”€â”€Spatial unit "cell"</span></span>
<span><span class="co"># â”‚  â””â”€â”€S4 spatLocsObj "raw" coordinates:   (167782 rows)</span></span>
<span><span class="co"># â”‚        An object of class spatLocsObj </span></span>
<span><span class="co"># â”‚        provenance: cell</span></span>
<span><span class="co"># â”‚            ------------------------</span></span>
<span><span class="co"># â”‚           cell_ID    sdimx    sdimy</span></span>
<span><span class="co"># â”‚        1:       1 377.6355 843.5235</span></span>
<span><span class="co"># â”‚        2:       2 382.0902 858.9148</span></span>
<span><span class="co"># â”‚        3:       3 319.8592 869.1546</span></span>
<span><span class="co"># â”‚        4:       4 259.2721 851.8312</span></span>
<span><span class="co"># â”‚        </span></span>
<span><span class="co"># â”‚        ranges:</span></span>
<span><span class="co"># â”‚                   sdimx       sdimy</span></span>
<span><span class="co"># â”‚        [1,]    2.189156    1.406448</span></span>
<span><span class="co"># â”‚        [2,] 7523.162860 5476.466538</span></span>
<span><span class="co"># â”‚        </span></span>
<span><span class="co"># â”‚        </span></span>
<span><span class="co"># â”‚     </span></span>
<span><span class="co"># â””â”€â”€Spatial unit "nucleus"</span></span>
<span><span class="co">#    â””â”€â”€S4 spatLocsObj "raw" coordinates:   (167782 rows)</span></span>
<span><span class="co">#          An object of class spatLocsObj </span></span>
<span><span class="co">#          provenance: nucleus</span></span>
<span><span class="co">#              ------------------------</span></span>
<span><span class="co">#             cell_ID    sdimx    sdimy</span></span>
<span><span class="co">#          1:       1 377.8125 842.8358</span></span>
<span><span class="co">#          2:       2 384.3298 858.9976</span></span>
<span><span class="co">#          3:       3 321.9175 869.2366</span></span>
<span><span class="co">#          4:       4 257.3259 851.5493</span></span>
<span><span class="co">#          </span></span>
<span><span class="co">#          ranges:</span></span>
<span><span class="co">#                     sdimx        sdimy</span></span>
<span><span class="co">#          [1,]    2.596845    0.8111559</span></span>
<span><span class="co">#          [2,] 7523.503184 5477.3741650</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="plot-the-generated-centroids-information">6.3 Plot the generated centroids information<a class="anchor" aria-label="anchor" href="#plot-the-generated-centroids-information"></a>
</h3>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://drieslab.github.io/GiottoVisuals/reference/spatPlot.html">spatPlot2D</a></span><span class="op">(</span><span class="va">xenium_gobj</span>,</span>
<span>           spat_unit <span class="op">=</span> <span class="st">'cell'</span>,</span>
<span>           point_shape <span class="op">=</span> <span class="st">'no_border'</span>,</span>
<span>           point_size <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>           point_alpha <span class="op">=</span> <span class="fl">0.4</span>,</span>
<span>           save_param <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>             base_width <span class="op">=</span> <span class="fl">7</span>,</span>
<span>             base_height <span class="op">=</span> <span class="fl">7</span>,</span>
<span>             save_name <span class="op">=</span> <span class="st">'1_spatplot'</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/xenium_breast_cancer/1_spatplot.png"></p>
</div>
</div>
<div class="section level2">
<h2 id="generate-aggregated-expression-based-on-feature-and-boundary-polygon-information">7. Generate aggregated expression based on feature and boundary
(polygon) information<a class="anchor" aria-label="anchor" href="#generate-aggregated-expression-based-on-feature-and-boundary-polygon-information"></a>
</h2>
<div class="section level3">
<h3 id="calculate-the-overlaps-of-the-rna-feature-data-within-the-cell-polygon-boundary-info-">7.1 Calculate the overlaps of the <code>'rna'</code> feature data
within the <code>'cell'</code> polygon boundary info.<a class="anchor" aria-label="anchor" href="#calculate-the-overlaps-of-the-rna-feature-data-within-the-cell-polygon-boundary-info-"></a>
</h3>
<p>This updates the <code>'cell'</code> <code>giottoPolygon</code>
overlaps slot with features that are overlapping the <code>'cell'</code>
polygons.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">xenium_gobj</span> <span class="op">=</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/calculateOverlapRaster.html">calculateOverlapRaster</a></span><span class="op">(</span><span class="va">xenium_gobj</span>,</span>
<span>                                     spatial_info <span class="op">=</span> <span class="st">'cell'</span>,</span>
<span>                                     feat_info <span class="op">=</span> <span class="st">'rna'</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/showGiottoSpatialInfo.html">showGiottoSpatialInfo</a></span><span class="op">(</span><span class="va">xenium_gobj</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># For Spatial info:  cell </span></span>
<span><span class="co"># </span></span>
<span><span class="co"># An object of class "giottoPolygon"</span></span>
<span><span class="co"># Slot "name":</span></span>
<span><span class="co"># [1] "cell"</span></span>
<span><span class="co"># </span></span>
<span><span class="co"># Slot "spatVector":</span></span>
<span><span class="co">#  class       : SpatVector </span></span>
<span><span class="co">#  geometry    : polygons </span></span>
<span><span class="co">#  dimensions  : 167782, 1  (geometries, attributes)</span></span>
<span><span class="co">#  extent      : 0, 7525.9, 0, 5478.038  (xmin, xmax, ymin, ymax)</span></span>
<span><span class="co">#  coord. ref. :  </span></span>
<span><span class="co">#  names       : poly_ID</span></span>
<span><span class="co">#  type        :   &lt;chr&gt;</span></span>
<span><span class="co">#  values      :       1</span></span>
<span><span class="co">#                      2</span></span>
<span><span class="co">#                      3</span></span>
<span><span class="co"># </span></span>
<span><span class="co"># Slot "spatVectorCentroids":</span></span>
<span><span class="co">#  class       : SpatVector </span></span>
<span><span class="co">#  geometry    : points </span></span>
<span><span class="co">#  dimensions  : 167782, 1  (geometries, attributes)</span></span>
<span><span class="co">#  extent      : 2.189156, 7523.163, 1.406448, 5476.467  (xmin, xmax, ymin, ymax)</span></span>
<span><span class="co">#  coord. ref. :  </span></span>
<span><span class="co">#  names       : poly_ID</span></span>
<span><span class="co">#  type        :   &lt;chr&gt;</span></span>
<span><span class="co">#  values      :       1</span></span>
<span><span class="co">#                      2</span></span>
<span><span class="co">#                      3</span></span>
<span><span class="co"># </span></span>
<span><span class="co"># Slot "overlaps":</span></span>
<span><span class="co"># $rna</span></span>
<span><span class="co">#  class       : SpatVector </span></span>
<span><span class="co">#  geometry    : points </span></span>
<span><span class="co">#  dimensions  : 43664530, 3  (geometries, attributes)</span></span>
<span><span class="co">#  extent      : -1.874261, 7522.837, 4.415276, 5473.721  (xmin, xmax, ymin, ymax)</span></span>
<span><span class="co">#  coord. ref. :  </span></span>
<span><span class="co">#  names       : poly_ID    feat_ID feat_ID_uniq</span></span>
<span><span class="co">#  type        :   &lt;chr&gt;      &lt;chr&gt;        &lt;int&gt;</span></span>
<span><span class="co">#  values      :   18790 BLANK_0180            1</span></span>
<span><span class="co">#                    370        LUM            2</span></span>
<span><span class="co">#                  18183     CLECL1            3</span></span>
<span><span class="co"># </span></span>
<span><span class="co"># </span></span>
<span><span class="co"># -----------------------------</span></span>
<span><span class="co">#  </span></span>
<span><span class="co"># For Spatial info:  nucleus </span></span>
<span><span class="co">#         ... </span></span>
<span><span class="co">#      truncated</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="assign-polygon-overlaps-information-to-expression-matrix">7.2 Assign polygon overlaps information to expression matrix<a class="anchor" aria-label="anchor" href="#assign-polygon-overlaps-information-to-expression-matrix"></a>
</h3>
<p>In order to create an aggregated expression matrix, the
<code>'rna'</code> features overlapped by the <code>'cell'</code>
polygon boundaries are sent to be combined into a cell/feature matrix
(named as <code>'raw'</code>) in the Giotto objectâ€™s
<code>expression</code> slot.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">xenium_gobj</span> <span class="op">=</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/overlapToMatrix.html">overlapToMatrix</a></span><span class="op">(</span><span class="va">xenium_gobj</span>,</span>
<span>                              poly_info <span class="op">=</span> <span class="st">'cell'</span>,</span>
<span>                              feat_info <span class="op">=</span> <span class="st">'rna'</span>,</span>
<span>                              name <span class="op">=</span> <span class="st">'raw'</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/showGiottoExpression.html">showGiottoExpression</a></span><span class="op">(</span><span class="va">xenium_gobj</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># â””â”€â”€Spatial unit "cell"</span></span>
<span><span class="co">#    â””â”€â”€Feature type "rna"</span></span>
<span><span class="co">#       â””â”€â”€Expression data "raw" values:</span></span>
<span><span class="co">#             An object of class exprObj </span></span>
<span><span class="co">#             for spatial unit: "cell" and feature type: "rna" </span></span>
<span><span class="co">#               Provenance:  cell </span></span>
<span><span class="co">#             </span></span>
<span><span class="co">#             contains:</span></span>
<span><span class="co">#             313 x 167782 sparse Matrix of class "dgCMatrix"</span></span>
<span><span class="co">#                                                    </span></span>
<span><span class="co">#             LUM   2 . 3 . 1 1  2 . . 2 . 5 . ......</span></span>
<span><span class="co">#             TCIM  1 1 . 4 1 1 13 . . . . . . ......</span></span>
<span><span class="co">#             RUNX1 . . . . . .  . . . . 1 . . ......</span></span>
<span><span class="co">#             </span></span>
<span><span class="co">#              ..............................</span></span>
<span><span class="co">#              ........suppressing 167769 columns and 307 rows </span></span>
<span><span class="co">#              ..............................</span></span>
<span><span class="co">#                                                    </span></span>
<span><span class="co">#             CD1C   1 . . . . . . . . . . . . ......</span></span>
<span><span class="co">#             CYP1A1 . . . . . . . . . . . . . ......</span></span>
<span><span class="co">#             CRHBP  . . . . . . . . . . . . . ......</span></span>
<span><span class="co">#             </span></span>
<span><span class="co">#              First four colnames:</span></span>
<span><span class="co">#              1 2 3 4 </span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="feature-metadata">7.3 Feature metadata<a class="anchor" aria-label="anchor" href="#feature-metadata"></a>
</h3>
<p>Append features metadata from <code>panel.tsv</code> which includes
information on what cell types the features are commonly markers for.
There are 313 rows in this file. One for each of the gene expression
probes, thus these metadata should be appended only to feat_type
â€˜rnaâ€™.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">panel_meta</span> <span class="op">=</span> <span class="fu">data.table</span><span class="fu">::</span><span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="va">panel_meta_path</span><span class="op">)</span></span>
<span><span class="fu">data.table</span><span class="fu">::</span><span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/setattr.html" class="external-link">setnames</a></span><span class="op">(</span><span class="va">panel_meta</span>, <span class="st">'Name'</span>, <span class="st">'feat_ID'</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Append this metadata</span></span>
<span><span class="va">xenium_gobj</span> <span class="op">=</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/addFeatMetadata.html">addFeatMetadata</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">xenium_gobj</span>,</span>
<span>                              feat_type <span class="op">=</span> <span class="st">'rna'</span>,</span>
<span>                              spat_unit <span class="op">=</span> <span class="st">'cell'</span>,</span>
<span>                              new_metadata <span class="op">=</span> <span class="va">panel_meta</span>,</span>
<span>                              by_column <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                              column_feat_ID <span class="op">=</span> <span class="st">'feat_ID'</span><span class="op">)</span></span>
<span><span class="va">xenium_gobj</span> <span class="op">=</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/addFeatMetadata.html">addFeatMetadata</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">xenium_gobj</span>,</span>
<span>                              feat_type <span class="op">=</span> <span class="st">'rna'</span>,</span>
<span>                              spat_unit <span class="op">=</span> <span class="st">'nucleus'</span>,</span>
<span>                              new_metadata <span class="op">=</span> <span class="va">panel_meta</span>,</span>
<span>                              by_column <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                              column_feat_ID <span class="op">=</span> <span class="st">'feat_ID'</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># to return a specific metadata as data.table </span></span>
<span><span class="co"># (spat_unit = 'cell', feat_type = 'rna' are default)</span></span>
<span><span class="co"># fDataDT(xenium_gobj)</span></span>
<span></span>
<span><span class="co"># Print a preview of all available features metadata</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/showGiottoFeatMetadata.html">showGiottoFeatMetadata</a></span><span class="op">(</span><span class="va">xenium_gobj</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># â”œâ”€â”€Spatial unit "cell"</span></span>
<span><span class="co"># â”‚  â”œâ”€â”€Feature type "rna"</span></span>
<span><span class="co"># â”‚  â”‚     An object of class featMetaObj </span></span>
<span><span class="co"># â”‚  â”‚     Provenance: cell </span></span>
<span><span class="co"># â”‚  â”‚        feat_ID      Ensembl ID             Annotation</span></span>
<span><span class="co"># â”‚  â”‚     1:     LUM ENSG00000139329            Fibroblasts</span></span>
<span><span class="co"># â”‚  â”‚     2:    TCIM ENSG00000176907 Breast glandular cells</span></span>
<span><span class="co"># â”‚  â”‚     3:   RUNX1 ENSG00000159216          Breast cancer</span></span>
<span><span class="co"># â”‚  â”‚     4: RAPGEF3 ENSG00000079337             Adipocytes</span></span>
<span><span class="co"># â”‚  â”‚  </span></span>
<span><span class="co"># â”‚  â”œâ”€â”€Feature type "blank_code"</span></span>
<span><span class="co"># â”‚  â”‚     An object of class featMetaObj </span></span>
<span><span class="co"># â”‚  â”‚     Provenance: cell </span></span>
<span><span class="co"># â”‚  â”‚           feat_ID</span></span>
<span><span class="co"># â”‚  â”‚     1: BLANK_0424</span></span>
<span><span class="co"># â”‚  â”‚     2: BLANK_0401</span></span>
<span><span class="co"># â”‚  â”‚     3: BLANK_0447</span></span>
<span><span class="co"># â”‚  â”‚     4: BLANK_0449</span></span>
<span><span class="co"># â”‚  â”‚  </span></span>
<span><span class="co"># â”‚  â”œâ”€â”€Feature type "neg_code"</span></span>
<span><span class="co"># â”‚  â”‚     An object of class featMetaObj </span></span>
<span><span class="co"># â”‚  â”‚     Provenance: cell </span></span>
<span><span class="co"># â”‚  â”‚                        feat_ID</span></span>
<span><span class="co"># â”‚  â”‚     1: NegControlCodeword_0503</span></span>
<span><span class="co"># â”‚  â”‚     2: NegControlCodeword_0514</span></span>
<span><span class="co"># â”‚  â”‚     3: NegControlCodeword_0535</span></span>
<span><span class="co"># â”‚  â”‚     4: NegControlCodeword_0519</span></span>
<span><span class="co"># â”‚  â”‚  </span></span>
<span><span class="co"># â”‚  â””â”€â”€Feature type "neg_probe"</span></span>
<span><span class="co"># â”‚        An object of class featMetaObj </span></span>
<span><span class="co"># â”‚        Provenance: cell </span></span>
<span><span class="co"># â”‚                         feat_ID</span></span>
<span><span class="co"># â”‚        1: NegControlProbe_00003</span></span>
<span><span class="co"># â”‚        2:       antisense_SCRIB</span></span>
<span><span class="co"># â”‚        3: NegControlProbe_00012</span></span>
<span><span class="co"># â”‚        4:        antisense_LGI3</span></span>
<span><span class="co"># â”‚     </span></span>
<span><span class="co"># â””â”€â”€Spatial unit "nucleus"</span></span>
<span><span class="co">#    â”œâ”€â”€Feature type "rna"</span></span>
<span><span class="co">#    â”‚     An object of class featMetaObj </span></span>
<span><span class="co">#    â”‚     Provenance: nucleus </span></span>
<span><span class="co">#    â”‚        feat_ID      Ensembl ID             Annotation</span></span>
<span><span class="co">#    â”‚     1:     LUM ENSG00000139329            Fibroblasts</span></span>
<span><span class="co">#    â”‚     2:    TCIM ENSG00000176907 Breast glandular cells</span></span>
<span><span class="co">#    â”‚     3:   RUNX1 ENSG00000159216          Breast cancer</span></span>
<span><span class="co">#    â”‚     4: RAPGEF3 ENSG00000079337             Adipocytes</span></span>
<span><span class="co">#    â”‚  </span></span>
<span><span class="co">#    â”œâ”€â”€Feature type "blank_code"</span></span>
<span><span class="co">#    â”‚     An object of class featMetaObj </span></span>
<span><span class="co">#    â”‚     Provenance: nucleus </span></span>
<span><span class="co">#    â”‚           feat_ID</span></span>
<span><span class="co">#    â”‚     1: BLANK_0424</span></span>
<span><span class="co">#    â”‚     2: BLANK_0401</span></span>
<span><span class="co">#    â”‚     3: BLANK_0447</span></span>
<span><span class="co">#    â”‚     4: BLANK_0449</span></span>
<span><span class="co">#    â”‚  </span></span>
<span><span class="co">#    â”œâ”€â”€Feature type "neg_code"</span></span>
<span><span class="co">#    â”‚     An object of class featMetaObj </span></span>
<span><span class="co">#    â”‚     Provenance: nucleus </span></span>
<span><span class="co">#    â”‚                        feat_ID</span></span>
<span><span class="co">#    â”‚     1: NegControlCodeword_0503</span></span>
<span><span class="co">#    â”‚     2: NegControlCodeword_0514</span></span>
<span><span class="co">#    â”‚     3: NegControlCodeword_0535</span></span>
<span><span class="co">#    â”‚     4: NegControlCodeword_0519</span></span>
<span><span class="co">#    â”‚  </span></span>
<span><span class="co">#    â””â”€â”€Feature type "neg_probe"</span></span>
<span><span class="co">#          An object of class featMetaObj </span></span>
<span><span class="co">#          Provenance: nucleus </span></span>
<span><span class="co">#                           feat_ID</span></span>
<span><span class="co">#          1: NegControlProbe_00003</span></span>
<span><span class="co">#          2:       antisense_SCRIB</span></span>
<span><span class="co">#          3: NegControlProbe_00012</span></span>
<span><span class="co">#          4:        antisense_LGI3</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="data-filtering">7.4 Data filtering<a class="anchor" aria-label="anchor" href="#data-filtering"></a>
</h3>
<p>Now that an aggregated expression matrix is generated the usual data
filtering and processing can be applied We start by setting a count of 1
to be the minimum to consider a feature expressed. A feature must be
detected in at least 3 cells to be included. Lastly, a cell must have a
minimum of 5 features detected to be included.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">xenium_gobj</span> <span class="op">=</span> <span class="fu"><a href="../reference/filterGiotto.html">filterGiotto</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">xenium_gobj</span>,</span>
<span>                           spat_unit <span class="op">=</span> <span class="st">'cell'</span>,</span>
<span>                           poly_info <span class="op">=</span> <span class="st">'cell'</span>,</span>
<span>                           expression_threshold <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                           feat_det_in_min_cells <span class="op">=</span> <span class="fl">3</span>,</span>
<span>                           min_det_feats_per_cell <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#    truncated</span></span>
<span><span class="co">#      ...</span></span>
<span><span class="co"># Feature type:  rna </span></span>
<span><span class="co"># Number of cells removed:  2945  out of  167782 </span></span>
<span><span class="co"># Number of feats removed:  0  out of  313 </span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="add-data-statistics">7.5 Add data statistics<a class="anchor" aria-label="anchor" href="#add-data-statistics"></a>
</h3>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">xenium_gobj</span> <span class="op">=</span> <span class="fu"><a href="../reference/addStatistics.html">addStatistics</a></span><span class="op">(</span><span class="va">xenium_gobj</span>, expression_values <span class="op">=</span> <span class="st">'raw'</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/showGiottoCellMetadata.html">showGiottoCellMetadata</a></span><span class="op">(</span><span class="va">xenium_gobj</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/showGiottoFeatMetadata.html">showGiottoFeatMetadata</a></span><span class="op">(</span><span class="va">xenium_gobj</span><span class="op">)</span></span></code></pre></div>
<details><summary>cell metadata
</summary><div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># â”œâ”€â”€Spatial unit "cell"</span></span>
<span><span class="co"># â”‚  â”œâ”€â”€Feature type "rna"</span></span>
<span><span class="co"># â”‚  â”‚     An object of class cellMetaObj </span></span>
<span><span class="co"># â”‚  â”‚     Provenance: cell </span></span>
<span><span class="co"># â”‚  â”‚        cell_ID nr_feats perc_feats total_expr</span></span>
<span><span class="co"># â”‚  â”‚     1:       1       62   19.80831        156</span></span>
<span><span class="co"># â”‚  â”‚     2:       2       41   13.09904         63</span></span>
<span><span class="co"># â”‚  â”‚     3:       3       38   12.14058         54</span></span>
<span><span class="co"># â”‚  â”‚     4:       4       47   15.01597        114</span></span>
<span><span class="co"># â”‚  â”‚  </span></span>
<span><span class="co"># â”‚  â”œâ”€â”€Feature type "blank_code"</span></span>
<span><span class="co"># â”‚  â”‚     An object of class cellMetaObj </span></span>
<span><span class="co"># â”‚  â”‚     Provenance: cell </span></span>
<span><span class="co"># â”‚  â”‚        cell_ID</span></span>
<span><span class="co"># â”‚  â”‚     1:       1</span></span>
<span><span class="co"># â”‚  â”‚     2:       2</span></span>
<span><span class="co"># â”‚  â”‚     3:       3</span></span>
<span><span class="co"># â”‚  â”‚     4:       4</span></span>
<span><span class="co"># â”‚  â”‚  </span></span>
<span><span class="co"># â”‚  â”œâ”€â”€Feature type "neg_code"</span></span>
<span><span class="co"># â”‚  â”‚     An object of class cellMetaObj </span></span>
<span><span class="co"># â”‚  â”‚     Provenance: cell </span></span>
<span><span class="co"># â”‚  â”‚        cell_ID</span></span>
<span><span class="co"># â”‚  â”‚     1:       1</span></span>
<span><span class="co"># â”‚  â”‚     2:       2</span></span>
<span><span class="co"># â”‚  â”‚     3:       3</span></span>
<span><span class="co"># â”‚  â”‚     4:       4</span></span>
<span><span class="co"># â”‚  â”‚  </span></span>
<span><span class="co"># â”‚  â””â”€â”€Feature type "neg_probe"</span></span>
<span><span class="co"># â”‚        An object of class cellMetaObj </span></span>
<span><span class="co"># â”‚        Provenance: cell </span></span>
<span><span class="co"># â”‚           cell_ID</span></span>
<span><span class="co"># â”‚        1:       1</span></span>
<span><span class="co"># â”‚        2:       2</span></span>
<span><span class="co"># â”‚        3:       3</span></span>
<span><span class="co"># â”‚        4:       4</span></span>
<span><span class="co"># â”‚     </span></span>
<span><span class="co"># â””â”€â”€Spatial unit "nucleus"</span></span>
<span><span class="co">#    â”œâ”€â”€Feature type "rna"</span></span>
<span><span class="co">#    â”‚     An object of class cellMetaObj </span></span>
<span><span class="co">#    â”‚     Provenance: nucleus </span></span>
<span><span class="co">#    â”‚        cell_ID</span></span>
<span><span class="co">#    â”‚     1:       1</span></span>
<span><span class="co">#    â”‚     2:       2</span></span>
<span><span class="co">#    â”‚     3:       3</span></span>
<span><span class="co">#    â”‚     4:       4</span></span>
<span><span class="co">#    â”‚  </span></span>
<span><span class="co">#    â”œâ”€â”€Feature type "blank_code"</span></span>
<span><span class="co">#    â”‚     An object of class cellMetaObj </span></span>
<span><span class="co">#    â”‚     Provenance: nucleus </span></span>
<span><span class="co">#    â”‚        cell_ID</span></span>
<span><span class="co">#    â”‚     1:       1</span></span>
<span><span class="co">#    â”‚     2:       2</span></span>
<span><span class="co">#    â”‚     3:       3</span></span>
<span><span class="co">#    â”‚     4:       4</span></span>
<span><span class="co">#    â”‚  </span></span>
<span><span class="co">#    â”œâ”€â”€Feature type "neg_code"</span></span>
<span><span class="co">#    â”‚     An object of class cellMetaObj </span></span>
<span><span class="co">#    â”‚     Provenance: nucleus </span></span>
<span><span class="co">#    â”‚        cell_ID</span></span>
<span><span class="co">#    â”‚     1:       1</span></span>
<span><span class="co">#    â”‚     2:       2</span></span>
<span><span class="co">#    â”‚     3:       3</span></span>
<span><span class="co">#    â”‚     4:       4</span></span>
<span><span class="co">#    â”‚  </span></span>
<span><span class="co">#    â””â”€â”€Feature type "neg_probe"</span></span>
<span><span class="co">#          An object of class cellMetaObj </span></span>
<span><span class="co">#          Provenance: nucleus </span></span>
<span><span class="co">#             cell_ID</span></span>
<span><span class="co">#          1:       1</span></span>
<span><span class="co">#          2:       2</span></span>
<span><span class="co">#          3:       3</span></span>
<span><span class="co">#          4:       4</span></span></code></pre></div>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># â”œâ”€â”€Spatial unit "cell"</span></span>
<span><span class="co"># â”‚  â”œâ”€â”€Feature type "rna"</span></span>
<span><span class="co"># â”‚  â”‚     An object of class featMetaObj </span></span>
<span><span class="co"># â”‚  â”‚     Provenance: cell </span></span>
<span><span class="co"># â”‚  â”‚        feat_ID      Ensembl ID             Annotation nr_cells perc_cells</span></span>
<span><span class="co"># â”‚  â”‚     1:     LUM ENSG00000139329            Fibroblasts   101666  61.676687</span></span>
<span><span class="co"># â”‚  â”‚     2:    TCIM ENSG00000176907 Breast glandular cells    84842  51.470240</span></span>
<span><span class="co"># â”‚  â”‚     3:   RUNX1 ENSG00000159216          Breast cancer    94086  57.078205</span></span>
<span><span class="co"># â”‚  â”‚     4: RAPGEF3 ENSG00000079337             Adipocytes    13286   8.060084</span></span>
<span><span class="co"># â”‚  â”‚        total_expr mean_expr mean_expr_det</span></span>
<span><span class="co"># â”‚  â”‚     1:     946217 5.7403192      9.307113</span></span>
<span><span class="co"># â”‚  â”‚     2:     300377 1.8222668      3.540428</span></span>
<span><span class="co"># â”‚  â”‚     3:     229633 1.3930914      2.440671</span></span>
<span><span class="co"># â”‚  â”‚     4:      16645 0.1009785      1.252823</span></span>
<span><span class="co"># â”‚  â”‚  </span></span>
<span><span class="co"># â”‚  â”œâ”€â”€Feature type "blank_code"</span></span>
<span><span class="co"># â”‚  â”‚     An object of class featMetaObj </span></span>
<span><span class="co"># â”‚  â”‚     Provenance: cell </span></span>
<span><span class="co"># â”‚  â”‚           feat_ID</span></span>
<span><span class="co"># â”‚  â”‚     1: BLANK_0424</span></span>
<span><span class="co"># â”‚  â”‚     2: BLANK_0401</span></span>
<span><span class="co"># â”‚  â”‚     3: BLANK_0447</span></span>
<span><span class="co"># â”‚  â”‚     4: BLANK_0449</span></span>
<span><span class="co"># â”‚  â”‚  </span></span>
<span><span class="co"># â”‚  â”œâ”€â”€Feature type "neg_code"</span></span>
<span><span class="co"># â”‚  â”‚     An object of class featMetaObj </span></span>
<span><span class="co"># â”‚  â”‚     Provenance: cell </span></span>
<span><span class="co"># â”‚  â”‚                        feat_ID</span></span>
<span><span class="co"># â”‚  â”‚     1: NegControlCodeword_0503</span></span>
<span><span class="co"># â”‚  â”‚     2: NegControlCodeword_0514</span></span>
<span><span class="co"># â”‚  â”‚     3: NegControlCodeword_0535</span></span>
<span><span class="co"># â”‚  â”‚     4: NegControlCodeword_0519</span></span>
<span><span class="co"># â”‚  â”‚  </span></span>
<span><span class="co"># â”‚  â””â”€â”€Feature type "neg_probe"</span></span>
<span><span class="co"># â”‚        An object of class featMetaObj </span></span>
<span><span class="co"># â”‚        Provenance: cell </span></span>
<span><span class="co"># â”‚                         feat_ID</span></span>
<span><span class="co"># â”‚        1: NegControlProbe_00003</span></span>
<span><span class="co"># â”‚        2:       antisense_SCRIB</span></span>
<span><span class="co"># â”‚        3: NegControlProbe_00012</span></span>
<span><span class="co"># â”‚        4:        antisense_LGI3</span></span>
<span><span class="co"># â”‚     </span></span>
<span><span class="co"># â””â”€â”€Spatial unit "nucleus"</span></span>
<span><span class="co">#    â”œâ”€â”€Feature type "rna"</span></span>
<span><span class="co">#    â”‚     An object of class featMetaObj </span></span>
<span><span class="co">#    â”‚     Provenance: nucleus </span></span>
<span><span class="co">#    â”‚        feat_ID      Ensembl ID             Annotation</span></span>
<span><span class="co">#    â”‚     1:     LUM ENSG00000139329            Fibroblasts</span></span>
<span><span class="co">#    â”‚     2:    TCIM ENSG00000176907 Breast glandular cells</span></span>
<span><span class="co">#    â”‚     3:   RUNX1 ENSG00000159216          Breast cancer</span></span>
<span><span class="co">#    â”‚     4: RAPGEF3 ENSG00000079337             Adipocytes</span></span>
<span><span class="co">#    â”‚  </span></span>
<span><span class="co">#    â”œâ”€â”€Feature type "blank_code"</span></span>
<span><span class="co">#    â”‚     An object of class featMetaObj </span></span>
<span><span class="co">#    â”‚     Provenance: nucleus </span></span>
<span><span class="co">#    â”‚           feat_ID</span></span>
<span><span class="co">#    â”‚     1: BLANK_0424</span></span>
<span><span class="co">#    â”‚     2: BLANK_0401</span></span>
<span><span class="co">#    â”‚     3: BLANK_0447</span></span>
<span><span class="co">#    â”‚     4: BLANK_0449</span></span>
<span><span class="co">#    â”‚  </span></span>
<span><span class="co">#    â”œâ”€â”€Feature type "neg_code"</span></span>
<span><span class="co">#    â”‚     An object of class featMetaObj </span></span>
<span><span class="co">#    â”‚     Provenance: nucleus </span></span>
<span><span class="co">#    â”‚                        feat_ID</span></span>
<span><span class="co">#    â”‚     1: NegControlCodeword_0503</span></span>
<span><span class="co">#    â”‚     2: NegControlCodeword_0514</span></span>
<span><span class="co">#    â”‚     3: NegControlCodeword_0535</span></span>
<span><span class="co">#    â”‚     4: NegControlCodeword_0519</span></span>
<span><span class="co">#    â”‚  </span></span>
<span><span class="co">#    â””â”€â”€Feature type "neg_probe"</span></span>
<span><span class="co">#          An object of class featMetaObj </span></span>
<span><span class="co">#          Provenance: nucleus </span></span>
<span><span class="co">#                           feat_ID</span></span>
<span><span class="co">#          1: NegControlProbe_00003</span></span>
<span><span class="co">#          2:       antisense_SCRIB</span></span>
<span><span class="co">#          3: NegControlProbe_00012</span></span>
<span><span class="co">#          4:        antisense_LGI3</span></span></code></pre></div>
</details>
</div>
<div class="section level3">
<h3 id="normalize-expression">7.6 Normalize expression<a class="anchor" aria-label="anchor" href="#normalize-expression"></a>
</h3>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">xenium_gobj</span> <span class="op">=</span> <span class="fu"><a href="../reference/normalizeGiotto.html">normalizeGiotto</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">xenium_gobj</span>,</span>
<span>                              spat_unit <span class="op">=</span> <span class="st">'cell'</span>,</span>
<span>                              scalefactor <span class="op">=</span> <span class="fl">5000</span>,</span>
<span>                              verbose <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="calculate-highly-variable-features">7.7 Calculate highly variable features<a class="anchor" aria-label="anchor" href="#calculate-highly-variable-features"></a>
</h3>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">xenium_gobj</span> <span class="op">=</span> <span class="fu"><a href="../reference/calculateHVF.html">calculateHVF</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">xenium_gobj</span>,</span>
<span>                           spat_unit <span class="op">=</span> <span class="st">'cell'</span>,</span>
<span>                           save_param <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>                             save_name <span class="op">=</span> <span class="st">'2_HVF'</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/fDataDT.html">fDataDT</a></span><span class="op">(</span><span class="va">xenium_gobj</span><span class="op">)</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">hvf</span> <span class="op">==</span> <span class="st">'yes'</span><span class="op">)</span><span class="op">]</span>, <span class="st">'hvf found'</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># 22 hvf found</span></span></code></pre></div>
<p><img src="images/xenium_breast_cancer/2_HVF.png"></p>
<p>Basing clustering on only 22 highly variable genes will not be
helpful. Will be using all 313 genes for clustering.</p>
</div>
</div>
<div class="section level2">
<h2 id="dimension-reduction-and-clustering">8. Dimension reduction and clustering<a class="anchor" aria-label="anchor" href="#dimension-reduction-and-clustering"></a>
</h2>
<div class="section level3">
<h3 id="pca">8.1 PCA<a class="anchor" aria-label="anchor" href="#pca"></a>
</h3>
<p>By default, <code><a href="../reference/runPCA.html">runPCA()</a></code> uses the subset of genes discovered
to be highly variable and then assigned as such in the feature metadata.
Instead, this time, using all genes is desireable, so
<code>feats_to_use</code> will be set to <code>NULL</code>.</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">xenium_gobj</span> <span class="op">=</span> <span class="fu"><a href="../reference/runPCA.html">runPCA</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">xenium_gobj</span>,</span>
<span>                     spat_unit <span class="op">=</span> <span class="st">'cell'</span>,</span>
<span>                     expression_values <span class="op">=</span> <span class="st">'scaled'</span>,</span>
<span>                     feats_to_use <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>                     scale_unit <span class="op">=</span> <span class="cn">F</span>,</span>
<span>                     center <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Visualize Screeplot and PCA</span></span>
<span><span class="fu"><a href="../reference/screePlot.html">screePlot</a></span><span class="op">(</span><span class="va">xenium_gobj</span>,</span>
<span>          ncp <span class="op">=</span> <span class="fl">20</span>,</span>
<span>          save_param <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>            save_name <span class="op">=</span> <span class="st">'3a_screePlot'</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/showGiottoDimRed.html">showGiottoDimRed</a></span><span class="op">(</span><span class="va">xenium_gobj</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/GiottoVisuals/reference/plotPCA.html">plotPCA</a></span><span class="op">(</span><span class="va">xenium_gobj</span>,</span>
<span>        spat_unit <span class="op">=</span> <span class="st">'cell'</span>,</span>
<span>        dim_reduction_name <span class="op">=</span> <span class="st">'pca'</span>,</span>
<span>        dim1_to_use <span class="op">=</span> <span class="fl">1</span>,</span>
<span>        dim2_to_use <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Dim reduction on cells:</span></span>
<span><span class="co">#  ------------------------- </span></span>
<span><span class="co"># </span></span>
<span><span class="co"># .</span></span>
<span><span class="co"># â””â”€â”€Spatial unit "cell"</span></span>
<span><span class="co">#    â””â”€â”€Feature type "rna"</span></span>
<span><span class="co">#       â””â”€â”€Dim reduction type "pca"</span></span>
<span><span class="co">#          â””â”€â”€S4 dimObj "pca" coordinates:   (165019 rows 21 cols)</span></span>
<span><span class="co">#                      Dim.1     Dim.2</span></span>
<span><span class="co">#                1 -2.459551 -0.418277</span></span>
<span><span class="co">#                2 -2.002147 -3.239184</span></span>
<span><span class="co">#                3 -1.588884 -2.195826</span></span></code></pre></div>
<p><img src="images/xenium_breast_cancer/3a_screePlot.png"><img src="images/xenium_breast_cancer/3b_PCA.png"></p>
</div>
<div class="section level3">
<h3 id="tsne-and-umap">8.2 tSNE and UMAP<a class="anchor" aria-label="anchor" href="#tsne-and-umap"></a>
</h3>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">xenium_gobj</span> <span class="op">=</span> <span class="fu"><a href="../reference/runtSNE.html">runtSNE</a></span><span class="op">(</span><span class="va">xenium_gobj</span>,</span>
<span>                      dimensions_to_use <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span>,</span>
<span>                      spat_unit <span class="op">=</span> <span class="st">'cell'</span><span class="op">)</span></span>
<span><span class="va">xenium_gobj</span> <span class="op">=</span> <span class="fu"><a href="../reference/runUMAP.html">runUMAP</a></span><span class="op">(</span><span class="va">xenium_gobj</span>,</span>
<span>                      dimensions_to_use <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span>,</span>
<span>                      spat_unit <span class="op">=</span> <span class="st">'cell'</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://drieslab.github.io/GiottoVisuals/reference/plotTSNE.html">plotTSNE</a></span><span class="op">(</span><span class="va">xenium_gobj</span>,</span>
<span>         point_size <span class="op">=</span> <span class="fl">0.01</span>,</span>
<span>         save_param <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>           save_name <span class="op">=</span> <span class="st">'4a_tSNE'</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/GiottoVisuals/reference/plotUMAP.html">plotUMAP</a></span><span class="op">(</span><span class="va">xenium_gobj</span>,</span>
<span>         point_size <span class="op">=</span> <span class="fl">0.01</span>,</span>
<span>         save_param <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>           save_name <span class="op">=</span> <span class="st">'4b_UMAP'</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/xenium_breast_cancer/4a_tSNE.png"><img src="images/xenium_breast_cancer/4b_UMAP.png"></p>
</div>
<div class="section level3">
<h3 id="snn-and-leiden-clustering">8.3 sNN and Leiden clustering<a class="anchor" aria-label="anchor" href="#snn-and-leiden-clustering"></a>
</h3>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">xenium_gobj</span> <span class="op">=</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/createNearestNetwork.html">createNearestNetwork</a></span><span class="op">(</span><span class="va">xenium_gobj</span>,</span>
<span>                                   dimensions_to_use <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span>,</span>
<span>                                   k <span class="op">=</span> <span class="fl">10</span>,</span>
<span>                                   spat_unit <span class="op">=</span> <span class="st">'cell'</span><span class="op">)</span></span>
<span><span class="va">xenium_gobj</span> <span class="op">=</span> <span class="fu"><a href="../reference/doLeidenCluster.html">doLeidenCluster</a></span><span class="op">(</span><span class="va">xenium_gobj</span>,</span>
<span>                              resolution <span class="op">=</span> <span class="fl">0.25</span>,</span>
<span>                              n_iterations <span class="op">=</span> <span class="fl">100</span>,</span>
<span>                              spat_unit <span class="op">=</span> <span class="st">'cell'</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># visualize UMAP cluster results</span></span>
<span><span class="fu"><a href="https://drieslab.github.io/GiottoVisuals/reference/plotUMAP.html">plotUMAP</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">xenium_gobj</span>,</span>
<span>         spat_unit <span class="op">=</span> <span class="st">'cell'</span>,</span>
<span>         cell_color <span class="op">=</span> <span class="st">'leiden_clus'</span>,</span>
<span>         show_legend <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>         point_size <span class="op">=</span> <span class="fl">0.01</span>,</span>
<span>         point_shape <span class="op">=</span> <span class="st">'no_border'</span>,</span>
<span>         save_param <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>save_name <span class="op">=</span> <span class="st">'5_umap_leiden'</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/xenium_breast_cancer/5_umap_leiden.png"></p>
</div>
<div class="section level3">
<h3 id="visualize-umap-and-spatial-results">8.4 Visualize UMAP and spatial results<a class="anchor" aria-label="anchor" href="#visualize-umap-and-spatial-results"></a>
</h3>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://drieslab.github.io/GiottoVisuals/reference/spatPlot.html">spatPlot2D</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">xenium_gobj</span>,</span>
<span>           spat_unit <span class="op">=</span> <span class="st">'cell'</span>,</span>
<span>           cell_color <span class="op">=</span> <span class="st">'leiden_clus'</span>,</span>
<span>           point_size <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>           point_shape <span class="op">=</span> <span class="st">'no_border'</span>,</span>
<span>           background_color <span class="op">=</span> <span class="st">'black'</span>,</span>
<span>           show_legend <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>           save_param <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>             save_name <span class="op">=</span> <span class="st">'6_spat_leiden'</span>,</span>
<span>             base_width <span class="op">=</span> <span class="fl">15</span>,</span>
<span>             base_height <span class="op">=</span> <span class="fl">15</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/xenium_breast_cancer/6_spat_leiden.png"></p>
</div>
</div>
<div class="section level2">
<h2 id="subcellular-visualization">9. Subcellular visualization<a class="anchor" aria-label="anchor" href="#subcellular-visualization"></a>
</h2>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://drieslab.github.io/GiottoVisuals/reference/spatInSituPlotPoints.html">spatInSituPlotPoints</a></span><span class="op">(</span><span class="va">xenium_gobj</span>,</span>
<span>                     show_image <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                     feats <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>                     point_size <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>                     show_polygon <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                     polygon_feat_type <span class="op">=</span> <span class="st">'cell'</span>,</span>
<span>                     polygon_alpha <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                     polygon_color <span class="op">=</span> <span class="st">'black'</span>,</span>
<span>                     polygon_line_size <span class="op">=</span> <span class="fl">0.01</span>,</span>
<span>                     polygon_fill <span class="op">=</span> <span class="st">'leiden_clus'</span>,</span>
<span>                     polygon_fill_as_factor <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                     coord_fix_ratio <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                     save_para <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>                       save_name <span class="op">=</span> <span class="st">'7_polys'</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/xenium_breast_cancer/7_polys.png"></p>
<p>The dataset is too large to visualize with subcellular polygons and
features clearly so a spatial subset is needed.</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">xenium_gobj_subset</span> <span class="op">=</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/subsetGiottoLocs.html">subsetGiottoLocs</a></span><span class="op">(</span><span class="va">xenium_gobj</span>,</span>
<span>                                      x_max <span class="op">=</span> <span class="fl">4500</span>,</span>
<span>                                      x_min <span class="op">=</span> <span class="fl">3500</span>,</span>
<span>                                      y_max <span class="op">=</span> <span class="fl">5500</span>,</span>
<span>                                      y_min <span class="op">=</span> <span class="fl">4500</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://drieslab.github.io/GiottoVisuals/reference/spatInSituPlotPoints.html">spatInSituPlotPoints</a></span><span class="op">(</span><span class="va">xenium_gobj_subset</span>,</span>
<span>                     show_image <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                     feats <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">'rna'</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>                       <span class="st">"LUM"</span>, <span class="st">"CXCR4"</span>, <span class="st">"ITGAX"</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                     feats_color_code <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>                       <span class="st">"LUM"</span> <span class="op">=</span> <span class="st">'green'</span>,</span>
<span>                       <span class="st">'CXCR4'</span> <span class="op">=</span> <span class="st">'blue'</span>,</span>
<span>                       <span class="st">'ITGAX'</span> <span class="op">=</span> <span class="st">'red'</span><span class="op">)</span>,</span>
<span>                     point_size <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>                     show_polygon <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                     polygon_feat_type <span class="op">=</span> <span class="st">'cell'</span>,</span>
<span>                     polygon_color <span class="op">=</span> <span class="st">'black'</span>,</span>
<span>                     polygon_line_size <span class="op">=</span> <span class="fl">0.01</span>,</span>
<span>                     polygon_fill <span class="op">=</span> <span class="st">'leiden_clus'</span>,</span>
<span>                     polygon_fill_as_factor <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                     coord_fix_ratio <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                     save_param <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>                       save_name <span class="op">=</span> <span class="st">'8_subset_in_situ'</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/xenium_breast_cancer/8_subset_in_situ.png"></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by <a href="https://www.drieslab.com" class="external-link">Ruben Dries</a>, Jiaji George Chen, <a href="https://josschavezf.github.io" class="external-link">Joselyn C. ChÃ¡vez-Fuentes</a>, <a href="https://labs.icahn.mssm.edu/yuanlab/" class="external-link">Guo-Cheng Yuan</a>, Matthew Oâ€™Brien, Edward Ruiz, Wen Wang, Natalie Del Rossi.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
